diff -ruN gcc-3.4.0/gcc/config/i386/i386.c gcc-3.4.0-new/gcc/config/i386/i386.c
--- gcc-3.4.0/gcc/config/i386/i386.c	2004-06-03 10:16:53.357355576 +0200
+++ gcc-3.4.0-new/gcc/config/i386/i386.c	2004-06-03 10:17:57.661579848 +0200
@@ -477,7 +477,7 @@
 const int x86_double_with_add = ~m_386;
 const int x86_use_bit_test = m_386;
 const int x86_unroll_strlen = m_486 | m_PENT | m_PPRO | m_ATHLON_K8 | m_K6;
-const int x86_cmove = m_PPRO | m_ATHLON_K8 | m_PENT4;
+const int x86_cmove = m_ATHLON_K8 | m_PENT4;
 const int x86_3dnow_a = m_ATHLON_K8;
 const int x86_deep_branch = m_PPRO | m_K6 | m_ATHLON_K8 | m_PENT4;
 const int x86_branch_hints = m_PENT4;
