--- ./lib/getline.c~	Sun Apr 13 19:51:47 2003
+++ ./lib/getline.c	Sun Apr 13 19:52:09 2003
@@ -44,7 +44,7 @@
 #else /* ! have getdelim */
 
 # include "getstr.h"
-
+#if 0
 int
 getline (char **lineptr, size_t *n, FILE *stream)
 {
@@ -56,4 +56,5 @@
 {
   return getstr (lineptr, n, stream, delimiter, 0, 0);
 }
+#endif
 #endif
--- ./src/dircolors.c~	Sun Dec  1 11:18:23 2002
+++ ./src/dircolors.c	Sat Apr 19 06:15:51 2003
@@ -28,7 +28,7 @@
 #include "dircolors.h"
 #include "dirname.h"
 #include "error.h"
-#include "getline.h"
+/*#include "getline.h"*/
 #include "obstack.h"
 #include "quote.h"
 
--- ./src/md5sum.c~	Sat Sep 14 09:37:49 2002
+++ ./src/md5sum.c	Sat Apr 19 09:25:00 2003
@@ -28,7 +28,7 @@
 #include "md5.h"
 #include "sha.h"
 #include "checksum.h"
-#include "getline.h"
+/*#include "getline.h"*/
 #include "closeout.h"
 #include "error.h"
 
--- ./src/date.c~	Sun Dec 15 21:47:03 2002
+++ ./src/date.c	Sat Apr 19 17:04:59 2003
@@ -30,7 +30,7 @@
 #include "closeout.h"
 #include "error.h"
 #include "getdate.h"
-#include "getline.h"
+/*#include "getline.h"*/
 #include "posixtm.h"
 #include "posixver.h"
 #include "strftime.h"
