--- gawk-3.1.5/io.c~	2005-07-26 20:07:43.000000000 +0200
+++ gawk-3.1.5/io.c	2005-09-02 13:08:50.784242397 +0200
@@ -41,8 +41,10 @@
 #include <termios.h>
 #endif
 #ifdef HAVE_STROPTS_H
+#ifndef __dietlibc__
 #include <stropts.h>
 #endif
+#endif
 
 #ifdef HAVE_SOCKETS
 #ifdef HAVE_SYS_SOCKET_H
--- gawk-3.1.5/regex_internal.h~	2005-05-11 17:28:17.000000000 +0200
+++ gawk-3.1.5/regex_internal.h	2005-09-02 13:14:44.954816214 +0200
@@ -37,6 +37,8 @@
 # include <wchar.h>
 #endif /* HAVE_WCHAR_H || _LIBC */
 #if defined HAVE_WCTYPE_H || defined _LIBC
+#define __BEGIN_NAMESPACE_C99
+#define __END_NAMESPACE_C99
 # include <wctype.h>
 #endif /* HAVE_WCTYPE_H || _LIBC */
 
--- gawk-3.1.5/configure.ac~	2005-07-26 20:43:51.000000000 +0200
+++ gawk-3.1.5/configure.ac	2005-09-02 13:23:40.040012522 +0200
@@ -264,20 +264,20 @@
 dnl check for dynamic linking
 dnl This is known to be very primitive
 AC_CHECK_HEADER(dlfcn.h,
-	[AC_DEFINE([DYNAMIC], 1, [dynamic loading is possible])
-	if test "$GCC" = yes
-	then
-		# Add others here as appropriate,
-		# one day use GNU libtool.
-		if uname | $EGREP -i 'linux|freebsd|cygwin' > /dev/null
-		then
-			LDFLAGS="$LDFLAGS -export-dynamic"
-		fi
-	fi
-
 	# Check this separately. Some systems have dlopen
 	# in libc. Notably freebsd and cygwin.
-	AC_CHECK_LIB(dl, dlopen)
+	AC_CHECK_LIB(dl, dlopen, [
+		[AC_DEFINE([DYNAMIC], 1, [dynamic loading is possible])
+		if test "$GCC" = yes
+		then
+			# Add others here as appropriate,
+			# one day use GNU libtool.
+			if uname | $EGREP -i 'linux|freebsd|cygwin' > /dev/null
+			then
+				LDFLAGS="$LDFLAGS -export-dynamic"
+			fi
+		fi
+	])
 ])
 
 dnl check for how to use getpgrp
--- gawk-3.1.5/main.c~	2005-07-06 16:05:14.000000000 +0200
+++ gawk-3.1.5/main.c	2005-09-02 13:34:36.176763197 +0200
@@ -226,11 +226,6 @@
 	if (getenv("WHINY_USERS") != NULL)
 		whiny_users = TRUE;
 
-#ifdef HAVE_MCHECK_H
-	if (do_tidy_mem)
-		mtrace();
-#endif /* HAVE_MCHECK_H */
-	
 #if defined(LC_CTYPE)
 	setlocale(LC_CTYPE, "");
 #endif
