--- ./src/progress.c~	Fri May 24 14:58:17 2002
+++ ./src/progress.c	Sat Mar 15 20:48:35 2003
@@ -690,7 +690,7 @@
   else if (bp->total_length > 0)
     APPEND_LITERAL ("    ETA --:--");
 
-  assert (p - bp->buffer <= bp->width);
+  assert ((p - bp->buffer) <= bp->width);
 
   while (p < bp->buffer + bp->width)
     *p++ = ' ';
--- ./src/url.c~	Sat Mar 15 19:50:51 2003
+++ ./src/url.c	Sat Mar 15 20:49:43 2003
@@ -205,7 +205,7 @@
 	*p2++ = *p1++;
     }
   *p2 = '\0';
-  assert (p2 - newstr == newlen);
+  assert ((p2 - newstr) == newlen);
 
   return newstr;
 }
@@ -408,7 +408,7 @@
 	}
     }
   *p2 = '\0';
-  assert (p2 - newstr == newlen);
+  assert ((p2 - newstr) == newlen);
   return newstr;
 }
 
@@ -1846,7 +1846,7 @@
   p += fplen;
   *p++ = '\0';
 
-  assert (p - result == size);
+  assert ((p - result) == size);
 
   if (quoted_user && quoted_user != url->user)
     xfree (quoted_user);
@@ -2337,7 +2337,7 @@
 	      *to++ = 'F';
 	    }
 	}
-      assert (to - newname == fsqlen);
+      assert ((to - newname) == fsqlen);
       *to = '\0';
 
       file_sans_qmark = newname;
