--- linux-2.4.22/scripts/mkdep.c.orig	2003-09-20 20:54:57.000000000 +0200
+++ linux-2.4.22/scripts/mkdep.c	2003-09-20 20:55:43.000000000 +0200
@@ -38,7 +38,9 @@
 #include <string.h>
 #include <unistd.h>
 
+#ifndef __dietlibc__
 #include <sys/fcntl.h>
+#endif
 #include <sys/mman.h>
 #include <sys/stat.h>
 #include <sys/types.h>
--- linux-2.4.22/lib/gen_crc32table.c.orig	2003-09-20 21:51:19.000000000 +0200
+++ linux-2.4.22/lib/gen_crc32table.c	2003-09-20 21:51:48.000000000 +0200
@@ -1,3 +1,6 @@
+#ifdef __dietlibc__
+#define _GNU_SOURCE
+#endif
 #include <stdio.h>
 #include "crc32defs.h"
 #include <sys/types.h>
