diff -ruN ppp-2.4.3.orig/include/net/ppp_defs.h ppp-2.4.3/include/net/ppp_defs.h
--- ppp-2.4.3.orig/include/net/ppp_defs.h	2002-12-06 10:49:15.000000000 +0100
+++ ppp-2.4.3/include/net/ppp_defs.h	2005-07-13 22:30:36.000000000 +0200
@@ -107,7 +107,7 @@
  */
 
 #if !defined(__BIT_TYPES_DEFINED__) && !defined(_BITYPES) \
- && !defined(__FreeBSD__) && (NS_TARGET < 40)
+&& !defined(__FreeBSD__) && !defined(_GNU_SOURCE) && (NS_TARGET < 40) 
 #ifdef	UINT32_T
 typedef UINT32_T	u_int32_t;
 #else
diff -ruN ppp-2.4.3.orig/pppd/Makefile.linux ppp-2.4.3/pppd/Makefile.linux
--- ppp-2.4.3.orig/pppd/Makefile.linux	2004-11-13 13:02:22.000000000 +0100
+++ ppp-2.4.3/pppd/Makefile.linux	2005-07-14 01:34:39.000000000 +0200
@@ -65,7 +65,7 @@
 #HAVE_INET6=y
 
 # Enable plugins
-PLUGIN=y
+#PLUGIN=y
 
 # Enable Microsoft proprietary Callback Control Protocol
 #CBCP=y
@@ -121,7 +121,7 @@
 CFLAGS   += -DHAVE_CRYPT_H=1
 endif
 ifneq ($(wildcard /usr/lib/libcrypt.*),)
-LIBS	+= -lcrypt
+LIBS   += -lcrypt
 endif
 
 ifdef NEEDDES
@@ -167,13 +167,13 @@
 ifdef PLUGIN
 CFLAGS	+= -DPLUGIN
 LDFLAGS	+= -Wl,-E
-LIBS	+= -ldl
+LIBS	+=  
 endif
 
 ifdef FILTER
 ifneq ($(wildcard /usr/include/pcap-bpf.h),)
-LIBS    += -lpcap
-CFLAGS  += -DPPP_FILTER
+LIBS    += /usr/lib/libpcap.a
+CFLAGS  += -DPPP_FILTER 
 endif
 endif
 
diff -ruN ppp-2.4.3.orig/pppd/pppcrypt.h ppp-2.4.3/pppd/pppcrypt.h
--- ppp-2.4.3.orig/pppd/pppcrypt.h	2002-12-04 23:44:07.000000000 +0100
+++ ppp-2.4.3/pppd/pppcrypt.h	2005-07-14 00:42:06.000000000 +0200
@@ -33,7 +33,7 @@
 #ifndef PPPCRYPT_H
 #define	PPPCRYPT_H
 
-#ifdef HAVE_CRYPT_H
+#if defined(HAVE_CRYPT_H) && !defined(__dietlibc__) 
 #include <crypt.h>
 #endif
 
diff -ruN ppp-2.4.3.orig/pppd/pppd.h ppp-2.4.3/pppd/pppd.h
--- ppp-2.4.3.orig/pppd/pppd.h	2004-11-13 13:02:22.000000000 +0100
+++ ppp-2.4.3/pppd/pppd.h	2005-07-14 01:39:48.000000000 +0200
@@ -54,7 +54,7 @@
 #include <sys/param.h>		/* for MAXPATHLEN and BSD4_4, if defined */
 #include <sys/types.h>		/* for u_int32_t, if defined */
 #include <sys/time.h>		/* for struct timeval */
-#include <net/ppp_defs.h>
+#include <net/ppp_defs.h> 
 #include "patchlevel.h"
 
 #if defined(__STDC__)
diff -ruN ppp-2.4.3.orig/pppd/sys-linux.c ppp-2.4.3/pppd/sys-linux.c
--- ppp-2.4.3.orig/pppd/sys-linux.c	2004-11-12 11:24:43.000000000 +0100
+++ ppp-2.4.3/pppd/sys-linux.c	2005-07-14 00:22:39.000000000 +0200
@@ -73,7 +73,7 @@
 #include <sys/types.h>
 #include <sys/socket.h>
 #include <sys/time.h>
-#include <sys/errno.h>
+#include <errno.h>
 #include <sys/file.h>
 #include <sys/stat.h>
 #include <sys/utsname.h>
@@ -102,7 +102,7 @@
 #define MAX_ADDR_LEN 7
 #endif
 
-#if __GLIBC__ >= 2
+#if __GLIBC__ >= 2 || defined(__dietlibc__) 
 #include <asm/types.h>		/* glibc 2 conflicts with linux/types.h */
 #include <net/if.h>
 #include <net/if_arp.h>
