--- cpio-2.5/rtapelib.c~	Thu Dec  6 05:55:12 2001
+++ cpio-2.5/rtapelib.c	Thu Aug 21 01:17:19 2003
@@ -202,7 +202,8 @@
   return atoi (cp + 1);
 }
 
-#ifdef HAVE_NETDB_H
+#if 0
+/*#ifdef HAVE_NETDB_H*/
 /* Execute /etc/rmt as user USER on remote system HOST using rexec.
    Return a file descriptor of a bidirectional socket for stdin and stdout.
    If USER is NULL, or an empty string, use the current username.
@@ -340,7 +341,8 @@
     }
   *dev = '\0';
 
-#ifdef HAVE_NETDB_H
+#if 0
+/*#ifdef HAVE_NETDB_H*/
   /* Execute the remote command using rexec.  */
   READ (i) = WRITE (i) = _rmt_rexec (system, login);
   if (READ (i) < 0)
