diff -ruN ppp-2.4.3.orig/include/net/ppp_defs.h ppp-2.4.3/include/net/ppp_defs.h
--- ppp-2.4.3.orig/include/net/ppp_defs.h	2002-12-06 10:49:15.000000000 +0100
+++ ppp-2.4.3/include/net/ppp_defs.h	2005-07-13 22:30:36.000000000 +0200
@@ -107,7 +107,7 @@
  */
 
 #if !defined(__BIT_TYPES_DEFINED__) && !defined(_BITYPES) \
- && !defined(__FreeBSD__) && (NS_TARGET < 40)
+&& !defined(__FreeBSD__) && !defined(_GNU_SOURCE) && (NS_TARGET < 40) 
 #ifdef	UINT32_T
 typedef UINT32_T	u_int32_t;
 #else
diff -ruN ppp-2.4.3.orig/pppd/Makefile.linux ppp-2.4.3/pppd/Makefile.linux
--- ppp-2.4.3.orig/pppd/Makefile.linux	2004-11-13 13:02:22.000000000 +0100
+++ ppp-2.4.3/pppd/Makefile.linux	2005-07-13 22:28:25.000000000 +0200
@@ -65,7 +65,7 @@
 #HAVE_INET6=y
 
 # Enable plugins
-PLUGIN=y
+#PLUGIN=y
 
 # Enable Microsoft proprietary Callback Control Protocol
 #CBCP=y
@@ -172,8 +172,8 @@
 
 ifdef FILTER
 ifneq ($(wildcard /usr/include/pcap-bpf.h),)
-LIBS    += -lpcap
-CFLAGS  += -DPPP_FILTER
+LIBS    += /usr/lib/libpcap.a
+CFLAGS  += -DPPP_FILTER 
 endif
 endif
 
diff -ruN ppp-2.4.3.orig/pppd/sys-linux.c ppp-2.4.3/pppd/sys-linux.c
--- ppp-2.4.3.orig/pppd/sys-linux.c	2004-11-12 11:24:43.000000000 +0100
+++ ppp-2.4.3/pppd/sys-linux.c	2005-07-13 22:16:58.000000000 +0200
@@ -102,7 +102,7 @@
 #define MAX_ADDR_LEN 7
 #endif
 
-#if __GLIBC__ >= 2
+#if __GLIBC__ >= 2 || defined(__dietlibc__) 
 #include <asm/types.h>		/* glibc 2 conflicts with linux/types.h */
 #include <net/if.h>
 #include <net/if_arp.h>
