--- ./configure.in.orig	2004-01-26 22:59:15.000000000 +0000
+++ ./configure.in	2004-01-27 18:17:47.675298984 +0000
@@ -46,12 +46,12 @@
 fi
 
 CHECK_WX_BUILT_WITH_GTK2
-
+ 
 if test x"$GTK_USEDVERSION" = x"2" ; then
-  AM_PATH_GTK_2_0(2.0.3, havegtk2=yes, havegtk2=no)
+  AM_PATH_GTK_2_0(2.0.3, havegtk2=yes, havegtk2=no, gthread)
   GTK_DEFS="-D__GTK2__"
 else
-  AM_PATH_GTK(1.2.0, havegtk=yes, havegtk=no)
+  AM_PATH_GTK(1.2.0, havegtk=yes, havegtk=no, gthread)
   GTK_DEFS=""
 fi
 
 
