--- ./selector.c.orig	2005-02-09 13:13:48.000000000 -0300
+++ ./selector.c	2005-02-09 13:24:01.514859312 -0300
@@ -18,6 +18,7 @@
  */
 
 #include "gasgui.h"
+#define __DIALOG_OLD_NAMES__
 #include <dialog.h>
 #include <limits.h>
 #include <sys/ioctl.h>
@@ -535,7 +536,7 @@
 		screensize_y = ws.ws_row - 3;
 	}
 
-	init_dialog();
+	init_dialog(stdin,stdout);
 
 	/* Yup - that's a label for goto.... */
 main_loop:
