diff -Nur cups-1.1.18-org/man/Makefile cups-1.1.18/man/Makefile
--- cups-1.1.18-org/man/Makefile	2003-04-21 14:12:02.000000000 +0200
+++ cups-1.1.18/man/Makefile	2003-04-21 14:36:55.000000000 +0200
@@ -70,26 +70,30 @@
 install:	all
 	$(INSTALL_DIR) $(MANDIR)/man1
 	for file in $(MAN1); do \
-		$(INSTALL_MAN) $$file $(MANDIR)/man1/`basename $$file man`1; \
+	        inst=`echo $$file | sed -e 's/\\.[0-9a-z]*$$//'`; \
+		$(INSTALL_MAN) $$file $(MANDIR)/man1/$$inst-cups.1; \
 	done
-	$(RM) $(MANDIR)/man1/cancel.1
-	$(LN) lp.1 $(MANDIR)/man1/cancel.1
+	$(RM) $(MANDIR)/man1/cancel-cups.1
+	$(LN) lp-cups.1 $(MANDIR)/man1/cancel-cups.1
 	$(INSTALL_DIR) $(PMANDIR)/man3
 	for file in $(MAN3); do \
-		$(INSTALL_MAN) $$file $(PMANDIR)/man3/`basename $$file man`3; \
+	        inst=`echo $$file | sed -e 's/\\.[0-9a-z]*$$//'`; \
+		$(INSTALL_MAN) $$file $(PMANDIR)/man3/$$inst-cups.3; \
 	done
 	$(INSTALL_DIR) $(MANDIR)/man5
 	for file in $(MAN5); do \
-		$(INSTALL_MAN) $$file $(MANDIR)/man5/`basename $$file man`5; \
+	        inst=`echo $$file | sed -e 's/\\.[0-9a-z]*$$//'`; \
+		$(INSTALL_MAN) $$file $(MANDIR)/man5/$$inst-cups.5; \
 	done
 	$(INSTALL_DIR) $(AMANDIR)/man$(MAN8EXT)
 	for file in $(MAN8); do \
-		$(INSTALL_MAN) $$file $(AMANDIR)/man$(MAN8EXT)/`basename $$file man`$(MAN8EXT); \
+	        inst=`echo $$file | sed -e 's/\\.[0-9a-z]*$$//'`; \
+		$(INSTALL_MAN) $$file $(AMANDIR)/man$(MAN8EXT)/$$inst-cups.$(MAN8EXT); \
 	done
-	$(RM) $(AMANDIR)/man$(MAN8EXT)/reject.$(MAN8EXT)
-	$(LN) accept.$(MAN8EXT) $(AMANDIR)/man$(MAN8EXT)/reject.$(MAN8EXT)
-	$(RM) $(AMANDIR)/man$(MAN8EXT)/disable.$(MAN8EXT)
-	$(LN) enable.$(MAN8EXT) $(AMANDIR)/man$(MAN8EXT)/disable.$(MAN8EXT)
+	$(RM) $(AMANDIR)/man$(MAN8EXT)/reject-cups.$(MAN8EXT)
+	$(LN) accept-cups.$(MAN8EXT) $(AMANDIR)/man$(MAN8EXT)/reject-cups.$(MAN8EXT)
+	$(RM) $(AMANDIR)/man$(MAN8EXT)/disable-cups.$(MAN8EXT)
+	$(LN) enable-cups.$(MAN8EXT) $(AMANDIR)/man$(MAN8EXT)/disable-cups.$(MAN8EXT)
 	$(INSTALL_DIR) $(MANDIR)/cat1
 	for file in $(CAT1); do \
 		$(INSTALL_MAN) $$file $(MANDIR)/cat1; \
