--- ./cups.list.in.orig	2003-03-12 21:47:30.000000000 +0100
+++ ./cups.list.in	2003-03-12 21:50:39.000000000 +0100
@@ -142,7 +142,7 @@
 f 0555 root sys $SBINDIR/accept systemv/accept
 f 0555 root sys $SBINDIR/cupsaddsmb systemv/cupsaddsmb
 f 0555 root sys $SBINDIR/lpadmin systemv/lpadmin
-f 0555 root sys $SBINDIR/lpc berkeley/lpc
+f 0555 root sys $SBINDIR/lpc_cups berkeley/lpc
 f 0555 root sys $SBINDIR/lpinfo systemv/lpinfo
 f 0555 root sys $SBINDIR/lpmove systemv/lpmove
 l 0555 root sys $SBINDIR/reject accept
@@ -152,14 +152,14 @@
 %system all
 
 # User commands
-f 0555 root sys $BINDIR/cancel systemv/cancel
-f 0555 root sys $BINDIR/lp systemv/lp
+f 0555 root sys $BINDIR/cancel_cups systemv/cancel
+f 0555 root sys $BINDIR/lp_cups systemv/lp
 f 0555 root sys $BINDIR/lpoptions systemv/lpoptions
 f 4555 $CUPS_USER sys $BINDIR/lppasswd systemv/lppasswd
-f 0555 root sys $BINDIR/lpq berkeley/lpq
-f 0555 root sys $BINDIR/lpr berkeley/lpr
-f 0555 root sys $BINDIR/lprm berkeley/lprm
-f 0555 root sys $BINDIR/lpstat systemv/lpstat
+f 0555 root sys $BINDIR/lpq_cups berkeley/lpq
+f 0555 root sys $BINDIR/lpr_cups berkeley/lpr
+f 0555 root sys $BINDIR/lprm_cups berkeley/lprm
+f 0555 root sys $BINDIR/lpstat_cups systemv/lpstat
 
 %system irix
 l 0555 root sys /usr/bsd/lpq $BINDIR/lpq
@@ -359,12 +359,12 @@
 f 0444 root sys $MANDIR/man1/filter.$MAN1EXT man/filter.man
 f 0444 root sys $MANDIR/man1/lpoptions.$MAN1EXT man/lpoptions.man
 f 0444 root sys $MANDIR/man1/lppasswd.$MAN1EXT man/lppasswd.man
-f 0444 root sys $MANDIR/man1/lpq.$MAN1EXT man/lpq.man
-f 0444 root sys $MANDIR/man1/lprm.$MAN1EXT man/lprm.man
-f 0444 root sys $MANDIR/man1/lpr.$MAN1EXT man/lpr.man
-f 0444 root sys $MANDIR/man1/lpstat.$MAN1EXT man/lpstat.man
-f 0444 root sys $MANDIR/man1/lp.$MAN1EXT man/lp.man
-l 0444 root sys $MANDIR/man1/cancel.$MAN1EXT lp.$MAN1EXT
+f 0444 root sys $MANDIR/man1/lpq_cups.$MAN1EXT man/lpq.man
+f 0444 root sys $MANDIR/man1/lprm_cups.$MAN1EXT man/lprm.man
+f 0444 root sys $MANDIR/man1/lpr_cups.$MAN1EXT man/lpr.man
+f 0444 root sys $MANDIR/man1/lpstat_cups.$MAN1EXT man/lpstat.man
+f 0444 root sys $MANDIR/man1/lp_cups.$MAN1EXT man/lp.man
+l 0444 root sys $MANDIR/man1/cancel_cups.$MAN1EXT lp.$MAN1EXT
 
 f 0444 root sys $PMANDIR/man3/cups-config.$MAN3EXT man/cups-config.man
 
