--- tools/xenstat/xentop/Makefile	2006-04-07 14:24:59.000000000 +0200
+++ tools/xenstat/xentop/Makefile	2006-04-09 16:37:55.634761500 +0200
@@ -26,8 +26,8 @@
 sbindir=$(prefix)/sbin
 
 CFLAGS += -DGCC_PRINTF -Wall -Werror -I$(XEN_LIBXENSTAT)
-LDFLAGS += -L$(XEN_LIBXENSTAT)
-LDLIBS += -lxenstat -lncurses
+LDFLAGS += -L$(XEN_LIBXENSTAT) -L$(XEN_XENSTORE)
+LDLIBS += -lxenstat -lncurses -lxenstore
 
 all: xentop
 
