--- ./docs/kdeedu/kturtle/using-kturtle.docbook.orig	2005-03-11 23:04:55.008715448 -0300
+++ ./docs/kdeedu/kturtle/using-kturtle.docbook	2005-03-11 23:06:08.570532352 -0300
@@ -1011,7 +1011,7 @@
 ></term>
       <listitem
 ><para
->This is a very useful function: it provides help on the code where the cursor in the code editor is at. So, &eg;, you have used the <userinput
+>This is a very useful function: it provides help on the code where the cursor in the code editor is at. So, &zb;, you have used the <userinput
 >print</userinput
 > command in your code, and you want to read and to know what the handbook says on this command. You just move your cursor so it is in the <userinput
 >print</userinput
