--- ./subversion/bindings/swig/perl/native/Core.pm.orig	2005-01-10 14:09:35.000000000 -0300
+++ ./subversion/bindings/swig/perl/native/Core.pm	2005-01-10 14:09:47.000000000 -0300
@@ -3,8 +3,7 @@
 
 package SVN::Core;
 use SVN::Base qw(Core svn_ VERSION);
-$SVN::Core::VERSION = "$SVN::Core::VER_MAJOR.$SVN::Core::VER_MINOR." .
-                      "$SVN::Core::VER_MICRO";
+$SVN::Core::VERSION = "$SVN::Core::VER_MAJOR.$SVN::Core::VER_MINOR." . "$SVN::Core::VER_MICRO";
 
 =head1 NAME
 
