--- ./libparted/linux.c.orig	2004-02-11 00:18:55.623764288 +0000
+++ ./libparted/linux.c	2004-02-11 00:19:35.662677448 +0000
@@ -164,7 +164,7 @@
 #define BLKSSZGET  _IO(0x12,104) /* get block device sector size */
 #define BLKGETLASTSECT  _IO(0x12,108) /* get last sector of block device */
 #define BLKSETLASTSECT  _IO(0x12,109) /* set last sector of block device */
-#define BLKGETSIZE64 _IOR(0x12,114,sizeof(uint64_t))	/* return device size in bytes (u64 *arg) */
+#define BLKGETSIZE64 _IOR(0x12,114,size_t)	/* return device size in bytes (u64 *arg) */
 
 struct blkdev_ioctl_param {
 	unsigned int block;
