diff --git a/src/tseng_driver.c b/src/tseng_driver.c
index fe4238b..62852c4 100644
--- a/src/tseng_driver.c
+++ b/src/tseng_driver.c
@@ -1015,7 +1015,7 @@ TsengPreInit(ScrnInfoPtr pScrn, int flags)
      * Our default depth is 8, so pass it to the helper function.
      * Our preference for depth 24 is 24bpp, so tell it that too.
      */
-    if (!xf86SetDepthBpp(pScrn, 8, 8, 8, Support24bppFb | Support32bppFb |
+    if (!xf86SetDepthBpp(pScrn, 16, 16, 16, Support24bppFb | Support32bppFb |
                          SupportConvert32to24 | PreferConvert32to24)) {
 	return FALSE;
     } else {
