--- ../j2sdk-1_4_2_08-linux-i586.bin	2005-04-14 18:17:17.000000000 +0200
+++ j2sdk-1_4_2_08-linux-i586.bin	2005-04-14 18:45:09.000000000 +0200
@@ -11,7 +11,7 @@
 localinstall=$1
 diskSpaceRequired=105996
 MIME_TYPE=application/x-java-jnlp-file
-more <<"EOF"
+cat <<"EOF"
                 Sun Microsystems, Inc.
              Binary Code License Agreement
                      for the
@@ -341,7 +341,7 @@
 Network Circle, Santa Clara, California 95054, U.S.A.
 (LFI#129530/Form ID#011801)
 EOF
-agreed=
+agreed=1
 while [ x$agreed = x ]; do
     echo
     echo "Do you agree to the above license terms? [yes or no] "
@@ -363,7 +363,7 @@
 fi
 trap 'rm -f $outname; exit 1' HUP INT QUIT TERM
 echo "Unpacking..."
-tail +511 $0 > $outname
+tail -n +511 $0 > $outname
 if [ -x /usr/bin/sum ] ; then
     echo "Checksumming..."
 
