--- ./Setup/Unix/Makefile.app.in.orig	2005-02-23 01:42:15.000000000 -0300
+++ ./Setup/Unix/Makefile.app.in	2005-02-23 01:48:40.000000000 -0300
@@ -10,9 +10,8 @@
 	@install -m 0644 Sources/API/application.h $(INC_PREFIX)/ClanLib/application.h
 	@install -d $(LIB_PREFIX)
 	@install Libs/libclanApp.so.$(D_VERSION_MINOR) $(LIB_PREFIX)
-	@install Libs/libclanApp.a $(LIB_PREFIX)
+	@install Libs/libclanApp.a $(LIB_PREFIX)/libclanApp.a.$(D_VERSION_MAJOR)
 	@ln -s -f libclanApp.so.$(D_VERSION_MINOR) $(LIB_PREFIX)/libclanApp.so.$(D_VERSION_MAJOR)
-	@ln -s -f libclanApp.so.$(D_VERSION_MAJOR) $(LIB_PREFIX)/libclanApp.so
 
 
 OBJF_LIB_App = Libs/Intermediate/clanapp.o
--- ./Setup/Unix/Makefile.core.in.orig	2005-02-23 01:42:15.000000000 -0300
+++ ./Setup/Unix/Makefile.core.in	2005-02-23 01:49:31.000000000 -0300
@@ -14,9 +14,8 @@
 	@install -m 0644 Sources/API/efence.h $(INC_PREFIX)/ClanLib/efence.h
 	@install -d $(LIB_PREFIX)
 	@install Libs/libclanCore.so.$(D_VERSION_MINOR) $(LIB_PREFIX)
-	@install Libs/libclanCore.a $(LIB_PREFIX)
+	@install Libs/libclanCore.a $(LIB_PREFIX)/libclanCore.a.$(D_VERSION_MAJOR)
 	@ln -s -f libclanCore.so.$(D_VERSION_MINOR) $(LIB_PREFIX)/libclanCore.so.$(D_VERSION_MAJOR)
-	@ln -s -f libclanCore.so.$(D_VERSION_MAJOR) $(LIB_PREFIX)/libclanCore.so
 	@install -d $(BIN_PREFIX)
 	@install Setup/Unix/clanlib-config $(BIN_PREFIX)
 
--- ./Setup/Unix/Makefile.display.in.orig	2005-02-23 01:42:15.000000000 -0300
+++ ./Setup/Unix/Makefile.display.in	2005-02-23 01:49:51.000000000 -0300
@@ -9,9 +9,8 @@
 	@install -m 0644 Sources/API/display.h $(INC_PREFIX)/ClanLib/display.h
 	@install -d $(LIB_PREFIX)
 	@install Libs/libclanDisplay.so.$(D_VERSION_MINOR) $(LIB_PREFIX)
-	@install Libs/libclanDisplay.a $(LIB_PREFIX)
+	@install Libs/libclanDisplay.a $(LIB_PREFIX)/libclanDisplay.a.$(D_VERSION_MAJOR)
 	@ln -s -f libclanDisplay.so.$(D_VERSION_MINOR) $(LIB_PREFIX)/libclanDisplay.so.$(D_VERSION_MAJOR)
-	@ln -s -f libclanDisplay.so.$(D_VERSION_MAJOR) $(LIB_PREFIX)/libclanDisplay.so
 
 
 OBJF_DISP_GENERIC = \
--- ./Setup/Unix/Makefile.gl.in.orig	2005-02-23 01:42:16.000000000 -0300
+++ ./Setup/Unix/Makefile.gl.in	2005-02-23 01:50:10.000000000 -0300
@@ -10,9 +10,8 @@
 	@install -m 0644 Sources/API/gl.h $(INC_PREFIX)/ClanLib/gl.h
 	@install -d $(LIB_PREFIX)
 	@install Libs/libclanGL.so.$(D_VERSION_MINOR) $(LIB_PREFIX)
-	@install Libs/libclanGL.a $(LIB_PREFIX)
+	@install Libs/libclanGL.a $(LIB_PREFIX)/libclanGL.a.$(D_VERSION_MAJOR)
 	@ln -s -f libclanGL.so.$(D_VERSION_MINOR) $(LIB_PREFIX)/libclanGL.so.$(D_VERSION_MAJOR)
-	@ln -s -f libclanGL.so.$(D_VERSION_MAJOR) $(LIB_PREFIX)/libclanGL.so
 
 OBJF_LIB_GL = Libs/Intermediate/texture_gl.o \
 		Libs/Intermediate/opengl.o \
--- ./Setup/Unix/Makefile.gui.in.orig	2005-02-23 01:42:17.000000000 -0300
+++ ./Setup/Unix/Makefile.gui.in	2005-02-23 01:50:28.000000000 -0300
@@ -10,9 +10,8 @@
 	@install -m 0644 Sources/API/gui.h $(INC_PREFIX)/ClanLib/gui.h
 	@install -d $(LIB_PREFIX)
 	@install Libs/libclanGUI.so.$(D_VERSION_MINOR) $(LIB_PREFIX)
-	@install Libs/libclanGUI.a $(LIB_PREFIX)
+	@install Libs/libclanGUI.a $(LIB_PREFIX)/libclanGUI.a.$(D_VERSION_MAJOR)
 	@ln -s -f libclanGUI.so.$(D_VERSION_MINOR) $(LIB_PREFIX)/libclanGUI.so.$(D_VERSION_MAJOR)
-	@ln -s -f libclanGUI.so.$(D_VERSION_MAJOR) $(LIB_PREFIX)/libclanGUI.so
 
 OBJF_LIB_GUI = \
 	Libs/Intermediate/button.o \
--- ./Setup/Unix/Makefile.jpeg.in.orig	2005-02-23 01:42:17.000000000 -0300
+++ ./Setup/Unix/Makefile.jpeg.in	2005-02-23 01:54:55.493692512 -0300
@@ -10,9 +10,8 @@
 	@install -m 0644 Sources/API/jpeg.h $(INC_PREFIX)/ClanLib/jpeg.h
 	@install -d $(LIB_PREFIX)
 	@install Libs/libclanJPEG.so.$(D_VERSION_MINOR) $(LIB_PREFIX)
-	@install Libs/libclanJPEG.a $(LIB_PREFIX)
+	@install Libs/libclanJPEG.a $(LIB_PREFIX)/libclanJPEG.a.$(D_VERSION_MAJOR)
 	@ln -s -f libclanJPEG.so.$(D_VERSION_MINOR) $(LIB_PREFIX)/libclanJPEG.so.$(D_VERSION_MAJOR)
-	@ln -s -f libclanJPEG.so.$(D_VERSION_MAJOR) $(LIB_PREFIX)/libclanJPEG.so
 
 
 OBJF_LIB_JPEG = \
--- ./Setup/Unix/Makefile.lua.in.orig	2005-02-23 01:42:17.000000000 -0300
+++ ./Setup/Unix/Makefile.lua.in	2005-02-23 01:51:10.000000000 -0300
@@ -10,9 +10,8 @@
 	@install -m 0644 Sources/API/lua.h $(INC_PREFIX)/ClanLib/lua.h
 	@install -d $(LIB_PREFIX)
 	@install Libs/libclanLua.so.$(D_VERSION_MINOR) $(LIB_PREFIX)
-	@install Libs/libclanLua.a $(LIB_PREFIX)
+	@install Libs/libclanLua.a $(LIB_PREFIX)/libclanLua.a.$(D_VERSION_MAJOR)
 	@ln -s -f libclanLua.so.$(D_VERSION_MINOR) $(LIB_PREFIX)/libclanLua.so.$(D_VERSION_MAJOR)
-	@ln -s -f libclanLua.so.$(D_VERSION_MAJOR) $(LIB_PREFIX)/libclanLua.so
 
 OBJF_LIB_LUA = \
 	Libs/Intermediate/clanbindings.o 
--- ./Setup/Unix/Makefile.mikmod.in.orig	2005-02-23 01:42:17.000000000 -0300
+++ ./Setup/Unix/Makefile.mikmod.in	2005-02-23 01:51:30.000000000 -0300
@@ -10,9 +10,8 @@
 	@install -m 0644 Sources/API/mikmod.h $(INC_PREFIX)/ClanLib/mikmod.h
 	@install -d $(LIB_PREFIX)
 	@install Libs/libclanMikMod.so.$(D_VERSION_MINOR) $(LIB_PREFIX)
-	@install Libs/libclanMikMod.a $(LIB_PREFIX)
+	@install Libs/libclanMikMod.a $(LIB_PREFIX)/libclanMikMod.a.$(D_VERSION_MAJOR)
 	@ln -s -f libclanMikMod.so.$(D_VERSION_MINOR) $(LIB_PREFIX)/libclanMikMod.so.$(D_VERSION_MAJOR)
-	@ln -s -f libclanMikMod.so.$(D_VERSION_MAJOR) $(LIB_PREFIX)/libclanMikMod.so
 
 OBJF_LIB_MIKMOD = \
 	Libs/Intermediate/resourcetype_module.o \
--- ./Setup/Unix/Makefile.mpeg.in.orig	2005-02-23 01:42:17.000000000 -0300
+++ ./Setup/Unix/Makefile.mpeg.in	2005-02-23 01:51:41.000000000 -0300
@@ -10,9 +10,8 @@
 	@install -m 0644 Sources/API/mpeg.h $(INC_PREFIX)/ClanLib/mpeg.h
 	@install -d $(LIB_PREFIX)
 	@install Libs/libclanMPEG.so.$(D_VERSION_MINOR) $(LIB_PREFIX)
-	@install Libs/libclanMPEG.a $(LIB_PREFIX)
+	@install Libs/libclanMPEG.a $(LIB_PREFIX)/libclanMPEG.a.$(D_VERSION_MAJOR)
 	@ln -s -f libclanMPEG.so.$(D_VERSION_MINOR) $(LIB_PREFIX)/libclanMPEG.so.$(D_VERSION_MAJOR)
-	@ln -s -f libclanMPEG.so.$(D_VERSION_MAJOR) $(LIB_PREFIX)/libclanMPEG.so
 
 OBJF_LIB_MPEG =	\
 	Libs/Intermediate/mpeg_soundprovider.o
--- ./Setup/Unix/Makefile.network.in.orig	2005-02-23 01:42:17.000000000 -0300
+++ ./Setup/Unix/Makefile.network.in	2005-02-23 01:51:55.000000000 -0300
@@ -10,9 +10,8 @@
 	@install -m 0644 Sources/API/network.h $(INC_PREFIX)/ClanLib/network.h
 	@install -d $(LIB_PREFIX)
 	@install Libs/libclanNetwork.so.$(D_VERSION_MINOR) $(LIB_PREFIX)
-	@install Libs/libclanNetwork.a $(LIB_PREFIX)
+	@install Libs/libclanNetwork.a $(LIB_PREFIX)/libclanNetwork.a.$(D_VERSION_MAJOR)
 	@ln -s -f libclanNetwork.so.$(D_VERSION_MINOR) $(LIB_PREFIX)/libclanNetwork.so.$(D_VERSION_MAJOR)
-	@ln -s -f libclanNetwork.so.$(D_VERSION_MAJOR) $(LIB_PREFIX)/libclanNetwork.so
 
 OBJF_LIB_NETWORK = Libs/Intermediate/netcomputer.o \
 	Libs/Intermediate/browse_client.o \
--- ./Setup/Unix/Makefile.png.in.orig	2005-02-23 01:42:17.000000000 -0300
+++ ./Setup/Unix/Makefile.png.in	2005-02-23 01:52:08.000000000 -0300
@@ -10,9 +10,8 @@
 	@install -m 0644 Sources/API/png.h $(INC_PREFIX)/ClanLib/png.h
 	@install -d $(LIB_PREFIX)
 	@install Libs/libclanPNG.so.$(D_VERSION_MINOR) $(LIB_PREFIX)
-	@install Libs/libclanPNG.a $(LIB_PREFIX)
+	@install Libs/libclanPNG.a $(LIB_PREFIX)/libclanPNG.a.$(D_VERSION_MAJOR)
 	@ln -s -f libclanPNG.so.$(D_VERSION_MINOR) $(LIB_PREFIX)/libclanPNG.so.$(D_VERSION_MAJOR)
-	@ln -s -f libclanPNG.so.$(D_VERSION_MAJOR) $(LIB_PREFIX)/libclanPNG.so
 
 
 OBJF_LIB_PNG = \
--- ./Setup/Unix/Makefile.smalljpeg.in.orig	2005-02-23 01:42:17.000000000 -0300
+++ ./Setup/Unix/Makefile.smalljpeg.in	2005-02-23 01:52:33.000000000 -0300
@@ -10,9 +10,8 @@
 	@install -m 0644 Sources/API/jpeg.h $(INC_PREFIX)/ClanLib/jpeg.h
 	@install -d $(LIB_PREFIX)
 	@install Libs/libclanSmallJPEG.so.$(D_VERSION_MINOR) $(LIB_PREFIX)
-	@install Libs/libclanSmallJPEG.a $(LIB_PREFIX)
+	@install Libs/libclanSmallJPEG.a $(LIB_PREFIX)/libclanSmallJPEG.a.$(D_VERSION_MAJOR)
 	@ln -s -f libclanSmallJPEG.so.$(D_VERSION_MINOR) $(LIB_PREFIX)/libclanSmallJPEG.so.$(D_VERSION_MAJOR)
-	@ln -s -f libclanSmallJPEG.so.$(D_VERSION_MAJOR) $(LIB_PREFIX)/libclanSmallJPEG.so
 
 
 OBJF_LIB_SMALLJPEG = \
--- ./Setup/Unix/Makefile.sound.in.orig	2005-02-23 01:42:17.000000000 -0300
+++ ./Setup/Unix/Makefile.sound.in	2005-02-23 01:52:46.000000000 -0300
@@ -9,9 +9,8 @@
 	@install -m 0644 Sources/API/sound.h $(INC_PREFIX)/ClanLib/sound.h
 	@install -d $(LIB_PREFIX)
 	@install Libs/libclanSound.so.$(D_VERSION_MINOR) $(LIB_PREFIX)
-	@install Libs/libclanSound.a $(LIB_PREFIX)
+	@install Libs/libclanSound.a $(LIB_PREFIX)/libclanSound.a.$(D_VERSION_MAJOR)
 	@ln -s -f libclanSound.so.$(D_VERSION_MINOR) $(LIB_PREFIX)/libclanSound.so.$(D_VERSION_MAJOR)
-	@ln -s -f libclanSound.so.$(D_VERSION_MAJOR) $(LIB_PREFIX)/libclanSound.so
 
 OBJF_SOUND_GENERIC = \
 	Libs/Intermediate/sound.o \
--- ./Setup/Unix/Makefile.ttf.in.orig	2005-02-23 01:42:17.000000000 -0300
+++ ./Setup/Unix/Makefile.ttf.in	2005-02-23 01:53:50.000000000 -0300
@@ -10,9 +10,8 @@
 	@install -m 0644 Sources/API/ttf.h $(INC_PREFIX)/ClanLib/ttf.h
 	@install -d $(LIB_PREFIX)
 	@install Libs/libclanTTF.so.$(D_VERSION_MINOR) $(LIB_PREFIX)
-	@install Libs/libclanTTF.a $(LIB_PREFIX)
+	@install Libs/libclanTTF.a $(LIB_PREFIX)/libclanTTF.a.$(D_VERSION_MAJOR)
 	@ln -s -f libclanTTF.so.$(D_VERSION_MINOR) $(LIB_PREFIX)/libclanTTF.so.$(D_VERSION_MAJOR)
-	@ln -s -f libclanTTF.so.$(D_VERSION_MAJOR) $(LIB_PREFIX)/libclanTTF.so
 
 
 OBJF_LIB_TTF = \
--- ./Setup/Unix/Makefile.vorbis.in.orig	2005-02-23 01:42:17.000000000 -0300
+++ ./Setup/Unix/Makefile.vorbis.in	2005-02-23 01:54:03.446604872 -0300
@@ -10,9 +10,8 @@
 	@install -m 0644 Sources/API/vorbis.h $(INC_PREFIX)/ClanLib/vorbis.h
 	@install -d $(LIB_PREFIX)
 	@install Libs/libclanVorbis.so.$(D_VERSION_MINOR) $(LIB_PREFIX)
-	@install Libs/libclanVorbis.a $(LIB_PREFIX)
+	@install Libs/libclanVorbis.a $(LIB_PREFIX)/libclanVorbis.a.$(D_VERSION_MAJOR)
 	@ln -s -f libclanVorbis.so.$(D_VERSION_MINOR) $(LIB_PREFIX)/libclanVorbis.so.$(D_VERSION_MAJOR)
-	@ln -s -f libclanVorbis.so.$(D_VERSION_MAJOR) $(LIB_PREFIX)/libclanVorbis.so
 
 OBJF_LIB_VORBIS = \
 	Libs/Intermediate/vorbis_soundprovider_generic.o \
