--- madwifi-old-r1417-20060128/Makefile.orig	2006-02-06 22:00:01.000000000 -0800
+++ madwifi-old-r1417-20060128/Makefile	2006-02-06 21:59:31.000000000 -0800
@@ -56,13 +56,13 @@
 	for i in ${DIRS}; do \
 		$(MAKE) -C $$i install || exit 1; \
 	done
-	@if [ -z $(DESTDIR) ]; then \
-	    /sbin/depmod -ae ; \
-	elif [ -f $(SYSTEMMAP) -a -z "$(NO_DEPMOD)" ]; then \
-	    /sbin/depmod -ae -b $(DESTDIR) -F $(SYSTEMMAP) $(KERNELRELEASE) ; \
-	else \
-	    echo "Don't forget to run \"depmod -ae\" on the target system."; \
-	fi
+	#@if [ -z $(DESTDIR) ]; then \
+	#    /sbin/depmod -ae ; \
+	#elif [ -f $(SYSTEMMAP) -a -z "$(NO_DEPMOD)" ]; then \
+	#    /sbin/depmod -ae -b $(DESTDIR) -F $(SYSTEMMAP) $(KERNELRELEASE) ; \
+	#else \
+	#    echo "Don't forget to run \"depmod -ae\" on the target system."; \
+	#fi
 
 FILES=	ath_hal COPYRIGHT ath include Makefile Makefile.inc \
 	README release.h share net80211 ath_rate
