--- madwifi/Makefile.orig	2005-06-24 10:00:38.000000000 -0700
+++ madwifi/Makefile	2005-06-24 10:01:12.000000000 -0700
@@ -56,13 +56,13 @@
 	for i in ${DIRS}; do \
 		(cd $$i; make install) || exit 1; \
 	done
-	@if [ -z $(DESTDIR) ]; then \
-	    /sbin/depmod -ae ; \
-	elif [ -f $(SYSTEMMAP) ]; then \
-	    /sbin/depmod -ae -b $(DESTDIR) -F $(SYSTEMMAP) $(KERNELRELEASE) ; \
-	else \
-	    echo "Don't forget to run \"depmod -ae\" on the target system."; \
-	fi
+	#@if [ -z $(DESTDIR) ]; then \
+	#    /sbin/depmod -ae ; \
+	#elif [ -f $(SYSTEMMAP) ]; then \
+	#    /sbin/depmod -ae -b $(DESTDIR) -F $(SYSTEMMAP) $(KERNELRELEASE) ; \
+	#else \
+	#    echo "Don't forget to run \"depmod -ae\" on the target system."; \
+	#fi
 
 FILES=	ath_hal COPYRIGHT ath include Makefile Makefile.inc \
 	README release.h share net80211 ath_rate
