--- postfix/conf/postfix-files	2003-01-12 18:08:12.000000000 +0100
+++ postfix/conf/postfix-files	2003-05-21 18:46:01.000000000 +0200
@@ -101,8 +101,8 @@
 $config_directory/virtual:f:root:-:644:p
 $config_directory/postfix-script:f:root:-:755
 $config_directory/post-install:f:root:-:755
-$manpage_directory/man1/mailq.1:f:root:-:644
-$manpage_directory/man1/newaliases.1:f:root:-:644
+$manpage_directory/man1/mailq_postfix.1:f:root:-:644
+$manpage_directory/man1/newaliases_postfix.1:f:root:-:644
 $manpage_directory/man1/postalias.1:f:root:-:644
 $manpage_directory/man1/postcat.1:f:root:-:644
 $manpage_directory/man1/postconf.1:f:root:-:644
@@ -116,7 +116,7 @@
 $manpage_directory/man1/postsuper.1:f:root:-:644
 $manpage_directory/man1/sendmail.1:f:root:-:644
 $manpage_directory/man5/access.5:f:root:-:644
-$manpage_directory/man5/aliases.5:f:root:-:644
+$manpage_directory/man5/aliases_postfix.5:f:root:-:644
 $manpage_directory/man5/canonical.5:f:root:-:644
 $manpage_directory/man5/pcre_table.5:f:root:-:644
 $manpage_directory/man5/regexp_table.5:f:root:-:644
