--- ./Sources/Display/Display/DirectFB/target_directfb.cpp.orig	2003-08-05 15:59:59.000000000 +0200
+++ ./Sources/Display/Display/DirectFB/target_directfb.cpp	2003-08-05 16:00:38.000000000 +0200
@@ -104,7 +104,7 @@
 
 unsigned int CL_Target_DirectFB::get_width() const
 {
-  unsigned int width = 0;
+  int width = 0;
 
   surface->GetSize (surface, &width, NULL);
 
@@ -113,7 +113,7 @@
 
 unsigned int CL_Target_DirectFB::get_height() const
 {
-  unsigned int height = 0;
+  int height = 0;
 
   surface->GetSize (surface, NULL, &height);
 
@@ -263,7 +263,7 @@
 
 void CL_Target_DirectFB::Clear (float r, float g, float b, float a)
 {
-  unsigned int width, height;
+  int width, height;
 
   surface->GetSize (surface, &width, &height);
 
