diff -ur wine-20030508/dlls/winmm/winejack/Makefile.in.old wine-20030508/dlls/winmm/winejack/Makefile.in
--- wine-20030508/dlls/winmm/winejack/Makefile.in.old	2002-12-13 03:26:18.000000000 +0100
+++ wine-20030508/dlls/winmm/winejack/Makefile.in	2003-05-19 21:03:49.000000000 +0200
@@ -4,6 +4,8 @@
 VPATH     = @srcdir@
 MODULE    = winejack.drv
 IMPORTS   = winmm user32 kernel32
+LIBUUID   = -L$(TOPOBJDIR)/libs/uuid -lwine_uuid
+EXTRALIBS = $(LIBUUID)
 
 LDDLLFLAGS = @LDDLLFLAGS@
 SYMBOLFILE = $(MODULE).tmp.o
