--- gcc-4.0.2/configure.orig	2005-10-29 22:12:08.153060784 +0200
+++ gcc-4.0.2/configure	2005-10-29 22:12:49.993700040 +0200
@@ -7,6 +7,9 @@
 # This configure script is free software; the Free Software Foundation
 # gives unlimited permission to copy, distribute and modify it.
 
+# declare following option so we can pass it from gcc.conf
+true --with-multilib-list
+
 # Defaults:
 ac_help=
 ac_default_prefix=/usr/local
