diff -Naur ucd-snmp-4.2.6-/snmplib/scapi.c ucd-snmp-4.2.6/snmplib/scapi.c
--- ucd-snmp-4.2.6-/snmplib/scapi.c	2001-10-01 10:52:49.000000000 +0100
+++ ucd-snmp-4.2.6/snmplib/scapi.c	2003-04-08 15:13:09.000000000 +0100
@@ -609,7 +609,7 @@
 	memset(my_iv, 0, sizeof(my_iv));
 	memset(pad_block, 0, sizeof(pad_block));
 	memset(key_struct, 0, sizeof(key_struct));
-	memset(key_sch, 0, sizeof(key_sch));
+	memset((char*)&key_sch, 0, sizeof(key_sch));
 	return rval;
 
 }  /* end sc_encrypt() */
@@ -722,7 +722,7 @@
 
 /* exit cond */
 sc_decrypt_quit:
-	memset(key_sch, 0, sizeof(key_sch));
+	memset((char*)&key_sch, 0, sizeof(key_sch));
 	memset(key_struct, 0, sizeof(key_struct));
 	memset(my_iv, 0, sizeof(my_iv));
 	return rval;
