diff -Naur ssh-gui-0.7.1/Makefile ssh-gui-0.7.1-patched1/Makefile
--- ssh-gui-0.7.1/Makefile	2000-05-13 12:44:23.000000000 +0800
+++ ssh-gui-0.7.1-patched1/Makefile	2007-12-03 01:44:31.000000000 +0800
@@ -4,14 +4,14 @@
 VERSION="0.5"
 
 #Get GTK+ and GLIB CC flags
-GTKFLAGS=`gtk-config --cflags`
-GLIBFLAGS=`glib-config --cflags`
+GTKFLAGS=`pkg-config --cflags "gtk+-2.0"`
+GLIBFLAGS=`pkg-config --cflags "glib-2.0"`
 
 CFLAGS=-Wall $(GTKFLAGS) $(GLIBFLAGS) -DPACKAGE=$(PACKAGE) -DVERSION=$(VERSION)
 
 #Get GTK+ and GLIB library flags
-GTKLIBS=`gtk-config --libs`
-GLIBLIBS=`glib-config --libs`
+GTKLIBS=`pkg-config --libs "gtk+-2.0"`
+GLIBLIBS=`pkg-config --libs "glib-2.0"`
 
 LIBS=$(GTKLIBS) $(GLIBLIBS)
 
