--- ./lib/mc.ext.in.orig	2000-07-19 09:25:48.000000000 +0300
+++ ./lib/mc.ext.in	2003-06-24 19:20:50.000000000 +0300
@@ -348,26 +348,26 @@
 
 # Manual page
 regex/(([^0-9]|^[^\.]*)\.([1-9][A-Za-z]*|n)|\.man)$
-	Open=nroff @TROFFASCII@ @MANDOC@ %f | %var{PAGER:more}
-	View=%view{ascii,nroff} nroff @TROFFASCII@ @MANDOC@ %f
+	Open=GROFF_NO_SGR= nroff @TROFFASCII@ @MANDOC@ %f | %var{PAGER:more}
+	View=%view{ascii,nroff} GROFF_NO_SGR= nroff @TROFFASCII@ @MANDOC@ %f
 	Icon=man.xpm
 
 # Troff with me macros.
 shell/.me
-	Open=nroff @TROFFASCII@ -me %f | %var{PAGER:more}
-	View=%view{ascii,nroff} nroff @TROFFASCII@ -me %f
+	Open=GROFF_NO_SGR= nroff @TROFFASCII@ -me %f | %var{PAGER:more}
+	View=%view{ascii,nroff} GROFF_NO_SGR= nroff @TROFFASCII@ -me %f
 	Icon=man.xpm
 
 # Troff with ms macros.
 shell/.ms
-	Open=nroff @TROFFASCII@ -ms %f | %var{PAGER:more}
-	View=%view{ascii,nroff} nroff @TROFFASCII@ -ms %f
+	Open=GROFF_NO_SGR= nroff @TROFFASCII@ -ms %f | %var{PAGER:more}
+	View=%view{ascii,nroff} GROFF_NO_SGR= nroff @TROFFASCII@ -ms %f
 	Icon=man.xpm
 
 # Manual page - compressed
 regex/([^0-9]|^[^\.]*)\.([1-9][a-z]?|n)\.g?[Zz]$
-	Open=gunzip -dc %f | nroff @TROFFASCII@ @MANDOC@ | %var{PAGER:more}
-	View=%view{ascii,nroff} gunzip -dc %f | nroff @TROFFASCII@ @MANDOC@
+	Open=gunzip -dc %f | GROFF_NO_SGR= nroff @TROFFASCII@ @MANDOC@ | %var{PAGER:more}
+	View=%view{ascii,nroff} gunzip -dc %f | GROFF_NO_SGR= nroff @TROFFASCII@ @MANDOC@
 	Icon=man.xpm
 
 regex/(read\.?me$)|(README)
