diff -Nur pyslsk-1.2.3.orig/pyslsk pyslsk-1.2.3/pyslsk
--- ./pyslsk	2003-06-23 19:24:12.000000000 +0200
+++ ./pyslsk	2003-08-04 02:45:19.000000000 +0200
@@ -55,7 +55,7 @@
 
 def usage():
     print """Usage: pyslsk [OPTION]...
-  -c file,	--config=file	Use non-defult configuration file
+  -c file,	--config=file	Use non-default configuration file
   -h,		--help		Display this help and exit
 
 Report bugs to <ak@sensi.org>""" 	
