--- ./ext/DB_File/Makefile.PL.orig	2002-06-01 19:02:53.000000000 +0200
+++ ./ext/DB_File/Makefile.PL	2003-05-03 11:58:02.000000000 +0200
@@ -8,7 +8,7 @@
 # OS2 is a special case, so check for it now.
 my $OS2 = "-DOS2" if $Config{'osname'} eq 'os2' ;
 
-my $LIB = "-ldb" ;
+my $LIB = "-ldb-3" ;
 # so is win32
 $LIB = "-llibdb" if $^O eq 'MSWin32' ;
 
@@ -16,12 +16,11 @@
 	NAME 		=> 'DB_File',
 	LIBS 		=> ["-L/usr/local/lib $LIB"],
         MAN3PODS        => {},         # Pods will be built by installman.
-	#INC		=> '-I/usr/local/include',
+	INC		=> '-I/usr/include/db3',
 	VERSION_FROM	=> 'DB_File.pm',
 	OBJECT		=> 'version$(OBJ_EXT) DB_File$(OBJ_EXT)',
 	XSPROTOARG	=> '-noprototypes',
 	DEFINE		=> $OS2 || "",
-	INC		=> ($^O eq "MacOS" ? "-i ::::db:include" : ""),
 	'depend'	=> {'version$(OBJ_EXT)' => 'version.c'},
 	'clean'		=> {FILES => 'constants.h constants.xs'},
 	);
