diff -ruN kismet-2004-04-R1/configure.in kismet-2004-04-R1-new/configure.in
--- kismet-2004-04-R1/configure.in	2004-03-17 22:59:34.000000000 +0100
+++ kismet-2004-04-R1-new/configure.in	2004-05-04 13:03:49.265078384 +0200
@@ -164,6 +164,13 @@
 AC_FUNC_STAT
 AC_CHECK_FUNCS([gettimeofday memset select socket strcasecmp strftime strstr])
 
+needsnprintf=no
+AC_CHECK_FUNCS(vsnprintf snprintf,,
+        [needsnprintf=yes])
+if test $needsnprintf = yes; then
+        AC_LIBOBJ(snprintf)
+fi
+
 # Do we have getopt_long natively?
 AC_MSG_CHECKING([for system-level getopt_long()])
 AC_LINK_IFELSE([AC_LANG_PROGRAM([[
