diff -ruN kismet-2004-04-R1/packet.h kismet-2004-04-R1-new/packet.h
--- kismet-2004-04-R1/packet.h	2004-04-02 18:53:36.000000000 +0200
+++ kismet-2004-04-R1-new/packet.h	2004-05-24 13:58:03.682859328 +0200
@@ -550,7 +550,7 @@
     // This isn't quite like STL iterators, because I'm too damned lazy to deal with all
     // the nasty STL hoop-jumping.  This does provide a somewhat-stl-ish interface to
     // iterating through the singleton and masked maps
-    friend class iterator {
+    class iterator {
         friend class macmap;
 
     public:
@@ -643,6 +643,7 @@
         int vector_itr;
         macmap<T> *owner;
     };
+    friend class iterator;
 
     iterator begin() {
         iterator ret(this);
