diff -ruN kismet-2004-04-R1/pcapsource.h kismet-2004-04-R1-new/pcapsource.h
--- kismet-2004-04-R1/pcapsource.h	2004-03-21 00:57:04.000000000 +0100
+++ kismet-2004-04-R1-new/pcapsource.h	2004-05-01 12:02:50.683464560 +0200
@@ -34,7 +34,7 @@
 extern "C" {
 #ifndef HAVE_PCAPPCAP_H
 #include <pcap.h>
-#include <net/bpf.h>
+#include <pcap-bpf.h>
 #else
 #include <pcap/pcap.h>
 #include <pcap/net/bpf.h>
