Index: configure.ac
===================================================================
RCS file: /cvs/xorg/app/xdm/configure.ac,v
retrieving revision 1.36
diff -u -r1.36 configure.ac
--- ./configure.ac	27 Mar 2006 23:59:21 -0000	1.36
+++ ./configure.ac	28 Mar 2006 21:33:51 -0000
@@ -43,7 +43,6 @@
 AM_CONDITIONAL(HAVE_MKTEMP_COMMAND, test x$MKTEMP_COMMAND != x)
 
 AC_FUNC_FORK
-AC_FUNC_SETPGRP
 
 dnl If you're looking here because you got a message that said
 dnl 'error: possibly undefined macro: AC_DEFINE', the source of
Index: daemon.c
===================================================================
RCS file: /cvs/xorg/app/xdm/daemon.c,v
retrieving revision 1.3
diff -u -r1.3 daemon.c
--- ./daemon.c	16 Mar 2006 21:46:55 -0000	1.3
+++ ./daemon.c	28 Mar 2006 21:33:51 -0000
@@ -40,62 +40,9 @@
 #include <stdlib.h>
 #include <string.h>
 
-
-#if defined(USG)
-# include <termios.h>
-#else
-# include <sys/ioctl.h>
-#endif
-#ifdef hpux
-# include <sys/ptyio.h>
-#endif
-
-#ifdef X_NOT_POSIX
-# define Pid_t int
-#else
-# define Pid_t pid_t
-#endif
-
 #include "dm.h"
 #include "dm_error.h"
 
-#ifndef X_NOT_POSIX
-#define HAS_SETSID
-#endif
-
-#ifndef HAS_SETSID
-#define setsid() MySetsid()
-
-static Pid_t
-MySetsid(void)
-{
-#if defined(TIOCNOTTY) || defined(TCCLRCTTY) || defined(TIOCTTY)
-    int fd;
-#endif
-    int stat;
-
-    fd = open("/dev/tty", O_RDWR);
-    if (fd >= 0) {
-#if defined(USG) && defined(TCCLRCTTY)
-       int zero = 0;
-       (void) ioctl (fd, TCCLRCTTY, &zero);
-#elif (defined(SYSV) || defined(SVR4)) && defined(TIOCTTY)
-       int zero = 0;
-       (void) ioctl (i, TIOCTTY, &zero);
-#elif defined(TIOCNOTTY)
-       (void) ioctl (i, TIOCNOTTY, (char *) 0);    /* detach, BSD style */
-#endif
-        close(fd);
-    }
-
-#if defined(SYSV) || defined(__QNXNTO__)
-    return setpgrp();
-#else
-    return setpgid(0, getpid());
-#endif
-}
-#endif /* !HAS_SETSID */
-
 /* detach */
 void
 BecomeDaemon (void)
@@ -103,7 +50,11 @@
 
     /* If our C library has the daemon() function, just use it. */
 #ifdef HAVE_DAEMON
-    daemon (0, 0);
+    if (daemon (0, 0) < 0) {
+       /* error */
+       LogError("daemon() failed, %s\n", strerror(errno));
+       exit(1);
+    }
 #else
     switch (fork()) {
     case -1:
@@ -138,5 +89,5 @@
     (void) open ("/dev/null", O_RDWR);
     (void) dup2 (0, 1);
     (void) dup2 (0, 2);
-#endif /* HAS_DAEMON */
+#endif /* HAVE_DAEMON */
 }
Index: util.c
===================================================================
RCS file: /cvs/xorg/app/xdm/util.c,v
retrieving revision 1.4
diff -u -r1.4 util.c
--- ./util.c	16 Mar 2006 21:46:55 -0000	1.4
+++ ./util.c	28 Mar 2006 21:33:51 -0000
@@ -56,9 +56,6 @@
 #undef _POSIX_SOURCE
 #endif
 #endif
-#if defined(__osf__) || defined(linux) || defined(__QNXNTO__) || defined(__GLIBC__)
-#define setpgrp setpgid
-#endif
 
 void
 printEnv (char **e)
@@ -241,18 +238,7 @@
 void
 CleanUpChild (void)
 {
-#ifdef CSRG_BASED
 	setsid();
-#else
-#ifdef SETPGRP_VOID
-#if !(defined(SVR4) && defined(i386))
-	setpgrp ();
-#endif
-#else
-	setpgrp (0, getpid ());
-	sigsetmask (0);
-#endif
-#endif
 #ifdef SIGCHLD
 	(void) Signal (SIGCHLD, SIG_DFL);
 #endif
