--- src.drwright.ref.20050826.120247.22451.jaguar/drwright-0.18/src/dbus.c.orig	2005-08-26 14:00:03.940452976 -0400
+++ src.drwright.ref.20050826.120247.22451.jaguar/drwright-0.18/src/dbus.c	2005-08-26 14:00:23.953533912 -0400
@@ -56,7 +56,7 @@
 		return FALSE;
 	}
   
-	dbus_bus_acquire_service (bus_conn, DRWRIGHT_DBUS_SERVICE, 0, &error);
+	dbus_bus_request_name (bus_conn, DRWRIGHT_DBUS_SERVICE, 0, &error);
 	if (dbus_error_is_set (&error)) {
 		g_warning ("Failed to acquire drwright service");
 		dbus_error_free (&error);
