--- man-pages-3.05/Makefile.orig	2008-07-25 12:33:02.000000000 -0700
+++ man-pages-3.05/Makefile	2008-07-25 12:33:23.000000000 -0700
@@ -50,7 +50,7 @@
 install:
 	for i in man?; do \
 		install -d -m 755 $(DESTDIR)$(MANDIR)/"$$i" || exit $$?; \
-		install -m 644 "$$i"/* $(DESTDIR)$(MANDIR)/"$$i" || exit $$?; \
+		install -m 644 "$$i"/* $(DESTDIR)$(MANDIR)/"$$i" || echo $$?; \
 		true; \
 	done; \
 
