From 1059526e357da1aa92e5c020332f4b39ceb37503 Mon Sep 17 00:00:00 2001
From: Sergey Poznyakoff <gray@gnu.org>
Date: Tue, 31 Jul 2018 07:39:53 +0000
Subject: Bugfix

* src/gdbmsync.c (gdbm_sync): Always return a meaningful value
---
diff --git a/src/gdbmsync.c b/src/gdbmsync.c
index a8d1002..a75952b 100644
--- a/src/gdbmsync.c
+++ b/src/gdbmsync.c
@@ -28,7 +28,7 @@ int
 gdbm_sync (GDBM_FILE dbf)
 {
   /* Return immediately if the database needs recovery */	
-  GDBM_ASSERT_CONSISTENCY (dbf, );
+  GDBM_ASSERT_CONSISTENCY (dbf, -1);
 
   /* Initialize the gdbm_errno variable. */
   gdbm_set_errno (dbf, GDBM_NO_ERROR, FALSE);
--
cgit v0.9.0.3

From 6311c6b4d7ebe989447151a3e0f8b8ae0b5997d3 Mon Sep 17 00:00:00 2001
From: Sergey Poznyakoff <gray@gnu.org>
Date: Fri, 03 Aug 2018 08:47:01 +0000
Subject: Bugfix

* src/gdbmopen.c (gdbm_avail_table_valid_p): Fix improper use of
the GDBM_DEBUG macro.
---
diff --git a/src/gdbmopen.c b/src/gdbmopen.c
index dc5a89e..908887c 100644
--- a/src/gdbmopen.c
+++ b/src/gdbmopen.c
@@ -99,7 +99,7 @@ gdbm_avail_table_valid_p (GDBM_FILE dbf, avail_elem *av, int count)
 
   if (needs_sorting && dbf->read_write)
     {
-      GDBM_DEBUG (GDBM_DEBUG_ERR, "restoring sort order");
+      GDBM_DEBUG (GDBM_DEBUG_ERR, "%s", "restoring sort order");
       qsort (av, count, sizeof av[0], avail_comp);
     }
   
--
cgit v0.9.0.3
