--- util-linux-2.37.2/sys-utils/blkzone.c.vanilla	2023-11-30 11:54:51.010339280 +0100
+++ util-linux-2.37.2/sys-utils/blkzone.c	2023-11-30 11:54:58.909339789 +0100
@@ -275,10 +275,10 @@
  * blk_zone_report hasn't been packed since https://github.com/torvalds/linux/commit/b3e7e7d2d668de0102264302a4d10dd9d4438a42
  * was merged. See https://github.com/karelzak/util-linux/issues/1083
  */
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Waddress-of-packed-member"
+//#pragma GCC diagnostic push
+//#pragma GCC diagnostic ignored "-Waddress-of-packed-member"
 			const struct blk_zone *entry = &zi->zones[i];
-#pragma GCC diagnostic pop
+//#pragma GCC diagnostic pop
 			unsigned int type = entry->type;
 			uint64_t start = entry->start;
 			uint64_t wp = entry->wp;
