--- xc/programs/Xserver/hw/xfree86/drivers/imstt/imstt_reg.h	2002-08-06 19:57:30.000000000 +0000
+++ xc/programs/Xserver/hw/xfree86/drivers/imstt/imstt_reg.h	2003-04-08 22:10:30.000000000 +0000
@@ -9,22 +9,11 @@
 
 static __inline__ void regw(unsigned long base_addr, unsigned long regindex, unsigned long regdata)
 {
-	asm volatile ("stwbrx %1,%2,%3; eieio"
-		      : "=m" (*(volatile unsigned *)(base_addr+regindex))
-		      : "r" (regdata), "b" (regindex), "r" (base_addr));
 }
 
 
 static __inline__ unsigned long regr(unsigned long base_addr, unsigned long regindex)
 {
-	register unsigned long val;
-
-	asm volatile ("lwbrx %0,%1,%2; eieio"
-		      : "=r" (val)
-		      : "b" (regindex), "r" (base_addr),
-		      "m" (*(volatile unsigned *)(base_addr+regindex)));
-
-	return (val);
 }
 
 
