

Enable the nvidia driver.

  - Rene Rebe <rene@rocklinux.org>

--- xc/config/cf/xfree86.cf.orig	2003-08-24 12:09:33.000000000 +0000
+++ xc/config/cf/xfree86.cf	2003-09-14 23:34:27.000000000 +0000
@@ -928,7 +928,7 @@
 
 /* Pure PCI drivers should go first */
 # ifndef XF86CardDrivers
-#  define XF86CardDrivers	ati mga glint s3 s3virge sis savage\
+#  define XF86CardDrivers	ati mga glint nv s3 s3virge sis savage\
 				trident chips tdfx fbdev \
 				DevelDrivers vga \
 				XF86OSCardDrivers XF86ExtraCardDrivers
