--- linux-5.12/drivers/ps3/ps3stor_lib.c.vanilla	2021-05-12 18:23:04.051975860 +0200
+++ linux-5.12/drivers/ps3/ps3stor_lib.c	2021-05-12 18:29:08.923961312 +0200
@@ -135,6 +135,7 @@
 {
 	int error, res, alignment;
 	enum ps3_dma_page_size page_size;
+	static u64 dummy_mask = DMA_BIT_MASK(32);
 
 	error = ps3stor_open_hv_device(&dev->sbd);
 	if (error) {
@@ -183,7 +184,8 @@
 		error = -ENOMEM;
 		goto fail_free_irq;
 	}
-
+	if (!dev->sbd.core.dma_mask)
+		dev->sbd.core.dma_mask = &dummy_mask;
 	dev->bounce_lpar = ps3_mm_phys_to_lpar(__pa(dev->bounce_buf));
 	dev->bounce_dma = dma_map_single(&dev->sbd.core, dev->bounce_buf,
 					 dev->bounce_size, DMA_BIDIRECTIONAL);
