--- a/drivers/ps3/ps3-sys-manager.c	2012-01-27 19:35:54.017195402 +0100
+++ b/drivers/ps3/ps3-sys-manager.c	2012-01-27 19:36:18.990903099 +0100
@@ -656,7 +656,7 @@
 	ps3_vuart_cancel_async(dev);
 
 	ps3_sys_manager_send_attr(dev, 0);
-	ps3_sys_manager_send_next_op(dev, PS3_SM_NEXT_OP_SYS_REBOOT,
+	ps3_sys_manager_send_next_op(dev, PS3_SM_NEXT_OP_LPAR_REBOOT,
 		user_wake_sources);
 
 	ps3_sys_manager_fin(dev);
