--- a/arch/powerpc/include/asm/lv1call.h	2012-02-03 21:39:51.000000000 +0100
+++ b/arch/powerpc/include/asm/lv1call.h	2012-02-05 17:34:26.973058855 +0100
@@ -270,6 +270,8 @@
 LV1_CALL(read_htab_entries,                             2, 5,  95 )
 LV1_CALL(set_dabr,                                      2, 0,  96 )
 LV1_CALL(get_total_execution_time,                      2, 1, 103 )
+LV1_CALL(undocumented_function_114,                     3, 1, 114 )
+LV1_CALL(undocumented_function_115,                     1, 0, 115 )
 LV1_CALL(allocate_io_segment,                           3, 1, 116 )
 LV1_CALL(release_io_segment,                            2, 0, 117 )
 LV1_CALL(construct_io_irq_outlet,                       1, 1, 120 )
