--- a/arch/powerpc/include/asm/lv1call.h	2012-02-07 18:22:20.218981765 +0100
+++ b/arch/powerpc/include/asm/lv1call.h	2012-02-08 19:44:40.636013978 +0100
@@ -327,6 +327,10 @@
 LV1_CALL(storage_write,                                 6, 1, 246 )
 LV1_CALL(storage_send_device_command,                   6, 1, 248 )
 LV1_CALL(storage_get_async_status,                      1, 2, 249 )
+LV1_CALL(storage_create_region,                         5, 2, 250 )
+LV1_CALL(storage_delete_region,                         2, 1, 251 )
+LV1_CALL(storage_set_region_acl,                        4, 1, 252 )
+LV1_CALL(storage_get_region_acl,                        3, 2, 253 )
 LV1_CALL(storage_check_async_status,                    2, 1, 254 )
 LV1_CALL(panic,                                         1, 0, 255 )
 LV1_CALL(construct_lpm,                                 6, 3, 140 )
