--- a/drivers/net/ethernet/toshiba/ps3_gelic_wireless.c	2012-01-26 01:39:32.000000000 +0100
+++ b/drivers/net/ethernet/toshiba/ps3_gelic_wireless.c	2012-02-03 15:42:14.996356826 +0100
@@ -183,7 +183,7 @@
 				      &cmd->tag, &cmd->size);
 	if (cmd->status) {
 		complete(&cmd->done);
-		pr_info("%s: cmd issue failed\n", __func__);
+		pr_info("%s: cmd (%d) issue failed (%d)\n", __func__, cmd->cmd, cmd->status);
 		return;
 	}
 
