--- a/arch/powerpc/platforms/ps3/interrupt.c	2012-06-22 20:37:50.000000000 +0200
+++ b/arch/powerpc/platforms/ps3/interrupt.c	2012-07-15 11:54:14.122392086 +0200
@@ -608,6 +608,7 @@
 
 	return result;
 }
+EXPORT_SYMBOL_GPL(ps3_spe_irq_setup);
 
 int ps3_spe_irq_destroy(unsigned int virq)
 {
@@ -620,6 +621,7 @@
 
 	return result;
 }
+EXPORT_SYMBOL_GPL(ps3_spe_irq_destroy);
 
 
 #define PS3_INVALID_OUTLET ((irq_hw_number_t)-1)
