--- a/arch/powerpc/include/asm/lv1call.h	2013-02-21 14:04:37.204534305 +0100
+++ b/arch/powerpc/include/asm/lv1call.h	2013-02-21 14:12:07.299635321 +0100
@@ -274,6 +274,11 @@
 LV1_CALL(set_dabr,                                      2, 0,  96 )
 LV1_CALL(undocumented_function_99,                      2, 0,  99 )
 LV1_CALL(get_total_execution_time,                      2, 1, 103 )
+LV1_CALL(undocumented_function_105,                     7, 0, 105 )
+LV1_CALL(undocumented_function_106,                     1, 0, 106 )
+LV1_CALL(undocumented_function_107,                     6, 1, 107 )
+LV1_CALL(undocumented_function_108,                     1, 5, 108 )
+LV1_CALL(undocumented_function_109,                     1, 0, 109 )
 LV1_CALL(undocumented_function_114,                     3, 1, 114 )
 LV1_CALL(undocumented_function_115,                     1, 0, 115 )
 LV1_CALL(allocate_io_segment,                           3, 1, 116 )
