--- ./mkdosfs.c.orig	2004-02-10 22:44:38.602762200 +0000
+++ ./mkdosfs.c	2004-02-10 22:43:57.846958024 +0000
@@ -43,7 +43,7 @@
 
 #include <fcntl.h>
 #include <linux/hdreg.h>
-#include <linux/fs.h>
+//#include <linux/fs.h>
 #include <sys/stat.h>
 #include <linux/fd.h>
 #include <mntent.h>
@@ -61,6 +61,9 @@
 
 /* Constant definitions */
 
+#define BLOCK_SIZE_BITS 10
+#define BLOCK_SIZE (1<<BLOCK_SIZE_BITS)
+
 #define TRUE 1			/* Boolean constants */
 #define FALSE 0
 
