--- ./Makefile.orig	2003-09-23 22:19:55.000000000 -0400
+++ ./Makefile	2003-09-23 22:28:29.000000000 -0400
@@ -518,7 +518,9 @@
 ifdef CONFIG_MODVERSIONS
 	$(MAKE) update-modverfile
 endif
-	scripts/mkdep -- `find $(FINDHPATH) \( -name SCCS -o -name .svn \) -prune -o -follow -name \*.h ! -name modversions.h -print` > .hdepend
+	for x in $(FINDHPATH); do \
+		scripts/mkdep -- `find $$x \( -name SCCS -o -name .svn \) -prune -o -follow -name \*.h ! -name modversions.h -print` >> .hdepend ; \
+	done
 	scripts/mkdep -- init/*.c > .depend
 
 ifdef CONFIG_MODVERSIONS
