--- linux-2.4.21/drivers/char/agp/agpgart_be.c.orig	2003-06-13 16:51:32.000000000 +0200
+++ linux-2.4.21/drivers/char/agp/agpgart_be.c	2003-08-26 03:20:43.000000000 +0200
@@ -71,7 +71,7 @@
 {
 #if defined(__i386__) || defined(__x86_64__)
 	asm volatile ("wbinvd":::"memory");
-#elif defined(__alpha__) || defined(__ia64__) || defined(__sparc__)
+#elif defined(__alpha__) || defined(__ia64__) || defined(__sparc__) || defined(__powerpc__)
 	/* ??? I wonder if we'll really need to flush caches, or if the
 	   core logic can manage to keep the system coherent.  The ARM
 	   speaks only of using `cflush' to get things in memory in
