--- ./drivers/atm/horizon.c.orig	2003-09-04 20:22:19.000000000 +0200
+++ ./drivers/atm/horizon.c	2003-09-04 20:24:36.000000000 +0200
@@ -481,7 +481,7 @@
   return;
 }
 
-static inline void dump_regs (hrz_dev * dev) {
+static inline void dump_horizon_regs (hrz_dev * dev) {
 #ifdef DEBUG_HORIZON
   PRINTD (DBG_REGS, "CONTROL 0: %#x", rd_regl (dev, CONTROL_0_REG));
   PRINTD (DBG_REGS, "RX CONFIG: %#x", rd_regw (dev, RX_CONFIG_OFF));
