--- ./configure.orig	2005-05-28 17:29:07.481582808 -0400
+++ ./configure	2005-05-28 17:33:38.487479752 -0400
@@ -20572,23 +20572,23 @@
   else
      PKG_CONFIG_MIN_VERSION=0.9.0
      if $PKG_CONFIG --atleast-pkgconfig-version $PKG_CONFIG_MIN_VERSION; then
-        echo "$as_me:$LINENO: checking for gedit-2.8 >= 2.8.0" >&5
-echo $ECHO_N "checking for gedit-2.8 >= 2.8.0... $ECHO_C" >&6
+        echo "$as_me:$LINENO: checking for gedit-2.10 >= 2.10.0" >&5
+echo $ECHO_N "checking for gedit-2.10 >= 2.10.0... $ECHO_C" >&6
 
-        if $PKG_CONFIG --exists "gedit-2.8 >= 2.8.0" ; then
+        if $PKG_CONFIG --exists "gedit-2.10 >= 2.10.0" ; then
             echo "$as_me:$LINENO: result: yes" >&5
 echo "${ECHO_T}yes" >&6
             succeeded=yes
 
             echo "$as_me:$LINENO: checking GEDIT_CFLAGS" >&5
 echo $ECHO_N "checking GEDIT_CFLAGS... $ECHO_C" >&6
-            GEDIT_CFLAGS=`$PKG_CONFIG --cflags "gedit-2.8 >= 2.8.0"`
+            GEDIT_CFLAGS=`$PKG_CONFIG --cflags "gedit-2.10 >= 2.10.0"`
             echo "$as_me:$LINENO: result: $GEDIT_CFLAGS" >&5
 echo "${ECHO_T}$GEDIT_CFLAGS" >&6
 
             echo "$as_me:$LINENO: checking GEDIT_LIBS" >&5
 echo $ECHO_N "checking GEDIT_LIBS... $ECHO_C" >&6
-            GEDIT_LIBS=`$PKG_CONFIG --libs "gedit-2.8 >= 2.8.0"`
+            GEDIT_LIBS=`$PKG_CONFIG --libs "gedit-2.10 >= 2.10.0"`
             echo "$as_me:$LINENO: result: $GEDIT_LIBS" >&5
 echo "${ECHO_T}$GEDIT_LIBS" >&6
         else
@@ -20596,7 +20596,7 @@
             GEDIT_LIBS=""
             ## If we have a custom action on failure, don't print errors, but
             ## do set a variable so people can do so.
-            GEDIT_PKG_ERRORS=`$PKG_CONFIG --errors-to-stdout --print-errors "gedit-2.8 >= 2.8.0"`
+            GEDIT_PKG_ERRORS=`$PKG_CONFIG --errors-to-stdout --print-errors "gedit-2.10 >= 2.10.0"`
 
         fi
 
@@ -20665,23 +20665,23 @@
   else
      PKG_CONFIG_MIN_VERSION=0.9.0
      if $PKG_CONFIG --atleast-pkgconfig-version $PKG_CONFIG_MIN_VERSION; then
-        echo "$as_me:$LINENO: checking for gedit-2.6 >= 2.6.0" >&5
-echo $ECHO_N "checking for gedit-2.6 >= 2.6.0... $ECHO_C" >&6
+        echo "$as_me:$LINENO: checking for gedit-2.8 >= 2.8.0" >&5
+echo $ECHO_N "checking for gedit-2.8 >= 2.8.0... $ECHO_C" >&6
 
-        if $PKG_CONFIG --exists "gedit-2.6 >= 2.6.0" ; then
+        if $PKG_CONFIG --exists "gedit-2.8 >= 2.8.0" ; then
             echo "$as_me:$LINENO: result: yes" >&5
 echo "${ECHO_T}yes" >&6
             succeeded=yes
 
             echo "$as_me:$LINENO: checking GEDIT_CFLAGS" >&5
 echo $ECHO_N "checking GEDIT_CFLAGS... $ECHO_C" >&6
-            GEDIT_CFLAGS=`$PKG_CONFIG --cflags "gedit-2.6 >= 2.6.0"`
+            GEDIT_CFLAGS=`$PKG_CONFIG --cflags "gedit-2.8 >= 2.8.0"`
             echo "$as_me:$LINENO: result: $GEDIT_CFLAGS" >&5
 echo "${ECHO_T}$GEDIT_CFLAGS" >&6
 
             echo "$as_me:$LINENO: checking GEDIT_LIBS" >&5
 echo $ECHO_N "checking GEDIT_LIBS... $ECHO_C" >&6
-            GEDIT_LIBS=`$PKG_CONFIG --libs "gedit-2.6 >= 2.6.0"`
+            GEDIT_LIBS=`$PKG_CONFIG --libs "gedit-2.8 >= 2.8.0"`
             echo "$as_me:$LINENO: result: $GEDIT_LIBS" >&5
 echo "${ECHO_T}$GEDIT_LIBS" >&6
         else
@@ -20689,7 +20689,7 @@
             GEDIT_LIBS=""
             ## If we have a custom action on failure, don't print errors, but
             ## do set a variable so people can do so.
-            GEDIT_PKG_ERRORS=`$PKG_CONFIG --errors-to-stdout --print-errors "gedit-2.6 >= 2.6.0"`
+            GEDIT_PKG_ERRORS=`$PKG_CONFIG --errors-to-stdout --print-errors "gedit-2.8 >= 2.8.0"`
             echo $GEDIT_PKG_ERRORS
         fi
 
@@ -20704,8 +20704,8 @@
   if test $succeeded = yes; then
      :
   else
-     { { echo "$as_me:$LINENO: error: Library requirements (gedit-2.6 >= 2.6.0) not met; consider adjusting the PKG_CONFIG_PATH environment variable if your libraries are in a nonstandard prefix so pkg-config can find them." >&5
-echo "$as_me: error: Library requirements (gedit-2.6 >= 2.6.0) not met; consider adjusting the PKG_CONFIG_PATH environment variable if your libraries are in a nonstandard prefix so pkg-config can find them." >&2;}
+     { { echo "$as_me:$LINENO: error: Library requirements (gedit-2.8 >= 2.8.0) not met; consider adjusting the PKG_CONFIG_PATH environment variable if your libraries are in a nonstandard prefix so pkg-config can find them." >&5
+echo "$as_me: error: Library requirements (gedit-2.8 >= 2.8.0) not met; consider adjusting the PKG_CONFIG_PATH environment variable if your libraries are in a nonstandard prefix so pkg-config can find them." >&2;}
    { (exit 1); exit 1; }; }
   fi
 
