
Fixed compilation with gcc >= 3.4.

  - Rene Rebe <rene@exactcode.de>

--- flite-1.2-release/include/cst_sts.h.vanilla	2005-03-23 16:10:16.000000000 +0100
+++ flite-1.2-release/include/cst_sts.h	2005-03-23 16:10:23.000000000 +0100
@@ -47,9 +47,9 @@
 /* else where, this information plus the indexes in the Unit relation */
 /* allow reconstruction of the signal itself                          */
 struct cst_sts_struct {
-    const unsigned short *frame;  
-    const int size;      /* in samples */
-    const unsigned char *residual;
+    unsigned short *frame;  
+    int size;      /* in samples */
+    unsigned char *residual;
 };
 typedef struct cst_sts_struct cst_sts;
 
