--- ./Makefile.orig	2006-07-07 21:40:12.000000000 +0000
+++ ./Makefile	2006-07-07 21:40:24.000000000 +0000
@@ -5,7 +5,6 @@
 DEFINES= 
 
 #options if you have a bind>=4.9.4 libresolv (or, maybe, glibc)
-LDLIBS=-lresolv
 ADDLIB=
 
 ifeq ($(LIBC_INCLUDE)/socketbits.h,$(wildcard $(LIBC_INCLUDE)/socketbits.h))
--- ./ping.c.orig	2006-07-07 21:56:01.000000000 +0000
+++ ./ping.c	2006-07-07 21:56:20.000000000 +0000
@@ -61,7 +61,6 @@
 #include "ping_common.h"
 
 #include <netinet/ip.h>
-#include <netinet/ip_icmp.h>
 
 
 #define	MAXIPLEN	60
--- ./ping_common.c.orig	2002-09-20 16:02:32.000000000 +0000
+++ ./ping_common.c	2006-07-07 21:58:57.000000000 +0000
@@ -1,6 +1,7 @@
 #include "ping_common.h"
 #include <ctype.h>
 #include <sched.h>
+#include <termios.h>
 
 int options;
 
--- ./ping_common.h.orig	2006-07-07 21:48:15.000000000 +0000
+++ ./ping_common.h	2006-07-07 22:06:28.000000000 +0000
@@ -9,6 +9,8 @@
 #include <sys/time.h>
 #include <sys/signal.h>
 #include <sys/ioctl.h>
+#include <sys/types.h>
+#include <linux/icmp.h>
 #include <net/if.h>
 #include <sys/uio.h>
 #include <sys/poll.h>
@@ -37,6 +38,8 @@
 #define	CLR(bit)	(A(bit) &= (~B(bit)))
 #define	TST(bit)	(A(bit) & B(bit))
 
+#define __constant_htons(N)	htons(N)
+
 /* various options */
 extern int options;
 #define	F_FLOOD		0x001
--- ./tracepath6.c.orig	2006-07-07 22:08:22.000000000 +0000
+++ ./tracepath6.c	2006-07-07 22:08:48.000000000 +0000
@@ -15,7 +15,6 @@
 #include <sys/socket.h>
 #include <netinet/in.h>
 
-#include <linux/in6.h>
 #include <linux/errqueue.h>
 #include <errno.h>
 #include <string.h>
--- ./traceroute6.c.orig	2006-07-07 22:09:53.000000000 +0000
+++ ./traceroute6.c	2006-07-07 22:10:30.000000000 +0000
@@ -247,7 +247,4 @@
 #include <netinet/udp.h>
-
-#include <linux/ipv6.h>
-#include <linux/in6.h>
 
 #include <linux/icmpv6.h>
 
--- ./ping6.c.orig	2006-07-07 22:12:25.000000000 +0000
+++ ./ping6.c	2006-07-07 22:12:44.000000000 +0000
@@ -69,7 +69,5 @@
 #include "ping_common.h"
 
-#include <linux/in6.h>
-#include <linux/ipv6.h>
 #include <linux/icmpv6.h>
 
 #define BIT_CLEAR(nr, addr) do { ((__u32 *)(addr))[(nr) >> 5] &= ~(1U << ((nr) & 31)); } while(0)
