--- ./include/misc.h.orig	2006-06-20 20:36:27.000000000 +0200
+++ ./include/misc.h	2006-06-20 20:36:48.000000000 +0200
@@ -105,7 +105,7 @@
 __u64 mask64 (int from, int count);
 
 int reiserfs_bin_search (void * key, void * base, __u32 num, int width,
-			 __u32 *ppos, __compar_fn_t comp_func);
+			 __u32 *ppos, int (*comp_func)(const void *, const void *));
 
 struct block_handler {
     __u32 blocknr;
--- ./lib/misc.c.orig	2006-06-20 20:40:00.000000000 +0200
+++ ./lib/misc.c	2006-06-20 20:42:07.000000000 +0200
@@ -604,7 +604,7 @@
 /* this implements binary search in the array 'base' among 'num' elements each
    of those is 'width' bytes long. 'comp_func' is used to compare keys */
 int reiserfs_bin_search (void * key, void * base, __u32 num, int width,
-			 __u32 * ppos, comparison_fn_t comp_func)
+			 __u32 * ppos, int (*comp_func)(const void *, const void *))
 {
     __u32 rbound, lbound, j;
     int ret;
