--- ./lib/metadata/metadata.c.orig	2006-06-21 18:03:02.000000000 +0200
+++ ./lib/metadata/metadata.c	2006-06-21 18:10:12.000000000 +0200
@@ -24,6 +24,9 @@
 #include "pv_alloc.h"
 #include "activate.h"
 
+#include <stdint.h> 
+#define UINT32_MAX         (4294967295U)
+
 static int _add_pv_to_vg(struct format_instance *fid, struct volume_group *vg,
 			 const char *pv_name)
 {
--- ./lib/misc/lib.h.orig	2006-07-07 20:17:21.000000000 +0000
+++ ./lib/misc/lib.h	2006-07-07 20:17:48.000000000 +0000
@@ -28,5 +28,6 @@
 #include "lvm-types.h"
 
 #include <libdevmapper.h>
+#include <stdint.h>
 
 #endif
