--- ./src/exec.c.orig	2006-07-07 23:21:42.000000000 +0000
+++ ./src/exec.c	2006-07-07 23:24:08.000000000 +0000
@@ -47,6 +47,9 @@
 
 #include <sys/types.h>
 #include <sys/stat.h>
+#ifdef __NO_STAT64
+#	define stat64	stat
+#endif
 #include <unistd.h>
 #include <fcntl.h>
 #include <stdlib.h>
--- ./src/expand.c.orig	2006-07-07 23:29:07.000000000 +0000
+++ ./src/expand.c	2006-07-07 23:30:31.000000000 +0000
@@ -48,6 +48,10 @@
 #include <sys/types.h>
 #include <sys/time.h>
 #include <sys/stat.h>
+#ifdef __NO_STAT64
+#	define stat64	stat
+#	define lstat64	lstat
+#endif
 #include <dirent.h>
 #include <unistd.h>
 #include <pwd.h>
--- ./src/mail.c.orig	2006-07-07 23:33:17.000000000 +0000
+++ ./src/mail.c	2006-07-07 23:33:51.000000000 +0000
@@ -50,6 +50,9 @@
  */
 #include <sys/types.h>
 #include <sys/stat.h>
+#ifdef __NO_STAT64
+#	define stat64	stat
+#endif
 #include <stdlib.h>
 
 #include "shell.h"
--- ./src/redir.c.orig	2006-07-07 23:35:14.000000000 +0000
+++ ./src/redir.c	2006-07-07 23:35:56.000000000 +0000
@@ -46,6 +46,10 @@
 #endif /* not lint */
 
 #include <sys/stat.h>
+#ifdef __NO_STAT64
+#	define stat64	stat
+#	define fstat64	fstat
+#endif
 #include <sys/types.h>
 #include <sys/param.h>	/* PIPE_BUF */
 #include <signal.h>
--- ./src/bltin/test.c.orig	2006-07-07 23:38:24.000000000 +0000
+++ ./src/bltin/test.c	2006-07-07 23:38:54.000000000 +0000
@@ -16,6 +16,10 @@
 #endif
 
 #include <sys/stat.h>
+#ifdef __NO_STAT64
+#	define stat64	stat
+#	define lstat64	lstat
+#endif
 #include <sys/types.h>
 
 #include <ctype.h>
