--- SDL-1.2.6/configure	2003-08-30 21:13:23.000000000 +0200
+++ SDL-1.2.6-rock/configure	2004-02-23 18:47:28.000000000 +0100
@@ -3595,6 +3595,7 @@
         else
             ARTSC_CFLAGS=`$ARTSCCONFIG --cflags`
             ARTSC_LIBS=`$ARTSCCONFIG --libs`
+	    ARTSC_PREFIX=`$ARTSCCONFIG --arts-prefix`
             echo $ac_n "checking for aRts development environment""... $ac_c" 1>&6
 echo "configure:3600: checking for aRts development environment" >&5
             audio_arts=no
@@ -3633,7 +3634,7 @@
   enable_arts_shared=yes
 fi
 
-                arts_lib_spec=`echo $ARTSC_LIBS | sed 's/.*-L\([^ ]*\).*/\1\/libartsc.so.*/'`
+                arts_lib_spec="$ARTSC_PREFIX/lib/libartsc.so.*"
                 arts_lib=`ls $arts_lib_spec | head -1 | sed 's/.*\/\(.*\)/\1/'`
                 echo "-- $arts_lib_spec -> $arts_lib"
                 if test x$enable_dlopen != xyes && \
