--- gtk+-2.2.4/gtk/gtkfontsel.c.orig	2003-10-07 01:54:42.000000000 +0200
+++ gtk+-2.2.4/gtk/gtkfontsel.c	2003-10-07 01:55:07.000000000 +0200
@@ -1007,6 +1007,8 @@
   if (fontsel->face)
     {
       font_desc = pango_font_face_describe (fontsel->face);
+      pango_font_description_set_family(font_desc,
+	  pango_font_family_get_name (fontsel->family));
       pango_font_description_set_size (font_desc, fontsel->size);
     }
   else
