--- uClibc/libc/misc/error/error.c.orig	2009-02-19 15:38:22.000000000 +0100
+++ uClibc/libc/misc/error/error.c	2009-02-19 15:39:16.000000000 +0100
@@ -44,7 +44,7 @@
 /* If NULL, error will flush stdout, then print on stderr the program
    name, a colon and a space.  Otherwise, error will call this
    function without parameters instead.  */
-/* void (*error_print_progname) (void) = NULL; */
+void (*error_print_progname) (void) = NULL;
 
 extern __typeof(error) __error attribute_hidden;
 void __error (int status, int errnum, const char *message, ...)
