--- openMotif-2.2.2-old/tools/wml/Makefile.in	Thu Feb 14 08:15:49 2002
+++ openmotif-2.2.2/tools/wml/Makefile.in	Fri Aug 16 16:49:20 2002
@@ -121,7 +121,7 @@
 
 wmldbcreate_LDADD = ../../lib/Xm/libXm.la
 
-INCLUDES = ${X_CFLAGS} -I$(srcdir)/../../lib -I$(srcdir)/../../include
+INCLUDES = -I$(srcdir)/../../lib -I$(srcdir)/../../include ${X_CFLAGS}
 
 noinst_LIBRARIES = libwml.a
 libwml_a_SOURCES = wmlparse.y wml.c wmloutkey.c wmlouth.c wmloutmm.c wmloutp1.c wmlresolve.c wmlsynbld.c wmlutils.c
