--- ./configure	2003-02-04 20:58:52.000000000 +0100
+++ ./configure.new	2003-03-28 22:50:33.000000000 +0100
@@ -20496,7 +20496,7 @@
   # Remove --cache-file and --srcdir arguments so they do not pile up.
   ac_sub_configure_args=
   ac_prev=
-  for ac_arg in $ac_configure_args; do
+  for ac_arg in `echo $ac_configure_args | egrep -v "CC=|CXX="`; do
     if test -n "$ac_prev"; then
       ac_prev=
       continue
