--- avifile-0.7-0.7.38/configure.orig	2004-03-27 11:10:46.000000000 +0100
+++ avifile-0.7-0.7.38/configure	2004-03-27 11:11:38.000000000 +0100
@@ -5364,7 +5364,6 @@
 #endif
 
 #ifdef __cplusplus
-extern "C" void exit (int);
 #endif
 
 void fnord() { int i=42;}
@@ -5458,7 +5457,6 @@
 #endif
 
 #ifdef __cplusplus
-extern "C" void exit (int);
 #endif
 
 void fnord() { int i=42;}
@@ -10230,7 +10228,6 @@
 #line 10231 "configure"
 #include "confdefs.h"
 #ifdef __cplusplus
-extern "C" void exit(int);
 #endif
 
 	    /*#include <qapplication.h>*/
