--- comedi-0.7.74/comedi/Makefile.am	2007-08-01 22:26:30.000000000 +0300
+++ comedi-0.7.74/comedi/Makefile.orig	2007-09-30 17:42:11.000000000 +0300
@@ -12,7 +12,7 @@
 #the kernel's build system doesn't get lost.
 	if test "$(srcdir)" != "."; then \
 		for dir in . drivers kcomedilib drivers/addi-data; do \
-			for file in `ls @abs_srcdir@/$$dir/*\.[ch] @abs_srcdir@/$$dir/Kbuild | $(GREP) -v \.mod\.c`; do \
+			for file in `ls @abs_srcdir@/$$dir/*\.[ch] @abs_srcdir@/$$dir/Kbuild | $(EGREP) -v \.mod\.c`; do \
 				LINK_NAME=$$dir/`basename "$$file"`; \
 				if test ! -e $$LINK_NAME; then ln -vs $$file $$LINK_NAME; fi; \
 			done; \
@@ -26,12 +26,12 @@
 install-data-local:
 	$(MAKE) -I@abs_builddir@ -C $(LINUX_DIR) M=@abs_builddir@ CC="$(LINUX_CC) -I@abs_top_builddir@/ \
 		-I@abs_top_srcdir@/include $(RTAI_CFLAGS) $(RTLINUX_CFLAGS)" INSTALL_MOD_DIR=$(COMEDI_MOD_DIR) \
-		INSTALL_MOD_PATH="$(prefix)" modules_install
+		INSTALL_MOD_PATH="$(DESTDIR)$(prefix)" modules_install
 clean-local:
 	$(MAKE) -I@abs_builddir@ -C $(LINUX_DIR) M=@abs_builddir@ CC="$(LINUX_CC) -I@abs_top_builddir@/ \
 		-I@abs_top_srcdir@/include $(RTAI_CFLAGS) $(RTLINUX_CFLAGS)" clean
 uninstall-local:
-	find $(prefix)/$(LINUX_MODLIB)/$(COMEDI_MOD_DIR) -name '*.ko' -exec $(RM) {} \;
+	find $(DESTDIR)$(prefix)/$(LINUX_MODLIB)/$(COMEDI_MOD_DIR) -name '*.ko' -exec $(RM) {} \;
 else
 module_PROGRAMS = comedi.ko
 all-local:
