--- comedi-0.7.69/comedi/drivers/cb_pcimdda.c~	2003-04-28 19:27:19.000000000 +0200
+++ comedi-0.7.69/comedi/drivers/cb_pcimdda.c	2005-01-26 15:42:03.008293000 +0100
@@ -209,7 +209,17 @@
 -----------------------------------------------------------------------------*/
 
 /* returns a maxdata value for a given n_bits */
-static inline lsampl_t figure_out_maxdata(int bits);
+static inline lsampl_t figure_out_maxdata(int bits)
+{
+    lsampl_t max = 0;
+    int i;
+
+    for (i = 0; i < bits; i++) {
+      max <<= 1;
+      max |=  1U;
+    }
+    return max;
+}
 
 /* 
  *  Probes for a supported device.
@@ -486,18 +496,6 @@
 }
 
 
-/* returns a maxdata value for a given n_bits */
-static inline lsampl_t figure_out_maxdata(int bits)
-{
-    lsampl_t max = 0;
-    int i;
-
-    for (i = 0; i < bits; i++) {
-      max <<= 1;
-      max |=  1U;
-    }
-    return max;
-}
 
 
 
