--- jhead-2.5/exif.c.orig	2006-03-10 10:34:51.000000000 -0800
+++ jhead-2.5/exif.c	2006-03-10 10:37:13.000000000 -0800
@@ -305,6 +305,7 @@
 void PrintFormatNumber(void * ValuePtr, int Format, int ByteCount)
 {
     int s,n;
+    char * vp = (char *)ValuePtr;
 
     for(n=0;n<16;n++){
         switch(Format){
@@ -329,7 +330,7 @@
         ByteCount -= s;
         if (ByteCount <= 0) break;
         printf(", ");
-        ((char *)ValuePtr) += s;
+        vp += s;
     }
     if (n >= 16) printf("...");
 }
