--- linux-2.6.11/include/linux/oprofile.h.orig	2005-03-02 08:38:10.000000000 +0100
+++ linux-2.6.11/include/linux/oprofile.h	2005-04-02 23:00:31.825217856 +0200
@@ -48,7 +48,7 @@
  *
  * If an error occurs, the fields should be left untouched.
  */
-int oprofile_arch_init(struct oprofile_operations * ops);
+int oprofile_arch_init(struct oprofile_operations ** ops);
  
 /**
  * One-time exit/cleanup for the arch.
