diff -u fribidi-0.10.4/fribidi.pc.in fribidi-0.10.4-fixed/fribidi.pc.in
--- fribidi-0.10.4/fribidi.pc.in	2002-01-03 16:31:24.000000000 +0100
+++ fribidi-0.10.4-fixed/fribidi.pc.in	2003-05-18 22:29:08.000000000 +0200
@@ -7,4 +7,4 @@
 Description: Unicode BiDirectional algorithm library
 Version: @VERSION@
 Libs: -L${libdir} -lfribidi
-Cflags: -I${includedir}/fribidi
+Cflags: -I${includedir} -I${includedir}/fribidi
Common subdirectories: fribidi-0.10.4/tests and fribidi-0.10.4-fixed/tests
Common subdirectories: fribidi-0.10.4/unidata and fribidi-0.10.4-fixed/unidata
