diff -urN openldap-2.1.28/build/openldap.m4 openldap-2.1.28-jsaw/build/openldap.m4
--- openldap-2.1.28/build/openldap.m4	2003-12-02 19:05:49.000000000 +0100
+++ openldap-2.1.28-jsaw/build/openldap.m4	2004-03-29 00:57:32.000000000 +0200
@@ -1204,7 +1204,7 @@
 [AC_CACHE_CHECK([Cyrus SASL library version], [ol_cv_sasl_compat],[
 	AC_EGREP_CPP(__sasl_compat,[
 #ifdef HAVE_SASL_SASL_H
-#include <sasl/sasl.h>
+#include <sasl2/sasl.h>
 #else
 #include <sasl.h>
 #endif
diff -urN openldap-2.1.28/configure openldap-2.1.28-jsaw/configure
--- openldap-2.1.28/configure	2004-01-18 22:25:00.000000000 +0100
+++ openldap-2.1.28-jsaw/configure	2004-03-29 00:57:32.000000000 +0200
@@ -20834,7 +20834,7 @@
 ol_link_sasl=no
 ol_link_spasswd=no
 if test $ol_with_cyrus_sasl != no ; then
-	for ac_hdr in sasl/sasl.h sasl.h
+	for ac_hdr in sasl2/sasl.h sasl.h
 do
 ac_safe=`echo "$ac_hdr" | sed 'y%./+-%__p_%'`
 echo $ac_n "checking for $ac_hdr""... $ac_c" 1>&6
@@ -20980,7 +20980,7 @@
 #include "confdefs.h"
 
 #ifdef HAVE_SASL_SASL_H
-#include <sasl/sasl.h>
+#include <sasl2/sasl.h>
 #else
 #include <sasl.h>
 #endif
diff -urN openldap-2.1.28/configure.in openldap-2.1.28-jsaw/configure.in
--- openldap-2.1.28/configure.in	2004-01-18 22:25:26.000000000 +0100
+++ openldap-2.1.28-jsaw/configure.in	2004-03-29 00:57:32.000000000 +0200
@@ -2129,7 +2129,7 @@
 ol_link_sasl=no
 ol_link_spasswd=no
 if test $ol_with_cyrus_sasl != no ; then
-	AC_CHECK_HEADERS(sasl/sasl.h sasl.h)
+	AC_CHECK_HEADERS(sasl2/sasl.h sasl.h)
 
 	if test $ac_cv_header_sasl_sasl_h = yes -o $ac_cv_header_sasl_h = yes; then
 		AC_CHECK_LIB(sasl2, sasl_client_init,
diff -urN openldap-2.1.28/include/portable.h.in openldap-2.1.28-jsaw/include/portable.h.in
--- openldap-2.1.28/include/portable.h.in	2003-03-10 03:36:10.000000000 +0100
+++ openldap-2.1.28-jsaw/include/portable.h.in	2004-03-29 00:57:32.000000000 +0200
@@ -476,7 +476,7 @@
 /* Define if you have the <sasl.h> header file.  */
 #undef HAVE_SASL_H
 
-/* Define if you have the <sasl/sasl.h> header file.  */
+/* Define if you have the <sasl2/sasl.h> header file.  */
 #undef HAVE_SASL_SASL_H
 
 /* Define if you have the <sched.h> header file.  */
diff -urN openldap-2.1.28/libraries/libldap/cyrus.c openldap-2.1.28-jsaw/libraries/libldap/cyrus.c
--- openldap-2.1.28/libraries/libldap/cyrus.c	2004-02-12 03:41:19.000000000 +0100
+++ openldap-2.1.28-jsaw/libraries/libldap/cyrus.c	2004-03-29 00:57:32.000000000 +0200
@@ -24,7 +24,7 @@
 #endif
 
 #ifdef HAVE_SASL_SASL_H
-#include <sasl/sasl.h>
+#include <sasl2/sasl.h>
 #else
 #include <sasl.h>
 #endif
diff -urN openldap-2.1.28/libraries/libldap/ldap-int.h openldap-2.1.28-jsaw/libraries/libldap/ldap-int.h
--- openldap-2.1.28/libraries/libldap/ldap-int.h	2003-04-29 01:41:55.000000000 +0200
+++ openldap-2.1.28-jsaw/libraries/libldap/ldap-int.h	2004-03-29 00:57:32.000000000 +0200
@@ -25,7 +25,7 @@
 #ifdef HAVE_CYRUS_SASL
 	/* the need for this should be removed */
 #ifdef HAVE_SASL_SASL_H
-#include <sasl/sasl.h>
+#include <sasl2/sasl.h>
 #else
 #include <sasl.h>
 #endif
diff -urN openldap-2.1.28/libraries/liblutil/passwd.c openldap-2.1.28-jsaw/libraries/liblutil/passwd.c
--- openldap-2.1.28/libraries/liblutil/passwd.c	2003-12-18 03:36:40.000000000 +0100
+++ openldap-2.1.28-jsaw/libraries/liblutil/passwd.c	2004-03-29 00:57:32.000000000 +0200
@@ -35,7 +35,7 @@
 
 #ifdef SLAPD_SPASSWD
 #	ifdef HAVE_SASL_SASL_H
-#		include <sasl/sasl.h>
+#		include <sasl2/sasl.h>
 #	else
 #		include <sasl.h>
 #	endif
diff -urN openldap-2.1.28/libraries/liblutil/sasl.c openldap-2.1.28-jsaw/libraries/liblutil/sasl.c
--- openldap-2.1.28/libraries/liblutil/sasl.c	2003-04-26 16:56:37.000000000 +0200
+++ openldap-2.1.28-jsaw/libraries/liblutil/sasl.c	2004-03-29 00:57:32.000000000 +0200
@@ -14,7 +14,7 @@
 #include <ac/unistd.h>
 
 #ifdef HAVE_SASL_SASL_H
-#include <sasl/sasl.h>
+#include <sasl2/sasl.h>
 #else
 #include <sasl.h>
 #endif
diff -urN openldap-2.1.28/servers/slapd/sasl.c openldap-2.1.28-jsaw/servers/slapd/sasl.c
--- openldap-2.1.28/servers/slapd/sasl.c	2004-01-21 08:56:34.000000000 +0100
+++ openldap-2.1.28-jsaw/servers/slapd/sasl.c	2004-03-29 00:57:33.000000000 +0200
@@ -19,14 +19,14 @@
 
 #ifdef HAVE_CYRUS_SASL
 # ifdef HAVE_SASL_SASL_H
-#  include <sasl/sasl.h>
+#  include <sasl2/sasl.h>
 # else
 #  include <sasl.h>
 # endif
 
 # if SASL_VERSION_MAJOR >= 2
 # ifdef HAVE_SASL_SASL_H
-#  include <sasl/saslplug.h>
+#  include <sasl2/saslplug.h>
 # else
 #  include <saslplug.h>
 # endif
