--- ./libegg/menu/egg-action.c.orig	2004-05-14 10:36:25.076328384 +0200
+++ ./libegg/menu/egg-action.c	2004-05-14 10:53:26.393064640 +0200
@@ -1,6 +1,7 @@
 #include <config.h>
 #include "egg-action.h"
 #include "../toolbar/eggtoolbutton.h"
+#include "../toolbar/eggtoolbar.h"
 #include "../util/eggintl.h"
 
 //#ifndef _
@@ -921,7 +922,7 @@
       /* gross hack!!! */
       for (tmp = GTK_TOOLBAR (button->parent)->children; tmp; tmp = tmp->next)
 	{
-	  GtkToolbarChild *tool_child = tmp->data;
+	  EggToolbarChild *tool_child = tmp->data;
 
 	  if (tool_child->widget == button)
 	    {
--- ./libegg/toolbar/eggtoolbar.c.orig	2004-05-14 10:32:25.233789992 +0200
+++ ./libegg/toolbar/eggtoolbar.c	2004-05-14 10:32:51.857742536 +0200
@@ -2570,7 +2570,7 @@
 
 GtkWidget*
 egg_toolbar_append_element (EggToolbar          *toolbar,
-			    GtkToolbarChildType  type,
+			    EggToolbarChildType  type,
 			    GtkWidget           *widget,
 			    const char          *text,
 			    const char          *tooltip_text,
@@ -2587,7 +2587,7 @@
 
 GtkWidget *
 egg_toolbar_prepend_element (EggToolbar          *toolbar,
-			     GtkToolbarChildType  type,
+			     EggToolbarChildType  type,
 			     GtkWidget           *widget,
 			     const char          *text,
 			     const char          *tooltip_text,
