--- ./demos/lineart.c.orig	2000-02-19 11:12:28.000000000 +0200
+++ ./demos/lineart.c	2003-10-19 08:40:01.000000000 +0300
@@ -19,8 +19,7 @@
   
   if(mode == 0)
     {
-      printf("Usage:linp [mode mode ...]\n
-              where mode is an integer.\n");
+      printf("Usage:linp [mode mode ...]\n\nwhere mode is an integer.\n");
       return;
     }
   if(! vga_hasmode(mode)) {diff -Nru3 trunk-1582/package/jimmy/svgalib/installmodule.patch trunk/package/jimmy/svgalib/installmodule.patch
