--- ./Makefile.cfg.orig	2003-01-24 13:56:05.000000000 +0200
+++ ./Makefile.cfg	2003-10-06 21:52:14.000000000 +0300
@@ -128,6 +128,9 @@
 # Comment this out if you use devfs only and don't want to make svga nodes
 INSTALLDEV = installdev
 
+# Comment this out if you are going to install the kernel module manually
+INSTALLMODULE = installmodule
+
 #
 # Comment out any driver that you don't want included in the library.
 #
--- ./Makefile.orig	2002-12-22 12:58:47.000000000 +0200
+++ ./Makefile	2003-10-06 21:52:15.000000000 +0300
@@ -211,7 +211,7 @@
 	(cd threeDKit/; $(MAKE) install)
 	 
 install: uninstall installheaders $(INSTALLSHAREDLIB) installconfig \
-	$(INSTALLSTATICLIB) $(INSTALLUTILS) $(INSTALLMAN) installmodule $(INSTALLDEV) \
+	$(INSTALLSTATICLIB) $(INSTALLUTILS) $(INSTALLMAN) $(INSTALLMODULE) $(INSTALLDEV) \
 	lib3dkit-install
 	@echo
 	@echo
