--- ./kernel/svgalib_helper/kernel26compat.h.orig	2003-10-04 16:42:01.000000000 +0300
+++ ./kernel/svgalib_helper/kernel26compat.h	2004-04-17 15:04:45.000000000 +0300
@@ -9,9 +9,12 @@
 
 # define my_io_remap_page_range(vma, start, ofs, len, prot) io_remap_page_range(start,ofs,len,prot)
 
+#ifndef IRQ_NONE
+/* The newest 2.4 kernels define these ... */
 typedef void irqreturn_t;
 # define IRQ_NONE
 # define IRQ_HANDLED
+#endif
 
 #else /* Kernel 2.6 */
 
@@ -44,10 +47,5 @@
 #endif
 
 
-/* Our main.c really do not like MINOR from 2.6 kernels ... */
-#if (defined MINOR) && (!defined KERNEL_2_6)
-# define my_minor(x) MINOR(x)
-#else
-# define my_minor(x) minor(x)
-#endif
+#define my_minor(x) MINOR(x)
 	
