--- ./kernel/svgalib_helper/kernel26compat.h.orig	2003-10-04 16:42:01.000000000 +0300
+++ ./kernel/svgalib_helper/kernel26compat.h	2004-01-11 17:30:25.000000000 +0200
@@ -9,9 +9,12 @@
 
 # define my_io_remap_page_range(vma, start, ofs, len, prot) io_remap_page_range(start,ofs,len,prot)
 
+#ifndef IRQ_NONE
+/* The newest 2.4 kernels define these ... */
 typedef void irqreturn_t;
 # define IRQ_NONE
 # define IRQ_HANDLED
+#endif
 
 #else /* Kernel 2.6 */
 
