--- ./demos/vgatest.c.orig	2002-01-02 12:41:07.000000000 +0200
+++ ./demos/vgatest.c	2004-06-13 17:01:56.386844392 +0300
@@ -56,6 +56,7 @@
 
     modeinfo = vga_getmodeinfo(mode);
 
+#if 0
     printf("Width: %d  Height: %d  Colors: %d\n",
 	   modeinfo->width,
 	   modeinfo->height,
@@ -64,6 +65,7 @@
 	   modeinfo->startaddressrange,
 	   modeinfo->maxpixels,
 	   modeinfo->haveblit ? "YES" : "NO");
+#endif
 
 #ifdef TEST_MODEX
     if (modeinfo->colors == 256)
