
Imported from Fedora Core.

  - Rene Rebe <rene@exactcode.de>

--- cups-1.1.23/pdftops/GfxFont.cxx.gcc4	2005-02-18 14:04:54.252480058 +0000
+++ cups-1.1.23/pdftops/GfxFont.cxx	2005-02-18 14:05:28.154003304 +0000
@@ -969,7 +969,7 @@
   if ((nameToGID = ff->getNameToGID())) {
     for (i = 0; i < 256; ++i) {
       if (!map[i] && (charName = enc[i])) {
-	map[i] = (Gushort)(int)nameToGID->lookup(charName);
+	map[i] = (Gushort)(long)nameToGID->lookup(charName);
       }
     }
     delete nameToGID;
