--- librsvg-2.8.1/librsvg-2.0.pc.in.orig	2004-10-04 14:30:01.570190768 +0200
+++ librsvg-2.8.1/librsvg-2.0.pc.in	2004-10-04 14:29:41.407256000 +0200
@@ -6,6 +6,6 @@
 Name: librsvg
 Description: library that renders svg files
 Version: @VERSION@
-Requires: glib-2.0 gdk-pixbuf-2.0
+Requires: glib-2.0 gdk-pixbuf-2.0 libgsf-1 libcroco-0.6 gnome-vfs-2.0
 Libs: -L${libdir} -lrsvg-2
 Cflags: -I${includedir}/librsvg-2
