diff -dur avifile-0.7-0.7.38/configure avifile-0.7-0.7.38-patched/configure
--- avifile-0.7-0.7.38/configure	2003-07-10 13:40:59.000000000 +0000
+++ avifile-0.7-0.7.38-patched/configure	2004-06-25 17:23:54.000000000 +0000
@@ -13301,7 +13301,7 @@
   AMM_USE_ARTSC_FALSE=
 fi
 
-last_cvs_update=`find . -name Entries -printf '%Ty%Tm%Td-%TH:%TM\n' | sort | tail -1`
+last_cvs_update=`find . -name Entries -printf '%Ty%Tm%Td-%TH:%TM\n' | sort | tail -n1`
 if test $? -ne 0 -o -z "$last_cvs_update" ; then
     # probably no gnu date installed(?), use current date
     # or this is release
diff -dur avifile-0.7-0.7.38/ffmpeg/libavcodec/i386/dsputil_mmx.c avifile-0.7-0.7.38-patched/ffmpeg/libavcodec/i386/dsputil_mmx.c
--- avifile-0.7-0.7.38/ffmpeg/libavcodec/i386/dsputil_mmx.c	2003-06-23 18:52:34.000000000 +0000
+++ avifile-0.7-0.7.38-patched/ffmpeg/libavcodec/i386/dsputil_mmx.c	2004-06-25 17:23:54.000000000 +0000
@@ -29,8 +29,8 @@
 static const uint64_t mm_wone __attribute__ ((aligned(8))) = 0x0001000100010001ULL;
 static const uint64_t mm_wtwo __attribute__ ((aligned(8))) = 0x0002000200020002ULL;
 
-static const uint64_t ff_pw_20 __attribute__ ((aligned(8))) = 0x0014001400140014ULL;
-static const uint64_t ff_pw_3  __attribute__ ((aligned(8))) = 0x0003000300030003ULL;
+static const uint64_t ff_pw_20 __attribute__ ((used)) __attribute__ ((aligned(8))) = 0x0014001400140014ULL;
+static const uint64_t ff_pw_3  __attribute__ ((used)) __attribute__ ((aligned(8))) = 0x0003000300030003ULL;
 static const uint64_t ff_pw_16 __attribute__ ((aligned(8))) = 0x0010001000100010ULL;
 static const uint64_t ff_pw_15 __attribute__ ((aligned(8))) = 0x000F000F000F000FULL;
 
diff -dur avifile-0.7-0.7.38/ffmpeg/libavcodec/i386/motion_est_mmx.c avifile-0.7-0.7.38-patched/ffmpeg/libavcodec/i386/motion_est_mmx.c
--- avifile-0.7-0.7.38/ffmpeg/libavcodec/i386/motion_est_mmx.c	2003-03-03 17:07:35.000000000 +0000
+++ avifile-0.7-0.7.38-patched/ffmpeg/libavcodec/i386/motion_est_mmx.c	2004-06-25 17:23:54.000000000 +0000
@@ -26,7 +26,7 @@
 0x0002000200020002,
 };
 
-static __attribute__ ((aligned(8))) uint64_t bone= 0x0101010101010101LL;
+static __attribute__ ((used)) __attribute__ ((aligned(8))) uint64_t bone= 0x0101010101010101LL;
 
 static inline void sad8_mmx(uint8_t *blk1, uint8_t *blk2, int stride, int h)
 {
diff -dur avifile-0.7-0.7.38/ffmpeg/libavcodec/i386/simple_idct_mmx.c avifile-0.7-0.7.38-patched/ffmpeg/libavcodec/i386/simple_idct_mmx.c
--- avifile-0.7-0.7.38/ffmpeg/libavcodec/i386/simple_idct_mmx.c	2003-02-14 13:58:00.000000000 +0000
+++ avifile-0.7-0.7.38-patched/ffmpeg/libavcodec/i386/simple_idct_mmx.c	2004-06-25 17:23:54.000000000 +0000
@@ -45,8 +45,8 @@
 #define ROW_SHIFT 11
 #define COL_SHIFT 20 // 6
 
-static const uint64_t __attribute__((aligned(8))) wm1010= 0xFFFF0000FFFF0000ULL;
-static const uint64_t __attribute__((aligned(8))) d40000= 0x0000000000040000ULL;
+static const uint64_t __attribute__ ((used)) __attribute__((aligned(8))) wm1010= 0xFFFF0000FFFF0000ULL;
+static const uint64_t __attribute__ ((used)) __attribute__((aligned(8))) d40000= 0x0000000000040000ULL;
 static int16_t __attribute__((aligned(8))) temp[64];
 static int16_t __attribute__((aligned(8))) coeffs[]= {
 	1<<(ROW_SHIFT-1), 0, 1<<(ROW_SHIFT-1), 0,
diff -dur avifile-0.7-0.7.38/ffmpeg/libavcodec/liba52/resample_mmx.c avifile-0.7-0.7.38-patched/ffmpeg/libavcodec/liba52/resample_mmx.c
--- avifile-0.7-0.7.38/ffmpeg/libavcodec/liba52/resample_mmx.c	2003-04-16 20:03:07.000000000 +0000
+++ avifile-0.7-0.7.38-patched/ffmpeg/libavcodec/liba52/resample_mmx.c	2004-06-25 17:23:54.000000000 +0000
@@ -7,10 +7,11 @@
 	and it would mean (C / MMX2 / MMX / 3DNOW) versions 
 */
 
-static uint64_t __attribute__((aligned(8))) magicF2W= 0x43c0000043c00000LL;
-static uint64_t __attribute__((aligned(8))) wm1010= 0xFFFF0000FFFF0000LL;
-static uint64_t __attribute__((aligned(8))) wm0101= 0x0000FFFF0000FFFFLL;
-static uint64_t __attribute__((aligned(8))) wm1100= 0xFFFFFFFF00000000LL;
+static uint64_t __attribute__((used)) __attribute__((aligned(8))) magicF2W= 0x43c0000043c00000LL;
+static uint64_t __attribute__((used)) __attribute__((aligned(8))) wm1010= 0xFFFF0000FFFF0000LL;
+static uint64_t __attribute__((used)) __attribute__((aligned(8))) wm0101= 0x0000FFFF0000FFFFLL;
+static uint64_t __attribute__((used)) __attribute__((aligned(8))) wm1100= 0xFFFFFFFF00000000LL;
+
 
 static int a52_resample_MONO_to_5_MMX(float * _f, int16_t * s16){
     int32_t * f = (int32_t *) _f;
diff -dur avifile-0.7-0.7.38/ffmpeg/libavcodec/libpostproc/postprocess.c avifile-0.7-0.7.38-patched/ffmpeg/libavcodec/libpostproc/postprocess.c
--- avifile-0.7-0.7.38/ffmpeg/libavcodec/libpostproc/postprocess.c	2003-06-26 19:14:31.000000000 +0000
+++ avifile-0.7-0.7.38-patched/ffmpeg/libavcodec/libpostproc/postprocess.c	2004-06-25 17:23:54.000000000 +0000
@@ -104,8 +104,8 @@
 //#define NUM_BLOCKS_AT_ONCE 16 //not used yet
 
 #ifdef ARCH_X86
-static uint64_t __attribute__((aligned(8))) w05=		0x0005000500050005LL;
-static uint64_t __attribute__((aligned(8))) w20=		0x0020002000200020LL;
+static uint64_t __attribute__ ((used)) __attribute__((aligned(8))) w05=		0x0005000500050005LL;
+static uint64_t __attribute__ ((used)) __attribute__((aligned(8))) w20=		0x0020002000200020LL;
 static uint64_t __attribute__((aligned(8))) b00= 		0x0000000000000000LL;
 static uint64_t __attribute__((aligned(8))) b01= 		0x0101010101010101LL;
 static uint64_t __attribute__((aligned(8))) b02= 		0x0202020202020202LL;
diff -dur avifile-0.7-0.7.38/include/aviplay.h avifile-0.7-0.7.38-patched/include/aviplay.h
--- avifile-0.7-0.7.38/include/aviplay.h	2003-05-24 22:51:30.000000000 +0000
+++ avifile-0.7-0.7.38-patched/include/aviplay.h	2004-06-25 17:23:54.000000000 +0000
@@ -12,8 +12,8 @@
 #include "image.h"
 #include "avm_args.h"
 
-typedef void (*KILLHANDLER)(int, void* p = 0);
-typedef int (*AUDIOFUNC)(void* srcdata, unsigned int size, void* p = 0);
+typedef void (*KILLHANDLER)(int, void* p);
+typedef int (*AUDIOFUNC)(void* srcdata, unsigned int size, void* p);
 
 AVM_BEGIN_NAMESPACE;
 
diff -dur avifile-0.7-0.7.38/include/avm_map.h avifile-0.7-0.7.38-patched/include/avm_map.h
--- avifile-0.7-0.7.38/include/avm_map.h	2003-03-27 01:01:22.000000000 +0000
+++ avifile-0.7-0.7.38-patched/include/avm_map.h	2004-06-25 17:23:54.000000000 +0000
@@ -198,7 +198,8 @@
 {
     // cast to the needed type - used to prevent internal compiler error
     // for old egcc
-    avm_map<Key, Value, Compare, Equal>::_Tnode* node = (avm_map<Key, Value, Compare, Equal>::_Tnode*) n;
+//    avm_map<Key, Value, Compare, Equal>::_Tnode* node = (avm_map<Key, Value, Compare, Equal>::_Tnode*) n;
+    binary_tree_node<Key,Value> * node = (binary_tree_node<Key,Value>*) n;
 
     if(node->entry)
     {
diff -dur avifile-0.7-0.7.38/include/avm_stl.h avifile-0.7-0.7.38-patched/include/avm_stl.h
--- avifile-0.7-0.7.38/include/avm_stl.h	2003-06-08 15:57:05.000000000 +0000
+++ avifile-0.7-0.7.38-patched/include/avm_stl.h	2004-06-25 17:23:54.000000000 +0000
@@ -149,7 +149,7 @@
 
     // disabled for now
     qring<Type>() {}
-    qring<Type>(const qring<Type>& t) {}
+//    qring<Type>(const qring<Type>& t) {}
 
     uint_t bpos() const { return (m_uiPos > 0) ? m_uiPos - 1 : m_uiCapacity - 1; }
     uint_t fpos() const { return (m_uiSize > m_uiPos) ? m_uiPos + m_uiCapacity - m_uiSize: m_uiPos - m_uiSize; }
diff -dur avifile-0.7-0.7.38/lib/aviread/AsfNetworkInputStream.cpp avifile-0.7-0.7.38-patched/lib/aviread/AsfNetworkInputStream.cpp
--- avifile-0.7-0.7.38/lib/aviread/AsfNetworkInputStream.cpp	2003-05-27 15:26:31.000000000 +0000
+++ avifile-0.7-0.7.38-patched/lib/aviread/AsfNetworkInputStream.cpp	2004-06-25 17:26:02.000000000 +0000
@@ -801,7 +801,8 @@
 	int i = ::write(m_iSocket, buffer, wsize);
 	if (i <= 0)
             return i;
-	(const char*)buffer += i;
+ 	const char* cbuffer = (const char*) buffer;
+ 	buffer = (void*) (cbuffer+i);
         wsize -= i;
     }
     return size;
@@ -843,7 +844,8 @@
 	int i = ::write(m_lfd, buffer, size);
 	if (i < 0)
             return i;
-	(const char*)buffer += i;
+ 	const char* cbuffer = (const char*) buffer;
+ 	buffer = (void*) (cbuffer+i);
 	size -= i;
     }
     fsync(m_lfd);
Nur in avifile-0.7-0.7.38-patched/lib/aviread: AsfNetworkInputStream.cpp~.
Nur in avifile-0.7-0.7.38-patched/lib/aviread: AsfNetworkInputStream.cpp.rej.
diff -dur avifile-0.7-0.7.38/lib/common/image.cpp avifile-0.7-0.7.38-patched/lib/common/image.cpp
--- avifile-0.7-0.7.38/lib/common/image.cpp	2003-05-24 23:00:58.000000000 +0000
+++ avifile-0.7-0.7.38-patched/lib/common/image.cpp	2004-06-25 17:23:54.000000000 +0000
@@ -954,7 +954,7 @@
     fillMembers();
     if (!copy)
     {
-	(const uint8_t*) m_pPlane[0] = data;
+	m_pPlane[0] = (unsigned char*) data;
     }
     else
     {
diff -dur avifile-0.7-0.7.38/plugins/libmp3lame_audioenc/lame3.70/quantize-pvt.c avifile-0.7-0.7.38-patched/plugins/libmp3lame_audioenc/lame3.70/quantize-pvt.c
--- avifile-0.7-0.7.38/plugins/libmp3lame_audioenc/lame3.70/quantize-pvt.c	2001-03-27 17:04:29.000000000 +0000
+++ avifile-0.7-0.7.38-patched/plugins/libmp3lame_audioenc/lame3.70/quantize-pvt.c	2004-06-25 17:23:54.000000000 +0000
@@ -1007,18 +1007,18 @@
         "dec %4\n\t"
 
         "movl %5, %%eax\n\t"
-        "movl 4+%5, %%ebx\n\t"
+        "movl 4+%5, %%ecx\n\t"
         "fxch %%st(1)\n\t"
         "fadd" F8type " (%2,%%eax," F8size ")\n\t"
         "fxch %%st(3)\n\t"
-        "fadd" F8type " (%2,%%ebx," F8size ")\n\t"
+        "fadd" F8type " (%2,%%ecx," F8size ")\n\t"
 
         "movl 8+%5, %%eax\n\t"
-        "movl 12+%5, %%ebx\n\t"
+        "movl 12+%5, %%ecx\n\t"
         "fxch %%st(2)\n\t"
         "fadd" F8type " (%2,%%eax," F8size ")\n\t"
         "fxch %%st(1)\n\t"
-        "fadd" F8type " (%2,%%ebx," F8size ")\n\t"
+        "fadd" F8type " (%2,%%ecx," F8size ")\n\t"
 
         "fxch %%st(3)\n\t"
         "fistpl -16(%3)\n\t"
@@ -1030,7 +1030,7 @@
         "jnz loop1\n\n"
         : /* no outputs */
         : "t" (istep), "r" (xr), "r" (adj43asm), "r" (ix), "r" (576 / 4), "m" (rx)
-        : "%eax", "%ebx", "memory", "cc"
+        : "%eax", "%ecx", "memory", "cc"
       );
   }
 #elif defined (USE_MSC_ASM)
diff -dur avifile-0.7-0.7.38/plugins/libvorbis/libvorbis.cpp avifile-0.7-0.7.38-patched/plugins/libvorbis/libvorbis.cpp
--- avifile-0.7-0.7.38/plugins/libvorbis/libvorbis.cpp	2003-05-20 14:33:13.000000000 +0000
+++ avifile-0.7-0.7.38-patched/plugins/libvorbis/libvorbis.cpp	2004-06-25 17:23:54.000000000 +0000
@@ -96,7 +96,7 @@
 	op.packet = NULL;
 	op.b_o_s  = 1; /* beginning of stream for first packet */
 	op.bytes  = hdrsizes[0];
-	(const void*) op.packet = vorbishdr;
+	op.packet = (unsigned char*) vorbishdr;
 	vorbishdr += op.bytes;
 	if (vorbis_synthesis_headerin(&vi, &vc, &op) < 0)
 	{
@@ -106,7 +106,7 @@
 
 	op.b_o_s  = 0;
 	op.bytes  = hdrsizes[1];
-	(const void*) op.packet = vorbishdr;
+	op.packet = (unsigned char*) vorbishdr;
 	vorbishdr += op.bytes;
 	if (vorbis_synthesis_headerin(&vi, &vc, &op) < 0)
 	{
@@ -115,7 +115,7 @@
 	}
 
 	op.bytes  = hdrsizes[2];
-	(const void*) op.packet = vorbishdr;
+	op.packet = (unsigned char*) vorbishdr;
 	vorbishdr += op.bytes;
 	if (vorbis_synthesis_headerin(&vi, &vc, &op) < 0)
 	{
