--- ./ffmpeg/libavcodec/i386/h264dsp_mmx.c.orig	2000-01-01 00:00:00.000000000 +0100
+++ ./ffmpeg/libavcodec/i386/h264dsp_mmx.c	2005-09-14 13:57:24.000000000 +0200
@@ -260,6 +260,7 @@
 }
 static void h264_h_loop_filter_luma_mmx2(uint8_t *pix, int stride, int alpha, int beta, int8_t *tc0)
 {
+#if 0
     //FIXME: could cut some load/stores by merging transpose with filter
     // also, it only needs to transpose 6x8
     uint8_t trans[8*8];
@@ -275,6 +276,7 @@
         transpose4x4(pix-2,          trans  +2*8, stride, 8);
         transpose4x4(pix-2+4*stride, trans+4+2*8, stride, 8);
     }
+#endif
 }
 
 static inline void h264_loop_filter_chroma_mmx2(uint8_t *pix, int stride, int alpha1, int beta1, int8_t *tc0)
