diff -ur avifile-0.6/Makefile.am avifile-0.6-fixed/Makefile.am
--- avifile-0.6/Makefile.am	2000-01-01 00:00:00.000000000 +0100
+++ avifile-0.6-fixed/Makefile.am	2005-10-26 11:28:21.000000000 +0200
@@ -1,4 +1,4 @@
-SUBDIRS = . admin m4 ffmpeg drivers lib plugins bin \
+SUBDIRS = . admin m4 drivers lib plugins bin \
  doc include debian libavqt player samples
 
 EXTRA_DIST = INSTALL autogen.sh avifile.spec.in avifile.pc.in \
diff -ur avifile-0.6/lib/Makefile.am avifile-0.6-fixed/lib/Makefile.am
--- avifile-0.6/lib/Makefile.am	2000-01-01 00:00:00.000000000 +0100
+++ avifile-0.6-fixed/lib/Makefile.am	2005-10-26 12:07:58.000000000 +0200
@@ -2,9 +2,9 @@
 noinst_HEADERS = Uncompressed.h
 
 COMMONCFLAGS = $(DIVX4_CFLAGS) $(XVID_CFLAGS) -DPLUGIN_PATH=\"$(pkglibdir)\" \
- -I$(srcdir)/../ffmpeg/libavcodec \
- -I$(srcdir)/../ffmpeg/libavformat \
- -I$(srcdir)/../ffmpeg/libavutil
+ -I/usr/include/ffmpeg \
+ -I/usr/include/ffmpeg \
+ -I/usr/include/ffmpeg
 if AMM_USE_FAST_BUILD
 AVISOURCES = libaviplay.cpp
 AVILIBADD =
diff -ur avifile-0.6/plugins/Makefile.am avifile-0.6-fixed/plugins/Makefile.am
--- avifile-0.6/plugins/Makefile.am	2000-01-01 00:00:00.000000000 +0100
+++ avifile-0.6-fixed/plugins/Makefile.am	2005-10-26 12:07:58.000000000 +0200
@@ -20,7 +20,7 @@
 
 AM_CPPFLAGS = -Ilibac3pass -Ilibwin32 -Ilibwin32/audiodec \
  -Ilibwin32/videocodec -Ilibwin32/loader \
- -Ilibwin32/loader/dshow -I../ffmpeg/libavcodec \
+ -Ilibwin32/loader/dshow -I/usr/include/ffmpeg \
  -DWIN32_PATH=\"$(WIN32_PATH)\" -D__WINE__ -DUNICODE \
  -Ddbg_printf=__vprintf -DTRACE=__vprintf \
  -Wmissing-prototypes -Wimplicit-function-declaration \
diff -ur avifile-0.6/plugins/libffmpeg/Makefile.am avifile-0.6-fixed/plugins/libffmpeg/Makefile.am
--- avifile-0.6/plugins/libffmpeg/Makefile.am	2000-01-01 00:00:00.000000000 +0100
+++ avifile-0.6-fixed/plugins/libffmpeg/Makefile.am	2005-10-26 12:07:58.000000000 +0200
@@ -18,10 +18,10 @@
 #KPROF_LIBS = -L/var/tmp/kprof/lib -lkprof
 #INSTRUMENT = -finstrument-functions
 AM_CPPFLAGS = $(LTNOPIC) $(INSTRUMENT) \
- -I$(srcdir)/../../ffmpeg/libavcodec \
- -I$(srcdir)/../../ffmpeg/libavformat \
- -I$(srcdir)/../../ffmpeg/libavutil \
- -I$(srcdir)/../../ffmpeg
+ -I/usr/include/ffmpeg \
+ -I/usr/include/ffmpeg \
+ -I/usr/include/ffmpeg \
+ -I/usr/include/ffmpeg
 AM_CXXFLAGS = $(CXXRTTIEXCEPT)
 
 TEST_LIBS = ../../lib/libaviplay.la $(pkglib_LTLIBRARIES) $(SDL_LIBS) $(MPATROL_LIBS) $(EFENCE_LIBS) $(KPROF_LIBS)
--- ./configure.in	2005-10-26 12:32:57.000000000 +0200
+++ ./configure.in	2005-10-26 12:41:18.000000000 +0200
@@ -1002,26 +1002,11 @@
 AC_SUBST(AVIFILE_BUILD)
 
 dnl mpglib/Makefile
-dnl ffmpeg/libavfilter/Makefile
 dnl libpp/Makefile
 
 AC_OUTPUT(
  Makefile
  admin/Makefile
- ffmpeg/Makefile
- ffmpeg/libavformat/Makefile
- ffmpeg/libavcodec/Makefile
- ffmpeg/libavcodec/alpha/Makefile
- ffmpeg/libavcodec/armv4l/Makefile
- ffmpeg/libavcodec/i386/Makefile
- ffmpeg/libavcodec/liba52/Makefile
- ffmpeg/libavcodec/libpostproc/Makefile
- ffmpeg/libavcodec/mlib/Makefile
- ffmpeg/libavcodec/ppc/Makefile
- ffmpeg/libavcodec/ps2/Makefile
- ffmpeg/libavcodec/sh4/Makefile
- ffmpeg/libavcodec/sparc/Makefile
- ffmpeg/libavutil/Makefile
  drivers/Makefile
  drivers/libdha/Makefile
  drivers/libdha/bin/Makefile
--- ./lib/aviread/Makefile.am.orig	2000-01-01 00:00:00.000000000 +0100
+++ ./lib/aviread/Makefile.am	2005-10-26 12:53:36.000000000 +0200
@@ -43,24 +43,14 @@
  StreamInfo.cpp
 
 
-if AMM_LINK_SHARED
-FF_LIB = \
- ../../ffmpeg/libavformat/libaviplayavformat.la \
- ../../ffmpeg/libavcodec/libaviplayavcodec.la \
- ../../ffmpeg/libavutil/libaviplayavutil.la
-else 
-FF_LIB = \
- ../../ffmpeg/libavformat/libavformat.la \
- ../../ffmpeg/libavcodec/libavcodec.la \
- ../../ffmpeg/libavutil/libavutil.la
-endif
+FF_LIB = 
 
 libaviread_la_LIBADD = $(FF_LIB)
 
 AM_CPPFLAGS = $(LTNOPIC) \
  -I$(srcdir)/../common \
- -I$(srcdir)/../../ffmpeg/libavformat \
- -I$(srcdir)/../../ffmpeg/libavcodec \
- -I$(srcdir)/../../ffmpeg/libavutil
+ -I/usr/include/ffmpeg \
+ -I/usr/include/ffmpeg \
+ -I/usr/include/ffmpeg
 
 MAINTAINERCLEANFILES = Makefile.in
