diff -ur libpng-1.2.8/scripts/libpng.pc.in libpng-1.2.8-t2/scripts/libpng.pc.in
--- libpng-1.2.8/scripts/libpng.pc.in	2004-12-03 01:14:31.000000000 +0100
+++ libpng-1.2.8-t2/scripts/libpng.pc.in	2006-02-16 22:05:33.000000000 +0100
@@ -1,7 +1,7 @@
 
 prefix=@PREFIX@
 exec_prefix=${prefix}
-libdir=${exec_prefix}/lib
+libdir=@LIBDIR@
 includedir=${exec_prefix}/include
 
 Name: libpng12
diff -ur libpng-1.2.8/scripts/makefile.linux libpng-1.2.8-t2/scripts/makefile.linux
--- libpng-1.2.8/scripts/makefile.linux	2004-12-03 01:14:31.000000000 +0100
+++ libpng-1.2.8-t2/scripts/makefile.linux	2006-02-16 22:07:06.000000000 +0100
@@ -78,7 +78,9 @@
 	$(RANLIB) $@
 
 libpng.pc:
-	cat scripts/libpng.pc.in | sed -e s\!@PREFIX@!$(prefix)! > libpng.pc
+	cat scripts/libpng.pc.in | \
+		sed -e s\!@PREFIX@!$(prefix)! \
+		    -e s\!@LIBDIR@!$(LIBPATH)! > libpng.pc
 
 libpng-config:
 	( cat scripts/libpng-config-head.in; \
