--- ./gcc/ch/loop.c.orig	2003-08-03 12:26:04.000000000 -0400
+++ ./gcc/ch/loop.c	2003-08-03 12:28:30.000000000 -0400
@@ -318,8 +318,8 @@
 (associated with all iterators).
 
  NOTE: Z.200, section 10.1 says that a block is ...
-       "the actions statement list in a do action, including any
-       loop counter and while control".  This means that an exp-
+       the actions statement list in a do action, including any
+       loop counter and while control.  This means that an exp-
        ression in a WHILE control can include references to the
        loop counters created for the loop''s exclusive use.  
        Example:
