--- ./panel-plugin/sample.c.orig	2003-07-18 18:58:50.000000000 +0300
+++ ./panel-plugin/sample.c	2003-12-26 12:17:26.000000000 +0200
@@ -127,7 +127,7 @@
 
 /* options dialog */
 static void
-create_options (Control *ctrl, GtkContainer *con, GtkWidget *done)
+sample_create_options (Control *ctrl, GtkContainer *con, GtkWidget *done)
 {
 }
 
@@ -147,7 +147,7 @@
 	/* options; don't define if you don't have any ;) */
 	cc->read_config		= sample_read_config;
 	cc->write_config	= sample_write_config;
-	cc->create_options		= sample_create_options;
+	cc->create_options	= sample_create_options;
 
 	/* Don't use this function at all if you want xfce to
 	 * do the sizing.
