--- gcc-4.1/libstdc++-v3/config/locale/uclibc/monetary_members.cc.vanilla	2007-05-23 14:34:02.593766069 +0200
+++ gcc-4.1/libstdc++-v3/config/locale/uclibc/monetary_members.cc	2007-05-23 14:34:50.232480845 +0200
@@ -406,8 +406,8 @@
 	  _M_data->_M_decimal_point = __cloc->decimal_point_wc;
 	  _M_data->_M_thousands_sep = __cloc->thousands_sep_wc;
 # else
-	  _M_data->_M_decimal_point = __global_locale->decimal_point_wc;
-	  _M_data->_M_thousands_sep = __global_locale->thousands_sep_wc;
+//	  _M_data->_M_decimal_point = __global_locale->decimal_point_wc;
+//	  _M_data->_M_thousands_sep = __global_locale->thousands_sep_wc;
 # endif
 #else
 	  union { char *__s; wchar_t __w; } __u;
@@ -561,8 +561,8 @@
 	  _M_data->_M_decimal_point = __cloc->decimal_point_wc;
 	  _M_data->_M_thousands_sep = __cloc->thousands_sep_wc;
 # else
-	  _M_data->_M_decimal_point = __global_locale->decimal_point_wc;
-	  _M_data->_M_thousands_sep = __global_locale->thousands_sep_wc;
+//	  _M_data->_M_decimal_point = __global_locale->decimal_point_wc;
+//	  _M_data->_M_thousands_sep = __global_locale->thousands_sep_wc;
 # endif
 #else
           union { char *__s; wchar_t __w; } __u;
--- gcc-4.1/libstdc++-v3/config/locale/uclibc/numeric_members.cc.vanilla	2007-05-23 14:27:29.167345975 +0200
+++ gcc-4.1/libstdc++-v3/config/locale/uclibc/numeric_members.cc	2007-05-23 14:27:41.352040342 +0200
@@ -138,8 +138,8 @@
 	  _M_data->_M_decimal_point = __cloc->decimal_point_wc;
 	  _M_data->_M_thousands_sep = __cloc->thousands_sep_wc;
 # else
-	  _M_data->_M_decimal_point = __global_locale->decimal_point_wc;
-	  _M_data->_M_thousands_sep = __global_locale->thousands_sep_wc;
+	  //_M_data->_M_decimal_point = __global_locale->decimal_point_wc;
+	  //_M_data->_M_thousands_sep = __global_locale->thousands_sep_wc;
 # endif
 #else
 	  union { char *__s; wchar_t __w; } __u;
