--- libgtop-2.8.0/examples/Makefile.in~	2004-09-13 16:43:58.000000000 +0200
+++ libgtop-2.8.0/examples/Makefile.in	2004-10-04 13:12:23.791280792 +0200
@@ -144,7 +144,7 @@
 sysdeps_dir = @sysdeps_dir@
 sysdeps_suid_lib = @sysdeps_suid_lib@
 
-LINK = $(LIBTOOL) --mode=link $(CC) $(CFLAGS) $(LDFLAGS) -o $@
+LINK = $(LIBTOOL) --mode=link $(CC) $(CFLAGS) $(LDFLAGS) $(GLIB_LIBS) -o $@
 
 INCLUDES = @INCLUDES@
 
