--- rivatv-0.8.6/Makefile.in	2005-07-23 22:55:35.191472688 +0200
+++ rivatv-0.8.6-patch/Makefile.in	2005-07-23 22:56:55.609247328 +0200
@@ -70,14 +70,14 @@
 	@make -C $(KERNEL) SUBDIRS=`pwd`/src EXTRA_CFLAGS="-I`pwd`/bttv @CFLAGS@" modules
 
 devices:
-	@if ! test -c /dev/video0 ; then \
-	  echo Creating video devices.; \
-	  mknod /dev/video0 c 81 0; chmod 660 /dev/video0; \
-	  mknod /dev/video1 c 81 1; chmod 660 /dev/video1; \
-	  mknod /dev/video2 c 81 2; chmod 660 /dev/video2; \
-	  mknod /dev/video3 c 81 3; chmod 660 /dev/video3; \
-	  ln -s /dev/video0 /dev/video; \
-	fi
+#	@if ! test -c /dev/video0 ; then \
+#	  echo Creating video devices.; \
+#	  mknod /dev/video0 c 81 0; chmod 660 /dev/video0; \
+#	  mknod /dev/video1 c 81 1; chmod 660 /dev/video1; \
+#	  mknod /dev/video2 c 81 2; chmod 660 /dev/video2; \
+#	  mknod /dev/video3 c 81 3; chmod 660 /dev/video3; \
+#	  ln -s /dev/video0 /dev/video; \
+#	fi
 
 config:
 	@-./configure
