--- nvi-1.79/build/Makefile.in.vanilla	2005-09-24 22:04:16.000000000 +0200
+++ nvi-1.79/build/Makefile.in	2005-09-24 22:05:36.000000000 +0200
@@ -137,19 +137,6 @@
 	    $(chmod) $(dmode) $(datadir)/vi/catalog
 	(cd $(srcdir)/catalog && $(cp) $(cat) $(datadir)/vi/catalog && \
 	    cd $(datadir)/vi/catalog && $(chmod) $(fmode) *)
-	@echo "Installing Perl scripts: $(datadir)/vi/perl ..."
-	$(mkdir) $(datadir)/vi/perl && $(chmod) $(dmode) $(datadir)/vi/perl
-	[ -f VI.pm ] && $(cp) VI.pm $(datadir)/vi/perl && \
-	    cd $(datadir)/vi/perl && $(chmod) $(fmode) VI.pm)
-	(cd $(srcdir)/perl_scripts && $(cp) *.pl $(datadir)/vi/perl && \
-	    cd $(datadir)/vi/perl && $(chmod) $(fmode) *.pl)
-	@echo "Installing Tcl scripts: $(datadir)/vi/tcl ..."
-	$(mkdir) $(datadir)/vi/tcl && $(chmod) $(dmode) $(datadir)/vi/tcl
-	(cd $(srcdir)/tcl_scripts && $(cp) *.tcl $(datadir)/vi/tcl && \
-	    cd $(datadir)/vi/tcl && $(chmod) $(fmode) *.tcl)
-	@echo "Installing recover script: $(datadir)/vi/recover ..."
-	($(cp) recover $(datadir)/vi/recover && \
-	    $(chmod) $(emode) $(datadir)/vi/recover)
 
 uninstall:
 	$(rm) -rf $(datadir)/vi
