
One more type conflict ...

  - Rene Rebe <rene@exactcode.de>

diff -ur nvi.1.49/PORT/linux/db/hash/hash.c nvi.1.49-fixed/PORT/linux/db/hash/hash.c
--- nvi.1.49/PORT/linux/db/hash/search.h	1993-06-05 00:22:44.000000000 +0200
+++ nvi.1.49-fixed/PORT/linux/db/hash/search.h	2005-02-02 13:52:06.221408376 +0100
@@ -46,6 +46,6 @@
 	FIND, ENTER
 } ACTION;
 
-int	 hcreate __P((unsigned int));
+int	 hcreate __P((u_int));
 void	 hdestroy __P((void));
 ENTRY	*hsearch __P((ENTRY, ACTION));
--- nvi.1.49/ex/filter.c.vanilla	2005-02-02 15:21:01.656298984 +0100
+++ nvi.1.49/ex/filter.c	2005-02-02 15:21:10.680927032 +0100
@@ -342,7 +342,6 @@
 	const char *cmd;
 	int okpipe;
 {
-	extern const char *const sys_siglist[];
 	size_t len;
 	int nf, pstat;
 	char *p;
