--- ./src/form_field.cpp.org	Mon Jan 20 13:56:48 2003
+++ ./src/form_field.cpp	Mon Jan 20 13:58:23 2003
@@ -133,7 +133,7 @@
 }
 
 
-Form::Form( Slots* slots = 0 ) : m_slots( slots ), m_form(0), m_visible( false )
+Form::Form( Slots* slots ) : m_slots( slots ), m_form(0), m_visible( false )
 {
 	m_instances.push_back( this );
 }
--- ./src/form_field.h.org	Mon Jan 20 13:56:52 2003
+++ ./src/form_field.h	Mon Jan 20 13:58:36 2003
@@ -86,7 +86,7 @@
 		virtual void slot_fieldChanged( Field* field ) {}
 	};
 	
-	Form( Slots* slots = 0 );
+	Form( Slots* slots );
 	~Form();
 	
 	vector<Field *>& fields();
