--- openh323/src/Makefile.orig	2004-10-05 23:50:07.678489264 +0200
+++ openh323/src/Makefile	2004-10-05 23:52:23.289873200 +0200
@@ -875,6 +875,7 @@
 HOST_PLATFORM_TYPE=$(PLATFORM_TYPE)
 endif
 
+asnparser.version:
 
 # Set library path so asnparser will run
 
@@ -889,28 +890,12 @@
 # If we're cross compiling, we want the host's asnparser
 # otherwise use the one for the current platform
 ASNPARSE_DIR = $(HOSTPWLIBDIR)/tools/asnparser
-ASNPARSER = $(ASNPARSE_DIR)/obj_$(HOST_PLATFORM_TYPE)_r/asnparser
-
-
-# If not cross compiling then make sure asnparser is built
-ifeq ($(PLATFORM_TYPE),$(HOST_PLATFORM_TYPE))
-$(ASNPARSER):
-	$(MAKE) -C $(ASNPARSE_DIR) opt
-endif
-
+ASNPARSER = asnparser
 
 ifdef ASN_EXCLUDE
 ASNPARSER += -x $(ASN_EXCLUDE)
 endif
 
-asnparser.version: $(ASNPARSER)
-	$(ASNPARSER) --version | awk '{print $$1,$$2,$$3}' > asnparser.version.new
-	if test -f asnparser.version && diff asnparser.version.new asnparser.version >/dev/null 2>&1 ; \
-		then rm asnparser.version.new ; \
-		else mv asnparser.version.new asnparser.version ; \
-	fi
-
-
 # Build the .cxx files from the .asn files
 
 #### h245
