--- openh323/include/gkserver.h~	2004-02-15 04:36:30.000000000 +0100
+++ openh323/include/gkserver.h	2004-10-06 00:11:48.707702760 +0200
@@ -400,9 +400,12 @@
     virtual Response OnHandlePDU();
 };
 
+class H323GatekeeperServer;
 
 class H323GatekeeperRRQ : public H323GatekeeperRequest
 {
+  friend class H323GatekeeperServer;
+
     PCLASSINFO(H323GatekeeperRRQ, H323GatekeeperRequest);
   public:
     H323GatekeeperRRQ(
@@ -465,6 +468,8 @@
 
 class H323GatekeeperARQ : public H323GatekeeperRequest
 {
+  friend class H323GatekeeperServer;
+
     PCLASSINFO(H323GatekeeperARQ, H323GatekeeperRequest);
   public:
     H323GatekeeperARQ(
@@ -1330,6 +1335,9 @@
   */
 class H323GatekeeperServer : public H323TransactionServer
 {
+  friend H323GatekeeperRequest::Response H323GatekeeperRRQ::OnHandlePDU();
+  friend H323GatekeeperRequest::Response H323GatekeeperARQ::OnHandlePDU();
+
     PCLASSINFO(H323GatekeeperServer, H323TransactionServer);
   public:
   /**@name Construction */
@@ -1912,9 +1920,6 @@
     PINDEX peakCalls;
     PINDEX totalCalls;
     PINDEX rejectedCalls;
-
-  friend H323GatekeeperRequest::Response H323GatekeeperRRQ::OnHandlePDU();
-  friend H323GatekeeperRequest::Response H323GatekeeperARQ::OnHandlePDU();
 };
 
 
