diff -urN -x '*~' at-spi-1.1.9-old/configure at-spi-1.1.9-new/configure
--- at-spi-1.1.9-old/configure	2003-03-27 14:52:12.000000000 +0100
+++ at-spi-1.1.9-new/configure	2003-07-01 15:34:26.000000000 +0200
@@ -10000,11 +10000,6 @@
    { (exit 1); exit 1; }; }
   fi
 
-ATK_BRIDGE_LIBS="$ATK_BRIDGE_LIBS -lspi"
-
-
-
-
   succeeded=no
 
   if test -z "$PKG_CONFIG"; then
diff -urN -x '*~' at-spi-1.1.9-old/configure.in at-spi-1.1.9-new/configure.in
--- at-spi-1.1.9-old/configure.in	2003-03-27 14:50:05.000000000 +0100
+++ at-spi-1.1.9-new/configure.in	2003-07-01 15:34:32.000000000 +0200
@@ -128,7 +128,6 @@
 PKG_CHECK_MODULES(ATK_BRIDGE, \
 	libbonobo-2.0 >= 1.107.0 \
 	atk >= 1.1.0)
-ATK_BRIDGE_LIBS="$ATK_BRIDGE_LIBS -lspi" 
 AC_SUBST(ATK_BRIDGE_LIBS)
 AC_SUBST(ATK_BRIDGE_CFLAGS)
 
