--- ./bgpd/bgp_aspath.c	Wed Jun 19 12:24:24 2002
+++ ./bgpd/bgp_aspath.c	Tue Aug  6 11:21:29 2002
@@ -1111,19 +1111,22 @@
   return aspath;
 }
 
-/* Make hash value by raw aspath data. */
+/* Make hash value by raw aspath data. Produces a 16-19 bit hash */
 unsigned int
 aspath_key_make (struct aspath *aspath)
 {
   unsigned int key = 0;
   int length;
-  caddr_t pnt;
+  unsigned short *pnt;
 
-  length = aspath->length;
-  pnt = aspath->data;
+  length = aspath->length / 2;
+  pnt = (unsigned short *) aspath->data;
 
   while (length)
-    key += pnt[--length];
+    {
+      key += *pnt++;
+      length--;
+    }
 
   return key;
 }
@@ -1143,7 +1146,7 @@
 void
 aspath_init ()
 {
-  ashash = hash_create (aspath_key_make, aspath_cmp);
+  ashash = hash_create_size (131072, aspath_key_make, aspath_cmp);
 }
 
 /* return and as path value */
@@ -1167,7 +1170,7 @@
 
   as = (struct aspath *) backet->data;
 
-  vty_out (vty, "[%p:%d] (%ld) ", backet, backet->key, as->refcnt);
+  vty_out (vty, "[%p:%u] (%ld) ", backet, backet->key, as->refcnt);
   vty_out (vty, "%s%s", as->str, VTY_NEWLINE);
 }
 
