--- ./ospfd/ospf_packet.c	Thu Jul 25 19:06:00 2002
+++ ./ospfd/ospf_packet.c	Tue Oct  8 08:34:59 2002
@@ -306,6 +306,7 @@
   struct md5_ctx ctx;
   void *ibuf;
   unsigned long oldputp;
+  u_int32_t t;
   struct crypt_key *ck;
   char *auth_key;
 
@@ -317,7 +318,9 @@
 
   /* We do this here so when we dup a packet, we don't have to
      waste CPU rewriting other headers. */
-  ospfh->u.crypt.crypt_seqnum = htonl (oi->crypt_seqnum++);
+  t = (time(NULL) & 0xFFFFFFFF);
+  oi->crypt_seqnum = ( t > oi->crypt_seqnum ? t : oi->crypt_seqnum++);
+  ospfh->u.crypt.crypt_seqnum = htonl (oi->crypt_seqnum); 
 
   /* Get MD5 Authentication key from auth_key list. */
   if (list_isempty (OSPF_IF_PARAM (oi, auth_crypt)))
