--- ./ospfd/ospf_vty.c	Thu Jul 25 19:06:00 2002
+++ ./ospfd/ospf_vty.c	Sat Aug 24 19:37:43 2002
@@ -3764,9 +3764,9 @@
 #ifdef HAVE_OPAQUE_LSA
   else if (strncmp (argv[0], "opaque-l", 8) == 0)
     type = OSPF_OPAQUE_LINK_LSA;
-  else if (strncmp (argv[0], "opaque-a", 8) == 0)
+  else if (strncmp (argv[0], "opaque-ar", 9) == 0)
     type = OSPF_OPAQUE_AREA_LSA;
-  else if (strncmp (argv[0], "opaque-e", 8) == 0)
+  else if (strncmp (argv[0], "opaque-as", 9) == 0)
     type = OSPF_OPAQUE_AS_LSA;
 #endif /* HAVE_OPAQUE_LSA */
   else
@@ -3888,9 +3888,9 @@
 #ifdef HAVE_OPAQUE_LSA
   else if (strncmp (argv[0], "opaque-l", 8) == 0)
     type = OSPF_OPAQUE_LINK_LSA;
-  else if (strncmp (argv[0], "opaque-a", 8) == 0)
+  else if (strncmp (argv[0], "opaque-ar", 9) == 0)
     type = OSPF_OPAQUE_AREA_LSA;
-  else if (strncmp (argv[0], "opaque-e", 8) == 0)
+  else if (strncmp (argv[0], "opaque-as", 9) == 0)
     type = OSPF_OPAQUE_AS_LSA;
 #endif /* HAVE_OPAQUE_LSA */
   else
