diff -ur rt2500-1.1.0_orig/Module/rtmp_main.c rt2500-1.1.0/Module/rtmp_main.c
--- rt2500-1.1.0_orig/Module/rtmp_main.c	2005-02-22 11:13:39.000000000 +0100
+++ rt2500-1.1.0/Module/rtmp_main.c	2005-06-21 04:45:31.000000000 +0200
@@ -100,7 +100,7 @@
 
     printk("%s %s %s http://rt2x00.serialmonkey.com\n", KERN_INFO DRV_NAME, DRV_VERSION, DRV_RELDATE);
 
-    print_name = pPci_Dev ? pPci_Dev->slot_name : "rt2500";
+    print_name = pPci_Dev ? pci_name(pPci_Dev) : "rt2500";
 
     // alloc_etherdev() will set net_dev->name
     net_dev = alloc_etherdev(sizeof(RTMP_ADAPTER));
