diff -ur rt2500-1.1.0_orig/Module/Makefile rt2500-1.1.0/Module/Makefile
--- rt2500-1.1.0_orig/Module/Makefile	2005-02-22 11:13:39.000000000 +0100
+++ rt2500-1.1.0/Module/Makefile	2005-06-17 07:04:37.000000000 +0200
@@ -62,7 +62,7 @@
  KERNEL_SOURCES := /lib/modules/$(shell uname -r)/build
 endif
 
-MODULE_ROOT := /lib/modules/$(shell uname -r)/kernel/drivers/net/wireless
+MODULE_ROOT := /lib/modules/$(KERNVER)/kernel/drivers/net/wireless
 
 src ?= .
 obj ?= .
