--- rt2500-1.1.0_orig/Module/Makefile	2005-02-22 11:13:39.000000000 +0100
+++ rt2500-1.1.0/Module/Makefile	2005-07-19 21:08:33.532024816 +0200
@@ -106,7 +106,6 @@
 	@echo "install '$(MODULE_OBJECT)' to $(MODULE_ROOT)"
 	install -m 755 -o 0 -g 0 -d $(MODULE_ROOT)
 	install -m 644 -o 0 -g 0 $(MODULE_OBJECT) $(MODULE_ROOT)
-	/sbin/depmod -a
 	
 	@if ! grep -q 'ra0' /etc/modules.conf ; then \
 		echo "append 'alias ra0 rt2500' to /etc/modules.conf"; \
