--- procps-3.1.14/Makefile.orig	2004-08-25 18:41:35.297550360 +0200
+++ procps-3.1.14/Makefile	2004-08-25 18:41:07.463781736 +0200
@@ -51,7 +51,7 @@
             $(usr/proc/bin)pgrep $(usr/proc/bin)pkill
 
 MANFILES := $(man1)uptime.1 $(man1)tload.1 $(man1)free.1 $(man1)w.1 \
-            $(man1)top.1 $(man1)watch.1 $(man1)skill.1 $(man1)kill.1 \
+            $(man1)top.1 $(man1)watch.1 $(man1)skill.1 \
             $(man1)snice.1 $(man1)pgrep.1 $(man1)pkill.1 $(man1)pmap.1 \
             $(man5)sysctl.conf.5 $(man8)vmstat.8 $(man8)sysctl.8
 
