without HAVE_LSTAT cpio will always dereference symlinks
 - Alejandro

--- ./configure.ac.orig	2006-06-25 15:58:15.000000000 +0200
+++ ./configure.ac	2006-06-25 15:57:56.000000000 +0200
@@ -92,7 +92,7 @@
   AC_CONFIG_LINKS(src/fnmatch.h:headers/fnmatch.h)
 fi
 
-AC_CHECK_FUNCS(lchown endpwent endgrent)
+AC_CHECK_FUNCS(lchown endpwent endgrent lstat)
 AC_FUNC_VPRINTF
 AC_FUNC_ALLOCA
 AC_CHECK_LIB(nsl, gethostname, [LIBS="$LIBS -lnsl"])
