diff -ur sharpconstruct-0.09_orig/src/Brush.cc sharpconstruct-0.09/src/Brush.cc
--- sharpconstruct-0.09_orig/src/Brush.cc	2005-06-20 04:42:12.000000000 +0200
+++ sharpconstruct-0.09/src/Brush.cc	2005-06-23 12:13:12.354247192 +0200
@@ -30,7 +30,7 @@
 
 	// FIXME!
 	_alpha = new Texture();
-	_alpha->Load( "/home/muvlo/SharpConstruct/textures/roundalpha.BMP" );
+	_alpha->Load( "/usr/share/sharpconstruct/brushes/Default.png" );
 }
 
 Brush::~Brush()
