
Speedup.

  - Rene Rebe <rene@exactcode.de>

--- sed-1.4-orig/sedexec.c	2003-12-29 22:36:33.000000000 +0100
+++ sed-1.4/sedexec.c	2005-06-26 23:00:22.000000000 +0200
@@ -127,8 +129,7 @@
 		/* here's where the transformed line is output */
 		if (!nflag && !delete)
 		{
-			for(p1 = linebuf; p1 < spend; p1++)
-				putc(*p1, stdout);
+			fwrite(linebuf, spend - linebuf, 1, stdout);
 			putc('\n', stdout);
 		}
