--- ./Makefile.orig	2004-06-27 01:18:28.370283816 +0200
+++ ./Makefile	2004-06-27 01:25:31.611941304 +0200
@@ -1,5 +1,6 @@
 CC = gcc
 CFLAGS = -O2 -Wall
+PREFIX = /usr
 
 all: elftoaout
 
@@ -7,8 +8,8 @@
 	$(CC) $(CFLAGS) -o elftoaout elftoaout.c
 
 install: elftoaout
-	install -m 0755 -s elftoaout /usr/bin/elftoaout
-	install -m 0644 elftoaout.1 /usr/man/man1/elftoaout.1
+	install -m 0755 -s elftoaout $(PREFIX)/bin/elftoaout
+	install -m 0644 elftoaout.1 $(PREFIX)/man/man1/elftoaout.1
 
 clean:
 	rm -f *~ *.[soa] core
