--- pmud-0.10/Makefile.orig	2003-05-15 00:01:04.000000000 +0200
+++ pmud-0.10/Makefile	2003-05-15 00:02:55.000000000 +0200
@@ -52,9 +52,7 @@
 	install -c xmouse /usr/X11R6/bin
 	install -c Batmon /usr/bin
 	install -c -D pwrctl /etc/power/pwrctl
-	cp pmud.rc /etc/rc.d/init.d/pmud
-	ln -s /sbin/snooze /usr/bin/apm
-	chkconfig --add pmud
+	ln -sf /sbin/snooze /usr/bin/apm
 
 clean:
 	rm -f $(PROGS) *.o
