--- ./mono/DBusType/Array.cs.orig	2005-03-19 18:03:55.342314760 -0400
+++ ./mono/DBusType/Array.cs	2005-03-19 18:04:19.056709624 -0400
@@ -59,7 +59,7 @@
       IntPtr arrayIter = Marshal.AllocCoTaskMem (Arguments.DBusMessageIterSize);
 
       if (!dbus_message_iter_open_container (iter,
-					     (int) this.Code,
+					     (int) Array.Code,
 					     Arguments.GetCodeAsString (elementType),
 					     arrayIter)) {
 	throw new ApplicationException("Failed to append array argument: " + val);
