--- ./second/Makefile.orig	2004-06-28 23:34:28.348996456 +0200
+++ ./second/Makefile	2004-06-28 23:34:36.316785168 +0200
@@ -85,7 +85,7 @@
 	$(CC) $(CFLAGS) -c -o $@ $<
 
 util:	util.c
-	$(CC) -DSMALL_RELOC=$(SMALL_RELOC) -DLARGE_RELOC=$(LARGE_RELOC) -o $@ $<
+	$(HOST_CC) -DSMALL_RELOC=$(SMALL_RELOC) -DLARGE_RELOC=$(LARGE_RELOC) -o $@ $<
 
 clean:
 	$(RM) *.o fs/*.o second* silotftp* util fs/libfs.a
--- ./common/Makefile.orig	2004-06-28 23:32:28.987142216 +0200
+++ ./common/Makefile	2004-06-28 23:33:46.917295032 +0200
@@ -12,12 +12,15 @@
 
 OBJS = sdiv.o rem.o udiv.o urem.o jmp.o printf.o console.o prom.o tree.o stringops2.o\
 	stringops1.o ffs.o divdi3.o udivdi3.o
-PROGRAMS = bin2h
 
-all: $(OBJS) $(PROGRAMS)
+bin2h:
+	$(HOST_CC) $(CFLAGS) -o bin2h bin2h.c
+
+all: $(OBJS) bin2h
 
 prom.o: prom.c
 	$(CC) $(CFLAGS) -c -Wa,-Av9 -o prom.o prom.c
 
 clean:
 	$(RM) $(PROGRAMS) *.o
+
