--- ./common/udivdi3.S.orig	2004-06-28 21:23:18.727522472 +0200
+++ ./common/udivdi3.S	2004-06-28 21:24:00.399187416 +0200
@@ -17,6 +17,8 @@
 the Free Software Foundation, 59 Temple Place - Suite 330,
 Boston, MA 02111-1307, USA.  */
 
+	.register %g2,#scratch
+
 	.text
 	.align 4
 	.globl __udivdi3
--- ./second/muldi3.S.orig	2004-06-28 21:29:22.416233384 +0200
+++ ./second/muldi3.S	2004-06-28 21:29:44.663851232 +0200
@@ -17,6 +17,8 @@
 the Free Software Foundation, 59 Temple Place - Suite 330,
 Boston, MA 02111-1307, USA.  */
 
+	.register %g2,#scratch
+
 	.text
 	.align 4
 	.globl __muldi3
--- ./common/divdi3.S.orig	2004-06-28 21:21:26.144637672 +0200
+++ ./common/divdi3.S	2004-06-28 21:21:52.257667888 +0200
@@ -17,6 +17,9 @@
 the Free Software Foundation, 59 Temple Place - Suite 330,
 Boston, MA 02111-1307, USA.  */
 
+	.register %g3,#scratch
+	.register %g2,#scratch
+ 
 	.data
 	.align 8
 	.globl	__clz_tab
