diff -Nur MPlayer-0.92/mplayer.c MPlayer-0.92.sj/mplayer.c
--- MPlayer-0.92/mplayer.c	2003-08-09 16:12:39.000000000 +0200
+++ MPlayer-0.92.sj/mplayer.c	2003-11-11 00:27:58.000000000 +0100
@@ -980,8 +980,8 @@
   if(!nortc)
   {
     // seteuid(0); /* Can't hurt to try to get root here */
-    if ((rtc_fd = open("/dev/rtc", O_RDONLY)) < 0)
-	mp_msg(MSGT_CPLAYER, MSGL_WARN, "Failed to open /dev/rtc: %s (mplayer should be setuid root or /dev/rtc should be readable by the user.)\n", strerror(errno));
+    if ((rtc_fd = open("/dev/misc/rtc", O_RDONLY)) < 0)
+	mp_msg(MSGT_CPLAYER, MSGL_WARN, "Failed to open /dev/misc/rtc: %s (mplayer should be setuid root or /dev/misc/rtc should be readable by the user.)\n", strerror(errno));
      else {
 	unsigned long irqp = 1024; /* 512 seemed OK. 128 is jerky. */
 
