--- j2re-1.4.2-01-linux-i586.bin	2005-03-24 12:03:57.000000000 +0100
+++ j2re-1.4.2-01-linux-i586.bin	2005-03-24 12:26:21.000000000 +0100
@@ -3,7 +3,7 @@
 # This script was generated using Makeself 1.5.5-Blackdown-j2se with modifications from Blackdown (source code is available from Blackdown mirrors)
 TAR_CMD=tar
 CRCsum=1315459154
-MD5=e2dd4b31085b43d2deae62f22ba14bcc
+MD5=\"00000000000000000000000000000000\"; SETUP_NOCHECK=1
 PATH=/usr/bin:/bin
 label="Blackdown Java 2 Standard Edition Runtime Environment v1.4.2-01 for Linux/i586"
 NO_DOT_VERSION=142
@@ -112,7 +112,7 @@
         fi
     fi
 fi
-more << EOF_LICENSE
+agreed=1; cat << EOF_LICENSE
 Sun Microsystems, Inc. Binary Code License Agreement
 for the JAVATM 2 RUNTIME ENVIRONMENT (J2RE), STANDARD
 EDITION, VERSION 1.4.2_X
