--- disktype/linux.c.vanilla	2026-05-24 14:21:02.902426705 +0200
+++ disktype/linux.c	2026-05-24 14:21:47.794686292 +0200
@@ -579,7 +579,7 @@
 
       major = get_ve_short(en, buf + 28);
       minor = get_ve_short(en, buf + 30);
-      print_line(level, "Linux squashfs, version %d.%d, %s",
+      print_line(level, "Squashfs file system, version %d.%d, %s",
                  major, minor, get_ve_name(en));
 
       if (major > 2)
