I don't know why, but it works (TM)
	-- jsaw

diff -ur kdebase-3.4.2/khotkeys/kcontrol/ui/action_list_widget_ui.ui kdebase-3.4.2-jsaw/khotkeys/kcontrol/ui/action_list_widget_ui.ui
--- kdebase-3.4.2/khotkeys/kcontrol/ui/action_list_widget_ui.ui	2005-05-23 14:14:50.000000000 +0200
+++ kdebase-3.4.2-jsaw/khotkeys/kcontrol/ui/action_list_widget_ui.ui	2005-09-20 21:05:28.007645649 +0200
@@ -169,6 +169,6 @@
 <layoutdefaults spacing="6" margin="11"/>
 <layoutfunctions spacing="KDialog::spacingHint" margin="KDialog::marginHint"/>
 <includehints>
-    <includehint>khotkeys::khlistview.h</includehint>
+    <includehint>khlistview.h</includehint>
 </includehints>
 </UI>
diff -ur kdebase-3.4.2/khotkeys/kcontrol/ui/actions_listview_widget_ui.ui kdebase-3.4.2-jsaw/khotkeys/kcontrol/ui/actions_listview_widget_ui.ui
--- kdebase-3.4.2/khotkeys/kcontrol/ui/actions_listview_widget_ui.ui	2005-05-23 14:14:50.000000000 +0200
+++ kdebase-3.4.2-jsaw/khotkeys/kcontrol/ui/actions_listview_widget_ui.ui	2005-09-20 21:05:47.964290573 +0200
@@ -80,7 +80,4 @@
 </includes>
 <layoutdefaults spacing="6" margin="11"/>
 <layoutfunctions spacing="KDialog::spacingHint" margin="KDialog::marginHint"/>
-<includehints>
-    <includehint>khotkeys::actions_listview.h</includehint>
-</includehints>
 </UI>
diff -ur kdebase-3.4.2/khotkeys/kcontrol/ui/condition_list_widget_ui.ui kdebase-3.4.2-jsaw/khotkeys/kcontrol/ui/condition_list_widget_ui.ui
--- kdebase-3.4.2/khotkeys/kcontrol/ui/condition_list_widget_ui.ui	2005-05-23 14:14:50.000000000 +0200
+++ kdebase-3.4.2-jsaw/khotkeys/kcontrol/ui/condition_list_widget_ui.ui	2005-09-20 21:05:28.011644977 +0200
@@ -168,6 +168,6 @@
 <layoutdefaults spacing="6" margin="11"/>
 <layoutfunctions spacing="KDialog::spacingHint" margin="KDialog::marginHint"/>
 <includehints>
-    <includehint>khotkeys::khlistview.h</includehint>
+    <includehint>khlistview.h</includehint>
 </includehints>
 </UI>
diff -ur kdebase-3.4.2/khotkeys/kcontrol/ui/gestures_settings_tab_ui.ui kdebase-3.4.2-jsaw/khotkeys/kcontrol/ui/gestures_settings_tab_ui.ui
--- kdebase-3.4.2/khotkeys/kcontrol/ui/gestures_settings_tab_ui.ui	2005-05-23 14:14:50.000000000 +0200
+++ kdebase-3.4.2-jsaw/khotkeys/kcontrol/ui/gestures_settings_tab_ui.ui	2005-09-20 21:05:29.105461090 +0200
@@ -157,6 +157,6 @@
 <includehints>
     <includehint>knuminput.h</includehint>
     <includehint>knuminput.h</includehint>
-    <includehint>khotkeys::windowdef_list_widget.h</includehint>
+    <includehint>windowdef_list_widget.h</includehint>
 </includehints>
 </UI>
diff -ur kdebase-3.4.2/khotkeys/kcontrol/ui/keyboard_input_widget_ui.ui kdebase-3.4.2-jsaw/khotkeys/kcontrol/ui/keyboard_input_widget_ui.ui
--- kdebase-3.4.2/khotkeys/kcontrol/ui/keyboard_input_widget_ui.ui	2005-05-23 14:14:50.000000000 +0200
+++ kdebase-3.4.2-jsaw/khotkeys/kcontrol/ui/keyboard_input_widget_ui.ui	2005-09-20 21:05:29.108460586 +0200
@@ -189,6 +189,6 @@
 <layoutfunctions spacing="KDialog::spacingHint" margin="KDialog::marginHint"/>
 <includehints>
     <includehint>ktextedit.h</includehint>
-    <includehint>khotkeys::windowdef_list_widget.h</includehint>
+    <includehint>windowdef_list_widget.h</includehint>
 </includehints>
 </UI>
diff -ur kdebase-3.4.2/khotkeys/kcontrol/ui/triggers_tab_ui.ui kdebase-3.4.2-jsaw/khotkeys/kcontrol/ui/triggers_tab_ui.ui
--- kdebase-3.4.2/khotkeys/kcontrol/ui/triggers_tab_ui.ui	2005-05-23 14:14:50.000000000 +0200
+++ kdebase-3.4.2-jsaw/khotkeys/kcontrol/ui/triggers_tab_ui.ui	2005-09-20 21:05:29.112459914 +0200
@@ -169,6 +169,6 @@
 <layoutdefaults spacing="6" margin="11"/>
 <layoutfunctions spacing="KDialog::spacingHint" margin="KDialog::marginHint"/>
 <includehints>
-    <includehint>khotkeys::khlistview.h</includehint>
+    <includehint>khlistview.h</includehint>
 </includehints>
 </UI>
diff -ur kdebase-3.4.2/khotkeys/kcontrol/ui/windowdef_list_widget_ui.ui kdebase-3.4.2-jsaw/khotkeys/kcontrol/ui/windowdef_list_widget_ui.ui
--- kdebase-3.4.2/khotkeys/kcontrol/ui/windowdef_list_widget_ui.ui	2005-05-23 14:14:50.000000000 +0200
+++ kdebase-3.4.2-jsaw/khotkeys/kcontrol/ui/windowdef_list_widget_ui.ui	2005-09-20 21:05:29.115459410 +0200
@@ -168,6 +168,6 @@
 <layoutdefaults spacing="6" margin="11"/>
 <layoutfunctions spacing="KDialog::spacingHint" margin="KDialog::marginHint"/>
 <includehints>
-    <includehint>khotkeys::khlistview.h</includehint>
+    <includehint>khlistview.h</includehint>
 </includehints>
 </UI>
