--- u-boot-2021.07/arch/riscv/dts/fu740-c000-u-boot.dtsi.vanilla	2021-11-14 17:13:35.562301035 +0100
+++ u-boot-2021.07/arch/riscv/dts/fu740-c000-u-boot.dtsi	2021-11-14 17:13:45.414300642 +0100
@@ -8,7 +8,7 @@
 / {
 	cpus {
 		assigned-clocks = <&prci PRCI_CLK_COREPLL>;
-		assigned-clock-rates = <1200000000>;
+		assigned-clock-rates = <1400000000>;
 		u-boot,dm-spl;
 		cpu0: cpu@0 {
 			clocks = <&prci PRCI_CLK_COREPLL>;
--- u-boot-2021.07/arch/riscv/dts/fu740-c000-u-boot.dtsi.vanilla	2021-11-14 22:20:08.037567674 +0100
+++ u-boot-2021.07/arch/riscv/dts/fu740-c000-u-boot.dtsi	2021-11-15 10:17:17.083851981 +0100
@@ -77,7 +77,7 @@
 			       0x0 0x100b2000 0x0 0x2000
 			       0x0 0x100b8000 0x0 0x1000>;
 			clocks = <&prci PRCI_CLK_DDRPLL>;
-			clock-frequency = <933333324>;
+			clock-frequency = <933333324>;
 			u-boot,dm-spl;
 		};
 	};
