--- ./configure	2003-02-07 20:54:10.000000000 +0100
+++ ./configure	2003-04-10 23:06:26.000000000 +0200
@@ -859,13 +859,13 @@
 # Check whether --with-mysql or --without-mysql was given.
 if test "${with_mysql+set}" = set; then
   withval="$with_mysql"
-  mysqlheadername="$withval"
+  mysqlheadername="$withval/include/mysql"
 fi
 
 # Check whether --with-pgsql or --without-pgsql was given.
 if test "${with_pgsql+set}" = set; then
   withval="$with_pgsql"
-  pgsqlheadername="$withval"
+  pgsqlheadername="$withval/include"
 fi
 
 
