--- ./xjdclient.c.old	2003-03-30 20:04:13.000000000 +0100
+++ ./xjdclient.c	2003-03-30 20:05:12.000000000 +0100
@@ -36,6 +36,7 @@
 #include <stdlib.h>
 #include <unistd.h>
 #include <signal.h>
+#include <errno.h>
 #include "xjdic.h"
 
 #define CVERBOSE 0
--- ./xjdserver.c.old	2003-03-30 20:04:22.000000000 +0100
+++ ./xjdserver.c	2003-03-30 20:05:31.000000000 +0100
@@ -28,6 +28,7 @@
 #include <netdb.h>
 #include <stdlib.h>
 #include <signal.h>
+#include <errno.h>
 #include "xjdic.h"
 
 #define SVERBOSE 0
