--- ./linc2/src/Makefile.in.orig	2005-04-04 10:10:13.888399912 -0400
+++ ./linc2/src/Makefile.in	2005-04-04 10:11:10.581557008 -0400
@@ -189,7 +189,7 @@
 noinst_LTLIBRARIES = liblinc.la
 
 #lib_LTLIBRARIES=liblinc.la
-@OS_WIN32_FALSE@bin_PROGRAMS = linc-cleanup-sockets
+@OS_WIN32_FALSE@bin_PROGRAMS = 
 
 #	-I$(top_builddir)/include  
 #	-I$(top_srcdir)/include    
@@ -241,7 +241,7 @@
 	linc-protocols.lo linc-server.lo linc-source.lo
 liblinc_la_OBJECTS = $(am_liblinc_la_OBJECTS)
 @OS_WIN32_TRUE@bin_PROGRAMS =
-@OS_WIN32_FALSE@bin_PROGRAMS = linc-cleanup-sockets$(EXEEXT)
+@OS_WIN32_FALSE@bin_PROGRAMS = 
 PROGRAMS = $(bin_PROGRAMS)
 
 am_linc_cleanup_sockets_OBJECTS = cleanup.$(OBJEXT)
