--- ORBit2-2.12.0/ORBit-2.0.pc.in.orig	2004-10-04 18:01:31.193074864 +0200
+++ ORBit2-2.12.0/ORBit-2.0.pc.in	2004-10-04 18:01:42.845303456 +0200
@@ -8,6 +8,6 @@
 Name: ORBit-2.0
 Description: High-performance CORBA Object Request Broker.
 Version: @ORBIT_VERSION@
-Requires: glib-2.0 gmodule-2.0 gthread-2.0
+Requires: glib-2.0 gmodule-2.0 gobject-2.0 gthread-2.0
 Libs: -L${libdir} -lORBit-2 -lm
 Cflags: -I${includedir}/orbit-2.0 -DORBIT2=1
