diff -rduN bluez-utils-2.4-orig/rfcomm/Makefile.am bluez-utils-2.4/rfcomm/Makefile.am
--- bluez-utils-2.4-orig/rfcomm/Makefile.am	2003-07-18 11:25:50.000000000 +0200
+++ bluez-utils-2.4/rfcomm/Makefile.am	2004-01-02 21:05:24.964888336 +0100
@@ -2,7 +2,6 @@
 #  $Id: Makefile.am,v 1.5 2003/07/18 09:25:50 holtmann Exp $
 #
 
-mandir = $(prefix)/usr/share/man
 confdir = $(sysconfdir)/bluetooth
 
 bin_PROGRAMS = rfcomm
diff -rduN bluez-utils-2.4-orig/rfcomm/Makefile.in bluez-utils-2.4/rfcomm/Makefile.in
--- bluez-utils-2.4-orig/rfcomm/Makefile.in	2003-12-18 19:36:54.000000000 +0100
+++ bluez-utils-2.4/rfcomm/Makefile.in	2004-01-02 21:05:46.092676424 +0100
@@ -89,7 +89,7 @@
 VERSION = @VERSION@
 YACC = @YACC@
 
-mandir = $(prefix)/usr/share/man
+mandir = @mandir@
 confdir = $(sysconfdir)/bluetooth
 
 bin_PROGRAMS = rfcomm
diff -rduN bluez-utils-2.4-orig/tools/Makefile.am bluez-utils-2.4/tools/Makefile.am
--- bluez-utils-2.4-orig/tools/Makefile.am	2003-07-03 00:27:41.000000000 +0200
+++ bluez-utils-2.4/tools/Makefile.am	2004-01-02 21:06:10.017039368 +0100
@@ -2,8 +2,6 @@
 #  $Id: Makefile.am,v 1.9 2003/07/02 22:27:41 holtmann Exp $
 #
 
-mandir = $(prefix)/usr/share/man
-
 sbin_PROGRAMS = hciattach hciconfig
 bin_PROGRAMS  = hcitool l2ping
 
diff -rduN bluez-utils-2.4-orig/tools/Makefile.in bluez-utils-2.4/tools/Makefile.in
--- bluez-utils-2.4-orig/tools/Makefile.in	2003-12-18 19:36:53.000000000 +0100
+++ bluez-utils-2.4/tools/Makefile.in	2004-01-02 21:06:21.549286200 +0100
@@ -89,7 +89,7 @@
 VERSION = @VERSION@
 YACC = @YACC@
 
-mandir = $(prefix)/usr/share/man
+mandir = @mandir@
 
 sbin_PROGRAMS = hciattach hciconfig
 bin_PROGRAMS = hcitool l2ping
