diff -rduN ./rfcomm/Makefile.am ../bluez-utils-2.3/rfcomm/Makefile.am
--- ./rfcomm/Makefile.am	2002-12-21 04:07:42.000000000 +0100
+++ ../bluez-utils-2.3/rfcomm/Makefile.am	2003-07-14 19:21:36.000000000 +0200
@@ -2,7 +2,6 @@
 #  $Id: Makefile.am,v 1.4 2002/10/17 22:15:48 holtmann Exp $
 #
 
-mandir = $(prefix)/usr/share/man
 confdir = /etc/bluetooth
 
 bin_PROGRAMS = rfcomm
diff -rduN ./rfcomm/Makefile.in ../bluez-utils-2.3/rfcomm/Makefile.in
--- ./rfcomm/Makefile.in	2003-03-20 07:40:30.000000000 +0100
+++ ../bluez-utils-2.3/rfcomm/Makefile.in	2003-07-14 19:26:05.000000000 +0200
@@ -89,7 +89,7 @@
 VERSION = @VERSION@
 YACC = @YACC@
 
-mandir = $(prefix)/usr/share/man
+mandir = @mandir@
 confdir = /etc/bluetooth
 
 bin_PROGRAMS = rfcomm
diff -rduN ./tools/Makefile.am ../bluez-utils-2.3/tools/Makefile.am
--- ./tools/Makefile.am	2002-12-21 04:07:42.000000000 +0100
+++ ../bluez-utils-2.3/tools/Makefile.am	2003-07-14 19:26:58.000000000 +0200
@@ -2,8 +2,6 @@
 #  $Id: Makefile.am,v 1.8 2002/11/16 01:44:53 holtmann Exp $
 #
 
-mandir = $(prefix)/usr/share/man
-
 sbin_PROGRAMS = hciattach hciconfig
 bin_PROGRAMS  = hcitool l2ping
 
diff -rduN ./tools/Makefile.in ../bluez-utils-2.3/tools/Makefile.in
--- ./tools/Makefile.in	2003-03-20 07:40:30.000000000 +0100
+++ ../bluez-utils-2.3/tools/Makefile.in	2003-07-14 19:26:30.000000000 +0200
@@ -89,7 +89,7 @@
 VERSION = @VERSION@
 YACC = @YACC@
 
-mandir = $(prefix)/usr/share/man
+mandir = @mandir@
 
 sbin_PROGRAMS = hciattach hciconfig
 bin_PROGRAMS = hcitool l2ping
