--- ./hcid/Makefile.am.orig	2003-07-14 17:58:00.000000000 +0200
+++ ./hcid/Makefile.am	2003-07-14 17:58:22.000000000 +0200
@@ -16,7 +16,7 @@
 #
 # Install configuration files
 # 
-confdir   = $(prefix)/etc/bluetooth
+confdir   = $(sysconfdir)/bluetooth
 conf_FILE = $(confdir)/$(hcid_CONFIG)
 pin_FILE  = $(confdir)/pin
 
--- ./hcid/Makefile.in.orig	2003-07-14 18:23:12.000000000 +0200
+++ ./hcid/Makefile.in	2003-07-14 18:23:25.000000000 +0200
@@ -104,7 +104,7 @@
 #
 # Install configuration files
 # 
-confdir = $(prefix)/etc/bluetooth
+confdir = $(sysconfdir)/bluetooth
 conf_FILE = $(confdir)/$(hcid_CONFIG)
 pin_FILE = $(confdir)/pin
 mkinstalldirs = $(SHELL) $(top_srcdir)/mkinstalldirs
