diff -dur sodipodi-0.31.1/src/libnrtype/Makefile.in sodipodi-0.31.1-new/src/libnrtype/Makefile.in
--- sodipodi-0.31.1/src/libnrtype/Makefile.in	2003-02-24 09:41:55.000000000 +0000
+++ sodipodi-0.31.1-new/src/libnrtype/Makefile.in	2003-07-20 23:40:43.000000000 +0000
@@ -124,6 +124,7 @@
 
 INCLUDES = \
 	-I.. \
+	-I/usr/X11R6/include/freetype2 \
 	$(SODIPODI_CFLAGS) \
 	$(GNOME_PRINT_CFLAGS) \
 	$(FONTCONFIG_CFLAGS)
