--- ./boehm-gc/include/Makefile.in.orig	2003-07-23 23:26:22.000000000 -0400
+++ ./boehm-gc/include/Makefile.in	2003-07-23 23:37:07.000000000 -0400
@@ -120,13 +120,7 @@
 
 
 install-includeHEADERS: $(include_HEADERS)
-	@$(NORMAL_INSTALL)
-	$(mkinstalldirs) $(DESTDIR)$(includedir)
-	@list='$(include_HEADERS)'; for p in $$list; do \
-	  if test -f "$$p"; then d= ; else d="$(srcdir)/"; fi; \
-	  echo " $(INSTALL_DATA) $$d$$p $(DESTDIR)$(includedir)/$$p"; \
-	  $(INSTALL_DATA) $$d$$p $(DESTDIR)$(includedir)/$$p; \
-	done
+	@echo RockLinux: i will not install old headers if i install boehm-gc as package.
 
 uninstall-includeHEADERS:
 	@$(NORMAL_UNINSTALL)
