--- acx100-0.2.0pre8_plus_fixes_45/Configure	2004-11-23 09:00:58.000000000 +0000
+++ acx100-0.2.0pre8_plus_fixes_45-fixed/Configure	2005-02-21 02:06:18.869402736 +0000
@@ -9,12 +9,11 @@
 # some initial definitions
 
 CONFIG_MK=config.mk
-KERNEL_VER=`uname -r`
 KERNEL_BUILD=/lib/modules/$KERNEL_VER/build
 # If using 2.6 outputting to a seperate directory, make KERNEL_OUTPUT
 # point to that directory
 #KERNEL_OUTPUT=
-ECHO="`which echo` -e"
+ECHO="echo -e"
 WARNING_FLAG=0
 
 # functions
