--- ./class/ByteFX.Data/AssemblyInfo.cs.orig	2003-12-03 13:31:28.000000000 -0300
+++ ./class/ByteFX.Data/AssemblyInfo.cs	2003-12-03 13:33:40.000000000 -0300
@@ -54,5 +54,5 @@
 //       documentation for more information on this.
 //
 [assembly: AssemblyDelaySign(false)]
-[assembly: AssemblyKeyFile("..\\..\\bytefx.snk")]
+//[assembly: AssemblyKeyFile("..\\..\\bytefx.snk")]
 [assembly: AssemblyKeyName("")]
