--- parted-3.0/parted/ui.c.vanilla	2014-04-11 13:12:30.692102893 +0200
+++ parted-3.0/parted/ui.c	2014-04-11 13:12:50.904102194 +0200
@@ -1442,7 +1442,7 @@
 #ifdef HAVE_LIBREADLINE
   if (!opt_script_mode) {
     rl_initialize ();
-    rl_attempted_completion_function = (CPPFunction*) complete_function;
+    rl_attempted_completion_function = (rl_completion_func_t*) complete_function;
     readline_state.in_readline = 0;
   }
 #endif
