
The usual gcc-3.4 fixes ...

  - Rene Rebe <rene@rocklinux.org>

--- imlib2-2004-01-18-orig/src/rgbadraw.c	2003-07-13 03:09:17.000000000 +0200
+++ imlib2-2004-01-18/src/rgbadraw.c	2004-05-05 18:31:54.920199456 +0200
@@ -2358,6 +2358,7 @@
                        if (ps)
                           ps->next = s;
                      nospans:
+                       ;
                     }
                }
              if (i == y2)
@@ -2365,6 +2366,7 @@
              i += step;
           }
       nolines:
+        ;
      }
    for (i = 0; i < h; i++)
      {
