--- xf86-video-ati-7.0.0/src/radeon_kms.c.vanilla	2012-11-19 10:54:59.838343700 +0100
+++ xf86-video-ati-7.0.0/src/radeon_kms.c	2012-11-19 10:55:18.444347985 +0100
@@ -977,6 +977,10 @@
 {
     return xf86_cursors_init (pScreen, CURSOR_WIDTH, CURSOR_HEIGHT,
 			      (HARDWARE_CURSOR_TRUECOLOR_AT_8BPP |
+#if X_BYTE_ORDER == X_BIG_ENDIAN
+                              HARDWARE_CURSOR_NIBBLE_SWAPPED |
+#endif
+
 			       HARDWARE_CURSOR_AND_SOURCE_WITH_MASK |
 			       HARDWARE_CURSOR_SOURCE_MASK_INTERLEAVE_1 |
 			       HARDWARE_CURSOR_UPDATE_UNHIDDEN |
