
The usual gcc-3.4 fixes.

  - Rene Rebe <rene@exactcode.de>

--- libgnomedb-1.0.1/libgnomedb/gnome-db-form.c.orig	2004-05-10 15:26:15.192229272 +0200
+++ libgnomedb-1.0.1/libgnomedb/gnome-db-form.c	2004-05-10 15:26:58.463651016 +0200
@@ -534,6 +534,7 @@
 	case GNOME_DB_FORM_TYPE_NAVIGATOR :
 		return form->priv->navigator_current_row;
 	case GNOME_DB_FORM_TYPE_LIST :
+		;
 	}
 
 	return -1;
@@ -559,5 +560,6 @@
 		show_navigator_row (form);
 		break;
 	case GNOME_DB_FORM_TYPE_LIST :
+		;
 	}
 }
--- libgnomedb-1.0.1/libgnomedb/utils.c.orig	2004-05-10 15:23:28.018643520 +0200
+++ libgnomedb-1.0.1/libgnomedb/utils.c	2004-05-10 15:24:10.804139136 +0200
@@ -45,6 +45,7 @@
 		gnome_db_error_dialog_show_errors (GNOME_DB_ERROR_DIALOG (dialog),
 						   gda_connection_get_errors (cnc));
 	default :
+		;
 	}
 }
 
