--- ./keepalived/Makefile.in.orig	2003-07-07 23:52:32.000000000 -0400
+++ ./keepalived/Makefile.in	2003-07-08 00:08:40.000000000 -0400
@@ -80,8 +80,6 @@
 install:
 	install -d $(sbindir)
 	install -m 700 $(BIN)/$(EXEC) $(sbindir)/
-	install -d $(sysconfdir)/init.d
-	install -m 755 $(init_script) $(sysconfdir)/init.d/keepalived
-	install -d $(sysconfdir)/keepalived/samples
+	install -d $(sysconfdir)/rc.d/init.d
+	install -m 755 $(init_script) $(sysconfdir)/rc.d/init.d/keepalived
 	install -m 644 $(conf_file) $(sysconfdir)/keepalived/
-	install -m 644 ../doc/samples/* $(sysconfdir)/keepalived/samples/
