diff -ruN slmodem-2.9.8/drivers/Makefile slmodem-2.9.8-new/drivers/Makefile
--- slmodem-2.9.8/drivers/Makefile	2003-12-21 22:15:54.000000000 +0100
+++ slmodem-2.9.8-new/drivers/Makefile	2004-06-24 15:52:10.855671664 +0200
@@ -75,14 +75,14 @@
 	cp /etc/modules.conf /etc/modules.conf.slamr && \
 	    echo 'alias char-major-212 slamr' >> /etc/modules.conf && \
 	    echo 'alias char-major-213 slusb' >> /etc/modules.conf 
-	/sbin/depmod -a
+	echo 'ROCK: Please run /etc/cron.d/00-updates afterwards!'
 uninstall:
 	/sbin/modprobe -r slamr slusb
 	cp /etc/modules.conf /etc/modules.conf.slamr && \
 	   egrep -ve 'alias char-major-212 slamr|alias char-major-213 slusb' /etc/modules.conf.slamr > /etc/modules.conf
 	$(RM) $(module-dir)/slamr.o
 	$(RM) $(module-dir)/slusb.o
-	/sbin/depmod -a
+	echo 'ROCK: Please run /etc/cron.d/00-updates afterwards!'
 uninstall-old:
 	$(RM) $(module-dir)/slmdm.o \
               $(module-dir)/slfax.o \
@@ -102,13 +102,13 @@
 install:
 	install -D -m 644 slamr.ko $(module-dir)/slamr.ko
 	install -D -m 644 slusb.ko $(module-dir)/slusb.ko
-	/sbin/depmod -a
+	echo 'ROCK: Please run /etc/cron.d/00-updates afterwards!'
 uninstall:
 	modprobe -r slamr ; echo -n
 	modprobe -r slusb ; echo -n
 	$(RM) $(module-dir)/slamr.ko
 	$(RM) $(module-dir)/slusb.ko
-	/sbin/depmod -a
+	echo 'ROCK: Please run /etc/cron.d/00-updates afterwards!'
 endif
 
 endif
