diff -urN pixieplus.orig/kdenonbeta/pixieplus/mimetypes/Makefile.am pixieplus/kdenonbeta/pixieplus/mimetypes/Makefile.am
--- pixieplus.orig/kdenonbeta/pixieplus/mimetypes/Makefile.am	2005-03-29 14:47:27.437200680 +0200
+++ pixieplus/kdenonbeta/pixieplus/mimetypes/Makefile.am	2005-03-29 15:51:20.607470224 +0200
@@ -1,5 +1,5 @@
 
-mime_DATA = x-pcx.desktop x-xwd.desktop x-miff.desktop x-tga.desktop \
+mime_DATA = x-xwd.desktop x-miff.desktop x-tga.desktop \
 	x-pict.desktop
 
 mimedir = $(kde_mimedir)/image
