--- ./src/Makefile.in.orig	2003-02-08 13:51:11.000000000 +0100
+++ ./src/Makefile.in	2003-02-08 13:51:14.000000000 +0100
@@ -144,8 +144,10 @@
 suidbins = su
 suidubins = chage chfn chsh expiry gpasswd newgrp passwd
 
+# Add $(LIBPAM) to $(LDADD) because libmisc and libshadow are using some
+# functions provided by the PAM library.
 LDADD = $(top_builddir)/libmisc/libmisc.la \
-	$(top_builddir)/lib/libshadow.la
+	$(top_builddir)/lib/libshadow.la $(LIBPAM)
 
 
 chpasswd_LDADD = $(LDADD) $(LIBPAM)
