--- sharutils-4.2.1/po/Makefile.in.in~	1995-12-04 17:19:48.000000000 +0100
+++ sharutils-4.2.1/po/Makefile.in.in	2003-06-20 13:05:32.000000000 +0200
@@ -117,14 +117,14 @@
 install-exec:
 install-data: all
 	$(top_srcdir)/mkinstalldirs $(datadir)
-	catalogs='$(CATALOGS)'; \
+	catalogs='$(GMOFILES)'; \
 	for cat in $$catalogs; do \
 	  if test "`echo $$cat | sed 's/.*\(\..*\)/\1/'`" = ".gmo"; then \
 	    destdir=$(gnulocaledir); \
 	  else \
 	    destdir=$(localedir); \
 	  fi; \
-	  lang=`echo $$cat | sed 's/$(CATOBJEXT)$$//'`; \
+	  lang=`echo $$cat | sed 's/\.[^.]*$$//'`; \
 	  dir=$$destdir/$$lang/LC_MESSAGES; \
 	  $(top_srcdir)/mkinstalldirs $$dir; \
 	  if test -r $$cat; then \
