--- bsd-games-2.16/dm/Makefrag.orig	2005-08-02 19:14:38.000000000 -0700
+++ bsd-games-2.16/dm/Makefrag	2005-08-02 19:14:54.000000000 -0700
@@ -27,7 +27,7 @@
 # SUCH DAMAGE.
 
 # Add -DLOG if you want logging
-dm_DEFS := $(GETLOADAVG_DEFS) -D_GNU_SOURCE -DSUPPORT_UTMPX -DSUPPORT_UTMP # -DLOG
+dm_DEFS := $(GETLOADAVG_DEFS) -D_GNU_SOURCE -DSUPPORT_UTMP # -DLOG
 dm_DIRS := $(GAMESDIR) $(MAN8DIR) $(MAN5DIR)
 
 dm_all:	dm/dm dm/dm.8 dm/dm.conf.5
