--- doxygen-1.3.9/src/config.l.orig	2004-10-12 11:26:04.676878072 +0200
+++ doxygen-1.3.9/src/config.l	2004-10-12 11:27:04.547776312 +0200
@@ -34,6 +34,10 @@
 
 #include "lang_cfg.h"
 
+#ifdef __cplusplus
+extern "C" int isatty(int fd);
+#endif
+
 #undef Config_getString
 #undef Config_getInt
 #undef Config_getList

--- doxygen-1.3.9/addon/doxywizard/config.l.orig	2004-10-12 11:26:04.676878072 +0200
+++ doxygen-1.3.9/addon/doxywizard/config.l	2004-10-12 11:27:04.547776312 +0200
@@ -34,6 +34,10 @@
 
 #include "lang_cfg.h"
 
+#ifdef __cplusplus
+extern "C" int isatty(int fd);
+#endif
+
 #undef Config_getString
 #undef Config_getInt
 #undef Config_getList
