--- bash-3.0/variables.c	2005-01-03 23:28:24.496779416 +0100
+++ bash-3.0-t2/variables.c	2005-01-04 00:01:47.485278880 +0100
@@ -986,6 +986,7 @@
       v = bind_variable (var, (val)); \
       v->dynamic_value = gfunc; \
       v->assign_func = afunc; \
+      VSETATTR(v, att_nounset|att_noassign); \
     } \
   while (0)
 
@@ -995,6 +996,7 @@
       v = make_new_array_variable (var); \
       v->dynamic_value = gfunc; \
       v->assign_func = afunc; \
+      VSETATTR(v, att_nounset|att_noassign); \
     } \
   while (0)
 
@@ -1386,7 +1388,7 @@
 #else
   INIT_DYNAMIC_VAR ("FUNCNAME", (char *)NULL, get_funcname, null_assign);
 #endif
-  VSETATTR (v, att_invisible|att_noassign);
+  VSETATTR (v, att_invisible|att_noassign|att_nounset);
   return v;
 }
 
