--- LibVNCServer-0.7.1/Makefile.in.vanilla	2005-10-05 18:15:43.000000000 +0200
+++ LibVNCServer-0.7.1/Makefile.in	2005-10-05 18:15:56.000000000 +0200
@@ -86,8 +86,8 @@
 with_ffmpeg = @with_ffmpeg@
 CFLAGS = -g -Wall
 
-SUBDIRS = libvncserver examples contrib x11vnc vncterm classes libvncclient client_examples test
-DIST_SUBDIRS = libvncserver examples contrib x11vnc vncterm classes libvncclient client_examples test
+SUBDIRS = libvncserver examples contrib vncterm classes libvncclient client_examples test
+DIST_SUBDIRS = libvncserver examples contrib vncterm classes libvncclient client_examples test
 
 bin_SCRIPTS = libvncserver-config
 
