--- gnome-vfs-2.8.0/gnome-vfs-2.0.pc.in~	2002-01-31 05:28:20.000000000 +0100
+++ gnome-vfs-2.8.0/gnome-vfs-2.0.pc.in	2004-10-04 15:21:23.849613072 +0200
@@ -7,6 +7,6 @@
 Name: gnome-vfs
 Description: The GNOME virtual file-system libraries
 Version: @VERSION@
-Requires: bonobo-activation-2.0,gthread-2.0,gmodule-2.0
+Requires: bonobo-activation-2.0 libbonobo-2.0 gconf-2.0 gthread-2.0 gmodule-2.0
 Libs: -L${libdir} -lgnomevfs-2
 Cflags: -I${includedir}/gnome-vfs-2.0 -I${libdir}/gnome-vfs-2.0/include
