diff -ruN jack-audio-connection-kit-0.61.0.org/Makefile.in jack-audio-connection-kit-0.61.0/Makefile.in
--- jack-audio-connection-kit-0.61.0.org/Makefile.in	2003-03-05 03:19:08.000000000 +0100
+++ jack-audio-connection-kit-0.61.0/Makefile.in	2003-03-14 16:51:36.000000000 +0100
@@ -170,9 +170,9 @@
 
 @HAVE_DOXYGEN_TRUE@DOC_DIR = doc
 
-SUBDIRS = jack libjack jackd drivers example-clients $(DOC_DIR)
+SUBDIRS = jack libjack jackd drivers $(DOC_DIR)
 
-DIST_SUBDIRS = jack libjack jackd drivers example-clients doc
+DIST_SUBDIRS = jack libjack jackd drivers doc
 
 pkgconfigdir = $(libdir)/pkgconfig
 pkgconfig_DATA = jack.pc
