--- koffice-1.2.90//kivio/kiviopart/kiviostencilsetaction.cpp.orig	2003-04-09 11:16:24.000000000 +0200
+++ koffice-1.2.90/kivio/kiviopart/kiviostencilsetaction.cpp	2003-04-29 14:12:00.000000000 +0200
@@ -36,7 +36,8 @@
 
 #include "kivio_stencil_spawner_set.h"
 
-static const char * default_plug_xpm[] = {
+/* this is only a make it compile hack ... */
+static const char * default_plug_xpm2[] = {
 "16 16 5 1",
 " 	c None",
 ".	c #000000",
