diff -BurN ./ext/mysql/libmysql.org/my_sys.h ./ext/mysql/libmysql/my_sys.h
--- ./ext/mysql/libmysql.org/my_sys.h	Fri Jun  7 17:07:54 2002
+++ ./ext/mysql/libmysql/my_sys.h	Wed Mar 19 16:58:01 2003
@@ -148,8 +148,4 @@
 
-#ifdef HAVE_ERRNO_AS_DEFINE
 #include <errno.h>			/* errno is a define */
-#else
-extern int errno;			/* declare errno */
-#endif
 extern const char ** NEAR my_errmsg[];
 extern char NEAR errbuff[NRERRBUFFS][ERRMSGSIZE];
