--- plasma-workspace-6.6.91/shell/desktopview.cpp.vanilla	2026-06-11 20:18:57.985766724 +0200
+++ plasma-workspace-6.6.91/shell/desktopview.cpp	2026-06-11 20:21:23.230379154 +0200
@@ -84,7 +84,7 @@
         }
     });
 
-#if PROJECT_VERSION_PATCH >= 80 || PROJECT_VERSION_MINOR >= 80
+#if 0
     m_showPreviewBanner = KConfigGroup(KSharedConfig::openConfig(u"kdeglobals"_s), u"General"_s).readEntry("ShowDesktopPreviewBanner", true);
     Q_EMIT showPreviewBannerChanged();
 #endif
@@ -191,7 +191,7 @@
     Q_EMIT accentColorChanged(Qt::transparent);
 }
 
-#if PROJECT_VERSION_PATCH >= 80 || PROJECT_VERSION_MINOR >= 80
+#if 0
 #include <KIO/JobUiDelegateFactory>
 #include <KIO/OpenUrlJob>
 #include <KMessageBox>
--- plasma-workspace-6.6.91/shell/desktopview.h.vanilla	2026-06-11 20:26:15.619343681 +0200
+++ plasma-workspace-6.6.91/shell/desktopview.h	2026-06-11 20:26:37.411354491 +0200
@@ -44,7 +44,7 @@
      */
     Q_PROPERTY(QRect strictAvailableScreenRect READ strictAvailableScreenRect NOTIFY strictAvailableScreenRectChanged)
 
-#if PROJECT_VERSION_PATCH >= 80 || PROJECT_VERSION_MINOR >= 80
+#if 0
     Q_PROPERTY(bool showPreviewBanner READ showPreviewBanner NOTIFY showPreviewBannerChanged)
     Q_PROPERTY(QString previewBannerTitle READ previewBannerTitle CONSTANT)
     Q_PROPERTY(QString previewBannerText READ previewBannerText CONSTANT)
@@ -76,7 +76,7 @@
 
     bool enteredEditMode() const;
 
-#if PROJECT_VERSION_PATCH >= 80 || PROJECT_VERSION_MINOR >= 80
+#if 0
     bool showPreviewBanner() const;
     QString previewBannerTitle() const;
     QString previewBannerText() const;
@@ -111,7 +111,7 @@
     void usedInAccentColorChanged();
     void accentColorChanged(const QColor &accentColor);
     void screenToFollowChanged(QScreen *screen);
-#if PROJECT_VERSION_PATCH >= 80 || PROJECT_VERSION_MINOR >= 80
+#if 0
     void showPreviewBannerChanged();
 #endif
     void strictAvailableScreenRectChanged();
@@ -136,7 +136,7 @@
 
     bool m_enteredEditMode;
 
-#if PROJECT_VERSION_PATCH >= 80 || PROJECT_VERSION_MINOR >= 80
+#if 0
     bool m_showPreviewBanner;
 #endif
 };
