--- iproute2-ahu-20021020/Makefile.orig	2004-07-29 01:43:53.083827680 +0200
+++ iproute2-ahu-20021020/Makefile	2004-07-29 01:44:27.182643872 +0200
@@ -37,7 +37,7 @@
 
 CC = gcc
 CCOPTS = -D_GNU_SOURCE -O2 -Wstrict-prototypes -Wall -g
-CFLAGS = $(CCOPTS) $(GLIBCFIX) -I$(KERNEL_INCLUDE) -I../include $(DEFINES)
+CFLAGS = $(CCOPTS) $(GLIBCFIX) -I../include -I$(KERNEL_INCLUDE) $(DEFINES)
 
 LDLIBS += -L../lib -lnetlink -lutil
 
