--- dvb-apps/include/linux/dvb/video.h.orig	2007-01-31 12:23:07.000000000 -0800
+++ dvb-apps/include/linux/dvb/video.h	2007-01-31 12:23:21.000000000 -0800
@@ -24,7 +24,7 @@
 #ifndef _DVBVIDEO_H_
 #define _DVBVIDEO_H_
 
-#include <linux/compiler.h>
+//#include <linux/compiler.h>
 
 #ifdef __KERNEL__
 #include <linux/types.h>
@@ -102,7 +102,7 @@
 
 
 struct video_still_picture {
-        char __user *iFrame;        /* pointer to a single iframe in memory */
+        char *iFrame;        /* pointer to a single iframe in memory */
         int32_t size; 
 };
 
