Andreas V. Meier for ROCKLinux:
	 dhcp: fix a gcc 3.4.3 regression

diff -ruN dhcp-3.0.2rc2/includes/dhcpd.h dhcp-3.0.2rc2-new/includes/dhcpd.h
--- dhcp-3.0.2rc2/includes/dhcpd.h	2004-10-01 20:43:21.000000000 +0200
+++ dhcp-3.0.2rc2-new/includes/dhcpd.h	2004-11-17 20:15:59.414035328 +0100
@@ -263,7 +263,7 @@
 	FTS_ABANDONED = 5,
 	FTS_RESET = 6,
 	FTS_BACKUP = 7
-} binding_state_t;
+} __attribute__((packed)) binding_state_t;
 
 /* FTS_LAST is the highest value that is valid for a lease binding state. */
 #define FTS_LAST FTS_BACKUP
@@ -306,9 +306,9 @@
 #	define EPHEMERAL_FLAGS		(MS_NULL_TERMINATION | \
 					 UNICAST_BROADCAST_HACK)
 
-	binding_state_t __attribute__ ((mode (__byte__))) binding_state;
-	binding_state_t __attribute__ ((mode (__byte__))) next_binding_state;
-	binding_state_t __attribute__ ((mode (__byte__))) desired_binding_state;
+	binding_state_t binding_state;
+	binding_state_t next_binding_state;
+	binding_state_t desired_binding_state;
 	
 	struct lease_state *state;
 
