diff -ruN centericq-4.9.5.orig/libicq2000-0.1/libicq2000/Xml.h centericq-4.9.5/libicq2000-0.1/libicq2000/Xml.h
--- centericq-4.9.5.orig/libicq2000-0.1/libicq2000/Xml.h	2003-07-13 12:37:00.000000000 +0200
+++ centericq-4.9.5/libicq2000-0.1/libicq2000/Xml.h	2003-08-02 14:19:51.000000000 +0200
@@ -24,6 +24,8 @@
 #ifndef XML_H
 #define XML_H
 
+using namespace std;
+
 #include <string>
 #include <list>
 #include <ctype.h>
