--- libmpeg3-1.5.4/video/output.c.firstpatch	2005-09-08 16:27:55.000000000 -0700
+++ libmpeg3-1.5.4/video/output.c	2005-09-08 16:32:47.000000000 -0700
@@ -497,10 +497,12 @@
 	*data++ = 0;
 
 #define STORE_PIXEL_RGB565 \
-	*((unsigned short*)data)++ = \
+	*((unsigned short*)data) = \
 		((CLIP(r_l) & 0xf8) << 8) | \
 		((CLIP(g_l) & 0xfc) << 3) | \
-		((CLIP(b_l) & 0xf8) >> 3);
+		((CLIP(b_l) & 0xf8) >> 3); \
+	data += sizeof(unsigned short); 
+  // as long as data is char* 
 
 #define STORE_PIXEL_RGB888 \
 	*data++ = CLIP(r_l); \
