--- sjog-0.6/src/sjog_brightness.c.orig	2004-09-17 13:39:19.218635212 +0200
+++ sjog-0.6/src/sjog_brightness.c	2004-09-17 13:39:32.634489573 +0200
@@ -51,8 +51,6 @@
         sjog_set_brightness(currentBrightness);
         break;
 
-     default:
-
    }
 
 }
--- sjog-0.6/src/sjog_scroll.c.orig	2004-09-17 13:42:49.557994523 +0200
+++ sjog-0.6/src/sjog_scroll.c	2004-09-17 13:43:11.730448309 +0200
@@ -63,8 +63,6 @@
         sjog_scroll_up();
         break;
 
-     default:
-
    }
 
 }
