--- linux-2.4.22/arch/ppc/config.in	2003-08-25 13:44:40.000000000 +0200
+++ linux-2.4.22-fixed/arch/ppc/config.in	2003-09-16 23:42:48.000000000 +0200
@@ -222,8 +222,10 @@
 
 source drivers/parport/Config.in
 
-tristate 'Generic /dev/rtc emulation' CONFIG_GEN_RTC
-if [ "$CONFIG_GEN_RTC" = "n" -a "$CONFIG_APUS" != "y" ]; then
+if [ "$CONFIG_PPC_RTC" != "y" ]; then
+  tristate 'Generic /dev/rtc emulation' CONFIG_GEN_RTC
+fi
+if [ "$CONFIG_GEN_RTC" != "y" -a "$CONFIG_APUS" != "y" ]; then
   tristate 'Support for /dev/rtc' CONFIG_PPC_RTC
 fi
 
