--- sed-4.1/sed/execute.c	2004-03-13 10:30:39.000000000 +0100
+++ sed-4.1-jsaw/sed/execute.c	2004-06-24 13:35:57.497606648 +0200
@@ -654,7 +654,7 @@
 
   input->read_fn = read_file_line;
 
-  if (in_place_extension)
+  if (in_place_extension && !(name[0] == '-' && name[1] == '\0'))
     {
       int output_fd;
       char *tmpdir = ck_strdup(name), *p;
@@ -701,7 +701,7 @@
   if (input->fp != stdin) /* stdin can be reused on tty and tape devices */
     ck_fclose(input->fp);
 
-  if (in_place_extension && output_file.fp != NULL)
+  if (in_place_extension && output_file.fp != NULL && output_file.fp != stdout)
     {
       ck_fclose (output_file.fp);
       if (strcmp(in_place_extension, "*") != 0)
