--- ./fs/Kconfig.orig	2003-01-14 14:52:18.000000000 +0100
+++ ./fs/Kconfig	2003-01-14 14:53:06.000000000 +0100
@@ -830,8 +830,8 @@
 	  programs depend on this, so everyone should say Y here.
 
 config DEVFS_FS
-	bool "/dev file system support (EXPERIMENTAL)"
-	depends on EXPERIMENTAL
+	bool "/dev file system support (Required for ROCK Linux)"
+	default y
 	---help---
 	  This is support for devfs, a virtual file system (like /proc) which
 	  provides the file system interface to device drivers, normally found
@@ -850,6 +850,7 @@
 config DEVFS_MOUNT
 	bool "Automatically mount at boot"
 	depends on DEVFS_FS
+	default y
 	help
 	  This option appears if you have CONFIG_DEVFS_FS enabled. Setting
 	  this to 'Y' will make the kernel automatically mount devfs onto /dev
