diff -urN -X ../scripts/old/dontdiff gDesklets-0.24.1/tools/Makefile.in gDesklets-0.24.1-libs-fix/tools/Makefile.in
--- gDesklets-0.24.1/tools/Makefile.in	2003-11-10 21:28:26.000000000 +0100
+++ gDesklets-0.24.1-libs-fix/tools/Makefile.in	2003-12-02 16:13:25.000000000 +0100
@@ -193,7 +193,9 @@
 bin_PROGRAMS = gdesklets-display-thumbnailer
 
 gdesklets_display_thumbnailer_SOURCES = gdesklets-display-thumbnailer.c
-gdesklets_display_thumbnailer_LDADD = -lglib-2.0 -lgnomevfs-2
+gdesklets_display_thumbnailer_LDADD = `pkg-config --libs glib-2.0 gnome-vfs-module-2.0`
+
+
 subdir = tools
 ACLOCAL_M4 = $(top_srcdir)/aclocal.m4
 mkinstalldirs = $(SHELL) $(top_srcdir)/mkinstalldirs
