--- ecore-2004-01-18/src/lib/ecore_config/edb.c.orig	2004-05-13 14:41:41.642699904 +0200
+++ ecore-2004-01-18/src/lib/ecore_config/edb.c	2004-05-13 14:41:55.229634376 +0200
@@ -106,6 +106,7 @@
         break;
       case PT_NIL:
         /* currently we do nothing for undefined ojects */
+	;
     }
 
     next=next->next;
--- ecore-2004-01-18/src/lib/ecore_fb/ecore_fb.c.orig	2004-05-13 14:37:15.076224168 +0200
+++ ecore-2004-01-18/src/lib/ecore_fb/ecore_fb.c	2004-05-13 14:37:30.233919848 +0200
@@ -1046,6 +1046,7 @@
 	     if (_ecore_fb_alt < 0) _ecore_fb_alt = 0;
 	  }
 	retry:
+		;
      }
    while (v > 0);
    return 1;
