diff -Nur NVIDIA_kernel-1.0-4191.orig/Makefile NVIDIA_kernel-1.0-4191/Makefile
--- NVIDIA_kernel-1.0-4191.orig/Makefile	2002-12-09 21:27:15.000000000 +0100
+++ NVIDIA_kernel-1.0-4191/Makefile	2003-02-17 00:02:12.000000000 +0100
@@ -87,7 +87,6 @@
 		rm -f $(INSTALLDIR)/NVdriver && \
 		$(INSTALL) -m 0664 -o root -g root nvidia.o $(INSTALLDIR)/nvidia.o$(O) && \
 		/sbin/depmod -a && \
-		/sbin/modprobe nvidia && \
 		sh makedevices.sh && \
 		echo "nvidia.o installed successfully."; \
 	fi
