--- cloop-2.01/Makefile	2004-05-14 21:17:51.000000000 +0200
+++ cloop-2.01-fixed/Makefile	2004-10-16 15:58:19.406490400 +0200
@@ -32,7 +31,7 @@
 MODULE:=cloop.o
 endif
 
-ALL_TARGETS = create_compressed_fs extract_compressed_fs
+ALL_TARGETS = 
 ifndef APPSONLY
 ALL_TARGETS += $(MODULE)
 endif
@@ -73,6 +73,4 @@
 
 install:
 	mkdir -p $(MODDIR)
-	install -m 755 cloop.o $(MODDIR)
-	install -m 755 create_compressed_fs $(PREFIX)/sbin
-	install -m 755 extract_compressed_fs $(PREFIX)/sbin
+	install -m 755 $(MODULE) $(MODDIR)/
