--- cloop-2.01/compressed_loop.c.orig	2004-10-16 16:33:17.765491344 +0200
+++ cloop-2.01/compressed_loop.c	2004-10-16 16:33:41.524879368 +0200
@@ -202,7 +202,7 @@
                           unsigned long offset, unsigned long size)
 {
  char *kaddr;
- struct clo_read_data *p = (struct clo_read_data*)desc->buf;
+ struct clo_read_data *p = (struct clo_read_data*)desc->arg.data;
  unsigned long count = desc->count;
  if (size > count) size = count;
  kaddr = kmap(page);
@@ -228,7 +228,7 @@
    read_descriptor_t desc;
    desc.written = 0;
    desc.count   = size;
-   desc.buf     = (char*)&cd;
+   desc.arg.data = (char*)&cd;
    desc.error   = 0;
 #ifdef REDHAT_KERNEL /* Greenshoe Linux */
    do_generic_file_read(f, &pos, &desc, clo_read_actor, 0);
--- cloop-2.01/Makefile.orig   2004-10-16 16:33:17.765491344 +0200
+++ cloop-2.01/Makefile        2004-10-16 16:33:41.524879368 +0200
@@ -11,7 +11,9 @@
 CKERNOPS:=-D__KERNEL__ -DMODULE -fno-builtin -nostdlib -DKBUILD_MODNAME="cloop" -DKBUILD_BASENAME="cloop"
 
 ifdef CONFIG_MODVERSIONS
+ifeq ($(PATCHLEVEL),4)
 MODVERSIONS:= -DMODVERSIONS -include $(KERNEL_DIR)/include/linux/modversions.h
+endif
 CKERNOPS += $(MODVERSIONS)
 endif
 
