--- ethereal-0.10.12/configure.orig	2005-08-31 08:58:05.153972750 +0200
+++ ethereal-0.10.12/configure	2005-08-31 09:20:21.961518000 +0200
@@ -27666,8 +27666,9 @@
 	  ac_mit_version=`grep 'Massachusetts Institute of Technology' $krb5_dir/include/krb5.h | head -n 1 | sed 's/^.*Massachusetts Institute of Technology.*$/MIT/'`
 	  ac_krb5_version="$ac_heimdal_version$ac_mit_version"
 	  if test "x$ac_krb5_version" = "xHEIMDAL"
-	      KRB5_LIBS="-L$krb5_dir/lib -lkrb5 -lasn1 $SSL_LIBS -lroken -lcrypt"
 	  then
+	      KRB5_LIBS="-L$krb5_dir/lib -lkrb5 -lasn1 $SSL_LIBS -lroken -lcrypt"
+          else
 	      KRB5_LIBS="-L$krb5_dir/lib -lkrb5 -lk5crypto -lcom_err"
 	  fi
 	  if test "x$ac_krb5_version" = "xMIT"
