--- xorg-server-9999/GL/glx/indirect_table.c.orig	2006-10-12 16:02:06.000000000 +0800
+++ xorg-server-9999/GL/glx/indirect_table.c	2006-10-12 16:15:35.000000000 +0800
@@ -1232,7 +1232,7 @@ const struct __glXDispatchInfo Render_di
 
 /*****************************************************************/
 /* tree depth = 13 */
-static const int_fast16_t VendorPriv_dispatch_tree[155] = {
+static const int_fast16_t VendorPriv_dispatch_tree[159] = {
     /* [0] -> opcode range [0, 131072], node depth 1 */
     2,
     5,
@@ -1480,11 +1480,16 @@ static const int_fast16_t VendorPriv_dis
     /* [152] -> opcode range [65536, 65552], node depth 13 */
     1,
     LEAF(88),
+    156,
     EMPTY_LEAF,
 
+    /* [156] -> opcode range [65544, 65552], node depth 13 */
+    1,
+    LEAF(96),
+    EMPTY_LEAF,
 };
 
-static const void *VendorPriv_function_table[96][2] = {
+static const void *VendorPriv_function_table[99][2] = {
     /* [  0] =     0 */ {NULL, NULL},
     /* [  1] =     1 */ {__glXDisp_GetConvolutionFilterEXT, __glXDispSwap_GetConvolutionFilterEXT},
     /* [  2] =     2 */ {__glXDisp_GetConvolutionParameterfvEXT, __glXDispSwap_GetConvolutionParameterfvEXT},
@@ -1581,6 +1586,9 @@ static const void *VendorPriv_function_t
     /* [ 93] = 65541 */ {__glXDisp_CreateContextWithConfigSGIX, __glXDispSwap_CreateContextWithConfigSGIX},
     /* [ 94] = 65542 */ {__glXDisp_CreateGLXPixmapWithConfigSGIX, __glXDispSwap_CreateGLXPixmapWithConfigSGIX},
     /* [ 95] = 65543 */ {NULL, NULL},
+    /* [ 96] = 65544 */ {NULL, NULL},
+    /* [ 97] = 65545 */ {NULL, NULL},
+    /* [ 98] = 65546 */ {__glXDisp_GetDrawableAttributesSGIX, __glXDispSwap_GetDrawableAttributesSGIX},
 };
 
 const struct __glXDispatchInfo VendorPriv_dispatch_info = {
