--- xorg-server-9999/GL/glx/glxdri.c.orig	2006-12-18 23:17:37.000000000 +0800
+++ xorg-server-9999/GL/glx/glxdri.c	2006-12-18 23:18:16.000000000 +0800
@@ -371,13 +371,13 @@ __glXDRIbindTexImage(__GLXcontext *baseC
 	type = GL_UNSIGNED_SHORT_5_6_5;
     }
 
-    if (!(glxCountBits(pixmap->drawable.width) == 1 &&
-	  glxCountBits(pixmap->drawable.height) == 1)
-	/* || strstr(CALL_GetString(GL_EXTENSIONS,
-	             "GL_ARB_texture_non_power_of_two")) */)
+    //if (!(glxCountBits(pixmap->drawable.width) == 1 &&
+	//  glxCountBits(pixmap->drawable.height) == 1)
+	///* || strstr(CALL_GetString(GL_EXTENSIONS,
+	//             "GL_ARB_texture_non_power_of_two")) */)
 	target = GL_TEXTURE_RECTANGLE_ARB;
-    else
-	target = GL_TEXTURE_2D;
+    //else
+	//target = GL_TEXTURE_2D;
 
     CALL_PixelStorei( GET_DISPATCH(), (GL_UNPACK_ROW_LENGTH,
 				       pixmap->devKind / bpp) );
