diff --git a/hw/xfree86/modes/xf86cvt.c b/hw/xfree86/modes/xf86cvt.c
index be5021c89..f617c2ace 100644
--- a/hw/xfree86/modes/xf86Modes.c
+++ b/hw/hw/xfree86/modes/xf86Modes.c
@@ -273,9 +273,12 @@ xf86CVTMode(int HDisplay, int VDisplay, float VRefresh, Bool Reduced,
     Mode->VSyncStart = libxcvt_mode_info->vsync_start;
     Mode->VSyncEnd   = libxcvt_mode_info->vsync_end;
     Mode->VTotal     = libxcvt_mode_info->vtotal;
+    Mode->HSync      = libxcvt_mode_info->hsync;
     Mode->VRefresh   = libxcvt_mode_info->vrefresh;
     Mode->Flags      = libxcvt_mode_info->mode_flags;
 
+    XNFasprintf((char**)&Mode->name, "%dx%d", HDisplay, VDisplay);
+
     free(libxcvt_mode_info);
 
     return Mode;
