--- ./rep-gtk.h.orig	2005-03-28 23:48:38.000000000 -0400
+++ ./rep-gtk.h	2005-03-28 23:49:35.000000000 -0400
@@ -291,10 +291,10 @@
 
 GtkWidget*
 gtk_radio_menu_item_new_with_label_from_widget (GtkRadioMenuItem *group,
-						gchar            *label);
+						const gchar *label);
 GtkWidget*
 gtk_radio_menu_item_new_with_mnemonic_from_widget (GtkRadioMenuItem *group,
-						   gchar            *label);
+						   const gchar *label);
 GtkWidget* gtk_radio_menu_item_new_from_widget (GtkRadioMenuItem *group);
 GtkWidget* gtk_pixmap_new_interp (char *file, GtkWidget *intended_parent);
 
--- ./gtk-compat.c.orig	2005-03-28 23:49:54.000000000 -0400
+++ ./gtk-compat.c	2005-03-28 23:50:39.363196776 -0400
@@ -64,7 +64,7 @@
 
 GtkWidget*
 gtk_radio_menu_item_new_with_label_from_widget (GtkRadioMenuItem *group,
-						gchar            *label)
+					const gchar *label)
 {
   GSList *g = group? gtk_radio_menu_item_group (group) : NULL;
   return gtk_radio_menu_item_new_with_label (g, label);
@@ -72,7 +72,7 @@
 
 GtkWidget*
 gtk_radio_menu_item_new_with_mnemonic_from_widget (GtkRadioMenuItem *group,
-						   gchar            *label)
+						   const gchar *label)
 {
   GSList *g = group? gtk_radio_menu_item_group (group) : NULL;
   return gtk_radio_menu_item_new_with_mnemonic (g, label);
