--- postgresql-9.0.4/src/pl/plperl/plperl.h.vanilla	2011-05-24 15:44:30.106239909 +0200
+++ postgresql-9.0.4/src/pl/plperl/plperl.h	2011-05-24 15:45:56.034249068 +0200
@@ -42,6 +42,11 @@
 #undef bool
 #endif
 
+/* supply GvCV_set if it's missing - ppport.h doesn't supply it, unfortunately */
+#ifndef GvCV_set
+#define GvCV_set(gv, cv)		(GvCV(gv) = cv)
+#endif
+
 /* declare routines from plperl.c for access by .xs files */
 HV		   *plperl_spi_exec(char *, int);
 void		plperl_return_next(SV *);
--- postgresql-9.0.4/src/pl/plperl/plperl.c.vanilla	2011-05-24 15:43:47.810231490 +0200
+++ postgresql-9.0.4/src/pl/plperl/plperl.c	2011-05-24 15:44:27.726237405 +0200
@@ -874,7 +874,7 @@
 		if (!isGV_with_GP(sv) || !GvCV(sv))
 			continue;
 		SvREFCNT_dec(GvCV(sv)); /* free the CV */
-		GvCV(sv) = NULL;		/* prevent call via GV */
+	        GvCV_set(sv, NULL);     /* prevent call via GV */
 	}
 	hv_clear(stash);
 
