--- comm-1.9.2/mozilla/js/src/config/system-headers.vanilla	2011-12-22 17:16:46.498986108 +0100
+++ comm-1.9.2/mozilla/js/src/config/system-headers	2011-12-22 17:16:51.159986101 +0100
@@ -105,7 +105,6 @@
 fusion/ref.h
 fusion/fusion_internal.h
 fusion/lock.h
-fusion/types.h
 fusion/vector.h
 fusion/call.h
 fusion/shmalloc.h
--- comm-1.9.2/mozilla/config/system-headers.vanilla	2011-12-22 17:16:27.151986107 +0100
+++ comm-1.9.2/mozilla/config/system-headers	2011-12-22 17:16:32.220986107 +0100
@@ -105,7 +105,6 @@
 fusion/ref.h
 fusion/fusion_internal.h
 fusion/lock.h
-fusion/types.h
 fusion/vector.h
 fusion/call.h
 fusion/shmalloc.h
--- comm-1.9.2/mozilla/toolkit/crashreporter/google-breakpad/src/common/linux/libcurl_wrapper.cc.vanilla	2011-12-22 17:16:04.706986108 +0100
+++ comm-1.9.2/mozilla/toolkit/crashreporter/google-breakpad/src/common/linux/libcurl_wrapper.cc	2011-12-22 17:16:10.077986106 +0100
@@ -29,7 +29,6 @@
 
 #include <curl/curl.h>
 #include <curl/easy.h>
-#include <curl/types.h>
 #include <dlfcn.h>
 
 #include <string>
--- comm-1.9.2/mozilla/toolkit/crashreporter/google-breakpad/src/common/linux/http_upload.cc.vanilla	2011-12-22 17:15:40.761986108 +0100
+++ comm-1.9.2/mozilla/toolkit/crashreporter/google-breakpad/src/common/linux/http_upload.cc	2011-12-22 17:15:50.592986107 +0100
@@ -31,7 +31,6 @@
 #include <dlfcn.h>
 #include <curl/curl.h>
 #include <curl/easy.h>
-#include <curl/types.h>
 
 #include "common/linux/http_upload.h"
 
