--- thunderbird-151.0.1/comm/build/moz.configure/gecko_source.configure.vanilla	2026-05-28 15:19:49.831333609 +0200
+++ thunderbird-151.0.1/comm/build/moz.configure/gecko_source.configure	2026-05-28 15:20:42.994732833 +0200
@@ -51,7 +51,7 @@
                 pass
 
             if len(lines) != 3:
-                log.warn("sourcestamp.txt is corrupt!")
+                #log.warn("sourcestamp.txt is corrupt!")
                 return
 
             if lines and lines[line2read].startswith("http"):
