--- mozilla/security/coreconf/Linux.mk.vanilla	2006-05-15 13:02:02.000000000 +0200
+++ mozilla/security/coreconf/Linux.mk	2006-05-15 13:02:29.000000000 +0200
@@ -60,6 +60,10 @@
 	OS_REL_CFLAGS	= -DLINUX1_2 -D_XOPEN_SOURCE
 	CPU_ARCH	= ppc
 else
+ifeq ($(OS_TEST),ppc64)
+        OS_REL_CFLAGS   = -DLINUX1_2 -D_XOPEN_SOURCE
+        CPU_ARCH        = ppc
+else
 ifeq ($(OS_TEST),alpha)
         OS_REL_CFLAGS   = -D_ALPHA_ -DLINUX1_2 -D_XOPEN_SOURCE
 	CPU_ARCH	= alpha
@@ -119,6 +123,7 @@
 endif
 endif
 endif
+endif
 
 
 LIBC_TAG		= _glibc
