--- ctapi-cyberjack-2.0.8/ctapi/Makefile.orig	2005-04-19 14:57:02.000000000 +0200
+++ ctapi-cyberjack-2.0.8/ctapi/Makefile	2005-04-19 14:57:48.000000000 +0200
@@ -15,8 +15,8 @@
 install: all
 	install -o root -g root -d $(INSTALL_PREFIX)/lib
 	install -o root -g root libctapi-cyberjack.so $(INSTALL_PREFIX)/lib
-	install -o root -g root -d $(INSTALL_PREFIX)/include
-	install -o root -g root include/ctapi.h $(INSTALL_PREFIX)/include
+	install -o root -g root -d $(INSTALL_PREFIX)/include/cyberjack
+	install -o root -g root include/ctapi.h $(INSTALL_PREFIX)/include/cyberjack
 
 %.o: %.c
 	$(CC) $(CFLAGS) -o $@ -c $^
