--- cpufreqd-1.0/Makefile.in.orig	2003-07-29 23:11:06.000000000 +0200
+++ cpufreqd-1.0/Makefile.in	2003-07-29 23:12:24.000000000 +0200
@@ -563,7 +563,7 @@
 	if [ -f $(DESTDIR)/$(sysconfdir)/cpufreqd.conf ]; then \
 	 echo "** preserving old configuration file **"; \
 	else \
- 	 install -Dm 0644 -o root -g root cpufreqd.conf $(DESTDIR)/$(sysconfdir)/cpufreqd.conf; \
+ 	 install -m 0644 -o root -g root cpufreqd.conf $(DESTDIR)/$(sysconfdir)/cpufreqd.conf; \
 	fi;
 
 #sysconf_DATA = cpufreqd.conf
