
Author:
 Copyright (C) 2003 Rene Rebe (rene@rocklinux.org)

diff -urN cpufreqd-1.0/Makefile.in cpufreqd-1.0-pmu/Makefile.in
--- cpufreqd-1.0/Makefile.in	2003-06-23 22:01:36.000000000 +0200
+++ cpufreqd-1.0-pmu/Makefile.in	2003-07-24 18:47:17.000000000 +0200
@@ -88,7 +88,7 @@
 cpufreqd_SOURCES =                      main.c                     config_parser.c                     string_list.c
 
 
-lib_LTLIBRARIES =                    libsys_acpi.la                   libsys_apm.la
+lib_LTLIBRARIES =                    libsys_acpi.la                   libsys_apm.la libsys_pmu.la
 
 
 libsys_acpi_la_SOURCES = libsys_acpi.c
@@ -97,6 +97,9 @@
 libsys_apm_la_SOURCES = libsys_apm.c
 libsys_apm_la_LDFLAGS = -module -avoid-version
 
+libsys_pmu_la_SOURCES = libsys_pmu.c
+libsys_pmu_la_LDFLAGS = -module -avoid-version
+
 noinst_HEADERS =                  config_parser.h                 cpufreqd.h                 main.h                 string_list.h
 
 
@@ -117,6 +120,9 @@
 libsys_acpi_la_OBJECTS =  libsys_acpi.lo
 libsys_apm_la_LIBADD = 
 libsys_apm_la_OBJECTS =  libsys_apm.lo
+libsys_pmu_la_LIBADD =
+libsys_pmu_la_OBJECTS =  libsys_pmu.lo
+
 bin_PROGRAMS =  cpufreqd$(EXEEXT)
 PROGRAMS =  $(bin_PROGRAMS)
 
@@ -141,8 +147,8 @@
 
 TAR = tar
 GZIP_ENV = --best
-SOURCES = $(libsys_acpi_la_SOURCES) $(libsys_apm_la_SOURCES) $(cpufreqd_SOURCES)
-OBJECTS = $(libsys_acpi_la_OBJECTS) $(libsys_apm_la_OBJECTS) $(cpufreqd_OBJECTS)
+SOURCES = $(libsys_acpi_la_SOURCES) $(libsys_apm_la_SOURCES) $(libsys_pmu_la_SOURCES) $(cpufreqd_SOURCES)
+OBJECTS = $(libsys_acpi_la_OBJECTS) $(libsys_apm_la_OBJECTS) $(libsys_pmu_la_OBJECTS) $(cpufreqd_OBJECTS)
 
 all: all-redirect
 .SUFFIXES:
@@ -265,6 +271,10 @@
 libsys_apm.la: $(libsys_apm_la_OBJECTS) $(libsys_apm_la_DEPENDENCIES)
 	$(LINK) -rpath $(libdir) $(libsys_apm_la_LDFLAGS) $(libsys_apm_la_OBJECTS) $(libsys_apm_la_LIBADD) $(LIBS)
 
+libsys_pmu.la: $(libsys_pmu_la_OBJECTS) $(libsys_pmu_la_DEPENDENCIES)
+	$(LINK) -rpath $(libdir) $(libsys_pmu_la_LDFLAGS) $(libsys_pmu_la_OBJECTS) $(libsys_pmu_la_LIBADD) $(LIBS)
+
+
 mostlyclean-binPROGRAMS:
 
 clean-binPROGRAMS:
@@ -448,6 +458,9 @@
 	config.h main.h config_parser.h
 libsys_apm.lo libsys_apm.o : libsys_apm.c cpufreqd.h string_list.h \
 	config.h main.h config_parser.h
+libsys_pmu.lo libsys_pmu.o : libsys_pmu.c cpufreqd.h string_list.h \
+	config.h main.h config_parser.h
+
 main.o: main.c cpufreqd.h string_list.h config.h main.h config_parser.h
 string_list.o: string_list.c string_list.h

