--- zebra-0.91a/lib/vty.c.orig	Thu Feb  1 02:18:05 2001
+++ zebra-0.91a/lib/vty.c	Wed Oct 15 19:50:34 2003
@@ -1103,13 +1103,16 @@
       break;
     case SE: 
       {
-	char *buffer = (char *)vty->sb_buffer->head->data;
-	int length = vty->sb_buffer->length;
+	char *buffer;
+	int length;
 
-	if (buffer == NULL)
+	if (!vty->iac_sb_in_progress)
 	  return 0;
 
-	if (!vty->iac_sb_in_progress)
+	buffer = (char *)vty->sb_buffer->head->data;
+	length = vty->sb_buffer->length;
+
+	if (buffer == NULL)
 	  return 0;
 
 	if (buffer[0] == '\0')
