diff --git a/arch/mips/Kconfig.debug b/arch/mips/Kconfig.debug
index 0749c3724543..b5e3163024f1 100644
--- a/arch/mips/Kconfig.debug
+++ b/arch/mips/Kconfig.debug
@@ -21,18 +21,34 @@ config EARLY_PRINTK
 	  doesn't cooperate with an X server. You should normally say N here,
 	  unless you want to debug such a crash.
 
+menu "Early Printk Drivers"
+
+choice
+	prompt "Early Printk Drivers"
+	depends on EARLY_PRINTK
+
+config EARLY_PRINTK_IP27
+	bool "Early Serial printk for IP27"
+	depends on EARLY_PRINTK && SGI_IP27
+	help
+	  This is a hack to get early console output on IP27 under the new
+	  early printk framework added by IP30 support.
+
 config EARLY_PRINTK_8250
-	bool
+	bool "8250 Serial printk"
 	depends on EARLY_PRINTK && USE_GENERIC_EARLY_PRINTK_8250
-	default y
 	help
 	  "8250/16550 and compatible serial early printk driver"
 	  If you say Y here, it will be possible to use a 8250/16550 serial
 	  port as the boot console.
 
+endchoice
+
 config USE_GENERIC_EARLY_PRINTK_8250
 	bool
 
+endmenu
+
 config CMDLINE_BOOL
 	bool "Built-in kernel command line"
 	default n
