diff --git a/arch/mips/include/asm/pci/bridge.h b/arch/mips/include/asm/pci/bridge.h
index 5d4eef676d8e..f891a44169a0 100644
--- a/arch/mips/include/asm/pci/bridge.h
+++ b/arch/mips/include/asm/pci/bridge.h
@@ -484,6 +484,8 @@ struct bridge_widget {
 #define BRIDGE_REV_B			0x2
 #define BRIDGE_REV_C			0x3
 #define BRIDGE_REV_D			0x4
+#define XBRIDGE_REV_A			0x1
+#define XBRIDGE_REV_B			0x2
 
 /* Bridge widget status register bits definition */
 #define BRIDGE_STAT_LLP_REC_CNT		GENMASK(31, 24)
@@ -943,6 +945,7 @@ struct bridge_err_cmdword_s {
 #define PCI64_ATTR_PREC		BIT_ULL(58)	/* Precise */
 #define PCI64_ATTR_VIRTUAL	BIT_ULL(57)	/* Virtual Request */
 #define PCI64_ATTR_BAR		BIT_ULL(56)	/* Barrier */
+#define PCI64_ATTR_SWAP		BIT_ULL(55)	/* Byte swap (XBRIDGE) */
 #define PCI64_ATTR_RMF_MASK	GENMASK_ULL(55, 48)
 #define PCI64_ATTR_RMF_SHFT	48
 
