--- isdn4k-utils/isdnlog/tools/cdb/Makefile.in~	2000-08-15 17:16:23.000000000 +0200
+++ isdn4k-utils/isdnlog/tools/cdb/Makefile.in	2004-12-18 12:42:55.354247840 +0100
@@ -90,4 +90,7 @@
 depend: .depend
 
 .depend:
-	$(CPP) -MM $(CFLAGS) $(DEFS) $(INCLUDE) *.c >.depend
+	rm -f .depend; for i in *.c; do \
+		$(CPP) -MM $(CFLAGS) $(DEFS) $(INCLUDE) $$i >>.depend; \
+	done
+
--- isdn4k-utils/lib/Makefile~	1999-10-26 20:17:17.000000000 +0200
+++ isdn4k-utils/lib/Makefile	2004-12-18 12:48:37.638212768 +0100
@@ -53,7 +53,9 @@
 depend dep: .depend
 
 .depend:
-		$(CPP) -M $(CFLAGS) $(DEFS) *.c >.depend
+	rm -f .depend; for i in *.c; do \
+		$(CPP) -MM $(CFLAGS) $(DEFS) $(INCLUDE) $$i >>.depend; \
+	done
 
 clean:
 		-rm -f *.o *.a
--- isdn4k-utils/isdnlog/Makefile.in~	2002-07-19 21:03:55.000000000 +0200
+++ isdn4k-utils/isdnlog/Makefile.in	2004-12-18 12:59:43.924921792 +0100
@@ -1903,7 +1903,9 @@
 depend: .depend
 
 .depend:
-		$(CPP) -M $(CFLAGS) $(DEFS) $(INCLUDE) */*.c >.depend
+		rm -f .depend; for i in */*.c; do \
+			$(CPP) -MM $(CFLAGS) $(DEFS) $(INCLUDE) $$i >>.depend; \
+		done
 ifeq ($(ORACLE),1)
 		$(CPP) -M $(CFLAGS) $(DEFS) $(INCLUDE) - < isdnlog/oracle.pc | \
 		sed s/-:/oracle.o:/ >>.depend
