--- isdn4k-utils/isdnlog/isdnlog/processor.c~	2002-04-01 20:06:10.000000000 +0200
+++ isdn4k-utils/isdnlog/isdnlog/processor.c	2004-10-14 14:23:54.634300688 +0200
@@ -5293,7 +5293,7 @@
     tei = BROADCAST; /* Wenn nach einer tei-Zeile keine hex:-Zeile kommt, tei ungueltig machen! */
 
     if ((type == SETUP) && !replay) { /* fetch additional info from "/dev/isdninfo" */
-      static void moreinfo(); /* soviel zu Objektorientiertem Denken ;-) */
+      /*static void moreinfo(); ??? soviel zu Objektorientiertem Denken ;-) */
       moreinfo();
     } /* if */
 
