--- ./Makefile.orig	2001-02-14 11:53:22.000000000 +0100
+++ ./Makefile	2003-07-12 13:50:17.000000000 +0200
@@ -1,3 +1,8 @@
+
+PREFIX = /usr
+DESTBIN = $(PREFIX)/sbin
+DESTMAN = $(PREFIX)/man/man1
+
 all: longrun README
 
 longrun: longrun.c
@@ -8,3 +13,14 @@
 
 clean:
 	-rm -f longrun README
+
+install:
+	install -m 755 -d $(DESTBIN)
+	install -m 755 longrun $(DESTBIN)
+	install -m 755 -d $(DESTMAN)
+	install -m 644 longrun.1 $(DESTMAN)
+
+uninstall:
+	rm -f $(DESTBIN)/longrun
+	rm -f $(DESTMAN)/longrun.1
+
