diff -Naur freetype-2.1.7.orig/builds/unix/ltmain.sh freetype-2.1.7/builds/unix/ltmain.sh
--- freetype-2.1.7.orig/builds/unix/ltmain.sh	2003-11-09 10:37:13.000000000 +0200
+++ freetype-2.1.7/builds/unix/ltmain.sh	2005-03-05 22:56:07.969688072 +0200
@@ -3264,7 +3264,7 @@
 	      fi
 	      if test -n "$a_deplib" ; then
 		libname=`eval \\$echo \"$libname_spec\"`
-		for i in $lib_search_path $sys_lib_search_path $shlib_search_path; do
+		for i in $lib_search_path $sys_lib_search_path $shlib_search_path /lib64 /usr/lib64; do
 		  potential_libs=`ls $i/$libname[.-]* 2>/dev/null`
 		  for potent_lib in $potential_libs; do
 		      # Follow soft links.
