diff -u screen-4.0.2.old/configure screen-4.0.2/configure
--- screen-4.0.2.old/configure	2003-12-05 14:46:53.000000000 +0100
+++ screen-4.0.2/configure	2004-03-23 18:27:13.000000000 +0100
@@ -7991,6 +7991,9 @@
 if test -n "$prefix"; then
 ETCSCREENRC="\"$prefix/etc/screenrc\""
 fi
+if test -n "$sysconfdir"; then
+ETCSCREENRC="\"$sysconfdir/screenrc\""
+fi
 echo "$as_me:$LINENO: checking for the global screenrc file" >&5
 echo $ECHO_N "checking for the global screenrc file... $ECHO_C" >&6
 
