--- ./configure.orig	2003-03-10 10:10:09.000000000 +0100
+++ ./configure	2003-03-10 10:10:21.000000000 +0100
@@ -7017,7 +7017,7 @@
     echo "$as_me:$LINENO: checking for PAM support" >&5
 echo $ECHO_N "checking for PAM support... $ECHO_C" >&6
     oldlibs="$LIBS"
-    LIBS="$LIBS -lpam"
+    LIBS="$LIBS -lpam -lpam_misc"
     cat >conftest.$ac_ext <<_ACEOF
 #line $LINENO "configure"
 #include "confdefs.h"
