--- ./Makefile.orig	2003-09-29 00:26:40.000000000 +0200
+++ ./Makefile	2003-09-29 00:26:42.000000000 +0200
@@ -79,10 +79,10 @@
 all: $(targets)
 
 mine: $(MINE_ALL_OBJS)
-	gcc $(MINE_ALL_OBJS) $(MINE_LDFLAGS) -o mine
+	$(CC) $(MINE_ALL_OBJS) $(MINE_LDFLAGS) -o mine
 
 gasgui: $(GAS_OBJ)
-	gcc $(GAS_OBJ) -ldialog -lcurses -lm -o gasgui
+	$(CC) $(GAS_OBJ) -ldialog -lcurses -lm -o gasgui
 
 install:
 	install mine $(sysprefix)/bin/mine
--- ./selector.c	(revision 15)
+++ ./selector.c	(revision 16)
@@ -551,7 +551,7 @@
 		}
 
 		if ( !pkgcount && !bigmain ) {
-			char *dirname = "All Packages";
+			char *dirname = "all";
 			struct directory_entry * e;
 			if ( instmenu1() == 1 ) dirname = "base";
 			for (e = rootdir->list; e != NULL; e = e->next) {
