--- mine-0.13/md5sum.c.orig	2003-05-21 11:55:20.000000000 +0200
+++ mine-0.13/md5sum.c	2003-06-06 14:52:52.000000000 +0200
@@ -74,8 +74,10 @@
 			while (fgets(buffer, 1024, f) != NULL) {
 				md5sum = strtok(buffer, " \t\n");
 				filename = strtok(NULL, "\n");
-				while ( *filename == ' ' ) filename++;
-				memdb_put(&md5_memdb, filename, md5sum);
+				if ( md5sum && filename) {
+					while ( *filename == ' ' ) filename++;
+					memdb_put(&md5_memdb, filename, md5sum);
+				}
 			}
 			fclose(f);
 		}
