--- ./Makefile.orig	2003-10-18 11:26:04.000000000 +0300
+++ ./Makefile	2003-10-18 11:26:13.000000000 +0300
@@ -23,7 +23,6 @@
 MINE_VER  = 0.15
 MINE_OBJ  = create.o install.o remove.o pkglist.o
 MINE_OBJ += showfile.o showkey.o memdb.o md5.o md5sum.o mine.o
-MINE_LDFLAGS = -static
 
 # GAS
 #
@@ -63,7 +62,7 @@
 
 # What do we want to build
 #
-targets = mine
+targets = mine mine.static
 ifeq ($(USE_GAS), 1)
   targets += gasgui
 endif
@@ -79,13 +78,17 @@
 all: $(targets)
 
 mine: $(MINE_ALL_OBJS)
-	$(CC) $(MINE_ALL_OBJS) $(MINE_LDFLAGS) -o mine
+	$(CC) $(MINE_ALL_OBJS) -o mine
+
+mine.static:
+	$(CC) $(MINE_ALL_OBJS) -static -o mine.static
 
 gasgui: $(GAS_OBJ)
 	$(CC) $(GAS_OBJ) -ldialog -lcurses -lm -o gasgui
 
 install:
 	install mine $(sysprefix)/bin/mine
+	install mine.static $(sysprefix)/bin/mine.static
 	install tarbz2gem.sh $(prefix)/sbin/tarbz2gem
 	install rocket.sh $(prefix)/sbin/rocket
 ifeq ($(USE_GAS), 1)
