diff -urN bison-1.875-old/data/yacc.c bison-1.875-new/data/yacc.c
--- bison-1.875-old/data/yacc.c	2002-12-28 09:36:02.000000000 +0100
+++ bison-1.875-new/data/yacc.c	2003-07-16 15:26:30.000000000 +0200
@@ -1114,8 +1114,10 @@
   /* Suppress GCC warning that yyerrlab1 is unused when no action
      invokes YYERROR.  */
 #if defined (__GNUC_MINOR__) && 2093 <= (__GNUC__ * 1000 + __GNUC_MINOR__)
+#if ! defined(__cplusplus)
   __attribute__ ((__unused__))
 #endif
+#endif
 
 ]b4_location_if([  yylerrsp = yylsp;
   *++yylerrsp = yyloc;])[
