
Some glibc-2.3.4 / gcc-3.4 changes.

  - Rene Rebe <rene@exactcode.de>

--- ncompress-4.2.4/compress42.c.orig	2004-05-03 19:39:12.000000000 +0200
+++ ncompress-4.2.4/compress42.c	2004-05-03 19:42:08.000000000 +0200
@@ -172,7 +172,7 @@
 	extern	void	*malloc	LARGS((int));
 	extern	void	free	LARGS((void *));
 #ifndef _IBMR2
-	extern	int		open	LARGS((char const *,int,...));
+	/* extern	int		open	LARGS((const char *,int,...)); */
 #endif
 	extern	int		close	LARGS((int));
 	extern	int		read	LARGS((int,void *,int));
@@ -638,7 +638,7 @@
 	} ;
 #endif
 
-void  	main			ARGS((int,char **));
+int   	main			ARGS((int,char **));
 void  	Usage			ARGS((void));
 void  	comprexx		ARGS((char **));
 void  	compdir			ARGS((char *));
@@ -691,7 +691,7 @@
  *   deterministic, and can be done on the fly.  Thus, the decompression
  *   procedure needs no input table, but tracks the way the table was built.
  */ 
-void
+int
 main(argc, argv)
 	REG1	int 	 argc;
 	REG2	char	*argv[];
