--- linux-6.1/drivers/ata/pata_pcmcia.c.orig	2023-02-27 18:51:20.641993816 +0100
+++ linux-6.1/drivers/ata/pata_pcmcia.c	2023-02-27 18:56:22.022992807 +0100
@@ -362,6 +362,7 @@
 	PCMCIA_DEVICE_PROD_ID12("STI", "Flash 5.0", 0xbf2df18d, 0x8cb57a0e),
 	PCMCIA_MFC_DEVICE_PROD_ID12(1, "SanDisk", "ConnectPlus", 0x7a954bd9, 0x74be00c6),
 	PCMCIA_DEVICE_PROD_ID2("Flash Card", 0x5a362506),
+	PCMCIA_DEVICE_PROD_ID2("PCMCIA ATA/ATAPI Adapter", 0x888D7B73),
 	PCMCIA_DEVICE_NULL,
 };
 
