Let brcmsmac/dma.c use new generic parity32 function.

Signed-off-by: René Rebe <rene@exactcode.de>

--- linux-6.3/drivers/net/wireless/broadcom/brcm80211/brcmsmac/dma.c.vanilla	2023-06-05 14:49:43.028688133 +0200
+++ linux-6.3/drivers/net/wireless/broadcom/brcm80211/brcmsmac/dma.c	2023-06-05 14:49:49.391688056 +0200
@@ -283,21 +283,6 @@
 	bool aligndesc_4k;
 };
 
-/* Check for odd number of 1's */
-static u32 parity32(__le32 data)
-{
-	/* no swap needed for counting 1's */
-	u32 par_data = *(u32 *)&data;
-
-	par_data ^= par_data >> 16;
-	par_data ^= par_data >> 8;
-	par_data ^= par_data >> 4;
-	par_data ^= par_data >> 2;
-	par_data ^= par_data >> 1;
-
-	return par_data & 1;
-}
-
 static bool dma64_dd_parity(struct dma64desc *dd)
 {
 	return parity32(dd->addrlow ^ dd->addrhigh ^ dd->ctrl1 ^ dd->ctrl2);
