@@ -, +, @@ 
---
 drivers/char/tpm/tpm-chip.c | 26 ++------------------------
 1 file changed, 2 insertions(+), 24 deletions(-)
--- a/drivers/char/tpm/tpm-chip.c	
+++ a/drivers/char/tpm/tpm-chip.c	
@@ -513,16 +513,11 @@ static int tpm_add_legacy_sysfs(struct tpm_chip *chip)
 /*
  * Some AMD fTPM versions may cause stutter
  * https://www.amd.com/en/support/kb/faq/pa-410
- *
- * Fixes are available in two series of fTPM firmware:
- * 6.x.y.z series: 6.0.18.6 +
- * 3.x.y.z series: 3.57.y.5 +
  */
 #ifdef CONFIG_X86
 static bool tpm_amd_is_rng_defective(struct tpm_chip *chip)
 {
-	u32 val1, val2;
-	u64 version;
+	u32 val1;
 	int ret;
 
 	if (!(chip->flags & TPM_CHIP_FLAG_TPM2))
@@ -539,31 +534,14 @@ static bool tpm_amd_is_rng_defective(struct tpm_chip *chip)
 		ret = -ENODEV;
 		goto release;
 	}
-	ret = tpm2_get_tpm_pt(chip, TPM2_PT_FIRMWARE_VERSION_1, &val1, NULL);
-	if (ret)
-		goto release;
-	ret = tpm2_get_tpm_pt(chip, TPM2_PT_FIRMWARE_VERSION_2, &val2, NULL);
-
 release:
 	tpm_relinquish_locality(chip);
 
 	if (ret)
 		return false;
 
-	version = ((u64)val1 << 32) | val2;
-	if ((version >> 48) == 6) {
-		if (version >= 0x0006000000180006ULL)
-			return false;
-	} else if ((version >> 48) == 3) {
-		if (version >= 0x0003005700000005ULL)
-			return false;
-	} else {
-		return false;
-	}
-
 	dev_warn(&chip->dev,
-		 "AMD fTPM version 0x%llx causes system stutter; hwrng disabled\n",
-		 version);
+		 "AMD fTPM may cause system stutter; hwrng disabled\n");
 
 	return true;
 }
-- 
