--- avview-0.80.3/configure.orig	2004-11-22 01:59:15.900006000 +0100
+++ avview-0.80.3/configure	2004-11-22 02:06:43.829910304 +0100
@@ -7582,8 +7582,8 @@
 #define USE_FFMPEG 1
 _ACEOF
 
-FFMPEG_LIBS="$with_ffmpeg/libavformat/libavformat.a $with_ffmpeg/libavcodec/libavcodec.a"
-FFMPEG_CFLAGS="-I$with_ffmpeg/libavformat -I$with_ffmpeg/libavcodec"
+FFMPEG_LIBS="-L$with_ffmpeg/lib -lavformat -lavcodec"
+FFMPEG_CFLAGS="-I$with_ffmpeg/include/ffmpeg"
 echo -n "checking ffmpeg version..."
 CFLAGS="$CFLAGS $FFMPEG_CFLAGS"
 #LDFLAGS="$LDFLAGS $FFMPEG_LIBS"
