--- emacs-21.3/Makefile.in.orig	2003-05-20 06:00:40.000000000 +0200
+++ emacs-21.3/Makefile.in	2003-05-20 06:00:47.000000000 +0200
@@ -448,7 +448,7 @@
 	-chmod -R a+r ${datadir}/emacs ${COPYDESTS} ${infodir}
 	thisdir=`/bin/pwd`; \
 	cd ${srcdir}/etc; \
-	for page in emacs etags ctags gfdl ; do \
+	for page in emacs gfdl ; do \
 	  (cd $${thisdir}; \
 	   ${INSTALL_DATA} ${srcdir}/etc/$${page}.1 ${man1dir}/$${page}${manext}; \
 	   chmod a+r ${man1dir}/$${page}${manext}); \
