diff -urN emacs-21.3/etc/Makefile emacs-21.3-fixed/etc/Makefile
--- emacs-21.3/etc/Makefile	1995-04-26 22:59:05.000000000 +0200
+++ emacs-21.3-fixed/etc/Makefile	2003-03-29 02:05:11.000000000 +0100
@@ -12,7 +12,7 @@
 clean distclean maintainer-clean:
 	-rm -f DOC* core
 
-SOURCES = [0-9A-QS-Z]* README *.[ch16] emacs.* etags.* ledit.l ms-* \
+SOURCES = [0-9A-QS-Z]* README *.[ch16] emacs.* ledit.l ms-* \
 	news.texi rc2log refcard.tex spook-lines termcap.* ulimit.hack \
 	vcdiff vipcard.tex xmouse.doc
 
diff -urN emacs-21.3/lib-src/Makefile.in emacs-21.3-fixed/lib-src/Makefile.in
--- emacs-21.3/lib-src/Makefile.in	2002-05-31 18:45:17.000000000 +0200
+++ emacs-21.3-fixed/lib-src/Makefile.in	2003-03-29 02:03:28.000000000 +0100
@@ -294,15 +294,8 @@
 
 install: ${archlibdir}
 	@echo
-	@echo "Installing utilities for users to run."
-	for file in ${INSTALLABLES} ; do \
-	  $(INSTALL_PROGRAM) $(INSTALL_STRIP) $${file} ${bindir}/$${file} ; \
-	  chmod a+rx ${bindir}/$${file}; \
-	done
-	for file in ${INSTALLABLE_SCRIPTS} ; do \
-	  $(INSTALL_PROGRAM) ${srcdir}/$${file} ${bindir}/$${file} ; \
-	  chmod a+rx ${bindir}/$${file}; \
-	done
+	@echo "Installing no utilities for users."
+	@echo "They are shared with emacs and in a seperate ROCK package."
 
 uninstall:
 	(cd ${bindir}; \
