Index: configure.ac
===================================================================
--- cinelerra-5.1/configure.ac	(.../configure.ac)	(working copy)
+++ cinelerra-5.1/configure.ac	(.../home/user/Desktop/patch/configure.ac)	(working copy)
@@ -652,7 +652,7 @@
 CHECK_LIB([libjpeg], [jpeg], [jpeg_start_decompress])
 CHECK_HEADERS([libjpeg], [jpeg headers], [stdio.h jpeglib.h])
 CHECK_LIB([openjpeg], [openjp2], [opj_version])
-CHECK_HEADERS([openjpeg], [openjpeg headers], [openjpeg.h])
+CHECK_HEADERS([openjpeg], [openjpeg headers], [openjpeg-2.5/openjpeg.h])
 CHECK_LIB([libsndfile], [sndfile], [sf_open])
 CHECK_HEADERS([libsndfile], [sndfile headers], [sndfile.h])
 CHECK_LIB([ilmbase], [IlmImf], [ImfOpenInputFile])
@@ -829,7 +829,7 @@
  CHECK_HEADERS([pulse], [pulse headers], [pulse/error.h])])
 
 CHECK_WANT([ISOFS], [auto], [use isofs], [
- CHECK_HEADERS([isofs], [isofs headers], [linux/isofs.h])])
+ CHECK_HEADERS([isofs], [isofs headers], [libisofs/libisofs.h])])
 
 if test "x$WANT_LADSPA_BUILD" != "xyes" ; then
  CHECK_WANT([LADSPA], [auto], [ladspa plugin clients], [

