--- ./dist/Makefile.in.orig	2003-12-16 19:42:57.000000000 -0300
+++ ./dist/Makefile.in	2003-12-16 19:52:04.000000000 -0300
@@ -61,12 +61,10 @@
 LIBSO_LIBS=	@LIBSO_LIBS@
 
 libdb_base=	libdb
-libdb=		$(libdb_base).a
+libdb=		$(libdb_base)-$(LIBVERSION).a
 libdb_version=	$(libdb_base)-$(LIBVERSION).a
 libso=		$(libdb_base)-$(LIBVERSION)@SOSUFFIX@
 libso_target=	$(libdb_base)-$(LIBVERSION).la
-libso_default=	$(libdb_base)@SOSUFFIX@
-libso_major=	$(libdb_base)-$(LIBMAJOR)@SOSUFFIX@
 
 ##################################################
 # C++ API.
@@ -80,12 +78,10 @@
 LIBXSO_LIBS=	@LIBXSO_LIBS@
 
 libcxx_base=	libdb_cxx
-libcxx=		$(libcxx_base).a
+libcxx=		$(libcxx_base)-$(LIBVERSION).a
 libcxx_version=	$(libcxx_base)-$(LIBVERSION).a
 libxso=		$(libcxx_base)-$(LIBVERSION)@SOSUFFIX@
 libxso_target=	$(libcxx_base)-$(LIBVERSION).la
-libxso_default=	$(libcxx_base)@SOSUFFIX@
-libxso_major=	$(libcxx_base)-$(LIBMAJOR)@SOSUFFIX@
 
 ##################################################
 # Java API.
@@ -115,8 +111,6 @@
 libjso=		$(libjso_base)-$(LIBVERSION)@JMODSUFFIX@
 libjso_static=	$(libjso_base)-$(LIBVERSION).a
 libjso_target=	$(libjso_base)-$(LIBVERSION).la
-libjso_default=	$(libjso_base)@JMODSUFFIX@
-libjso_major=	$(libjso_base)-$(LIBMAJOR)@JMODSUFFIX@
 libjso_g=	$(libjso_base)-$(LIBVERSION)_g@JMODSUFFIX@
 
 ##################################################
@@ -130,8 +124,6 @@
 libtso=		$(libtso_base)-$(LIBVERSION)@MODSUFFIX@
 libtso_static=	$(libtso_base)-$(LIBVERSION).a
 libtso_target=	$(libtso_base)-$(LIBVERSION).la
-libtso_default=	$(libtso_base)@MODSUFFIX@
-libtso_major=	$(libtso_base)-$(LIBMAJOR)@MODSUFFIX@
 
 ##################################################
 # db_dump185 UTILITY
@@ -592,25 +584,17 @@
 LIB_INSTALL_FILE_LIST=\
 	$(libdb) \
 	$(libso) \
-	$(libso_default) \
-	$(libso_major) \
 	$(libdb_version) \
 	$(libso_target) \
 	$(libcxx) \
 	$(libxso) \
-	$(libxso_default) \
-	$(libxso_major) \
 	$(libcxx_version) \
 	$(libxso_target) \
 	$(libtso) \
-	$(libtso_default) \
-	$(libtso_major) \
 	$(libtso_static) \
 	$(libtso_target) \
 	$(libjso) \
-	$(libjso_default) \
 	$(libjso_g) \
-	$(libjso_major) \
 	$(libjso_static) \
 	$(libjso_target) \
 	$(libj_exjarfile) \
@@ -811,14 +795,6 @@
 	@cd $(DESTDIR)$(libdir) && $(rm) -f $(LIB_INSTALL_FILE_LIST)
 	@$(INSTALLER) @INSTALL_LIBS@ $(DESTDIR)$(libdir)
 	@(cd $(DESTDIR)$(libdir) && \
-	    test -f $(libso) && $(ln) -s $(libso) $(libso_default); \
-	    test -f $(libso) && $(ln) -s $(libso) $(libso_major); \
-	    test -f $(libxso) && $(ln) -s $(libxso) $(libxso_default); \
-	    test -f $(libxso) && $(ln) -s $(libxso) $(libxso_major); \
-	    test -f $(libtso) && $(ln) -s $(libtso) $(libtso_default); \
-	    test -f $(libtso) && $(ln) -s $(libtso) $(libtso_major); \
-	    test -f $(libjso) && $(ln) -s $(libjso) $(libjso_default); \
-	    test -f $(libjso) && $(ln) -s $(libjso) $(libjso_major); \
 	    test -f $(libjso) && $(ln) -s $(libjso) $(libjso_g)) || true
 	@(test -f $(libj_jarfile) && \
 	    $(cp) $(libj_jarfile) $(DESTDIR)$(libdir) && \
