--- xmkmf/Makefile.in.vanilla	2005-10-04 16:24:39.000000000 +0200
+++ xmkmf/Makefile.in	2005-10-04 15:38:48.000000000 +0200
@@ -163,7 +163,7 @@
 sharedstatedir = @sharedstatedir@
 sysconfdir = @sysconfdir@
 target_alias = @target_alias@
-XCONFDIR = $(sysconfdir)/X11/config/cf
+XCONFDIR = $(sysconfdir)/X11/config
 bin_SCRIPTS = xmkmf
 dist_man1_MANS = \
         xmkmf.man
@@ -520,7 +520,7 @@
 
 
 xmkmf: xmkmf.cpp
-	$(RAWCPP) -DCONFIGDIRSPEC=$(XCONFDIR) $(srcdir)/xmkmf.cpp | $(SED) -e /^\#/d | $(SED) -e s/XCOMM/\#/ > $@
+	$(RAWCPP) -DCONFIGDIRSPEC=-I$(XCONFDIR) $(srcdir)/xmkmf.cpp | $(SED) -e /^\#/d | $(SED) -e s/XCOMM/\#/ > $@
 # Tell versions [3.59,3.63) of GNU make to not export all variables.
 # Otherwise a system limit (for SysV at least) may be exceeded.
 .NOEXPORT:
