--- rock-generic_etc_profile	2003-05-18 00:12:08.000000000 +0200
+++ rock-desktop_etc_profile	2003-05-18 00:54:13.000000000 +0200
@@ -43,7 +43,8 @@
 #
 if [ "$PS1" ] ; then
 	PS1='$ ' ; PS2='> ' ; PS4='+ '
-	[ "$BASH_VERSION" ] && PS1='\u@\h:\w\$ '
+	color="1;32m" ; [ "$UID" = 0 ] && color="0;31m"
+	[ "$BASH_VERSION" ] && PS1="\[\033[$color\]\u@\h:\w\$\[\033[0m\] "
 	[ "$KSH_VERSION" -a   -w / ] && PS1='$USER:$PWD# '
 	[ "$KSH_VERSION" -a ! -w / ] && PS1='$USER:$PWD# '
 fi
