--- kdebase-3.5.4/startkde~	2006-07-23 16:23:24.000000000 +0200
+++ kdebase-3.5.4/startkde	2006-09-11 10:46:20.000000000 +0200
@@ -263,10 +263,13 @@
     kpersonalizer --before-session
     # handle kpersonalizer restarts (language change)
     while test $? -eq 1; do
+        setxkbmap `kreadconfig --file kdeglobals --group Locale --key Language`
         kpersonalizer --r --before-session
     done
     dcopquit kwin
     dcopserver_shutdown --wait
+else
+    setxkbmap `kreadconfig --file kdeglobals --group Locale --key Language`
 fi
 
 # the splashscreen and progress indicator
--- kdebase-3.5.4/kdm/config.def.vanilla	2006-09-11 10:51:08.000000000 +0200
+++ kdebase-3.5.4/kdm/config.def	2006-09-11 10:52:34.000000000 +0200
@@ -1854,7 +1854,7 @@
 
 Key: AutoLoginEnable
 Type: bool
-Default: false
+Default: true
 User: dep
 Instance: #:0/true
 Comment: &
@@ -1886,7 +1886,7 @@
 
 Key: AutoLoginUser
 Type: string
-Default: ""
+Default: "guest"
 PostProc: PautoLoginX
 User: core(autoUser)
 User: greeter
@@ -2096,7 +2096,7 @@
 # This needs to come _in front_ of the font settings to be effective!
 Key: AntiAliasing
 Type: bool
-Default: false
+Default: true
 User: greeter
 Instance: #*/!
 Comment: &

Prettification.

diff -ur kdebase-3.5.5-old/kcontrol/style/kcmstyle.cpp kdebase-3.5.5/kcontrol/style/kcmstyle.cpp
--- kdebase-3.5.5-old/kcontrol/style/kcmstyle.cpp.orig	2006-10-01 19:31:49.000000000 +0200
+++ kdebase-3.5.0/kcontrol/style/kcmstyle.cpp	2006-10-12 13:47:41.000000000 +0200
@@ -229,6 +229,14 @@
 	containerLayout->addWidget( lblTooltipEffect, 1, 0 );
 	containerLayout->addWidget( comboTooltipEffect, 1, 1 );
 
+	comboRubberbandEffect = new QComboBox( FALSE, containerFrame );
+	comboRubberbandEffect->insertItem( i18n("Disable") );
+	comboRubberbandEffect->insertItem( i18n("Make translucent") );
+	lblRubberbandEffect = new QLabel( i18n("&Rubberband effect:"), containerFrame );
+	lblRubberbandEffect->setBuddy( comboRubberbandEffect );
+	containerLayout->addWidget( lblRubberbandEffect, 2, 0 );
+	containerLayout->addWidget( comboRubberbandEffect, 2, 1 );
+
 	comboMenuEffect = new QComboBox( FALSE, containerFrame );
 	comboMenuEffect->insertItem( i18n("Disable") );
 	comboMenuEffect->insertItem( i18n("Animate") );
@@ -236,8 +244,8 @@
 	comboMenuEffect->insertItem( i18n("Make Translucent") );
 	lblMenuEffect = new QLabel( i18n("&Menu effect:"), containerFrame );
 	lblMenuEffect->setBuddy( comboMenuEffect );
-	containerLayout->addWidget( lblMenuEffect, 2, 0 );
-	containerLayout->addWidget( comboMenuEffect, 2, 1 );
+	containerLayout->addWidget( lblMenuEffect, 3, 0 );
+	containerLayout->addWidget( comboMenuEffect, 3, 1 );
 
 	comboMenuHandle = new QComboBox( FALSE, containerFrame );
 	comboMenuHandle->insertItem( i18n("Disable") );
@@ -245,11 +253,11 @@
 //	comboMenuHandle->insertItem( i18n("Enable") );
 	lblMenuHandle = new QLabel( i18n("Me&nu tear-off handles:"), containerFrame );
 	lblMenuHandle->setBuddy( comboMenuHandle );
-	containerLayout->addWidget( lblMenuHandle, 3, 0 );
-	containerLayout->addWidget( comboMenuHandle, 3, 1 );
+	containerLayout->addWidget( lblMenuHandle, 4, 0 );
+	containerLayout->addWidget( comboMenuHandle, 4, 1 );
 
 	cbMenuShadow = new QCheckBox( i18n("Menu &drop shadow"), containerFrame );
-	containerLayout->addWidget( cbMenuShadow, 4, 0 );
+	containerLayout->addWidget( cbMenuShadow, 5, 0 );
 
 	// Push the [label combo] to the left.
 	comboSpacer = new QSpacerItem( 20, 20, QSizePolicy::Expanding, QSizePolicy::Minimum );
@@ -356,22 +364,23 @@
 	// Do all the setDirty connections.
 	connect(cbStyle, SIGNAL(activated(int)), this, SLOT(setStyleDirty()));
 	// Page2
-	connect( cbEnableEffects,     SIGNAL(toggled(bool)),    this, SLOT(setEffectsDirty()));
-	connect( cbEnableEffects,     SIGNAL(toggled(bool)),    this, SLOT(setStyleDirty()));
-	connect( comboTooltipEffect,  SIGNAL(activated(int)), this, SLOT(setEffectsDirty()));
-	connect( comboComboEffect,    SIGNAL(activated(int)), this, SLOT(setEffectsDirty()));
-	connect( comboMenuEffect,     SIGNAL(activated(int)), this, SLOT(setStyleDirty()));
-	connect( comboMenuHandle,     SIGNAL(activated(int)), this, SLOT(setStyleDirty()));
-	connect( comboMenuEffectType, SIGNAL(activated(int)), this, SLOT(setStyleDirty()));
-	connect( slOpacity,           SIGNAL(valueChanged(int)),this, SLOT(setStyleDirty()));
-	connect( cbMenuShadow,        SIGNAL(toggled(bool)),    this, SLOT(setStyleDirty()));
+	connect( cbEnableEffects,       SIGNAL(toggled(bool)),    this, SLOT(setEffectsDirty()));
+	connect( cbEnableEffects,       SIGNAL(toggled(bool)),    this, SLOT(setStyleDirty()));
+	connect( comboTooltipEffect,    SIGNAL(activated(int)),   this, SLOT(setEffectsDirty()));
+	connect( comboRubberbandEffect, SIGNAL(activated(int)),   this, SLOT(setStyleDirty()));
+	connect( comboComboEffect,      SIGNAL(activated(int)),   this, SLOT(setEffectsDirty()));
+	connect( comboMenuEffect,       SIGNAL(activated(int)),   this, SLOT(setStyleDirty()));
+	connect( comboMenuHandle,       SIGNAL(activated(int)),   this, SLOT(setStyleDirty()));
+	connect( comboMenuEffectType,   SIGNAL(activated(int)),   this, SLOT(setStyleDirty()));
+	connect( slOpacity,             SIGNAL(valueChanged(int)),this, SLOT(setStyleDirty()));
+	connect( cbMenuShadow,          SIGNAL(toggled(bool)),    this, SLOT(setStyleDirty()));
 	// Page3
 	connect( cbHoverButtons,        SIGNAL(toggled(bool)),    this, SLOT(setToolbarsDirty()));
 	connect( cbTransparentToolbars, SIGNAL(toggled(bool)),    this, SLOT(setToolbarsDirty()));
 	connect( cbEnableTooltips,      SIGNAL(toggled(bool)),    this, SLOT(setEffectsDirty()));
 	connect( cbIconsOnButtons,      SIGNAL(toggled(bool)),    this, SLOT(setEffectsDirty()));
 	connect( cbTearOffHandles,      SIGNAL(toggled(bool)),    this, SLOT(setEffectsDirty()));
-	connect( comboToolbarIcons,     SIGNAL(activated(int)), this, SLOT(setToolbarsDirty()));
+	connect( comboToolbarIcons,     SIGNAL(activated(int)),   this, SLOT(setToolbarsDirty()));
 
 	addWhatsThis();
 
@@ -556,6 +565,11 @@
 	item = comboTooltipEffect->currentItem();
 	config.writeEntry( "EffectAnimateTooltip", item == 1);
 	config.writeEntry( "EffectFadeTooltip", item == 2 );
+	item = comboRubberbandEffect->currentItem();
+	{
+		QSettings settings;	// Only for KStyle stuff
+		settings.writeEntry("/KStyle/Settings/SemiTransparentRubberband", item == 1);
+	}
 	item = comboMenuHandle->currentItem();
 	config.writeEntry( "InsertTearOffHandle", item );
 	item = comboMenuEffect->currentItem();
@@ -888,6 +902,12 @@
 	else
 		comboTooltipEffect->setCurrentItem( 0 );
 
+	comboRubberbandEffect->setCurrentItem(1);
+
+	QSettings settings;
+	bool semiTransparentRubberband = settings.readBoolEntry("/KStyle/Settings/SemiTransparentRubberband", true);
+	comboRubberbandEffect->setCurrentItem( semiTransparentRubberband ? 1 : 0 );
+
 	if ( config.readBoolEntry( "EffectAnimateMenu", false) )
 		comboMenuEffect->setCurrentItem( 1 );
 	else if ( config.readBoolEntry( "EffectFadeMenu", false) )
@@ -898,7 +918,6 @@
 	comboMenuHandle->setCurrentItem(config.readNumEntry("InsertTearOffHandle", 0));
 
 	// KStyle Menu transparency and drop-shadow options...
-	QSettings settings;
 	QString effectEngine = settings.readEntry("/KStyle/Settings/MenuTransparencyEngine", "Disabled");
 
 #ifdef HAVE_XRENDER
@@ -1027,6 +1046,8 @@
 	QWhatsThis::add( comboTooltipEffect, i18n( "<p><b>Disable: </b>do not use any tooltip effects.</p>\n"
 							"<p><b>Animate: </b>Do some animation.</p>\n"
 							"<b>Fade: </b>Fade in tooltips using alpha-blending.") );
+	QWhatsThis::add( comboRubberbandEffect, i18n( "<p><b>Disable: </b>do not use any rubberband effects.</p>\n"
+							"<b>Make Translucent: </b>Draw a translucent rubberband.") );
 	QWhatsThis::add( comboMenuEffect, i18n( "<p><b>Disable: </b>do not use any menu effects.</p>\n"
 							"<p><b>Animate: </b>Do some animation.</p>\n"
 							"<p><b>Fade: </b>Fade in menus using alpha-blending.</p>\n"
diff -ur kdebase-3.5.0-old/kcontrol/style/kcmstyle.h kdebase-3.5.0/kcontrol/style/kcmstyle.h
--- kdebase-3.5.0-old/kcontrol/style/kcmstyle.h	2005-10-10 10:03:48.000000000 -0500
+++ kdebase-3.5.0/kcontrol/style/kcmstyle.h	2005-12-17 16:00:46.773291232 -0600
@@ -125,11 +125,13 @@
 	QFrame* containerFrame;
 	QGridLayout* containerLayout;
 	QComboBox* comboTooltipEffect;
+	QComboBox* comboRubberbandEffect;
 	QComboBox* comboComboEffect;
 	QComboBox* comboMenuEffect;
 	QComboBox* comboMenuHandle;
 
 	QLabel* lblTooltipEffect;
+	QLabel* lblRubberbandEffect;
 	QLabel* lblComboEffect;
 	QLabel* lblMenuEffect;
 	QLabel* lblMenuHandle;
diff -u kdebase-3.5.6/kdesktop/kfileividesktop.cpp.orig kdebase-3.5.6/kdesktop/kfileividesktop.cpp
--- kdebase-3.5.6/kdesktop/kfileividesktop.cpp.orig	2007-01-15 12:32:10.000000000 +0100
+++ kdebase-3.5.6/kdesktop/kfileividesktop.cpp	2007-01-26 11:36:05.096065775 +0100
@@ -101,6 +101,14 @@
   // draw the pixmap as in KIconViewItem::paintItem(...)
   paintPixmap(p, colors);
 
+  if ( isSelected() ) {
+    p->setBrush( QBrush( cg.highlight() ) );
+    p->setPen( QPen( cg.highlight() ) );
+    p->drawRoundRect( textRect( FALSE ),
+                      1000 / textRect( FALSE ).width(),
+                      1000 / textRect( FALSE ).height() );
+  }
+
   //
   // Paint the text as shadowed if the shadow is available
   //
@@ -167,7 +175,7 @@
 
 void KFileIVIDesktop::drawShadowedText( QPainter *p, const QColorGroup &cg )
 {
-  int textX = textRect( FALSE ).x() + 2;
+  int textX = textRect( FALSE ).x() + 4;
   int textY = textRect( FALSE ).y();
   int align = ((KIconView *) iconView())->itemTextPos() == QIconView::Bottom
     ? AlignHCenter : AlignAuto;
@@ -184,12 +192,15 @@
   int spread = shadowThickness();
 
   if ( isSelected() && settings->selectionType() != KShadowSettings::InverseVideoOnSelection ) {
-    // select using a filled rect
     text = cg.highlightedText();
     QRect rect = textRect( false );
     rect.setRight( rect.right() - spread );
     rect.setBottom( rect.bottom() - spread + 1 );
-    p->fillRect( rect, cg.highlight() );
+    p->setBrush( QBrush( cg.highlight() ) );
+    p->setPen( QPen( cg.highlight() ) );
+    p->drawRoundRect( rect,
+		      1000 / rect.width(),
+		      1000 / rect.height() );
   }
   else {
     // use shadow
diff -ur kdebase-3.5.0-old/kdesktop/kfileividesktop.h kdebase-3.5.0/kdesktop/kfileividesktop.h
--- kdebase-3.5.0-old/kdesktop/kfileividesktop.h	2005-10-10 10:04:11.000000000 -0500
+++ kdebase-3.5.0/kdesktop/kfileividesktop.h	2005-12-17 16:00:46.913269952 -0600
@@ -80,12 +80,6 @@
     virtual void paintItem(QPainter *p, const QColorGroup &cg);
 
     /**
-     * Reimplements QIconView::paintFocus() to take the shadow
-     * metrics into account();
-     */
-    virtual void paintFocus( QPainter *p, const QColorGroup &cg );
-
-    /**
      * Draws the shadow text.
      * @param p the painter for drawing the item
      * @param cg the base color group
@@ -101,6 +95,8 @@
      */
     virtual QImage *buildShadow(QPainter *p, const int align, QColor &shadowColor);
 
+    void paintFocus(QPainter*, const QColorGroup&);
+
  protected:
     void setNormalImage(QImage *newImage) { delete m_normalImage; m_normalImage = newImage; };
     void setSelectedImage(QImage *newImage) { delete m_selectedImage; m_selectedImage = newImage; };
diff -ur kdebase-3.5.0-old/konqueror/listview/konq_listviewwidget.cc kdebase-3.5.0/konqueror/listview/konq_listviewwidget.cc
--- kdebase-3.5.0-old/konqueror/listview/konq_listviewwidget.cc	2005-10-10 10:04:14.000000000 -0500
+++ kdebase-3.5.0/konqueror/listview/konq_listviewwidget.cc	2005-12-17 16:03:21.375788080 -0600
@@ -405,7 +405,6 @@
 {
    if ( m_rubber )
    {
-      drawRubber();
       delete m_rubber;
       m_rubber = 0;
    }
@@ -417,14 +416,17 @@
    KonqBaseListViewItem* item = isExecuteArea( vp ) ?
          static_cast<KonqBaseListViewItem*>( itemAt( vp ) ) : 0L;
 
-   if ( item )
+   if ( item ) {
       KListView::contentsMousePressEvent( e );
+   }
    else {
       if ( e->button() == LeftButton )
       {
          if ( !( e->state() & ControlButton ) )
             setSelected( itemAt( vp ), false );
          m_rubber = new QRect( e->x(), e->y(), 0, 0 );
+         clearSelection();
+         emit selectionChanged();
          m_fileTip->setItem( 0 );
       }
       if ( e->button() != RightButton )
@@ -440,9 +442,10 @@
 {
    if ( m_rubber )
    {
-      drawRubber();
+      QRect r( m_rubber->normalize() );
       delete m_rubber;
       m_rubber = 0;
+      repaintContents( r, FALSE );
    }
 
    if ( m_scrollTimer )
@@ -534,23 +537,21 @@
    KListView::leaveEvent( e );
 }
 
-void KonqBaseListViewWidget::drawRubber()
+void KonqBaseListViewWidget::drawRubber( QPainter *p )
 {
    if ( !m_rubber )
       return;
 
-   QPainter p;
-   p.begin( viewport() );
-   p.setRasterOp( NotROP );
-   p.setPen( QPen( color0, 1 ) );
-   p.setBrush( NoBrush );
+   p->setRasterOp( NotROP );
+   p->setPen( QPen( color0, 1 ) );
+   p->setBrush( NoBrush );
 
    QPoint pt( m_rubber->x(), m_rubber->y() );
    pt = contentsToViewport( pt );
-   style().drawPrimitive( QStyle::PE_FocusRect, &p,
+   style().drawPrimitive( QStyle::PE_FocusRect, p,
                           QRect( pt.x(), pt.y(), m_rubber->width(), m_rubber->height() ),
                           colorGroup(), QStyle::Style_Default, colorGroup().base() );
-   p.end();
+   p->end();
 }
 
 void KonqBaseListViewWidget::slotAutoScroll()
@@ -566,10 +567,11 @@
    if ( vc == m_rubber->bottomRight() )
       return;
 
+   QRect oldRubber = *m_rubber;
+
    const int oldTop = m_rubber->normalize().top();
    const int oldBottom = m_rubber->normalize().bottom();
 
-   drawRubber();
    m_rubber->setBottomRight( vc );
 
    QListViewItem *cur = itemAt( QPoint(0,0) );
@@ -577,10 +579,19 @@
    bool block = signalsBlocked();
    blockSignals( true );
 
+   QRect rr;
    QRect nr = m_rubber->normalize();
+   bool changed = FALSE;
    if ( cur )
    {
-      QRect rect = itemRect( cur );
+      QRect rect;
+      if ( allColumnsShowFocus() )
+          rect = itemRect( cur );
+      else {
+          rect = itemRect( cur );
+          rect.setWidth( executeArea( cur ) );
+      }
+
       if ( !allColumnsShowFocus() )
           rect.setWidth( executeArea( cur ) );
 
@@ -606,10 +617,25 @@
          if ( rect.intersects( nr ) )
          {
             if ( !cur->isSelected() && cur->isSelectable() )
+            {
                setSelected( cur, true );
-         } else if ( !m_selected || !m_selected->contains( (KonqBaseListViewItem*)cur ) )
-            setSelected( cur, false );
-
+               changed=TRUE;
+               rr = rr.unite( itemRect( cur ) );
+            }
+         }
+         else
+         {
+            if ( cur->isSelected() )
+            {
+               changed = TRUE;
+               rr = rr.unite( itemRect( cur ) );
+            }
+
+            if ( !m_selected || !m_selected->contains( (KonqBaseListViewItem*)cur ) )
+            {
+               setSelected( cur, false );
+            }
+         }
          cur = cur->itemBelow();
          if (cur && !allColumnsShowFocus())
             rect.setWidth( executeArea( cur ) );
@@ -625,9 +651,22 @@
          if ( rect.intersects( nr ) )
          {
             if ( !cur->isSelected() && cur->isSelectable() )
+            {
                setSelected( cur, true );
-         } else if ( !m_selected || !m_selected->contains( (KonqBaseListViewItem*)cur ) )
-            setSelected( cur, false );
+               changed = TRUE;
+               rr = rr.unite( itemRect( cur ) );
+            }
+         }
+         else
+         {
+            if ( cur->isSelected() )
+            {
+               changed = TRUE;
+               rr = rr.unite( itemRect( cur ) );
+            }
+            if ( !m_selected || !m_selected->contains( (KonqBaseListViewItem*)cur ) )
+               setSelected( cur, false );
+         }
 
          cur = cur->itemAbove();
          if (cur && !allColumnsShowFocus())
@@ -639,7 +678,31 @@
    blockSignals( block );
    emit selectionChanged();
 
-   drawRubber();
+   QRect allRect = oldRubber.normalize();
+   if ( changed )
+   {
+       allRect |= rr.normalize();
+   }
+   allRect |= m_rubber->normalize();
+   QPoint point = contentsToViewport( allRect.topLeft() );
+   allRect = QRect( point.x(), point.y(), allRect.width(), allRect.height() );
+   allRect &= viewport()->rect();
+   allRect.addCoords( -2, -2, 2, 2 );
+
+   QPixmap backrubber( viewport()->rect().size() );
+   backrubber.fill( viewport(), viewport()->rect().topLeft() );
+
+   QPainter p( &backrubber );
+   p.save();
+   drawContentsOffset( &p,
+      contentsX(),
+      contentsY(),
+      contentsX() + allRect.left(), contentsY() + allRect.top(),
+      allRect.width(), allRect.height() );
+   p.restore();
+   drawRubber( &p );
+   p.end();
+   bitBlt( viewport(), allRect.topLeft(), &backrubber, allRect );
 
    const int scroll_margin = 40;
    ensureVisible( vc.x(), vc.y(), scroll_margin, scroll_margin );
@@ -669,9 +732,11 @@
 
 void KonqBaseListViewWidget::viewportPaintEvent( QPaintEvent *e )
 {
-   drawRubber();
    KListView::viewportPaintEvent( e );
-   drawRubber();
+
+   QPainter p( viewport() );
+   drawRubber( &p );
+   p.end();
 }
 
 void KonqBaseListViewWidget::viewportResizeEvent(QResizeEvent * e)
Only in kdebase-3.5.0/konqueror/listview: konq_listviewwidget.cc.orig
diff -ur kdebase-3.5.0-old/konqueror/listview/konq_listviewwidget.h kdebase-3.5.0/konqueror/listview/konq_listviewwidget.h
--- kdebase-3.5.0-old/konqueror/listview/konq_listviewwidget.h	2005-10-10 10:04:14.000000000 -0500
+++ kdebase-3.5.0/konqueror/listview/konq_listviewwidget.h	2005-12-17 16:00:47.076245176 -0600
@@ -206,7 +206,7 @@
    virtual void viewportPaintEvent( QPaintEvent *e );
    virtual void viewportResizeEvent( QResizeEvent *e );
 
-   virtual void drawRubber();
+   virtual void drawRubber( QPainter * );
    virtual void contentsMousePressEvent( QMouseEvent *e );
    virtual void contentsMouseReleaseEvent( QMouseEvent *e );
    virtual void contentsMouseMoveEvent( QMouseEvent *e );
@@ -246,6 +246,7 @@
    QColor m_itemColor;
 
    QRect *m_rubber;
+   QPixmap *m_backrubber;
 
    bool m_bTopLevelComplete:1;
    bool m_showIcons:1;
diff -ur kdebase-3.5.0-old/konqueror/sidebar/konqsidebar.cpp kdebase-3.5.0/konqueror/sidebar/konqsidebar.cpp
--- kdebase-3.5.0-old/konqueror/sidebar/konqsidebar.cpp	2005-09-10 03:25:47.000000000 -0500
+++ kdebase-3.5.0/konqueror/sidebar/konqsidebar.cpp	2005-12-17 16:00:47.129237120 -0600
@@ -56,7 +56,8 @@
 	return true;
 }
 
-bool KonqSidebar::openURL(const KURL &url) {
+bool KonqSidebar::openURL(const KURL &url)
+{
 	if (m_widget)
 		return m_widget->openURL(url);
 	else return false;
@@ -73,8 +74,6 @@
 	}
 }
 
-
-
 // It's usually safe to leave the factory code alone.. with the
 // notable exception of the KAboutData data
 #include <kaboutdata.h>
diff -ur kdebase-3.5.0-old/konqueror/sidebar/konqsidebartng.rc kdebase-3.5.0/konqueror/sidebar/konqsidebartng.rc
--- kdebase-3.5.0-old/konqueror/sidebar/konqsidebartng.rc	2005-09-10 03:25:47.000000000 -0500
+++ kdebase-3.5.0/konqueror/sidebar/konqsidebartng.rc	2005-12-17 16:00:47.162232104 -0600
@@ -1,7 +1,12 @@
 [filemanagement]
+[filemanagement]
 OpenViews=home.desktop
 SingleWidgetMode=true
 
 [webbrowsing]
 OpenViews=bookmarks.desktop
 SingleWidgetMode=true
+
+[webbrowsing]
+OpenViews=bookmarks.desktop
+SingleWidgetMode=true
diff -ur kdebase-3.5.0-old/konqueror/sidebar/konqsidebartng.upd kdebase-3.5.0/konqueror/sidebar/konqsidebartng.upd
--- kdebase-3.5.0-old/konqueror/sidebar/konqsidebartng.upd	2005-09-10 03:25:47.000000000 -0500
+++ kdebase-3.5.0/konqueror/sidebar/konqsidebartng.upd	2005-12-17 16:00:47.171230736 -0600
@@ -10,3 +10,15 @@
 
 Id=konqsidebartng_entries
 Script=move_konqsidebartng_entries.sh,sh
+Id=konqsidebartng_rc
+File=konqsidebartng.rc
+Group=<default>,webbrowsing
+Options=Copy
+AllKeys
+Group=<default>,filemanagement
+Options=Copy
+AllKeys
+RemoveGroup=<default>
+
+Id=konqsidebartng_entries
+Script=move_konqsidebartng_entries.sh,sh
diff -ur kdebase-3.5.0-old/konqueror/sidebar/move_konqsidebartng_entries.sh kdebase-3.5.0/konqueror/sidebar/move_konqsidebartng_entries.sh
--- kdebase-3.5.0-old/konqueror/sidebar/move_konqsidebartng_entries.sh	2005-09-10 03:25:47.000000000 -0500
+++ kdebase-3.5.0/konqueror/sidebar/move_konqsidebartng_entries.sh	2005-12-17 16:00:47.173230432 -0600
@@ -16,3 +16,21 @@
 done
 
 rm -rf $source/entries
+#!/bin/sh
+
+prefix=$(kde-config --localprefix)
+source="$prefix/share/apps/konqsidebartng"
+
+[ -d "$source/entries" ] || exit 0
+
+profiles="filemanagement webbrowsing"
+for profile in $profiles; do
+	dest="$source/$profile/entries"
+	if [ ! -d "$dest" ]; then
+		mkdir -p "$dest" || exit 1
+		cp $source/entries/.version $dest/
+		cp $source/entries/* $dest/
+	fi
+done
+
+rm -rf $source/entries
diff -ur kdebase-3.5.0-old/konqueror/sidebar/sidebar_widget.cpp kdebase-3.5.0/konqueror/sidebar/sidebar_widget.cpp
--- kdebase-3.5.0-old/konqueror/sidebar/sidebar_widget.cpp	2005-09-10 03:25:47.000000000 -0500
+++ kdebase-3.5.0/konqueror/sidebar/sidebar_widget.cpp	2005-12-17 16:00:47.242219944 -0600
@@ -14,6 +14,7 @@
  *   (at your option) any later version.                                   *
  *                                                                         *
  ***************************************************************************/
+
 #include <config.h>
 
 #include <limits.h>
@@ -25,6 +26,11 @@
 #include <qwhatsthis.h>
 #include <qlayout.h>
 #include <qstringlist.h>
+#include <qdockwindow.h>
+#include <qlabel.h>
+#include <qtoolbox.h>
+#include <qpainter.h>
+#include <kpushbutton.h>
 
 #include <klocale.h>
 #include <ksimpleconfig.h>
diff -ur kdebase-3.5.0-old/konqueror/sidebar/sidebar_widget.h kdebase-3.5.0/konqueror/sidebar/sidebar_widget.h
--- kdebase-3.5.0-old/konqueror/sidebar/sidebar_widget.h	2005-09-10 03:25:47.000000000 -0500
+++ kdebase-3.5.0/konqueror/sidebar/sidebar_widget.h	2005-12-17 16:00:47.244219640 -0600
@@ -4,7 +4,8 @@
     begin                : Sat June 2 16:25:27 CEST 2001
     copyright            : (C) 2001 Joseph Wenninger
     email                : jowenn@kde.org
- ***************************************************************************/
+
+***************************************************************************/
 
 /***************************************************************************
  *                                                                         *
@@ -14,6 +15,7 @@
  *   (at your option) any later version.                                   *
  *                                                                         *
  ***************************************************************************/
+
 #ifndef _SIDEBAR_WIDGET_
 #define _SIDEBAR_WIDGET_
 
