--- kdebase-3.5.4/startkde~	2006-07-23 16:23:24.000000000 +0200
+++ kdebase-3.5.4/startkde	2006-09-11 10:46:20.000000000 +0200
@@ -263,10 +263,13 @@
     kpersonalizer --before-session
     # handle kpersonalizer restarts (language change)
     while test $? -eq 1; do
+        setxkbmap `kreadconfig --file kdeglobals --group Locale --key Language`
         kpersonalizer --r --before-session
     done
     dcopquit kwin
     dcopserver_shutdown --wait
+else
+    setxkbmap `kreadconfig --file kdeglobals --group Locale --key Language`
 fi
 
 # the splashscreen and progress indicator
--- kdebase-3.5.4/kdm/config.def.vanilla	2006-09-11 10:51:08.000000000 +0200
+++ kdebase-3.5.4/kdm/config.def	2006-09-11 10:52:34.000000000 +0200
@@ -1854,7 +1854,7 @@
 
 Key: AutoLoginEnable
 Type: bool
-Default: false
+Default: true
 User: dep
 Instance: #:0/true
 Comment: &
@@ -1886,7 +1886,7 @@
 
 Key: AutoLoginUser
 Type: string
-Default: ""
+Default: "live"
 PostProc: PautoLoginX
 User: core(autoUser)
 User: greeter
@@ -2096,7 +2096,7 @@
 # This needs to come _in front_ of the font settings to be effective!
 Key: AntiAliasing
 Type: bool
-Default: false
+Default: true
 User: greeter
 Instance: #*/!
 Comment: &
