--- httpd-2.0.48/srclib/apr-util/configure.orig	2003-10-24 19:51:13.000000000 +0200
+++ httpd-2.0.48/srclib/apr-util/configure	2004-01-14 17:10:38.000000000 +0100
@@ -13923,19 +13924,19 @@
 
   places="$all_places"
   if test -z "$places"; then
-    places="std /usr/local/BerkeleyDB.4.1"
+    places="std /usr/local/BerkeleyDB.4.2"
   fi
 
   bdb_version="4"
   if test ""1"" != "-1"; then
-    bdb_version="$bdb_version."1""
+    bdb_version="$bdb_version."2""
     if test ""-1"" != "-1"; then
       bdb_version="$bdb_version."-1""
     fi
   fi
   bdb_places="$places"
-  bdb_default_search_headers="db4/db.h db.h"
-  bdb_default_search_lib_names="db-4.1 db4 db"
+  bdb_default_search_headers="db42/db.h db4/db.h db.h"
+  bdb_default_search_lib_names="db-4.2 db-4.1 db4 db"
 
 
   apu_have_db=0
