--- gnomemeeting-1.2.1/configure.orig	2005-08-26 08:41:33.000000000 -0700
+++ gnomemeeting-1.2.1/configure	2005-08-26 08:41:58.000000000 -0700
@@ -21902,7 +21902,7 @@
 
 echo "$as_me:$LINENO: checking for Quicknet support in OpenH323" >&5
 echo $ECHO_N "checking for Quicknet support in OpenH323... $ECHO_C" >&6;
-OPENH323_IXJ_SUPPORT=`cat ${with_openh323_dir}/include/openh323/openh323buildopts.h | grep "HAS_IXJ 1" | cut -f2  -d ' '`
+OPENH323_IXJ_SUPPORT=`cat ${with_openh323_dir}/include/openh323/openh323buildopts.h | grep "HAS_IXJ 1" | cut -f3  -d ' '`
 if test "x${OPENH323_IXJ_SUPPORT}" != "x1"; then
 	echo "$as_me:$LINENO: result: no" >&5
 echo "${ECHO_T}no" >&6;
