--- alleyoop-0.8.2/src/symtab.c~	2004-04-02 07:05:05.000000000 +0200
+++ alleyoop-0.8.2/src/symtab.c	2004-10-25 00:16:12.644737016 +0200
@@ -115,7 +115,7 @@
 	
 	map->text_section = section;
 	map->text_start = load_addr + bfd_section_vma (map->abfd, section);
-	map->text_end = map->text_start + bfd_get_section_size_before_reloc (section);
+	map->text_end = map->text_start + bfd_get_section_size (section);
 	
 	map->filename = g_strdup (filename);
 	map->libname = g_strdup (libname);
