--- at76c503a/at76c503-fw_skel.c.vanilla	2006-05-18 10:46:37.772394500 +0200
+++ at76c503a/at76c503-fw_skel.c	2006-05-18 10:46:57.637636000 +0200
@@ -89,10 +89,10 @@
 /* structure for registering this driver with the usb subsystem */
 
 static struct usb_driver module_usb = {
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,4,20)
+	.driver = {
 	owner:      THIS_MODULE,
-#endif
 	name:	    DRIVER_NAME,
+	},
 	probe:	    at76c50x_probe,
 	disconnect: at76c50x_disconnect,
 	id_table:   dev_table,
