diff -ur tetex-src-2.0.2/Makefile.in tetex-src-2.0.2-fixed/Makefile.in
--- tetex-src-2.0.2/Makefile.in	2003-05-08 09:56:20.030430704 +0200
+++ tetex-src-2.0.2-fixed/Makefile.in	2003-05-08 09:48:43.000000000 +0200
@@ -44,7 +44,7 @@
   kpsewhich mag makeindex mf mf-nowin mft mfw mpost mpto newer \
   odvicopy odvips odvitype omega omfonts otangle otp2ocp outocp oxdvi.bin \
   patgen pdfetex pdftex pfb2pfa pk2bm pktogf pktype pltotf pooltype ps2pk \
-  readlink tangle tcdialog tex tftopl tie ttf2afm vftovp vptovf \
+  tangle tcdialog tex tftopl tie ttf2afm vftovp vptovf \
   weave xdvi.bin
 
 # Subdirectories that have makefiles
--- tetex-src-2.0.2/texk/kpathsea/Makefile.in	2003-05-08 11:03:28.292041856 +0200
+++ tetex-src-2.0.2/texk/kpathsea/Makefile.in.orig	2003-02-15 18:21:42.000000000 +0100
@@ -38,7 +38,7 @@
 .c.lo:
 	$(KPATHSEA_CC) $< $(ALL_CFLAGS)
 
-default all: kpsewhich kpsestat access $(library).info manfiles
+default all: kpsewhich kpsestat access readlink $(library).info manfiles
 
 kpsewhich: $(kpathsea) kpsewhich.o
 	$(kpathsea_link) kpsewhich.o $(LOADLIBES)
@@ -184,6 +184,7 @@
 	$(INSTALL_LIBTOOL_PROG) $(bindir) kpsewhich
 	$(INSTALL_PROGRAM) kpsestat $(bindir)
 	$(INSTALL_PROGRAM) access $(bindir)
+	$(INSTALL_PROGRAM) readlink $(bindir)
 uninstall-exec:
 	rm -f $(bindir)/kpsewhich
 
@@ -264,7 +265,7 @@
 	rm -f paths.h texmf.cnf texmf.sed stamp-paths
 
 # The manual pages
-manfiles = access.1 kpsestat.1 kpsewhich.1 \
+manfiles = access.1 kpsestat.1 kpsewhich.1 readlink.1 \
            mktexlsr.1 mktexmf.1 mktextfm.1 mktexpk.1
 
 ac_include ../make/man.mk
