--- tetex-src-2.0.2/Makefile.in.old	2003-02-02 10:48:11.000000000 +0100
+++ tetex-src-2.0.2/Makefile.in	2003-05-07 09:15:15.793848096 +0200
@@ -40,11 +40,11 @@
 
 ALLPROGS = access pdftosrc afm2tfm bibtex ctangle cweave dmp dvicopy \
   dvilj dvilj2p dvilj4 dvilj4l dvipdfm dvips dvitomp dvitype ebb etex \
-  gftodvi gftopk gftype gsftopk info infokey install-info kpsestat \
-  kpsewhich mag makeindex makeinfo mf mf-nowin mft mfw mpost mpto newer \
+  gftodvi gftopk gftype gsftopk kpsestat \
+  kpsewhich mag makeindex mf mf-nowin mft mfw mpost mpto newer \
   odvicopy odvips odvitype omega omfonts otangle otp2ocp outocp oxdvi.bin \
   patgen pdfetex pdftex pfb2pfa pk2bm pktogf pktype pltotf pooltype ps2pk \
-  readlink tangle tcdialog tex texindex tftopl tie ttf2afm vftovp vptovf \
+  readlink tangle tcdialog tex tftopl tie ttf2afm vftovp vptovf \
   weave xdvi.bin
 
 # Subdirectories that have makefiles
