--- clang-21.1.1.src/clang-tools-extra-21.1.1.src/CMakeLists.txt.vanilla	2025-09-16 17:19:13.610225268 +0200
+++ clang-21.1.1.src/clang-tools-extra-21.1.1.src/CMakeLists.txt	2025-09-16 17:19:19.005883182 +0200
@@ -21,7 +21,7 @@
 add_subdirectory(clang-tidy)
 
 add_subdirectory(clang-change-namespace)
-add_subdirectory(clang-doc)
+#add_subdirectory(clang-doc)
 add_subdirectory(clang-include-fixer)
 add_subdirectory(clang-move)
 add_subdirectory(clang-query)
