Index: llvm-9.0.0-20190629/clang/lib/Driver/ToolChains/CommonArgs.cpp
===================================================================
--- llvm-9.0.0-20190629/lib/Driver/ToolChains/CommonArgs.cpp
+++ llvm-9.0.0-20190629/lib/Driver/ToolChains/CommonArgs.cpp
@@ -389,8 +389,11 @@
     if (A->getOption().matches(options::OPT_O4) ||
         A->getOption().matches(options::OPT_Ofast))
       OOpt = "3";
-    else if (A->getOption().matches(options::OPT_O))
+    else if (A->getOption().matches(options::OPT_O)) {
       OOpt = A->getValue();
+      if(OOpt == "s" || OOpt == "z")
+        OOpt = "2";
+    }
     else if (A->getOption().matches(options::OPT_O0))
       OOpt = "0";
     if (!OOpt.empty())
