--- ./configure.orig	2006-03-20 11:14:09.000000000 +0100
+++ ./configure	2006-03-20 12:54:18.000000000 +0100
@@ -3218,9 +3218,9 @@
   echo 'Missing OpenSSL (-lssl) - aborted'; exit 1
 fi
 
-echo "$as_me:$LINENO: checking for pcreposix_regcomp in -lpcreposix" >&5
-echo $ECHO_N "checking for pcreposix_regcomp in -lpcreposix... $ECHO_C" >&6
-if test "${ac_cv_lib_pcreposix_pcreposix_regcomp+set}" = set; then
+echo "$as_me:$LINENO: checking for regcomp in -lpcreposix" >&5
+echo $ECHO_N "checking for regcomp in -lpcreposix... $ECHO_C" >&6
+if test "${ac_cv_lib_pcreposix_regcomp+set}" = set; then
   echo $ECHO_N "(cached) $ECHO_C" >&6
 else
   ac_check_lib_save_LIBS=$LIBS
@@ -3238,11 +3238,11 @@
 #endif
 /* We use char because int might match the return type of a gcc2
    builtin and then its argument prototype would still apply.  */
-char pcreposix_regcomp ();
+char regcomp ();
 int
 main ()
 {
-pcreposix_regcomp ();
+regcomp ();
   ;
   return 0;
 }
@@ -3268,20 +3268,20 @@
   ac_status=$?
   echo "$as_me:$LINENO: \$? = $ac_status" >&5
   (exit $ac_status); }; }; then
-  ac_cv_lib_pcreposix_pcreposix_regcomp=yes
+  ac_cv_lib_pcreposix_regcomp=yes
 else
   echo "$as_me: failed program was:" >&5
 sed 's/^/| /' conftest.$ac_ext >&5
 
-ac_cv_lib_pcreposix_pcreposix_regcomp=no
+ac_cv_lib_pcreposix_regcomp=no
 fi
 rm -f conftest.err conftest.$ac_objext \
       conftest$ac_exeext conftest.$ac_ext
 LIBS=$ac_check_lib_save_LIBS
 fi
-echo "$as_me:$LINENO: result: $ac_cv_lib_pcreposix_pcreposix_regcomp" >&5
-echo "${ECHO_T}$ac_cv_lib_pcreposix_pcreposix_regcomp" >&6
-if test $ac_cv_lib_pcreposix_pcreposix_regcomp = yes; then
+echo "$as_me:$LINENO: result: $ac_cv_lib_pcreposix_regcomp" >&5
+echo "${ECHO_T}$ac_cv_lib_pcreposix_regcomp" >&6
+if test $ac_cv_lib_pcreposix_regcomp = yes; then
   LIBS="${LIBS} -lpcreposix"
 fi
 
