diff -Naur DirectFB-0.9.21.orig/include/dfb_types.h DirectFB-0.9.21/include/dfb_types.h
--- DirectFB-0.9.21.orig/include/dfb_types.h	2004-04-29 22:08:37.000000000 +0300
+++ DirectFB-0.9.21/include/dfb_types.h	2005-03-06 15:58:22.664606216 +0200
@@ -6,11 +6,19 @@
 #define __u8  u_int8_t
 #define __u16 u_int16_t
 #define __u32 u_int32_t
-#define __u64 u_int64_t
+
+#ifdef __x86_64__
+	#define __u64 unsigned long long int
+#else
+	#define __u64 u_int64_t
+#endif
 
 #define __s8  int8_t
 #define __s16 int16_t
 #define __s32 int32_t
-#define __s64 int64_t
-
+#ifdef __x86_64__
+	#define __s64 long long int
+#else
+	#define __s64 int64_t
+#endif
 #endif
