--- DirectFB-0.9.21/inputdrivers/linux_input/Makefile.in.orig	2004-12-21 20:21:49.000000000 +0100
+++ DirectFB-0.9.21/inputdrivers/linux_input/Makefile.in	2005-01-04 12:32:31.000000000 +0100
@@ -326,13 +326,12 @@
 linux_input.lo linux_input.o : linux_input.c ../../include/dfb_types.h \
 	../../include/directfb.h ../../include/directfb_keyboard.h \
 	../../src/core/coredefs.h ../../config.h \
-	../../src/core/fusion/fusion.h \
-	../../src/core/fusion/fusion_types.h ../../src/misc/conf.h \
-	../../src/misc/debug.h ../../src/core/coretypes.h \
-	../../src/core/input.h ../../src/core/fusion/reactor.h \
-	../../src/core/fusion/list.h ../../src/core/modules.h \
-	../../src/core/system.h ../../src/core/thread.h \
-	../../src/misc/mem.h ../../src/misc/util.h \
+	../../lib/fusion/fusion.h \
+	../../src/misc/conf.h \
+	../../src/core/coretypes.h \
+	../../src/core/input.h ../../lib/fusion/reactor.h \
+	../../src/core/system.h \
+	../../src/misc/util.h \
 	../../src/core/input_driver.h
 
 info-am:
--- DirectFB-0.9.21/inputdrivers/linux_input/linux_input.c.orig	2004-12-20 02:22:57.000000000 +0100
+++ DirectFB-0.9.21/inputdrivers/linux_input/linux_input.c	2005-01-06 20:18:54.277719128 +0100
@@ -884,7 +884,7 @@
  * Fetch one entry from the kernel keymap.
  */
 static DFBResult
-driver_get_keymap_entry( InputDevice               *device,
+driver_get_keymap_entry( CoreInputDevice           *device,
                          void                      *driver_data,
                          DFBInputDeviceKeymapEntry *entry )
 {
