--- ./configure.org	2003-03-29 19:09:28.000000000 +0100
+++ ./configure	2003-04-06 21:59:05.000000000 +0200
@@ -31,9 +31,9 @@
 	QTDIR=$QTDIR make -C src
 
 install:
-	cp -a src/qtparted \${prefix}/usr/sbin
-	mkdir -p \${prefix}/usr/share/qtparted/locale
-	cp -a src/ts/*qm \${prefix}/usr/share/qtparted/locale
+	cp -a src/qtparted \${prefix}/sbin
+	mkdir -p \${prefix}/share/qtparted/locale
+	cp -a src/ts/*qm \${prefix}/share/qtparted/locale
 
 clean:
 	QTDIR=$QTDIR make -C src clean
