--- ./configure.orig	2003-06-08 19:12:01.000000000 +0300
+++ ./configure	2003-06-19 19:17:05.000000000 +0300
@@ -6307,18 +6307,6 @@
 	< $srcdir/po/POTFILES.in > po/POTFILES
 
 
-if test "$gt_cv_have_gettext" != "yes" ; then
-  { { echo "$as_me:$LINENO: error:
-*** You must have either have gettext support in your C library, or use the
-*** GNU gettext library. (http://www.gnu.org/software/gettext/gettext.html
-" >&5
-echo "$as_me: error:
-*** You must have either have gettext support in your C library, or use the
-*** GNU gettext library. (http://www.gnu.org/software/gettext/gettext.html
-" >&2;}
-   { (exit 1); exit 1; }; }
-fi
-
 LIBS="$INTLLIBS $LIBS"
 
 GETTEXT_PACKAGE=glib20
