--- ./Makefile.orig	2003-11-09 13:50:16.061560752 +0100
+++ ./Makefile	2003-11-09 13:51:45.827914192 +0100
@@ -7,6 +7,10 @@
 
 all: bacterium_ffd bacterium_gui
 
+# place -lncurses after bacterium_ffd.c
+bacterium_ffd: bacterium_ffd.c
+	$(CC) -Wall bacterium_ffd.c -lncurses -o bacterium_ffd
+
 bacterium_gui: bacterium_gui.sh
 	sed 's,./\(bacterium_ffd\),\1,' < bacterium_gui.sh > bacterium_gui
 	chmod +x bacterium_gui
