--- ./Makefile.in.orig	2004-01-10 13:24:42.000000000 +0100
+++ ./Makefile.in	2004-01-10 13:24:55.000000000 +0100
@@ -545,7 +545,6 @@
 @kernel_support_FALSE@install-exec-local:
 @kernel_support_FALSE@	-mkdir -p $(MODINSTALLDIR)/oprofile
 @kernel_support_FALSE@	cp module/oprofile-module.o $(MODINSTALLDIR)/oprofile/oprofile.o
-@kernel_support_FALSE@	if test $(KVERS) = "`uname -r`"; then depmod -a ; fi
 
 @kernel_support_FALSE@clean-local:
 @kernel_support_FALSE@	rm -f module/*.o module/@OPROFILE_MODULE_ARCH@/*.o \
