--- hal-0.5.7/tools/hal-device.c.orig	2006-08-14 18:14:42.000000000 -0700
+++ hal-0.5.7/tools/hal-device.c	2006-08-14 18:15:03.000000000 -0700
@@ -155,7 +155,7 @@
 
 	libhal_ctx_shutdown(hal_ctx, &error);
 	libhal_ctx_free(hal_ctx);
-	dbus_connection_disconnect(conn);
+	dbus_connection_close(conn);
 	dbus_connection_unref(conn);
 	dbus_error_free(&error);
 
--- hal-0.5.7/tools/lshal.c.orig	2006-08-14 18:15:36.000000000 -0700
+++ hal-0.5.7/tools/lshal.c	2006-08-14 18:16:14.000000000 -0700
@@ -701,7 +701,7 @@
 	libhal_ctx_shutdown (hal_ctx, &error);
 	libhal_ctx_free (hal_ctx);
 
-	dbus_connection_disconnect (conn);
+	dbus_connection_close (conn);
 	dbus_connection_unref (conn);
 
 	if (show_device)
