--- hal-0.5.14/hald/linux/addons/addon-imac-backlight.c.vanilla	2013-10-22 13:48:23.495494582 +0000
+++ hal-0.5.14/hald/linux/addons/addon-imac-backlight.c	2013-10-22 13:48:45.451603457 +0000
@@ -29,7 +29,7 @@
 #include <stdlib.h>
 #include <sys/io.h>
 
-#include <glib/gmain.h>
+#include <glib.h>
 #include <dbus/dbus-glib.h>
 #include <dbus/dbus-glib-lowlevel.h>
 
--- hal-0.5.14/hald/linux/addons/addon-rfkill-killswitch.c.vanilla	2013-10-22 13:42:57.249876818 +0000
+++ hal-0.5.14/hald/linux/addons/addon-rfkill-killswitch.c	2013-10-22 13:43:21.821998664 +0000
@@ -36,7 +36,6 @@
 #include <unistd.h> 
 
 #include <glib.h>
-#include <glib/gmain.h>
 #include <glib/gstdio.h>
 #include <dbus/dbus-glib.h>
 #include <dbus/dbus-glib-lowlevel.h>
--- hal-0.5.14/hald/linux/addons/addon-ipw-killswitch.c.vanilla	2013-10-22 13:40:56.177276451 +0000
+++ hal-0.5.14/hald/linux/addons/addon-ipw-killswitch.c	2013-10-22 13:41:27.097429776 +0000
@@ -36,7 +36,6 @@
 #include <unistd.h> 
 
 #include <glib.h>
-#include <glib/gmain.h>
 #include <glib/gstdio.h>
 #include <dbus/dbus-glib.h>
 #include <dbus/dbus-glib-lowlevel.h>
--- hal-0.5.14/hald/linux/addons/addon-generic-backlight.c.vanilla	2013-10-22 13:34:35.211387343 +0000
+++ hal-0.5.14/hald/linux/addons/addon-generic-backlight.c	2013-10-22 13:38:34.692574866 +0000
@@ -35,7 +35,7 @@
 #include <sys/stat.h>
 #include <unistd.h> 
 
-#include <glib/gmain.h>
+#include <glib.h>
 #include <dbus/dbus-glib.h>
 #include <dbus/dbus-glib-lowlevel.h>
 
--- hal-0.5.14/hald/linux/addons/addon-leds.c.vanilla	2013-10-22 13:41:58.441585203 +0000
+++ hal-0.5.14/hald/linux/addons/addon-leds.c	2013-10-22 13:42:27.497729285 +0000
@@ -36,7 +36,6 @@
 #include <unistd.h> 
 
 #include <glib.h>
-#include <glib/gmain.h>
 #include <glib/gstdio.h>
 #include <dbus/dbus-glib.h>
 #include <dbus/dbus-glib-lowlevel.h>
--- hal-0.5.14/hald/linux/addons/addon-macbookpro-backlight.c.vanilla	2013-10-22 13:50:32.136132477 +0000
+++ hal-0.5.14/hald/linux/addons/addon-macbookpro-backlight.c	2013-10-22 13:50:54.372242740 +0000
@@ -39,7 +39,7 @@
 #include <pci/pci.h>
 #include <unistd.h> 
 
-#include <glib/gmain.h>
+#include <glib.h>
 #include <dbus/dbus-glib.h>
 #include <dbus/dbus-glib-lowlevel.h>
 
--- hal-0.5.14/hald/linux/addons/addon-storage.c.vanilla	2013-10-22 13:44:33.534354267 +0000
+++ hal-0.5.14/hald/linux/addons/addon-storage.c	2013-10-22 13:45:04.374507195 +0000
@@ -39,7 +39,7 @@
 #include <string.h>
 #include <sys/ioctl.h>
 #include <unistd.h>
-#include <glib/gmain.h>
+#include <glib.h>
 #include <dbus/dbus-glib.h>
 #include <dbus/dbus-glib-lowlevel.h>
 
--- hal-0.5.14/hald/linux/addons/addon-input.c.vanilla	2013-10-22 13:39:29.280845555 +0000
+++ hal-0.5.14/hald/linux/addons/addon-input.c	2013-10-22 13:40:08.489039978 +0000
@@ -41,7 +41,7 @@
 #else
   #include <linux/input.h>
 #endif
-#include <glib/gmain.h>
+#include <glib.h>
 #include <glib/gprintf.h>
 #include <dbus/dbus-glib-lowlevel.h>
 
