--- h/fdi/policy/10osvendor/10-storage-policy.fdi	2005-05-23 16:06:27.000000000 +0200
+++ h/fdi/policy/10osvendor/10-storage-policy.fdi	2005-12-04 15:57:47.000000000 +0100
@@ -130,6 +130,11 @@
 	      <merge key="volume.policy.should_mount" type="bool">false</merge>
 	    </match>
 	  </match>
+
+          <!-- Attempt to form human readable names -->
+          <match key="@block.storage_device:info.vendor" exists="true">
+             <merge key="volume.policy.desired_mount_point" type="copy_property">@block.storage_device:info.vendor</merge>
+          </match>
 	  
 	  <!-- Attempt mount point 'ipod' for iPod's -->
 	  <match key="@block.storage_device:portable_audio_player.type" string="ipod">
