--- srm-1.2.8/src/sunlink.c.orig	2007-02-01 12:05:13.000000000 -0800
+++ srm-1.2.8/src/sunlink.c	2007-02-01 12:06:08.000000000 -0800
@@ -12,9 +12,6 @@
 #include <time.h>
 #include <unistd.h>
 
-#if HAVE_LINUX_EXT2_FS_H
-#include <linux/ext2_fs.h>
-#endif
 
 #if HAVE_SYS_VFS_H
 #include <sys/vfs.h>
@@ -24,6 +21,10 @@
 #include <sys/param.h>
 #include <sys/mount.h>
 #endif
+#if HAVE_LINUX_EXT2_FS_H
+#include <linux/fs.h>
+#include <linux/ext2_fs.h>
+#endif
 
 #include "srm.h"
 
