--- bacula-1.38.11/src/console/console.c.orig	2006-11-27 16:12:02.000000000 -0800
+++ bacula-1.38.11/src/console/console.c	2006-11-27 16:13:56.000000000 -0800
@@ -75,7 +75,7 @@
 static BSOCK *UA_sock = NULL;
 static DIRRES *dir;
 static FILE *output = stdout;
-static bool tee = false;                  /* output to output and stdout */
+static bool do_tee = false;                  /* output to output and stdout */
 static bool stop = false;
 static int argc;
 static int numdir;
@@ -702,7 +702,7 @@
 {
    int len;
    if (!stop) {
-      if (output == stdout || tee) {
+      if (output == stdout || do_tee) {
          sendit(prompt);
       }
    }
@@ -781,14 +781,14 @@
 /* Send output to both termina and specified file */
 static int teecmd(FILE *input, BSOCK *UA_sock)
 {
-   tee = true;
+   do_tee = true;
    return do_outputcmd(input, UA_sock);
 }
 
 /* Send output to specified "file" */
 static int outputcmd(FILE *input, BSOCK *UA_sock)
 {
-   tee = false;
+   do_tee = false;
    return do_outputcmd(input, UA_sock);
 }
 
@@ -806,7 +806,7 @@
       if (output != stdout) {
          fclose(output);
          output = stdout;
-         tee = false;
+         do_tee = false;
       }
       return 1;
    }
@@ -865,7 +865,7 @@
 void sendit(const char *buf)
 {
 #ifdef xHAVE_CONIO
-    if (output == stdout || tee) {
+    if (output == stdout || do_tee) {
        char *p, *q;
        /*
         * Here, we convert every \n into \r\n because the
@@ -890,10 +890,10 @@
 
     fputs(buf, output);
     fflush(output);
-    if (tee) {
+    if (do_tee) {
        fputs(buf, stdout);
     }
-    if (output != stdout || tee) {
+    if (output != stdout || do_tee) {
        fflush(stdout);
     }
 #endif
