--- pybliographer-1.0.11/compiled/Makefile.pre.in.orig	2003-10-08 00:11:33.000000000 +0200
+++ pybliographer-1.0.11/compiled/Makefile.pre.in	2003-10-08 00:20:01.000000000 +0200
@@ -267,6 +267,7 @@
 	echo "/^LINKPATH=/s%=.*%=	$(LINKPATH)%" >>sedscript
 	echo "/^BASELIB=/s%=.*%=	$(BASELIB)%" >>sedscript
 	echo "/^BASESETUP=/s%=.*%=	$(BASESETUP)%" >>sedscript
+	echo "s%@DEFS[@]%%" >>sedscript
 
 # Bootstrap target
 boot:	clobber
