--- pybliographer-1.0.11/configure.orig	2003-10-02 13:41:45.000000000 +0200
+++ pybliographer-1.0.11/configure	2003-10-02 13:42:06.000000000 +0200
@@ -4316,10 +4316,6 @@
 
 version=`${RECODE} --version | head -n 1 | cut -f 3 -d' '`
 
-if test ${version} != "3.5" ; then
-  { echo "configure: error: Please check that your version of GNU Recode is 3.5" 1>&2; exit 1; }
-fi
-
 cat >> confdefs.h <<\EOF
 #define USE_RECODE 1
 EOF
