--- ./Makefile.in.orig	2005-02-28 20:01:21.282452368 -0300
+++ ./Makefile.in	2005-02-28 20:02:48.512191432 -0300
@@ -444,11 +444,8 @@
 
 install-data-local:
 	@$(NORMAL_INSTALL)
-	@if test ! -f $(DESTDIR)$(sysconfdir)/ldap.conf; then \
-		$(mkinstalldirs) $(DESTDIR)$(sysconfdir); \
-		$(INSTALL_DATA) -o $(INST_UID) -g $(INST_GID) $(srcdir)/ldap.conf $(DESTDIR)$(sysconfdir)/ldap.conf; \
-	fi
-	$(INSTALL_DATA) -o $(INST_UID) -g $(INST_GID) $(srcdir)/nsswitch.ldap $(DESTDIR)$(sysconfdir)/nsswitch.ldap; 
+	$(INSTALL_DATA) -o $(INST_UID) -g $(INST_GID) $(srcdir)/ldap.conf $(DESTDIR)$(sharedstatedir)/ldap.conf
+	$(INSTALL_DATA) -o $(INST_UID) -g $(INST_GID) $(srcdir)/nsswitch.ldap $(DESTDIR)$(sharedstatedir)/nsswitch.ldap
 
 uninstall-local:
 	@$(NORMAL_UNINSTALL)
