--- ./makefiles/Makefile.linux.orig	2005-03-21 22:44:16.164568112 -0400
+++ ./makefiles/Makefile.linux	2005-03-21 22:44:37.173374288 -0400
@@ -21,7 +21,7 @@
 
 ARFLAGS=-urs
 
-LIBS= -L/usr/X11R6/lib -Wl,-Bstatic -lXm -Wl,-Bdynamic -lXp -lXpm -lXext -lXt -lSM -lICE -lX11 -lm
+LIBS= -L/usr/X11R6/lib -Wl,-Bstatic -Wl,-Bdynamic -lXp -lXpm -lXext -lXt -lSM -lICE -lX11 -lm -lXm
 
 include Makefile.common
 
