--- ./errhnd.hpp.orig	2004-01-05 06:52:30.000000000 +0200
+++ ./errhnd.hpp	2004-01-28 19:48:12.000000000 +0200
@@ -1,7 +1,7 @@
 #ifndef _RAR_ERRHANDLER_
 #define _RAR_ERRHANDLER_
 
-#if (defined(GUI) || !defined(_WIN_32)) && !defined(SFX_MODULE) && !defined(_WIN_CE) || defined(RARDLL)
+#if (defined(GUI) || !defined(_WIN_32)) && (!defined(__GNUG__) || defined(__EXCEPTIONS)) && !defined(SFX_MODULE) && !defined(_WIN_CE) || defined(RARDLL)
 #define ALLOW_EXCEPTIONS
 #endif
 
