--- ecdplay/COMPILE.orig-rock-maint	2003-08-29 10:19:30.000000000 +0200
+++ ecdplay/COMPILE	2003-08-29 10:20:03.000000000 +0200
@@ -1,6 +1,6 @@
 #!/bin/sh
 
-cc ecdplay.c epplet.c -o ecdplay.epplet `imlib-config --libs` `imlib-config --cflags` -lcdaudio
+$CC $CFLAGS ecdplay.c epplet.c -o ecdplay.epplet `imlib-config --libs` `imlib-config --cflags` -L$E16/lib -lepplet -lcdaudio
 
 cp -f ecdplay.epplet $EBIN/ecdplay.epplet
 chmod a+x $EBIN/ecdplay.epplet
