--- E-CPU2v0.2/COMPILE.orig-rock-maint	2003-08-29 08:34:25.000000000 +0200
+++ E-CPU2v0.2/COMPILE	2003-08-29 09:02:11.000000000 +0200
@@ -24,8 +24,8 @@
 echo $CC $CFLAGS `imlib-config --cflags` -c $EPPLET.c -o $EPPLET.o
 $CC $CFLAGS `imlib-config --cflags` -c $EPPLET.c -o $EPPLET.o
 
-echo $CC $CFLAGS `imlib-config --cflags` -o $EPPLET.epplet $EPPLET.o `imlib-config --libs` -lepplet
-$CC $CFLAGS `imlib-config --cflags` -o $EPPLET.epplet $EPPLET.o  `imlib-config --libs` -lepplet
+echo $CC $CFLAGS `imlib-config --cflags` -o $EPPLET.epplet $EPPLET.o `imlib-config --libs` -L$E16/lib -lepplet
+$CC $CFLAGS `imlib-config --cflags` -o $EPPLET.epplet $EPPLET.o  `imlib-config --libs` -L$E16/lib -lepplet
 
 # Install E-NetFlame to the Enlightenment /bin dir
 echo install -m 755 ./$EPPLET.epplet $EBIN/$EPPLET.epplet
