--- E-Dial/COMPILE.orig-rock-maint	2003-08-29 08:39:04.000000000 +0200
+++ E-Dial/COMPILE	2003-08-29 09:03:36.000000000 +0200
@@ -2,7 +2,7 @@
 # Run this to install E-Dial on your system.
 
 NAME=E-Dial
-cc edial.c epplet.c -o $NAME.epplet -O2 `imlib-config --libs --cflags`
+$CC $CFLAGS edial.c epplet.c -o $NAME.epplet -O2 `imlib-config --libs --cflags` -L$E16/lib -lepplet
 cp -f $NAME.epplet $EBIN/$NAME.epplet
 chmod a+x $EBIN/$NAME.epplet
 
