--- E-Mount_0.3/Makefile.orig-rock-maint	2003-08-29 09:31:58.000000000 +0200
+++ E-Mount_0.3/Makefile	2003-08-29 09:32:49.000000000 +0200
@@ -2,13 +2,13 @@
 
 IMLIB_CFLAGS = `imlib-config --cflags`
 IMLIB_LIBS = `imlib-config --libs`
-GCCOPTS = -MMD -Wall
+GCCOPTS = -MMD -Wall $(CFLAGS)
 GCC = gcc
 
 all : E-Mount.epplet
 
 E-Mount.epplet : E-Mount.c
-	${GCC} ${GCCOPTS}  E-Mount.c -o E-Mount.epplet -lepplet ${IMLIB_LIBS} ${IMLIB_CFLAGS}
+	${GCC} ${GCCOPTS}  E-Mount.c -o E-Mount.epplet -L$(E16)/lib -lepplet ${IMLIB_LIBS} ${IMLIB_CFLAGS}
 
 install : E-Mount.epplet
 	install -m 755 E-Mount.epplet ${EBIN}/E-Mount.epplet
