diff -ruN unison-2.9.20/Makefile.OCaml unison-2.9.20-new/Makefile.OCaml
--- unison-2.9.20/Makefile.OCaml	2002-08-19 23:15:04.000000000 +0200
+++ unison-2.9.20-new/Makefile.OCaml	2004-05-24 12:43:49.546990904 +0200
@@ -58,7 +58,7 @@
 # to Unix '/' separators, and extraneous control-M's are deleted.
 # Unfortunately there is a literal control-M buried in this, I'd rather
 # get rid of it...
-OCAMLLIBDIR=$(shell ocamlc -v | tail -1 | sed -e 's/.* //g' | sed -e 's/\\/\//g' | sed -e 's///g')
+OCAMLLIBDIR=$(shell ocamlc -v | tail -n 1 | sed -e 's/.* //g' | sed -e 's/\\/\//g' | sed -e 's///g')
 
 # User interface style: 
 #   Legal values are
