diff -ruN gzip-1.2.4a/getopt.c gzip-1.2.4a-new/getopt.c
--- gzip-1.2.4a/getopt.c	1993-08-17 20:24:08.000000000 +0200
+++ gzip-1.2.4a-new/getopt.c	2004-05-18 22:01:12.830713688 +0200
@@ -172,7 +172,7 @@
 #if __STDC__ || defined(PROTO)
 extern char *getenv(const char *name);
 extern int  strcmp (const char *s1, const char *s2);
-extern int  strncmp(const char *s1, const char *s2, int n);
+extern int  strncmp(const char *s1, const char *s2, size_t n);
 
 static int my_strlen(const char *s);
 static char *my_index (const char *str, int chr);
