--- ./Makefile.orig	Sat Apr 19 12:28:47 2003
+++ ./Makefile	Sat Apr 19 12:28:59 2003
@@ -103,7 +103,6 @@
 	rm -f $(DESTDIR)/usr/share/man/man8/smartd.8
 	install -m 755 -D smartctl         $(DESTDIR)/usr/sbin/smartctl
 	install -m 755 -D smartd           $(DESTDIR)/usr/sbin/smartd
-	install -m 755 -D smartd.initd     $(DESTDIR)/etc/rc.d/init.d/smartd
 	install -m 644 -D smartctl.8.gz    $(DESTDIR)/usr/share/man/man8/smartctl.8.gz
 	install -m 644 -D smartd.8.gz      $(DESTDIR)/usr/share/man/man8/smartd.8.gz
 	install -m 644 -D smartd.conf.5.gz $(DESTDIR)/usr/share/man/man5/smartd.conf.5.gz
