--- ./Makefile.orig	2003-04-23 15:56:17.000000000 +0200
+++ ./Makefile	2003-07-02 00:33:06.000000000 +0200
@@ -118,7 +118,6 @@
 	rm -f $(DESTDIR)/usr/share/man/man8/smartd.8
 	$(INSTALL_PROGRAM) -D smartctl         $(DESTDIR)/usr/sbin/smartctl
 	$(INSTALL_PROGRAM) -D smartd           $(DESTDIR)/usr/sbin/smartd
-	$(INSTALL_PROGRAM) -D smartd.initd     $(DESTDIR)/etc/rc.d/init.d/smartd
 	$(INSTALL_DATA) -D smartctl.8.gz    $(DESTDIR)/usr/share/man/man8/smartctl.8.gz
 	$(INSTALL_DATA) -D smartd.8.gz      $(DESTDIR)/usr/share/man/man8/smartd.8.gz
 	$(INSTALL_DATA) -D smartd.conf.5.gz $(DESTDIR)/usr/share/man/man5/smartd.conf.5.gz
