--- ./Makefile.orig	Sat Apr 19 12:14:19 2003
+++ ./Makefile	Sat Apr 19 12:24:09 2003
@@ -118,12 +118,8 @@
 	install -m 755 -D $(es)/Example1   $(DESTDIR)/usr/share/doc/smartmontools-5.1/$(es)/Example1
 	install -m 755 -D $(es)/Example2   $(DESTDIR)/usr/share/doc/smartmontools-5.1/$(es)/Example2
 	install -m 755 -D $(es)/Example3   $(DESTDIR)/usr/share/doc/smartmontools-5.1/$(es)/Example3
-	install -m 644 -D smartd.conf      $(DESTDIR)/etc/smartd.conf.example
-	if [ ! -f $(DESTDIR)/etc/smartd.conf ] ; then install -m 644 -D smartd.conf $(DESTDIR)/etc/smartd.conf ; fi
-	@echo -e "\n\nTo manually start smartd on bootup, run /etc/rc.d/init.d/smartd start"
-	@echo "To automatically start smartd on bootup, run /sbin/chkconfig --add smartd"
 	@echo -e "\n\nSmartd can now use a configuration file /etc/smartd.conf. Do:\nman 8 smartd\n."
-	@echo -e "A sample configuration file may be found in /usr/share/doc/smartmontools-5.1 and /etc/smartd.conf.example/\n\n"
+	@echo -e "A sample configuration file may be found in /usr/share/doc/smartmontools-5.1\n\n"
 
 # perhaps for consistency I should also have $(DESTDIR) for the uninstall...
 uninstall:
