--- xf86-video-via-0.1.31.1_orig/src/via_memory.c	2005-08-11 20:28:29.000000000 +0200
+++ xf86-video-via-0.1.31.1/src/via_memory.c	2005-11-02 12:16:43.718820592 +0100
@@ -110,7 +110,7 @@
 	    mem->drm_fd = pVia->drmFD;
 	    mem->drm.context = 1;
 	    mem->drm.size = size;
-	    mem->drm.type = VIDEO;
+	    mem->drm.type = VIA_MEM_VIDEO;
 	    ret = drmCommandWriteRead(mem->drm_fd, DRM_VIA_ALLOCMEM, &mem->drm, 
 				      sizeof(drm_via_mem_t));
 	    if (ret || (size != mem->drm.size)) {
--- xf86-video-via-0.1.31.1_orig/src/xvmc/unichrome/viaLowLevel.c	2005-06-29 00:49:21.000000000 +0200
+++ xf86-video-via-0.1.31.1/src/xvmc/unichrome/viaLowLevel.c	2005-11-02 12:07:07.991344472 +0100
@@ -285,7 +285,7 @@
     if (xl->use_agp) {
 	xl->tsMem.context = *(xl->drmcontext);
 	xl->tsMem.size = 64;
-	xl->tsMem.type = VIDEO;
+	xl->tsMem.type = VIA_MEM_VIDEO;
 	if (drmCommandWriteRead(xl->fd, DRM_VIA_ALLOCMEM, &xl->tsMem, sizeof(xl->tsMem)) < 0) 
 	    return ret;
 	if (xl->tsMem.size != 64)
--- xf86-video-via-0.1.31.1_orig/src/xvmc/unichromeProA/viaLowLevelPro.c	2005-08-11 21:32:48.000000000 +0200
+++ xf86-video-via-0.1.31.1/src/xvmc/unichromeProA/viaLowLevelPro.c	2005-11-02 12:13:11.003158288 +0100
@@ -535,7 +535,7 @@
     if (xl->use_agp) {
 	xl->tsMem.context = *(xl->drmcontext);
 	xl->tsMem.size = 64;
-	xl->tsMem.type = VIDEO;
+	xl->tsMem.type = VIA_MEM_VIDEO;
 	if ((ret = drmCommandWriteRead(xl->fd, DRM_VIA_ALLOCMEM, 
 				       &xl->tsMem, sizeof(xl->tsMem))) < 0) 
 	    return ret;
@@ -1339,7 +1339,7 @@
 	    drmCommandWrite(xl->fd, DRM_VIA_FREEMEM, mem, sizeof(*mem)); 
 	mem->context = *(xl->drmcontext);
 	mem->size = size;
-	mem->type = VIDEO;
+	mem->type = VIA_MEM_VIDEO;
 	
 	if (((ret = drmCommandWriteRead(xl->fd, DRM_VIA_ALLOCMEM, mem, sizeof(mem))) < 0)  ||
 	    mem->size != size) {
