--- linux-5.4/include/uapi/linux/mount.h.vanilla	2022-11-01 13:00:35.615893986 +0100
+++ linux-5.4/include/uapi/linux/mount.h	2022-11-01 13:00:50.726892413 +0100
@@ -85,6 +85,8 @@
 #define FSPICK_NO_AUTOMOUNT	0x00000004
 #define FSPICK_EMPTY_PATH	0x00000008
 
+#if 0
+
 /*
  * The type of fsconfig() call made.
  */
@@ -104,6 +106,8 @@
  */
 #define FSMOUNT_CLOEXEC		0x00000001
 
+#endif
+
 /*
  * Mount attributes.
  */
