--- qemu-0.7.1/configure~	2005-07-25 18:41:13.000000000 +0200
+++ qemu-0.7.1/configure	2005-07-25 18:45:01.000000000 +0200
@@ -718,7 +718,7 @@
   echo "#define TARGET_ARCH \"x86_64\"" >> $config_h
   echo "#define TARGET_I386 1" >> $config_h
   echo "#define TARGET_X86_64 1" >> $config_h
-  if test $kqemu = "yes" -a "$target_softmmu" = "yes" -a $cpu = "x86_64"  ; then
+  if test $kernel_module = "yes" -a "$target_softmmu" = "yes" -a $cpu = "x86_64"  ; then
     echo "#define USE_KQEMU 1" >> $config_h
   fi
 elif test "$target_cpu" = "mips" ; then
