2006-02-12  Gwenole Beauchesne  <gbeauchesne@mandriva.com>

	* Enforce 16-byte boundaries.

--- qemu-0.8.3/Makefile.target.enforce-16byte-boundaries	2007-01-24 22:40:21.000000000 +0100
+++ qemu-0.8.3/Makefile.target	2007-01-30 18:23:10.000000000 +0100
@@ -74,7 +74,7 @@ OP_CFLAGS = -Wall -O2 -g -fno-strict-ali
 
 ifeq ($(ARCH),i386)
 HELPER_CFLAGS+=-fomit-frame-pointer
-OP_CFLAGS+=-mpreferred-stack-boundary=2 -fomit-frame-pointer
+OP_CFLAGS+=-fomit-frame-pointer
 ifeq ($(HAVE_GCC3_OPTIONS),yes)
 OP_CFLAGS+= -falign-functions=0 -fno-gcse
 else
