--- libgnomeui-2.8.0/libgnomeui/libgnomeui-2.0.pc.in.orig	2004-10-04 12:59:07.781292704 +0200
+++ libgnomeui-2.8.0/libgnomeui/libgnomeui-2.0.pc.in	2004-10-04 13:00:07.363234872 +0200
@@ -6,7 +6,7 @@
 
 Name: libgnomeui
 Description: libgnomeui
-Requires: libgnome-2.0 libgnomecanvas-2.0 gtk+-2.0 gdk-pixbuf-2.0 libart-2.0 gconf-2.0 libbonoboui-2.0
+Requires: libgnome-2.0 libgnomecanvas-2.0 gtk+-2.0 gdk-pixbuf-2.0 libart-2.0 gconf-2.0 libbonoboui-2.0 gnome-keyring-1
 Version: @VERSION@
 Libs: -L${libdir} -lgnomeui-2 @SM_LIBS@
 Cflags: -I${includedir}/libgnomeui-2.0
