--- libgda-0.12.1/configure~	2003-06-08 20:17:08.000000000 +0200
+++ libgda-0.12.1/configure	2003-07-17 22:45:54.000000000 +0200
@@ -11808,6 +11808,7 @@
 		{ echo "$as_me:$LINENO: WARNING: LDAP backend not used" >&5
 echo "$as_me: WARNING: LDAP backend not used" >&2;}
 	else
+		LDAP_CFLAGS="-I${ldapdir}/include"
 		LDAP_LIBS="-L${ldapdir}/lib -lldap -llber"
 		cat >>confdefs.h <<\_ACEOF
 #define HAVE_LDAP 1
