--- indent-2.2.9/src/output.c.vanilla	2005-09-02 15:54:08.000000000 +0000
+++ indent-2.2.9/src/output.c	2005-09-02 15:54:24.000000000 +0000
@@ -23,7 +23,7 @@
 
 RCSTAG_CC ("$Id: output.c,v 1.5 2002/12/12 17:36:49 david Exp $");
 
-static FILE            * output       = NULL;
+FILE            * output       = NULL;
 static BOOLEAN           inhibited    = 0;
 static buf_break_st_ty * buf_break_list = NULL;
 
