--- qdvdauthor-0.0.10/qdvdauthor/log.cpp.orig	2005-07-31 17:52:47.000000000 +0200
+++ qdvdauthor-0.0.10/qdvdauthor/log.cpp	2005-07-31 17:41:39.000000000 +0200
@@ -1,5 +1,6 @@
 
 #include <qstring.h>
+#include <stdio.h>
 #include "log.h"
 
 Log::Log (int iLogLevel)
--- qdvdauthor-0.0.10/qdvdauthor/menupreview.cpp.orig	2005-07-31 17:51:59.000000000 +0200
+++ qdvdauthor-0.0.10/qdvdauthor/menupreview.cpp	2005-07-31 17:52:16.000000000 +0200
@@ -24,6 +24,8 @@
 #include <qtimer.h>
 #include <qimage.h>
 
+#include <stdlib.h>
+
 #include "global.h"
 #include "dvdmenuundoobject.h"
 #include "xml_dvd.h"
