--- clucene-core-2.3.3.4/src/core/CLucene/util/VoidMap.h.vanilla	2026-06-01 10:55:17.152738905 +0200
+++ clucene-core-2.3.3.4/src/core/CLucene/util/VoidMap.h	2026-06-01 10:55:52.592400183 +0200
@@ -83,7 +83,7 @@
 	_vt get( _kt k) const {
 		const_iterator itr = base::find(k);
 		if ( itr==base::end() )
-			return (_vt)NULL;
+			return (_vt)0;
 		else
 			return itr->second;
 	}
--- clucene-core-2.3.3.4/src/core/CLucene/index/DocumentsWriter.cpp.vanilla	2026-06-01 10:56:25.098174646 +0200
+++ clucene-core-2.3.3.4/src/core/CLucene/index/DocumentsWriter.cpp	2026-06-01 10:59:04.826320305 +0200
@@ -125,7 +125,7 @@
   if (this->postingsFreeListDW.values){
       if (this->postingsFreeCountDW < this->postingsFreeListDW.length) {
           memset(this->postingsFreeListDW.values + this->postingsFreeCountDW
-              , NULL
+              , 0
               , sizeof(Posting*));
       }
       postingsFreeListDW.deleteUntilNULL();
