--- ./configure.vanilla	2006-05-23 21:57:34.423083500 +0200
+++ ./configure	2006-05-23 21:58:02.460835750 +0200
@@ -507,7 +507,7 @@
     mingw32msvc) system_name=MINGW32 ;;
   esac
   # We need to convert underscores so that values like k6-2 and pentium-mmx can be passed
-  host_arch=`echo $_target | cut -d '-' -f 1 | tr '_' '-'`
+  host_arch=`echo $_target | cut -d '-' -f 1`
 fi
 
 echo "Detected operating system: $system_name"
