--- ./cpudetect.c.orig	2004-01-21 14:52:42.000000000 +0100
+++ ./cpudetect.c	2004-01-21 14:57:28.000000000 +0100
@@ -551,6 +551,7 @@
 /* no Darwin, do it the brute-force way */
 /* this is borrowed from the libmpeg2 library */
         {
+	/*
           signal (SIGILL, sigill_handler);
           if (sigsetjmp (jmpbuf, 1)) {
             signal (SIGILL, SIG_DFL);
@@ -565,6 +566,7 @@
             signal (SIGILL, SIG_DFL);
             caps->hasAltiVec = 1;
           }
+	*/
         }
 #endif /* SYS_DARWIN */
         mp_msg(MSGT_CPUDETECT,MSGL_INFO,"AltiVec %sfound\n", (caps->hasAltiVec ? "" : "not "));
