diff -Nur LPRng-3.8.20-org/man/Makefile.in LPRng-3.8.20/man/Makefile.in
--- LPRng-3.8.20-org/man/Makefile.in	2003-04-21 13:44:41.000000000 +0200
+++ LPRng-3.8.20/man/Makefile.in	2003-04-21 14:06:02.000000000 +0200
@@ -121,8 +121,10 @@
 		    ${SRC}/mkinstalldirs $(DESTDIR)$(MAN)/man$${suffix}; \
 	    fi; \
 	    for i in *.$$suffix; do \
-		echo   $(INSTALL) -m 644 $$i $(DESTDIR)$(MAN)/man$$suffix/`basename $$i`$(MANEXT); \
-		$(INSTALL)   -m 644 $$i $(DESTDIR)$(MAN)/man$$suffix/`basename $$i`$(MANEXT); \
+	        ext=`echo $$i | sed -e 's/^.*\\.//'`; \
+	        inst=`echo $$i | sed -e 's/\\.[0-9a-z]*$$//'`; \
+		echo   $(INSTALL) -m 644 $$i $(DESTDIR)$(MAN)/man$$suffix/`basename $$inst-lprng.$$ext`$(MANEXT); \
+		$(INSTALL)   -m 644 $$i $(DESTDIR)$(MAN)/man$$suffix/`basename $$inst-lprng.$$ext`$(MANEXT); \
 	    done; \
 	  fi; \
 	done;
