--- ./regex/Makefile.in.orig	2005-05-22 00:59:13.577959464 -0400
+++ ./regex/Makefile.in	2005-05-22 01:03:24.691260232 -0400
@@ -63,9 +63,11 @@
 RANLIB = @RANLIB@
 VERSION = @VERSION@
 
+KERNELSOURCE = /usr/src/linux
+MODVERSIONS_H = $(KERNELSOURCE)/include/linux/modversions.h
 AUTOMAKE_OPTIONS = foreign no-dependencies
 lib_LIBRARIES = librxspencer.a
-INCLUDES = -DPOSIX_MISTAKE -D__KERNEL__ -DMODULE -DEXPORT_SYMTAB -DMODVERSIONS  -I/usr/src/linux/include -include /usr/src/linux/include/linux/modversions.h
+INCLUDES = -DPOSIX_MISTAKE -D__KERNEL__ -DMODULE -DEXPORT_SYMTAB -DMODVERSIONS  -I$(KERNELSOURCE)/include -include $(MODVERSIONS_H)
 
 librxspencer_a_SOURCES = regcomp.c regerror.c regexec.c regfree.c kernel.c
 noinst_HEADERS = cclass.h cname.h engine.c regex2.h utils.h
