--- rivatv-0.8.4/Makefile.in	2004-02-24 20:44:54.000000000 +0100
+++ rivatv-0.8.4/Makefile.in.patched	2004-05-12 17:46:56.000000000 +0200
@@ -69,14 +69,14 @@
 	make -C $(KERNEL) SUBDIRS=`pwd`/src EXTRA_CFLAGS="-I`pwd`/bttv @CFLAGS@" modules
 
 devices:
-	@if ! test -c /dev/video0 ; then \
-	  echo Creating video devices. \
-	  mknod /dev/video0 c 81 0; chmod 660 /dev/video0; \
-	  mknod /dev/video1 c 81 1; chmod 660 /dev/video1; \
-	  mknod /dev/video2 c 81 2; chmod 660 /dev/video2; \
-	  mknod /dev/video3 c 81 3; chmod 660 /dev/video3; \
-	  ln -s /dev/video0 /dev/video; \
-	fi
+#	@if ! test -c /dev/video0 ; then \
+#	  echo Creating video devices. \
+#	  mknod /dev/video0 c 81 0; chmod 660 /dev/video0; \
+#	  mknod /dev/video1 c 81 1; chmod 660 /dev/video1; \
+#	  mknod /dev/video2 c 81 2; chmod 660 /dev/video2; \
+#	  mknod /dev/video3 c 81 3; chmod 660 /dev/video3; \
+#	  ln -s /dev/video0 /dev/video; \
+#	fi
 
 config:
 	@-./configure
