--- rivatv-0.8.4/Makefile.in	2004-02-24 20:44:54.000000000 +0100
+++ rivatv-0.8.4.patched/Makefile.in	2004-05-13 23:46:46.000000000 +0200
@@ -99,7 +99,7 @@
 	else \
 	install -m 644 -o 0 -g 0 $$ofiles $(MODULES)$(MOD_DEST); \
 	fi; \
-	$(DEPMOD); \
+	\
 	echo; \
 	echo "RivaTV installed."; \
 	echo "To load RivaTV, login as root and \`modprobe rivatv\`."; \
@@ -107,12 +107,12 @@
         fi
 
 install-ko:
-	install -m 644 -o 0 -g 0 $(TARGETS:.o=.ko) $(MODULES)$(MOD_DEST); \
-	$(DEPMOD);
+	install -m 644 -o 0 -g 0 $(TARGETS:.o=.ko) $(MODULES)$(MOD_DEST); 
+
 
 uninstall:
 	@(cd $(MODULES)$(MOD_DEST) && rm -f $(SOURCES:src/%.c=%.o) $(BTTV_SOURCES:bttv/%.c=%.o) rivatv.o)
-	@$(DEPMOD)
+
 	@echo
 	@echo RivaTV uninstalled.
 	@echo
