diff -dur rivatv-0.8.4/src/rivatv-convert.c rivatv-0.8.4-patched/src/rivatv-convert.c
--- rivatv-0.8.4/src/rivatv-convert.c	2004-02-23 17:27:12.000000000 +0100
+++ rivatv-0.8.4-patched/src/rivatv-convert.c	2004-05-19 20:54:32.000000000 +0200
@@ -73,7 +73,6 @@
 /* indicates whether MMX code is requested or not */
 int mmx = 1;
 
-#ifndef RIVATV_DISABLE_CONVERSION
 #if defined (__i386__) || defined (__x86_64__)
 
 /* kernel_fpu_begin() is not exported (properly).  Made a local
@@ -105,7 +104,6 @@
 
 #endif
 #endif
-#endif
 
 void rivatv_detect_mmx (void)
 {
