--- binutils-2.23.2/gas/config/tc-arm.c.vanilla	2013-07-25 16:00:37.856682814 +0200
+++ binutils-2.23.2/gas/config/tc-arm.c	2013-07-25 16:02:27.661679018 +0200
@@ -885,7 +885,8 @@
 static inline int
 skip_past_char (char ** str, char c)
 {
-  if (**str == c)
+    skip_whitespace (*str);
+    if (**str == c)
     {
       (*str)++;
       return SUCCESS;
@@ -5168,7 +5169,8 @@
       return PARSE_OPERAND_SUCCESS;
     }
 
-  if ((reg = arm_reg_parse (&p, REG_TYPE_RN)) == FAIL)
+    skip_whitespace (p);
+    if ((reg = arm_reg_parse (&p, REG_TYPE_RN)) == FAIL)
     {
       inst.error = _(reg_expected_msgs[REG_TYPE_RN]);
       return PARSE_OPERAND_FAIL;
