--- ./config.h	2003-04-06 00:42:38.000000000 +0200
+++ ./config.h	2003-04-06 00:43:18.000000000 +0200
@@ -5,7 +5,7 @@
 #define HAVE_LC_MESSAGES 1
 #define HAVE_STPCPY 1
 #define HAVE_LIBSM 1
-#define PACKAGE_LOCALE_DIR "/usr/local/share/locale"
+#define PACKAGE_LOCALE_DIR "/usr/share/locale"
 /* #undef PACKAGE_DATA_DIR */
 /* #undef PACKAGE_SOURCE_DIR */
 
--- ./lib/utils.h	2003-03-26 08:35:38.000000000 +0100
+++ ./lib/utils.h	2003-04-06 00:43:54.000000000 +0200
@@ -57,8 +57,8 @@
 
 #define DAYS_MONTH 30
 
-#define IMG_DIRECTORY "/usr/local/share/senken/img/"
-#define GAMES_DIRECTORY "/usr/local/share/senken/games/"
+#define IMG_DIRECTORY "/usr/share/senken/img/"
+#define GAMES_DIRECTORY "/usr/share/senken/games/"
 
 #include "game.h"
 #include "map.h"
