--- bluefish-1.0/data/Makefile.in.orig	2005-01-12 21:38:49.000000000 +0100
+++ bluefish-1.0/data/Makefile.in	2005-01-23 20:44:16.269972992 +0100
@@ -62,7 +62,8 @@
 	$(INSTALL_DATA) bluefish.xml $(DESTDIR)$(FDMIME)/packages
 
 install-update-mime:
-	$(UPDATE_MIME) $(FDMIME)
+#	$(UPDATE_MIME) $(FDMIME)
+	echo "NOTE: update-mime-database should be run later"
 
 install-gnome-2.4-mime:
 	$(INSTALL_DIR) $(DESTDIR)$(GMIMEINFO)
