--- ./conf_file.c.orig	2000-10-23 04:36:18.000000000 +0200
+++ ./conf_file.c	2003-05-28 08:42:42.000000000 +0200
@@ -1,4 +1,4 @@
-/* ============================================================================
+	/* ============================================================================
  * Copyright (C) 1999 Angus Mackay. All rights reserved; 
  *
  * This program is free software; you can redistribute it and/or modify
@@ -37,13 +37,16 @@
 
 #include <conf_file.h>
 
-#if HAVE_STRERROR
+#if HAVE_ERRNO_H
+#  include <errno.h>
+#elif HAVE_STRERROR
 extern int errno;
 #  define error_string strerror(errno)
-#elif HAVE_SYS_ERRLIST
-extern const char *const sys_errlist[];
-extern int errno;
-#  define error_string (sys_errlist[errno])
+#endif
+
+#if HAVE_SYS_ERRLIST
+//extern char *const strerror[];
+#  define error_string (strerror[errno])
 #else
 #  define error_string "error message not found"
 #endif
@@ -85,7 +88,7 @@ int parse_conf_file(char *fname, struct 
   {
     if((in=fopen(fname, "r")) == NULL)
     {
-      fprintf(stderr, "could not open config file \"%s\": %s\n", fname, error_string);
+      fprintf(stderr, "could not open config file \"%s\"\n", fname);
       return(-1);
     }
   }
