--- kdegraphics-3.5.2/kuickshow/src/Makefile.am.orig	2005-09-10 10:19:04.000000000 +0200
+++ kdegraphics-3.5.2/kuickshow/src/Makefile.am	2006-04-18 21:52:54.000000000 +0200
@@ -7,6 +7,7 @@
 # set the include path for X, qt and KDE
 INCLUDES = $(all_includes)
 KDE_CXXFLAGS = $(IMLIB_CFLAGS)
+LIBS=$(shell imlib-config --libs) -lm
 METASOURCES = AUTO
 
 kuickshow_la_LDFLAGS = $(all_libraries) -module -avoid-version
