diff -urN -x '*~' libcss-0.1.0/src/css_cat.c libcss-0.1.0-gcc-3.3.1/src/css_cat.c
--- libcss-0.1.0/src/css_cat.c	2001-02-13 22:59:27.000000000 +0100
+++ libcss-0.1.0-gcc-3.3.1/src/css_cat.c	2003-08-19 22:43:44.000000000 +0200
@@ -60,14 +60,14 @@
 
 #define STCODE(p,a,b,c,d) ((p)[0] == a && (p)[1] == b && (p)[2] == c && (p)[3] == d)
 
-u_long long offset = 0x0;
-u_long long offset_end = 0x249a4;
+unsigned long long offset = 0x0;
+unsigned long long offset_end = 0x249a4;
 
 /**
  *
  */
 
-static void super_seek (int fd, u_long long offset)
+static void super_seek (int fd, unsigned long long offset)
 {
 	int i;
 
