--- ./setup.sh	2003-02-20 11:40:14.000000000 +0000
+++ ./setup.sh	2003-04-05 16:37:29.000000000 +0000
@@ -382,7 +382,6 @@
 			echo "Webmin does not support being started at boot time on your system."
 		fi
 	fi
-	makeboot=$atboot
 
 	# Create webserver config file
 	echo $perl > $config_dir/perl-path
@@ -421,11 +420,7 @@
 	fi
 
 	ufile=$config_dir/miniserv.users
-	if [ "$crypt" != "" ]; then
-		echo "$login:$crypt:0" > $ufile
-	else
-		$perl -e 'print "$ARGV[0]:",crypt($ARGV[1], "XX"),":0\n"' "$login" "$password" > $ufile
-	fi
+	touch $ufile
 	chmod 600 $ufile
 	echo "userfile=$ufile" >> $cfile
 
@@ -462,7 +457,7 @@
 	echo "Creating access control file.."
 	afile=$config_dir/webmin.acl
 	rm -f $afile
-	echo "$login: $allmods" >> $afile
+	touch $afile
 	chmod 600 $afile
 	echo "..done"
 	echo ""
