--- ./configure.orig	2004-05-14 11:24:03.748744120 +0200
+++ ./configure	2004-05-14 11:24:15.483960096 +0200
@@ -1766,7 +1766,7 @@
  	for dir in /usr/local/freetype /usr/local /opt/freetype /usr/pkg /usr; do
  		ac_cv_ft_lib=no
  		ac_cv_ft_include=no
-		shlib1=`echo $dir/lib/libttf.so* | head -1 | sed -e 's/.*\///'`
+		shlib1=`echo $dir/lib/libttf.so* | head -n 1 | sed -e 's/.*\///'`
 		if test x$shlib1 = x; then
 			shlib1=no
 		fi
@@ -1962,8 +1962,8 @@
 		ac_cv_vf_libname=VFlib
 		ac_cv_vf_libdir=no
 		ac_cv_vf_hdrdir=no
-		shlib1=`echo $dir/lib/libVFlib.so* | head -1 | sed -e 's/.*\///'`
-		shlib2=`echo $dir/lib/libVFlib2.so* | head -1 | sed -e 's/.*\///'`
+		shlib1=`echo $dir/lib/libVFlib.so* | head -n 1 | sed -e 's/.*\///'`
+		shlib2=`echo $dir/lib/libVFlib2.so* | head -n 1 | sed -e 's/.*\///'`
 		if test x$shlib1 = x; then
 			$shlib1=no
 		fi
