--- ./src/Makefile.orig	2004-11-08 00:27:05.000000000 -0300
+++ ./src/Makefile	2004-11-08 00:29:35.000000000 -0300
@@ -14,7 +14,7 @@
 STATIC	=
 
 # For Debian we do not build all programs, otherwise we do.
-PROGS	= init halt shutdown killall5 runlevel sulogin \
+PROGS	= halt shutdown killall5 sulogin \
 		last mesg wall
 
 BIN_OWNER	= root
@@ -78,15 +78,14 @@
 distclean:	clobber
 
 install:
-		$(INSTALL) -m 755 halt init killall5 sulogin \
-			runlevel shutdown $(ROOT)/sbin
+		$(INSTALL) -m 755 halt killall5 sulogin \
+			shutdown $(ROOT)/sbin
 		# These are not installed by default
 		$(INSTALL) -m 555 wall $(ROOT)/usr/bin
 		# $(INSTALL) -m 755 etc/initscript.sample $(ROOT)/etc
 		$(INSTALL) -m 755 mesg last $(ROOT)/usr/bin
 		cd $(ROOT)/sbin; ln -sf halt reboot; chown $(BIN_COMBO) reboot
 		cd $(ROOT)/sbin; ln -sf halt poweroff; chown $(BIN_COMBO) poweroff
-		cd $(ROOT)/sbin; ln -sf init telinit; chown $(BIN_COMBO) telinit
 		cd $(ROOT)/bin; ln -sf ../sbin/killall5 pidof; chown $(BIN_COMBO) pidof
 		cd $(ROOT)/usr/bin; ln -sf last lastb; chown $(BIN_COMBO) lastb
 		$(INSTALL) -m 644 initreq.h $(ROOT)/usr/include
