diff -r -u -N loudmouth-0.17.2.orig/loudmouth/lm-ssl.c loudmouth-0.17.2/loudmouth/lm-ssl.c
--- loudmouth-0.17.2.orig/loudmouth/lm-ssl.c	2004-10-19 09:52:48.000000000 +0200
+++ loudmouth-0.17.2/loudmouth/lm-ssl.c	2005-03-28 02:41:04.520131624 +0200
@@ -71,7 +71,6 @@
 	}
 	
 	if (status & GNUTLS_CERT_INVALID
-	    || status & GNUTLS_CERT_NOT_TRUSTED
 	    || status & GNUTLS_CERT_REVOKED) {
 		if (ssl->func (ssl, LM_SSL_STATUS_UNTRUSTED_CERT,
 			       ssl->func_data) != LM_SSL_RESPONSE_CONTINUE) {
@@ -126,7 +125,7 @@
 
 		gnutls_x509_crt_deinit (cert);
 
-		if (gnutls_x509_fingerprint (GNUTLS_DIG_MD5, &cert_list[0],
+		if (gnutls_fingerprint (GNUTLS_DIG_MD5, &cert_list[0],
 					     ssl->fingerprint,
 					     &digest_size) >= 0) {
 			if (ssl->expected_fingerprint &&
