--- ./vncviewer/Makefile	Mon Jan 20 16:53:18 2003
+++ ./vncviewer/Makefile	Mon Jan 20 17:04:35 2003
@@ -970,8 +970,6 @@
 	fi
 	$(INSTALL) $(INSTALLFLAGS) $(INSTPGMFLAGS)  vncviewer $(DESTDIR)$(BINDIR)/vncviewer
 
-all:: vncviewer.$(MANSUFFIX).html
-
 vncviewer.$(MANSUFFIX).html: vncviewer.$(MANNEWSUFFIX)
 	$(RM) vncviewer.$(MANSUFFIX).html vncviewer.$(MANSUFFIX)-html
 	$(RMAN) $(RMANOPTIONS) < vncviewer.$(MANNEWSUFFIX) \
@@ -993,13 +991,12 @@
 cleandir::
 	$(RM) vncviewer.$(MANSUFFIX).html vncviewer.$(MANSUFFIX)-html
 
-all:: vncviewer.$(MANNEWSUFFIX)
-
 vncviewer.$(MANNEWSUFFIX): vncviewer.$(MANSRCSUFFIX)
 	$(RM) $@
 	cd `dirname vncviewer` && \
 	$(LN) `basename vncviewer.$(MANSRCSUFFIX)` `basename $@`
-
+	touch vncviewer.$(MANSRCSUFFIX)
+	
 cleandir::
 	$(RM) vncviewer.$(MANNEWSUFFIX)
 
