--- cinelerra-2.0_orig/build/Makefile.cinelerra	2005-09-11 21:21:42.000000000 +0200
+++ cinelerra-2.0/build/Makefile.cinelerra	2005-09-20 11:36:44.593712160 +0200
@@ -5,9 +5,7 @@
 
 DIRS := \
 	doc \
-	libmpeg3 \
 	mpeg2enc \
-	mplexlo \
 	guicast \
 	cinelerra \
 	plugins \
@@ -23,25 +21,10 @@
 
 
 all: $(OBJDIR)/soundtest
-ifeq ($(HAVE_FIREWIRE), y)
-	$(MAKE) -C libraw1394*
 	$(MAKE) -C libiec61883*
-	$(MAKE) -C libavc1394*
-endif
-	$(MAKE) -C fftw*
-	$(MAKE) -f build/Makefile.esound
-	$(MAKE) -f build/Makefile.freetype
-	$(MAKE) -f build/Makefile.tiff
 	$(MAKE) -f build/Makefile.toolame
 	$(MAKE) -f build/Makefile.uuid
-	MAKE='make -j1' make -j1 -C mjpegtools*
-	$(MAKE) -C libsndfile*
 	$(MAKE) -C quicktime
-	$(MAKE) -C libtheora*
-ifeq ($(HAVE_ALSA), y)
-	$(MAKE) -f build/Makefile.asound
-endif
-	$(MAKE) -f build/Makefile.exr
 	@ for i in $(DIRS) ; \
 	do \
 		 $(MAKE) -C $$i ; \
@@ -78,8 +61,6 @@
 install:
 	$(MAKE) -C cinelerra install
 	$(MAKE) -C plugins install
-	$(MAKE) -C libmpeg3 install
-	$(MAKE) -C mplexlo install
 	$(MAKE) -C po install
 	cp $(OBJDIR)/soundtest /usr/bin
 
diff -ur cinelerra-2.0_orig/cinelerra/Makefile cinelerra-2.0/cinelerra/Makefile
--- cinelerra-2.0_orig/cinelerra/Makefile	2005-09-11 21:21:39.000000000 +0200
+++ cinelerra-2.0/cinelerra/Makefile	2005-09-19 21:26:01.993366880 +0200
@@ -17,11 +17,11 @@
 UUID_DIR := ../uuid
 VORBIS_DIR := $(shell expr ../quicktime/libvorbis* )
 
-AVC_LIB := $(AVC_DIR)/libavc1394/.libs/libavc1394.a $(AVC_DIR)/librom1394/.libs/librom1394.a
-RAW_LIB := $(LIBRAW_DIR)/src/.libs/libraw1394.a
+AVC_LIB := -lavc1394 -lrom1394
+RAW_LIB := -lraw1394
 IEC_LIB := $(LIBIEC_DIR)/src/.libs/libiec61883.a
-SNDFILE_LIB := $(SNDFILE_DIR)/src/.libs/libsndfile.a
-THEORA_LIB := $(THEORA_DIR)/lib/.libs/libtheora.a
+SNDFILE_LIB := -lsndfile
+THEORA_LIB := -ltheora
 
 OBJS = \
 	$(OBJDIR)/aattachmentpoint.o \
@@ -306,7 +306,7 @@
 
 ifeq ($(HAVE_ALSA), y)
 ALSA_CFLAGS := -DHAVE_ALSA -I$(ALSA_DIR)/include
-ALSA_LIBS := $(ALSA_DIR)/$(OBJDIR)/libasound.a
+ALSA_LIBS := -lasound
 endif
 
 
@@ -317,11 +317,11 @@
 	$(RAW_LIB) \
 	$(IEC_LIB) \
 	../quicktime/$(OBJDIR)/libquicktime.a \
-        $(ESOUNDLIBS) \
-	$(EXR_DIR)/$(OBJDIR)/libexr.a \
-	$(TIFF_DIR)/$(OBJDIR)/libtiff.a \
+       -lesd \
+	-lHalf -lIex -lImath -lIlmImf \
+	-ltiff \
 	../guicast/$(OBJDIR)/libguicast.a \
-	../libmpeg3/$(OBJDIR)/libmpeg3.a \
+	/usr/lib/libmpeg3.a \
 	$(SNDFILE_LIB) \
 	$(UUID_DIR)/$(OBJDIR)/libuuid.a \
 	../mpeg2enc/$(OBJDIR)/mpeg2enc.a \
@@ -438,7 +438,7 @@
 
 $(shell echo "\#define BUILDDATE \"`date`\"" > builddate.h )
 $(shell echo $(CFLAGS) > $(OBJDIR)/c_flags)
-$(shell echo $(LDFLAGS1) $(OBJS) $(DCRAW) $(FILEEXR) $(AVIOBJS) $(CXXREPOSITORY) $(STATICLIBS) $(LIBS) $(LDFLAGS2) > $(OBJDIR)/objs)
+$(shell echo $(LDFLAGS1) $(OBJS) $(DCRAW) $(FILEEXR) $(AVIOBJS) $(CXXREPOSITORY) $(STATICLIBS) $(LIBS) $(LDFLAGS2) -lvorbis -lvorbisfile -lvorbisenc -logg -ltiff -lmp3lame -ldv -lavcodec > $(OBJDIR)/objs)
 
 
 
diff -ur cinelerra-2.0_orig/configure cinelerra-2.0/configure
--- cinelerra-2.0_orig/configure	2005-09-11 21:22:33.000000000 +0200
+++ cinelerra-2.0/configure	2005-09-19 21:26:01.994366728 +0200
@@ -87,46 +87,12 @@
 echo CONFIGURING QUICKTIME
 cd quicktime* && ./configure && cd ..
 
-echo CONFIGURING LIBMPEG3
-cd libmpeg3* && ./configure && cd ..
-
-echo CONFIGURING FFTW
-cd fftw* && CFLAGS=-fPIC ./configure && cd ..
-
-echo CONFIGURING MJPEGTOOLS
-ALSA_PATH=`expr $TOPDIR/alsa-lib*` && \
-cd mjpegtools* && CFLAGS="-I$ALSA_PATH/include -L$ALSA_PATH/$OBJDIR" ./configure --enable-shared=no && cd ..
-
-echo CONFIGURING SNDFILE
-cd libsndfile* && chmod a+x configure && ./configure && cd ..
-
-echo CONFIGURING RAW1394
-cd libraw1394* && \
-./configure --enable-shared=no && \
-ln -sf src libraw1394 && \
-cd ..
-
-echo CONFIGURING AVC1394
-cd libavc1394* && \
-RAW1394_PATH=`expr $TOPDIR/libraw1394*` && \
-PKG_CONFIG_PATH=$RAW1394_PATH CFLAGS=-I$RAW1394_PATH/ LDFLAGS=-L$RAW1394_PATH/src/.libs ./configure --enable-shared=no && \
-cd ..
-
 echo CONFIGURING IEC61883
 cd libiec61883* && \
 RAW1394_PATH=`expr $TOPDIR/libraw1394*` && \
 PKG_CONFIG_PATH=$RAW1394_PATH CFLAGS=-I$RAW1394_PATH/ LDFLAGS=-L$RAW1394_PATH/src/.libs ./configure --enable-shared=no && \
 cd ..
 
-echo CONFIGURING THEORA
-ALSA_PATH=`expr $TOPDIR/alsa-lib*` && \
-LIBOGG_PATH=`expr $TOPDIR/quicktime/libogg*` && \
-LIBVORBIS_PATH=`expr $TOPDIR/quicktime/libvorbis*` && \
-cd libtheora* && \
-PKG_CONFIG_PATH=$LIBOGG_PATH:$LIBVORBIS_PATH CFLAGS="-I$LIBOGG_PATH/include -I$ALSA_PATH -L$ALSA_PATH/$OBJDIR -L$LIBOGG_PATH/src -I$LIBVORBIS_PATH/include -L$LIBVORBIS_PATH/lib" ./configure --enable-shared=no && \
-cd ..
-
-
 
 # write configuration header
 echo "Writing hvirtual_config.h"
diff -ur cinelerra-2.0_orig/quicktime/Makefile cinelerra-2.0/quicktime/Makefile
--- cinelerra-2.0_orig/quicktime/Makefile	2005-08-17 05:56:05.000000000 +0200
+++ cinelerra-2.0/quicktime/Makefile	2005-09-19 21:26:02.011364144 +0200
@@ -91,11 +91,11 @@
 
 #include Makefile.jpeg
 #include Makefile.vorbis
-include Makefile.lame
+#include Makefile.lame
 #include Makefile.dv
 include Makefile.encore50
 
-
+LAME_DIR = /usr
 
 
 
@@ -213,7 +213,7 @@
 	-I$(DV_INCLUDE) \
 	-I$(JPEG_INCLUDE) \
 	-I$(MPEG3_DIR) \
-	-I$(LAME_DIR)/include \
+	-I$(LAME_DIR)/include/lame \
 	-I$(OGG_DIR)/include \
 	-I$(VORBIS_DIR)/include \
 	-ffast-math \
@@ -243,11 +243,11 @@
 $(shell if ! test -d $(OBJDIR) \; then mkdir -p $(DIRS) \; fi )
 $(shell echo  $(CFLAGS) > $(OBJDIR)/c_flags)
 $(shell echo  $(LAME_CFLAGS) > $(OBJDIR)/lame_flags)
-$(shell echo  $(OBJS) $(JPEG_LIB)/*.o $(DV_LIB)/*.o $(FAAD_LIB)/*.o $(FAAC_LIB)/*.o $(FFMPEG_LIB)/*.o $(X264_LIB)/*.o $(OGG_LIB)/*.o $(VORBIS_LIB)/*.o $(LAME_OBJS) > $(OBJDIR)/objs)
+$(shell echo  $(OBJS) $(FAAD_LIB)/*.o $(FAAC_LIB)/*.o $(X264_LIB)/*.o > $(OBJDIR)/objs)
 
 all: $(OBJDIR) $(OUTPUT)
 
-$(OUTPUT): $(OBJS) $(ENCOREMMX_OBJS) $(JPEG_NASM) $(LAME_OBJS) $(FAAD_LIB) $(FAAC_LIB) $(FFMPEG_LIB) $(X264_LIB) $(DV_LIB) $(JPEG_LIB) $(OGG_LIB)/ $(VORBIS_LIB)
+$(OUTPUT): $(OBJS) $(ENCOREMMX_OBJS) $(JPEG_NASM) $(FAAD_LIB) $(FAAC_LIB) $(X264_LIB)
 	ar rcs $(OUTPUT) `cat $(OBJDIR)/objs`
 
 # this is the only way we figured out to merge .a files
diff -ur cinelerra-2.0_orig/quicktime/configure cinelerra-2.0/quicktime/configure
--- cinelerra-2.0_orig/quicktime/configure	2005-09-11 21:21:34.000000000 +0200
+++ cinelerra-2.0/quicktime/configure	2005-09-20 09:44:24.853308440 +0200
@@ -58,20 +58,8 @@
 
 cd faac* && ./configure && cd ..
 
-cd ffmpeg* && ./configure && cd ..
-
 cd x264* && ./configure --enable-pthread && cd ..
 
-cd libdv* && ./configure --disable-gtk --enable-shared=no && cd ..
-
-cd libogg* && ./configure && cd ..
-
-LIBOGG_PATH=`expr libogg*` && \
-cd libvorbis* && \
-CFLAGS="-I../../$LIBOGG_PATH/include -L../../$LIBOGG_PATH/src/" ./configure --enable-shared=no&& \
-cd ..
-
-
 
 if [ `arch` == i686 ];
 then
--- cinelerra-2.0_orig/plugins/titler/Makefile	2005-06-16 04:39:35.000000000 +0200
+++ cinelerra-2.0/plugins/titler/Makefile	2005-09-20 10:00:09.679673000 +0200
@@ -15,7 +15,7 @@
 $(OUTPUT_PLUGIN): $(OBJS)
 	$(LDLINKER) -o $(OUTPUT_PLUGIN) \
 		$(OBJS) \
-		$(FREETYPE_DIR)/$(OBJDIR)/libfreetype.a
+		-lfreetype
 	strip $(OUTPUT_PLUGIN)
 
 include ../../plugin_config
