--- ./configure.in.orig	2003-01-26 19:28:40.000000000 +0100
+++ ./configure.in	2003-01-26 19:28:49.000000000 +0100
@@ -24,7 +24,7 @@
 AC_SUBST(MYSQLCC_HEADERS)
 AC_SUBST(MYSQLCC_SOURCES)
 AC_OUTPUT(mysqlcc.pro)
-QMAKESPEC=$QTDIR/mkspecs/$host_os
+QMAKESPEC=$QTDIR/mkspecs/linux-g++
 echo "Creating Makefile with qmake"
 QMAKESPEC=$QMAKESPEC QTDIR=$QTDIR $QTDIR/bin/qmake mysqlcc.pro
 if test ! $? = 0 ; then
