--- ./src/bdf2h.c.orig	2000-06-01 17:44:48.000000000 +0300
+++ ./src/bdf2h.c	2003-10-19 09:26:48.000000000 +0300
@@ -45,14 +45,14 @@
  */
 printf("#ifndef DEFINED_STRUCT_FONTINFO_TAG\n");
 printf("#define DEFINED_STRUCT_FONTINFO_TAG 1\n");
-printf("\
-struct fontinfo_tag
-  {
-  signed char *data;
-  int table[96];
-  int yofs,fh,oy;
-  };
-");
+printf(
+"struct fontinfo_tag\n"
+"  {\n"
+"  signed char *data;\n"
+"  int table[96];\n"
+"  int yofs,fh,oy;\n"
+"  };\n"
+);
 printf("#endif\n\n");
 
 /* generate font data */
