--- NVIDIA-Linux-x86-1.0-6629-pkg1/usr/src/nv/Makefile.kbuild.orig	2004-11-03 22:53:00.000000000 +0100
+++ NVIDIA-Linux-x86-1.0-6629-pkg1/usr/src/nv/Makefile.kbuild	2005-01-05 23:09:02.238876496 +0100
@@ -286,7 +286,7 @@
 module-install: suser-sanity-check module
 	@mkdir -p $(MODULE_ROOT)/video; \
 	install -m 0664 -o root -g root $(MODULE_OBJECT) $(MODULE_ROOT)/video; \
-	PATH="$(PATH):/bin:/sbin" depmod -ae;
+	PATH="$(PATH):/bin:/sbin" depmod -ae $(KERNEL_UNAME);
 
 #
 # This target builds, then installs, then creates device nodes and inserts
