--- NVIDIA-Linux-x86_64-100.14.19-pkg2/usr/src/nv/Makefile.nvidia.orig	2007-09-12 14:53:20.000000000 -0700
+++ NVIDIA-Linux-x86_64-100.14.19-pkg2/usr/src/nv/Makefile.nvidia	2007-10-18 15:51:53.000000000 -0700
@@ -130,9 +130,7 @@
 	mkdir -p $(MODULE_ROOT)/video && \
 	install -m 0664 -o root -g root $(MODULE_NAME).o \
 	 $(MODULE_ROOT)/video/$(MODULE_NAME).o && \
-	PATH="$(PATH):/bin:/sbin" depmod -a && \
-		PATH="$(PATH):/bin:/sbin" modprobe $(MODULE_NAME) && \
-	echo "$(MODULE_NAME).o installed successfully."; \
+	echo "$(MODULE_NAME).o installed successfully."
 
 $(VERSION_HDR): 
 	echo \#define NV_COMPILER \"`$(CC) -v 2>&1 | tail -n 1`\" > $@
