--- NVIDIA-Linux-x86_64-100.14.19-pkg2/usr/src/nv/Makefile.kbuild.orig	2007-10-18 10:42:17.000000000 -0700
+++ NVIDIA-Linux-x86_64-100.14.19-pkg2/usr/src/nv/Makefile.kbuild	2007-10-18 10:43:53.000000000 -0700
@@ -334,7 +334,8 @@
 module-install: suser-sanity-check module
 	@mkdir -p $(MODULE_ROOT)/video; \
 	install -m 0664 -o root -g root $(MODULE_OBJECT) $(MODULE_ROOT)/video; \
-	PATH="$(PATH):/bin:/sbin" depmod -ae;
+  echo "$(MODULE_OBJECT) installed successfully.";
+#	PATH="$(PATH):/bin:/sbin" depmod -ae;
 
 #
 # This target builds, then installs, then creates device nodes and inserts
@@ -342,7 +343,7 @@
 #
 
 package-install: module-install rmmod-sanity-check
-	PATH="$(PATH):/bin:/sbin" modprobe $(MODULE_NAME) && \
+#	PATH="$(PATH):/bin:/sbin" modprobe $(MODULE_NAME) && 
 	echo "$(MODULE_OBJECT) installed successfully.";
 
 #
