--- NVIDIA-Linux-x86-1.0-6629-pkg1/usr/src/nv/Makefile.nvidia~	2005-01-06 00:29:31.718683968 +0100
+++ NVIDIA-Linux-x86-1.0-6629-pkg1/usr/src/nv/Makefile.nvidia	2005-01-06 00:30:57.207687664 +0100
@@ -109,6 +109,11 @@
 module: gcc-sanity-check rivafb-support-sanity-check $(LINUX_MODULE) $(RESMAN_KERNEL_MODULE) 
 	$(LD) -r -o nvidia.o $(LINUX_MODULE) $(RESMAN_KERNEL_MODULE)
 
+module-install: suser-sanity-check module
+	mkdir -p $(MODULE_ROOT)/video; \
+	install -m 0664 -o root -g root nvidia.o $(MODULE_ROOT)/video; \
+	PATH="$(PATH):/bin:/sbin" depmod -ae $(KERNEL_UNAME)
+
 package-install: module suser-sanity-check rmmod-sanity-check 
 	mkdir -p $(MODULE_ROOT)/video && \
 	install -m 0664 -o root -g root nvidia.o $(MODULE_ROOT)/video/nvidia.o$(O) && \
