diff -ru usr/src/nv/nv.c usr/src/nv.1165235/nv.c
--- ./usr/src/nv/nv.c	2004-11-03 22:53:00.000000000 +0100
+++ ./usr/src/nv.1165235/nv.c	2004-11-25 16:45:04.000000000 +0100
@@ -1604,9 +1604,8 @@
         }
         nv_vm_list_page_count(at->page_table, at->num_pages);
 
-        /* prevent the swapper from swapping it out */
-        /* mark the memory i/o so the buffers aren't dumped on core dumps */
-        vma->vm_flags |= (VM_LOCKED | VM_IO);
+        // mark it as IO so that we don't dump it on core dump
+        vma->vm_flags |= VM_IO;
     }
 
     /* Magic allocator */
