--- WebKit-r70732/GNUmakefile.am.orig	2012-01-11 10:09:28.000000000 -0330
+++ WebKit-r70732/GNUmakefile.am	2012-01-11 10:10:11.000000000 -0330
@@ -91,7 +91,7 @@
 	-Wformat -Wformat-security -Wno-format-y2k -Wundef \
 	-Wmissing-format-attribute -Wpointer-arith -Wwrite-strings \
 	-Wno-unused-parameter -Wno-parentheses \
-	-fno-exceptions -DENABLE_GLIB_SUPPORT=1
+	-fno-exceptions -DENABLE_GLIB_SUPPORT=1 -I/opt/gnome2/include/libsoup-2.4
 
 
 global_cxxflags += \
