--- WebKit-r70732/WebCore/make-hash-tools.pl.orig	2012-01-11 09:26:45.000000000 -0330
+++ WebKit-r70732/WebCore/make-hash-tools.pl	2012-01-11 09:25:08.000000000 -0330
@@ -20,7 +20,8 @@
 #   Boston, MA 02110-1301, USA.
 
 use strict;
-use Switch;
+#use Switch;
+use feature "switch";
 use File::Basename;
 
 my $outdir = $ARGV[0];
@@ -82,9 +83,10 @@
 
 }
 
-switch ($option) {
+#switch ($option) {
+given ($option) {
 
-case "DocTypeStrings" {
+when ("DocTypeStrings") {
 
     createHashToolsHeader();
 
@@ -117,7 +119,7 @@
 
 } # case "DocTypeStrings"
 
-case "ColorData" {
+when ("ColorData") {
 
     my $colorDataImpl              = "$outdir/ColorData.cpp";
     my $colorDataGenerated         = "$outdir/ColorDataHash.h";
--- WebKit-r70732/WebCore/dom/make_names.pl.orig	2010-10-06 09:00:45.000000000 -0230
+++ WebKit-r70732/WebCore/dom/make_names.pl	2012-01-11 09:38:21.000000000 -0330
@@ -35,7 +35,7 @@
 use File::Path;
 use IO::File;
 use InFilesParser;
-use Switch;
+use feature "switch";
 
 sub readTags($$);
 sub readAttrs($$);
