--- root/utils/src/rootcint.cxx~	2005-06-14 01:26:45.000000000 +0200
+++ root/utils/src/rootcint.cxx	2005-09-02 21:09:45.000000000 +0200
@@ -2306,7 +2306,10 @@
    fprintf(fp, "   namespace ROOT {\n");
 
 #if !defined(R__SGI)
-   fprintf(fp, "      inline ::ROOT::TGenericClassInfo *GenerateInitInstance();\n");
+   if (nesting)
+     fprintf(fp, "      ::ROOT::TGenericClassInfo *GenerateInitInstance();\n");
+   else
+     fprintf(fp, "      inline ::ROOT::TGenericClassInfo *GenerateInitInstance();\n");
 #endif
 
    if (!cl.HasMethod("Dictionary") || cl.IsTmplt())
@@ -2317,7 +2320,10 @@
    fprintf(fp, "      // Function generating the singleton type initializer\n");
 
 #if !defined(R__SGI)
-   fprintf(fp, "      inline ::ROOT::TGenericClassInfo *GenerateInitInstance()\n      {\n");
+   if (nesting)
+     fprintf(fp, "      ::ROOT::TGenericClassInfo *GenerateInitInstance()\n      {\n");
+   else
+     fprintf(fp, "      inline ::ROOT::TGenericClassInfo *GenerateInitInstance()\n      {\n");
 #else
    fprintf(fp, "      ::ROOT::TGenericClassInfo *GenerateInitInstance()\n      {\n");
 #endif
