--- root/base/inc/Rtypes.h~	2005-04-07 16:43:35.000000000 +0200
+++ root/base/inc/Rtypes.h	2005-09-02 17:10:57.000000000 +0200
@@ -307,7 +307,7 @@
 #define NamespaceImpUnique(name,key) \
    namespace name { \
       namespace ROOT { \
-         ::ROOT::TGenericClassInfo *GenerateInitInstance(); \
+         using ::ROOT::TGenericClassInfo *GenerateInitInstance(); \
          static int _R__UNIQUE_(_NAME2_(R__dummyint,key)) = \
             GenerateInitInstance()->SetImplFile(__FILE__, __LINE__); \
          R__UseDummy(_R__UNIQUE_(_NAME2_(R__dummyint,key))); \
