--- rpm-4.1/lib/rpminstall.c.vanilla	2004-06-28 09:43:21.867093080 +0200
+++ rpm-4.1/lib/rpminstall.c	2004-06-28 09:43:46.076412704 +0200
@@ -724,7 +724,7 @@
 /*@=bounds@*/
 
 int rpmErase(rpmts ts,
-		const struct rpmInstallArguments_s * ia,
+		struct rpmInstallArguments_s * ia,
 		const char ** argv)
 {
     int count;
--- rpm-4.1/lib/rpmcli.h.vanilla	2004-06-28 09:44:37.724560992 +0200
+++ rpm-4.1/lib/rpmcli.h	2004-06-28 09:44:56.856652472 +0200
@@ -557,7 +557,7 @@
  * @param argv		array of package file names (NULL terminated)
  * @return		0 on success
  */
-int rpmErase(rpmts ts, const struct rpmInstallArguments_s * ia,
+int rpmErase(rpmts ts, struct rpmInstallArguments_s * ia,
 		/*@null@*/ const char ** argv)
 	/*@globals rpmGlobalMacroContext, fileSystem, internalState @*/
 	/*@modifies ts, ia, rpmGlobalMacroContext,
