--- rpm-4.1/lib/rpminstall.c	2002-09-16 19:06:08.000000000 +0000
+++ rpm-4.1-patch/lib/rpminstall.c	2004-06-19 19:44:52.000000000 +0000
@@ -747,9 +747,9 @@
     ovsflags = rpmtsSetVSFlags(ts, vsflags);
 
     if (rpmExpandNumeric("%{?_repackage_all_erasures}"))
-	ia->transFlags |= RPMTRANS_FLAG_REPACKAGE;
-
-    (void) rpmtsSetFlags(ts, ia->transFlags);
+	(void) rpmtsSetFlags(ts, ia->transFlags|RPMTRANS_FLAG_REPACKAGE);
+    else
+	(void) rpmtsSetFlags(ts, ia->transFlags);
 
 #ifdef	NOTYET	/* XXX no callbacks on erase yet */
     {	int notifyFlags;
