--- ./src/art_sort.c	Mon Mar 11 15:22:35 2002
+++ ./src/art_sort.c	Wed Mar 13 21:32:41 2002
@@ -1011,14 +1011,22 @@
    
    while (-1 != SLextract_list_element (sort_order, nth, ',', buf, sizeof(buf)))
      {
-	if (! strcasecmp(buf, "Subject")) add_sort_function(header_subject_cmp, isupper(buf[0]));
-	else if (! strcasecmp(buf, "Score")) add_sort_function(header_score_cmp, isupper(buf[0]));
-	else if (! strcasecmp(buf, "Highscore")) add_sort_function(header_highscore_cmp, isupper(buf[0]));
-	else if (! strcasecmp(buf, "Date")) add_sort_function(header_date_cmp, isupper(buf[0]));
-	else if (! strcasecmp(buf, "Author")) add_sort_function(header_author_cmp, isupper(buf[0]));
-	else if (! strcasecmp(buf, "Lines")) add_sort_function(header_lines_cmp, isupper(buf[0]));
-	else if (! strcasecmp(buf, "Number")) add_sort_function(header_num_cmp, isupper(buf[0]));
-	else if (! strcasecmp(buf, "Id")) add_sort_function(header_msgid_cmp, isupper(buf[0]));
+	if (! slrn_case_strcmp((unsigned char*)buf, (unsigned char*)"Subject"))
+	  add_sort_function(header_subject_cmp, isupper(buf[0]));
+	else if (! slrn_case_strcmp((unsigned char*)buf, (unsigned char*)"Score"))
+	  add_sort_function(header_score_cmp, isupper(buf[0]));
+	else if (! slrn_case_strcmp((unsigned char*)buf, (unsigned char*)"Highscore"))
+	  add_sort_function(header_highscore_cmp, isupper(buf[0]));
+	else if (! slrn_case_strcmp((unsigned char*)buf, (unsigned char*)"Date"))
+	  add_sort_function(header_date_cmp, isupper(buf[0]));
+	else if (! slrn_case_strcmp((unsigned char*)buf, (unsigned char*)"Author"))
+	  add_sort_function(header_author_cmp, isupper(buf[0]));
+	else if (! slrn_case_strcmp((unsigned char*)buf, (unsigned char*)"Lines"))
+	  add_sort_function(header_lines_cmp, isupper(buf[0]));
+	else if (! slrn_case_strcmp((unsigned char*)buf, (unsigned char*)"Number"))
+	  add_sort_function(header_num_cmp, isupper(buf[0]));
+	else if (! slrn_case_strcmp((unsigned char*)buf, (unsigned char*)"Id"))
+	  add_sort_function(header_msgid_cmp, isupper(buf[0]));
 	else /* Nonexistant sorting method */
 	  {
 	     slrn_error(_("Can't sort according to `%s'"), buf);
