From: Peter Robinson <pbrobinson@gmail.com>
Date: Apr 24 2024 14:24:44 +0000
Subject: Fix location of gconv conflicts

--- v4l-utils-1.30.1/contrib/gconv/meson.build.vanilla	2025-05-28 12:12:46.636526543 +0200
+++ v4l-utils-1.30.1/contrib/gconv/meson.build	2025-05-28 12:13:04.197454648 +0200
@@ -41,4 +41,4 @@
 
 dep_en300_468_tab00 = declare_dependency(link_with : en300_468_tab00)
 
-install_data('gconv-modules', install_dir : gconv_install_dir)
+install_data('gconv-modules', install_dir : gconv_install_dir / 'gconv-modules.d', rename : ['libdvbv5.conf'])
