--- linux-4.15/drivers/hwmon/k10temp.c.vanilla	2018-03-04 22:35:59.983889704 +0100
+++ linux-4.15/drivers/hwmon/k10temp.c	2018-03-04 22:53:09.795853424 +0100
@@ -89,6 +89,7 @@
 	{ 0x17, "AMD Ryzen Threadripper 1950", 10000 },
 	{ 0x17, "AMD Ryzen Threadripper 1920", 10000 },
 	{ 0x17, "AMD Ryzen Threadripper 1910", 10000 },
+	{ 0x17, "AMD Ryzen 5 2500U with Radeon Vega Mobile Gfx", 20000 },
 };
 
 static void read_tempreg_pci(struct pci_dev *pdev, u32 *regval)
@@ -288,6 +289,7 @@
 	{ PCI_VDEVICE(AMD, PCI_DEVICE_ID_AMD_16H_NB_F3) },
 	{ PCI_VDEVICE(AMD, PCI_DEVICE_ID_AMD_16H_M30H_NB_F3) },
 	{ PCI_VDEVICE(AMD, PCI_DEVICE_ID_AMD_17H_DF_F3) },
+	{ PCI_VDEVICE(AMD, 0x15d0) },
 	{}
 };
 MODULE_DEVICE_TABLE(pci, k10temp_id_table);
