--- ./themes/default.keys/keythemerc.orig	2003-02-27 23:02:48.000000000 +0200
+++ ./themes/default.keys/keythemerc	2003-08-30 17:49:47.000000000 +0300
@@ -29,15 +29,15 @@
 workspace_9_key=Control+F9
 move_window_next_workspace_key=Alt+Control+End
 move_window_prev_workspace_key=Alt+Control+Home
-move_window_workspace_1_key=Alt+Control+KP_1
-move_window_workspace_2_key=Alt+Control+KP_2
-move_window_workspace_3_key=Alt+Control+KP_3
-move_window_workspace_4_key=Alt+Control+KP_4
-move_window_workspace_5_key=Alt+Control+KP_5
-move_window_workspace_6_key=Alt+Control+KP_6
-move_window_workspace_7_key=Alt+Control+KP_7
-move_window_workspace_8_key=Alt+Control+KP_8
-move_window_workspace_9_key=Alt+Control+KP_9
+move_window_workspace_1_key=Control+Shift+F1
+move_window_workspace_2_key=Control+Shift+F2
+move_window_workspace_3_key=Control+Shift+F3
+move_window_workspace_4_key=Control+Shift+F4
+move_window_workspace_5_key=Control+Shift+F5
+move_window_workspace_6_key=Control+Shift+F6
+move_window_workspace_7_key=Control+Shift+F7
+move_window_workspace_8_key=Control+Shift+F8
+move_window_workspace_9_key=Control+Shift+F9
 shortcut_1_key=Alt+F1
 shortcut_1_exec=xfhelp4
 shortcut_2_key=Alt+F2
