diff -Nur nedit-5.4-old/makefiles/Makefile.linux nedit-5.4/makefiles/Makefile.linux
--- nedit-5.4-old/makefiles/Makefile.linux	2003-11-22 11:42:32.968085128 +0100
+++ nedit-5.4/makefiles/Makefile.linux	2003-11-22 11:45:14.896468272 +0100
@@ -17,11 +17,11 @@
 # To test if the Motif library exports the runtime version
 # add -DHAVE__XMVERSIONSTRING to CFLAGS
 #
-CFLAGS=-O -I/usr/X11R6/include -DUSE_DIRENT -DUSE_LPR_PRINT_CMD
+CFLAGS=-O -I/usr/X11R6/include -D__LINUX__ -DUSE_DIRENT -DUSE_LPR_PRINT_CMD
 
 ARFLAGS=-urs
 
-LIBS= -L/usr/X11R6/lib -Wl,-Bstatic -lXm -Wl,-Bdynamic -lXp -lXpm -lXext -lXt -lSM -lICE -lX11 -lm
+LIBS= -L/usr/X11R6/lib -lXm -lXp -lXext -lXt -lSM -lICE -lX11 -lm -lXmu
 
 include Makefile.common
 
diff -Nur nedit-5.4-old/source/preferences.c nedit-5.4/source/preferences.c
--- nedit-5.4-old/source/preferences.c	2003-11-22 11:42:33.023076768 +0100
+++ nedit-5.4/source/preferences.c	2003-11-22 11:47:31.771660080 +0100
@@ -318,7 +318,7 @@
 #ifndef VMS
 #ifdef linux
     {"shellCommands", "ShellCommands", PREF_ALLOC_STRING, "spell:Alt+B:s:EX:\n\
-    cat>spellTmp; xterm -e ispell -x spellTmp; cat spellTmp; rm spellTmp\n\
+    cat>spellTmp; xterm -e aspell -c -x spellTmp; cat spellTmp; rm spellTmp\n\
     wc::w:ED:\nwc | awk '{print $1 \" lines, \" $2 \" words, \" $3 \" characters\"}'\n\
     sort::o:EX:\nsort\nnumber lines::n:AW:\nnl -ba\nmake:Alt+Z:m:W:\nmake\n\
     expand::p:EX:\nexpand\nunexpand::u:EX:\nunexpand\n",
