
The usual gcc-3.4 fixes.

  Rene Rebe <rene@exactcode.de>

--- cdparanoia-III-alpha9.8-orig/interface/utils.h	2000-04-20 00:41:04.000000000 +0200
+++ cdparanoia-III-alpha9.8/interface/utils.h	2004-05-03 21:29:26.000000000 +0200
@@ -112,6 +112,7 @@
       break;
     case CDDA_MESSAGE_FORGETIT:
     default:
+      ((void)0);
     }
   }
 }
@@ -127,6 +128,7 @@
       break;
     case CDDA_MESSAGE_FORGETIT:
     default:
+      ((void)0);
     }
   }
 }
@@ -169,6 +171,7 @@
       break;
     case CDDA_MESSAGE_FORGETIT:
     default:
+      ((void)0);
     }
   }
   if(malloced)free(buffer);
@@ -205,6 +208,7 @@
       break;
     case CDDA_MESSAGE_FORGETIT:
     default:
+      ((void)0);
     }
   }
   if(malloced)free(buffer);
