
The usual gcc-3.4 trivialities.

  - Rene Rebe <rene@rocklinux.org>


--- libgda-1.0.1/libgda/gda-util.c.orig	2004-05-10 14:21:46.925295232 +0200
+++ libgda-1.0.1/libgda/gda-util.c	2004-05-10 14:22:59.587248936 +0200
@@ -62,10 +62,8 @@
 	case GDA_VALUE_TYPE_TIME : return "time";
 	case GDA_VALUE_TYPE_TIMESTAMP : return "timestamp";
 	case GDA_VALUE_TYPE_TINYINT : return "tinyint";
-	default:
+	default: return "string";
 	}
-
-	return "string";
 }
 
 /**
--- libgda-1.0.1/providers/xml/gda-xml-provider.c.orig	2004-05-10 14:31:07.936008640 +0200
+++ libgda-1.0.1/providers/xml/gda-xml-provider.c	2004-05-10 14:31:22.640773176 +0200
@@ -682,7 +682,6 @@
 	case GDA_CONNECTION_SCHEMA_TYPES :
 		return get_types (cnc);
 	default :
+		return NULL;
 	}
-
-	return NULL;
 }
--- libgda-1.0.1/providers/ldap/gda-ldap-provider.c.orig	2004-05-10 14:29:32.499517192 +0200
+++ libgda-1.0.1/providers/ldap/gda-ldap-provider.c	2004-05-10 14:30:36.051855768 +0200
@@ -276,9 +276,8 @@
 	case GDA_CONNECTION_FEATURE_SQL :
 		return TRUE;
 	default :
+		return FALSE;
 	}
-
-	return FALSE;
 }
 
 static void
@@ -467,9 +466,8 @@
 	case GDA_CONNECTION_SCHEMA_TYPES :
 		return get_ldap_types (cnc, params);
 	default :
+		return NULL;
 	}
-
-	return NULL;
 }
 
 /* begin_transaction handler for the GdaLdapProvider class */
--- libgda-1.0.1/providers/mysql/gda-mysql-provider.c.orig	2004-05-10 14:25:03.484413704 +0200
+++ libgda-1.0.1/providers/mysql/gda-mysql-provider.c	2004-05-10 14:26:35.245463904 +0200
@@ -510,6 +510,7 @@
 		g_free (str);
 		break;
 	default:
+		;
 	}
 
 	return reclist;
@@ -655,6 +656,7 @@
 	case GDA_CONNECTION_FEATURE_TRANSACTIONS :
 		return TRUE;
 	default :
+		;
 	}
 
 	return FALSE;
@@ -1187,7 +1189,6 @@
 	case GDA_CONNECTION_SCHEMA_TYPES :
 		return get_mysql_types (cnc, params);
 	default :
+		return NULL;
 	}
-
-	return NULL;
 }
--- libgda-1.0.1/providers/sqlite/gda-sqlite-provider.c.orig	2004-05-10 14:27:57.509957800 +0200
+++ libgda-1.0.1/providers/sqlite/gda-sqlite-provider.c	2004-05-10 14:28:56.085053032 +0200
@@ -589,9 +589,8 @@
 	case GDA_CONNECTION_FEATURE_VIEWS :
 		return TRUE;
 	default:
+		return FALSE;
 	}
-
-	return FALSE;
 }
 
 static void
@@ -775,7 +774,6 @@
 	case GDA_CONNECTION_SCHEMA_TYPES :
 		return get_types ();
 	default:
+		return NULL;
 	}
-
-	return NULL;
 }
