--- ./SpiralSound/Plugins/Widgets/Fl_VU_Meter.cxx.org	Mon Jan 20 12:35:00 2003
+++ ./SpiralSound/Plugins/Widgets/Fl_VU_Meter.cxx	Mon Jan 20 12:35:42 2003
@@ -2,7 +2,7 @@
 #include "Fl_VU_Meter.h"
 #include <FL/fl_draw.H>
 
-Fl_VU_Meter::Fl_VU_Meter (int x, int y, int w, int h, const char *l = 0) :
+Fl_VU_Meter::Fl_VU_Meter (int x, int y, int w, int h, const char *l ) :
 Fl_Progress (x, y, w, h, l) {
 }
 
--- ./SpiralSound/Plugins/Widgets/Fl_VU_Meter.h.org	Mon Jan 20 12:35:05 2003
+++ ./SpiralSound/Plugins/Widgets/Fl_VU_Meter.h	Mon Jan 20 12:36:08 2003
@@ -5,7 +5,7 @@
 
 class Fl_VU_Meter : public Fl_Progress {
   public:
-    FL_EXPORT Fl_VU_Meter (int x, int y, int w, int h, const char *l = 0);
+    FL_EXPORT Fl_VU_Meter (int x, int y, int w, int h, const char *l );
   protected:
     FL_EXPORT virtual void draw();
 };
