diff -Naur doxygen-1.2.18.000/configure doxygen-1.2.18/configure
--- doxygen-1.2.18.000/configure	2002-08-21 18:48:09.000000000 +0100
+++ doxygen-1.2.18/configure	2003-04-17 18:33:26.000000000 +0100
@@ -287,7 +287,7 @@
   for i in $install_names; do
     for j in $install_dirs; do
       if test -x "$j/$i"; then
-        if test -n "`$j/$i --version 2>/dev/null | grep fileutils`"; then
+        if test -n "`$j/$i --version 2>/dev/null | grep -E \"(file|core)utils\"`"; then
           install_found=YES
           install_prog="$j/$i"
           break 2
