diff -Naur kdemultimedia-3.4.0.orig/mpeglib/lib/input/cdromAccess.cpp kdemultimedia-3.4.0/mpeglib/lib/input/cdromAccess.cpp
--- kdemultimedia-3.4.0.orig/mpeglib/lib/input/cdromAccess.cpp	2005-01-03 16:53:36.000000000 +0200
+++ kdemultimedia-3.4.0/mpeglib/lib/input/cdromAccess.cpp	2005-03-06 15:04:12.252743928 +0200
@@ -29,7 +29,11 @@
   #if LINUX_VERSION_CODE >= KERNEL_VERSION(2,5,70)
   #include <bits/wordsize.h>
   #if __WORDSIZE == 64
-  typedef unsigned long __u64;
+  #ifdef __x86_64__
+	  typedef unsigned long long __u64;
+  #else
+	  typedef unsigned long __u64;
+  #endif
   #else
   typedef unsigned long long __u64;
   #endif 
