--- ./configure.orig	2005-03-18 00:08:16.199392800 -0400
+++ ./configure	2005-03-18 00:08:35.952389888 -0400
@@ -5103,7 +5103,7 @@
       fi
    fi
 fi
-OOO_CUSTOM_IMAGES=ooo_custom_images-12.tar.bz2
+OOO_CUSTOM_IMAGES=ooo_custom_images-13.tar.bz2
 OOO_EXTRA_ARTWORK=extras-1.tar.bz2
 OOO_CRYSTAL_IMAGES=ooo_crystal_images-5.tar.bz2
 
--- ./Makefile.shared.orig	2005-03-18 00:30:35.708756424 -0400
+++ ./Makefile.shared	2005-03-18 00:30:57.913380808 -0400
@@ -43,7 +43,7 @@
 		     $(OOBUILDDIR)/external/unicows/*.dll || true; \
 	fi
 	mkdir -p $(OOBUILDDIR)/libwpd/download
-	cp -af $(SRCDIR)/libwpd*.tar.gz $(OOBUILDDIR)/libwpd/download
+	cp -af $(SRCDIR)/libwpd*.tar.bz2 $(OOBUILDDIR)/libwpd/download
 	$(TOOLSDIR)/bin/transform --revert $(TOOLSDIR) $(OOBUILDDIR)
 	$(TOOLSDIR)/patches/apply.pl $(APPLY_DIR) $(OOBUILDDIR) --distro=$(DISTRO)
 	$(TOOLSDIR)/bin/transform --apply $(TOOLSDIR) $(OOBUILDDIR)
--- ./Makefile.in.orig	2005-03-18 00:53:45.010550448 -0400
+++ ./Makefile.in	2005-03-18 00:54:39.440275872 -0400
@@ -658,7 +658,7 @@
 		     $(OOBUILDDIR)/external/unicows/*.dll || true; \
 	fi
 	mkdir -p $(OOBUILDDIR)/libwpd/download
-	cp -af $(SRCDIR)/libwpd*.tar.gz $(OOBUILDDIR)/libwpd/download
+	cp -af $(SRCDIR)/libwpd*.tar.bz2 $(OOBUILDDIR)/libwpd/download
 	$(TOOLSDIR)/bin/transform --revert $(TOOLSDIR) $(OOBUILDDIR)
 	$(TOOLSDIR)/patches/apply.pl $(APPLY_DIR) $(OOBUILDDIR) --distro=$(DISTRO)
 	$(TOOLSDIR)/bin/transform --apply $(TOOLSDIR) $(OOBUILDDIR)
