--- OOO_1_1_3/config_office/configure.~1.36.6.9.~	2004-08-27 12:03:28.000000000 +0200
+++ OOO_1_1_3/config_office/configure	2004-10-27 16:07:17.932901600 +0200
@@ -5166,7 +5166,7 @@
 cat confdefs.h >>conftest.$ac_ext
 cat >>conftest.$ac_ext <<_ACEOF
 /* end confdefs.h.  */
-$STLPORT4/stlport/hash_map
+$STLPORT4/include/stlport/hash_map
 _ACEOF
 if { (eval echo "$as_me:$LINENO: \"$ac_cpp conftest.$ac_ext\"") >&5
   (eval $ac_cpp conftest.$ac_ext) 2>conftest.er1
@@ -5198,7 +5198,7 @@
 fi
 rm -f conftest.err conftest.$ac_ext
         else
-                      if test -f "$STLPORT4/stlport/hash_map"; then
+                      if test -f "$STLPORT4/include/stlport/hash_map"; then
               echo "$as_me:$LINENO: result: checked." >&5
 echo "${ECHO_T}checked." >&6
            else
