--- ooo-build-1.3.6/bin/build-ooo~	2004-11-12 12:44:40.000000000 +0100
+++ ooo-build-1.3.6/bin/build-ooo	2004-11-25 03:01:09.065129632 +0100
@@ -19,7 +19,7 @@
 
 # If we are forcing gcc33, make sure configure knows
 if test "z$FORCE_GCC33" = "zYES"; then
-CONFIGURE_OPTIONS="$CONFIGURE_OPTIONS CFLAGS= CXXFLAGS= CC=gcc33 CXX=g++33"
+CONFIGURE_OPTIONS="$CONFIGURE_OPTIONS --with-gxx-include-path=c++"
 fi
 
 ./configure $CONFIGURE_OPTIONS || exit 1;
