diff -urN -x '*~' blackbox-0.65.0/src/Window.cc blackbox-0.65.0-gcc-3.3.1/src/Window.cc
--- blackbox-0.65.0/src/Window.cc	2002-09-17 17:23:35.000000000 +0200
+++ blackbox-0.65.0-gcc-3.3.1/src/Window.cc	2003-08-14 22:22:57.000000000 +0200
@@ -44,6 +44,12 @@
 #endif // HAVE_STDLIB_H
 }
 
+#if defined(__GNUC__)
+#if __GNUC__ == 3
+#include <cassert>
+#endif
+#endif
+
 #include "i18n.hh"
 #include "blackbox.hh"
 #include "GCCache.hh"
