--- wxWidgets-3.3.2/build/cmake/init.cmake.vanilla	2026-03-20 18:44:27.482587257 +0100
+++ wxWidgets-3.3.2/build/cmake/init.cmake	2026-03-20 19:02:56.229342112 +0100
@@ -568,9 +568,12 @@
                 if(WXGTK2)
                     find_package(WEBKIT 1.0)
                 elseif(WXGTK3)
-                    find_package(WEBKIT2 4.1 QUIET)
+                    find_package(PkgConfig REQUIRED)
+                    pkg_check_modules(WEBKIT2 webkit2gtk-4.1)
                     if(WEBKIT2_FOUND)
                         set(WEBKIT_LIBSOUP_VERSION 3.0)
+			set(WEBKIT2_INCLUDE_DIR ${WEBKIT2_PREFIX}/include/${WEBKIT2_MODULE_NAME})
+			set(WEBKIT2_LIBRARY ${WEBKIT2_MODULE_NAME})
                     else()
                         find_package(WEBKIT2 4.0)
                     endif()
