diff -Naur wxWidgets-2.8.10-orig/src/common/strconv.cpp wxWidgets-2.8.10/src/common/strconv.cpp
--- wxWidgets-2.8.10-orig/src/common/strconv.cpp	2009-03-06 14:10:56.000000000 +0200
+++ wxWidgets-2.8.10/src/common/strconv.cpp	2009-12-12 09:02:06.000000000 +0200
@@ -1676,7 +1676,7 @@
         {
 #if SIZEOF_WCHAR_T == 4
             _T("UCS-4"),
-#elif SIZEOF_WCHAR_T = 2
+#elif SIZEOF_WCHAR_T == 2
             _T("UCS-2"),
 #endif
             NULL
