--- ./src/interface.c.orig	2003-09-01 22:50:23.000000000 +0300
+++ ./src/interface.c	2003-12-26 13:52:08.000000000 +0200
@@ -21,7 +21,7 @@
 #include "callbacks.h"
 #include "interface.h"
 #include "support.h"
-#include "xfcalendar-icon-inline.h"
+#include "calendar-icon.h"
 
 #define GLADE_HOOKUP_OBJECT(component,widget,name) \
   g_object_set_data_full (G_OBJECT (component), name, \
@@ -53,7 +53,7 @@
   GtkAccelGroup *accel_group;
   GdkPixbuf *xfcalendar_logo;
 
-  xfcalendar_logo = inline_icon_at_size(xfcalendar_icon, 48, 48);
+  xfcalendar_logo = inline_icon_at_size(calendar_icon_data, 48, 48);
   accel_group = gtk_accel_group_new ();
 
   XFCalendar = gtk_window_new (GTK_WINDOW_TOPLEVEL);
@@ -342,7 +342,7 @@
   GdkPixbuf *xfcalendar_logo;
   gchar *textInfo;
 
-  xfcalendar_logo = inline_icon_at_size(xfcalendar_icon, 48, 48);
+  xfcalendar_logo = inline_icon_at_size(calendar_icon_data, 48, 48);
 
   wInfo = gtk_dialog_new ();
   gtk_widget_set_size_request (wInfo, 300, 370);
