--- partimage-0.6.5_beta3/src/client/fs/fs_xfs.h.orig	2006-04-22 08:22:48.000000000 -0700
+++ partimage-0.6.5_beta3/src/client/fs/fs_xfs.h	2006-12-31 01:45:31.000000000 -0800
@@ -150,13 +150,13 @@
 typedef	__uint64_t	xfs_dfiloff_t;	// block number in a file 
 typedef	__uint64_t	xfs_dfilblks_t;	// number of blocks in a file 
 
-typedef uint64_t __u64;
-typedef int64_t  __s64;
+typedef uint64_t my__u64;
+typedef int64_t  my__s64;
 
-typedef __u64	xfs_off_t;
+typedef my__u64	xfs_off_t;
 //typedef __s32	xfs32_off_t;
-typedef __u64	xfs_ino_t;	// <inode> type 
-typedef __s64	xfs_daddr_t;	// <disk address> type 
+typedef my__u64	xfs_ino_t;	// <inode> type 
+typedef my__s64	xfs_daddr_t;	// <disk address> type 
 typedef char *	xfs_caddr_t;	// <core address> type 
 typedef off_t	linux_off_t;
 //typedef __kernel_ino_t	linux_ino_t;
