--- ./mpeg2enc/picture.cc.orig	2004-05-14 09:59:53.570487912 +0200
+++ ./mpeg2enc/picture.cc	2004-05-14 10:00:08.136273576 +0200
@@ -83,9 +83,9 @@
     }
 
 
-	curref = new (uint8_t *)[5];
-	curorg = new (uint8_t *)[5];
-	pred   = new (uint8_t *)[5];
+	curref = new uint8_t* [5];
+	curorg = new uint8_t* [5];
+	pred   = new uint8_t* [5];
 
 	for( i = 0 ; i<3; i++)
 	{
--- ./lavtools/lavpipe.c.orig	2004-05-14 10:27:18.105480592 +0200
+++ ./lavtools/lavpipe.c	2004-05-14 10:28:04.736391616 +0200
@@ -582,6 +582,7 @@
       mjpeg_info( "closing input %d (source %d)", i, current_index);
       decommission_pipe_source(source);
     KEEP_SOURCE:
+      ;
     }
   }
 }
