http://bugzilla.gnome.org/show_bug.cgi?id=342968

diff -Naur gnome-terminal-2.14.2.orig/src/terminal-window.c gnome-terminal-2.14.2/src/terminal-window.c
--- gnome-terminal-2.14.2.orig/src/terminal-window.c	2006-03-04 07:21:01.000000000 +0100
+++ gnome-terminal-2.14.2/src/terminal-window.c	2006-10-06 18:53:03.000000000 +0200
@@ -770,6 +770,8 @@
   window->priv->active_term = NULL;
   window->priv->menubar = gtk_menu_bar_new ();
   window->priv->menubar_visible = FALSE;
+  /* workaround for a resize bug, see bugzilla 342968 comment 45 */
+  gtk_widget_set_size_request (window->priv->menubar, 0, -1);
   
   window->priv->main_vbox = gtk_vbox_new (FALSE, 0);
   window->priv->notebook = terminal_notebook_new ();
