
  - Rene Rebe <rene@exactcode.de>

--- ash-linux-0.2/ash-linux-0.2/Makefile	2005-02-02 11:09:38.579275000 +0100
+++ ash-linux-0.2/Makefile	2005-02-02 11:13:54.344393016 +0100
@@ -27,8 +27,8 @@
 	$(CC) -o $(PROG) $(OBJS) $(LDFLAGS)
 
 install: all
-	install sh /bin/ash
-	install -m 644 sh.1 /usr/man/man1/ash.1
+	install sh $(DESTDIR)/bin/ash
+	install -m 644 sh.1 $(DESTDIR)/usr/man/man1/ash.1
 
 parser.o: token.def
 token.def: mktokens
