--- clockspeed-0.62.orig/error.h	1998-10-13 19:37:49.000000000 +0200
+++ clockspeed-0.62/error.h	2005-02-12 16:57:50.200179568 +0100
@@ -1,7 +1,7 @@
 #ifndef ERROR_H
 #define ERROR_H
 
-extern int errno;
+#include <errno.h>
 
 extern int error_intr;
 extern int error_nomem;
--- clockspeed-0.62.orig/leapsecs_read.c	1998-10-13 19:37:49.000000000 +0200
+++ clockspeed-0.62/leapsecs_read.c	2005-02-12 17:01:38.517470064 +0100
@@ -2,7 +2,6 @@
 #include <sys/stat.h>
 #include <fcntl.h>
 #include <errno.h>
-extern int errno;
 #include "tai.h"
 #include "leapsecs.h"
 
