--- clockspeed-0.62.orig/Makefile	1998-10-13 19:37:49.000000000 +0200
+++ clockspeed-0.62/Makefile	2005-02-12 16:49:11.867978040 +0100
@@ -7,8 +7,8 @@
 auto-ccld.sh: \
 conf-cc conf-ld warn-auto.sh
 	( cat warn-auto.sh; \
-	echo CC=\'`head -1 conf-cc`\'; \
-	echo LD=\'`head -1 conf-ld`\' \
+	echo CC=\'`head -n 1 conf-cc`\'; \
+	echo LD=\'`head -n 1 conf-ld`\' \
 	) > auto-ccld.sh
 
 auto-str: \
@@ -21,7 +21,7 @@
 
 auto_home.c: \
 auto-str conf-home
-	./auto-str auto_home `head -1 conf-home` > auto_home.c
+	./auto-str auto_home `head -n 1 conf-home` > auto_home.c
 
 auto_home.o: \
 compile auto_home.c
