--- clockspeed-0.62/Makefile	2006-01-02 22:00:37.000000000 +0100
+++ clockspeed-0.62/Makefile	2006-01-02 23:04:32.000000000 +0100
@@ -21,7 +21,7 @@
 
 auto_home.c: \
 auto-str conf-home
-	./auto-str auto_home `head -n 1 conf-home` > auto_home.c
+	echo auto_home=\"`head -n 1 conf-home`\"\; > auto_home.c
 
 auto_home.o: \
 compile auto_home.c
@@ -153,8 +153,8 @@
 
 hasrdtsc.h: \
 tryrdtsc.c compile load
-	( ( ./compile tryrdtsc.c && ./load tryrdtsc && ./tryrdtsc \
+	( ( ./compile tryrdtsc.c && ./load tryrdtsc \
 	) >/dev/null 2>&1 \
 	&& echo \#define HASRDTSC 1 || exit 0 ) > hasrdtsc.h
 	rm -f tryrdtsc.o tryrdtsc
 
