diff -Naur linux-2.6.11.orig/include/linux/cyclades.h linux-2.6.11/include/linux/cyclades.h
--- linux-2.6.11.orig/include/linux/cyclades.h	2005-03-02 09:38:33.000000000 +0200
+++ linux-2.6.11/include/linux/cyclades.h	2005-03-07 17:52:56.716885160 +0200
@@ -67,6 +67,8 @@
 #ifndef _LINUX_CYCLADES_H
 #define _LINUX_CYCLADES_H
 
+#include <linux/compiler.h>
+
 struct cyclades_monitor {
         unsigned long           int_count;
         unsigned long           char_count;
