--- ddd-3.3.1/ddd/configure.orig	2002-12-04 09:44:09.000000000 +0200
+++ ddd-3.3.1/ddd/configure	2002-12-04 10:45:04.000000000 +0200
@@ -13282,13 +13282,14 @@
 /* System header to define __stub macros and hopefully few prototypes,
     which can conflict with char $ac_func(); below.  */
 #include <assert.h>
+#include <math.h>
 /* Override any gcc2 internal prototype to avoid an error.  */
 #ifdef __cplusplus
 extern "C"
 #endif
 /* We use char because int might match the return type of a gcc2
     builtin and then its argument prototype would still apply.  */
-char $ac_func();
+/* char $ac_func(); */
 
 int main() {
 
@@ -13298,7 +13299,8 @@
 #if defined (__stub_$ac_func) || defined (__stub___$ac_func)
 choke me
 #else
-$ac_func();
+double x = 0, y =0;
+x = $ac_func(x, y);
 #endif
 
 ; return 0; }
