Description: Replace deprecated GTK macros with appropriate function
Author: Luca Falavigna <dktrkranz@debian.org>
Author: Stephen Kitt <steve@sk2.org>

--- gdmap.orig/src/gui_support.c
+++ gdmap/src/gui_support.c
@@ -85,7 +85,7 @@
   const char* text;
 
   (void)event;
-  if (GTK_WIDGET_SENSITIVE(label)) {
+  if (gtk_widget_get_sensitive(label)) {
     char* temp;
 
     text = g_object_get_data(G_OBJECT(label), "label");
@@ -129,7 +129,7 @@
 void ui_event_label_set_sensitive(GtkWidget* ebox, gboolean set) {
   gpointer callback = g_object_get_data(G_OBJECT(ebox), "callback");
   GtkWidget* child = GTK_BIN(ebox)->child;
-  if (GTK_WIDGET_SENSITIVE(child) == set) return;
+  if (gtk_widget_get_sensitive(child) == set) return;
 
   if (set && !callback) return;
   gtk_widget_set_sensitive(child, set);
@@ -164,7 +164,7 @@
                         // GDK_FOCUS_CHANGE_MASK |
                         GDK_ENTER_NOTIFY_MASK |
                         GDK_LEAVE_NOTIFY_MASK);
-  GTK_WIDGET_SET_FLAGS(ebox, GTK_CAN_FOCUS);
+  gtk_widget_set_can_focus(ebox, TRUE);
 
   gtk_widget_show(ebox);
 
