Description: Add missing libm
Author: Daniel T Chen <crimsun@ubuntu.com>

Index: gdmap-0.8.1/src/Makefile.am
===================================================================
--- gdmap-0.8.1.orig/src/Makefile.am	2008-05-23 08:46:24.000000000 -0400
+++ gdmap-0.8.1/src/Makefile.am	2013-04-07 18:18:34.052526861 -0400
@@ -16,5 +16,5 @@
 gdmap_CFLAGS = \
 	-DGDMAP_LOCALE_DIR=\""$(localedir)"\" -DPACKAGE_DATA_DIR=\""$(pkgdatadir)"\" @UI_CFLAGS@ @LIBXML_CFLAGS@
 
-gdmap_LDADD = @UI_LIBS@ @LIBXML_LIBS@
+gdmap_LDADD = @UI_LIBS@ @LIBXML_LIBS@ -lm
 
Index: gdmap-0.8.1/src/Makefile.in
===================================================================
--- gdmap-0.8.1.orig/src/Makefile.in	2008-06-15 09:44:10.000000000 -0400
+++ gdmap-0.8.1/src/Makefile.in	2013-04-07 18:18:44.276365301 -0400
@@ -213,7 +213,7 @@
 gdmap_CFLAGS = \
 	-DGDMAP_LOCALE_DIR=\""$(localedir)"\" -DPACKAGE_DATA_DIR=\""$(pkgdatadir)"\" @UI_CFLAGS@ @LIBXML_CFLAGS@
 
-gdmap_LDADD = @UI_LIBS@ @LIBXML_LIBS@
+gdmap_LDADD = @UI_LIBS@ @LIBXML_LIBS@ -lm
 all: all-am
 
 .SUFFIXES:
