--- ./lib/list.h.orig	2002-12-16 22:24:27.000000000 +0200
+++ ./lib/list.h	2004-06-12 13:28:00.069138384 +0300
@@ -111,8 +111,8 @@
     }
 private:
   const list<T>& lst;
-  const list<T>::node* prev;
-  const list<T>::node* curr;
+  const typename list<T>::node* prev;
+  const typename list<T>::node* curr;
 };
 
 template<class T>
@@ -164,8 +164,8 @@
     }
 private:
   list<T>& lst;
-  list<T>::node* prev;
-  list<T>::node* curr;
+  typename list<T>::node* prev;
+  typename list<T>::node* curr;
 };
 
 template<class T>
