--- ./Makefile.in.orig	2005-03-11 00:11:50.701033912 -0300
+++ ./Makefile.in	2005-03-11 00:12:23.462053480 -0300
@@ -584,11 +584,11 @@
 	chmod 600 $(DESTDIR)$(localstatedir)/trigger
 
 install-root:
-	chown nullmail $(DESTDIR)$(localstatedir)/*
-	chown nullmail $(DESTDIR)$(sbindir)/nullmailer-queue
+	chown 38 $(DESTDIR)$(localstatedir)/*
+	chown 38 $(DESTDIR)$(sbindir)/nullmailer-queue
 	chmod u+s $(DESTDIR)$(sbindir)/nullmailer-queue
-	chown nullmail $(DESTDIR)$(bindir)/mailq
-	chmod u+s $(DESTDIR)$(bindir)/mailq
+	chown 38 $(DESTDIR)$(bindir)/mailq_nullmailer
+	chmod u+s $(DESTDIR)$(bindir)/mailq_nullmailer
 
 dist-hook:
 	sed -e s/@VERSION\@/@VERSION@/ \
