--- ./panel-plugin/vc_oss.c.orig	2003-07-18 16:08:27.000000000 +0300
+++ ./panel-plugin/vc_oss.c	2003-08-14 13:55:03.000000000 +0300
@@ -74,7 +74,7 @@
 
 #define MAX_AMP 100
 
-static char dev_name[PATH_MAX] = { "/dev/mixer" };
+static char dev_name[PATH_MAX] = { "/dev/sound/mixer" };
 
 static int mixer_handle = -1;
 static int devmask = 0;                /* Bitmask for supported mixer devices */
