--- kdeadmin-3.1.4/configure.orig	2003-12-03 18:10:11.000000000 +0100
+++ kdeadmin-3.1.4/configure	2003-12-03 18:11:29.000000000 +0100
@@ -27469,7 +27469,7 @@
 #define HAVE_PAM 1
 _ACEOF
 
-    PAMLIBS="$PAM_MISC_LIB -lpam $LIBDL"
+    PAMLIBS="$PAM_MISC_LIB -lpam -lpam_misc $LIBDL"
 
         echo "$as_me:$LINENO: checking for const pam_message" >&5
 echo $ECHO_N "checking for const pam_message... $ECHO_C" >&6
--- kdeadmin-3.1.4/kuser/Makefile.in.orig	2003-12-03 18:09:10.000000000 +0100
+++ kdeadmin-3.1.4/kuser/Makefile.in	2003-12-03 18:10:02.000000000 +0100
@@ -280,7 +280,7 @@
 AM_CXXFLAGS = $(KU_NIS) $(KU_QUOTA) $(QUOTACTL) -D_KU_DEBUG -D_SVID_SOURCE -D_GNU_SOURCE -DQT_NO_ASCII_CAST -DQT_NO_CAST_ASCII
 
 kuser_LDFLAGS = $(all_libraries) $(KDE_RPATH)
-kuser_LDADD = $(LIB_KDEUI) $(LIBCRYPT) $(LIBSHADOW) $(LIBGEN)
+kuser_LDADD = $(LIB_KDEUI) $(LIBCRYPT) $(LIBSHADOW) $(PAMLIBS) $(LIBGEN)
 
 
 ####### Files
