--- libxml2-2.6.2/doc/Makefile.in.orig	2003-11-04 11:27:20.000000000 +0100
+++ libxml2-2.6.2/doc/Makefile.in	2003-11-25 20:29:40.000000000 +0100
@@ -415,14 +415,14 @@
 	-@(xsltproc --nonet xmlcatalog_man.xml)
 
 scan:
-	-gtkdoc-scan --module=libxml --source-dir=$(DOC_SOURCE_DIR) --ignore-headers="acconfig.h config.h win32config.h trio.h triostr.h triop.h config-mac.h XMLTestPrefix2.h  XMLTestPrefix.h triodef.h trionan.h xlink.h libxml.h libxml2-py.h libxml_wrap.h"
+	-gtkdoc-scan --module=libxml2 --source-dir=$(DOC_SOURCE_DIR) --ignore-headers="acconfig.h config.h win32config.h trio.h triostr.h triop.h config-mac.h XMLTestPrefix2.h  XMLTestPrefix.h triodef.h trionan.h xlink.h libxml.h libxml2-py.h libxml_wrap.h"
 
 templates: scan
-	-gtkdoc-mktmpl --module=libxml
+	-gtkdoc-mktmpl --module=libxml2
 
 xml: templates
 	if test -d xml ; then rm -f xml/* ; else mkdir xml ; fi
-	-gtkdoc-mkdb --module=libxml --source-dir=$(DOC_SOURCE_DIR) --output-format=xml --main-sgml-file=$(DOC_MAIN_XML_FILE)
+	-gtkdoc-mkdb --module=libxml2 --source-dir=$(DOC_SOURCE_DIR) --output-format=xml --main-sgml-file=$(DOC_MAIN_XML_FILE)
 
 libxml2-api.xml: apibuild.py ../include/libxml/*.h ../*.c
 	-(./apibuild.py)
@@ -436,7 +436,7 @@
 
 html:
 	if test -n -d html ; then mkdir html ; fi
-	-cd html && gtkdoc-mkhtml libxml ../$(DOC_MAIN_XML_FILE)
+	-cd html && gtkdoc-mkhtml libxml2 ../$(DOC_MAIN_XML_FILE)
 
 clean-local:
 	rm -f *~ *.bak *.hierarchy *.signals *-unused.txt
@@ -467,7 +467,7 @@
 	$(mkinstalldirs) $(DESTDIR)$(TARGET_DIR)/tutorial/images/callouts
 	-@INSTALL@ -m 0644 $(srcdir)/tutorial/images/callouts/*.* \
 		$(DESTDIR)$(TARGET_DIR)/tutorial/images/callouts
-	-(cd $(DESTDIR); gtkdoc-fixxref --module=libxml --html-dir=$(HTML_DIR))
+	-(cd $(DESTDIR); gtkdoc-fixxref --module=libxml2 --html-dir=$(HTML_DIR))
 
 .PHONY : html xml templates scan
 # Tell versions [3.59,3.63) of GNU make to not export all variables.
