--- esmart-2004-01-18/src/container/layout/default/Makefile.am.orig	2004-01-21 16:10:44.000000000 +0100
+++ esmart-2004-01-18/src/container/layout/default/Makefile.am	2004-01-21 16:10:52.000000000 +0100
@@ -1,5 +1,5 @@
 plugin_LTLIBRARIES = default.la
-plugindir = /usr/local/lib/esmart/layout
+plugindir = $(libdir)/esmart/layout
 
 default_la_SOURCES = default.c
 default_la_LDFLAGS = -module  -avoid-version
--- esmart-2004-01-18/src/container/layout/entice/Makefile.am.orig	2004-01-21 16:11:20.000000000 +0100
+++ esmart-2004-01-18/src/container/layout/entice/Makefile.am	2004-01-21 16:11:04.000000000 +0100
@@ -1,5 +1,5 @@
 plugin_LTLIBRARIES = entice.la
-plugindir = /usr/local/lib/esmart/layout
+plugindir = $(libdir)/esmart/layout
 
 entice_la_SOURCES = entice.c
 entice_la_LDFLAGS = -module  -avoid-version
