--- sysvinit-3.04/src/killall5.c.vanilla	2024-09-01 20:09:01.736907043 +0200
+++ sysvinit-3.04/src/killall5.c	2024-09-01 20:13:27.153580134 +0200
@@ -1081,7 +1081,7 @@
 int main(int argc, char **argv)
 {
 	PROC		*p;
-	int		pid, sid = -1;
+	pid_t		pid, sid;
 	int		sig = SIGKILL;
 	int		c;
 
@@ -1163,8 +1163,8 @@
 	}
 
 	/* Now kill all processes except init (pid 1) and our session. */
-	sid = (int)getsid(0);
-	pid = (int)getpid();
+	sid = getsid(0);
+	pid = getpid();
 	for (p = plist; p; p = p->next) {
 		if (p->pid == 1 || p->pid == pid || p->sid == sid || p->kernel)
 			continue;
