--- netcat-1.10.orig/netcat.c	Tue Jun  4 15:35:11 2002
+++ netcat-1.10/netcat.c	Tue Jun  4 15:35:26 2002
@@ -1640,8 +1640,16 @@
 /* if we're scanning at a "one -v" verbosity level, don't print refusals.
    Give it another -v if you want to see everything. */
 	if ((Single || (o_verbose > 1)) || (errno != ECONNREFUSED))
+	{
+	  /* bug 65413 - if we're not scanning, we always want an
+	   * error to be printed for refused connects. This is a
+	   * disgustingly ugly way to do it, I really should just
+	   * rewrite the holler() interface... */
+	  if (Single) o_verbose++;
 	  holler ("%s [%s] %d (%s)",
 	    whereto->name, whereto->addrs[0], curport, portpoop->name);
+	  if (Single) o_verbose--;
+	}
       } /* if netfd */
       close (netfd);			/* just in case we didn't already */
       if (o_interval)
