--- netcat-1.10.orig/netcat.c	Tue Jun  4 15:35:33 2002
+++ netcat-1.10/netcat.c	Tue Jun  4 15:35:41 2002
@@ -1394,7 +1394,8 @@
     argv[1] = cp;			/* head of new arg block */
     fprintf (stderr, "Cmd line: ");
     fflush (stderr);		/* I dont care if it's unbuffered or not! */
-    insaved = read (0, cp, BIGSIZ);	/* we're gonna fake fgets() here */
+    insaved = read (0, cp, BIGSIZ-1);	/* we're gonna fake fgets() here */
+    cp[BIGSIZ-1] = '\0';
     if (insaved <= 0)
       bail ("wrong");
     x = findline (cp, insaved);
