--- xchat-2.0.4/configure.orig	2003-07-31 07:24:30.000000000 +0200
+++ xchat-2.0.4/configure	2003-08-15 18:28:09.986630464 +0200
@@ -23467,8 +23467,8 @@
         else
                 echo "$as_me:$LINENO: checking Python version" >&5
 echo $ECHO_N "checking Python version... $ECHO_C" >&6
-                                PY_VERSION=`$pythonpath -c 'import sys; print sys.version[:3]'`
-                PY_PATH=`$pythonpath -c 'import sys ; print sys.path[1][5:]'`
+                                PY_VERSION=`$pythonpath -c 'import distutils.sysconfig; print distutils.sysconfig.get_python_version()'`
+                PY_PATH=`$pythonpath -c 'import distutils.sysconfig; print distutils.sysconfig.get_python_lib(standard_lib=1)'`
                 $pythonpath -c "import sys; map(int,sys.version[:3].split('.')) >= [2,2] or sys.exit(1)"
                                 echo "$as_me:$LINENO: result: $PY_VERSION" >&5
 echo "${ECHO_T}$PY_VERSION" >&6
@@ -23477,9 +23477,10 @@
 echo $ECHO_N "checking Python compile flags... $ECHO_C" >&6
                         PY_PREFIX=`$pythonpath -c 'import sys; print sys.prefix'`
                         PY_EXEC_PREFIX=`$pythonpath -c 'import sys; print sys.exec_prefix'`
-                        if test -f $PY_PREFIX/include/python$PY_VERSION/Python.h; then
-                                PY_LIBS="-L$PY_EXEC_PREFIX/$PY_PATH/config -lpython$PY_VERSION -lpthread -lutil"
-                                PY_CFLAGS="-I$PY_PREFIX/include/python$PY_VERSION"
+PY_INCLUDE=`$pythonpath -c 'import distutils.sysconfig; print distutils.sysconfig.get_python_inc()'`
+                        if test -f $PY_INCLUDE/Python.h; then
+                                PY_LIBS="-L$PY_PATH/config -lpython$PY_VERSION -lpthread -lutil"
+                                PY_CFLAGS="-I$PY_INCLUDE"
                                 echo "$as_me:$LINENO: result: ok" >&5
 echo "${ECHO_T}ok" >&6
                         else
