--- honeyd-1.5a/Makefile.in.orig	2006-02-20 12:55:22.000000000 +0100
+++ honeyd-1.5a/Makefile.in	2006-02-20 12:55:42.000000000 +0100
@@ -1317,6 +1317,7 @@
 	rm -f .libs/libtmp$@.*
 
 install-data-local:
+	test -z "$(honeyddatadir)" || $(mkdir_p) "$(DESTDIR)$(honeyddatadir)"
 	(cd $(top_srcdir) && tar -cf - $(top_srcdir)/webserver) | \
 	(cd $(honeyddatadir) && tar -xf -)
 	find $(honeyddatadir)/webserver -type f | xargs chmod a+r
