--- nfs-utils-1.0.7/support/rpc/svc_auth_gss.c.vanilla	2005-09-04 11:05:10.000000000 +0000
+++ nfs-utils-1.0.7/support/rpc/svc_auth_gss.c	2005-09-04 11:08:31.000000000 +0000
@@ -382,7 +382,7 @@
 			return (AUTH_FAILED);
 		}
 		auth->svc_ah_ops = &svc_auth_gss_ops;
-		SVCAUTH_PRIVATE(auth) = gd;
+		auth = gd;
 		rqst->rq_xprt->xp_auth = auth;
 	}
 	else gd = SVCAUTH_PRIVATE(rqst->rq_xprt->xp_auth);
