--- ./cuneiform_src/Kern/rout/src/html.cpp	2008-09-03 13:43:55 +0000
+++ ./cuneiform_src/Kern/rout/src/html.cpp	2008-09-07 11:39:43 +0000
@@ -73,7 +73,7 @@
 static BOOL Static_MakeHTML(Handle hObject, long reason);
 
 static BOOL FontStyle(ULONG newStyle);
-static BOOL BeginParagraph(ULONG alignment);
+static BOOL BeginParagraph(Handle hObject);
 static BOOL CellStart();
 static BOOL CalcCellSpan();
 static BOOL OptimizeTags();
@@ -121,37 +119,48 @@
 			FontStyle(CED_GetCharFontAttribs(hObject));
 
 			// Записать символ
-			ONE_CHAR(hObject);
+			{
+				char buf[256] = "";
+				edRect r = CED_GetCharLayout(hObject);
+				if (r.left != -1) {
+					sprintf (buf, "<span title=\"bbox %d %d %d %d\">", r.left, r.top, r.right, r.bottom);
+					PUT_STRING(buf);
+				}
+
+				ONE_CHAR(hObject);
+				if (r.left != -1)
+					PUT_STRING("</span>");
+			}
 
 			break;
 
 		case BROWSE_LINE_END:
 			// Конец строки текста
 			if ( gPreserveLineBreaks || gEdLineHardBreak )
-				PUT_STRING("<br>");
+				PUT_STRING("<br>\n");
 			break;
 
 		case BROWSE_PARAGRAPH_START:
 			// Начало абзаца
 			FontStyle(0);
-			BeginParagraph(CED_GetAlignment(hObject));
+			BeginParagraph(hObject);
 			break;
 
 		case BROWSE_PARAGRAPH_END:
 			// Конец абзаца
 			FontStyle(0);
-			PUT_STRING("</p>");
+			PUT_STRING("</p>\n");
 			break;
 
 		case BROWSE_PAGE_START:
 			// Начало страницы
 			FontStyle(0);
-			PUT_STRING("<html><body>");
+			PUT_STRING("<html><body>\n");
 			break;
 
 		case BROWSE_PAGE_END:
 			// Конец страницы
-			PUT_STRING("</body></html>");
+			PUT_STRING("</body></html>\n");
 			break;
 
 		case BROWSE_TABLE_START:
@@ -163,7 +172,7 @@
 		case BROWSE_TABLE_END:
 			// Конец таблицы
 			FontStyle(0);
-			PUT_STRING("</table>");
+			PUT_STRING("</table>\n");
 			break;
 
 		case BROWSE_ROW_START:
@@ -236,13 +236,16 @@
 sFontStyle = newStyle;
 return TRUE;
 }
-//********************************************************************
-static BOOL BeginParagraph(ULONG alignment)
+
+static BOOL BeginParagraph(Handle hObject)
 {
 	char *p = 0;
 	char buf[80] = "";
 
-switch ( alignment & ALIGN_MASK)
+	edBox b = CED_GetLayout(hObject);
+
+	ULONG alignment = CED_GetAlignment(hObject);
+	switch (alignment & ALIGN_MASK)
 	{
 	case ALIGN_CENTER:
 		p = "center";
@@ -244,14 +256,21 @@
 	case ALIGN_LEFT:
 	default:
 		// "left" by default
-		PUT_STRING("<p>");
-		return TRUE;
-	}
-
-sprintf(buf,"<p align=%s>",p);
-PUT_STRING(buf);
-
-return TRUE;
+		;
+	}
+	PUT_STRING("<p");
+	if (p) {
+		sprintf(buf," align=%s", p);
+		PUT_STRING(buf);
+	}
+	
+	if (b.x != -1) {
+		sprintf(buf," title=\"bbox %d %d %d %d\"", b.x, b.y, b.x + b.w, b.y + b.h);
+		PUT_STRING(buf);
+	}
+	PUT_STRING(">");
+
+	return TRUE;
 }
 //********************************************************************
 static BOOL CellStart()

=== modified file 'cuneiform_src/Kern/rout/src/text.cpp'
--- ./cuneiform_src/Kern/rout/src/text.cpp	2008-07-29 09:38:01 +0000
+++ ./cuneiform_src/Kern/rout/src/text.cpp	2008-09-08 13:24:23 +0000
@@ -243,14 +243,22 @@
 			break;
 
 		// Угловые скобки в HTML заменяются на круглые
+		case '&':
 		case '<':
-			if (gFormat == ROUT_FMT_HTML)
-				c2 = '(';
-			break;
-
 		case '>':
-			if (gFormat == ROUT_FMT_HTML)
-				c2 = ')';
+			if (gFormat == ROUT_FMT_HTML) {
+				*gMemCur++ = '&';
+				if (c1 == '<' || c1 == '>') {
+					*gMemCur++ = (c1 == '<' ? 'l' : 'g');
+					*gMemCur++ = 't';
+				} else  {
+					*gMemCur++ = 'a';
+					*gMemCur++ = 'm';
+					*gMemCur++ = 'p';
+				}
+				*gMemCur++ = ';';
+				return TRUE;
+			}
 			break;
 
 		// Дифтонг oe / OE

