diff -Naur kdepim-3.5.0-orig/config.h.in kdepim-3.5.0/config.h.in
--- kdepim-3.5.0-orig/config.h.in	2005-11-26 09:34:43.000000000 +0200
+++ kdepim-3.5.0/config.h.in	2005-12-25 12:53:06.885518000 +0200
@@ -177,8 +177,8 @@
 /* Define if you have the res_init prototype */
 #undef HAVE_RES_INIT_PROTO
 
-/* Define to 1 if you have the <sasl/sasl.h> header file. */
-#undef HAVE_SASL_SASL_H
+/* Define to 1 if you have the <sasl2/sasl.h> header file. */
+#undef HAVE_SASL2_SASL_H
 
 /* Define to 1 if you have the `setenv' function. */
 #undef HAVE_SETENV
diff -Naur kdepim-3.5.0-orig/configure kdepim-3.5.0/configure
--- kdepim-3.5.0-orig/configure	2005-11-26 09:36:58.000000000 +0200
+++ kdepim-3.5.0/configure	2005-12-25 12:53:06.901519000 +0200
@@ -38245,7 +38245,7 @@
 
 
 
-for ac_header in sasl/sasl.h
+for ac_header in sasl2/sasl.h
 do
 as_ac_Header=`echo "ac_cv_header_$ac_header" | $as_tr_sh`
 if eval "test \"\${$as_ac_Header+set}\" = set"; then
diff -Naur kdepim-3.5.0-orig/configure.in.rej kdepim-3.5.0/configure.in.rej
--- kdepim-3.5.0-orig/configure.in.rej	1970-01-01 02:00:00.000000000 +0200
+++ kdepim-3.5.0/configure.in.rej	2005-12-25 12:53:06.905519250 +0200
@@ -0,0 +1,17 @@
+***************
+*** 223,229 ****
+  sasl2_header="no"
+  SASL2_LIBS=""
+  
+- KDE_CHECK_HEADERS(sasl/sasl.h, sasl2_header="yes")
+  if test "$sasl2_header" = "yes" ; then
+    KDE_CHECK_LIB(sasl2, sasl_client_init, SASL2_LIBS="-lsasl2")
+  fi
+--- 223,229 ----
+  sasl2_header="no"
+  SASL2_LIBS=""
+  
++ KDE_CHECK_HEADERS(sasl2/sasl.h, sasl2_header="yes")
+  if test "$sasl2_header" = "yes" ; then
+    KDE_CHECK_LIB(sasl2, sasl_client_init, SASL2_LIBS="-lsasl2")
+  fi
diff -Naur kdepim-3.5.0-orig/kioslaves/configure.in.in kdepim-3.5.0/kioslaves/configure.in.in
--- kdepim-3.5.0-orig/kioslaves/configure.in.in	2005-10-10 18:02:21.000000000 +0300
+++ kdepim-3.5.0/kioslaves/configure.in.in	2005-12-25 12:56:10.084967250 +0200
@@ -8,7 +8,7 @@
 sasl2_header="no"
 SASL2_LIBS=""
 if test "x$with_sasl" != xno; then
-  KDE_CHECK_HEADERS(sasl/sasl.h, sasl2_header="yes")
+  KDE_CHECK_HEADERS(sasl2/sasl.h, sasl2_header="yes")
   if test "$sasl2_header" = "yes" ; then
     KDE_CHECK_LIB(sasl2, sasl_client_init, SASL2_LIBS="-lsasl2")
   fi
diff -Naur kdepim-3.5.0-orig/kioslaves/imap4/imap4.cc kdepim-3.5.0/kioslaves/imap4/imap4.cc
--- kdepim-3.5.0-orig/kioslaves/imap4/imap4.cc	2005-10-10 18:02:21.000000000 +0300
+++ kdepim-3.5.0/kioslaves/imap4/imap4.cc	2005-12-25 12:53:06.905519250 +0200
@@ -74,7 +74,7 @@
 
 #ifdef HAVE_LIBSASL2
 extern "C" {
-#include <sasl/sasl.h>
+#include <sasl2/sasl.h>
 }
 #endif
 
diff -Naur kdepim-3.5.0-orig/kioslaves/imap4/imapparser.cc kdepim-3.5.0/kioslaves/imap4/imapparser.cc
--- kdepim-3.5.0-orig/kioslaves/imap4/imapparser.cc	2005-09-10 11:24:05.000000000 +0300
+++ kdepim-3.5.0/kioslaves/imap4/imapparser.cc	2005-12-25 12:53:06.905519250 +0200
@@ -43,7 +43,7 @@
 
 #ifdef HAVE_LIBSASL2
 extern "C" {
-#include <sasl/sasl.h>
+#include <sasl2/sasl.h>
 }
 #endif
 
diff -Naur kdepim-3.5.0-orig/kioslaves/sieve/sieve.cpp kdepim-3.5.0/kioslaves/sieve/sieve.cpp
--- kdepim-3.5.0-orig/kioslaves/sieve/sieve.cpp	2005-09-10 11:24:05.000000000 +0300
+++ kdepim-3.5.0/kioslaves/sieve/sieve.cpp	2005-12-25 12:53:06.905519250 +0200
@@ -29,7 +29,7 @@
 #endif
 
 extern "C" {
-#include <sasl/sasl.h>
+#include <sasl2/sasl.h>
 }
 #include "sieve.h"
 
