--- db-3.3.11/dist/Makefile.in.orig	2003-04-25 20:52:21.000000000 +0200
+++ db-3.3.11/dist/Makefile.in	2003-04-25 20:53:06.000000000 +0200
@@ -401,7 +401,6 @@
 	@cd $(libdir) && $(rm) -f \
 	    $(libso) $(libso_target) $(libso_default) $(libso_major)
 	@$(INSTALLER) $(libso_target) $(libdir)
-	@cd $(libdir) && $(ln) -s $(libso) $(libso_default)
 	@cd $(libdir) && $(ln) -s $(libso) $(libso_major)
 	@$(LIBTOOL) --mode=finish $(libdir)
 	@cd $(libdir) && $(chmod) $(fmode) $(libso) $(libso_target)
@@ -421,7 +420,6 @@
 	@cd $(libdir) && $(rm) -f \
 	    $(libxso) $(libxso_target) $(libxso_default) $(libxso_major)
 	@$(INSTALLER) $(libxso_target) $(libdir)
-	@cd $(libdir) && $(ln) -s $(libxso) $(libxso_default)
 	@cd $(libdir) && $(ln) -s $(libxso) $(libxso_major)
 	@cd $(libdir) && $(chmod) $(fmode) $(libxso) $(libxso_target)
 
