--- ../j2re-1_4_2_08-linux-i586.bin	2005-04-14 16:25:21.000000000 +0200
+++ j2re-1_4_2_08-linux-i586.bin	2005-04-14 18:41:30.000000000 +0200
@@ -11,7 +11,7 @@
 localinstall=$1
 diskSpaceRequired=53732
 MIME_TYPE=application/x-java-jnlp-file
-more <<"EOF"
+cat <<"EOF"
              Sun Microsystems, Inc.
           Binary Code License Agreement
                   for the
@@ -307,7 +307,7 @@
 Network Circle, Santa Clara, California 95054, U.S.A.
 (LFI#129530/Form ID#011801)
 EOF
-agreed=
+agreed=1
 while [ x$agreed = x ]; do
     echo
     echo "Do you agree to the above license terms? [yes or no] "
@@ -329,7 +329,7 @@
 fi
 trap 'rm -f $outname; exit 1' HUP INT QUIT TERM
 echo "Unpacking..."
-tail +477 $0 > $outname
+tail -n +477 $0 > $outname
 if [ -x /usr/bin/sum ] ; then
     echo "Checksumming..."
 
