--- grub-0.97/stage2/char_io.c.orig	2007-09-10 15:43:53.000000000 -0700
+++ grub-0.97/stage2/char_io.c	2007-09-10 15:45:10.000000000 -0700
@@ -1175,11 +1175,11 @@
 #endif /* ! STAGE1_5 */
 
 int
-memcheck (int addr, int len)
+memcheck (unsigned long addr, unsigned long len)
 {
 #ifdef GRUB_UTIL
   auto int start_addr (void);
-  auto int end_addr (void);
+  auto unsigned long end_addr (void);
   
   auto int start_addr (void)
     {
@@ -1192,7 +1192,7 @@
       return ret;
     }
 
-  auto int end_addr (void)
+  auto unsigned long end_addr (void)
     {
       int ret;
 # if defined(HAVE_END_SYMBOL)
--- grub-0.97/stage2/shared.h.orig	2004-06-19 09:40:09.000000000 -0700
+++ grub-0.97/stage2/shared.h	2007-09-10 15:39:04.000000000 -0700
@@ -911,7 +911,7 @@
 int nul_terminate (char *str);
 int get_based_digit (int c, int base);
 int safe_parse_maxint (char **str_ptr, int *myint_ptr);
-int memcheck (int start, int len);
+int memcheck (unsigned long start, unsigned long len);
 void grub_putstr (const char *str);
 
 #ifndef NO_DECOMPRESSION
