--- g/stage2/stage2.c.orig	2005-03-19 18:51:57.000000000 +0100
+++ g/stage2/stage2.c	2005-05-23 23:59:20.000000000 +0200
@@ -97,7 +97,7 @@
 static void
 print_entry (int y, int highlight, char *entry)
 {
-  int x;
+  int x; int t = 0;
 
   if (current_term->setcolorstate)
     current_term->setcolorstate (COLOR_STATE_NORMAL);
@@ -116,10 +116,14 @@
 	  else
 	    grub_putchar (*entry++);
 	}
-      else
+      else {
 	grub_putchar (' ');
+	if (!t && current_term->setcolorstate)
+          current_term->setcolorstate (COLOR_STATE_NORMAL);
+	t = 1;
+      }
     }
-  gotoxy (74, y);
+  gotoxy (2, y);
 
   if (current_term->setcolorstate)
     current_term->setcolorstate (COLOR_STATE_STANDARD);
@@ -158,7 +162,7 @@
   else
     grub_putchar (' ');
 
-  gotoxy (74, y + entryno + 1);
+  gotoxy (2, y + entryno + 1);
 }
 
 static void
@@ -302,7 +307,8 @@
       if (current_term->flags & TERM_DUMB)
 	print_entries_raw (num_entries, first_entry, menu_entries);
       else
-	print_border (3, 12);
+	/*print_border (3, 12);*/
+	gotoxy (74, 3 + 12);
 
       grub_printf ("\n\
       Use the %c and %c keys to select which entry is highlighted.\n",
@@ -361,7 +367,7 @@
 	      gotoxy (3, 22);
 	      grub_printf ("The highlighted entry will be booted automatically in %d seconds.    ",
 			   grub_timeout);
-	      gotoxy (74, 4 + entryno);
+	      gotoxy (2, 4);
 	  }
 	  
 	  grub_timeout--;
