--- ./stage2/fsys_reiserfs.c.orig	2003-06-24 23:23:26.012894384 +0200
+++ ./stage2/fsys_reiserfs.c	2003-06-24 23:24:04.525039648 +0200
@@ -112,7 +112,7 @@
   /* offset in the log of where to start replay after a crash */
   __u32 j_first_unflushed_offset;
   /* mount id to detect very old transactions */
-  __u32 long j_mount_id;
+  __u32 j_mount_id;
 };
 
 /* magic string to find desc blocks in the journal */
