Description: Specify format strings where appropriate
Author: Stephen Kitt <steve@sk2.org>

Index: gdmap/src/main.c
===================================================================
--- gdmap.orig/src/main.c	2011-03-13 14:32:57.964160454 +0100
+++ gdmap/src/main.c	2013-06-22 17:22:24.000000000 +0200
@@ -54,7 +54,7 @@
   g_option_context_set_help_enabled(context, TRUE);
   g_option_context_add_main_entries(context, Options, NULL);
   if (!g_option_context_parse(context, &argc, &argv, &error)) {
-    g_warning(error->message);
+    g_warning("%s", error->message);
     g_option_context_free(context);
     return 0;
   }
Index: gdmap/src/preferences.c
===================================================================
--- gdmap.orig/src/preferences.c	2013-06-22 16:53:02.000000000 +0200
+++ gdmap/src/preferences.c	2013-06-22 17:23:01.110547024 +0200
@@ -660,7 +660,7 @@
     {
         if (error->code != G_FILE_ERROR_NOENT) {
 	    // The settings file is allowed to not exist
-            g_message(error->message);
+	    g_message("%s", error->message);
         }
         g_error_free(error);
         return;
