--- gtoaster/Makefile.in	2003-05-07 23:08:44.749331280 +0200
+++ gtoaster-fixed/Makefile.in	2003-05-07 23:35:00.910718328 +0200
@@ -184,7 +184,7 @@
 
 Makefile: $(srcdir)/Makefile.in  $(top_builddir)/config.status $(BUILT_SOURCES)
 	cd $(top_builddir) \
-	  && CONFIG_FILES=$@ CONFIG_HEADERS= $(SHELL) ./config.status
+	  && echo CONFIG_FILES=$@ CONFIG_HEADERS= $(SHELL) ./config.status
 
 $(ACLOCAL_M4):  configure.in  m4/codeset.m4 m4/gettext.m4 m4/glibc21.m4 \
 		m4/iconv.m4 m4/intdiv0.m4 m4/inttypes-pri.m4 \
