--- epiphany-1.2.6/help/C/Makefile.in.orig	2004-06-17 11:48:38.721481152 +0200
+++ epiphany-1.2.6/help/C/Makefile.in	2004-06-17 11:49:14.052110080 +0200
@@ -458,7 +458,7 @@
 install-data-hook-omf:
 	$(mkinstalldirs) $(DESTDIR)$(omf_dest_dir)
 	for file in $(omffile); do \
-		$(INSTALL_DATA) $$file.out $(DESTDIR)$(omf_dest_dir)/$$file; \
+		$(INSTALL_DATA) $$file $(DESTDIR)$(omf_dest_dir)/$$file; \
 	done
 	-scrollkeeper-update -p $(scrollkeeper_localstate_dir) -o $(DESTDIR)$(omf_dest_dir)
 
--- epiphany-1.2.6/help/es/Makefile.in.orig	2004-06-17 12:10:24.211016576 +0200
+++ epiphany-1.2.6/help/es/Makefile.in	2004-06-17 12:10:39.259728824 +0200
@@ -458,7 +458,7 @@
 install-data-hook-omf:
 	$(mkinstalldirs) $(DESTDIR)$(omf_dest_dir)
 	for file in $(omffile); do \
-		$(INSTALL_DATA) $$file.out $(DESTDIR)$(omf_dest_dir)/$$file; \
+		$(INSTALL_DATA) $$file $(DESTDIR)$(omf_dest_dir)/$$file; \
 	done
 	-scrollkeeper-update -p $(scrollkeeper_localstate_dir) -o $(DESTDIR)$(omf_dest_dir)
 
--- epiphany-1.2.6/help/eu/Makefile.in.orig	2004-06-17 12:11:07.968364448 +0200
+++ epiphany-1.2.6/help/eu/Makefile.in	2004-06-17 12:11:13.382541368 +0200
@@ -458,7 +458,7 @@
 install-data-hook-omf:
 	$(mkinstalldirs) $(DESTDIR)$(omf_dest_dir)
 	for file in $(omffile); do \
-		$(INSTALL_DATA) $$file.out $(DESTDIR)$(omf_dest_dir)/$$file; \
+		$(INSTALL_DATA) $$file $(DESTDIR)$(omf_dest_dir)/$$file; \
 	done
 	-scrollkeeper-update -p $(scrollkeeper_localstate_dir) -o $(DESTDIR)$(omf_dest_dir)
 
