--- ./dist/Makefile.in.orig	2005-05-31 15:40:50.195829504 -0400
+++ ./dist/Makefile.in	2005-05-31 16:15:59.974004664 -0400
@@ -858,15 +858,8 @@
 	@test -d $(DESTDIR)$(libdir) || \
 	    ($(mkdir) -p $(DESTDIR)$(libdir) && \
 	    $(chmod) $(dmode) $(DESTDIR)$(libdir))
-	@cd $(DESTDIR)$(libdir) && $(rm) -f $(LIB_INSTALL_FILE_LIST)
-	@$(INSTALLER) @INSTALL_LIBS@ $(DESTDIR)$(libdir)
+	@$(INSTALLER) $(libjso_target) $(DESTDIR)$(libdir)
 	@(cd $(DESTDIR)$(libdir) && \
-	    test -f $(libso) && $(ln) -s $(libso) $(libso_default); \
-	    test -f $(libso) && $(ln) -s $(libso) $(libso_major); \
-	    test -f $(libxso) && $(ln) -s $(libxso) $(libxso_default); \
-	    test -f $(libxso) && $(ln) -s $(libxso) $(libxso_major); \
-	    test -f $(libtso) && $(ln) -s $(libtso) $(libtso_default); \
-	    test -f $(libtso) && $(ln) -s $(libtso) $(libtso_major); \
 	    test -f $(libjso) && $(ln) -s $(libjso) $(libjso_default); \
 	    test -f $(libjso) && $(ln) -s $(libjso) $(libjso_major); \
 	    test -f $(libjso) && $(ln) -s $(libjso) $(libjso_g)) || true
