--- ./dist/Makefile.in.orig	2005-05-20 21:53:14.000000000 -0400
+++ ./dist/Makefile.in	2005-05-20 22:27:22.245343240 -0400
@@ -11,6 +11,7 @@
 bindir=	@bindir@
 includedir=@includedir@
 libdir=	@libdir@
+libjdir=@libdir@
 docdir=	$(prefix)/docs
 
 dmode=	755
@@ -859,20 +860,24 @@
 	    ($(mkdir) -p $(DESTDIR)$(libdir) && \
 	    $(chmod) $(dmode) $(DESTDIR)$(libdir))
 	@cd $(DESTDIR)$(libdir) && $(rm) -f $(LIB_INSTALL_FILE_LIST)
-	@$(INSTALLER) @INSTALL_LIBS@ $(DESTDIR)$(libdir)
+	@for i in @INSTALL_LIBS@; do \
+		[[ $$i = *java* ]] \
+		&& $(INSTALLER) $$i $(DESTDIR)$(libjdir) \
+		|| $(INSTALLER) $$i $(DESTDIR)$(libdir); done
 	@(cd $(DESTDIR)$(libdir) && \
 	    test -f $(libso) && $(ln) -s $(libso) $(libso_default); \
 	    test -f $(libso) && $(ln) -s $(libso) $(libso_major); \
 	    test -f $(libxso) && $(ln) -s $(libxso) $(libxso_default); \
 	    test -f $(libxso) && $(ln) -s $(libxso) $(libxso_major); \
 	    test -f $(libtso) && $(ln) -s $(libtso) $(libtso_default); \
-	    test -f $(libtso) && $(ln) -s $(libtso) $(libtso_major); \
+	    test -f $(libtso) && $(ln) -s $(libtso) $(libtso_major)) || true
+	@(cd $(DESTDIR)$(libjdir) && \
 	    test -f $(libjso) && $(ln) -s $(libjso) $(libjso_default); \
 	    test -f $(libjso) && $(ln) -s $(libjso) $(libjso_major); \
 	    test -f $(libjso) && $(ln) -s $(libjso) $(libjso_g)) || true
 	@(test -f $(libj_jarfile) && \
-	    $(cp) $(libj_jarfile) $(DESTDIR)$(libdir) && \
-	    $(chmod) $(fmode) $(DESTDIR)$(libdir)/$(libj_jarfile)) || true
+	    $(cp) $(libj_jarfile) $(DESTDIR)$(libjdir) && \
+	    $(chmod) $(fmode) $(DESTDIR)$(libjdir)/$(libj_jarfile)) || true
 
 uninstall_lib:
 	@cd $(DESTDIR)$(libdir) && $(rm) -f $(LIB_INSTALL_FILE_LIST)
