--- ./config/cf/Imake.rules.orig	2003-03-23 11:44:47.000000000 +0100
+++ ./config/cf/Imake.rules	2003-03-23 11:50:43.000000000 +0100
@@ -1693,8 +1693,9 @@
 		(cd Concat($(DESTDIR),npath/mdir); \			@@\
 		 tar xf -; exit 0); exit 0); \				@@\
 	  fi; \								@@\
-	  $(RM) -r rdir; \						@@\
-	  $(LN) Concat($${RELPATH},npath/mdir) ldir; \			@@\
+	  if [ ! -e rdir ]; then \					@@\
+	    $(LN) Concat($${RELPATH},npath/mdir) ldir; \		@@\
+	  fi; \								@@\
 	fi
 #else
 #define LinkConfDirectoryLong(mdir,cdir,rdir,ldir,opath,npath)
