--- ucspi-tcp-0.88.old/Makefile	2000-03-18 16:18:42.000000000 +0100
+++ ucspi-tcp-0.88/Makefile	2005-02-23 14:17:46.704390744 +0100
@@ -38,7 +38,7 @@
 
 auto_home.c: \
 auto-str conf-home
-	./auto-str auto_home `head -1 conf-home` > auto_home.c
+	./auto-str auto_home `head -n 1 conf-home` > auto_home.c
 
 auto_home.o: \
 compile auto_home.c
@@ -147,7 +147,7 @@
 choose: \
 warn-auto.sh choose.sh conf-home
 	cat warn-auto.sh choose.sh \
-	| sed s}HOME}"`head -1 conf-home`"}g \
+	| sed s}HOME}"`head -n 1 conf-home`"}g \
 	> choose
 	chmod 755 choose
 
@@ -159,14 +159,14 @@
 compile: \
 warn-auto.sh conf-cc
 	( cat warn-auto.sh; \
-	echo exec "`head -1 conf-cc`" '-c $${1+"$$@"}' \
+	echo exec "`head -n 1 conf-cc`" '-c $${1+"$$@"}' \
 	) > compile
 	chmod 755 compile
 
 date@: \
 warn-auto.sh date@.sh conf-home
 	cat warn-auto.sh date@.sh \
-	| sed s}HOME}"`head -1 conf-home`"}g \
+	| sed s}HOME}"`head -n 1 conf-home`"}g \
 	> date@
 	chmod 755 date@
 
@@ -288,7 +288,7 @@
 finger@: \
 warn-auto.sh finger@.sh conf-home
 	cat warn-auto.sh finger@.sh \
-	| sed s}HOME}"`head -1 conf-home`"}g \
+	| sed s}HOME}"`head -n 1 conf-home`"}g \
 	> finger@
 	chmod 755 finger@
 
@@ -342,7 +342,7 @@
 http@: \
 warn-auto.sh http@.sh conf-home
 	cat warn-auto.sh http@.sh \
-	| sed s}HOME}"`head -1 conf-home`"}g \
+	| sed s}HOME}"`head -n 1 conf-home`"}g \
 	> http@
 	chmod 755 http@
 
@@ -385,7 +385,7 @@
 warn-auto.sh conf-ld
 	( cat warn-auto.sh; \
 	echo 'main="$$1"; shift'; \
-	echo exec "`head -1 conf-ld`" \
+	echo exec "`head -n 1 conf-ld`" \
 	'-o "$$main" "$$main".o $${1+"$$@"}' \
 	) > load
 	chmod 755 load
@@ -412,7 +412,7 @@
 mconnect: \
 warn-auto.sh mconnect.sh conf-home
 	cat warn-auto.sh mconnect.sh \
-	| sed s}HOME}"`head -1 conf-home`"}g \
+	| sed s}HOME}"`head -n 1 conf-home`"}g \
 	> mconnect
 	chmod 755 mconnect
 
@@ -504,7 +504,7 @@
 rts: \
 warn-auto.sh rts.sh conf-home
 	cat warn-auto.sh rts.sh \
-	| sed s}HOME}"`head -1 conf-home`"}g \
+	| sed s}HOME}"`head -n 1 conf-home`"}g \
 	> rts
 	chmod 755 rts
 
@@ -661,8 +661,8 @@
 systype: \
 find-systype.sh conf-cc conf-ld trycpp.c x86cpuid.c
 	( cat warn-auto.sh; \
-	echo CC=\'`head -1 conf-cc`\'; \
-	echo LD=\'`head -1 conf-ld`\'; \
+	echo CC=\'`head -n 1 conf-cc`\'; \
+	echo LD=\'`head -n 1 conf-ld`\'; \
 	cat find-systype.sh; \
 	) | sh > systype
 
@@ -705,7 +705,7 @@
 tcpcat: \
 warn-auto.sh tcpcat.sh conf-home
 	cat warn-auto.sh tcpcat.sh \
-	| sed s}HOME}"`head -1 conf-home`"}g \
+	| sed s}HOME}"`head -n 1 conf-home`"}g \
 	> tcpcat
 	chmod 755 tcpcat
 
@@ -832,6 +832,6 @@
 who@: \
 warn-auto.sh who@.sh conf-home
 	cat warn-auto.sh who@.sh \
-	| sed s}HOME}"`head -1 conf-home`"}g \
+	| sed s}HOME}"`head -n 1 conf-home`"}g \
 	> who@
 	chmod 755 who@
