--- dxr3-0.2.2/Makefile.orig	2004-01-04 18:32:10.000000000 +0100
+++ dxr3-0.2.2/Makefile	2005-04-07 13:28:49.000000000 +0200
@@ -42,8 +42,8 @@
 
 ### Includes and Defines (add further entries here):
 
-INCLUDES += -I$(VDRDIR)/include -I$(DVBDIR)/include -I$(FFMDIR)
-LIBS     = -L$(FFMDIR)/libavcodec -lavcodec
+INCLUDES += -I$(VDRDIR)/include -I$(DVBDIR)/include -I/usr/include/ffmpeg
+LIBS     = -lavcodec
 DEFINES += -DPLUGIN_NAME_I18N='"$(PLUGIN)"'
 DEFINES += -DSOCKET_CHMOD=0660
 DEFINES += -D_GNU_SOURCE
--- dxr3-0.2.2/dxr3audiodecoder.h.orig	2005-04-07 13:51:42.000000000 +0200
+++ dxr3-0.2.2/dxr3audiodecoder.h	2005-04-07 13:51:58.000000000 +0200
@@ -10,7 +10,7 @@
 #define __DXR3AUDIODECODER_H
 
 extern "C" {
-#include "libavcodec/avcodec.h"
+#include "ffmpeg/avcodec.h"
 };
 
 #include <stdlib.h>
