--- image-0.2.3/libimage/pnm.h.orig	2005-02-14 18:01:00.464704296 +0100
+++ image-0.2.3/libimage/pnm.h	2005-02-14 18:01:11.924962072 +0100
@@ -151,7 +151,7 @@
   Get next significant character, i.e. jump over comments
   @param file - source file
   @param - Next significant character*/
-  bool getchar(FILE* f,char& nNext);
+  bool getchar_(FILE* f,char& nNext);
 
   /*****************************************************************************
   Get next bit from file
--- image-0.2.3/libimage/pnm.c.orig	2005-02-14 18:01:04.294122136 +0100
+++ image-0.2.3/libimage/pnm.c	2005-02-14 18:01:30.060205096 +0100
@@ -73,7 +73,7 @@
 Get next significant character, i.e. jump over comments
 @param file - source file
 @param - Next significant character*/
-bool cPNM::getchar(FILE* f, char& nNext)
+bool cPNM::getchar_(FILE* f, char& nNext)
 {
   register int ich;
   register char ch;
@@ -112,7 +112,7 @@
   register char ch;
   
   do {
-    if(!getchar( f, ch ))
+    if(!getchar_( f, ch ))
       return false;
   } while ( ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r' );
   
@@ -156,7 +156,7 @@
   register T i;
   
   do {
-    if(!getchar( f,ch ))
+    if(!getchar_( f,ch ))
        return false;
   } while ( ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r' );
   
@@ -168,7 +168,7 @@
   i = 0;
   do {
     i = i * 10 + ch - '0';
-    if(!getchar( f,ch ))
+    if(!getchar_( f,ch ))
       return false;
   } while ( ch >= '0' && ch <= '9' );
   nNext = i;
