--- ./mp3/decoder-ogg.h~	2005-02-06 13:42:26.000000000 +0200
+++ ./mp3/decoder-ogg.h	2006-01-08 19:08:05.000000000 +0200
@@ -49,7 +49,7 @@
   ~cOggFile();
   bool Open(bool log=true);
   void Close(void);
-  long long cOggFile::Seek(long long posMs=0, bool relativ=false);
+  long long Seek(long long posMs=0, bool relativ=false);
   int Stream(short *buffer, int samples);
   bool CanSeek(void) { return canSeek; }
   long long IndexMs(void);
