diff -ru softdevice/Makefile softdevice.work/Makefile
--- softdevice/Makefile	2005-03-04 21:04:20.000000000 +0100
+++ softdevice.work/Makefile	2005-04-10 12:42:42.000000000 +0200
@@ -51,7 +51,7 @@
 #   install -m 644 postprocess.h /usr/local/include/postproc/postprocess.h
 #   cp libpostproc.a /usr/local/lib/
 #
-PP_LIBAVCODEC=1
+#PP_LIBAVCODEC=1
 
 # -----------------------------------------------------------------------------
 # if you want output methods build as a single lib comment the following line
@@ -60,10 +60,10 @@
 # -----------------------------------------------------------------------------
 # framebuffer device name
 #
-#FBDEV = /dev/fb/0
+FBDEV = /dev/fb/0
 #FBDEV = /dev/fb0
 
-FBDEV = /dev/fb0
+#FBDEV = /dev/fb0
 
 # -----------------------------------------------------------------------------
 # VIDIX specific definitions
@@ -72,7 +72,7 @@
 #VIDIX_DIR    = /opt/vidix
 #VIDIX_DIR    = /usr/local/
 
-VIDIX_DIR    = /usr/local
+VIDIX_DIR    = /usr
 
 # -----------------------------------------------------------------------------
 # Set up these paths to your enviroment!
@@ -81,7 +81,7 @@
 #LIBAVCODEC = ../../../../ffmpeg/libavcodec
 #
 # If you installed ffmpeg's lib (make installlib)
-LIBAVCODEC=/usr/local/include/ffmpeg/
+LIBAVCODEC=/usr/include/ffmpeg/
 #LIBAVCODEC= /home/extra/src/video/ffmpeg-0.4.8/libavcodec
 #
 #
@@ -145,7 +145,7 @@
 INCLUDES += -I$(VDRDIR)/include -I$(DVBDIR)/include -I$(LIBAVCODEC)
 
 DEFINES += -DPLUGIN_NAME_I18N='"$(PLUGIN)"' -D_GNU_SOURCE
-DEFINES += -DPLUGINLIBDIR='"./PLUGINS/lib"'
+DEFINES += -DPLUGINLIBDIR='"/usr/lib/vdr"'
 
 ### The object files (add further files here):
 
@@ -160,16 +160,16 @@
 
 ifdef DFB_SUPPORT
   ifdef USE_SUBPLUGINS
-    DFB_LIBS  = -L/usr/local/lib -ldfb++
+    DFB_LIBS  = -ldfb++
     DFB_OBJS  = video.o video-dfb.o
     ALL_OBJS += $(DFB_OBJS)
     TARGETS  += libvdr-$(PLUGIN)-dfb.so
   else
     OBJS     += video-dfb.o
-    LIBS     += -L/usr/local/lib -ldfb++
+    LIBS     += -ldfb++
   endif
   DEFINES  += -DDFB_SUPPORT
-  INCLUDES += -I/usr/local/include/dfb++ -I/usr/local/include/directfb
+  INCLUDES += -I/usr/include/dfb++ -I/usr/include/directfb
 endif
 
 ifdef FB_SUPPORT
