diff -Naur plptools-0.12.orig/include/plp_inttypes.h plptools-0.12/include/plp_inttypes.h
--- plptools-0.12.orig/include/plp_inttypes.h	2002-02-14 20:56:41.000000000 +0200
+++ plptools-0.12/include/plp_inttypes.h	2005-03-24 20:59:34.822833400 +0200
@@ -27,7 +27,11 @@
 typedef unsigned char      u_int8_t;
 typedef unsigned short     u_int16_t;
 typedef unsigned int       u_int32_t;
+#ifdef __x86_64__
+typedef long unsigned int u_int64_t;
+#else
 typedef unsigned long long u_int64_t;
+#endif
 typedef signed long long   s_int64_t;
 #endif // ISO_INTTYPES
 #endif // GNU_INTTYPES
