--- luminocity/autogen.sh.vanilla	2005-04-18 00:09:30.000000000 +0200
+++ luminocity/autogen.sh	2005-04-18 00:09:39.000000000 +0200
@@ -20,9 +20,9 @@
 	DIE=1
 }
 
-if automake-1.7 --version < /dev/null > /dev/null 2>&1 ; then
-    AUTOMAKE=automake-1.7
-    ACLOCAL=aclocal-1.7
+if automake --version < /dev/null > /dev/null 2>&1 ; then
+    AUTOMAKE=automake
+    ACLOCAL=aclocal
 else
 	echo
 	echo "You must have automake 1.7.x installed to compile $PROJECT."
