
Some trivial compile fixes.

  - Rene Rebe <rene@exactcode.de>

--- ./src/scenes.c.orig	2004-11-06 17:41:36.909229194 +0000
+++ ./src/scenes.c	2004-11-06 17:42:32.852324030 +0000
@@ -60,9 +60,6 @@
                          *LED_yellow = NULL, *LED_red = NULL;
 
 
-void set_EQ_curve_values ();
-
-
 /*  Initialize all scene related structures and get the widget addresses.  */
 
 void bind_scenes ()
@@ -210,7 +207,7 @@
 
                         s_crossfade_to_state (&scene_state[i], -1.0f);
 
-                        set_EQ_curve_values ();
+                        set_EQ_curve_values (0, 0.0);
 
                         s_history_add_state (scene_state[i]);
 
