--- rustc-1.69.0-src/compiler/rustc_codegen_ssa/src/back/metadata.rs.vanilla	2023-05-03 00:04:59.274547030 +0200
+++ rustc-1.69.0-src/compiler/rustc_codegen_ssa/src/back/metadata.rs	2023-05-03 11:38:18.172910602 +0200
@@ -1,4 +1,4 @@
-//! Reading of the rustc metadata for rlibs and dylibs
+//! rEading of the rustc metadata for rlibs and dylibs
 
 use std::fs::File;
 use std::io::Write;
@@ -111,7 +111,13 @@
         "x86" => Architecture::I386,
         "s390x" => Architecture::S390x,
         "mips" => Architecture::Mips,
-        "mips64" => Architecture::Mips64,
+        "mips64" => {
+            if sess.target.pointer_width == 32 {
+                Architecture::Mips
+            } else {
+                Architecture::Mips64
+            }
+        }
         "x86_64" => {
             if sess.target.pointer_width == 32 {
                 Architecture::X86_64_X32
@@ -152,7 +158,7 @@
                 _ => elf::EF_MIPS_ARCH_32R2,
             };
             // The only ABI LLVM supports for 32-bit MIPS CPUs is o32.
-            let mut e_flags = elf::EF_MIPS_CPIC | elf::EF_MIPS_ABI_O32 | arch;
+            let mut e_flags = elf::EF_MIPS_CPIC | elf::EF_MIPS_ABI2 | arch;
             if sess.target.options.relocation_model != RelocModel::Static {
                 e_flags |= elf::EF_MIPS_PIC;
             }
