--- rustc-1.48.0-src/compiler/rustc_target/src/spec/mod.rs.vanilla	2020-12-25 18:38:24.276037477 +0100
+++ rustc-1.48.0-src/compiler/rustc_target/src/spec/mod.rs	2020-12-25 18:39:59.920035379 +0100
@@ -494,6 +494,7 @@
     ("x86_64-unknown-linux-gnux32", x86_64_unknown_linux_gnux32),
     ("i686-unknown-linux-gnu", i686_unknown_linux_gnu),
     ("i586-unknown-linux-gnu", i586_unknown_linux_gnu),
+    ("i486-unknown-linux-gnu", i486_unknown_linux_gnu),
     ("mips-unknown-linux-gnu", mips_unknown_linux_gnu),
     ("mips64-unknown-linux-gnuabi64", mips64_unknown_linux_gnuabi64),
     ("mips64el-unknown-linux-gnuabi64", mips64el_unknown_linux_gnuabi64),
--- rustc-1.48.0-src/compiler/rustc_target/src/spec/i486_unknown_linux_gnu.rs.vanilla	2020-12-25 18:40:42.916034436 +0100
+++ rustc-1.48.0-src/compiler/rustc_target/src/spec/i486_unknown_linux_gnu.rs	2020-12-25 18:41:12.220033793 +0100
@@ -0,0 +1,8 @@
+use crate::spec::TargetResult;
+
+pub fn target() -> TargetResult {
+    let mut base = super::i686_unknown_linux_gnu::target()?;
+    base.options.cpu = "pentium".to_string();
+    base.llvm_target = "i486-unknown-linux-gnu".to_string();
+    Ok(base)
+}
