--- rustc-1.60.0-src/compiler/rustc_target/src/spec/mips64_unknown_linux_gnuabi64.rs.vanilla	2022-06-16 14:25:42.309506167 +0200
+++ rustc-1.60.0-src/compiler/rustc_target/src/spec/mips64_unknown_linux_gnuabi64.rs	2022-06-16 14:27:26.132499978 +0200
@@ -11,8 +11,8 @@
             abi: "abi64".to_string(),
             endian: Endian::Big,
             // NOTE(mips64r2) matches C toolchain
-            cpu: "mips64r2".to_string(),
-            features: "+mips64r2".to_string(),
+            cpu: "mips4".to_string(),
+            features: "".to_string(),
             max_atomic_width: Some(64),
             mcount: "_mcount".to_string(),
 
