Index: ./Sources/API/Display/Input/key.h
===================================================================
RCS file: /var/lib/cvs/Libs/ClanLib-0.6/Sources/API/Display/Input/key.h,v
retrieving revision 1.11
diff -u -r1.11 key.h
--- ./Sources/API/Display/Input/key.h	18 Oct 2001 09:23:04 -0000	1.11
+++ ./Sources/API/Display/Input/key.h	1 Apr 2003 11:35:29 -0000
@@ -40,7 +40,7 @@
 	CL_KEY_INSERT, CL_KEY_DELETE, CL_KEY_HOME, CL_KEY_END, CL_KEY_PAGEUP,
 	CL_KEY_PAGEDOWN, CL_KEY_CAPSLOCK, CL_KEY_NUMLOCK, CL_KEY_SCRLOCK,
 	CL_KEY_PRINT, CL_KEY_PAUSE, CL_KEY_KP_DIV, CL_KEY_KP_MULT,
-	CL_KEY_KP_MINUS, CL_KEY_KP_PLUS, CL_KEY_KP_ENTER,
+	CL_KEY_KP_MINUS, CL_KEY_KP_PLUS, CL_KEY_KP_ENTER, CL_KEY_GRAVE,
 	
 	CL_KEY_KP_DECIMAL, 
 	CL_KEY_KP_0, CL_KEY_KP_1, CL_KEY_KP_2, CL_KEY_KP_3, CL_KEY_KP_4, 
Index: ./Sources/Display/Input/X11/keyboard_x11.cpp
===================================================================
RCS file: /var/lib/cvs/Libs/ClanLib-0.6/Sources/Display/Input/X11/keyboard_x11.cpp,v
retrieving revision 1.18
diff -u -r1.18 keyboard_x11.cpp
--- ./Sources/Display/Input/X11/keyboard_x11.cpp	2 Mar 2002 16:43:48 -0000	1.18
+++ ./Sources/Display/Input/X11/keyboard_x11.cpp	1 Apr 2003 11:35:41 -0000
@@ -197,6 +197,7 @@
 //	case XK_Clear: return CL_KEY_CLEAR;
 	case XK_Return: return CL_KEY_ENTER;
 	case XK_Pause: return CL_KEY_PAUSE;
+        case XK_grave: return CL_KEY_GRAVE;
 //	case XK_Scroll_Lock: return CL_KEY_SCROLL;
 	case XK_Escape: return CL_KEY_ESCAPE;
 	case XK_Delete: return CL_KEY_DELETE;
@@ -376,6 +377,7 @@
 	case CL_KEY_SCRLOCK: return 78;
 	case CL_KEY_PRINT: return 111;
 	case CL_KEY_PAUSE: return 110;
+	case CL_KEY_GRAVE: return XK_grave;
 	case CL_KEY_KP_DIV: return 112;
 	case CL_KEY_KP_MULT: return 63;
 	case CL_KEY_KP_MINUS: return 82;
