Index: ./Sources/GUI/window_default.cpp
===================================================================
RCS file: /usr/local/cvsroot/Libs/ClanLib-0.6/Sources/GUI/window_default.cpp,v
retrieving revision 1.31
diff -u -r1.31 window_default.cpp
--- ./Sources/GUI/window_default.cpp	16 Jan 2002 19:50:41 -0000	1.31
+++ ./Sources/GUI/window_default.cpp	24 Oct 2002 13:03:10 -0000
@@ -42,15 +42,15 @@
 */
 	// Create a close button on the titlebar
 	CL_Rect rect(window->get_width() - 19, 3, window->get_width() - 3, 19);
-	button_close = new CL_Button(rect, "X", window, style);
+	//	button_close = new CL_Button(rect, "X", window, style);
 
-	slot_close = button_close->sig_clicked().connect(this, &CL_Window_Default::on_close);
+	//slot_close = button_close->sig_clicked().connect(this, &CL_Window_Default::on_close);
 	slot_paint = window->sig_paint().connect(this, &CL_Window_Default::on_paint);
 	slot_resize = window->sig_resize().connect(this, &CL_Window_Default::on_resize);
 
 	// Move close button on window resize
-	layout.add_resize_position(button_close, CL_LayoutManager::x1, window, CL_LayoutManager::x2);
-	layout.add_resize_position(button_close, CL_LayoutManager::x2, window, CL_LayoutManager::x2);
+	//	layout.add_resize_position(button_close, CL_LayoutManager::x1, window, CL_LayoutManager::x2);
+	//	layout.add_resize_position(button_close, CL_LayoutManager::x2, window, CL_LayoutManager::x2);
 
 	// Resize client area on window resize
 	CL_Component *client_area = window->get_client_area();
@@ -65,7 +65,7 @@
 	delete font_disabled;
 //	delete resize_handler;
 	delete move_handler;
-	delete button_close;
+	//	delete button_close;
 }
 
 void CL_Window_Default::on_paint()
Index: ./Sources/GUI/window_default.h
===================================================================
RCS file: /usr/local/cvsroot/Libs/ClanLib-0.6/Sources/GUI/window_default.h,v
retrieving revision 1.14
diff -u -r1.14 window_default.h
--- ./Sources/GUI/window_default.h	28 Dec 2001 22:41:50 -0000	1.14
+++ ./Sources/GUI/window_default.h	24 Oct 2002 13:03:11 -0000
@@ -40,7 +40,7 @@
 	void on_resize(int old_width, int old_height);
 
 	CL_Window *window;
-	CL_Button *button_close;
+	//	CL_Button *button_close;
 
 	CL_ComponentMoveHandler *move_handler;
 	CL_ComponentResizeHandler *resize_handler;
