diff -Naur dosfstools-2.9.orig/dosfsck/io.c dosfstools-2.9/dosfsck/io.c
--- dosfstools-2.9.orig/dosfsck/io.c	2000-11-28 18:25:15.000000000 +0200
+++ dosfstools-2.9/dosfsck/io.c	2005-03-10 18:57:13.853587504 +0200
@@ -42,7 +42,7 @@
 /* Use the _llseek system call directly, because there (once?) was a bug in
  * the glibc implementation of it. */
 #include <linux/unistd.h>
-#if defined(__alpha) || defined (__ia64__)
+#if defined(__alpha) || defined (__ia64__) || defined (__x86_64__)
 /* On alpha, the syscall is simply lseek, because it's a 64 bit system. */
 static loff_t llseek( int fd, loff_t offset, int whence )
 {
diff -Naur dosfstools-2.9.orig/mkdosfs/mkdosfs.c dosfstools-2.9/mkdosfs/mkdosfs.c
--- dosfstools-2.9.orig/mkdosfs/mkdosfs.c	2002-01-08 10:27:54.000000000 +0200
+++ dosfstools-2.9/mkdosfs/mkdosfs.c	2005-03-10 18:30:27.200835848 +0200
@@ -95,7 +95,7 @@
 /* Use the _llseek system call directly, because there (once?) was a bug in
  * the glibc implementation of it. */
 #include <linux/unistd.h>
-#if defined(__alpha) || defined(__ia64__)
+#if defined(__alpha) || defined(__ia64__) || defined(__x86_64__)
 /* On alpha, the syscall is simply lseek, because it's a 64 bit system. */
 static loff_t llseek( int fd, loff_t offset, int whence )
 {
