diff -urN ../clearsilver-0.9.13/configure ./configure
--- ../clearsilver-0.9.13/configure	2004-11-25 00:05:58.000000000 +0100
+++ ./configure	2005-01-17 13:42:44.412237920 +0100
@@ -3047,7 +3047,7 @@
 echo "configure:3048: checking for python includes" >&5
   python_inc=no
   python_search_path="/neo/opt /usr/local /usr /c"
-  python_versions="2.3 2.2 2.1 2.0 1.5 22 21 20 15"
+  python_versions="2.4 2.3 2.2 2.1 2.0 1.5 22 21 20 15"
   if test $cs_cv_python_path != "no" -a -x $cs_cv_python_path; then
     python_bin=$cs_cv_python_path
     vers=`$python_bin -c "import sys; print sys.version[:3]"`
@@ -3084,7 +3084,7 @@
     PYTHON=$python_bin
     PYTHON_INC="-I$python_inc"
     PYTHON_LIB=$python_lib
-    PYTHON_SITE=`$python_bin -c "import site; print site.sitedirs[0]"`
+    PYTHON_SITE=`$python_bin -c "import os,sys; print os.path.join(sys.prefix, 'lib','python'+sys.version[:3],'site-packages')"`
     BUILD_WRAPPERS="$BUILD_WRAPPERS python"
   fi
 fi
