diff -urN -x '*~' gstreamer-0.6.5/docs/faq/Makefile.in gstreamer-0.6.5-rock/docs/faq/Makefile.in
--- gstreamer-0.6.5/docs/faq/Makefile.in	2004-02-12 15:10:44.000000000 +0100
+++ gstreamer-0.6.5-rock/docs/faq/Makefile.in	2004-02-24 23:30:07.000000000 +0100
@@ -682,8 +682,10 @@
 	echo Done
 
 install-data-local: html
-	mkdir -p $(docdir)
-	cp -pr $(HTML_DAT) $(docdir)
+	if [ -n "$(HTML_DAT)" ]; then \
+		mkdir -p $(docdir); \
+		cp -pr $(HTML_DAT) $(docdir); \
+	fi
 
 uninstall-local:
 	for part in $(HTML_DAT); do rm -rf $(docdir)/$$part; done
diff -urN -x '*~' gstreamer-0.6.5/docs/manual/Makefile.in gstreamer-0.6.5-rock/docs/manual/Makefile.in
--- gstreamer-0.6.5/docs/manual/Makefile.in	2004-02-12 15:10:45.000000000 +0100
+++ gstreamer-0.6.5-rock/docs/manual/Makefile.in	2004-02-24 23:30:24.000000000 +0100
@@ -682,8 +682,10 @@
 	echo Done
 
 install-data-local: html
-	mkdir -p $(docdir)
-	cp -pr $(HTML_DAT) $(docdir)
+	if [ -n "$(HTML_DAT)" ]; then \
+		mkdir -p $(docdir); \
+		cp -pr $(HTML_DAT) $(docdir); \
+	fi
 
 uninstall-local:
 	for part in $(HTML_DAT); do rm -rf $(docdir)/$$part; done
diff -urN -x '*~' gstreamer-0.6.5/docs/pwg/Makefile.in gstreamer-0.6.5-rock/docs/pwg/Makefile.in
--- gstreamer-0.6.5/docs/pwg/Makefile.in	2004-02-12 15:10:45.000000000 +0100
+++ gstreamer-0.6.5-rock/docs/pwg/Makefile.in	2004-02-24 23:31:22.000000000 +0100
@@ -682,8 +682,10 @@
 	echo Done
 
 install-data-local: html
-	mkdir -p $(docdir)
-	cp -pr $(HTML_DAT) $(docdir)
+	if [ -n "$(HTML_DAT)" ]; then \
+		mkdir -p $(docdir); \
+		cp -pr $(HTML_DAT) $(docdir); \
+	fi
 
 uninstall-local:
 	for part in $(HTML_DAT); do rm -rf $(docdir)/$$part; done
