--- ./Makefile.orig	2003-04-29 09:46:06.000000000 +0200
+++ ./Makefile	2003-04-29 09:46:22.000000000 +0200
@@ -50,18 +50,6 @@
 	$(INS) -m 644 $(MANPAGE) $(DESTDIR)$(MANDIR)
 	gzip -9f $(DESTDIR)$(MANDIR)/$(MANPAGE)
 
-	$(INS) -m 755 $(RCFILE) $(DESTDIR)$(RCFILETO)/$(RCFILEFINAL)
-
-ifndef DESTDIR
-		chkconfig --add $(RCFILEFINAL)
-else
-		echo "MAKE: Skipping chkconfig operation (rpm build?)"
-endif
-
-device:
-	mkdir -p $(DESTDIR)/dev/cpu
-	mknod $(DESTDIR)/dev/cpu/microcode c 10 184
-
 uninstall:
 ifndef DESTDIR
 	chkconfig --del $(RCFILEFINAL)
