diff -Nur net-tools-1.60.orig/Makefile net-tools-1.60/Makefile
--- net-tools-1.60.orig/Makefile	2001-04-15 16:34:31.000000000 +0200
+++ net-tools-1.60/Makefile	2003-09-02 01:34:49.000000000 +0200
@@ -240,9 +240,6 @@
 	install -m 0755 mii-tool   $(BASEDIR)/sbin
 endif
 	ln -fs hostname $(BASEDIR)/bin/dnsdomainname
-	ln -fs hostname $(BASEDIR)/bin/ypdomainname
-	ln -fs hostname $(BASEDIR)/bin/nisdomainname
-	ln -fs hostname $(BASEDIR)/bin/domainname
 ifeq ($(HAVE_AFDECnet),1)
 	ln -fs hostname $(BASEDIR)/bin/nodename
 endif
