--- linux-2.4.27/drivers/char/keyboard.c.vanilla	2005-02-01 23:05:43.391427456 +0100
+++ linux-2.4.27/drivers/char/keyboard.c	2005-02-01 23:06:31.297144680 +0100
@@ -419,6 +419,7 @@
 	}
 	rep = 0;
 out:
+	;
 }
 
 /*
--- linux-2.4.27/include/asm-arm/hardware/psionw.h.vanilla	2005-02-01 23:04:22.173774416 +0100
+++ linux-2.4.27/include/asm-arm/hardware/psionw.h	2005-02-01 23:04:28.740776080 +0100
@@ -399,4 +399,4 @@
 #define COEOI		(0x0a0c)	/* Codec end of interrupt, write only */
 #define COTEST		(0x0a10)	/* Codec test register */
 
-#endif /* __ASM_HARDWARE_PSIONW_H */					\
+#endif /* __ASM_HARDWARE_PSIONW_H */
--- linux-2.4.27/arch/arm/mach-psionw/psionw-power.c.vanilla	2005-02-01 23:26:23.158954144 +0100
+++ linux-2.4.27/arch/arm/mach-psionw/psionw-power.c	2005-02-01 23:27:37.000000000 +0100
@@ -166,11 +166,11 @@
 
 		/* Now go to sleep */
 		psionw_writel(1, STBY);
-		__asm__ __volatile__("
-			mov     r0, r0
-			mov     r0, r0 
-			mov     r0, r0
-		");
+		__asm__ __volatile__("\
+			mov     r0, r0\n\
+			mov     r0, r0\n\
+			mov     r0, r0"
+		);
 
 		/* We're back from sleep */
 
--- linux-2.4.27/drivers/char/keyboard_psion.c.vanilla	2005-02-01 23:14:21.000000000 +0100
+++ linux-2.4.27/drivers/char/keyboard_psion.c	2005-02-01 23:14:43.000000000 +0100
@@ -54,6 +54,8 @@
     "\000" "\000" "\000" ","	"\000" " "    "\000";
 #endif
 
+void debug_gpio(void);
+
 void psion_cleartable(void) 
 {
 	int key;
--- linux-2.4.27/include/asm-arm/arch-psionw/system.h.vanilla	2005-02-01 23:28:08.178988672 +0100
+++ linux-2.4.27/include/asm-arm/arch-psionw/system.h	2005-02-01 23:28:31.000000000 +0100
@@ -29,11 +29,11 @@
 arch_idle(void)
 {
 	psionw_writel(1, HALT);
-	__asm__ __volatile__("
-		mov	r0, r0
-		mov	r0, r0 
-		mov	r0, r0 
-	");
+	__asm__ __volatile__("\
+		mov	r0, r0\n\
+		mov	r0, r0\n\
+		mov	r0, r0"
+	);
 }
 
 static inline void arch_reset(char mode)
