--- ./makefile.orig	2002-07-06 13:34:20.000000000 -0400
+++ ./makefile	2004-02-21 01:48:43.000000000 -0500
@@ -51,12 +51,10 @@
 	@echo ---------------------------------------------
 	@echo First listen to the white noise input signal:
 	@echo ---------------------------------------------
-	play ../snd/noise.wav
 	@echo -------------------------
 	@echo Compare to plugin output.
 	@echo -------------------------
 	@echo Should be a noise band around 6000Hz, repeated quietly after 1s.
-	play /tmp/test.wav
 
 install:	targets
 	-mkdirhier	$(INSTALL_PLUGINS_DIR)				\
