--- pegasus-2.3.2/src/Pegasus/Common/Platform_LINUX_IX86_GNU.h	2003-10-22 14:26:03.000000000 +0000
+++ pegasus-2.3.2-patched/src/Pegasus/Common/Platform_LINUX_IX86_GNU.h	2004-06-20 01:17:37.000000000 +0000
@@ -59,7 +59,7 @@
 
 #define PEGASUS_HAVE_FOR_SCOPE
 
-// #define PEGASUS_HAVE_TEMPLATE_SPECIALIZATION
+#define PEGASUS_HAVE_TEMPLATE_SPECIALIZATION
 
 #ifndef _GNU_SOURCE
 #define _GNU_SOURCE
@@ -71,7 +71,7 @@
 #include <features.h>
 #include <stdio.h>
 #include <stdlib.h>
-#ifndef min
-#define min(a,b) (((a)<(b))?(a):(b))
-#endif
+// #ifndef min
+// #define min(a,b) (((a)<(b))?(a):(b))
+// #endif
 #endif /* Pegasus_Platform_LINUX_IX86_GNU_h */
