
The usual tail fixes.

  - Rene Rebe <rene@rocklinux.org>

diff -u exim-4.22-orig/scripts/Configure-config.h exim-4.22/scripts/Configure-config.h
--- exim-4.22-orig/scripts/Configure-config.h	2003-08-18 14:52:52.000000000 +0200
+++ exim-4.22/scripts/Configure-config.h	2004-05-03 21:01:25.000000000 +0200
@@ -41,7 +41,7 @@
 
 # Double-check that config.h is complete.
 
-if [ "`tail -1 config.h`" != "/* End of config.h */" ] ; then
+if [ "`tail -n 1 config.h`" != "/* End of config.h */" ] ; then
   echo "*** config.h appears to be incomplete"
   echo "*** unexpected failure in buildconfig program"
   exit 1
