--- ./examples/Makefile	2002-01-16 16:18:53.000000000 +0100
+++ ./examples/Makefile-new	2004-01-02 19:47:03.000000000 +0100
@@ -39,7 +39,7 @@
 
 
 ${DBIN}/DEVA_server:	DEVA_server.o
-	${CXX} ${CXXFLAG} DEVA_server.o -o ${DBIN}/DEVA_server ${LIBDIR} -ldeva_server -ldeva_comms -ldeva_kernel -lpthread ${DYNAMIC}
+	${CXX} ${CXXFLAG} DEVA_server.o -o ${DBIN}/DEVA_server ${LIBDIR} -ldeva_server -ldeva_comms -L${MAV_HOME}/lib -lmaverik -ldeva_kernel -lpthread ${DYNAMIC}
 
 ${DBIN}/DEVA_client:	DEVA_client.o
 	${CXX} ${CXXFLAGS} DEVA_client.o -o ${DBIN}/DEVA_client ${LIBDIR} -ldeva_viewer -L${MAV_HOME}/lib -lmaverik -ldeva_client -ldeva_comms -ldeva_kernel ${DYNAMIC}
