--- ./vfs-methods/fontilus/Makefile.in.orig	2005-03-17 12:21:42.083267976 -0400
+++ ./vfs-methods/fontilus/Makefile.in	2005-03-17 12:23:24.650675360 -0400
@@ -813,7 +813,6 @@
 @GCONF_SCHEMAS_INSTALL_FALSE@install-gconf-schemas:
 
 install-desktop-database: install-desktopDATA
-	update-desktop-database $(DESTDIR)$(desktopdir)
 
 uninstall-local:
 	rm -f $(DESTDIR)$(desktopdir)/mimeinfo.cache
--- ./vfs-methods/themus/Makefile.in.orig	2005-03-17 12:21:49.017213856 -0400
+++ ./vfs-methods/themus/Makefile.in	2005-03-17 12:24:42.122897792 -0400
@@ -868,7 +868,6 @@
 @GCONF_SCHEMAS_INSTALL_FALSE@install-gconf-schemas:
 
 install-desktop-database: install-desktopDATA
-	update-desktop-database $(DESTDIR)$(desktopdir)
 
 distclean-local:
 	rm -f $(DESTDIR)$(desktopdir)/mimeinfo.cache
