--- samba-4.20.8/lib/util/charset/wscript_configure.vanilla	2025-04-09 23:11:14.009109068 +0200
+++ samba-4.20.8/lib/util/charset/wscript_configure	2025-04-09 23:11:52.387997074 +0200
@@ -37,7 +37,7 @@
                 lib='iconv',
                 headers='errno.h iconv.h')
 
-if conf.CHECK_CFG(package='icu-i18n',
+if conf.CHECK_CFG(package='icu-i18n icu-uc',
                args='--cflags --libs',
                msg='Checking for icu-i18n',
                uselib_store='ICU_I18N'):
