
Taken from Frugalware fix compilation with heimdal.

--- ./include/includes.h.orig	2005-09-05 12:13:20.279231750 +0200
+++ ./include/includes.h	2005-09-05 12:09:02.931148500 +0200
@@ -27,7 +27,7 @@
 
 #ifndef __cplusplus
 #define class #error DONT_USE_CPLUSPLUS_RESERVED_NAMES
-#define private #error DONT_USE_CPLUSPLUS_RESERVED_NAMES
+// #define private #error DONT_USE_CPLUSPLUS_RESERVED_NAMES
 #define public #error DONT_USE_CPLUSPLUS_RESERVED_NAMES
 #define protected #error DONT_USE_CPLUSPLUS_RESERVED_NAMES
 #define template #error DONT_USE_CPLUSPLUS_RESERVED_NAMES
