diff -Nur Gtk-Perl-0.7008/GtkHTML/pkg.defs Gtk-Perl-0.7008.test/GtkHTML/pkg.defs
--- Gtk-Perl-0.7008/GtkHTML/pkg.defs	2001-01-21 12:21:06.000000000 +0100
+++ Gtk-Perl-0.7008.test/GtkHTML/pkg.defs	2003-08-05 20:13:34.000000000 +0200
@@ -61,7 +61,6 @@
   (remove-link GTK_HTML_COMMAND_REMOVE_LINK)
   (delete GTK_HTML_COMMAND_DELETE)
   (delete-back GTK_HTML_COMMAND_DELETE_BACK)
-  (set-mark GTK_HTML_COMMAND_SET_MARK)
   (disable-selection GTK_HTML_COMMAND_DISABLE_SELECTION)
   (bold-on GTK_HTML_COMMAND_BOLD_ON)
   (bold-off GTK_HTML_COMMAND_BOLD_OFF)
diff -Nur Gtk-Perl-0.7008/GtkHTML/xs/GtkHTML.xs Gtk-Perl-0.7008.test/GtkHTML/xs/GtkHTML.xs
--- Gtk-Perl-0.7008/GtkHTML/xs/GtkHTML.xs	2001-06-17 18:35:46.000000000 +0200
+++ Gtk-Perl-0.7008.test/GtkHTML/xs/GtkHTML.xs	2003-08-05 20:09:49.000000000 +0200
@@ -122,9 +122,10 @@
 	RETVAL
 
 void
-gtk_html_set_iframe_parent (html, parent)
+gtk_html_set_iframe_parent (html, parent, dummy)
 	Gtk::HTML	html
 	Gtk::Widget	parent
+        void*           dummy
 
 gulong
 gtk_html_begin (html)
@@ -167,8 +168,9 @@
 	bool	allow
 
 int
-gtk_html_request_paste (html, type, time)
+gtk_html_request_paste (html, dummy, type, time)
 	Gtk::HTML	html
+        void*           dummy
 	int	type
 	int	time
 
