--- ./Makefile.PL.orig	2003-08-05 20:00:52.000000000 +0200
+++ ./Makefile.PL	2003-08-05 20:01:29.000000000 +0200
@@ -22,9 +22,9 @@
 		check => sub {ccompile('#include <gtkgl/gtkglarea.h>', "", `gtk-config --cflags`,
 					"-lgtkgl  -lGL -lGLU " . `gtk-config --libs`)}},
 	gdkpixbuf => {order => 2, dir => 'GdkPixbuf', depends => [qw(gtk)],
-		check => sub {cconfig('gdk-pixbuf-config --version', '0\.(8|9|1\d)\.?')}},
+		check => sub {cconfig('gdk-pixbuf-config --version', '0\.(2|8|9|1\d)\.?')}},
 	gtkhtml => {order => 2, dir => 'GtkHTML', depends => [qw(gtk)],
-		check => sub {cconfig('gnome-config --modversion gtkhtml', 'gtkhtml-0\.[89]')}},
+		check => sub {cconfig('gnome-config --modversion gtkhtml', 'gtkhtml-((1\.0)|(0\.[89]))')}},
 	gtkxmhtml => {order => 2, dir => 'GtkXmHTML', depends => [qw(gtk)],
 		check => sub {cconfig('gnome-config --libs gtkxmhtml', '-lgtkxmhtml')}},
 	gnome => {order => 3, dir => 'Gnome', depends => [qw(gdkimlib)],
