--- SDL-1.2.9/src/video/SDL_blit_A.c.orig	2006-03-10 15:38:21.000000000 -0800
+++ SDL-1.2.9/src/video/SDL_blit_A.c	2006-03-10 15:40:13.000000000 -0800
@@ -1319,7 +1319,7 @@
 
 #ifdef MMX_ASMBLIT
 /* fast (as in MMX with prefetch) ARGB888->(A)RGB888 blending with pixel alpha */
-inline static void BlitRGBtoRGBPixelAlphaMMX3DNOW(SDL_BlitInfo *info)
+static void BlitRGBtoRGBPixelAlphaMMX3DNOW(SDL_BlitInfo *info)
 {
 	int width = info->d_width;
 	int height = info->d_height;
