--- SDL-1.2.11/src/video/fbcon/SDL_fbvideo.c.vanilla	2007-04-21 11:41:02.000000000 +0000
+++ SDL-1.2.11/src/video/fbcon/SDL_fbvideo.c	2007-04-21 11:41:23.000000000 +0000
@@ -29,7 +29,6 @@
 #include <unistd.h>
 #include <sys/ioctl.h>
 #include <sys/mman.h>
-#include <asm/page.h>		/* For definition of PAGE_SIZE */
 #include <linux/vt.h>
 
 #include "SDL_video.h"
@@ -547,7 +546,7 @@
 
 	/* Memory map the device, compensating for buggy PPC mmap() */
 	mapped_offset = (((long)finfo.smem_start) -
-	                (((long)finfo.smem_start)&~(PAGE_SIZE-1)));
+	                (((long)finfo.smem_start)&~(sysconf(_SC_PAGESIZE)-1)));
 	mapped_memlen = finfo.smem_len+mapped_offset;
 	mapped_mem = do_mmap(NULL, mapped_memlen,
 	                  PROT_READ|PROT_WRITE, MAP_SHARED, console_fd, 0);
