--- SDL-1.2.15/src/joystick/linux/SDL_sysjoystick.c.vanilla	2018-11-17 18:49:38.519999962 +0000
+++ SDL-1.2.15/src/joystick/linux/SDL_sysjoystick.c	2018-11-17 18:53:16.319999957 +0000
@@ -29,7 +29,7 @@
 #include <unistd.h>
 #include <fcntl.h>
 #include <sys/ioctl.h>
-#include <limits.h>		/* For the definition of PATH_MAX */
+#include <sys/param.h>		/* For the definition of PATH_MAX */
 #include <linux/joystick.h>
 #if SDL_INPUT_LINUXEV
 #include <linux/input.h>
