--- posix/lposix.c	2005-09-08 21:50:42.445019600 +0200
+++ posix-t2/lposix.c	2005-09-08 21:49:14.994723494 +0200
@@ -22,6 +22,7 @@
 #include <time.h>
 #include <unistd.h>
 #include <utime.h>
+#include <glob.h>
 
 #define MYNAME		"posix"
 #define MYVERSION	MYNAME " library for " LUA_VERSION " / Nov 2003"
@@ -177,6 +178,25 @@
 }
 
 
+static int Pglob(lua_State *L)			/** glob(pattern) */
+{
+	const char *pattern = luaL_optstring(L, 1, ".");
+	glob_t globres;
+
+	if (glob(pattern, GLOB_MARK, NULL, &globres))
+		return pusherror(L, pattern);
+	else
+	{
+		int i;
+		lua_newtable(L);
+		for (i=1; i<=globres.gl_pathc; i++)
+			storeindex(L, i, globres.gl_pathv[i-1]);
+		globfree(&globres);
+		return 1;
+	}
+}
+
+
 static int aux_files(lua_State *L)
 {
 	DIR *d = lua_touserdata(L, lua_upvalueindex(1));
@@ -773,6 +793,7 @@
 	{"getlogin",		Pgetlogin},
 	{"getpasswd",		Pgetpasswd},
 	{"getprocessid",	Pgetprocessid},
+	{"glob",                Pglob},
 	{"kill",		Pkill},
 	{"link",		Plink},
 	{"mkdir",		Pmkdir},
