--- ./data/themes/taillights/Makefile.am.orig	2005-03-01 15:54:36.026396912 -0300
+++ ./data/themes/taillights/Makefile.am	2005-03-01 15:55:03.271255064 -0300
@@ -1,7 +1,7 @@
 EXTRA_DIST= taillights.edc
 SUBDIRS = parts programs fonts images
 DB = taillights.eet
-CC = edje_cc
+ECC = edje_cc
 INCLUDES = -v -id ./images -fd ./fonts
 
 CLEANFILES = $(DB)
@@ -12,4 +12,4 @@
 all: $(DB)
 
 $(DB): taillights.edc
-	$(CC) $(INCLUDES) taillights.edc $@
+	$(ECC) $(INCLUDES) taillights.edc $@
