diff -ruN entrance-orig/data/config/autodetect.sh.in entrance/data/config/autodetect.sh.in
--- entrance-orig/data/config/autodetect.sh.in	2000-01-01 00:00:00.000000000 +0100
+++ entrance/data/config/autodetect.sh.in	2006-04-07 14:00:34.000000000 +0200
@@ -47,11 +47,11 @@
 
 You could pick a vt manually and run this command (assuming you picked vt7) -
 
-sudo ecore_config -c $DB -k /entranced/xserver -s \"/usr/X11R6/bin/X -quiet -nolisten tcp vt7\"
+sudo ecore_config -c $DB -k /entranced/xserver -s \"/usr/X11/bin/X -quiet -nolisten tcp vt7\"
 "
   else
     echo "Setting the X virtual terminal to vt$found."
-    sudo ecore_config -c $DB -k /entranced/xserver -s "/usr/X11R6/bin/X -quiet -nolisten tcp vt$found"
+    sudo ecore_config -c $DB -k /entranced/xserver -s "/usr/X11/bin/X -quiet -nolisten tcp vt$found"
   fi
 )
 echo ""
diff -ruN entrance-orig/data/config/build_config.sh.in entrance/data/config/build_config.sh.in
--- entrance-orig/data/config/build_config.sh.in	2000-01-01 00:00:00.000000000 +0100
+++ entrance/data/config/build_config.sh.in	2006-04-07 14:01:32.000000000 +0200
@@ -16,7 +16,7 @@
 ecore_config -c $DB -k /entrance/time_format -s "%l:%M:%S %p"
 #ecore_config -c $DB -k /entrance/fonts/count -i 2
 #ecore_config -c $DB -k /entrance/fonts/0/-s -s "/usr/share/fonts/truetype/"
-#ecore_config -c $DB -k /entrance/fonts/1/-s -s "/usr/X11R6/lib/X11/fonts/Truetype/"
+#ecore_config -c $DB -k /entrance/fonts/1/-s -s "/usr/X11/lib/X11/fonts/Truetype/"
 ecore_config -c $DB -k /entrance/greeting/before -s "Welcome to"
 ecore_config -c $DB -k /entrance/greeting/after -s ""
 ecore_config -c $DB -k /entrance/session/count -i 7
@@ -50,5 +50,5 @@
 ecore_config -c $DB -k /entrance/presel/mode -i 1
 
 # for Entranced
-ecore_config -c $DB -k /entranced/xserver -s "/usr/X11R6/bin/X -quiet -nolisten tcp"
+ecore_config -c $DB -k /entranced/xserver -s "/usr/X11/bin/X -quiet -nolisten tcp"
 ecore_config -c $DB -k /entranced/attempts -i 5
diff -ruN entrance-orig/src/client/entrance_session.c entrance/src/client/entrance_session.c
--- entrance-orig/src/client/entrance_session.c	2000-01-01 00:00:00.000000000 +0100
+++ entrance/src/client/entrance_session.c	2006-04-07 14:01:55.000000000 +0200
@@ -403,7 +403,7 @@
    {
       printf("Would have executed: %s\n", buf);
       fflush(stdout);
-      snprintf(buf, PATH_MAX, "/usr/X11R6/bin/xterm");
+      snprintf(buf, PATH_MAX, "/usr/X11/bin/xterm");
    }
 
    syslog(LOG_INFO, "Executing %s", buf);
diff -ruN entrance-orig/src/daemon/Entranced.h entrance/src/daemon/Entranced.h
--- entrance-orig/src/daemon/Entranced.h	2000-01-01 00:00:00.000000000 +0100
+++ entrance/src/daemon/Entranced.h	2006-04-07 14:03:06.000000000 +0200
@@ -19,8 +19,8 @@
 
 #define ENTRANCED_DEBUG 1
 
-#define DEFAULT_X_SERVER "/usr/X11R6/bin/X -quiet -nolisten tcp"
-//#define X_SERVER "/usr/X11R6/bin/Xnest -full"
+#define DEFAULT_X_SERVER "/usr/X11/bin/X -quiet -nolisten tcp"
+//#define X_SERVER "/usr/X11/bin/Xnest -full"
 #define X_DISP ":0"             /* only used if DISPLAY variable is NOT set */
 #define ENTRANCE PREFIX "/bin/entrance_wrapper"
 #define PIDFILE "/var/run/entranced.pid"
