--- EMILE-EMILE-0.14/libext2/ext2_utils.c.vanilla	2024-03-27 18:30:07.193791000 +0000
+++ EMILE-EMILE-0.14/libext2/ext2_utils.c	2024-03-27 18:30:19.120791000 +0000
@@ -148,7 +148,6 @@
         }
 	inode->i_generation = __le32_to_cpu(le_inode->i_generation);
 	inode->i_file_acl = __le32_to_cpu(le_inode->i_file_acl);
-	inode->i_dir_acl = __le32_to_cpu(le_inode->i_dir_acl);
 	inode->i_faddr = __le32_to_cpu(le_inode->i_faddr);
 	inode->osd2.linux2.l_i_blocks_hi =
                        __le16_to_cpu(le_inode->osd2.linux2.l_i_blocks_hi);
