--- j2sdk-1.4.2-01-linux-i586.bin	2005-03-24 15:25:49.000000000 +0100
+++ j2sdk-1.4.2-01-linux-i586.bin	2005-03-24 15:31:03.000000000 +0100
@@ -3,7 +3,7 @@
 # This script was generated using Makeself 1.5.5-Blackdown-j2se with modifications from Blackdown (source code is available from Blackdown mirrors)
 TAR_CMD=tar
 CRCsum=1459487342
-MD5=61635e4e5eea44f6be27823cea1cf12c
+MD5=\"00000000000000000000000000000000\"; SETUP_NOCHECK=1
 PATH=/usr/bin:/bin
 label="Blackdown Java 2 Standard Edition SDK v1.4.2-01 for Linux/i586"
 NO_DOT_VERSION=142
@@ -112,7 +112,7 @@
         fi
     fi
 fi
-more << EOF_LICENSE
+agreed=1; cat << EOF_LICENSE
 Sun Microsystems, Inc.  Binary Code License Agreement for
 the JAVATM 2 SOFTWARE DEVELOPMENT KIT (J2SDK), STANDARD
 EDITION, VERSION 1.4.2_X
