diff -Naur boost_1_40_0.orig/libs/graph/module.cmake boost_1_40_0/libs/graph/module.cmake
--- boost_1_40_0.orig/libs/graph/module.cmake	2009-06-08 15:45:46.000000000 +0000
+++ boost_1_40_0/libs/graph/module.cmake	2009-11-01 14:41:18.000000000 +0000
@@ -1,4 +1,4 @@
-boost_module(graph DEPENDS property_map tuple multi_index any random)
+boost_module(graph DEPENDS property_map tuple multi_index any random parameter)
 boost_module(graph_mpi DEPENDS mpi graph)
 
 # any is there because of the dependency on boost/property_map/dynamic_property_map.hpp
