--- ipsec-tools-0.6.5/src/racoon/oakley.c.orig	2006-02-11 14:13:53.183921968 +0100
+++ ipsec-tools-0.6.5/src/racoon/oakley.c	2006-02-11 14:15:35.584354736 +0100
@@ -2189,8 +2189,8 @@
 	case ISAKMP_CERT_CRL:
 	default:
 		plog(LLV_ERROR, LOCATION, NULL,
-			"Invalid CR type %d\n", type);
-		return -1;
+			"Warning: Ignoring invalid CR type %d\n", type);
+		return 0;
 	}
 
 	*c = save_certbuf(gen);
