--- kexec-tools-2.0.27/purgatory/arch/ppc64/Makefile.vanilla	2024-01-23 11:01:46.141986158 +0100
+++ kexec-tools-2.0.27/purgatory/arch/ppc64/Makefile	2024-01-23 11:04:26.115979780 +0100
@@ -14,8 +14,8 @@
 ppc64_PURGATORY_EXTRA_ASFLAGS += -m64
 ifeq ($(SUBARCH),BE)
 	ppc64_PURGATORY_EXTRA_LDFLAGS += -melf64ppc
-	ppc64_PURGATORY_EXTRA_CFLAGS += -mcall-aixdesc
-	ppc64_PURGATORY_EXTRA_ASFLAGS += -mcall-aixdesc
+	ppc64_PURGATORY_EXTRA_CFLAGS += #-mabi=elfv1 -mcall-aixdesc
+	ppc64_PURGATORY_EXTRA_ASFLAGS += #-mabi=elfv1 -mcall-aixdesc
 else
 	ppc64_PURGATORY_EXTRA_LDFLAGS += -melf64lppc
 	ppc64_PURGATORY_EXTRA_CFLAGS += -mlittle-endian
