Index: src/cpu/core_full/op.h
===================================================================
--- ./src/cpu/core_full/op.h	(revision 4019)
+++ ./src/cpu/core_full/op.h	(working copy)
@@ -358,7 +358,7 @@
 	case O_INT:
 #if C_DEBUG
 		FillFlags();
-		if (((inst.entry & 0xFF)==0xcc) && DEBUG_Breakpoint()) 
+		if (((inst.entry & 0xFF)==0xcc) && (DEBUG_Breakpoint() || DEBUG_IntBreakpoint(3)))
 			return debugCallback;
 		else if (DEBUG_IntBreakpoint(inst_op1_b)) 
 			return debugCallback;
Index: src/cpu/core_normal/prefix_none.h
===================================================================
--- ./src/cpu/core_normal/prefix_none.h	(revision 4019)
+++ ./src/cpu/core_normal/prefix_none.h	(working copy)
@@ -743,7 +743,7 @@
 	CASE_B(0xcc)												/* INT3 */
 #if C_DEBUG	
 		FillFlags();
-		if (DEBUG_Breakpoint())
+		if (DEBUG_Breakpoint() || DEBUG_IntBreakpoint(3))
 			return debugCallback;
 #endif			
 		CPU_SW_Interrupt_NoIOPLCheck(3,GETIP);
