--- mm2.7/src/fonts/Makefile.marc	Tue Feb 13 14:56:31 1996
+++ mm2.7/src/fonts/Makefile	Tue Feb 13 14:57:00 1996
@@ -28,11 +28,11 @@
 	-${DIRBUILDER}
 
 shownonascii: Xshownonascii
-	sed -e s@/u/nsb/lib/fonts@`pwd`@ < Xshownonascii > shownonascii
+	sed -e s@/u/nsb/lib/fonts@/usr/lib/metamail/fonts@ < Xshownonascii > shownonascii
 	chmod +x shownonascii
 
 mailto-hebrew: Xmailto-hebrew
-	sed -e s@/u/nsb/lib/fonts@`pwd`@ < Xmailto-hebrew > mailto-hebrew
+	sed -e s@/u/nsb/lib/fonts@/usr/lib/metamail/fonts@ < Xmailto-hebrew > mailto-hebrew
 	chmod +x mailto-hebrew
 
 clean:
