--- mm2.7/src/metamail/mailto.c~	Thu Sep  9 01:24:39 1999
+++ mm2.7/src/metamail/mailto.c	Thu Sep  9 01:28:07 1999
@@ -1796,6 +1796,8 @@
                 printf("\nEnter the MIME Content-type value for the data from file %s\n    (type '?' for a list of locally-valid content-types): ", sdum);
                 fflush(stdout);
                 fgets(LineBuf, sizeof(LineBuf), stdin);
+		if (LineBuf[strlen(LineBuf)-1] == '\n')
+		    LineBuf[strlen(LineBuf)-1] = '\0';
                 if (index(LineBuf, '/')) {
                     char lc[100], *s, AnsBuf[100];
                     strcpy(lc, LineBuf);
