--- mm2.7/src/config.h.linux	Sun Jun 18 21:59:38 2000
+++ mm2.7/src/config.h	Sun Jun 18 21:59:10 2000
@@ -100,9 +100,11 @@
 #endif
 
 #ifdef SYSV
-#define killpg(a, b) kill(-(a), (b))
-#define bcopy(a, b, c) memcpy(b, a, c)
-#define bzero(a, b) memset(a, 0, b)
+#ifndef linux
+# define killpg(a, b) kill(-(a), (b))
+# define bcopy(a, b, c) memcpy(b, a, c)
+# define bzero(a, b) memset(a, 0, b)
+#endif
 #define bcmp memcmp
 #define index strchr
 #define rindex strrchr
