--- mm2.7/src/bin/showpartial.fixpartial	Thu Feb  1 12:20:37 2001
+++ mm2.7/src/bin/showpartial	Thu Feb  1 12:21:29 2001
@@ -13,7 +13,7 @@
 endif
 set file="$1"
 # This next line is because message-id can contain weird chars
-set id=`echo $2 | tr -d  \!\$\&\*\(\)\|\'\"\;\/\<\>\\` 
+set id=`echo $2 | tr -d  \!\$\&\*\(\)\|\'\"\;\/\<\>\\\\` 
 @ partnum = "$3"
 if ($#argv == 3 || "$4" == "") then
     set totalnum=-1
@@ -74,9 +74,9 @@
 end
 if ("$found" == "$totalnum") then
     cd "${TREEROOT}/$id"
-    cat "$list" > "${TREEROOT}/$id/FULL"
+    cat $list > "${TREEROOT}/$id/FULL"
 #    cat ${TREEROOT}/$id/HDRS $list > ${TREEROOT}/$id/FULL
-    rm "$list"
+    rm $list
     echo All parts of this "${totalnum}"-part message have now been read.
     metamail -d "${TREEROOT}/$id/FULL"
     echo WARNING:  To save space, the full file is now being deleted.  
