diff -Nur mm2.7.orig/src/bin/showexternal mm2.7/src/bin/showexternal
--- mm2.7.orig/src/bin/showexternal	2003-05-14 23:58:27.000000000 +0200
+++ mm2.7/src/bin/showexternal	2003-05-14 23:59:21.000000000 +0200
@@ -219,7 +219,7 @@
 	echo "" >> "$NEWNAME"
 	sed -e 1,/^\$/d < $bodyfile >> "$NEWNAME"
 	echo "" >> "$NEWNAME"
-	/usr/lib/sendmail -t  < "$NEWNAME"
+	/usr/sbin/sendmail -t  < "$NEWNAME"
 	if ($status) then
 	    echo sendmail failed
 	    cd ${METAMAIL_TMPDIR}
