--- psmisc-21.6/src/Makefile.in.vanilla	2005-05-29 11:28:59.000000000 +0200
+++ psmisc-21.6/src/Makefile.in	2005-05-29 11:29:09.000000000 +0200
@@ -459,7 +459,7 @@
 
 install-exec-hook:
 		cd $(DESTDIR)$(bindir) && \
-			$(LN_S) pstree pstree.x11
+			$(LN_S) -f pstree pstree.x11
 # Tell versions [3.59,3.63) of GNU make to not export all variables.
 # Otherwise a system limit (for SysV at least) may be exceeded.
 .NOEXPORT:
