diff -urN -x '*~' DirectFB-0.9.18/gfxdrivers/matrox/matrox_maven.c DirectFB-0.9.18-fix/gfxdrivers/matrox/matrox_maven.c
--- DirectFB-0.9.18/gfxdrivers/matrox/matrox_maven.c	2003-06-02 17:28:24.000000000 +0200
+++ DirectFB-0.9.18-fix/gfxdrivers/matrox/matrox_maven.c	2003-10-02 07:53:46.000000000 +0200
@@ -21,14 +21,26 @@
    Boston, MA 02111-1307, USA.
 */
 
+/* TERRIBLE HACK */
+#define __KERNEL__
+
 #include <stdio.h>
 #include <unistd.h>
 #include <string.h>
 #include <ctype.h>
 #include <sys/types.h>
+#include <asm/types.h>
 #include <sys/stat.h>
 #include <fcntl.h>
 #include <sys/ioctl.h>
+
+// where the fsck does this get defined:
+typedef __s32 s32;
+typedef __u32 u32;
+typedef __u16 u16;
+typedef __u8 u8;
+
+#include <linux/i2c.h>
 #include <linux/i2c-dev.h>
 
 #include <misc/util.h>
