diff -urN pl240src.orig/src/gd13.c pl240src/src/gd13.c
--- pl240src.orig/src/gd13.c	2007-06-20 15:39:54.000000000 +0200
+++ pl240src/src/gd13.c	2008-01-28 17:00:06.000000000 +0100
@@ -2722,6 +2722,7 @@
 	im->interlace = interlaceArg;
 }
 
+static  /* added scg 1/18/08 */
 int cost[] = {
   1024,
   1023,
@@ -3085,6 +3086,7 @@
   1023
 };
 
+static  /* added scg 1/18/08 */
 int sint[] = {
   0,
   17,
diff -urN pl240src.orig/src/proc_axis.c pl240src/src/proc_axis.c
--- pl240src.orig/src/proc_axis.c	2008-02-16 18:02:27.000000000 +0100
+++ pl240src/src/proc_axis.c	2008-01-25 18:06:28.000000000 +0100
@@ -803,7 +803,7 @@
 				Ecblock( glemin, gbylast, glemax, Ea( Y, y ), gbcolor2, 0 );
 				gbstate = 0;
 				}
-			/* gbylast = Ea( Y, y ); */
+			gbylast = Ea( Y, y );
 			}
 		else	{
 			linedet( "grid", griddet, 0.5 );
@@ -823,7 +823,8 @@
 			}
 
 		/* urlencode the value now.. scg 5/29/06 */
-		GL_urlencode( cmtxt, buf );
+		/* GL_urlencode( cmtxt, buf ); */ /* don't encode it now.. procedures were changed in clickmap.c  scg 1/25/08 */
+		strcpy( buf, cmtxt ); 		  /* <-- do this instead */
 
 		if( Eflip ) {
 			clickmap_entry( 'r', buf, clickmap, (Ea(Y,y)-halfdown)+stubslide, cmemin, 
