--- ./zlib/Makefile.in.orig	2002-11-25 21:28:45.000000000 +0200
+++ ./zlib/Makefile.in	2003-01-25 15:39:06.000000000 +0200
@@ -227,15 +227,13 @@
 	$(mkinstalldirs) $(DESTDIR)$(toolexeclibdir)
 	@list='$(toolexeclib_LIBRARIES)'; for p in $$list; do \
 	  if test -f $$p; then \
-	    echo " $(INSTALL_DATA) $$p $(DESTDIR)$(toolexeclibdir)/$$p"; \
-	    $(INSTALL_DATA) $$p $(DESTDIR)$(toolexeclibdir)/$$p; \
+	    echo "flistdel: $$p should not install."; \
 	  else :; fi; \
 	done
 	@$(POST_INSTALL)
 	@list='$(toolexeclib_LIBRARIES)'; for p in $$list; do \
 	  if test -f $$p; then \
-	    echo " $(RANLIB) $(DESTDIR)$(toolexeclibdir)/$$p"; \
-	    $(RANLIB) $(DESTDIR)$(toolexeclibdir)/$$p; \
+	    echo "flistdel: $$p should not reindex."; \
 	  else :; fi; \
 	done
 
