--- ncurses-5.4/include/MKhashsize.sh.orig	2006-10-16 16:22:42.000000000 -0700
+++ ncurses-5.4/include/MKhashsize.sh	2006-10-16 16:21:25.000000000 -0700
@@ -7,7 +7,7 @@
 echo " */"
 
 CAPS="${1-Caps}"
-TABSIZE=`grep -v '^[ #]' $CAPS | grep -v "^$" | grep -v "^capalias"| grep -v "^infoalias" | wc -l`
+TABSIZE=$(grep -v '^[ #]' $CAPS | grep -v "^$" | grep -v "^capalias"| grep -v "^infoalias" | wc -l )
 
 echo ""
 echo "#define CAPTABSIZE	${TABSIZE}"
