--- ./ncurses/Makefile.orig	2005-01-19 18:44:35.000000000 -0300
+++ ./ncurses/Makefile	2005-01-19 18:46:32.000000000 -0300
@@ -617,9 +617,8 @@
 install.libs \
 install.ncurses :: $(DESTDIR)$(libdir) $(LIBRARIES)
 	@echo installing ../lib/libncurses.so.$(REL_VERSION) as $(DESTDIR)$(libdir)/libncurses.so.$(REL_VERSION)
-	-@rm -f $(DESTDIR)$(libdir)/libncurses.so.$(REL_VERSION)
 	$(INSTALL_LIB) ../lib/libncurses.so.$(REL_VERSION) $(DESTDIR)$(libdir)/libncurses.so.$(REL_VERSION)
-	cd $(DESTDIR)$(libdir) && (rm -f libncurses.so.$(ABI_VERSION); $(LN_S) libncurses.so.$(REL_VERSION) libncurses.so.$(ABI_VERSION); rm -f libncurses.so; $(LN_S) libncurses.so.$(ABI_VERSION) libncurses.so; )
+	cd $(DESTDIR)$(libdir) && ($(LN_S) -f libncurses.so.$(REL_VERSION) libncurses.so.$(ABI_VERSION); $(LN_S) -f libncurses.so.$(ABI_VERSION) libncurses.so; )
 	- test -z "$(DESTDIR)" && /sbin/ldconfig
 
 uninstall \
