--- ./mk-1st.awk.orig	2002-04-20 20:32:47.000000000 +0300
+++ ./mk-1st.awk	2003-09-22 16:15:38.000000000 +0300
@@ -224,8 +224,7 @@
 					src_name = sprintf("../lib/%s", end_name);
 					dst_name = sprintf("$(DESTDIR)$(libdir)/%s", end_name);
 					printf "\t@echo installing %s as %s\n", src_name, dst_name
-					printf "\t-@rm -f %s\n", dst_name
-					printf "\t$(INSTALL_LIB) %s %s\n", src_name, dst_name
+					printf "\tinstall %s %s\n", src_name, dst_name
 
 					sharedlinks("$(DESTDIR)$(libdir)")
 diff -Nru3 trunk-1418/package/base/liblockfile/force-symlink.patch trunk/package/base/liblockfile/force-symlink.patch
