--- ./mk-1st.awk.orig	2002-04-20 19:32:47.000000000 +0200
+++ ./mk-1st.awk	2003-11-10 09:04:39.651894512 +0100
@@ -52,8 +52,7 @@
 #
 function symlink(src,dst) {
 		if ( src != dst ) {
-			printf "rm -f %s; ", dst
-			printf "$(LN_S) %s %s; ", src, dst
+			printf "$(LN_S) -f %s %s; ", src, dst
 		}
 	}
 function rmlink(directory, dst) {
@@ -224,8 +223,7 @@
 					src_name = sprintf("../lib/%s", end_name);
 					dst_name = sprintf("$(DESTDIR)$(libdir)/%s", end_name);
 					printf "\t@echo installing %s as %s\n", src_name, dst_name
-					printf "\t-@rm -f %s\n", dst_name
-					printf "\t$(INSTALL_LIB) %s %s\n", src_name, dst_name
+					printf "\tinstall %s %s\n", src_name, dst_name
 
 					sharedlinks("$(DESTDIR)$(libdir)")
 
@@ -235,7 +233,7 @@
 				{
 					ovr_name = sprintf("libcurses%s", suffix)
 					printf "\t@echo linking %s to %s\n", end_name, ovr_name
-					printf "\tcd $(DESTDIR)$(libdir) && (rm -f %s; $(LN_S) %s %s; )\n", ovr_name, end_name, ovr_name
+					printf "\t$(LN_S) -f %s $(DESTDIR)$(libdir)/%s\n", end_name, ovr_name
 				}
 				if ( ldconfig != "" ) {
 					printf "\t- test -z \"$(DESTDIR)\" && %s\n", ldconfig
