--- fetchmail-6.2.5/configure.orig	2005-09-06 07:12:37.751306750 +0200
+++ fetchmail-6.2.5/configure	2005-09-06 07:13:30.398597000 +0200
@@ -13959,7 +13959,7 @@
 
   if test -f "$with_gssapi/include/roken.h"
   then
-     LIBS="$LIBS -lkrb5 -lasn1 -ldes -lroken -ldb -ldl"
+     LIBS="$LIBS -lkrb5 -lasn1 -lroken -ldb -ldl"
      echo "$as_me:$LINENO: checking for gss_init_sec_context in -lgssapi" >&5
 echo $ECHO_N "checking for gss_init_sec_context in -lgssapi... $ECHO_C" >&6
 if test "${ac_cv_lib_gssapi_gss_init_sec_context+set}" = set; then
