--- jahshaka/source/Jahshaka/JahLibraries/jahformatter/valueFormatters.cpp_orig	2006-05-17 14:52:37.368546250 +0200
+++ jahshaka/source/Jahshaka/JahLibraries/jahformatter/valueFormatters.cpp	2006-05-17 14:53:30.731881250 +0200
@@ -6,6 +6,7 @@
 
 // std
 #include <cmath>
+#include <cstdlib>
 
 TimecodeValueFormatter::TimecodeValueFormatter()
     : m_fps( 30 ),
--- jahshaka/source/OpenLibraries/openmedialib/mediaobject/mediaobject.cpp_orig	2006-05-07 22:39:17.000000000 +0200
+++ jahshaka/source/OpenLibraries/openmedialib/mediaobject/mediaobject.cpp	2006-05-17 13:52:13.174048250 +0200
@@ -8,6 +8,7 @@
 *******************************************************************************/
 
 #include "mediaobject.h"
+#include "mediaobject_.h"
 
 #ifdef JAHMPEGDECODER  
 #include "mediaobject_mpeg.h"
