--- jahshaka/source/Jahshaka/JahWidgets/jahfileloader/dirview.cpp_orig	2006-05-17 15:01:32.449986750 +0200
+++ jahshaka/source/Jahshaka/JahWidgets/jahfileloader/dirview.cpp	2006-05-17 15:03:25.409046250 +0200
@@ -1029,7 +1029,7 @@
     box->setSpacing(15);
     di->setCaption( "New collection" );
     int choice = di->exec();
-    const std::string thename = le->text();
+    const std::string thename = le->text().ascii();
     bool ok = true;
     
     if ( le->text().stripWhiteSpace().length() == 0 ||  le->text().contains( QRegExp( "[^a-zA-Z0-9_\\s]" ) ) ) 
@@ -1079,7 +1079,7 @@
     lab->setText( "Enter a new name for the collection.\nNames must be unique.");
     QLineEdit *le = new QLineEdit( di );
     box->addWidget( le );
-    std::string curname = dir->text(0, false);
+    std::string curname = dir->text(0, false).ascii();
     le->setText( QString(curname.c_str() ) );
     le->selectAll();
     QPushButton *pb = new QPushButton( di );
@@ -1092,7 +1092,7 @@
     box->setSpacing(15);
     di->setCaption( "Rename collection" );
     int choice = di->exec();
-    const std::string thename = le->text();
+    const std::string thename = le->text().ascii();
     bool ok = true;
 
     if ( le->text().contains( QRegExp( "[^a-zA-Z0-9_\\s]" ) ) ) 
@@ -1123,7 +1123,7 @@
     if ( dir->dirType() != Directory::Collection )
         return;
 
-    std::string curname = dir->text(0, false);
+    std::string curname = dir->text(0, false).ascii();
 
     bool ok = true; 
 
