--- ./etc/fb.modes.ATI.orig	2003-03-23 10:31:41.000000000 +0100
+++ ./etc/fb.modes.ATI	2003-03-23 10:32:15.000000000 +0100
@@ -430,7 +430,7 @@
 #	1024x768, 72 Hz, Non-Interlaced (75.00 MHz dotclock)
 #	
 #				Horizontal	Vertical
-#	Resolution		10224		768
+#	Resolution		1024		768
 #	Scan Frequency		58.230 kHz	72.245 Hz
 #	Sync Width		1.813 us	0.103 ms
 #				17 chars	6 lines
@@ -447,7 +447,7 @@
 
 mode "1024x768-72"
     # D: 75.00 MHz, H: 58.230 kHz, V: 72.245 Hz
-    geometry 10224 768 10224 768 8
+    geometry 1024 768 10224 768 8
     timings 13334 104 24 29 3 136 6
 endmode
  
