--- ./lib/blkid/blkid_types.h.in.vanilla	2005-10-27 01:03:35.146713072 +0300
+++ ./lib/blkid/blkid_types.h.in	2005-10-27 01:04:29.142625760 +0300
@@ -3,7 +3,7 @@
  * everything we need.  (cross fingers)  Other header files may have 
  * also defined the types that we need.
  */
-#if (!defined(_LINUX_TYPES_H) && !defined(_BLKID_TYPES_H) && \
+#if (!defined(_LINUX_TYPES_H) && !defined(_BLKID_TYPES_H) && !defined(_PPC64_TYPES_H) && \
 	!defined(_EXT2_TYPES_H))
 #define _BLKID_TYPES_H

--- ./lib/ext2fs/ext2_types.h.in.vanilla	2005-10-27 00:59:45.932682280 +0300
+++ ./lib/ext2fs/ext2_types.h.in	2005-10-27 01:02:44.810644824 +0300
@@ -3,7 +3,7 @@
  * everything we need.  (cross fingers)  Other header files may have 
  * also defined the types that we need.
  */
-#if (!defined(_LINUX_TYPES_H) && !defined(_BLKID_TYPES_H) && \
+#if (!defined(_LINUX_TYPES_H) && !defined(_BLKID_TYPES_H) && !defined(_PPC64_TYPES_H) && \
 	!defined(_EXT2_TYPES_H))
 #define _EXT2_TYPES_H
 
