--- Zope-2.10.6-final/configure.orig	2008-05-09 23:33:32.000000000 -0700
+++ Zope-2.10.6-final/configure	2008-06-18 11:20:27.000000000 -0700
@@ -18,11 +18,11 @@
 # sys.version) below in "best" to "worst" order, not including the
 # target version.  Up to six acceptable python versions are allowed.
 # Do not include the target version number in this list!
-ACCEPTABLE="2.4.4"
+ACCEPTABLE="2.5.2 2.4.4"
 
 # provide the executable names for all the acceptable versions
 # (and the target version) below
-EXENAMES="python python2 python2.4"
+EXENAMES="python python2 python2.5 python2.4"
 
 #####################################################################
 #                    END EDITABLE PARAMETERS                        #
