--- pl232src/src/Makefile.vanilla	2005-10-09 12:15:40.000000000 +0200
+++ pl232src/src/Makefile	2005-10-09 12:16:27.000000000 +0200
@@ -42,10 +42,10 @@
 #############################################################
 
 #### For Linux uncomment the following.. #######################
-XLIBS = -L/usr/X11R6/lib -lX11
-INCLUDEDIRS = -I/usr/X11R6/include
+XLIBS = -L/usr/X11/lib -lX11
+INCLUDEDIRS = -I/usr/X11/include
 
-#### Note: for 64 bit platform I used: XLIBS = -L/usr/X11R6/lib64
+#### Note: for 64 bit platform I used: XLIBS = -L/usr/X11/lib64
 
 
 #### For Solaris 2.x uncomment the following.. #################
