--- linux-firmware-20250311/copy-firmware.sh.vanilla	2025-03-23 08:51:17.159999833 +0100
+++ linux-firmware-20250311/copy-firmware.sh	2025-03-23 08:51:28.559999832 +0100
@@ -111,7 +111,7 @@
     fi
 done
 if [ "$num_jobs" -gt 1 ]; then
-    parallel -j"$num_jobs" -a "$parallel_args_file"
+    PARALLEL_HOME=${TMPDIR:-/tmp} parallel -j"$num_jobs" -a "$parallel_args_file"
     echo > "$parallel_args_file" # prepare for next run
 fi
 
@@ -138,7 +138,7 @@
     fi
 done
 if [ "$num_jobs" -gt 1 ]; then
-    parallel -j"$num_jobs" -a "$parallel_args_file"
+    PARALLEL_HOME=${TMPDIR:-/tmp} parallel -j"$num_jobs" -a "$parallel_args_file"
 fi
 
 # Verify no broken symlinks
