--- pcmcia-cs-3.2.8/Makefile	2003-05-16 07:59:44.000000000 +0200
+++ pcmcia-cs-3.2.8-fixed/Makefile	2004-10-16 23:29:13.054710728 +0200
@@ -6,10 +6,10 @@
 include config.mk
 endif
 
-ALL  = modules clients wireless cardmgr flash debug-tools man etc
+ALL  = clients wireless cardmgr flash debug-tools man etc
 DIRS = cardmgr flash debug-tools man etc
 ifndef CONFIG_PCMCIA
-DIRS := modules clients wireless $(DIRS)
+DIRS := $(DIRS)
 endif
 
 help:
