--- mesa-23.1.0/meson.build.vanilla	2023-05-25 21:21:38.953100947 +0200
+++ mesa-23.1.0/meson.build	2023-05-25 21:24:16.027118304 +0200
@@ -1429,15 +1429,15 @@
 endif
 
 if with_intel_hasvk and host_machine.cpu_family().startswith('x86') == false
-  error('Intel "hasvk" Vulkan driver requires x86 or x86_64 CPU family')
+  #error('Intel "hasvk" Vulkan driver requires x86 or x86_64 CPU family')
 endif
 
 if with_gallium_crocus and host_machine.cpu_family().startswith('x86') == false
-  error('Intel "crocus" Gallium driver requires x86 or x86_64 CPU family')
+  #error('Intel "crocus" Gallium driver requires x86 or x86_64 CPU family')
 endif
 
 if with_gallium_i915 and host_machine.cpu_family().startswith('x86') == false
-  error('Intel "i915" Gallium driver requires x86 or x86_64 CPU family')
+  #error('Intel "i915" Gallium driver requires x86 or x86_64 CPU family')
 endif
 
 # Determine whether or not the rt library is needed for time functions
