--- pango-1.28.3/pango/Makefile.in.vanilla	2011-02-20 17:34:33.854999851 +0100
+++ pango-1.28.3/pango/Makefile.in	2011-02-20 17:35:09.291999850 +0100
@@ -702,7 +702,7 @@
 PangoFT2_1_0_gir_PACKAGES = freetype2
 PangoFT2_1_0_gir_INCLUDES = GObject-2.0 cairo-1.0 freetype2-2.0 fontconfig-2.0
 PangoFT2_1_0_gir_LIBS = libpangoft2-1.0.la
-PangoFT2_1_0_gir_CFLAGS = -I$(top_srcdir)
+PangoFT2_1_0_gir_CFLAGS = -I$(top_srcdir) -I/usr/X11R7/include/
 PangoFT2_1_0_gir_SCANNERFLAGS = --include-uninstalled=$(builddir)/Pango-1.0.gir
 PangoFT2_1_0_gir_FILES = $(filter-out %-private.h, $(addprefix $(srcdir)/,$(pangoft2_introspection_files)))
 pangoxft_headers = pangoxft.h pangoxft-render.h
