Index: kvm-75/kernel/x86/svm.c
===================================================================
--- kvm-75.orig/kernel/x86/svm.c
+++ kvm-75/kernel/x86/svm.c
@@ -50,12 +50,8 @@ MODULE_LICENSE("GPL");
 #define DEBUGCTL_RESERVED_BITS (~(0x3fULL))
 
 /* enable NPT for AMD64 and X86 with PAE */
-#if defined(CONFIG_X86_64) || defined(CONFIG_X86_PAE)
-static bool npt_enabled = true;
-#else
 static bool npt_enabled = false;
-#endif
-static int npt = 1;
+static int npt = 0;
 
 module_param(npt, int, S_IRUGO);
 
@@ -439,19 +435,19 @@ static __init int svm_hardware_setup(voi
 
 	svm_features = cpuid_edx(SVM_CPUID_FUNC);
 
-	if (!svm_has(SVM_FEATURE_NPT))
-		npt_enabled = false;
+	if (npt)
+		npt_enabled = true;
 
-	if (npt_enabled && !npt) {
-		printk(KERN_INFO "kvm: Nested Paging disabled\n");
+	if (!svm_has(SVM_FEATURE_NPT))
 		npt_enabled = false;
-	}
 
 	if (npt_enabled) {
 		printk(KERN_INFO "kvm: Nested Paging enabled\n");
 		kvm_enable_tdp();
-	} else
+	} else {
+		printk(KERN_INFO "kvm: Nested Paging disabled\n");
 		kvm_disable_tdp();
+	}
 
 	return 0;
 
