
Index: kvm-83/qemu/target-i386/helper.c
===================================================================
--- qemu/target-i386/helper.c.vanilla	2009-05-05 17:36:13.000000000 +0200
+++ qemu/target-i386/helper.c	2009-05-05 17:36:37.000000000 +0200
@@ -29,6 +29,7 @@
 #include "exec-all.h"
 #include "qemu-common.h"
 #include "kvm.h"
+#include "hw/boards.h"
 
 #include "qemu-kvm.h"
 
@@ -1473,7 +1473,7 @@
          * isn't supported in compatibility mode on Intel.  so advertise the
          * actuall cpu, and say goodbye to migration between different vendors
          * is you use compatibility mode. */
-        if (kvm_enabled())
+        if (kvm_enabled() && current_machine != &mac_machine)
             host_cpuid(0, 0, NULL, ebx, ecx, edx);
         break;
     case 1:
