
Index: kvm-75/kernel/x86/ioapic.c
===================================================================
--- kvm-75.orig/kvm/kernel/x86/ioapic.c
+++ kvm-75/kvm/kernel/x86/ioapic.c
@@ -274,7 +274,8 @@ void kvm_ioapic_set_irq(struct kvm_ioapi
 
 	if (irq >= 0 && irq < IOAPIC_NUM_PINS) {
 		entry = ioapic->redirtbl[irq];
-		level ^= entry.fields.polarity;
+// polarity is always active high in qemu
+//		level ^= entry.fields.polarity;
 		if (!level)
 			ioapic->irr &= ~mask;
 		else {
