--- q/hw/ide/ahci.c.orig	2012-09-06 10:31:27.000000000 +0200
+++ q/hw/ide/ahci.c	2012-09-25 20:54:53.000000000 +0200
@@ -926,6 +926,7 @@
              * do, I simply assume non-used fields as reserved and OR everything
              * together, independent of the command.
              */
+ide_state->select |= 0x40;
             ide_set_sector(ide_state, ((uint64_t)cmd_fis[10] << 40)
                                     | ((uint64_t)cmd_fis[9] << 32)
                                     /* This is used for LBA48 commands */
