--- flac-1.1.2/configure.in.orig	2006-03-06 16:08:00.000000000 +0100
+++ flac-1.1.2/configure.in	2006-03-06 16:08:11.000000000 +0100
@@ -157,7 +157,7 @@
 AC_DEFINE(FLAC__HAS_OGG)
 fi
 
-AM_PATH_XMMS(0.9.5.1, , AC_MSG_WARN([*** XMMS >= 0.9.5.1 not installed - xmms support will not be built]))
+#AM_PATH_XMMS(0.9.5.1, , AC_MSG_WARN([*** XMMS >= 0.9.5.1 not installed - xmms support will not be built]))
 AM_CONDITIONAL(FLaC__HAS_XMMS, test x$XMMS_INPUT_PLUGIN_DIR != x)
 
 dnl check for i18n(internationalization); these are from libiconv/gettext
