--- flac-1.1.2_orig/src/plugin_xmms/configure.c	2005-01-25 05:26:29.000000000 +0100
+++ flac-1.1.2/src/plugin_xmms/configure.c	2005-04-03 16:56:57.987305008 +0200
@@ -55,13 +55,13 @@
 		100 /* KB */, /* http_buffer_size */
 		50, /* http_prebuffer */
 		FALSE, /* use_proxy */
-		"", /* proxy_host */
+		FALSE, /* proxy_host */
 		0, /* proxy_port */
 		FALSE, /* proxy_use_auth */
-		"", /* proxy_user */
-		"", /* proxy_pass */
+		FALSE, /* proxy_user */
+		FALSE, /* proxy_pass */
 		FALSE, /* save_http_stream */
-		"", /* save_http_path */
+		FALSE, /* save_http_path */
 		FALSE, /* cast_title_streaming */
 		FALSE /* use_udp_channel */
 	},
@@ -666,7 +666,8 @@
 	gtk_box_pack_start(GTK_BOX(streaming_proxy_hbox), streaming_proxy_host_label, FALSE, FALSE, 0);
 
 	streaming_proxy_host_entry = gtk_entry_new();
-	gtk_entry_set_text(GTK_ENTRY(streaming_proxy_host_entry), flac_cfg.stream.proxy_host);
+	if(flac_cfg.stream.proxy_host)
+		gtk_entry_set_text(GTK_ENTRY(streaming_proxy_host_entry), flac_cfg.stream.proxy_host);
 	gtk_box_pack_start(GTK_BOX(streaming_proxy_hbox), streaming_proxy_host_entry, TRUE, TRUE, 0);
 
 	streaming_proxy_port_label = gtk_label_new(_("Port:"));
