
The usual gcc-4.0 fixes.

  - Rene Rebe <rene@exactcode.de>

--- xmms-1.2.10/General/ir/ir.c	2001-03-05 13:17:44.000000000 +0000
+++ xmms-1.2.10-fixed/General/ir/ir.c	2005-09-03 15:44:41.000000000 +0000
@@ -16,10 +16,10 @@
 #include "ir.h"
 
 /* Important stuff to know */
-static gboolean keepGoing = FALSE;
+gboolean keepGoing = FALSE;
 
 /* The thread handle */
-static pthread_t irapp_thread;
+pthread_t irapp_thread;
 
 /* Declarations for calls that we need to mention in the plugin struct */
 static void init(void);
