--- ./dist/Makefile.in.orig	2005-05-20 23:35:22.088112864 -0400
+++ ./dist/Makefile.in	2005-05-20 23:36:52.062434688 -0400
@@ -11,6 +11,7 @@
 bindir=	@bindir@
 includedir=@includedir@
 libdir=	@libdir@
+libjdir=@libdir@
 docdir=	$(prefix)/docs
 
 dmode=	755
@@ -437,17 +438,17 @@
 	@cd $(libdir) && $(chmod) $(fmode) $(libtso) $(libtso_target)
 
 install_java:
-	@echo "Installing DB Java library: $(libdir) ..."
-	@test -d $(libdir) || \
-	    ($(mkdir) -p $(libdir) && $(chmod) $(dmode) $(libdir))
-	@cd $(libdir) && $(rm) -f $(libjso_g) \
+	@echo "Installing DB Java library: $(libjdir) ..."
+	@test -d $(libjdir) || \
+	    ($(mkdir) -p $(libjdir) && $(chmod) $(dmode) $(libjdir))
+	@cd $(libjdir) && $(rm) -f $(libjso_g) \
 	    $(libjso) $(libjso_target) $(libjso_default) $(libjso_major)
-	@$(INSTALLER) $(libjso_target) $(libdir)
-	@cd $(libdir) && $(ln) -s $(libjso) $(libjso_default)
-	@cd $(libdir) && $(ln) -s $(libjso) $(libjso_major)
-	@cd $(libdir) && $(ln) -s $(libjso) $(libjso_g)
-	@cd $(libdir) && $(chmod) $(fmode) $(libjso) $(libjso_target)
-	@$(cp) $(JAVA_CLASSTOP)/$(libj_jarfile) $(libdir)
+	@$(INSTALLER) $(libjso_target) $(libjdir)
+	@cd $(libjdir) && $(ln) -s $(libjso) $(libjso_default)
+	@cd $(libjdir) && $(ln) -s $(libjso) $(libjso_major)
+	@cd $(libjdir) && $(ln) -s $(libjso) $(libjso_g)
+	@cd $(libjdir) && $(chmod) $(fmode) $(libjso) $(libjso_target)
+	@$(cp) $(JAVA_CLASSTOP)/$(libj_jarfile) $(libjdir)
 
 install_utilities:
 	@echo "Installing DB utilities: $(bindir) ..."
