--- ogrenew/OgreMain/include/OgreImage.h.orig	2006-03-14 13:17:13.000000000 -0800
+++ ogrenew/OgreMain/include/OgreImage.h	2006-03-14 13:18:28.000000000 -0800
@@ -136,7 +136,7 @@
 			@remarks 
 				The size of the buffer must be numFaces*PixelUtil::getMemorySize(width, height, depth, format)
          */
-		Image& Image::loadDynamicImage( uchar* pData, size_t uWidth, size_t uHeight, 
+		Image& loadDynamicImage( uchar* pData, size_t uWidth, size_t uHeight, 
 							size_t depth,
 							 PixelFormat eFormat, bool autoDelete = false, 
 							 size_t numFaces = 1, size_t numMipMaps = 0);
--- ogrenew/OgreMain/include/OgreEntity.h.orig	2006-03-14 13:21:11.000000000 -0800
+++ ogrenew/OgreMain/include/OgreEntity.h	2006-03-14 13:21:28.000000000 -0800
@@ -568,7 +568,7 @@
 	
         /** Stops sharing the SkeletonInstance with other entities.
         */
-        void Entity::stopSharingSkeletonInstance();
+        void stopSharingSkeletonInstance();
 
 
         /**
--- ogrenew/OgreMain/include/OgreHardwareOcclusionQuery.h.orig	2006-03-14 13:23:45.000000000 -0800
+++ ogrenew/OgreMain/include/OgreHardwareOcclusionQuery.h	2006-03-14 13:24:30.000000000 -0800
@@ -113,7 +113,7 @@
 	  * Lets you know when query is done, or still be processed by the Hardware
 	  * @return true if query isn't finished.
 	  */
-	 virtual bool HardwareOcclusionQuery::isStillOutstanding(void) = 0; 
+	 virtual bool isStillOutstanding(void) = 0; 
 	/**
 	  *   
 	  * @Remarks This function allows you to set how often the hardware occlusion really are sent to the driver
--- ogrenew/PlugIns/OctreeSceneManager/include/OgreOctreeSceneManager.h.orig	2006-03-14 15:41:27.000000000 -0800
+++ ogrenew/PlugIns/OctreeSceneManager/include/OgreOctreeSceneManager.h	2006-03-14 15:41:45.000000000 -0800
@@ -193,8 +193,8 @@
     /** Overridden from SceneManager */
     void clearScene(void);
 
-    AxisAlignedBoxSceneQuery* OctreeSceneManager::createAABBQuery(const AxisAlignedBox& box, unsigned long mask);
-    SphereSceneQuery* OctreeSceneManager::createSphereQuery(const Sphere& sphere, unsigned long mask);
+    AxisAlignedBoxSceneQuery* createAABBQuery(const AxisAlignedBox& box, unsigned long mask);
+    SphereSceneQuery* createSphereQuery(const Sphere& sphere, unsigned long mask);
     PlaneBoundedVolumeListSceneQuery* createPlaneBoundedVolumeQuery(const PlaneBoundedVolumeList& volumes, unsigned long mask);
     RaySceneQuery* createRayQuery(const Ray& ray, unsigned long mask);
     IntersectionSceneQuery* createIntersectionQuery(unsigned long mask);
--- ogrenew/Tools/XMLConverter/include/tinyxml.h.orig	2006-03-14 20:51:41.000000000 -0800
+++ ogrenew/Tools/XMLConverter/include/tinyxml.h	2006-03-14 20:52:52.000000000 -0800
@@ -790,7 +790,7 @@
 #endif
 
 	/// Construct.
-	TiXmlDeclaration::TiXmlDeclaration( const char * _version,
+	TiXmlDeclaration( const char * _version,
 										const char * _encoding,
 										const char * _standalone );
 
