diff -Naur inkscapelite-0.36.3ORIG/src/dialogs/tool-attributes.c inkscapelite-0.36.3/src/dialogs/tool-attributes.c
--- inkscapelite-0.36.3ORIG/src/dialogs/tool-attributes.c	2003-10-31 14:28:39.000000000 +0800
+++ inkscapelite-0.36.3/src/dialogs/tool-attributes.c	2014-12-10 23:02:30.000000000 +0800
@@ -13,7 +13,8 @@
 
 #include <config.h>
 #include <string.h>
-#include <glib/gtypes.h>
+/*BK#include <glib/gtypes.h>*/
+#include <glib.h>
 #include <gtk/gtksignal.h>
 #include <gtk/gtklabel.h>
 #include "macros.h"
