diff -rduN scummvm-old/base/main.cpp scummvm-new/base/main.cpp
--- scummvm-old/base/main.cpp	2004-03-14 14:45:12.000000000 +0100
+++ scummvm-new/base/main.cpp	2004-05-24 00:15:18.000000000 +0200
@@ -262,7 +262,7 @@
 
 	// Free up memory
 	delete engine;
-};
+}
 
 #ifndef _WIN32_WCE
 int main(int argc, char *argv[]) {
diff -rduN scummvm-old/scumm/actor.cpp scummvm-new/scumm/actor.cpp
--- scummvm-old/scumm/actor.cpp	2004-02-25 05:37:28.000000000 +0100
+++ scummvm-new/scumm/actor.cpp	2004-05-23 21:36:52.000000000 +0200
@@ -789,14 +789,14 @@
 		return _V1_talkingActor;
 	else
 		return VAR(VAR_TALK_ACTOR);
-};
+}
 
 void ScummEngine::talkingActor(int value) {
 	if (_gameId == GID_MANIAC && _version == 1)
 		_V1_talkingActor = value;
 	else
 		VAR(VAR_TALK_ACTOR) = value;
-};
+}
 
 void ScummEngine::showActors() {
 	int i;
diff -rduN scummvm-old/scumm/debugger.cpp scummvm-new/scumm/debugger.cpp
--- scummvm-old/scumm/debugger.cpp	2004-02-13 11:26:40.000000000 +0100
+++ scummvm-new/scumm/debugger.cpp	2004-05-23 21:37:21.000000000 +0200
@@ -57,7 +57,7 @@
 	va_end(va);
 
 	debug(buf);
-};
+}
 	
 ScummDebugger::ScummDebugger(ScummEngine *s)
 	: Common::Debugger<ScummDebugger>() {
@@ -592,7 +592,7 @@
 
 	DebugPrintf("Unknown flag. Type 'Debug ?' for syntax\n");
 	return true;
-};
+}
 
 bool ScummDebugger::Cmd_DebugLevel(int argc, const char **argv) {
 	if (argc == 1) {
