--- ./tools/Makefile.in.orig	2005-02-18 11:34:25.440484160 -0300
+++ ./tools/Makefile.in	2005-02-18 11:44:34.835841984 -0300
@@ -30,6 +30,7 @@
 srcdir = @srcdir@
 VPATH = @srcdir@
 STATIC_LINK = @STATIC_LINK@
+SUFFIX =
 
 # Link decision (archive or shared library)
 # See './configure --help' for info on setting this - default is dynamic
@@ -74,15 +75,15 @@
 	@echo "*** Installing $(ALL_TARGETS) $(ALL_SCRIPTS) in $(DESTDIR)$(sbindir) ***"
 	@for f in $(ALL_TARGETS); \
 	do \
-  	   rm -f $(DESTDIR)$(sbindir)/$$f; \
+  	   rm -f $(DESTDIR)$(sbindir)/$$f$(SUFFIX); \
 	   @INSTALL@ -d $(DESTDIR)$(sbindir); \
-	   @INSTALL@ -o $(OWNER) -g $(GROUP) -m 555 -s $$f $(DESTDIR)$(sbindir)/$$f; \
+	   @INSTALL@ -o $(OWNER) -g $(GROUP) -m 555 -s $$f $(DESTDIR)$(sbindir)/$$f$(SUFFIX); \
 	done
 	@for f in $(ALL_SCRIPTS); \
 	do \
-	   rm -f $(DESTDIR)$(sbindir)/$$f; \
+	   rm -f $(DESTDIR)$(sbindir)/$$f$(SUFFIX); \
 	   @INSTALL@ -d $(DESTDIR)$(sbindir); \
-	   @INSTALL@ -o $(OWNER) -g $(GROUP) -m 555 $(srcdir)/$$f $(DESTDIR)$(sbindir)/$$f; \
+	   @INSTALL@ -o $(OWNER) -g $(GROUP) -m 555 $(srcdir)/$$f $(DESTDIR)$(sbindir)/$$f$(SUFFIX); \
 	done
 
 remove-bin:	$(ALL_TARGETS) $(ALL_SCRIPTS)
