diff -dur vamps-0.95/Makefile vamps-0.95-patched/Makefile
--- vamps-0.95/Makefile	2003-12-15 07:55:45.000000000 +0100
+++ vamps-0.95-patched/Makefile	2004-05-18 22:35:37.000000000 +0200
@@ -14,7 +14,7 @@
 
 CC     = gcc
 CFLAGS = -Wall -O3 -fomit-frame-pointer -fexpensive-optimizations \
-         -funroll-loops -march=i686 -mcpu=i686 -malign-functions=4
+        -funroll-loops -march=i686 -mcpu=i686 -falign-functions=4
 STATIC =
 #STATIC = -static
 STRIP  = strip
@@ -23,7 +23,7 @@
 all:			vamps vamps-play_title lsdvd
 
 vamps:			vamps.o requant.o
-			$(CC) -o $@ $(STATIC) vamps.o requant.o -lpthread
+			$(CC) -o $@ $(STATIC) vamps.o requant.o -lpthread -lm
 			$(STRIP) $@
 
 requant.o:		requant.h putvlc.h getvlc.h
Nur in vamps-0.95-patched: Makefile~.
