--- mozilla/gfx/public/nsRegion.h	2003-03-25 03:57:42.000000000 +0100
+++ mozilla-fixed/gfx/public/nsRegion.h	2004-05-05 23:52:57.000000000 +0200
@@ -233,8 +233,9 @@
   void MoveInto (nsRegion& aDestRegion, const RgnRect* aStartRect);
   void MoveInto (nsRegion& aDestRegion)
   {    MoveInto (aDestRegion, mRectListHead.next);  }
-
+public:
   nsRegion (const nsRegion& aRegion);       // Prevent copying of regions
+private:
   nsRegion& operator = (const nsRegion& aRegion);
 };
 
