diff -Nur motif-vanilla/config/Imakefile motif/config/Imakefile
--- motif-vanilla/config/Imakefile	2003-11-22 15:54:21.032310256 +0100
+++ motif/config/Imakefile	2003-11-22 15:55:04.876644904 +0100
@@ -5,8 +5,7 @@
 #undef BootstrapCleanSubdirs
 #define BootstrapCleanSubdirs BOOTSTRAPSUBDIRS="$(BOOTSTRAPSUBDIRS)"
 
-BOOTSTRAPSUBDIRS = imake makedepend
-SUBDIRS = cf $(BOOTSTRAPSUBDIRS) util
+SUBDIRS = 
 
 MakeSubdirs($(SUBDIRS))
 DependSubdirs($(SUBDIRS))
