--- xfig.3.2.4/u_fonts.h.vanilla	2005-09-02 17:50:09.000000000 +0000
+++ xfig.3.2.4/u_fonts.h	2005-09-02 17:51:13.000000000 +0000
@@ -32,9 +32,6 @@
 
 extern int		psfontnum();
 extern int		latexfontnum();
-extern struct _xfstruct	x_fontinfo[], x_backup_fontinfo[];
-extern struct _fstruct	ps_fontinfo[];
-extern struct _fstruct	latex_fontinfo[];
 
 /* element of linked list for each font
    The head of list is for the different font NAMES,
@@ -62,4 +59,9 @@
 };
 
 int		x_fontnum();
+
+extern struct _xfstruct x_fontinfo[], x_backup_fontinfo[];
+extern struct _fstruct  ps_fontinfo[];
+extern struct _fstruct  latex_fontinfo[];
+
 #endif /* U_FONTS_H */
