--- clisp-2.33.2/modules/bindings/glibc/linux.lisp~	2004-03-10 18:52:06.000000000 +0100
+++ clisp-2.33.2/modules/bindings/glibc/linux.lisp	2005-01-03 02:52:22.634096240 +0100
@@ -2489,7 +2489,7 @@
 ;; miscellaneous signal stuff
 (def-call-out kill (:arguments (pid pid_t) (sig int)) (:return-type int))
 (def-call-out raise (:arguments (sig int)) (:return-type int))
-(def-call-out sigpause (:arguments (sig int)) (:return-type int))
+(def-call-out sigsuspend (:arguments (sig int)) (:return-type int))
 (def-call-out killpg (:arguments (pgrp pid_t) (sig int))
   (:return-type int))
 
