diff -ur rt2570_orig/Module/Makefile rt2570/Module/Makefile
--- rt2570_orig/Module/Makefile	2000-01-01 00:00:00.000000000 +0100
+++ rt2570/Module/Makefile	2005-07-20 14:27:11.560244512 +0200
@@ -64,7 +64,7 @@
  KERNEL_SOURCES := /lib/modules/$(shell uname -r)/build
 endif
 
-MODULE_ROOT := /lib/modules/$(shell uname -r)/extra
+MODULE_ROOT := $(MODDIR)/extra
 
 src ?= .
 obj ?= .
@@ -104,18 +104,13 @@
 	@rm -fr .tmp_versions
 
 modules_install:
-ifeq ($(PATCHLEVEL),4)
 	if ! [ -f $(MODULE_OBJECT) ]; then \
 	module; \
 	fi
 	@echo "install '$(MODULE_OBJECT)' to $(MODULE_ROOT)"
 	install -m 755 -o 0 -g 0 -d $(MODULE_ROOT)
 	install -m 644 -o 0 -g 0 $(MODULE_OBJECT) $(MODULE_ROOT)
-	/sbin/depmod -a
-else
-	echo "2.6 module install"
-	make $(KBUILD_PARAMS) modules_install
-endif
+
 
 install: modules_install
 	@if ! grep -q 'rausb0' $(MODULE_CONF) ; then \
Only in rt2570/Module: Makefile~
