--- ./modules/ssl/ssl_engine_pphrase.c.orig	2005-07-14 23:32:39.040813840 -0400
+++ ./modules/ssl/ssl_engine_pphrase.c	2005-07-14 23:35:34.285860200 -0400
@@ -30,6 +30,13 @@
                                            -- Clifford Stoll     */
 #include "mod_ssl.h"
 
+#ifndef PEM_F_DEF_CALLBACK
+# ifdef PEM_F_PEM_DEF_CALLBACK
+/* In OpenSSL 0.9.8 PEM_F_DEF_CALLBACK was renamed */
+#  define PEM_F_DEF_CALLBACK PEM_F_PEM_DEF_CALLBACK
+# endif
+#endif
+
 /*
  * Return true if the named file exists and is readable
  */
