--- pdksh-5.2.14/configure	1999-04-20 22:01:59.000000000 +0200
+++ pdksh-5.2.14-patched/configure	2005-05-17 23:02:51.000000000 +0200
@@ -623,7 +623,7 @@
 fi
 
 case $enable_shell in
-  ksh) cat >> confdefs.h <<\EOF
+  *ksh) cat >> confdefs.h <<\EOF
 #define KSH 1
 EOF
  ;;
@@ -640,7 +640,7 @@
 fi
 
 case $enable_emacs:$enable_shell in
-  yes:*|:ksh) enable_emacs=yes; cat >> confdefs.h <<\EOF
+  yes:*|:*ksh) enable_emacs=yes; cat >> confdefs.h <<\EOF
 #define EMACS 1
 EOF
  ;;
@@ -654,7 +654,7 @@
 fi
 
 case $enable_vi:$enable_shell in
-  yes:*|:ksh) enable_vi=yes; cat >> confdefs.h <<\EOF
+  yes:*|:*ksh) enable_vi=yes; cat >> confdefs.h <<\EOF
 #define VI 1
 EOF
  ;;
@@ -682,7 +682,7 @@
 fi
 
 case $enable_brace_expand:$enable_shell in
-  yes:*|:ksh) enable_brace_expand=yes; cat >> confdefs.h <<\EOF
+  yes:*|:*ksh) enable_brace_expand=yes; cat >> confdefs.h <<\EOF
 #define BRACE_EXPAND 1
 EOF
  ;;
@@ -696,7 +696,7 @@
 fi
 
 case $enable_history:$enable_shell in
-  simple:*|:ksh) enable_history=simple; ;;
+  simple:*|:*ksh) enable_history=simple; ;;
   complex:*) enable_history=complex; cat >> confdefs.h <<\EOF
 #define COMPLEX_HISTORY 1
 EOF
