--- unionfs-20060503-0408/Makefile.vanilla	2006-07-01 16:35:28.000000000 +0000
+++ unionfs-20060503-0408/Makefile	2006-07-01 16:35:33.000000000 +0000
@@ -29,7 +29,7 @@
 
 CC	= gcc
 KERNELVERSION=$(shell echo $(KVERS) | cut -d. -f1,2)
-EXTRA_CFLAGS += ${TOPINC} -Wall -Werror ${EXTRACFLAGS} ${UNIONFS_DEBUG_CFLAG} ${UNIONFS_OPT_CFLAG} ${UNIONFS_VERSION_CFLAG}
+EXTRA_CFLAGS += ${TOPINC} -Wall ${EXTRACFLAGS} ${UNIONFS_DEBUG_CFLAG} ${UNIONFS_OPT_CFLAG} ${UNIONFS_VERSION_CFLAG}
 UCFLAGS = -I. ${UNIONFS_DEBUG_CFLAG} ${UNIONFS_OPT_CFLAG} -Wall -Werror ${EXTRAUCFLAGS}  ${UNIONFS_VERSION_CFLAG}
 
 obj-m := unionfs.o
--- union/mmap.c.vanilla	2006-07-01 17:46:08.000000000 +0000
+++ union/mmap.c	2006-07-01 17:46:29.000000000 +0000
@@ -293,13 +293,12 @@
  * This function is copied verbatim from mm/filemap.c.
  * XXX: It should be simply moved to some header file instead -- bug Al about it!
  */
-static inline int sync_page(struct page *page)
+static void sync_page(struct page *page)
 {
 	struct address_space *mapping = page->mapping;
 
 	if (mapping && mapping->a_ops && mapping->a_ops->sync_page)
-		return mapping->a_ops->sync_page(page);
-	return 0;
+		mapping->a_ops->sync_page(page);
 }
 
 int unionfs_sync_page(struct page *page)
@@ -320,7 +319,7 @@
 		err = -ENOMEM;
 		goto out;
 	}
-	err = sync_page(lower_page);
+	sync_page(lower_page);
 
 	unlock_page(lower_page);	/* b/c grab_cache_page locked it */
 	page_cache_release(lower_page);	/* b/c grab_cache_page increased refcnt */
--- union/Makefile.vanilla	2006-07-01 17:45:17.000000000 +0000
+++ union/Makefile	2006-07-01 17:45:26.000000000 +0000
@@ -30,7 +30,7 @@
 CC	= gcc
 KERNELVERSION=$(shell echo $(KVERS) | cut -d. -f1,2)
 EXTRA_CFLAGS += ${TOPINC} -Wall ${EXTRACFLAGS} ${UNIONFS_DEBUG_CFLAG} ${UNIONFS_OPT_CFLAG} ${UNIONFS_VERSION_CFLAG}
-UCFLAGS = -I. ${UNIONFS_DEBUG_CFLAG} ${UNIONFS_OPT_CFLAG} -Wall -Werror ${EXTRAUCFLAGS}  ${UNIONFS_VERSION_CFLAG}
+UCFLAGS = -I. ${UNIONFS_DEBUG_CFLAG} ${UNIONFS_OPT_CFLAG} -Wall ${EXTRAUCFLAGS}  ${UNIONFS_VERSION_CFLAG}
 
 obj-m := unionfs.o
 unionfs-objs := subr.o dentry.o file.o inode.o main.o super.o \
