diff -ur cvs/archivista/apcl/Archivista/Config.pm cvs-fixed/archivista/apcl/Archivista/Config.pm
--- cvs/archivista/apcl/Archivista/Config.pm	2005-07-19 10:26:41.000000000 +0000
+++ cvs-fixed/archivista/apcl/Archivista/Config.pm	2005-07-21 13:15:49.000000000 +0000
@@ -44,7 +44,7 @@
 	$self->{'MYSQL_HOST'} = "localhost";
 	$self->{'MYSQL_DB'} = "archivista";
 	$self->{'MYSQL_UID'} = "root";
-  $self->{'MYSQL_PWD'} = "";
+  	$self->{'MYSQL_PWD'} = "archivista";
   $self->{'MYSQL_BIN'} = "/usr/local/bin/mysql";
 	$self->{'MYSQL_DUMP'} = "/usr/local/bin/mysqldump --add-drop-table";
 	# Please note: the base path directory MUST be writable for the web user 
diff -ur cvs/archivista/webadmin/perl/ASConfig.pm cvs-fixed/archivista/webadmin/perl/ASConfig.pm
--- cvs/archivista/webadmin/perl/ASConfig.pm	2005-07-19 09:14:22.000000000 +0000
+++ cvs-fixed/archivista/webadmin/perl/ASConfig.pm	2005-07-21 14:01:46.000000000 +0000
@@ -41,8 +41,8 @@
   my $self = shift;
 
   # REQUIRED
-	$self->{'WWW_DIR'} = "";
-  $self->{'PERL_DIR'} = "/perl";
+	$self->{'WWW_DIR'} = "/webadmin";
+  $self->{'PERL_DIR'} = "/cgi-bin/webadmin";
 	$self->{'STYLES'} = "/css/styles.css";
   $self->{'LANGUAGES'} = "";
 	$self->{'DEFAULT_LOGIN_DB'} = "archivista";
diff -ur cvs/archivista/webclient/perl/frame.pl cvs-fixed/archivista/webclient/perl/frame.pl
--- cvs/archivista/webclient/perl/frame.pl	2005-07-19 09:26:04.000000000 +0000
+++ cvs-fixed/archivista/webclient/perl/frame.pl	2005-07-21 14:11:31.000000000 +0000
@@ -12,6 +12,8 @@
 # andere Frames starten, aufgerufen werden.
 ###
 
+use lib qw (/home/cvs/archivista/webclient/perl);
+
 use strict;
 use DBI;
 use Digest::MD5 qw(md5_hex);
diff -ur cvs/archivista/webclient/perl/inc/Global.pm cvs-fixed/archivista/webclient/perl/inc/Global.pm
--- cvs/archivista/webclient/perl/inc/Global.pm	2005-07-19 09:26:04.000000000 +0000
+++ cvs-fixed/archivista/webclient/perl/inc/Global.pm	2005-07-21 13:54:23.000000000 +0000
@@ -277,7 +277,7 @@
 
 # DocumentRoot des WebServers
 sub getDOCRoot {
-    return "/home/web/www";
+    return "/home/cvs/archivista/webclient/www";
 }
 
 # Felder Ordner und Eigentuemer ausblenden (1/0)
@@ -341,7 +341,7 @@
 }
 
 sub avdb_pwd {
-    return "";
+    return "archivista";
 }
 
 sub avdb_db {
diff -ur cvs/archivista/webclient/perl/index.pl cvs-fixed/archivista/webclient/perl/index.pl
--- cvs/archivista/webclient/perl/index.pl	2005-07-19 09:26:04.000000000 +0000
+++ cvs-fixed/archivista/webclient/perl/index.pl	2005-07-21 13:53:25.000000000 +0000
@@ -6,6 +6,7 @@
 
 # ============= Archivista Web Client --- v1.00 --- 2004-01-25 ============== 
 
+use lib qw (/home/cvs/archivista/webclient/perl);
 use strict;
 use inc::Main;
 
diff -ur cvs/archivista/webclient/perl/leer.pl cvs-fixed/archivista/webclient/perl/leer.pl
--- cvs/archivista/webclient/perl/leer.pl	2005-07-19 09:26:04.000000000 +0000
+++ cvs-fixed/archivista/webclient/perl/leer.pl	2005-07-21 14:13:13.000000000 +0000
@@ -6,6 +6,8 @@
 
 # ============= Archivista Web Client --- v1.00 --- 2004-01-25 ============== 
 
+use lib qw (/home/cvs/archivista/webclient/perl);
+
 use strict;
 use inc::HTML;
 
diff -ur cvs-old/archivista/webclient/perl/inc/Global.pm cvs/archivista/webclient/perl/inc/Global.pm
--- cvs-old/archivista/webclient/perl/inc/Global.pm	2005-07-22 18:15:46.000000000 +0000
+++ cvs/archivista/webclient/perl/inc/Global.pm	2005-07-22 14:01:02.000000000 +0000
@@ -59,7 +59,7 @@
 
 # Dieser user wird, sofern gesetzt standardmässig bei der login maske angezeigt
 sub defaultLoginUser {
-  return "SYSOP";
+  return "Admin";
 }
 
 # Datenbank der jobs tabelle
@@ -113,7 +113,7 @@
 
 # Nur localhost zugriffe zulassen 
 sub onlyLocalhost {
-    return 0;
+    return 1;
 }
 
 # Detail Frame anzeigen (1 = ja / 0 = nein)
--- cvs-old/archivista/webclient/perl/inc/Global.pm.old	2005-08-11 12:59:49.000000000 +0000
+++ cvs-old/archivista/webclient/perl/inc/Global.pm	2005-08-11 13:01:45.000000000 +0000
@@ -345,11 +345,11 @@
 }
 
 sub avdb_db {
-    return "avclient";
+    return "archivista";
 }
 
 sub avdb_table {
-    return "avclient";
+    return "sessionweb";
 }
 
 sub mysql_port {
