diff -urN hotplug-2002_08_26/Makefile hotplug-2002_08_26-mine/Makefile
--- hotplug-2002_08_26/Makefile	Tue Mar 26 23:13:35 2002
+++ hotplug-2002_08_26-mine/Makefile	Wed Apr 16 14:39:13 2003
@@ -75,9 +75,9 @@
 install: installdirs
 	for i in sbin/*;do [[ -d $$i ]] && continue;install -m755 $$i $(sbindir)/;done
 	for i in etc/hotplug/*;do [[ $$i == *CVS* ]] && continue;cp -a -r $$i $(etcdir)/hotplug;done
-	for i in etc/rc.d/init.d/*;do [[ $$i == *CVS* ]] && continue;cp -a -r $$i $(etcdir)/rc.d/init.d/;done
+	chown -R root.root $(etcdir)/hotplug 
 	$(INSTALL_DATA) hotplug.8  $(mandir)/man8/hotplug.8
 
 installdirs: mkinstalldirs
-	$(srcdir)/mkinstalldirs $(sbindir) $(etcdir)/hotplug $(etcdir)/rc.d/init.d $(mandir)/man8 $(prefix)/var/run/usb
+	$(srcdir)/mkinstalldirs $(sbindir) $(etcdir)/hotplug $(mandir)/man8 $(prefix)/var/run/usb
 	chmod 0700 $(prefix)/var/run/usb
