--- gnome-icon-theme-2.16.0/Makefile.in.orig	2006-09-03 19:19:36.000000000 -0700
+++ gnome-icon-theme-2.16.0/Makefile.in	2006-09-08 10:23:46.000000000 -0700
@@ -199,7 +199,7 @@
 target_alias = @target_alias@
 themedir = @themedir@
 SUBDIRS = 8x8 16x16 22x22 24x24 32x32 48x48 scalable po
-pkgconfigdir = $(datadir)/pkgconfig
+pkgconfigdir = $(libdir)/pkgconfig
 pkgconfig_DATA = gnome-icon-theme.pc
 THEME_IN_FILES = index.theme.in
 theme_DATA = index.theme
