--- k3b-0.11.21/src/audiodecoding/flac/k3bflacdecoder.cpp.orig	2005-03-12 17:52:22.000000000 +0100
+++ k3b-0.11.21/src/audiodecoding/flac/k3bflacdecoder.cpp	2005-03-16 10:23:19.000000000 +0100
@@ -310,7 +310,7 @@
   if( d->comments != 0 ) {
     if( info == i18n("Vendor") )
 #ifdef FLAC_NEWER_THAN_1_1_1
-      return QString::fromUtf8(d->comments->get_vendor_string());
+      return QString::fromUtf8( (const char *) d->comments->get_vendor_string() );
 #else
       return QString::fromUtf8(d->comments->get_vendor_string().get_field());
 #endif
