--- hydrogen-0.9.3/plugins/wasp/Makefile.in_orig	2006-07-22 10:45:25.000000000 +0100
+++ hydrogen-0.9.3/plugins/wasp/Makefile.in	2006-07-22 14:06:34.000000000 +0100
@@ -23,9 +23,9 @@
 
 install:
 	@echo "Installing WASP plugins"
-	mkdir -p $(DESTDIR)$(prefix)/lib/hydrogen/plugins/
-	@install -vD *.so $(DESTDIR)$(prefix)/lib/hydrogen/plugins
+	mkdir -p $(LIBDIR)/hydrogen/plugins/
+	@install -vD *.so $(LIBDIR)/hydrogen/plugins
 
 uninstall:
-	@rm -rf $(DESTDIR)$(prefix)/lib/hydrogen
+	@rm -rf $(LIBDIR)/hydrogen
 	
