--- wavemon-0.4.0b/llist.c.vanilla	2005-09-04 16:54:40.000000000 +0000
+++ wavemon-0.4.0b/llist.c	2005-09-04 16:58:51.000000000 +0000
@@ -352,7 +352,7 @@
 					*((int *)l->e) = va_arg(*ap, int);
 					break;
 		case 's':	l->e = (void *)malloc(sizeof(char *));
-					(char *)l->e = strdup(va_arg(*ap, char *));
+					/*(char *)*/l->e = strdup(va_arg(*ap, char *));
 					break;
 		case 'f':	l->e = (void *)malloc(sizeof(double));
 					*((double *)l->e) = va_arg(*ap, double);
