--- hotplug++-0.0.5/Makefile.vanilla	2006-03-29 12:32:35.902220250 +0200
+++ hotplug++-0.0.5/Makefile	2006-03-29 12:33:47.378687250 +0200
@@ -1,14 +1,10 @@
 
 # the use of gcc for C++ files is intentional in this file!
 
-DIET		= diet
-ifeq ($(DIET),)
-CC		= gcc
-CXX		= gcc -fno-exceptions -fno-rtti -fno-use-cxa-atexit
-else
-CC		= $(DIET) gcc
-CXX		= $(DIET) gcc -fno-exceptions -fno-rtti -fno-use-cxa-atexit
-endif
+CC		= diet gcc
+CXX		= diet gcc
+
+override CXX	+= -fno-exceptions -fno-rtti -fno-use-cxa-atexit
 
 #LIBS		= -lsupc++ -lc -lgcc
 CXXFLAGS = -Os -s
