--- firefox-107.0/build/moz.configure/init.configure.vanilla	2022-11-17 17:55:31.419221937 +0100
+++ firefox-107.0/build/moz.configure/init.configure	2022-11-17 17:57:55.802208026 +0100
@@ -529,7 +529,7 @@
         endianness = "big"
     elif cpu.startswith("arm"):
         canonical_cpu = "arm"
-        endianness = "big" if cpu.startswith(("armeb", "armbe")) else "little"
+        endianness = "big" if cpu.endswith(("eb", "be")) else "little"
     elif cpu in ("m68k"):
         canonical_cpu = "m68k"
         endianness = "big"
--- firefox-107.0/third_party/libwebrtc/build/build_config.h.vanilla	2022-11-17 18:02:57.834178924 +0100
+++ firefox-107.0/third_party/libwebrtc/build/build_config.h	2022-11-17 18:03:29.928175831 +0100
@@ -165,6 +165,11 @@
 #define ARCH_CPU_ARMEL 1
 #define ARCH_CPU_32_BITS 1
 #define ARCH_CPU_LITTLE_ENDIAN 1
+#elif defined(__ARMEB__)  
+#define ARCH_CPU_ARM_FAMILY 1
+#define ARCH_CPU_ARMEL 1
+#define ARCH_CPU_32_BITS 1
+#define ARCH_CPU_BIG_ENDIAN 1
 #elif defined(__aarch64__) || defined(_M_ARM64)
 #define ARCH_CPU_ARM_FAMILY 1
 #define ARCH_CPU_ARM64 1
--- firefox-107.0/ipc/chromium/src/build/build_config.h.vanilla	2022-11-17 17:59:57.767196274 +0100
+++ firefox-107.0/ipc/chromium/src/build/build_config.h	2022-11-17 18:00:45.275191696 +0100
@@ -75,7 +75,7 @@
 #  define ARCH_CPU_X86_FAMILY 1
 #  define ARCH_CPU_X86 1
 #  define ARCH_CPU_32_BITS 1
-#elif defined(__ARMEL__)
+#elif defined(__ARMEL__) || defined(__ARMEB__)
 #  define ARCH_CPU_ARM_FAMILY 1
 #  define ARCH_CPU_ARMEL 1
 #  define ARCH_CPU_32_BITS 1
