--- firefox-142.0/gfx/harfbuzz/src/hb.hh.vanilla	2025-08-24 12:01:24.609021283 +0200
+++ firefox-142.0/gfx/harfbuzz/src/hb.hh	2025-08-24 12:01:35.204416340 +0200
@@ -89,7 +89,6 @@
 #pragma GCC diagnostic error   "-Wstring-conversion"
 #pragma GCC diagnostic error   "-Wswitch-enum"
 #pragma GCC diagnostic error   "-Wtautological-overlap-compare"
-#pragma GCC diagnostic error   "-Wuninitialized"
 #pragma GCC diagnostic error   "-Wunneeded-internal-declaration"
 #pragma GCC diagnostic error   "-Wunused"
 #pragma GCC diagnostic error   "-Wunused-local-typedefs"
