--- firefox-122.0.1/python/mozbuild/mozbuild/configure/constants.py.vanilla	2024-02-17 18:20:27.661337485 +0100
+++ firefox-122.0.1/python/mozbuild/mozbuild/configure/constants.py	2024-02-17 18:24:45.834312609 +0100
@@ -126,8 +126,8 @@
         ("sparc64", "__sparc__ && __arch64__"),
         ("sparc", "__sparc__"),
         ("m68k", "__m68k__"),
-        ("mips64", "__mips64"),
-        ("mips32", "__mips__"),
+        ("mips64", "__mips64 && _ABI64"),
+        ("mips32", "__mips__ || (__mips64 && _ABIN32)"),
         ("riscv64", "__riscv && __riscv_xlen == 64"),
         ("loongarch64", "__loongarch64"),
         ("sh4", "__sh__"),
--- firefox-122.0.1/build/moz.configure/init.configure.vanilla	2024-02-17 18:15:36.219365567 +0100
+++ firefox-122.0.1/build/moz.configure/init.configure	2024-02-17 18:17:42.422353407 +0100
@@ -579,7 +579,7 @@
         canonical_cpu = "mips32"
         endianness = "little" if "el" in cpu else "big"
     elif cpu in ("mips64", "mips64el"):
-        canonical_cpu = "mips64"
+        canonical_cpu = "mips32"
         endianness = "little" if "el" in cpu else "big"
     elif cpu.startswith("aarch64"):
         canonical_cpu = "aarch64"
