--- firefox-57.0/security/sandbox/linux/broker/SandboxBrokerPolicyFactory.cpp.vanilla	2017-11-16 17:56:45.091951171 +0100
+++ firefox-57.0/security/sandbox/linux/broker/SandboxBrokerPolicyFactory.cpp	2017-11-16 17:57:11.287951481 +0100
@@ -161,7 +161,8 @@
   policy->AddDir(rdonly, "/usr/tmp");
   policy->AddDir(rdonly, "/var/tmp");
   // Various places where fonts reside
-  policy->AddDir(rdonly, "/usr/X11R6/lib/X11/fonts");
+  policy->AddDir(rdonly, "/usr/X11/share/X11/fonts");
+  policy->AddDir(rdonly, "/usr/X11/lib/X11/fonts");
   policy->AddDir(rdonly, "/nix/store");
   policy->AddDir(rdonly, "/run/host/fonts");
   policy->AddDir(rdonly, "/run/host/user-fonts");
