--- firefox-127.0/security/sandbox/linux/broker/SandboxBrokerPolicyFactory.cpp.vanilla	2024-06-14 14:48:51.851243892 +0200
+++ firefox-127.0/security/sandbox/linux/broker/SandboxBrokerPolicyFactory.cpp	2024-06-14 15:24:11.735169676 +0200
@@ -446,6 +446,9 @@
   policy->AddDir(rdonly, "/usr/lib");
   policy->AddDir(rdonly, "/usr/lib32");
   policy->AddDir(rdonly, "/usr/lib64");
+  policy->AddDir(rdonly, "/usr/X11/lib");
+  policy->AddDir(rdonly, "/usr/X11/lib32");
+  policy->AddDir(rdonly, "/usr/X11/lib64");
   policy->AddDir(rdonly, "/etc");
   policy->AddDir(rdonly, "/usr/share");
   policy->AddDir(rdonly, "/usr/local/share");
@@ -461,7 +462,6 @@
 
   // Bug 1848615
   policy->AddPath(rdonly, "/usr");
-  policy->AddPath(rdonly, "/nix");
 
   AddLdconfigPaths(policy);
   AddLdLibraryEnvPaths(policy);
@@ -984,6 +984,9 @@
   policy->AddDir(rdonly, "/usr/lib");
   policy->AddDir(rdonly, "/usr/lib32");
   policy->AddDir(rdonly, "/usr/lib64");
+  policy->AddDir(rdonly, "/usr/X11/lib");
+  policy->AddDir(rdonly, "/usr/X11/lib32");
+  policy->AddDir(rdonly, "/usr/X11/lib64");
   policy->AddDir(rdonly, "/usr/share");
   policy->AddDir(rdonly, "/usr/local/share");
   policy->AddDir(rdonly, "/etc");
@@ -1038,6 +1041,9 @@
   policy->AddDir(rdonly, "/usr/lib");
   policy->AddDir(rdonly, "/usr/lib32");
   policy->AddDir(rdonly, "/usr/lib64");
+  policy->AddDir(rdonly, "/usr/X11/lib");
+  policy->AddDir(rdonly, "/usr/X11/lib32");
+  policy->AddDir(rdonly, "/usr/X11/lib64");
   policy->AddDir(rdonly, "/usr/share");
   policy->AddDir(rdonly, "/usr/local/share");
   policy->AddDir(rdonly, "/etc");
