diff --git a/arch/mips/Kbuild.platforms b/arch/mips/Kbuild.platforms
index ac7ad54f984f..50c7b8e68547 100644
--- a/arch/mips/Kbuild.platforms
+++ b/arch/mips/Kbuild.platforms
@@ -27,6 +27,7 @@ platforms += pmcs-msp71xx
 platforms += pnx833x
 platforms += ralink
 platforms += rb532
+platforms += sgi-common
 platforms += sgi-ip22
 platforms += sgi-ip27
 platforms += sgi-ip32
diff --git a/arch/mips/Kconfig b/arch/mips/Kconfig
index 08c10c518f83..8708bf24c359 100644
--- a/arch/mips/Kconfig
+++ b/arch/mips/Kconfig
@@ -620,6 +620,7 @@ config RALINK
 
 config SGI_IP22
 	bool "SGI IP22 (Indy/Indigo2)"
+	select MIPS_SGI
 	select FW_ARC
 	select FW_ARC32
 	select ARCH_MIGHT_HAVE_PC_SERIO
@@ -662,6 +663,7 @@ config SGI_IP22
 
 config SGI_IP27
 	bool "SGI IP27 (Origin200/2000)"
+	select MIPS_SGI
 	select FW_ARC
 	select FW_ARC64
 	select BOOT_ELF64
@@ -683,6 +685,7 @@ config SGI_IP27
 
 config SGI_IP28
 	bool "SGI IP28 (Indigo2 R10k)"
+	select MIPS_SGI
 	select FW_ARC
 	select FW_ARC64
 	select ARCH_MIGHT_HAVE_PC_SERIO
@@ -721,6 +724,7 @@ config SGI_IP28
 
 config SGI_IP32
 	bool "SGI IP32 (O2)"
+	select MIPS_SGI
 	select FW_ARC
 	select FW_ARC32
 	select BOOT_ELF32
@@ -1283,6 +1287,9 @@ config MIPS_SPRAM
 config SWAP_IO_SPACE
 	bool
 
+config MIPS_SGI
+	bool
+
 config SGI_HAS_INDYDOG
 	bool
 
diff --git a/arch/mips/sgi-common/Makefile b/arch/mips/sgi-common/Makefile
new file mode 100644
index 000000000000..bb8e588093eb
--- /dev/null
+++ b/arch/mips/sgi-common/Makefile
@@ -0,0 +1,3 @@
+#
+# Makefile for common bits shared by various SGI systems
+#
diff --git a/arch/mips/sgi-common/Platform b/arch/mips/sgi-common/Platform
new file mode 100644
index 000000000000..142a76a2880b
--- /dev/null
+++ b/arch/mips/sgi-common/Platform
@@ -0,0 +1 @@
+platform-$(CONFIG_MIPS_SGI)		+= sgi-common/
