diff --git a/arch/mips/include/asm/xtalk/xtalk.h b/arch/mips/include/asm/xtalk/xtalk.h
index 9125bd85514d..627ed91b2880 100644
--- a/arch/mips/include/asm/xtalk/xtalk.h
+++ b/arch/mips/include/asm/xtalk/xtalk.h
@@ -21,24 +21,15 @@
 #define XWIDGET_MFG_NUM_NONE	-1
 
 /* It is often convenient to fold the XIO target port */
-#define XIO_NOWHERE	(0xFFFFFFFFFFFFFFFFull)
-#define XIO_ADDR_BITS	(0x0000FFFFFFFFFFFFull)
-#define XIO_PORT_BITS	(0xF000000000000000ull)
+#define XIO_NOWHERE	(0xffffffffffffffffULL)
+#define XIO_ADDR_BITS	(0x0000ffffffffffffULL)
+#define XIO_PORT_BITS	(0xf000000000000000ULL)
 #define XIO_PORT_SHIFT	(60)
 
-#define XIO_PACKED(x)	(((x)&XIO_PORT_BITS) != 0)
-#define XIO_ADDR(x)	((x)&XIO_ADDR_BITS)
-#define XIO_PORT(x)	((s8)(((x)&XIO_PORT_BITS) >> XIO_PORT_SHIFT))
-#define XIO_PACK(p, o)	((((uint64_t)(p))<<XIO_PORT_SHIFT) | ((o)&XIO_ADDR_BITS))
-
-#ifdef CONFIG_PCI
-extern int bridge_probe(nasid_t nasid, int widget, int masterwid);
-#else
-static inline int bridge_probe(nasid_t nasid, int widget, int masterwid)
-{
-	return 0;
-}
-#endif
+#define XIO_PACKED(x)	(((x) & XIO_PORT_BITS) != 0)
+#define XIO_ADDR(x)	((x) & XIO_ADDR_BITS)
+#define XIO_PORT(x)	((s8)(((x) & XIO_PORT_BITS) >> XIO_PORT_SHIFT))
+#define XIO_PACK(p, o)	((((u64)(p)) << XIO_PORT_SHIFT) | ((o) & XIO_ADDR_BITS))
 
 #endif /* !__ASSEMBLY__ */
 
