--- torcs-1.2.2/Make-default.mk.orig	2005-01-02 23:38:15.724695600 +0100
+++ torcs-1.2.2/Make-default.mk	2005-01-02 23:52:18.544567504 +0100
@@ -134,7 +134,10 @@
 endif
 
 .depend:
-	$(CPP) $(INCFLAGS) $(CFLAGSD) $(COMPILFLAGS)  $(SOURCES) > .depend
+	rm -f .depend; \
+	list="$(SOURCES)" ; for i in $$list; do \
+		$(CPP) -MM $(INCFLAGS) $(CFLAGSD) $(COMPILFLAGS) $$i >> .depend; \
+	done
 
 dep:	.depend
 
