diff -pruN subversion-1.0.2_orig/subversion/libsvn_ra_svn/marshal.c subversion-1.0.2/subversion/libsvn_ra_svn/marshal.c
--- subversion-1.0.2_orig/subversion/libsvn_ra_svn/marshal.c	2004-05-13 11:03:26.000000000 +0200
+++ subversion-1.0.2/subversion/libsvn_ra_svn/marshal.c	2004-05-13 11:04:08.000000000 +0200
@@ -731,10 +731,6 @@ svn_error_t *svn_ra_svn_read_cmd_respons
                                     "Malformed error list");
           SVN_ERR(svn_ra_svn_parse_tuple(elt->u.list, pool, "nccn", &apr_err,
                                           &message, &file, &line));
-          /* The message field should have been optional, but we can't
-             easily change that, so "" means a nonexistent message. */
-          if (!*message)
-            message = NULL;
           err = svn_error_create(apr_err, err, message);
           err->file = apr_pstrdup(err->pool, file);
           err->line = line;
@@ -835,12 +831,10 @@ svn_error_t *svn_ra_svn_write_cmd_failur
   SVN_ERR(svn_ra_svn_start_list(conn, pool));
   for (; err; err = err->child)
     {
-      /* The message string should have been optional, but we can't
-         easily change that, so marshal nonexistent messages as "". */
       SVN_ERR(svn_ra_svn_write_tuple(conn, pool, "nccn",
                                      (apr_uint64_t) err->apr_err,
-                                     err->message ? err->message : "",
-                                     err->file, (apr_uint64_t) err->line));
+                                     err->message, err->file,
+                                     (apr_uint64_t) err->line));
     }
   SVN_ERR(svn_ra_svn_end_list(conn, pool));
   SVN_ERR(svn_ra_svn_end_list(conn, pool));
