--- ./src/callbacks.c.orig	2003-12-29 22:43:25.000000000 +0200
+++ ./src/callbacks.c	2004-02-01 16:11:44.000000000 +0200
@@ -581,19 +581,20 @@
 	    if(gtk_tree_view_get_path_at_pos(treeview, event->x, event->y, 
 				    &treepath, NULL, NULL, NULL))
 	    {
-		gchar *readable_path;
 		gtk_tree_model_get_iter(treemodel, &iter, treepath);
-
 		gtk_tree_model_get(treemodel, &iter, ENTRY_COLUMN, &en, -1);
-		readable_path=g_path_get_basename(en->path);
-		if (IS_NETWORK_TYPE(en->type)&&!IS_SAMBA_SERVER(en->subtype))
-			ascii_readable(readable_path);
-		print_status_tmp(treeview, resolve_icon_small(en), 
-				readable_path,
-				NULL);
-		g_free(readable_path);
-		readable_path=NULL;
-		set_path_reference(treeview,treepath);
+		if (en->path)
+		{
+		    gchar *readable_path=g_path_get_basename(en->path);
+		    if (IS_NETWORK_TYPE(en->type)&&!IS_SAMBA_SERVER(en->subtype))
+			    ascii_readable(readable_path);
+		    print_status_tmp(treeview, resolve_icon_small(en), 
+				    readable_path,
+				    NULL);
+		    g_free(readable_path);
+		    readable_path=NULL;
+		    set_path_reference(treeview,treepath);
+		}
 		
 		gtk_tree_path_free(treepath);
 		turn_on(treeview);
