--- ./libs/run.c.orig	2003-09-15 00:52:02.000000000 +0300
+++ ./libs/run.c	2003-11-09 16:01:31.000000000 +0200
@@ -527,8 +527,8 @@
     
      /* filter element is double used for arguments here */
     if (IS_APP_TYPE(en->type) && en->filter) a=g_strconcat(en->path," ",en->filter,NULL);
-    else if (IS_EXE(en->type)) a=NULL;
-    else a=g_strdup(en->path);
+    else if (IS_APP_TYPE(en->type)) a=g_strdup(en->path);
+    else a=NULL;
    
     print_status(tree_details->treeview, "xf_INFO_ICON",
 		   _("Executing"), " : ", 
