--- ./menubox.c.orig	2003-03-08 18:40:50.000000000 +0200
+++ ./menubox.c	2003-06-28 19:33:18.000000000 +0300
@@ -77,14 +77,14 @@
     /* Draw the input field box (only for inputmenu) */
     (void) wmove(win, y + 1, item_x);
     if (dialog_vars.input_menu)
-	draw_box(win, y, item_x, INPUT_ROWS, menu_width - 2 - item_x - tag_x,
+	draw_box(win, y, item_x, INPUT_ROWS, menu_width - item_x - tag_x,
 		 selected ? item_selected_attr : item_attr,
 		 selected ? item_selected_attr : item_attr);
 
     /* print actual item */
     wmove(win, y + 1, item_x + 1);	/* go into inputbox */
     wattrset(win, selected ? item_selected_attr : item_attr);
-    dlg_print_text(win, ItemText(0), menu_width - 2 - item_x - 2, &attr);
+    dlg_print_text(win, ItemText(0), menu_width - item_x - 2, &attr);
 
     if (selected) {
 	dlg_item_help(ItemHelp(0));
@@ -132,7 +132,7 @@
 	     *                      ( see in dialog_menu() )
 	     */
 	    dlg_show_string(win, result, offset, item_selected_attr,
-			    y + 1, item_x + 1, menu_width - 2 - item_x - 2,
+			    y + 1, item_x + 1, menu_width - item_x - 2,
 			    FALSE, first);
 	    first = FALSE;
 	}
