--- ./include/dirent.h~	Sat Apr 19 09:12:38 2003
+++ ./include/dirent.h	Sat Apr 19 09:13:24 2003
@@ -61,4 +61,33 @@
 #define alphasort alphasort64
 #endif
 
+#ifdef _BSD_SOURCE
+/* File types for `d_type'.  */
+enum
+  {
+    DT_UNKNOWN = 0,
+# define DT_UNKNOWN     DT_UNKNOWN
+    DT_FIFO = 1,
+# define DT_FIFO        DT_FIFO
+    DT_CHR = 2,
+# define DT_CHR         DT_CHR
+    DT_DIR = 4,
+# define DT_DIR         DT_DIR
+    DT_BLK = 6,
+# define DT_BLK         DT_BLK
+    DT_REG = 8,
+# define DT_REG         DT_REG
+    DT_LNK = 10,
+# define DT_LNK         DT_LNK
+    DT_SOCK = 12,
+# define DT_SOCK        DT_SOCK
+    DT_WHT = 14
+# define DT_WHT         DT_WHT
+  };
+
+/* Convert between stat structure types and directory types.  */
+# define IFTODT(mode)   (((mode) & 0170000) >> 12)
+# define DTTOIF(dirtype)        ((dirtype) << 12)
+#endif
+
 #endif
