--- dietlibc-0.29/dyn_start.c.orig	2005-07-07 15:09:10.000000000 +0200
+++ dietlibc-0.29/dyn_start.c	2005-07-07 15:09:31.000000000 +0200
@@ -24,7 +24,7 @@
 
 __attribute__((section(".eh_frame")))
 __attribute_used
-char __EH_FRAME_BEGIN__[] = { };
+static const char __EH_FRAME_BEGIN__[] = { };
 
 
 static void __do_global_dtors_aux(void)
--- dietlibc-0.29/dyn_stop.c.orig	2005-07-07 15:14:23.000000000 +0200
+++ dietlibc-0.29/dyn_stop.c	2005-07-07 15:17:58.000000000 +0200
@@ -17,9 +17,9 @@
 __attribute__((section(".eh_frame")))
 __attribute_used
 #if __WORDSIZE == 32
-char __FRAME_END__[4] = { 0, 0, 0, 0 };
+static const char __FRAME_END__[4] = { 0, 0, 0, 0 };
 #else
-char __FRAME_END__[8] = { 0, 0, 0, 0, 0, 0, 0, 0 };
+static const char __FRAME_END__[8] = { 0, 0, 0, 0, 0, 0, 0, 0 };
 #endif
 
 static void __do_global_ctors_aux(void)
