Terminate shell_argv - bye Gernot Tenchio, dietlibc list.

--- ./lib/exec_lib.c	2001-08-14 18:56:50.000000000 +0200
+++ ./lib/exec_lib.c	2006-04-04 09:07:50.581997500 +0200
@@ -10,11 +10,12 @@
   for (i = 0; argv[i]; i++);
 
   {
-    const char *shell_argv[i + 1];
+    const char *shell_argv[i + 2];
     shell_argv[0] = _PATH_BSHELL;
     shell_argv[1] = file;
     for (; i > 1; i--)
       shell_argv[i] = argv[i - 1];
+    shell_argv[i + 1] = NULL;
     return execve(_PATH_BSHELL, (char*const*)shell_argv, environ);
   }
 }
