diff -ur dietlibc-0.30/libm/fpmacros.c dietlibc-0.30-libm-i386/libm/fpmacros.c
--- dietlibc-0.30/libm/fpmacros.c	2006-06-28 20:15:02.000000000 +0200
+++ dietlibc-0.30-libm-i386/libm/fpmacros.c	2006-06-28 19:51:31.000000000 +0200
@@ -20,6 +20,8 @@
 **
 ***********************************************************************/
 
+#if ! defined (__i386__) && ! defined (__x86_64__)
+
 #include     <features.h>
 #define _GNU_SOURCE
 #include     <sys/types.h>
@@ -289,3 +291,4 @@
 weak_alias (__isnanl, isnanl);
 #endif
 
+#endif
diff -ur dietlibc-0.30/libm/s_asinh.c dietlibc-0.30-libm-i386/libm/s_asinh.c
--- dietlibc-0.30/libm/s_asinh.c	2006-06-28 20:15:02.000000000 +0200
+++ dietlibc-0.30-libm-i386/libm/s_asinh.c	2006-06-28 20:03:38.000000000 +0200
@@ -38,9 +38,9 @@
 huge=  1.00000000000000000000e+300;
 
 #ifdef __STDC__
-	double asinh(double x)
+	double __asinh(double x)
 #else
-	double asinh(x)
+	double __asinh(x)
 	double x;
 #endif
 {
@@ -63,3 +63,4 @@
 	}
 	if(hx>0) return w; else return -w;
 }
+weak_alias (__asinh, asinh)
diff -ur dietlibc-0.30/libm/s_atan.c dietlibc-0.30-libm-i386/libm/s_atan.c
--- dietlibc-0.30/libm/s_atan.c	2006-06-28 20:15:02.000000000 +0200
+++ dietlibc-0.30-libm-i386/libm/s_atan.c	2006-06-28 20:10:26.000000000 +0200
@@ -86,9 +86,9 @@
 huge   = 1.0e300;
 
 #ifdef __STDC__
-	double atan(double x)
+	double __atan(double x)
 #else
-	double atan(x)
+	double __atan(x)
 	double x;
 #endif
 {
@@ -137,3 +137,4 @@
 	    return (hx<0)? -z:z;
 	}
 }
+weak_alias (__atan, atan)
diff -ur dietlibc-0.30/libm/s_cbrt.c dietlibc-0.30-libm-i386/libm/s_cbrt.c
--- dietlibc-0.30/libm/s_cbrt.c	2006-06-28 20:15:02.000000000 +0200
+++ dietlibc-0.30-libm-i386/libm/s_cbrt.c	2006-06-28 20:05:18.000000000 +0200
@@ -40,9 +40,9 @@
 G =  3.57142857142857150787e-01; /* 5/14      = 0x3FD6DB6D, 0xB6DB6DB7 */
 
 #ifdef __STDC__
-	double cbrt(double x)
+	double __cbrt(double x)
 #else
-	double cbrt(x)
+	double __cbrt(x)
 	double x;
 #endif
 {
@@ -91,3 +91,4 @@
 	SET_HIGH_WORD(t,high|sign);
 	return(t);
 }
+weak_alias (__cbrt, cbrt)
diff -ur dietlibc-0.30/libm/s_copysign.c dietlibc-0.30-libm-i386/libm/s_copysign.c
--- dietlibc-0.30/libm/s_copysign.c	2006-06-28 20:15:02.000000000 +0200
+++ dietlibc-0.30-libm-i386/libm/s_copysign.c	2006-06-28 20:01:55.000000000 +0200
@@ -24,9 +24,9 @@
 #include "math_private.h"
 
 #ifdef __STDC__
-	double copysign(double x, double y)
+	double __copysign(double x, double y)
 #else
-	double copysign(x,y)
+	double __copysign(x,y)
 	double x,y;
 #endif
 {
@@ -36,4 +36,4 @@
 	SET_HIGH_WORD(x,(hx&0x7fffffff)|(hy&0x80000000));
         return x;
 }
-
+weak_alias(__copysign, copysign)
diff -ur dietlibc-0.30/libm/s_cos.c dietlibc-0.30-libm-i386/libm/s_cos.c
--- dietlibc-0.30/libm/s_cos.c	2006-06-28 20:15:02.000000000 +0200
+++ dietlibc-0.30-libm-i386/libm/s_cos.c	2006-06-28 20:10:03.000000000 +0200
@@ -49,9 +49,9 @@
 #include "math_private.h"
 
 #ifdef __STDC__
-	double cos(double x)
+	double __cos(double x)
 #else
-	double cos(x)
+	double __cos(x)
 	double x;
 #endif
 {
@@ -80,3 +80,4 @@
 	    }
 	}
 }
+weak_alias (__cos, cos)
diff -ur dietlibc-0.30/libm/s_expm1.c dietlibc-0.30-libm-i386/libm/s_expm1.c
--- dietlibc-0.30/libm/s_expm1.c	2006-06-28 20:15:02.000000000 +0200
+++ dietlibc-0.30-libm-i386/libm/s_expm1.c	2006-06-28 20:04:13.000000000 +0200
@@ -132,9 +132,9 @@
 Q5  =  -2.01099218183624371326e-07; /* BE8AFDB7 6E09C32D */
 
 #ifdef __STDC__
-	double expm1(double x)
+	double __expm1(double x)
 #else
-	double expm1(x)
+	double __expm1(x)
 	double x;
 #endif
 {
@@ -227,3 +227,4 @@
 	}
 	return y;
 }
+weak_alias (__expm1, expm1)
diff -ur dietlibc-0.30/libm/s_fabs.c dietlibc-0.30-libm-i386/libm/s_fabs.c
--- dietlibc-0.30/libm/s_fabs.c	2006-06-28 20:15:02.000000000 +0200
+++ dietlibc-0.30-libm-i386/libm/s_fabs.c	2006-06-28 20:09:37.000000000 +0200
@@ -22,9 +22,9 @@
 #include "math_private.h"
 
 #ifdef __STDC__
-	double fabs(double x)
+	double __fabs(double x)
 #else
-	double fabs(x)
+	double __fabs(x)
 	double x;
 #endif
 {
@@ -33,3 +33,4 @@
 	SET_HIGH_WORD(x,high&0x7fffffff);
         return x;
 }
+weak_alias (__fabs, fabs)
diff -ur dietlibc-0.30/libm/s_floor.c dietlibc-0.30-libm-i386/libm/s_floor.c
--- dietlibc-0.30/libm/s_floor.c	2006-06-28 20:15:02.000000000 +0200
+++ dietlibc-0.30-libm-i386/libm/s_floor.c	2006-06-28 20:09:19.000000000 +0200
@@ -33,9 +33,9 @@
 #endif
 
 #ifdef __STDC__
-	double floor(double x)
+	double __floor(double x)
 #else
-	double floor(x)
+	double __floor(x)
 	double x;
 #endif
 {
@@ -79,3 +79,4 @@
 	INSERT_WORDS(x,i0,i1);
 	return x;
 }
+weak_alias (__floor, floor)
diff -ur dietlibc-0.30/libm/s_ilogb.c dietlibc-0.30-libm-i386/libm/s_ilogb.c
--- dietlibc-0.30/libm/s_ilogb.c	2006-06-28 20:15:02.000000000 +0200
+++ dietlibc-0.30-libm-i386/libm/s_ilogb.c	2006-06-28 19:58:36.000000000 +0200
@@ -24,9 +24,9 @@
 #include "math_private.h"
 
 #ifdef __STDC__
-	int ilogb(double x)
+	int __ilogb(double x)
 #else
-	int ilogb(x)
+	int __ilogb(x)
 	double x;
 #endif
 {
@@ -49,3 +49,4 @@
 	else if (hx<0x7ff00000) return (hx>>20)-1023;
 	else return 0x7fffffff;
 }
+weak_alias (__ilogb, ilogb)
diff -ur dietlibc-0.30/libm/s_ldexp.c dietlibc-0.30-libm-i386/libm/s_ldexp.c
--- dietlibc-0.30/libm/s_ldexp.c	2006-06-28 20:15:02.000000000 +0200
+++ dietlibc-0.30-libm-i386/libm/s_ldexp.c	2006-06-28 20:06:45.000000000 +0200
@@ -19,9 +19,9 @@
 #include <errno.h>
 
 #ifdef __STDC__
-	double ldexp(double value, int exp)
+	double __ldexp(double value, int exp)
 #else
-	double ldexp(value, exp)
+	double __ldexp(value, exp)
 	double value; int exp;
 #endif
 {
@@ -30,3 +30,4 @@
 	if(!finite(value)||value==0.0) errno = ERANGE;
 	return value;
 }
+weak_alias (__ldexp, ldexp)
diff -ur dietlibc-0.30/libm/s_log1p.c dietlibc-0.30-libm-i386/libm/s_log1p.c
--- dietlibc-0.30/libm/s_log1p.c	2006-06-28 20:15:02.000000000 +0200
+++ dietlibc-0.30-libm-i386/libm/s_log1p.c	2006-06-28 20:06:16.000000000 +0200
@@ -105,9 +105,9 @@
 #endif
 
 #ifdef __STDC__
-	double log1p(double x)
+	double __log1p(double x)
 #else
-	double log1p(x)
+	double __log1p(x)
 	double x;
 #endif
 {
@@ -172,3 +172,4 @@
 	if(k==0) return f-(hfsq-s*(hfsq+R)); else
 		 return k*ln2_hi-((hfsq-(s*(hfsq+R)+(k*ln2_lo+c)))-f);
 }
+weak_alias (__log1p, log1p)
diff -ur dietlibc-0.30/libm/s_logb.c dietlibc-0.30-libm-i386/libm/s_logb.c
--- dietlibc-0.30/libm/s_logb.c	2006-06-28 20:15:02.000000000 +0200
+++ dietlibc-0.30-libm-i386/libm/s_logb.c	2006-06-28 19:59:16.000000000 +0200
@@ -24,9 +24,9 @@
 #include "math_private.h"
 
 #ifdef __STDC__
-	double logb(double x)
+	double __logb(double x)
 #else
-	double logb(x)
+	double __logb(x)
 	double x;
 #endif
 {
@@ -40,3 +40,4 @@
 	else
 		return (double) (ix-1023);
 }
+weak_alias (__logb, logb)
diff -ur dietlibc-0.30/libm/s_sin.c dietlibc-0.30-libm-i386/libm/s_sin.c
--- dietlibc-0.30/libm/s_sin.c	2006-06-28 20:15:02.000000000 +0200
+++ dietlibc-0.30-libm-i386/libm/s_sin.c	2006-06-28 20:07:36.000000000 +0200
@@ -49,9 +49,9 @@
 #include "math_private.h"
 
 #ifdef __STDC__
-	double sin(double x)
+	double __sin(double x)
 #else
-	double sin(x)
+	double __sin(x)
 	double x;
 #endif
 {
@@ -80,3 +80,4 @@
 	    }
 	}
 }
+weak_alias (__sin, sin)
diff -ur dietlibc-0.30/libm/s_tan.c dietlibc-0.30-libm-i386/libm/s_tan.c
--- dietlibc-0.30/libm/s_tan.c	2006-06-28 20:15:02.000000000 +0200
+++ dietlibc-0.30-libm-i386/libm/s_tan.c	2006-06-28 19:57:55.000000000 +0200
@@ -48,9 +48,9 @@
 #include "math_private.h"
 
 #ifdef __STDC__
-	double tan(double x)
+	double __tan(double x)
 #else
-	double tan(x)
+	double __tan(x)
 	double x;
 #endif
 {
@@ -74,3 +74,4 @@
 							-1 -- n odd */
 	}
 }
+weak_alias (__tan, tan)
diff -ur dietlibc-0.30/libm/w_acos.c dietlibc-0.30-libm-i386/libm/w_acos.c
--- dietlibc-0.30/libm/w_acos.c	2006-06-28 20:15:02.000000000 +0200
+++ dietlibc-0.30-libm-i386/libm/w_acos.c	2006-06-28 20:10:58.000000000 +0200
@@ -23,9 +23,9 @@
 
 
 #ifdef __STDC__
-	double acos(double x)		/* wrapper acos */
+	double __acos(double x)		/* wrapper acos */
 #else
-	double acos(x)			/* wrapper acos */
+	double __acos(x)			/* wrapper acos */
 	double x;
 #endif
 {
@@ -41,3 +41,4 @@
 	    return z;
 #endif
 }
+weak_alias (__acos, acos)
diff -ur dietlibc-0.30/libm/w_acosh.c dietlibc-0.30-libm-i386/libm/w_acosh.c
--- dietlibc-0.30/libm/w_acosh.c	2006-06-28 20:15:02.000000000 +0200
+++ dietlibc-0.30-libm-i386/libm/w_acosh.c	2006-06-28 20:02:44.000000000 +0200
@@ -22,9 +22,9 @@
 #include "math_private.h"
 
 #ifdef __STDC__
-	double acosh(double x)		/* wrapper acosh */
+	double __acosh(double x)		/* wrapper acosh */
 #else
-	double acosh(x)			/* wrapper acosh */
+	double __acosh(x)			/* wrapper acosh */
 	double x;
 #endif
 {
@@ -40,3 +40,4 @@
 	    return z;
 #endif
 }
+weak_alias (__acosh, acosh)
diff -ur dietlibc-0.30/libm/w_asin.c dietlibc-0.30-libm-i386/libm/w_asin.c
--- dietlibc-0.30/libm/w_asin.c	2006-06-28 20:15:02.000000000 +0200
+++ dietlibc-0.30-libm-i386/libm/w_asin.c	2006-06-28 20:03:08.000000000 +0200
@@ -24,9 +24,9 @@
 
 
 #ifdef __STDC__
-	double asin(double x)		/* wrapper asin */
+	double __asin(double x)		/* wrapper asin */
 #else
-	double asin(x)			/* wrapper asin */
+	double __asin(x)			/* wrapper asin */
 	double x;
 #endif
 {
@@ -42,3 +42,4 @@
 	    return z;
 #endif
 }
+weak_alias (__asin, asin)
diff -ur dietlibc-0.30/libm/w_atan2.c dietlibc-0.30-libm-i386/libm/w_atan2.c
--- dietlibc-0.30/libm/w_atan2.c	2006-06-28 20:15:02.000000000 +0200
+++ dietlibc-0.30-libm-i386/libm/w_atan2.c	2006-06-28 19:57:11.000000000 +0200
@@ -22,9 +22,9 @@
 
 
 #ifdef __STDC__
-	double atan2(double y, double x)	/* wrapper atan2 */
+	double __atan2(double y, double x)	/* wrapper atan2 */
 #else
-	double atan2(y,x)			/* wrapper atan2 */
+	double __atan2(y,x)			/* wrapper atan2 */
 	double y,x;
 #endif
 {
@@ -40,3 +40,4 @@
 	    return z;
 #endif
 }
+weak_alias (__atan2, atan2)
diff -ur dietlibc-0.30/libm/w_atanh.c dietlibc-0.30-libm-i386/libm/w_atanh.c
--- dietlibc-0.30/libm/w_atanh.c	2006-06-28 20:15:02.000000000 +0200
+++ dietlibc-0.30-libm-i386/libm/w_atanh.c	2006-06-28 20:02:23.000000000 +0200
@@ -23,9 +23,9 @@
 
 
 #ifdef __STDC__
-	double atanh(double x)		/* wrapper atanh */
+	double __atanh(double x)		/* wrapper atanh */
 #else
-	double atanh(x)			/* wrapper atanh */
+	double __atanh(x)			/* wrapper atanh */
 	double x;
 #endif
 {
@@ -45,3 +45,4 @@
 	    return z;
 #endif
 }
+weak_alias (__atanh, atanh)
diff -ur dietlibc-0.30/libm/w_drem.c dietlibc-0.30-libm-i386/libm/w_drem.c
--- dietlibc-0.30/libm/w_drem.c	2006-06-28 20:15:02.000000000 +0200
+++ dietlibc-0.30-libm-i386/libm/w_drem.c	2006-06-28 20:01:26.000000000 +0200
@@ -6,10 +6,10 @@
  */
 
 #include <math.h>
+#include "math_private.h"
 
-double
-drem(x, y)
-	double x, y;
+double __drem(double x, double y)
 {
 	return remainder(x, y);
 }
+weak_alias (__drem, drem)
diff -ur dietlibc-0.30/libm/w_fmod.c dietlibc-0.30-libm-i386/libm/w_fmod.c
--- dietlibc-0.30/libm/w_fmod.c	2006-06-28 20:15:02.000000000 +0200
+++ dietlibc-0.30-libm-i386/libm/w_fmod.c	2006-06-28 20:05:43.000000000 +0200
@@ -23,9 +23,9 @@
 
 
 #ifdef __STDC__
-	double fmod(double x, double y)	/* wrapper fmod */
+	double __fmod(double x, double y)	/* wrapper fmod */
 #else
-	double fmod(x,y)		/* wrapper fmod */
+	double __fmod(x,y)		/* wrapper fmod */
 	double x,y;
 #endif
 {
@@ -41,3 +41,4 @@
 	    return z;
 #endif
 }
+weak_alias (__fmod, fmod)
diff -ur dietlibc-0.30/libm/w_hypot.c dietlibc-0.30-libm-i386/libm/w_hypot.c
--- dietlibc-0.30/libm/w_hypot.c	2006-06-28 20:15:02.000000000 +0200
+++ dietlibc-0.30-libm-i386/libm/w_hypot.c	2006-06-28 20:08:54.000000000 +0200
@@ -23,9 +23,9 @@
 
 
 #ifdef __STDC__
-	double hypot(double x, double y)/* wrapper hypot */
+	double __hypot(double x, double y)/* wrapper hypot */
 #else
-	double hypot(x,y)		/* wrapper hypot */
+	double __hypot(x,y)		/* wrapper hypot */
 	double x,y;
 #endif
 {
@@ -41,3 +41,4 @@
 	    return z;
 #endif
 }
+weak_alias (__hypot, hypot)
diff -ur dietlibc-0.30/libm/w_log.c dietlibc-0.30-libm-i386/libm/w_log.c
--- dietlibc-0.30/libm/w_log.c	2006-06-28 20:15:02.000000000 +0200
+++ dietlibc-0.30-libm-i386/libm/w_log.c	2006-06-28 20:08:32.000000000 +0200
@@ -23,9 +23,9 @@
 
 
 #ifdef __STDC__
-	double log(double x)		/* wrapper log */
+	double __log(double x)		/* wrapper log */
 #else
-	double log(x)			/* wrapper log */
+	double __log(x)			/* wrapper log */
 	double x;
 #endif
 {
@@ -41,3 +41,4 @@
 	    return __kernel_standard(x,x,17); /* log(x<0) */
 #endif
 }
+weak_alias (__log, log)
diff -ur dietlibc-0.30/libm/w_log10.c dietlibc-0.30-libm-i386/libm/w_log10.c
--- dietlibc-0.30/libm/w_log10.c	2006-06-28 20:15:02.000000000 +0200
+++ dietlibc-0.30-libm-i386/libm/w_log10.c	2006-06-28 20:08:11.000000000 +0200
@@ -23,9 +23,9 @@
 
 
 #ifdef __STDC__
-	double log10(double x)		/* wrapper log10 */
+	double __log10(double x)		/* wrapper log10 */
 #else
-	double log10(x)			/* wrapper log10 */
+	double __log10(x)			/* wrapper log10 */
 	double x;
 #endif
 {
@@ -44,3 +44,4 @@
 	    return z;
 #endif
 }
+weak_alias (__log10, log10)
diff -ur dietlibc-0.30/libm/w_sqrt.c dietlibc-0.30-libm-i386/libm/w_sqrt.c
--- dietlibc-0.30/libm/w_sqrt.c	2006-06-28 20:15:02.000000000 +0200
+++ dietlibc-0.30-libm-i386/libm/w_sqrt.c	2006-06-28 20:07:09.000000000 +0200
@@ -22,9 +22,9 @@
 #include "math_private.h"
 
 #ifdef __STDC__
-	double sqrt(double x)		/* wrapper sqrt */
+	double __sqrt(double x)		/* wrapper sqrt */
 #else
-	double sqrt(x)			/* wrapper sqrt */
+	double __sqrt(x)			/* wrapper sqrt */
 	double x;
 #endif
 {
@@ -40,3 +40,4 @@
 	    return z;
 #endif
 }
+weak_alias (__sqrt, sqrt)
