--- dietlibc/include/sys/swap.h.orig	1970-01-01 01:00:00.000000000 +0100
+++ dietlibc/include/sys/swap.h	2004-12-29 16:26:49.767145264 +0100
@@ -0,0 +1,19 @@
+#ifndef _SYS_SWAP_H
+#define _SYS_SWAP_H
+
+#include <sys/cdefs.h>
+
+#define SWAP_FLAG_PREFER	0x8000	/* set if swap priority specified */
+#define SWAP_FLAG_PRIO_MASK	0x7fff
+#define SWAP_FLAG_PRIO_SHIFT	0
+
+#define MAX_SWAPFILES 32
+
+__BEGIN_DECLS
+
+extern int swapon (const char *path, int flags) __THROW;
+extern int swapoff (const char *path) __THROW;
+
+__END_DECLS
+
+#endif /* _SYS_SWAP_H */
