--- ./include/sys/utsname.h.vanilla	2006-07-19 13:08:26.000000000 +0200
+++ ./include/sys/utsname.h	2006-07-19 13:08:54.000000000 +0200
@@ -14,6 +14,8 @@
    to a value != 0 to activate this entry.  */
 #define _UTSNAME_DOMAIN_LENGTH _UTSNAME_LENGTH
 
+#define SYS_NMLN _UTSNAME_LENGTH
+
 struct utsname {
   /* Name of the implementation of the operating system.  */
   char sysname[_UTSNAME_LENGTH];
