
Mismatch with kernel headers ...

  - Rene Rebe <rene@exactcode.de>

--- dietlibc/include/asm/types.h	2005-08-24 16:57:45.000000000 +0000
+++ dietlibc/include/asm/types.h	2005-08-24 21:53:49.000000000 +0000
@@ -25,7 +25,7 @@
 typedef int64_t __s64;
 #endif
 
-#if defined(__alpha__)
+#if defined(__alpha__) || defined(__x86_64__)
 typedef unsigned long __kernel_size_t;
 #else
 typedef unsigned int __kernel_size_t;
