diff -Naur dietlibc_cvs-orig/include/sys/stat.h dietlibc_cvs/include/sys/stat.h
--- dietlibc_cvs-orig/include/sys/stat.h	2005-03-15 10:51:22.000000000 +0200
+++ dietlibc_cvs/include/sys/stat.h	2005-04-30 18:50:40.000000000 +0300
@@ -520,11 +520,11 @@
 	unsigned long	st_size;
 	unsigned long	st_blksize;
 	unsigned long	st_blocks;
-	unsigned long	st_atime;
+	signed long	st_atime;
 	unsigned long	__reserved0;
-	unsigned long	st_mtime;
+	signed long	st_mtime;
 	unsigned long	__reserved1;
-	unsigned long	st_ctime;
+	signed long	st_ctime;
 	unsigned long	__reserved2;
 	long		__unused[3];
 };
