diff -ur webmin-1.180_orig/setup.sh webmin-1.180/setup.sh
--- webmin-1.180_orig/setup.sh	2005-01-24 23:15:16.000000000 +0100
+++ webmin-1.180/setup.sh	2005-05-26 11:53:51.524877296 +0200
@@ -55,14 +55,14 @@
 			echo ""
 			exit 1
 		fi
-	else
+	###else
 		# Make sure dest dir is not in use
-		ls "$wadir" | grep -v rpmsave >/dev/null 2>&1
-		if [ "$?" = "0" -a ! -r "$wadir/setup.sh" ]; then
-			echo "ERROR: Installation directory $wadir contains other files"
-			echo ""
-			exit 1
-		fi
+		###ls "$wadir" | grep -v rpmsave >/dev/null 2>&1
+		###if [ "$?" = "0" -a ! -r "$wadir/setup.sh" ]; then
+		###	echo "ERROR: Installation directory $wadir contains other files"
+		###	echo ""
+		###	exit 1
+		###fi
 	fi
 else
 	echo "Installing Webmin in $wadir ..."
