diff -Nur chrony-1.19.orig/conf.c chrony-1.19/conf.c
--- chrony-1.19.orig/conf.c	2002-03-01 00:27:09.000000000 +0100
+++ chrony-1.19/conf.c	2003-03-07 20:21:56.000000000 +0100
@@ -85,7 +85,7 @@
 /* ================================================== */
 /* Configuration variables */
 
-static char *rtc_device = "/dev/rtc";
+static char *rtc_device = "/dev/misc/rtc";
 static int ntp_port = 123;
 static char *keys_file = NULL;
 static char *drift_file = NULL;
