diff -Nur chrony-1.19.orig/conf.c chrony-1.19/conf.c
--- chrony-1.19.orig/conf.c	2003-03-07 20:21:56.000000000 +0100
+++ chrony-1.19/conf.c	2003-03-07 21:50:46.000000000 +0100
@@ -45,7 +45,7 @@
 
 /* ================================================== */
 
-#define DEFAULT_CONF_FILE "/etc/chrony.conf"
+#define DEFAULT_CONF_FILE "/etc/chrony/chrony.conf"
 
 /* ================================================== */
 /* Forward prototypes */
