--- uucp-1.06.1/Makefile.in.cfgmake	Sun Aug 20 15:36:06 1995
+++ uucp-1.06.1/Makefile.in	Thu Aug  9 22:55:23 2001
@@ -18,10 +18,11 @@
 # The user name that should own the resulting executables, several of
 # which are suid.  If you change this, you must run `make clean'.
 owner = uucp
+group = uucp
 
 # Where to install uucico, uuxqt, uuchk and uuconv.  If you use this
 # definition, $(exec_prefix)/lib must exist.
-sbindir = $(exec_prefix)/lib/uucp
+sbindir = $(exec_prefix)/sbin
 #sbindir = $(exec_prefix)/libexec/uucp
 
 # Where to install uucp, uux, uustat, uuname, uulog, uuto, uupick and
@@ -43,12 +44,12 @@
 # The directory to look in for new style configuration files (when
 # using HAVE_TAYLOR_CONFIG).  Note that by default this is different
 # from sbindir, unlike traditional UUCP packages.
-newconfigdir = $(prefix)/conf/uucp
-#newconfigdir = /etc/uucp
+#newconfigdir = $(prefix)/conf/uucp
+newconfigdir = /etc/uucp
 
 # The directory to look in for HDB (when using HAVE_HDB_CONFIG) or
 # V2 (when using HAVE_V2_CONFIG) style configuration files.
-oldconfigdir = /usr/lib/uucp
+oldconfigdir = /etc/uucp/oldconfig
 
 # If you don't want to compile tstuu, which requires BSD style ptys
 # and the Taylor configuration files, comment out the following line
@@ -108,9 +109,9 @@
 
 VERSION = 1.06.1
 
-MORECFLAGS = -I. -I$(srcdir) -DOWNER=\"$(owner)\" -DVERSION=\"$(VERSION)\" $(UUDIRFLAGS)
+MORECFLAGS = -I. -I$(srcdir) -DOWNER=\"$(owner)\" -DGROUP=\"$(group)\" -DVERSION=\"$(VERSION)\" $(UUDIRFLAGS)
 
-MDEFINES = CC='$(CC)' CPPFLAGS='$(CPPFLAGS)' CFLAGS='$(CFLAGS)' AR='$(AR)' RANLIB='$(RANLIB)' LN_S='$(LN_S)' newconfigdir='$(newconfigdir)' oldconfigdir='$(oldconfigdir)' owner='$(owner)' sbindir='$(sbindir)' UUDIRFLAGS='$(UUDIRFLAGS)' VERSION='$(VERSION)'
+MDEFINES = CC='$(CC)' CPPFLAGS='$(CPPFLAGS)' CFLAGS='$(CFLAGS)' AR='$(AR)' RANLIB='$(RANLIB)' LN_S='$(LN_S)' newconfigdir='$(newconfigdir)' oldconfigdir='$(oldconfigdir)' owner='$(owner)' group='$(group)' sbindir='$(sbindir)' UUDIRFLAGS='$(UUDIRFLAGS)' VERSION='$(VERSION)'
 
 PROGRAMS = uucico uuxqt uux uucp uuchk uuconv uustat uuname uusched \
 	uulog uuto uupick cu
@@ -151,16 +152,16 @@
 	$(INSTALL_PROGRAM) uuchk $(sbindir)/uuchk
 	$(INSTALL_PROGRAM) uuconv $(sbindir)/uuconv
 	$(INSTALL_PROGRAM) uusched $(sbindir)/uusched
-	chown $(owner) $(sbindir)/uucico $(sbindir)/uuxqt
-	chmod 4555 $(sbindir)/uucico $(sbindir)/uuxqt
-	chown $(owner) $(sbindir)/uuchk $(sbindir)/uuconv
+	chown $(owner).$(group) $(sbindir)/uucico $(sbindir)/uuxqt
+	chmod 6555 $(sbindir)/uucico $(sbindir)/uuxqt
+	chown $(owner).$(group) $(sbindir)/uuchk $(sbindir)/uuconv
 	if test "$(UUDIR)X" != "X"; then \
 	  if test -d $(uudirdir); then \
 	    true \
 	  else \
 	    mkdir $(uudirdir); \
 	  fi; \
-	  chown $(owner) $(uudirdir); \
+	  chown $(owner).$(group) $(uudirdir); \
 	  chmod 100 $(uudirdir); \
 	  $(INSTALL_PROGRAM) uudir $(uudirdir)/uudir; \
 	  chown root $(uudirdir)/uudir; \
@@ -176,10 +177,10 @@
 	  fi; \
 	  $(INSTALL_PROGRAM) $$i $(bindir)/$$i; \
 	done
-	chown $(owner) $(bindir)/uux $(bindir)/uucp $(bindir)/uustat 
-	chown $(owner) $(bindir)/uuname $(bindir)/cu
+	chown $(owner).$(group) $(bindir)/uux $(bindir)/uucp $(bindir)/uustat 
+	chown $(owner).$(group) $(bindir)/uuname $(bindir)/cu
 	chmod 4555 $(bindir)/uux $(bindir)/uucp $(bindir)/uustat 
-	chmod 4555 $(bindir)/uuname $(bindir)/cu
+	chmod 6555 $(bindir)/uuname $(bindir)/cu
 	if test -d $(man1dir); then true; else mkdir $(man1dir); fi
 	if test -d $(man8dir); then true; else mkdir $(man8dir); fi
 	$(INSTALL_DATA) $(srcdir)/uucico.8 $(man8dir)/uucico$(man8ext)
@@ -195,9 +196,9 @@
 	-cp $(sbindir)/uucico.old $(sbindir)/uucico
 	-cp $(sbindir)/uuxqt.old $(sbindir)/uuxqt
 	-cp $(sbindir)/uusched.old $(sbindir)/uusched
-	-chown $(owner) $(sbindir)/uucico $(sbindir)/uuxqt
+	-chown $(owner).$(group) $(sbindir)/uucico $(sbindir)/uuxqt
 	-chmod 4555 $(sbindir)/uucico $(sbindir)/uuxqt
-	-chown $(owner) $(sbindir)/uuchk $(sbindir)/uuconv
+	-chown $(owner).$(group) $(sbindir)/uuchk $(sbindir)/uuconv
 	rm -f $(bindir)/uux $(bindir)/uucp $(bindir)/uustat
 	rm -f $(bindir)/uuname $(bindir)/uulog $(bindir)/uuto
 	rm -f $(bindir)/uupick $(bindir)/cu
@@ -209,8 +210,8 @@
 	-cp $(bindir)/uuto.old $(bindir)/uuto
 	-cp $(bindir)/uupick.old $(bindir)/uupick
 	-cp $(bindir)/cu.old $(bindir)/cu
-	-chown $(owner) $(bindir)/uux $(bindir)/uucp $(bindir)/uustat 
-	-chown $(owner) $(bindir)/uuname $(bindir)/cu
+	-chown $(owner).$(group) $(bindir)/uux $(bindir)/uucp $(bindir)/uustat 
+	-chown $(owner).$(group) $(bindir)/uuname $(bindir)/cu
 	-chmod 4555 $(bindir)/uux $(bindir)/uucp $(bindir)/uustat 
 	-chmod 4555 $(bindir)/uuname $(bindir)/cu
 
