--- uucp-1.06.1/uuxqt.c.vetargs	Sun Aug 20 15:29:59 1995
+++ uucp-1.06.1/uuxqt.c	Mon Jan 28 14:19:48 2002
@@ -928,6 +928,32 @@
 	 don't permit multiple arguments.  */
       for (i = 1; azQargs[i] != NULL; i++)
 	{
+	  if (azQargs[i][0] == '-' && azQargs[i][1] == '-')
+	    {
+		char *zopts = azQargs[i] + 2;
+		char *zeq = strchr(zopts, '=');
+
+		/* The -g, -n, and -s options take an argument.  */
+		if (!(strncmp(zopts, "not", 3)	/* --notify */
+		   && strncmp(zopts, "g", 1)	/* --grade */
+		   && strncmp(zopts, "s", 1))	/* --status */
+		&& zeq == NULL)
+		  {
+		    if (azQargs[i+1] != NULL)
+		      ++i;
+		  }
+
+		/* The -I, -u and -x options are not permitted.  */
+		if (!(strncmp(zopts, "con", 3)	/* --config */
+		   && strncmp(zopts, "us", 2)	/* --user */
+		   && strncmp(zopts, "de", 2)))	/* --debug */
+		  {
+		    if (azQargs[i+1] != NULL && zeq == NULL)
+		      ++i;
+		    azQargs[i] = zbufcpy ("--nouucico");
+		  }
+	    }
+	  else
 	  if (azQargs[i][0] == '-')
 	    {
 	      char *zopts;
