--- uucp-1.06.1/contrib/uureroute.perl.orig	2003-05-12 22:40:45.000000000 +0200
+++ uucp-1.06.1/contrib/uureroute.perl	2003-05-12 22:41:27.000000000 +0200
@@ -1,5 +1,5 @@
-#!/usr/local/bin/perl
-eval ' exec /usr/local/bin/perl $0 "$@" '
+#!/usr/bin/perl
+eval ' exec /usr/bin/perl $0 "$@" '
 	if $running_under_some_shell;
 
 # From a script by <Bill.Campbell@celestial.com>
@@ -34,7 +34,7 @@
 #
 ";
 
-$UUSTAT = "/usr/local/bin/uustat";
+$UUSTAT = "/usr/bin/uustat";
 $SHELL = "/bin/sh";
 $SMAIL = "/bin/smail";
 
