--- ./configure.in.orig	2005-02-09 11:24:21.713397650 -0300
+++ ./configure.in	2005-02-09 11:25:31.069420341 -0300
@@ -39,14 +39,6 @@
 ])
 fi
 
-if ! grep -q '\(\*release\)' $fuseinst/include/fuse.h ; then
-	AC_MSG_RESULT([old])
-	AC_MSG_ERROR([
-*** You need fuse version 1.0 or later.
-*** Please go to http://sourceforge.net/projects/avf
-*** and download the latest version
-])
-fi
 AC_MSG_RESULT(${fuseinst})
 AC_SUBST(fuseinst)
 
--- ./siefs/Makefile.am.orig	2005-02-09 11:31:40.016076350 -0300
+++ ./siefs/Makefile.am	2005-02-09 11:33:11.394638747 -0300
@@ -1,6 +1,6 @@
 ## Process this file with automake to produce Makefile.in
 
-CFLAGS = -I$(fuseinst)/include
+CFLAGS = -I$(fuseinst)/include -D_FILE_OFFSET_BITS=64 -DFUSE_USE_VERSION=11
 
 bin_PROGRAMS = siefs slink
 
