--- gnomemeeting-0.98.5/lib/xdap/xdap.c.orig	2003-12-02 14:43:58.000000000 +0100
+++ gnomemeeting-0.98.5/lib/xdap/xdap.c	2003-12-02 14:47:13.000000000 +0100
@@ -156,7 +156,7 @@
 	   xmlEntityPtr (**oldgetent) (void *, const xmlChar *), int noerr)
 {
   xmlDocPtr xp;
-  xmlSAXHandler sax;
+  xmlSAXHandlerV1 sax;
 
   memset (&sax, 0, sizeof sax);
   sax = xmlDefaultSAXHandler;	/* Copy original handler */
