diff -u misc.org/lpt.c misc/lpt.c
--- ./src/base/dev/misc.org/lpt.c	Sun Mar 11 14:27:24 2001
+++ ./src/base/dev/misc/lpt.c	Wed Mar 19 16:28:43 2003
@@ -17,6 +17,7 @@
 #include <unistd.h>
 #include <stdlib.h>
 #include <sys/wait.h>
+#include <errno.h>
 
 #include "emu.h"
 #include "dosio.h"
@@ -137,7 +138,6 @@
 printer_flush(int prnum)
 {
   int returnstat;
-  extern int errno;
 
   p_printf("LPT: flushing printer %d, %s\n", prnum, lpt[prnum].dev);
 
