--- nxcomp/Makefile.in.orig	2003-10-26 16:15:37.000000000 +0100
+++ nxcomp/Makefile.in	2003-10-27 00:28:39.000000000 +0100
@@ -1,6 +1,7 @@
 #
 # Get values from configure script.
 #
+
 VERSION=@VERSION@
 LIBVERSION=@LIBVERSION@
 
@@ -195,8 +196,10 @@
 		touch depend.status
 
 install:	install.bin install.man
+		$(INSTALL) -v NX*.h $(prefix)/include/
 
 install.bin:
+		$(INSTALL) -v $(LIBFULL) $(LIBLOAD) $(prefix)/lib/
 
 install.man:
 
