--- uClibc-0.9.27/libm/s_sin.c.orig	2005-01-28 10:47:40.000000000 -0800
+++ uClibc-0.9.27/libm/s_sin.c	2005-01-28 10:55:01.000000000 -0800
@@ -80,3 +80,17 @@
 	    }
 	}
 }
+
+#ifdef __STDC__
+        float sinf(float x)
+#else   
+        float sinf(x)
+        float x;
+#endif
+{
+        double y;
+        y=x;
+        y=sin(y);
+        return (float)y;
+}
+
--- uClibc-0.9.27/libm/s_cos.c.orig	2005-01-28 10:44:30.000000000 -0800
+++ uClibc-0.9.27/libm/s_cos.c	2005-01-28 10:52:15.000000000 -0800
@@ -80,3 +80,16 @@
 	    }
 	}
 }
+
+#ifdef __STDC__
+	float cosf(float x)
+#else
+	float cosf(x)
+	float x;
+#endif
+{
+	double y;
+	y=x;
+	y=cos(y);
+	return (float)y;
+}
