--- ./Makefile.orig	2004-02-07 10:30:44.000000000 -0300
+++ ./Makefile	2004-02-07 10:32:53.000000000 -0300
@@ -121,11 +121,6 @@
 	@./extra/config/conf -o extra/Configs/Config.in
 
 headers: include/bits/uClibc_config.h
-ifeq ($(strip $(UCLIBC_HAS_MMU)),y)
-	@set -x; ./extra/scripts/fix_includes.sh -k $(KERNEL_SOURCE) -t $(TARGET_ARCH)
-else
-	@set -x; ./extra/scripts/fix_includes.sh -k $(KERNEL_SOURCE) -t $(TARGET_ARCH) -n
-endif
 	@cd include/bits; \
 	set -e; \
 	for i in `ls ../../libc/sysdeps/linux/common/bits/*.h` ; do \
