--- gnumeric-1.3.1/src/workbook-control-gui-priv.h~	2004-07-19 08:06:18.000000000 +0200
+++ gnumeric-1.3.1/src/workbook-control-gui-priv.h	2004-07-23 16:35:58.642484856 +0200
@@ -111,7 +111,7 @@
 void	 wbcg_insert_sheet	      (GtkWidget *ignored, WorkbookControlGUI *wbcg);
 void	 wbcg_append_sheet	      (GtkWidget *ignored, WorkbookControlGUI *wbcg);
 void	 wbcg_set_selection_halign    (WorkbookControlGUI *wbcg, StyleHAlignFlags halign);
-void	 wbcg_set_selection_valign    (WorkbookControlGUI *wbcg, StyleHAlignFlags halign);
+void	 wbcg_set_selection_valign    (WorkbookControlGUI *wbcg, StyleVAlignFlags halign);
 
 enum {
 	WBCG_MARKUP_CHANGED,
