--- libdvb-0.5.5.1/sample_progs/cam_menu.cc.orig	2005-09-08 10:55:29.000000000 -0700
+++ libdvb-0.5.5.1/sample_progs/cam_menu.cc	2005-09-08 11:08:25.000000000 -0700
@@ -46,7 +46,13 @@
 
 void byebye(int i, void *f)
 {
-	int fd = (int) f;
+	int fd;
+	if (sizeof(void *) == sizeof(int))
+		fd = (long) f;
+	else {
+		long fd_long = (long) f;
+		fd = (int) fd_long;
+	}
 
 	CamPacket cp(fd, CMENU_CLOSE, 0, NULL);
 	cp.psend();
