diff -urN util-linux-2.11z.orig/disk-utils/blockdev.c util-linux-2.12/disk-utils/blockdev.c
--- util-linux-2.11z.orig/disk-utils/blockdev.c	2002-03-08 22:57:02.000000000 +0000
+++ util-linux-2.11z/disk-utils/blockdev.c	2003-09-11 06:16:46.000000000 +0100
@@ -24,8 +24,8 @@
 #define BLKRASET   _IO(0x12,98)
 #define BLKRAGET   _IO(0x12,99)
 #define BLKSSZGET  _IO(0x12,104)
-#define BLKBSZGET  _IOR(0x12,112,sizeof(int))
-#define BLKBSZSET  _IOW(0x12,113,sizeof(int))
+#define BLKBSZGET  _IOR(0x12,112,int)
+#define BLKBSZSET  _IOW(0x12,113,int)
 #endif
 
 /* Maybe <linux/hdreg.h> could be included */
diff -urN util-linux-2.11z.orig/disk-utils/elvtune.c util-linux-2.12/disk-utils/elvtune.c
--- util-linux-2.11z.orig/disk-utils/elvtune.c	2002-03-08 22:57:49.000000000 +0000
+++ util-linux-2.11z/disk-utils/elvtune.c	2003-09-11 06:16:46.000000000 +0100
@@ -37,8 +37,8 @@
 	int max_bomb_segments;
 } blkelv_ioctl_arg_t;
 
-#define BLKELVGET   _IOR(0x12,106,sizeof(blkelv_ioctl_arg_t))
-#define BLKELVSET   _IOW(0x12,107,sizeof(blkelv_ioctl_arg_t))
+#define BLKELVGET   _IOR(0x12,106,blkelv_ioctl_arg_t)
+#define BLKELVSET   _IOW(0x12,107,blkelv_ioctl_arg_t)
 
 static void
 usage(void) {
