--- ./Makefile.orig	2003-07-31 11:57:31.000000000 -0400
+++ ./Makefile	2003-07-31 12:08:53.000000000 -0400
@@ -253,14 +253,7 @@
 user_install::
 all :: user
 install :: all user_install
-	@echo "*** The depmod command below may generate errors. We are aware of the"
-	@echo "*** problem and are working on a solution. You should ignore them for now."
-ifeq ($(DESTDIR),)
-	-/sbin/depmod -a
-else
-	-/sbin/depmod -a -b $(DESTDIR)
-endif
-	@echo "*** Installation successful!"
+	@echo "*** RockLinux: depmod -a is pending!"
 	@echo "*** Important notes:"
 	@echo "*** 1* The libsensors configuration file ($(ETCDIR)/sensors.conf) is never"
 	@echo "***    overwritten by our installation process, so that you won't lose"
