--- ./kernel/include/Module.mk.orig	2003-07-31 23:36:24.000000000 -0400
+++ ./kernel/include/Module.mk	2003-07-31 23:37:10.000000000 -0400
@@ -21,7 +21,7 @@
 MODULE_DIR := kernel/include
 KERNELINCLUDEDIR := $(MODULE_DIR)
 
-KERNELINCLUDEFILES := $(MODULE_DIR)/i2c-dev.h $(MODULE_DIR)/sensors.h
+KERNELINCLUDEFILES := $(MODULE_DIR)/sensors.h
 
 $(KERNELINCLUDEDIR)/sensors.h: $(KERNELINCLUDEDIR)/sensors.h.template
 	cat $@.template > $@
