diff -urN live/groupsock/Groupsock.cpp live-jsaw/groupsock/Groupsock.cpp
--- live/groupsock/Groupsock.cpp	2003-11-25 10:04:41.000000000 +0100
+++ live-jsaw/groupsock/Groupsock.cpp	2004-03-29 01:37:48.000000000 +0200
@@ -26,9 +26,9 @@
 #if defined(__WIN32__) || defined(_WIN32)
 #include <strstrea.h>
 #else
-#include <strstream.h>
+#include <strstream>
 #endif
-#include <stdio.h>
+#include <cstdio>
 
 ///////// OutputSocket //////////
 
diff -urN live/groupsock/NetInterface.cpp live-jsaw/groupsock/NetInterface.cpp
--- live/groupsock/NetInterface.cpp	2003-11-25 10:04:41.000000000 +0100
+++ live-jsaw/groupsock/NetInterface.cpp	2004-03-29 01:37:34.000000000 +0200
@@ -24,7 +24,7 @@
 #if defined(__WIN32__) || defined(_WIN32)
 #include <strstrea.h>
 #else
-#include <strstream.h>
+#include <strstream>
 #endif
 
 ////////// NetInterface //////////
