--- heimdal-0.6-20031010/lib/hdb/hdb-ldap.c.orig	2003-10-30 04:12:47.000000000 +0000
+++ heimdal-0.6-20031010/lib/hdb/hdb-ldap.c	2003-10-30 04:13:35.000000000 +0000
@@ -310,6 +310,17 @@
 	if (ret != 0) {
 	    goto out;
 	}
+	/* inetOrgPerson is the 2nd structural object class */
+	ret = LDAP_addmod(&mods, LDAP_MOD_ADD, "objectClass", "inetOrgPerson");
+	if (ret != 0) {
+	    goto out;
+	}
+	/* hordePerson is the 3rd structural object class */
+	ret = LDAP_addmod(&mods, LDAP_MOD_ADD, "objectClass", "hordePerson");
+	if (ret != 0) {
+	    goto out;
+	}
+	/* sambaSamAccount should go here when SID generation is figured out */
 	ret =
 	    LDAP_addmod(&mods, LDAP_MOD_ADD, "objectClass",
 			"krb5Principal");
