--- xf86-video-sis/src/sis_memcpy.c.vanilla	2024-04-10 11:15:01.331968429 +0200
+++ xf86-video-sis/src/sis_memcpy.c	2024-04-10 11:20:49.686966900 +0200
@@ -486,7 +486,7 @@
     return(eax);
 }
 
-#elif defined(__AMD64__) || defined(__amd64__) || defined(__x86_64__) /***************** AMD64 */
+#elif (defined(__AMD64__) || defined(__amd64__) || defined(__x86_64__)) && !defined(__ILP32__)  /***************** AMD64 */
 
 #define SiS_checkosforsse	/* Does this cpu support sse and do we need to check os? */
 #define SiS_canBenchmark	/* Can we perform a benchmark? */
@@ -1014,7 +1014,7 @@
     return flags;
 }
 
-#elif defined(__AMD64__) || defined(__amd64__) || defined(__x86_64__) /* AMD64 specific ***** */
+#elif (defined(__AMD64__) || defined(__amd64__) || defined(__x86_64__)) && !defined(__ILP32__) /* AMD64 specific ***** */
 
 PREFETCH_FUNC(SiS_sse,SSE64,SSE,,FENCE,small_memcpy_amd64)
 
