--- nxproxy/configure.orig	2003-10-26 14:13:29.000000000 +0100
+++ nxproxy/configure	2003-10-26 14:14:07.000000000 +0100
@@ -2094,8 +2094,8 @@
 
 LIBVERSION=`echo ${VERSION} | cut -d '.' -f 1`
 
-CXXFLAGS="$CXXFLAGS -DVERSION=\\\"${VERSION}\\\""
-CPPFLAGS="$CPPFLAGS -DVERSION=\\\"${VERSION}\\\""
+CXXFLAGS="$CXXFLAGS -DVERSION=\\\"${VERSION}\\\" -DLIBVERSION=\\\"${LIBVERSION}\\\""
+CPPFLAGS="$CPPFLAGS -DVERSION=\\\"${VERSION}\\\" -DLIBVERSION=\\\"${LIBVERSION}\\\""
 
 
 
--- nxproxy/Main.c.orig	2003-10-26 14:10:16.000000000 +0100
+++ nxproxy/Main.c	2003-10-26 14:14:18.000000000 +0100
@@ -137,7 +137,7 @@
   }
   else
   {
-    version = VERSION;
+    version = LIBVERSION;
 
     if (version == NULL)
     {
