--- firefox-127.0/dom/media/platforms/ffmpeg/FFmpegRuntimeLinker.cpp.vanilla	2024-06-14 13:56:22.965354132 +0200
+++ firefox-127.0/dom/media/platforms/ffmpeg/FFmpegRuntimeLinker.cpp	2024-06-14 15:29:47.123157935 +0200
@@ -50,9 +50,11 @@
   "libavcodec.so.60",
   "libavcodec.so.59",
   "libavcodec.so.58",
+#if 0
   "libavcodec-ffmpeg.so.58",
   "libavcodec-ffmpeg.so.57",
   "libavcodec-ffmpeg.so.56",
+#endif
   "libavcodec.so.57",
   "libavcodec.so.56",
   "libavcodec.so.55",
