--- firefox-151.0/build/moz.configure/toolchain.configure.vanilla	2026-05-18 16:11:25.836619492 +0200
+++ firefox-151.0/build/moz.configure/toolchain.configure	2026-05-18 16:12:08.857540514 +0200
@@ -697,7 +697,7 @@
         toolchain = "%s%d" % (target.toolchain, android_version)
     elif target.kernel == "WASI" and info.type == "clang" and info.version >= Version("22.1"):
         # The wasm32-wasi target was renamed to wasm32-wasip1 in LLVM 22.1.
-        toolchain = "wasm32-wasip1"
+        toolchain = "wasm32-wasi"
     else:
         toolchain = target.toolchain
 
