--- ./pine/osdep/debuging.tim.orig	2002-11-28 01:19:37.000000000 +0100
+++ ./pine/osdep/debuging.tim	2003-08-16 13:53:23.000000000 +0200
@@ -195,20 +195,22 @@
 {
     va_list args;
 
-    va_start(args, fmt);
-
-    if(debug_fp && debug >= ps_global->dlevel && do_debug(debug_fp))
-      vfprintf(debug_fp, fmt, args);
+    if(debug_fp && debug >= ps_global->dlevel && do_debug(debug_fp)) {
+        va_start(args, fmt);
+        vfprintf(debug_fp, fmt, args);
+        va_end(args);
+    }
 
     if(ps_global->dlevel <= 9 || ps_global->dlevel <= debug){
 	char    b[64000];
 
+        va_start(args, fmt);
 	vsnprintf(b, sizeof(b), fmt, args);
+        va_end(args);
+
 	b[sizeof(b)-1] = '\0';
 	add_review_message(b, ps_global->dlevel);
     }
-
-    va_end(args);
 }
 #endif /* DEBUGJOURNAL */
 
