diff -ur coreutils-5.2.1/man/Makefile.in coreutils-5.2.1-patched/man/Makefile.in
--- coreutils-5.2.1/man/Makefile.in	2004-03-11 09:58:00.000000000 +0100
+++ coreutils-5.2.1-patched/man/Makefile.in	2004-09-13 12:06:52.229272249 +0200
@@ -265,14 +265,14 @@
 dist_man_MANS = \
   basename.1 cat.1 chgrp.1 chmod.1 chown.1 chroot.1 cksum.1 comm.1 \
   cp.1 csplit.1 cut.1 date.1 dd.1 df.1 dir.1 dircolors.1 dirname.1 du.1 \
-  echo.1 env.1 expand.1 expr.1 factor.1 false.1 fmt.1 fold.1 groups.1 \
-  head.1 hostid.1 hostname.1 id.1 install.1 join.1 kill.1 \
+  echo.1 env.1 expand.1 expr.1 factor.1 false.1 fmt.1 fold.1 \
+  head.1 hostid.1 id.1 install.1 join.1 kill.1 \
   link.1 ln.1 logname.1 \
   ls.1 md5sum.1 mkdir.1 mkfifo.1 mknod.1 mv.1 nice.1 nl.1 nohup.1 od.1 \
   paste.1 pathchk.1 pinky.1 pr.1 printenv.1 printf.1 ptx.1 pwd.1 readlink.1 \
   rm.1 rmdir.1 seq.1 sha1sum.1 shred.1 sleep.1 sort.1 split.1 stat.1 stty.1 \
-  su.1 sum.1 sync.1 tac.1 tail.1 tee.1 test.1 touch.1 tr.1 true.1 tsort.1 \
-  tty.1 uname.1 unexpand.1 uniq.1 unlink.1 uptime.1 users.1 vdir.1 wc.1 \
+  sum.1 sync.1 tac.1 tail.1 tee.1 test.1 touch.1 tr.1 true.1 tsort.1 \
+  tty.1 uname.1 unexpand.1 uniq.1 unlink.1 users.1 vdir.1 wc.1 \
   who.1 whoami.1 yes.1
 
 man_aux = $(dist_man_MANS:.1=.x)
diff -ur coreutils-5.2.1/src/Makefile.in coreutils-5.2.1-patched/src/Makefile.in
--- coreutils-5.2.1/src/Makefile.in	2004-03-11 09:59:23.000000000 +0100
+++ coreutils-5.2.1-patched/src/Makefile.in	2004-09-13 12:05:50.967013786 +0200
@@ -39,8 +39,8 @@
 POST_UNINSTALL = :
 host_triplet = @host@
 EXTRA_PROGRAMS = chroot$(EXEEXT) df$(EXEEXT) hostid$(EXEEXT) \
-	nice$(EXEEXT) pinky$(EXEEXT) stty$(EXEEXT) su$(EXEEXT) \
-	uname$(EXEEXT) uptime$(EXEEXT) users$(EXEEXT) who$(EXEEXT)
+	nice$(EXEEXT) pinky$(EXEEXT) stty$(EXEEXT) \
+	uname$(EXEEXT) users$(EXEEXT) who$(EXEEXT)
 bin_PROGRAMS = [$(EXEEXT) chgrp$(EXEEXT) chown$(EXEEXT) chmod$(EXEEXT) \
 	cp$(EXEEXT) dd$(EXEEXT) dircolors$(EXEEXT) du$(EXEEXT) \
 	ginstall$(EXEEXT) link$(EXEEXT) ln$(EXEEXT) dir$(EXEEXT) \
@@ -56,7 +56,7 @@
 	tail$(EXEEXT) tr$(EXEEXT) tsort$(EXEEXT) unexpand$(EXEEXT) \
 	uniq$(EXEEXT) wc$(EXEEXT) basename$(EXEEXT) date$(EXEEXT) \
 	dirname$(EXEEXT) echo$(EXEEXT) env$(EXEEXT) expr$(EXEEXT) \
-	factor$(EXEEXT) false$(EXEEXT) hostname$(EXEEXT) id$(EXEEXT) \
+	factor$(EXEEXT) false$(EXEEXT) id$(EXEEXT) \
 	kill$(EXEEXT) logname$(EXEEXT) pathchk$(EXEEXT) \
 	printenv$(EXEEXT) printf$(EXEEXT) pwd$(EXEEXT) seq$(EXEEXT) \
 	sleep$(EXEEXT) tee$(EXEEXT) test$(EXEEXT) true$(EXEEXT) \
@@ -225,7 +225,7 @@
 MSGMERGE = /usr/bin/msgmerge
 NEED_SETGID = false
 OBJEXT = o
-OPTIONAL_BIN_PROGS =  uname$(EXEEXT) chroot$(EXEEXT) hostid$(EXEEXT) nice$(EXEEXT) pinky$(EXEEXT) users$(EXEEXT) who$(EXEEXT) uptime$(EXEEXT) stty$(EXEEXT)
+OPTIONAL_BIN_PROGS =  uname$(EXEEXT) chroot$(EXEEXT) hostid$(EXEEXT) nice$(EXEEXT) pinky$(EXEEXT) users$(EXEEXT) who$(EXEEXT) stty$(EXEEXT)
 PACKAGE = coreutils
 PACKAGE_BUGREPORT = bug-coreutils@gnu.org
 PACKAGE_NAME = GNU coreutils
@@ -816,7 +816,7 @@
 sharedstatedir = @sharedstatedir@
 sysconfdir = @sysconfdir@
 target_alias = @target_alias@
-bin_SCRIPTS = groups
+bin_SCRIPTS = 
 noinst_HEADERS = \
   system.h checksum.h copy.h cp-hash.h ls.h dircolors.h remove.h \
   chown-core.h fs.h \
