--- gnucash-1.8.9/configure~	2004-04-12 04:11:07.000000000 +0200
+++ gnucash-1.8.9/configure	2004-10-24 22:51:02.808550192 +0200
@@ -11473,7 +11473,7 @@
   ac_save_CFLAGS="$CFLAGS"
   ac_save_LIBS="$LIBS"
   CFLAGS="$CFLAGS $OPENHBCI_CFLAGS"
-  LIBS="$OPENHBCI_LIBS $LIBS"
+  LIBS="$OPENHBCI_LIBS -lstdc++ $LIBS"
   rm -f conf.openhbcitest
   if test "$cross_compiling" = yes; then
   echo $ac_n "cross compiling; assumed OK... $ac_c"
@@ -11608,7 +11608,7 @@
     ac_save_CFLAGS="$CFLAGS"
     ac_save_LIBS="$LIBS"
     CFLAGS="$CFLAGS $OPENHBCI_CFLAGS"
-    LIBS="$LIBS $OPENHBCI_LIBS"
+    LIBS="$LIBS $OPENHBCI_LIBS -lstdc++"
     cat > conftest.$ac_ext <<EOF
 #line 11614 "configure"
 #include "confdefs.h"
