--- gnucash-1.8.10/src/import-export/hbci/druid-hbci-initial.c~	2004-12-12 00:04:28.000000000 +0100
+++ gnucash-1.8.10/src/import-export/hbci/druid-hbci-initial.c	2004-12-30 14:06:58.873958432 +0100
@@ -129,9 +129,13 @@
 
 static gchar *gnc_hbci_account_longname(const AB_ACCOUNT *hacc)
 {
+  const char *bankname;
+  const char *bankcode;
+
   g_assert(hacc);
-  const char *bankname = AB_Account_GetBankName (hacc);
-  const char *bankcode = AB_Account_GetBankCode (hacc);
+  bankname = AB_Account_GetBankName (hacc);
+  bankcode = AB_Account_GetBankCode (hacc);
+
   /* Translators: Strings are 1. Account code, 2. Bank name, 3. Bank code. */
   if (bankname)
     return g_strdup_printf(_("%s at %s (code %s)"),
