bash dislikes that '' -ne 0
Alejandro

--- ./alsaconf/alsaconf.in.orig	2005-02-19 23:03:30.000000000 -0300
+++ ./alsaconf/alsaconf.in	2005-02-19 23:04:55.000000000 -0300
@@ -54,9 +54,9 @@
   distribution="debian"
 elif [ -f /etc/mandrake-release ]; then
   distribution="mandrake"
-elif [ -f /etc/redhat-release -a `grep -c "Red Hat" /etc/redhat-release 2>/dev/null` -ne 0 ]; then
+elif [ -f /etc/redhat-release -a "`grep -c 'Red Hat' /etc/redhat-release 2>/dev/null`" != 0 ]; then
   distribution="redhat"
-elif [ -f /etc/fedora-release -a `grep -c "Fedora" /etc/fedora-release 2>/dev/null` -ne 0 ]; then
+elif [ -f /etc/fedora-release -a "`grep -c 'Fedora' /etc/fedora-release 2>/dev/null`" != 0 ]; then
   distribution="fedora"
 else
   distribution="unknown"
