--- alsa-utils-1.2.9/alsactl/init_sysdeps.c.vanilla	2023-08-04 23:14:42.829273016 +0200
+++ alsa-utils-1.2.9/alsactl/init_sysdeps.c	2023-08-04 23:22:11.137281323 +0200
@@ -17,7 +17,8 @@
  *
  */
 
-#if defined(__GLIBC__) && !(defined(__UCLIBC__) && defined(__USE_BSD))
+#if defined __GLIBC__ && ((__GLIBC__ >= 2) && (__GLIBC_MINOR__ >= 27))
+#else
 static size_t strlcpy(char *dst, const char *src, size_t size)
 {
 	size_t bytes = 0;
