diff -pruN e2fsprogs-1.32_orig/e2fsck/Makefile.in e2fsprogs-1.32/e2fsck/Makefile.in
--- e2fsprogs-1.32_orig/e2fsck/Makefile.in	2003-03-27 11:47:03.000000000 +0100
+++ e2fsprogs-1.32/e2fsck/Makefile.in	2003-03-27 11:46:39.000000000 +0100
@@ -95,7 +95,6 @@ SRCS= $(srcdir)/e2fsck.c \
 	$(MTRACE_SRC)
 
 all:: profiled $(PROGS) e2fsck.static e2fsck.shared $(MANPAGES)
-	@sync
 
 @PROFILE_CMT@all:: e2fsck.profiled
 
