--- ./bindings/python/Makefile.in.org	Fri Jan  3 17:29:08 2003
+++ ./bindings/python/Makefile.in	Tue May 27 10:14:37 2003
@@ -180,7 +180,7 @@
 	$(SDL_LIBS) \
 	$(IMAGE_LIBS) \
 	$(LIBSTDCPP) \
-	-lparagui -L$(top_srcdir)/src
+	-lparagui -L$(top_srcdir)/src/.libs
 
 
 INCLUDES = \
--- ./bindings/ruby/Makefile.in.org	Fri Jan  3 17:29:08 2003
+++ ./bindings/ruby/Makefile.in	Tue May 27 10:15:32 2003
@@ -177,7 +177,7 @@
 
 LINK = $(COMPILE) -shared \
 	-L@prefix@/lib \
-	-L$(top_srcdir)/src \
+	-L$(top_srcdir)/src/.libs \
 	$(PARAGUI_LIBS)
 
 
