--- kbd-1.08/src/kbdrate.c	2002-10-11 13:09:01.000000000 +0200
+++ kbd-1.08-fixed/src/kbdrate.c	2003-07-19 11:03:05.000000000 +0200
@@ -237,6 +237,7 @@
 	if(KIOCSRATE_ioctl_ok(rate, delay, silent))	/* sparc? */
 		return 0;
 
+#ifdef __i386__ 
 
 	/* The ioport way */
 
@@ -286,4 +287,8 @@
 			valid_delays[ (value & 0x60) >> 5 ] );
 
 	return 0;
+#else
+	perror( _("No alternative method for this platform") );
+	exit( 1 );
+#endif
 }
