--- ./configure.in (revision 7881) 
+++ ./configure.in (revision 7960)
@@ -167,4 +167,8 @@
 
 	AC_CHECK_HEADER($PGSQL_INCLUDEDIR/libpq-fe.h, pgsql_hdr=true)
+	if test x$pgsql_hdr != xtrue; then
+	   PGSQL_INCLUDEDIR="$PGSQL_INCLUDEDIR/postgresql"
+	   AC_CHECK_HEADER($PGSQL_INCLUDEDIR/libpq-fe.h, pgsql_hdr=true)
+	fi
 fi
 
--- ./plugins/sql/pgsql/pgsql.c (revision 7799)
+++ ./plugins/sql/pgsql/pgsql.c (revision 7960)
@@ -37,5 +37,5 @@
 #include <libprelude/prelude-error.h>
 #include <libprelude/idmef.h>
-#include <postgresql/libpq-fe.h>
+#include <libpq-fe.h>
 
 #include "preludedb-sql-settings.h"
