--- diskdev_cmds-332.25/fsck_hfs.tproj/Makefile.lnx	2017-08-09 19:59:41.782161654 +0000
+++ diskdev_cmds-332.25/fsck_hfs.tproj/Makefile.lnx	2017-08-09 19:42:09.138179153 +0000
@@ -4,7 +4,7 @@
 all: fsck_hfs
 
 fsck_hfs: $(OFILES) dfalib/libdfa.a
-	$(CC) $(CFLAGS) $(LDFLAGS) -o $@ $(OFILES) dfalib/libdfa.a -lbsd
+	$(CC) $(CFLAGS) $(LDFLAGS) -o $@ $(OFILES) dfalib/libdfa.a # -lbsd
 
 dfalib/libdfa.a: FORCE
 	$(MAKE) -C dfalib -f Makefile.lnx CFLAGS="$(CFLAGS)" LDFLAGS="$(LDFLAGS)" libdfa.a
--- diskdev_cmds-332.25/fsck_hfs.tproj/dfalib/SVerify1.c	2017-08-09 19:59:41.818161653 +0000
+++ diskdev_cmds-332.25/fsck_hfs.tproj/dfalib/SVerify1.c	2017-08-09 19:48:03.398173264 +0000
@@ -2176,7 +2176,8 @@
 	/* Always update the last recordType, fileID and attribute name before exiting */
 	prevAttr->recordType = rec->recordType;
 	prevAttr->fileID = fileID;
-	(void) strlcpy((char *)prevAttr->attrname, (char *)attrname, sizeof(prevAttr->attrname));
+	(void) strncpy((char *)prevAttr->attrname, (char *)attrname, sizeof(prevAttr->attrname));
+	((char *)prevAttr->attrname)[sizeof(prevAttr->attrname) - 1] = 0;
 
 	goto out;
 
@@ -3201,7 +3202,8 @@
 		if (extentInfo.attrname == NULL) {
 			return(memFullErr);
 		}
-		strlcpy(extentInfo.attrname, attrname, attrlen);
+		strncpy(extentInfo.attrname, attrname, attrlen);
+		extentInfo.attrname[attrlen - 1] = 0;
 	}
 	
 	//	If it's uninitialized
--- diskdev_cmds-332.25/include/missing.h	2017-08-09 19:59:41.822161653 +0000
+++ diskdev_cmds-332.25/include/missing.h	2017-08-09 19:41:34.210179734 +0000
@@ -6,7 +6,7 @@
 #include <errno.h>
 #include <stdint.h>
 #include <string.h>
-#include <bsd/string.h>
+//#include <bsd/string.h>
 
 #define MAXBSIZE		(256 * 4096)
 
