--- subversion-0.21.0/configure	2003-04-16 03:16:06.000000000 +0200
+++ subversion-0.21.0-new/configure	2003-05-04 11:31:27.000000000 +0200
@@ -8844,7 +8844,7 @@
         ;;
       esac
 
-      for db_libname in db4 db; do
+      for db_libname in db40 db4 db; do
         # We generate a separate cache variable for each prefix and libname
         # we search under.  That way, we avoid caching information that
         # changes if the user runs `configure' with a different set of
