diff -pruN koules-cvs-2003-04-16-orig/koules.c koules-cvs-2003-04-16/koules.c
--- koules-cvs-2003-04-16-orig/koules.c	2003-07-31 07:33:29.000000000 +0200
+++ koules-cvs-2003-04-16/koules.c	2003-07-31 07:28:35.000000000 +0200
@@ -1213,8 +1213,8 @@ creators_points (int radius, int x1, int
 	r=sqrt((double)((x-midX)*(x-midX) + (y-midY)*(y-midY)));
 
 /* fix by Alexey Bokov 25-06-98 */
-//        r=(r*radius/r1)/r*0.9; // why do we multyply by r then divide by r?
-				 // causing SIGFPE when r==0...
+/*        r=(r*radius/r1)/r*0.9; // why do we multyply by r then divide by r?
+				 // causing SIGFPE when r==0... */
 	r=(radius/r1)*0.9;
 	x2=x1+(x-midX)*r;
 	y2=y1+(y-midY)*r;
