--- alsa-tools-1.0.9/ac3dec/libac3/bitstream.c.orig	2005-09-08 15:43:54.000000000 -0700
+++ alsa-tools-1.0.9/ac3dec/libac3/bitstream.c	2005-09-08 15:48:41.000000000 -0700
@@ -94,7 +94,8 @@
 static inline void
 bitstream_fill_current()
 {
-	current_word = *((uint_32*)buffer_start)++;
+	current_word = *((uint_32*)buffer_start);
+	buffer_start += sizeof(uint_32);
 	current_word = swab32(current_word);
 }
 
