--- gst-plugins-0.8.2/ext/arts/Makefile.in~	2004-06-23 17:01:29.000000000 +0200
+++ gst-plugins-0.8.2/ext/arts/Makefile.in	2004-07-23 03:08:15.528997968 +0200
@@ -1013,7 +1013,7 @@
 # If someone has time and energy, please check if -pthread should be allowed
 # in compile flags at all.
 .idl.cc:
-	mcopidl -t `echo $(ARTS_CXXFLAGS) | sed 's/-pthread//g'` $<
+	mcopidl -t `echo $(ARTS_CXXFLAGS) | sed 's/-pthread//g; s/-D_REENTRANT//g;'` $<
 
 dist-hook:
 	rm -f $(distdir)/gst_artsio.cc
