--- ./configure.orig	2003-02-10 12:00:22.000000000 +0100
+++ ./configure	2003-02-10 12:01:29.000000000 +0100
@@ -5149,14 +5149,14 @@
 fi
 
 
-echo $ac_n "checking for db_version in -ldb""... $ac_c" 1>&6
-echo "configure:5154: checking for db_version in -ldb" >&5
+echo $ac_n "checking for db_version in -ldb-3""... $ac_c" 1>&6
+echo "configure:5154: checking for db_version in -ldb-3" >&5
 ac_lib_var=`echo db'_'db_version | sed 'y%./+-%__p_%'`
 if eval "test \"`echo '$''{'ac_cv_lib_$ac_lib_var'+set}'`\" = set"; then
   echo $ac_n "(cached) $ac_c" 1>&6
 else
   ac_save_LIBS="$LIBS"
-LIBS="-ldb  $LIBS"
+LIBS="-ldb-3  $LIBS"
 cat > conftest.$ac_ext <<EOF
 #line 5162 "configure"
 #include "confdefs.h"
@@ -5190,7 +5190,7 @@
 #define $ac_tr_lib 1
 EOF
 
-  LIBS="-ldb $LIBS"
+  LIBS="-ldb-3 $LIBS"
 
 else
   echo "$ac_t""no" 1>&6
