--- gpdf-2.10.0/xpdf/gpdf-control.cc.orig	2006-08-29 19:48:00.000000000 +0200
+++ gpdf-2.10.0/xpdf/gpdf-control.cc	2006-08-29 19:49:14.000000000 +0200
@@ -208,12 +208,12 @@
 			   const gchar * status)
 {
 	GPdfControlPrivate *priv;
-	guint to_id;
+	unsigned long to_id;
 
 	g_return_if_fail (GPDF_IS_NON_NULL_CONTROL (control));
 	priv = control->priv;
 
-	if ((to_id = (unsigned int)
+	if ((to_id = (unsigned long)
 	     g_object_get_data (G_OBJECT (control),
 				"status-timeout-id")) != 0) {
 		g_object_set_data (G_OBJECT (control),
