--- gpdf-2.10.0/xpdf/TextOutputDev.h.old	2004-05-17 21:37:57.000000000 +0200
+++ gpdf-2.10.0/xpdf/TextOutputDev.h	2006-03-24 23:40:41.000000000 +0100
@@ -26,6 +26,10 @@
 class GfxState;
 class UnicodeMap;
 
+class TextBlock;
+class TextPage;
+class TextLineFrag;
+
 //------------------------------------------------------------------------
 
 typedef void (*TextOutputFunc)(void *stream, char *text, int len);
