--- taper-6.9b/Makefile.orig	2005-09-06 12:58:10.000000000 -0700
+++ taper-6.9b/Makefile	2005-09-06 12:58:33.000000000 -0700
@@ -83,7 +83,7 @@
 	install --mode 755 --strip taper $(BINDIR) 
 	install --mode 755 --strip bg_backup $(BINDIR)
 	install --mode 755 --strip bg_restore $(BINDIR)
-	install --mode 644 docs/taper.8 $(MANDIR)
+	install -D --mode 644 docs/taper.8 $(MANDIR)/taper.8
 	ln -sf $(BINDIR)/taper $(ALTBIN)/taper
 	ln -sf $(BINDIR)/bg_backup $(ALTBIN)/bg_backup
 	ln -sf $(BINDIR)/bg_restore $(ALTBIN)/bg_restore
