--- ./configure.orig	Thu Feb 13 21:55:07 2003
+++ ./configure	Thu Feb 13 22:07:52 2003
@@ -441,6 +441,11 @@
   -docdir=* | --docdir=* | --infodi=* | --infod=* | --doc=* | --inf=*)
     docdir=$ac_optarg ;;
 
+  -infodir | --infodir | --infodi | --infod | --info | --inf)
+    ac_prev=infodir ;;
+  -infodir=* | --infodir=* | --infodi=* | --infod=* | --info=* | --inf=*)
+    infodir=$ac_optarg ;;
+
   -libdir | --libdir | --libdi | --libd)
     ac_prev=libdir ;;
   -libdir=* | --libdir=* | --libdi=* | --libd=*)
