diff -urN straw.orig/Makefile straw/Makefile
--- straw.orig/Makefile	2003-11-25 11:22:22.000000000 +0100
+++ straw/Makefile	2003-11-27 07:11:26.000000000 +0100
@@ -1,4 +1,4 @@
-PYTHON ?= python2.2
+PYTHON ?= python2.3
 INSTALL ?= install
 RM ?= rm
 MSGFMT ?= msgfmt
@@ -11,14 +11,14 @@
 PREFIX ?= `pkg-config libgnome-2.0 --variable=prefix || echo /usr/local`
 
 # If you edited PYTHON above, edit this to the corresponding python lib dir
-LIBDIR = $(PREFIX)/lib/python2.2/site-packages/straw
+LIBDIR = $(PREFIX)/lib/python2.3/site-packages/straw
 BINDIR = $(PREFIX)/bin
 DATADIR = $(PREFIX)/share/straw
 APPLICATIONSDIR = $(PREFIX)/share/applications
 ICONDIR = $(PREFIX)/share/pixmaps
 LOCALEDIR = $(PREFIX)/share/locale
 # change this to just "/etc" if you're installing in /usr
-SYSCONFDIR = $(PREFIX)/etc
+SYSCONFDIR = /etc/$(PREFIX)
 
 POFILES := $(shell echo po/*.po)
 MOFILES := $(POFILES:.po=.mo)
@@ -27,8 +27,8 @@
 schema_DATA = data/straw.schemas 
 # modify this if you're running your own build environment (garnome/jhbuild)
 # change it to "xml://path/to/gnome_build_env/etc/gconf/gconf.xml.defaults"
-GCONF_SCHEMA_CONFIG_SOURCE = xml::/etc/gconf/gconf.xml.defaults 
-GCONF_SCHEMA_FILE_DIR = /etc/gconf/schemas/
+GCONF_SCHEMA_CONFIG_SOURCE = xml::$(SYSCONFDIR)/gconf/gconf.xml.defaults 
+GCONF_SCHEMA_FILE_DIR = $(SYSCONFDIR)/gconf/schemas/
 GCONFTOOL = gconftool-2
 
 SCHEMADIR = $(GCONF_SCHEMA_FILE_DIR)
diff -urN straw.orig/src/straw straw/src/straw
--- straw.orig/src/straw	2002-09-12 01:15:49.000000000 +0200
+++ straw/src/straw	2003-11-27 07:00:11.000000000 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/env python2.2
+#!/usr/bin/env python
 #
 # Copyright (c) 2002 Juri Pakaste
 # You may use and distribute this software under the terms of the
