--- straw-0.19.1/Makefile.orig	2003-09-30 21:45:17.000000000 +0200
+++ straw-0.19.1/Makefile	2003-09-30 21:45:26.000000000 +0200
@@ -1,4 +1,4 @@
-PYTHON ?= python2.2
+PYTHON ?= python2.3
 INSTALL ?= install
 RM ?= rm
 MSGFMT ?= msgfmt
@@ -10,7 +10,7 @@
 # autodetect GNOME prefix, change this if you want it elsewhere
 PREFIX ?= `pkg-config libgnome-2.0 --variable=prefix || echo /usr/local`
 
-LIBDIR = $(PREFIX)/lib/python2.2/site-packages/straw
+LIBDIR = $(PREFIX)/lib/python2.3/site-packages/straw
 BINDIR = $(PREFIX)/bin
 DATADIR = $(PREFIX)/share/straw
 APPLICATIONSDIR = $(PREFIX)/share/applications
