diff --git a/configure.ac b/configure.ac
index 7700b53..bb8ab36 100644
--- a/configure.ac
+++ b/configure.ac
@@ -31,7 +31,7 @@ AC_CONFIG_HEADERS([config.h])
 AC_CONFIG_AUX_DIR(.)
 
 # Initialize Automake
-AM_INIT_AUTOMAKE([foreign dist-bzip2])
+AM_INIT_AUTOMAKE([foreign dist-xz])
 
 # Require X.Org macros 1.8 or later for MAN_SUBSTS set by XORG_MANPAGE_SECTIONS
 m4_ifndef([XORG_MACROS_VERSION],
@@ -78,24 +78,6 @@ if test "x$XSERVER_LIBPCIACCESS" = xyes; then
 fi
 AM_CONDITIONAL(XSERVER_LIBPCIACCESS, test "x$XSERVER_LIBPCIACCESS" = xyes)
 
-AC_ARG_ENABLE(xaa,
-              AS_HELP_STRING([--enable-xaa],
-                             [Enable legacy X Acceleration Architecture (XAA) [default=auto]]),
-              [XAA="$enableval"],
-              [XAA=auto])
-if test "x$XAA" != xno; then
-        save_CFLAGS=$CFLAGS
-        save_CPPFLAGS=$CPPFLAGS
-        CFLAGS=$XORG_CFLAGS
-        CPPFLAGS="$XORG_CFLAGS"
-        AC_CHECK_HEADERS([xaa.h], XAA=yes, XAA=no)
-        CFLAGS=$save_CFLAGS
-        CPPFLAGS=$save_CPPFLAGS
-fi
-AC_MSG_CHECKING([whether to include XAA support])
-AM_CONDITIONAL(XAA, test "x$XAA" = xyes)
-AC_MSG_RESULT([$XAA])
-
 AC_SUBST([moduledir])
 
 DRIVER_NAME=s3
diff --git a/man/s3.man b/man/s3.man
index 732e4f3..e02646a 100644
--- a/man/s3.man
+++ b/man/s3.man
@@ -1,8 +1,8 @@
 .\" shorthand for double quote that works everywhere.
 .ds q \N'34'
-.TH s3 __drivermansuffix__ __vendorversion__
+.TH S3 __drivermansuffix__ 2009-06-25 __vendorversion__
 .SH NAME
-s3 \- S3 video driver
+s3 \- S3 video driver for Xorg
 .SH SYNOPSIS
 .B Section \*qDevice\*q
 .RS 2
@@ -15,7 +15,7 @@ s3 \- S3 video driver
 [\fBOption\fP "optionname" ["optionvalue"]]
 .RE
 .B EndSection
-
+.
 .SH SUPPORTED HARDWARE
 The
 .B s3
@@ -47,36 +47,35 @@ driver supports PCI video cards based on the following S3 chips:
 .TP 12
 .B Vision968
 86C968
-
+.sp
+.PP
+The driver also supports the following RAMDACs:
 .PP
-Also driver supports the following RAMDACs:
-.IP 12
 .B IBM 524, IBM 524A, IBM 526, IBM 526DB
-.IP 12
-.B TI ViewPoint 3025 
-
+.PP
+.B TI ViewPoint 3025
+.
 .SH DESCRIPTION
-\fBs3\fP is an __xservername__ driver for S3 based video cards. The
-driver provides full accelerated support for the following colour
-depths: 8, 15, 16 and 24. The overlay video (Xv) is supported in
-depths 16 and 24 for the Trio64V+, Trio64UV+, Trio64V2/DX and
-Trio64V2/GX chips.
-
+\fBs3\fP is an Xorg driver for S3 based video cards.
+The driver provides full accelerated support for the following colour depths:
+8, 15, 16, and 24.
+The overlay video (Xv) is supported in depths 16 and 24 for the
+Trio64V+, Trio64UV+, Trio64V2/DX, and Trio64V2/GX chips.
+.
 .SH CONFIGURATION DETAILS
-Please refer to __xconfigfile__(__filemansuffix__) for general
-configuration details.  This section only covers configuration details
-specific to this driver. All options names are case and white space
-insensitive when parsed by the server, for example, "trio 32/64" and
-"Trio32/64" are equivalent.
-
-.SS
-Chip overriding
-.PP
+Please refer to
+.BR xorg.conf (__filemansuffix__)
+for general configuration details.
+This section only covers configuration details specific to this driver.
+All options names are case and white space insensitive
+when parsed by the server, for example,
+"trio 32/64" and "Trio32/64" are equivalent.
+.sp
+.SS Chip overriding
 The driver auto-detects the chipset and RAMDAC, but the following
 \fBChipset\fP names may optionally be specified in the configuration
 file \fB\*qDevice\*q\fP section, and will override the auto-detection:
-
-.PP
+.sp
 .RS 4
 .TP 18
 "964-0", "964-1"
@@ -87,47 +86,47 @@ file \fB\*qDevice\*q\fP section, and will override the auto-detection:
 .TP 18
 "Trio32/64"
 86C732/86C764
-.TP 18 
+.TP 18
 "Aurora64V+"
 86CM65
 .TP 18
 "Trio64UV+"
-86C767 
+86C767
 .TP 18
 "Trio64V2/DX/GX"
 86C775/86C785
 .RE
-
+.sp
 .PP
 The 86C765 (Trio64V+) is \fBChipset\fP "Trio32/64" with \fBChipRev\fP
 greater or equal 0x40.
-
-An overriding of RAMDAC auto-detection currently is not implemented.
-
-.SS
-Colour depth
 .PP
+An overriding of RAMDAC auto-detection currently is not implemented.
+.
+.SS Colour depth
+.
 For every supported colour depth the X server automatically selects an
-appropriate number of bits per pixel (bpp) for framebuffer. The depth
-8 is represented by 8 bpp framebuffer (1 byte/pixel), the depths 15
-and 16 (respectively 5.5.5 and 5.6.5 pixel formats) are represented by
-the 16 bpp framebuffer (2 bytes/pixel). Whenever the depth 24 is
-configured, the X server by default uses the 32 bpp framebuffer (4
-bytes/pixel), and the hardware is adjusted accordingly to the X.8.8.8
-pixel format, where X is the ignored upper byte.
-
+appropriate number of bits per pixel (bpp) for framebuffer.
+The depth 8 is represented by 8 bpp framebuffer (1 byte/pixel),
+the depths 15 and 16 (respectively 5.5.5 and 5.6.5 pixel formats)
+are represented by the 16 bpp framebuffer (2 bytes/pixel).
+Whenever the depth 24 is configured,
+the X server by default uses the 32 bpp framebuffer (4 bytes/pixel),
+and the hardware is adjusted accordingly to the X.8.8.8 pixel format,
+where X is the ignored upper byte.
+.
 .PP
 As the second option for the depth 24, the X server can be configured
 to use 24 bpp framebuffer with the 8.8.8 pixel format (3 bytes/pixel),
-which is supported by Trio64V+, Trio64UV+, Trio64V2/DX and
-Trio64V2/GX. If the video board has limited video RAM the 24 bpp
+which is supported by Trio64V+, Trio64UV+, Trio64V2/DX, and Trio64V2/GX.
+If the video board has limited video RAM the 24 bpp
 framebuffer has an advantage over 32 bpp framebuffer as it requires
 less memory to store screen, and, therefore, more space will be
-available for a video frame (see "Overlay video" section).  The 24 bpp
-framebuffer can be selected by X server's option \fB-fbbpp 24\fP or by
-specifying \fBDefaultFbBpp\fP option in
-__xconfigfile__(__filemansuffix__):
-
+available for a video frame (see "Overlay video" section).
+The 24 bpp framebuffer can be selected by X server's option \fB-fbbpp 24\fP
+or by specifying \fBDefaultFbBpp\fP option in
+.BR xorg.conf (__filemansuffix__):
+.sp
 .RS 4
 .B Section \*qScreen\*q
 .RS 2
@@ -139,94 +138,99 @@ __xconfigfile__(__filemansuffix__):
 .RE
 .B EndSection
 .RE
-
+.sp
 .PP
-However, the 24 bpp framebuffer mode has the hardware limitations: (i)
-the 2D acceleration doesn't work with this mode (use the
-\fB\*qshadowFB\*q\fP option to speed up drawing routines in this
-case); (ii) 24 bpp framebuffer cannot be used with either interlaced
+However, the 24 bpp framebuffer mode has the hardware limitations:
+(i)\ the 2D acceleration doesn't work with this mode (use the
+\fB\*qshadowFB\*q\fP option to speed up drawing routines in this case);
+(ii)\ 24 bpp framebuffer cannot be used with either interlaced
 or doublescan graphics modes.
-
-.SS
-Overlay video
-.PP
+.
+.SS Overlay video
+.
 If your video board has limited RAM it would be useful to estimate how
-large video frame might be placed in offscreen video memory. Suppose
-that the video board has 2 MB of RAM, and X server is configured to
-display 800x600 with the depth 24. By default the 32 bpp framebuffer
-will be chosen for this depth, so 800 x 600 x 4 = 1875 kB will be
-reserved for screen, and 2 MB - 1875 kB = 173 kB will remain for the
-offscreen area. This is sufficient space, for example, for the VCD
-NTSC 352x240 frame. If you need to upscale video with a bigger frame,
+large video frame might be placed in offscreen video memory.
+Suppose that the video board has 2 MB of RAM,
+and X server is configured to display 800x600 with the depth 24.
+By default the 32 bpp framebuffer will be chosen for this depth,
+so 800 x 600 x 4 = 1875 kB will be reserved for screen,
+and 2 MB - 1875 kB = 173 kB will remain for the offscreen area.
+This is sufficient space, for example, for the VCD NTSC 352x240 frame.
+If you need to upscale video with a bigger frame,
 but it doesn't fit the offscreen area, the only way to do this is to
 lower either the resolution or framebuffer's depth or both before the
 viewing, otherwise, you will get the allocation error and will not see
-the video. For example, with 24 bpp framebuffer and the same
-resolution 800x600 the offscreen area will be about 640 kB - it's
-enough for the 640x480 frame. 800x600 with the 16 bpp framebuffer
-gives about 1110 kB of offscreen area - this allows to upscale up to
-768x576 movies. Note, that all movie resolutions in examples above are
-provided as a reference; the movies can have the different aspect
-ratios and non-standard dimensions. A total pixel amount is the main
-consideration not the certain width and height.
+the video.
+For example, with 24 bpp framebuffer and the same resolution 800x600
+the offscreen area will be about 640 kB \(em it's enough for the 640x480 frame.
+800x600 with the 16 bpp framebuffer gives about 1110 kB of offscreen area
+\(em this allows to upscale up to 768x576 movies.
+Note, that all movie resolutions in examples above are provided as a reference;
+the movies can have the different aspect ratios and non-standard dimensions.
+A total pixel amount is the main consideration,
+not the certain width and height.
 .PP
 Due to hardware limitation the overlay video will not work with the
-interlaced/doublescan modes. Downscaling is not implemented in
-hardware.
-
-.SS
-Configuration options
-.PP
+interlaced/doublescan modes.
+Downscaling is not implemented in hardware.
+.sp
+.SS Configuration options
+.
 The following display
 .B Options
 are supported:
 .TP
 .BI "Option \*qHWCursor\*q \*q" boolean \*q
-Enable or disable the hardware cursor. Currently, hardware cursor is
-not implemented, so the option will be ignored. Default: off (software
-cursor).
+Enable or disable the hardware cursor.
+Currently, hardware cursor is not implemented, so the option will be ignored.
+Default: off (software cursor).
 .TP
 .BI "Option \*qNoAccel\*q \*q" boolean \*q
-Disable acceleration. Very useful for determining if the driver has
-problems with drawing and acceleration routines. This is the first
-option to try if your server runs but you see graphic corruption on
-the screen. Using it decreases performance, as it uses software
-emulation for drawing operations the video driver can accelerate with
-hardware. Default: off (acceleration is enabled).
+Disable acceleration.
+Very useful for determining if the driver has
+problems with drawing and acceleration routines.
+This is the first option to try if your server runs
+but you see graphic corruption on the screen.
+Using it decreases performance, as it uses software emulation for
+drawing operations the video driver can accelerate with hardware.
+Default: off (acceleration is enabled).
 .TP
 .BI "Option \*qShadowFB\*q \*q" boolean \*q
-Use shadow framebuffer. Disables hardware acceleration. Use this
-option when the hardware acceleration is not available or
-undesirable. Default: off.
+Use shadow framebuffer.
+Disables hardware acceleration.
+Use this option when the hardware acceleration is not available or undesirable.
+Default: off.
 .TP
 .BI "Option \*qXVideo\*q \*q" boolean \*q
-Enable or disable Xv support. Default: on.
-
+Enable or disable Xv support.
+Default: on.
+.sp
 .PP
 The following video memory \fBOptions\fP are supported:
 .TP
 .BI "Option \*qslow_dram_refresh\*q \*q" boolean \*q
-Enable three refresh cycles per scanline. Default: off (one refresh
-cycle).
+Enable three refresh cycles per scanline.
+Default: off (one refresh cycle).
 .TP
 .BI "Option \*qslow_edodram\*q \*q" boolean \*q
-Switch to 2-cycle EDO mode. Try this if you encounter pixel
-corruption. Using this option will cause a decrease in
-performance. Default: off (BIOS defaults).
+Switch to 2-cycle EDO mode.
+Try this if you encounter pixel corruption.
+Using this option will cause a decrease in performance.
+Default: off (BIOS defaults).
 .TP
 .BI "Option \*qslow_dram\*q \*q" boolean \*q
-For Trio and Aurora64V+ chips: increase -RAS Precharge Timing to 3.5
-MCLK. Try this option if you encounter pixel errors. Default: off
-(BIOS defaults).
+For Trio and Aurora64V+ chips: increase -RAS Precharge Timing to 3.5\ MCLK.
+Try this option if you encounter pixel errors.
+Default: off (BIOS defaults).
 .TP
 .BI "Option \*qslow_vram\*q \*q" boolean \*q
-For Vision964, Vision968 chips: increase -RAS Low Timing to 4.5
-MCLK. Default: off (BIOS defaults).
-
+For Vision964, Vision968 chips: increase -RAS Low Timing to 4.5\ MCLK.
+Default: off (BIOS defaults).
 .SH SEE ALSO
-__xservername__(1), __xconfigfile__(__filemansuffix__), Xserver(1),
-X(__miscmansuffix__)
-
+.BR Xorg (1),
+.BR xorg.conf (__filemansuffix__),
+.BR Xserver (1),
+.BR X (__miscmansuffix__)
 .SH AUTHORS
 Thomas Roell, Mark Vojkovich, Kevin E. Martin, Amancio Hasty, Jon
 N. Tombs and others were the original authors of driver for XFree86
@@ -235,6 +239,5 @@ modifications were made by the following contributors: Adam Jackson,
 Alan Coopersmith, Dave Airlie, Andrew Radrianasulu, Paulo Cesar
 Pereira de Andrade, Eric Anholt, Søren Sandmann Pedersen, Alex
 Deucher, Evgeny M. Zubok, Daniel Stone and others.
-
 .PP
 The manual was written by Evgeny M. Zubok <evgeny.zubok@tochka.ru>
diff --git a/src/s3.h b/src/s3.h
index 72a1072..fc746a5 100644
--- a/src/s3.h
+++ b/src/s3.h
@@ -35,10 +35,7 @@
 #include "xf86.h"
 #include "xf86Pci.h"
 #include "xf86RamDac.h"
-#ifdef HAVE_XAA_H
 #include "xaa.h"
-#endif
-#include "xf86fbman.h"
 #include "vbe.h"
 #include "vgaHW.h"
 
@@ -131,9 +128,7 @@ typedef struct _S3Rec {
 	Bool			PCIRetry;
 	Bool			ColorExpandBug;
 
-#ifdef HAVE_XAA_H
         XAAInfoRecPtr           pXAA;
-#endif
 	xf86CursorInfoPtr	pCurs;
 	xf86Int10InfoPtr	pInt10;
         XF86VideoAdaptorPtr     adaptor;
@@ -176,7 +171,7 @@ typedef struct _S3Rec {
 	Bool			(*CursorInit)(ScreenPtr pScreen);
 
 	void			(*LoadPalette)(ScrnInfoPtr pScrn, int numColors,
-					       int *indicies, LOCO *colors,
+					       int *indices, LOCO *colors,
 					       VisualPtr pVisual);
 
 	Bool                    (*CloseScreen)(CLOSE_SCREEN_ARGS_DECL);
@@ -244,7 +239,7 @@ void S3TiDAC_PreInit(ScrnInfoPtr pScrn);
 void S3TiDAC_Init(ScrnInfoPtr pScrn, DisplayModePtr mode);
 void S3TiDAC_Save(ScrnInfoPtr pScrn);
 void S3TiDAC_Restore(ScrnInfoPtr pScrn);
-void S3TiLoadPalette(ScrnInfoPtr pScrn, int numColors, int *indicies, LOCO *colors,
+void S3TiLoadPalette(ScrnInfoPtr pScrn, int numColors, int *indices, LOCO *colors,
 		     VisualPtr pVisual);
 Bool S3Ti_CursorInit(ScreenPtr pScreen);
 void S3OutTiIndReg(ScrnInfoPtr pScrn, CARD32 reg, unsigned char mask,
diff --git a/src/s3_Ti.c b/src/s3_Ti.c
index bb0ce51..5c346e7 100644
--- a/src/s3_Ti.c
+++ b/src/s3_Ti.c
@@ -577,7 +577,7 @@ void S3TiDAC_Init(ScrnInfoPtr pScrn, DisplayModePtr mode)
 }
 
 
-void S3TiLoadPalette(ScrnInfoPtr pScrn, int numColors, int *indicies, LOCO *colors,
+void S3TiLoadPalette(ScrnInfoPtr pScrn, int numColors, int *indices, LOCO *colors,
 		     VisualPtr pVisual)
 {
 	int i;
diff --git a/src/s3_accel.c b/src/s3_accel.c
index 5d11ac4..9b784ab 100644
--- a/src/s3_accel.c
+++ b/src/s3_accel.c
@@ -36,7 +36,6 @@
 #include "s3.h"
 #include "s3_reg.h"
 
-#ifdef HAVE_XAA_H
 
 #if 0
 static Bool NicePattern;
@@ -539,7 +538,6 @@ static void S3SubsequentDashedBresenhamLine32(ScrnInfoPtr pScrn,
 	}
 }
 
-#endif
 #endif
  	
 #ifdef S3_NEWMMIO
@@ -549,7 +547,7 @@ Bool S3AccelInitPIO(ScreenPtr pScreen)
 #endif
 {
 #ifdef HAVE_XAA_H
-	ScrnInfoPtr pScrn = xf86ScreenToScrn(pScreen);
+	ScrnInfoPtr pScrn = xf86Screens[pScreen->myNum];
 	S3Ptr pS3 = S3PTR(pScrn);
 	XAAInfoRecPtr pXAA;
 
@@ -628,7 +626,4 @@ Bool S3AccelInitPIO(ScreenPtr pScreen)
 #endif
 
 	return XAAInit(pScreen, pXAA);
-#else
-	return FALSE;
-#endif
 }
diff --git a/src/s3_dga.c b/src/s3_dga.c
index e384c45..84556b9 100644
--- a/src/s3_dga.c
+++ b/src/s3_dga.c
@@ -41,16 +41,14 @@
 static Bool S3_SetMode(ScrnInfoPtr pScrn, DGAModePtr pMode);
 static int S3_GetViewport(ScrnInfoPtr pScrn);
 static void S3_SetViewport(ScrnInfoPtr pScrn, int x, int y, int flags);
-#ifdef HAVE_XAA_H
 static void S3_FillRect(ScrnInfoPtr pScrn, int x, int y, int w, int h,
 		        unsigned long color);
 static void S3_BltRect(ScrnInfoPtr pScrn, int srcx, int srcy, int w, int h,
 		       int dstx, int dsty);
-static void S3_Sync(ScrnInfoPtr pScrn);
-#endif
 static Bool S3_OpenFramebuffer(ScrnInfoPtr pScrn, char **name,
 			       unsigned char **mem, int *size, int *offset,
 			       int *flags);
+static void S3_Sync(ScrnInfoPtr pScrn);
 
 
 static DGAFunctionRec S3_DGAFuncs = {
@@ -59,13 +57,9 @@ static DGAFunctionRec S3_DGAFuncs = {
 	S3_SetMode,
 	S3_SetViewport,
 	S3_GetViewport,
-#ifdef HAVE_XAA_H
 	S3_Sync,
 	S3_FillRect,
 	S3_BltRect,
-#else
-	NULL, NULL, NULL,
-#endif
 	NULL
 };
 
@@ -111,10 +105,8 @@ SECOND_PASS:
 		currentMode->flags = DGA_CONCURRENT_ACCESS;
 		if (pixmap)
 			currentMode->flags |= DGA_PIXMAP_AVAILABLE;
-#ifdef HAVE_XAA_H
 		if (pS3->pXAA)
 			currentMode->flags |= DGA_FILL_RECT | DGA_BLIT_RECT;
-#endif
 		if (pMode->Flags & V_DBLSCAN)
 			currentMode->flags |= DGA_DOUBLESCAN;
 		if (pMode->Flags & V_INTERLACE)
@@ -275,7 +267,6 @@ static void S3_SetViewport(ScrnInfoPtr pScrn, int x, int y, int flags)
 }
 
 
-#ifdef HAVE_XAA_H
 static void S3_FillRect(ScrnInfoPtr pScrn, int x, int y, int w, int h,
 		        unsigned long color)
 {
@@ -305,7 +296,7 @@ static void S3_BltRect(ScrnInfoPtr pScrn, int srcx, int srcy, int w, int h,
 		SET_SYNC_FLAG(pS3->pXAA);
 	}
 }
-#endif
+
 
 static Bool S3_OpenFramebuffer(ScrnInfoPtr pScrn, char **name,
 			       unsigned char **mem, int *size, int *offset,
@@ -322,9 +313,8 @@ static Bool S3_OpenFramebuffer(ScrnInfoPtr pScrn, char **name,
 	return TRUE;
 }
 
-#ifdef HAVE_XAA_H
+
 static void S3_Sync(ScrnInfoPtr pScrn)
 {
 	WaitIdle();
 }
-#endif
diff --git a/src/s3_driver.c b/src/s3_driver.c
index 305b564..6a3c972 100644
--- a/src/s3_driver.c
+++ b/src/s3_driver.c
@@ -49,6 +49,7 @@
 #include "xf86RAC.h"
 #endif
 #include "compiler.h"
+#include "xaa.h"
 #include "mipointer.h"
 #include "micmap.h"
 #include "fb.h"
@@ -103,7 +104,7 @@ Bool S3CloseScreen(CLOSE_SCREEN_ARGS_DECL);
 Bool S3SaveScreen(ScreenPtr pScreen, int mode);
 static void S3FreeScreen(FREE_SCREEN_ARGS_DECL);
 static void S3GenericLoadPalette(ScrnInfoPtr pScrn, int numColors,
-                                 int *indicies, LOCO *colors,
+                                 int *indices, LOCO *colors,
                                  VisualPtr pVisual);
 static void S3Restore(ScrnInfoPtr pScrn);
 void S3BankZero(ScrnInfoPtr pScrn);
@@ -1112,13 +1113,13 @@ Bool S3SwitchMode(SWITCH_MODE_ARGS_DECL)
 
 
 static void S3GenericLoadPalette(ScrnInfoPtr pScrn, int numColors,
-                                 int *indicies, LOCO *colors,
+                                 int *indices, LOCO *colors,
                                  VisualPtr pVisual)
 {               
         int i, index;
 	
         for (i=0; i<numColors; i++) {
-                index = indicies[i];
+                index = indices[i];
                 outb(0x3c8, index);
                 outb(0x3c9, colors[index].red);
                 outb(0x3c9, colors[index].green);
@@ -1670,7 +1671,7 @@ static Bool S3ModeInit(ScrnInfoPtr pScrn, DisplayModePtr mode)
 	if (pS3->SlowVRAM) {
 		/*
 		 * some Diamond Stealth 64 VRAM cards have a problem with
-		 * VRAM timing, increas -RAS low timing from 3.5 MCLKs
+		 * VRAM timing, increase -RAS low timing from 3.5 MCLKs
 		 * to 4.5 MCLKs
 		 */
 		outb(vgaCRIndex, 0x39);
