--- pan-0.14.2.91/pan/base/msort.c.orig	2005-09-20 22:40:20.167293327 +0200
+++ pan-0.14.2.91/pan/base/msort.c	2005-09-20 22:42:45.465692264 +0200
@@ -65,14 +65,18 @@
 			if ((*cmp) (b1, b2) <= 0)
 			{
 				--n1;
-				*((unsigned long int *) tmp)++ =
-					*((unsigned long int *) b1)++;
+				*((unsigned long int *) tmp) =
+					*((unsigned long int *) b1);
+				tmp += sizeof(unsigned long int);
+				b1 += sizeof(unsigned long int);
 			}
 			else
 			{
 				--n2;
-				*((unsigned long int *) tmp)++ =
-					*((unsigned long int *) b2)++;
+				*((unsigned long int *) tmp) =
+					*((unsigned long int *) b2);
+				tmp += sizeof(unsigned long int);
+				b2 += sizeof(unsigned long int);
 			}
 		}
 	else
