diff -urN mutt-1.4.2.1.orig/configure mutt-1.4.2.1/configure
--- mutt-1.4.2.1.orig/configure	2004-02-12 09:51:57.000000000 +0100
+++ mutt-1.4.2.1/configure	2005-08-15 15:32:09.599760750 +0200
@@ -5089,7 +5089,7 @@
     GSSAPI_LIBS="$MUTTLIBS `$KRB5CFGPATH --libs gssapi`"
     case "`$KRB5CFGPATH --version`" in
       "Kerberos 5 "*)	GSSAPI_IMPL="MIT";;
-      hHeimdal*)		GSSAPI_IMPL="Heimdal";;
+      [hH]eimdal*)		GSSAPI_IMPL="Heimdal";;
       *)		GSSAPI_IMPL="Unknown";;
    esac
   else
