--- ./imap/Makefile.in.orig	2005-02-03 16:30:34.000000000 -0300
+++ ./imap/Makefile.in	2005-02-03 16:32:05.000000000 -0300
@@ -120,7 +120,7 @@
 noinst_LIBRARIES = libimap.a
 noinst_HEADERS = auth.h imap_private.h message.h
 
-libimap_a_SOURCES = auth.c auth_login.c browse.c command.c imap.c imap.h 	message.c utf7.c util.c $(AUTHENTICATORS) $(GSSSOURCES)
+libimap_a_SOURCES = auth.c auth_login.c browse.c command.c imap.c imap.h imap_headercache.c imap_headercache.h message.c utf7.c util.c $(AUTHENTICATORS) $(GSSSOURCES)
 
 mkinstalldirs = $(SHELL) $(top_srcdir)/mkinstalldirs
 CONFIG_HEADER = ../config.h
@@ -134,17 +134,17 @@
 LIBS = @LIBS@
 libimap_a_LIBADD = 
 @USE_SASL_TRUE@@USE_GSS_TRUE@libimap_a_OBJECTS =  auth.o auth_login.o \
-@USE_SASL_TRUE@@USE_GSS_TRUE@browse.o command.o imap.o message.o utf7.o \
+@USE_SASL_TRUE@@USE_GSS_TRUE@browse.o command.o imap.o imap_headercache.o message.o utf7.o \
 @USE_SASL_TRUE@@USE_GSS_TRUE@util.o auth_sasl.o auth_gss.o
 @USE_SASL_FALSE@@USE_GSS_TRUE@libimap_a_OBJECTS =  auth.o auth_login.o \
-@USE_SASL_FALSE@@USE_GSS_TRUE@browse.o command.o imap.o message.o \
+@USE_SASL_FALSE@@USE_GSS_TRUE@browse.o command.o imap.o imap_headercache.o message.o \
 @USE_SASL_FALSE@@USE_GSS_TRUE@utf7.o util.o auth_anon.o auth_cram.o \
 @USE_SASL_FALSE@@USE_GSS_TRUE@auth_gss.o
 @USE_SASL_FALSE@@USE_GSS_FALSE@libimap_a_OBJECTS =  auth.o auth_login.o \
-@USE_SASL_FALSE@@USE_GSS_FALSE@browse.o command.o imap.o message.o \
+@USE_SASL_FALSE@@USE_GSS_FALSE@browse.o command.o imap.o imap_headercache.o message.o \
 @USE_SASL_FALSE@@USE_GSS_FALSE@utf7.o util.o auth_anon.o auth_cram.o
 @USE_SASL_TRUE@@USE_GSS_FALSE@libimap_a_OBJECTS =  auth.o auth_login.o \
-@USE_SASL_TRUE@@USE_GSS_FALSE@browse.o command.o imap.o message.o \
+@USE_SASL_TRUE@@USE_GSS_FALSE@browse.o command.o imap.o imap_headercache.o message.o \
 @USE_SASL_TRUE@@USE_GSS_FALSE@utf7.o util.o auth_sasl.o
 CFLAGS = @CFLAGS@
 COMPILE = $(CC) $(DEFS) $(INCLUDES) $(AM_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS)
