Imported from Gentoo for latest freetype versions.

diff -ur SDL_ttf-2.0.8.orig/SDL_ttf.c SDL_ttf-2.0.8/SDL_ttf.c
--- SDL_ttf-2.0.8.orig/SDL_ttf.c	2006-08-31 13:44:21.000000000 +0200
+++ SDL_ttf-2.0.8/SDL_ttf.c	2006-08-31 13:44:34.000000000 +0200
@@ -43,16 +43,6 @@
 #include FT_FREETYPE_H
 #include FT_OUTLINE_H
 #include FT_TRUETYPE_IDS_H
-/*
-#include <freetype/freetype.h>
-#include <freetype/ftoutln.h>
-#include <freetype/ttnameid.h>
-*/
-#include <freetype/internal/ftobjs.h>
-
-#ifndef FT_OPEN_STREAM
-#define FT_OPEN_STREAM ft_open_stream
-#endif
 
 #include "SDL.h"
 #include "SDL_endian.h"
@@ -278,7 +268,7 @@
 	}
 	memset(stream, 0, sizeof(*stream));
 
-	stream->memory = library->memory;
+	stream->memory = NULL;  /* set by FT_Open_Face */
 	stream->read = RWread;
 	stream->descriptor.pointer = src;
 	stream->pos = (unsigned long)position;
