diff -Naur installwatch-0.7.0beta7PREV1/installwatch.c installwatch-0.7.0beta7/installwatch.c
--- installwatch-0.7.0beta7PREV1/installwatch.c	2014-12-20 08:20:24.000000000 +0800
+++ installwatch-0.7.0beta7/installwatch.c	2014-12-20 08:31:17.000000000 +0800
@@ -95,18 +95,11 @@
 static int (*true_xstat)(int,const char *,struct stat *);
 static int (*true_lxstat)(int,const char *,struct stat *);
 
-#if(GLIBC_MINOR >= 10)
 
 static int (*true_scandir)(	const char *,struct dirent ***,
 				int (*)(const struct dirent *),
 				int (*)(const struct dirent **,const struct dirent **));
 
-#else
-
-static int (*true_scandir)(	const char *,struct dirent ***,
-				int (*)(const struct dirent *),
-				int (*)(const void *,const void *));
-#endif
 
 static int (*true_symlink)(const char *, const char *);
 static int (*true_truncate)(const char *, TRUNCATE_T);
@@ -124,15 +117,9 @@
 static int (*true_open64)(const char *, int, ...);
 static struct dirent64 *(*true_readdir64)(DIR *dir);
 
-#if(GLIBC_MINOR >= 10)
 static int (*true_scandir64)(	const char *,struct dirent64 ***,
 				int (*)(const struct dirent64 *),
 				int (*)(const struct dirent64 **,const struct dirent64 **));
-#else
-static int (*true_scandir64)(	const char *,struct dirent64 ***,
-				int (*)(const struct dirent64 *),
-				int (*)(const void *,const void *));
-#endif
 static int (*true_xstat64)(int,const char *, struct stat64 *);
 static int (*true_lxstat64)(int,const char *, struct stat64 *);
 static int (*true_truncate64)(const char *, __off64_t);
@@ -3071,11 +3058,7 @@
 
 int scandir(	const char *dir,struct dirent ***namelist,
 		int (*select)(const struct dirent *),
-#if (GLIBC_MINOR >= 10)
 		int (*compar)(const struct dirent **,const struct dirent **)	) {
-#else
-		int (*compar)(const void *,const void *)	) {
-#endif
 	int result;
 
 	if (!libc_handle)
@@ -3686,11 +3669,7 @@
 
 int scandir64(	const char *dir,struct dirent64 ***namelist,
 		int (*select)(const struct dirent64 *),
-#if (GLIBC_MINOR >= 10)
 		int (*compar)(const struct dirent64 **,const struct dirent64 **)	) {
-#else
-		int (*compar)(const void *,const void *)	) {
-#endif
 	int result;
 
 	if (!libc_handle)
