diff -Naur spicctrl-1.8-orig/Makefile spicctrl-1.8/Makefile
--- spicctrl-1.8-orig/Makefile	2005-02-10 17:21:47.000000000 +0200
+++ spicctrl-1.8/Makefile	2005-05-28 18:35:46.718907544 +0300
@@ -6,7 +6,7 @@
 LIBS =
 DOCS = AUTHORS CHANGES LICENSE
 VERSION = 1.8
-PREFIX = /usr/local
+PREFIX = /usr
 DESTBIN = $(PREFIX)/bin
 DESTMAN = $(PREFIX)/man/man1
 DESTDOC = $(PREFIX)/share/doc/spicctrl-$(VERSION)
