--- pwlib_v1_9_1.orig/src/ptclib/pssl.cxx	2004-04-09 08:52:17.000000000 +0200
+++ pwlib_v1_9_1/src/ptclib/pssl.cxx	2005-08-08 20:07:17.661786704 +0200
@@ -297,14 +297,14 @@
 
 PSSLPrivateKey::PSSLPrivateKey(const BYTE * keyData, PINDEX keySize)
 {
-  key = d2i_AutoPrivateKey(NULL, (BYTE **)&keyData, keySize);
+  key = d2i_AutoPrivateKey(NULL, (const unsigned char**)&keyData, keySize);
 }
 
 
 PSSLPrivateKey::PSSLPrivateKey(const PBYTEArray & keyData)
 {
   const BYTE * keyPtr = keyData;
-  key = d2i_AutoPrivateKey(NULL, (BYTE **)&keyPtr, keyData.GetSize());
+  key = d2i_AutoPrivateKey(NULL, (const unsigned char**)&keyPtr, keyData.GetSize());
 }
 
 
@@ -472,14 +472,14 @@
 
 PSSLCertificate::PSSLCertificate(const BYTE * certData, PINDEX certSize)
 {
-  certificate = d2i_X509(NULL, (unsigned char **)&certData, certSize);
+  certificate = d2i_X509(NULL, (const unsigned char **)&certData, certSize);
 }
 
 
 PSSLCertificate::PSSLCertificate(const PBYTEArray & certData)
 {
   const BYTE * certPtr = certData;
-  certificate = d2i_X509(NULL, (unsigned char **)&certPtr, certData.GetSize());
+  certificate = d2i_X509(NULL, (const unsigned char **)&certPtr, certData.GetSize());
 }
 
 
@@ -489,7 +489,7 @@
   PBase64::Decode(certStr, certData);
   if (certData.GetSize() > 0) {
     const BYTE * certPtr = certData;
-    certificate = d2i_X509(NULL, (unsigned char **)&certPtr, certData.GetSize());
+    certificate = d2i_X509(NULL, (const unsigned char **)&certPtr, certData.GetSize());
   }
   else
     certificate = NULL;
