===================================================================
--- ./frontends/hocr2pdf.cc	(revision 1391)
+++ ./frontends/hocr2pdf.cc	(revision 1394)
@@ -190,7 +190,7 @@
 	std::string text = htmlDecode(it->text);
 	BBox bbox = it->bbox;
 	
-	// one might imprecicely place text sloppyly in favour of "sometimes"
+	// one might imprecicely place text sloppily in favour of "sometimes"
 	// improved cut'n paste-able text in not so advanced PDF Viewers
 	if (sloppy) {
 	  span_iterator it2 = it;
@@ -371,14 +371,14 @@
   arglist.Add(&arg_no_image);
 
   Argument<bool> arg_sloppy_text("s", "sloppy-text",
-				 "sloppyly place text, combine words, do not draw single glyphs",
+				 "sloppily place text, group words, do not draw single glyphs",
 				 0, 0, true, true);
   arglist.Add(&arg_sloppy_text);
   
   // parse the specified argument list - and maybe output the Usage
   if (!arglist.Read(argc, argv) || arg_help.Get() == true)
     {
-      std::cerr << "hOCR to search-able PDF convertert" << std::endl
+      std::cerr << "hOCR to search-able PDF converter" << std::endl
                 <<  "    - Copyright 2008 by René Rebe, ExactCODE" << std::endl
                 << "Usage:" << std::endl;
       
