--- udev-167/udev/udevd.c.vanilla	2013-04-16 12:45:10.946003262 +0200
+++ udev-167/udev/udevd.c	2013-04-16 12:45:43.934994140 +0200
@@ -43,6 +43,8 @@
 #include <sys/ioctl.h>
 #include <sys/inotify.h>
 #include <sys/utsname.h>
+#include <sys/resource.h>
+#include <sys/resource.h>
 
 #include "udev.h"
 #include "sd-daemon.h"
--- udev-167/libudev/libudev-util-private.c.vanilla	2013-04-16 12:44:40.593989482 +0200
+++ udev-167/libudev/libudev-util-private.c	2013-04-16 12:44:57.648987445 +0200
@@ -21,6 +21,7 @@
 #include <grp.h>
 #include <sys/wait.h>
 #include <sys/param.h>
+#include <sys/resource.h>
 #include <limits.h>
 
 #include "libudev.h"
