--- udev-167/udev/udevd.c.vanilla	2013-12-12 12:55:05.393138957 +0100
+++ udev-167/udev/udevd.c	2013-12-12 12:55:28.047138173 +0100
@@ -43,6 +43,7 @@
 #include <sys/ioctl.h>
 #include <sys/inotify.h>
 #include <sys/utsname.h>
+#include <sys/resource.h>
 
 #include "udev.h"
 #include "sd-daemon.h"
--- udev-167/libudev/libudev-util-private.c.vanilla	2013-12-12 12:54:18.159140589 +0100
+++ udev-167/libudev/libudev-util-private.c	2013-12-12 12:54:31.588140125 +0100
@@ -21,6 +21,7 @@
 #include <grp.h>
 #include <sys/wait.h>
 #include <sys/param.h>
+#include <sys/resource.h>
 
 #include "libudev.h"
 #include "libudev-private.h"
