diff -ur udev-083/extras/ata_id/Makefile udev-083-fixed/extras/ata_id/Makefile
--- udev-083/extras/ata_id/Makefile	2006-01-27 01:41:07.000000000 +0100
+++ udev-083-fixed/extras/ata_id/Makefile	2006-05-17 11:22:08.000000000 +0200
@@ -48,7 +48,7 @@
 .PHONY: clean
 
 install-bin: all
-	$(INSTALL_PROGRAM) $(PROG) $(DESTDIR)$(sbindir)/$(PROG)
+	$(INSTALL_PROGRAM) $(PROG) $(DESTDIR)$(libudevdir)/$(PROG)
 .PHONY: install-bin
 
 uninstall-bin:
diff -ur udev-083/extras/cdrom_id/Makefile udev-083-fixed/extras/cdrom_id/Makefile
--- udev-083/extras/cdrom_id/Makefile	2006-01-27 01:41:07.000000000 +0100
+++ udev-083-fixed/extras/cdrom_id/Makefile	2006-05-17 11:21:47.000000000 +0200
@@ -48,7 +48,7 @@
 .PHONY: clean
 
 install-bin: all
-	$(INSTALL_PROGRAM) $(PROG) $(DESTDIR)$(sbindir)/$(PROG)
+	$(INSTALL_PROGRAM) $(PROG) $(DESTDIR)$(libudevdir)/$(PROG)
 .PHONY: install-bin
 
 uninstall-bin:
diff -ur udev-083/extras/dasd_id/Makefile udev-083-fixed/extras/dasd_id/Makefile
--- udev-083/extras/dasd_id/Makefile	2006-01-27 01:41:07.000000000 +0100
+++ udev-083-fixed/extras/dasd_id/Makefile	2006-05-17 11:22:22.000000000 +0200
@@ -48,7 +48,7 @@
 .PHONY: clean
 
 install-bin: all
-	$(INSTALL_PROGRAM) $(PROG) $(DESTDIR)$(sbindir)/$(PROG)
+	$(INSTALL_PROGRAM) $(PROG) $(DESTDIR)$(libudevdir)/$(PROG)
 .PHONY: install-bin
 
 uninstall-bin:
diff -ur udev-083/extras/edd_id/Makefile udev-083-fixed/extras/edd_id/Makefile
--- udev-083/extras/edd_id/Makefile	2006-01-27 01:41:07.000000000 +0100
+++ udev-083-fixed/extras/edd_id/Makefile	2006-05-17 11:22:31.000000000 +0200
@@ -48,7 +48,7 @@
 .PHONY: clean
 
 install-bin: all
-	$(INSTALL_PROGRAM) $(PROG) $(DESTDIR)$(sbindir)/$(PROG)
+	$(INSTALL_PROGRAM) $(PROG) $(DESTDIR)$(libudevdir)/$(PROG)
 .PHONY: install-bin
 
 uninstall-bin:
diff -ur udev-083/extras/scsi_id/Makefile udev-083-fixed/extras/scsi_id/Makefile
--- udev-083/extras/scsi_id/Makefile	2006-01-27 01:41:07.000000000 +0100
+++ udev-083-fixed/extras/scsi_id/Makefile	2006-05-17 11:22:52.000000000 +0200
@@ -66,7 +66,7 @@
 .PHONY: clean
 
 install-bin: all
-	$(INSTALL_PROGRAM) $(PROG) $(DESTDIR)$(sbindir)/$(PROG)
+	$(INSTALL_PROGRAM) $(PROG) $(DESTDIR)$(libudevdir)/$(PROG)
 .PHONY: install-bin
 
 uninstall-bin:
diff -ur udev-083/extras/usb_id/Makefile udev-083-fixed/extras/usb_id/Makefile
--- udev-083/extras/usb_id/Makefile	2006-01-27 01:41:07.000000000 +0100
+++ udev-083-fixed/extras/usb_id/Makefile	2006-05-17 11:22:59.000000000 +0200
@@ -48,7 +48,7 @@
 .PHONY: clean
 
 install-bin: all
-	$(INSTALL_PROGRAM) $(PROG) $(DESTDIR)$(sbindir)/$(PROG)
+	$(INSTALL_PROGRAM) $(PROG) $(DESTDIR)$(libudevdir)/$(PROG)
 .PHONY: install-bin
 
 uninstall-bin:
diff -ur udev-083/extras/volume_id/Makefile udev-083-fixed/extras/volume_id/Makefile
--- udev-083/extras/volume_id/Makefile	2006-01-27 01:41:07.000000000 +0100
+++ udev-083-fixed/extras/volume_id/Makefile	2006-05-17 11:23:07.000000000 +0200
@@ -51,7 +51,7 @@
 .PHONY: clean
 
 install-bin: all
-	$(INSTALL_PROGRAM) $(PROG) $(DESTDIR)$(sbindir)/$(PROG)
+	$(INSTALL_PROGRAM) $(PROG) $(DESTDIR)$(libudevdir)/$(PROG)
 .PHONY: install-bin
 
 uninstall-bin:
