--- mol/src/main/misc.h.vanilla	2023-08-31 15:05:56.344264179 +0100
+++ mol/src/main/misc.h	2023-08-31 15:06:03.360246709 +0100
@@ -20,7 +20,7 @@
 /* arch/misc.c */
 #ifdef __linux__
 extern void	signal_handler( int sig_num, siginfo_t *sinfo,
-				struct ucontext *puc, ulong rt_sf );
+				struct ucontext_t *puc, ulong rt_sf );
 #else
 extern void	signal_handler( int sig_num );
 #endif
--- mol/src/main/linux/misc.c.vanilla	2023-08-31 15:05:30.556328393 +0100
+++ mol/src/main/linux/misc.c	2023-08-31 15:05:52.008274976 +0100
@@ -68,7 +68,7 @@
 #endif
 
 void 
-signal_handler( int sig_num, siginfo_t *sinfo, struct ucontext *puc, ulong rt_sf )
+signal_handler( int sig_num, siginfo_t *sinfo, struct ucontext_t *puc, ulong rt_sf )
 {
 	/* handles SIGINT, SIGPIPE */
 	if( common_signal_handler(sig_num) )
