--- mol/src/main/main.c.vanilla	2023-08-31 15:06:49.992130592 +0100
+++ mol/src/main/main.c	2023-08-31 15:10:03.319649193 +0100
@@ -112,7 +112,7 @@
 	
 	if( sig_num != SIGINT && sig_num != SIGTRAP ) {
 		aprint("***** SIGNAL %d [%s] in thread %s *****\n", 
-		       sig_num, (sig_num < NSIG)? sys_siglist[sig_num]:"?",
+		       sig_num, (sig_num < NSIG)? strsignal(sig_num):"?",
 		       get_thread_name() );
 	}
 
