--- mol/src/kmod/Linux/Makefile.26
+++ mol/src/kmod/Linux/Makefile.26
@@ -9,7 +9,7 @@
 obj-m		:= _kuname.o mol.o 
 
 PERFOBJS	= $(addprefix $(obj)/, $(PERFOBJS_))
-MOL_ASMFLAGS	= $(ASMFLAGS) $(INCLUDES) -D__ASSEMBLY__
+MOL_ASMFLAGS	= $(CPPFLAGS) $(ASMFLAGS) $(INCLUDES) -D__ASSEMBLY__
 
 
 $(obj)/hook.o: 		$(src)/reloc_table.h
@@ -37,7 +37,7 @@
 $(src)/asm_offsets.h:	$(src)/archinclude.h $(src)/kernel_vars.h $(src)/mac_registers.h
 $(src)/asm_offsets.h:	$(src)/asm_offsets.c $(src)/asm_offsets.inc
 	@$(RM) $(src)/tmp-offsets.c $@ ; cat $^ > $(src)/tmp-offsets.c
-	@$(CC) $(CFLAGS) -I$(src) -Wall -S $(src)/tmp-offsets.c
+	@$(CC) $(CPPFLAGS) $(CFLAGS) -I$(src) -Wall -S $(src)/tmp-offsets.c
 	@echo "/* WARNING! Automatically generated from 'shared/asm_offsets.c' - DO NOT EDIT! */" > $@
 	@grep '^#' tmp-offsets.s >> $@
 	@$(RM) $(src)/tmp-offsets.*
