--- mol-2003-06-20/src/startmol.in	2003-06-12 13:00:59.000000000 +0200
+++ mol-2003-06-20-devfs/src/startmol.in	2003-06-21 17:58:58.000000000 +0200
@@ -256,7 +256,7 @@
     [ "$LINUX" -a ! -f `getres -F console_video_prefs` ] && {
 	echo
 	echo "*************************************************************"
-	echo " No video modes has been configured. Please run 'molvconfig'"
+	echo " No video modes have been configured. Please run 'molvconfig'"
 	echo " as root to configure full screen video or disable console"
 	echo " video in /etc/mol/molrc.video."
 	echo "*************************************************************"
