--- mol-2003-06-20/src/vconfig/main.c	2003-06-12 13:00:59.000000000 +0200
+++ mol-2003-06-20-devfs/src/vconfig/main.c	2003-06-21 13:37:54.000000000 +0200
@@ -99,10 +99,12 @@
 	struct vt_stat 	vts;
 	int 		fd;
 	char 		vtname[32];
+	char		vtname2[32];
 	struct termios 	tio;	
 
-	if ((fd = open("/dev/tty0", O_WRONLY)) < 0) {
-		fprintf(stderr, "failed to open /dev/tty0: %s\n", strerror(errno));
+	if ( (fd = open("/dev/tty0", O_WRONLY)) < 0 &&
+	     (fd = open("/dev/vc/0", O_WRONLY)) < 0 ) {
+		fprintf(stderr, "failed to open /dev/tts0 or /dev/vc/0: %s\n", strerror(errno));
 		return 1;
 	}
 
@@ -128,10 +130,12 @@
 	/* open our vt */
 	sprintf(vtname,"/dev/tty%d",vt_no);
 	if ((console_fd = open(vtname, O_RDWR )) < 0) {
-		fprintf(stderr, "Cannot open %s (%s)\n",vtname, strerror(errno));
-		return 1;
+		sprintf(vtname2,"/dev/vc/%d",vt_no);
+		if ((console_fd = open(vtname2, O_RDWR )) < 0) {
+			fprintf(stderr, "Cannot open %s or %s (%s)\n",vtname, vtname, strerror(errno));
+			return 1;
+		}
 	}
-
 	printf("Running MOL video configurator on VT%d\n", vt_no );
 #if 0
 	/* set controlling tty */
