--- ./Makefile.in.orig	2004-10-23 17:40:37.788483608 +0000
+++ ./Makefile.in	2004-10-23 17:40:48.944787592 +0000
@@ -215,7 +215,7 @@
 sharedstatedir = @sharedstatedir@
 sysconfdir = @sysconfdir@
 target_alias = @target_alias@
-SUBDIRS = libexif libjpeg test po m4
+SUBDIRS = libexif libjpeg test m4
 EXTRA_DIST = @PACKAGE@.spec
 
 ########################################################################
