--- libexif-0.6.13/doc/Makefile.in.orig	2006-01-04 19:50:04.000000000 +0200
+++ libexif-0.6.13/doc/Makefile.in	2006-01-04 19:50:10.000000000 +0200
@@ -444,9 +444,11 @@
 
 all-local: $(DOXYGEN_STAMPS)
 
-install-data-local: install-apidocs install-apidocs-internals
+@HAVE_DOXYGEN_TRUE@install-data-local: install-apidocs install-apidocs-internals
+@HAVE_DOXYGEN_FALSE@install-data-local:
 
-uninstall-local: uninstall-apidocs uninstall-apidocs-internals
+@HAVE_DOXYGEN_TRUE@uninstall-local: uninstall-apidocs uninstall-apidocs-internals
+@HAVE_DOXYGEN_FALSE@uninstall-local:
 
 clean-local: clean-apidocs
 
