--- icecream/client/Makefile.in	2004-08-29 13:27:41.000000000 +0200
+++ icecream-fixed/client/Makefile.in	2004-10-16 14:26:16.627709839 +0200
@@ -671,11 +671,6 @@
 	$(LN_S) icecc icecc++
 
 install-exec-local:
-	$(mkinstalldirs) $(DESTDIR)$(bindir)
-	for link in g++ gcc c++ cc; do \
-	  rm -f $(DESTDIR)$(bindir)/$$link ;\
-	  $(LN_S) icecc $(DESTDIR)$(bindir)/$$link ;\
-	done
 
 uninstall-local:
 	for link in g++ gcc c++ cc; do \
