diff -r -u -N libsoup-2.2.2.orig/libsoup/soup-gnutls.c libsoup-2.2.2/libsoup/soup-gnutls.c
--- libsoup-2.2.2.orig/libsoup/soup-gnutls.c	2005-03-28 02:58:00.256716192 +0200
+++ libsoup-2.2.2/libsoup/soup-gnutls.c	2005-03-28 02:59:15.378295976 +0200
@@ -63,7 +63,6 @@
 	}
 
 	if (status & GNUTLS_CERT_INVALID ||
-	    status & GNUTLS_CERT_NOT_TRUSTED ||
 	    status & GNUTLS_CERT_REVOKED)
 	{
 		g_set_error (err, SOUP_SSL_ERROR,
