diff -ruN snownews-1.4.4/Makefile.inc snownews-1.4.4-new/Makefile.inc
--- snownews-1.4.4/Makefile.inc	2004-01-23 13:13:29.000000000 +0100
+++ snownews-1.4.4-new/Makefile.inc	2004-02-05 19:05:25.594030184 +0100
@@ -1,7 +1,7 @@
 CC = gcc
 MAKE = make
 INSTALL = install
-PREFIX = /usr/local
+PREFIX = /usr
 LOCALEPATH = $(PREFIX)/share/locale
 DISTFILES = AUTHOR COPYING CREDITS Changelog README README.de README.colors INSTALL INSTALL.solaris INSTALL.FreeOpenNetBSD TODO opml2snow \
 	Makefile Makefile.inc Makefile.solaris \
