diff -Nur syslog-ng-1.6.8-orig/configure syslog-ng-1.6.8/configure
--- syslog-ng-1.6.8-orig/configure	2005-05-26 10:34:01.000000000 +0200
+++ syslog-ng-1.6.8/configure	2005-05-31 18:08:00.000000000 +0200
@@ -7410,7 +7410,7 @@
 
 echo "$as_me:$LINENO: checking how to enable static linking for certain libraries" >&5
 echo $ECHO_N "checking how to enable static linking for certain libraries... $ECHO_C" >&6
-ldversion=`ld -V 2>&1 | head -1 `
+ldversion=`ld -V 2>&1 | head -n 1 `
 if echo $ldversion | egrep "GNU|Solaris" > /dev/null; then
 	LD_START_STATIC="-Wl,-Bstatic"
 	LD_END_STATIC="-Wl,-Bdynamic"
diff -Nur syslog-ng-1.6.8-orig/configure.in syslog-ng-1.6.8/configure.in
--- syslog-ng-1.6.8-orig/configure.in	2005-05-26 10:33:54.000000000 +0200
+++ syslog-ng-1.6.8/configure.in	2005-05-31 18:08:13.000000000 +0200
@@ -296,7 +296,7 @@
 fi
 
 AC_MSG_CHECKING(how to enable static linking for certain libraries)
-ldversion=`ld -V 2>&1 | head -1 `
+ldversion=`ld -V 2>&1 | head -n 1 `
 if echo $ldversion | egrep "GNU|Solaris" > /dev/null; then
 	LD_START_STATIC="-Wl,-Bstatic"
 	LD_END_STATIC="-Wl,-Bdynamic"
