--- runit-shutdown-0.2/src/halt.sh.orig	2006-04-09 01:53:16.000000000 +0200
+++ runit-shutdown-0.2/src/halt.sh	2006-04-09 01:55:07.000000000 +0200
@@ -27,7 +27,11 @@
 done
 
 if [ "$force" ]; then
-	echo "forced $PROG not implemented, ignored."
+	if [ "$PROG" == 'reboot' ]; then
+		exec /command/forced_reboot
+	else
+		exec /command/forced_poweroff
+	fi
 elif [ "$wtmp" ]; then
 	echo "wtmp record-only for $PROG not implemented, skiped."
 	exit 0
