--- runit-shutdown-0.2/src/killall5.sh.orig	2006-04-08 23:48:04.000000000 +0200
+++ runit-shutdown-0.2/src/killall5.sh	2006-04-09 00:13:51.000000000 +0200
@@ -35,13 +35,13 @@
 	fi
 	
  	# pid (comm) state ppid pgrp session tty_nr tpgid flags ...
-	eval `sed -e 's,^\([0-9]*\) ([^)]*) . \([0-9]*\) \([0-9]*\) \([0-9]*\) .*,my_pid=\1 my_ppid=\2 my_pgrp=\3 my_session=\4,g' \
+	eval `sed -n -e 's,^\([0-9]*\) ([^)]*) . \([0-9]*\) \([0-9]*\) \([0-9]*\) .*,my_pid=\1 my_ppid=\2 my_pgrp=\3 my_session=\4,p' \
 		/proc/$$/stat | tr ' ' '\n'`
 
 	# who am i?
 	#echo -e "$my_pid\t$my_ppid\t$my_pgrp\t$my_session\tMYSELF"
 
-	sed -e 's,^\([0-9]*\) ([^)]*) . \([0-9]*\) \([0-9]*\) \([0-9]*\) .*,\1 \2 \3 \4,g' \
+	sed -n -e 's,^\([0-9]*\) ([^)]*) . \([0-9]*\) \([0-9]*\) \([0-9]*\) .*,\1 \2 \3 \4,p' \
 		`ls -d1 /proc/*/stat` 2> /dev/null | while read proc_pid proc_ppid proc_pgrp proc_session; do
 			case "$proc_session" in
 				0|1|$my_session)	action=''	;;
