--- pixieplus-0.5.4/mimetypes/Makefile.in	2003-02-09 05:58:13.000000000 +0100
+++ pixieplus-0.5.4-fixed/mimetypes/Makefile.in	2003-08-22 22:52:05.000000000 +0200
@@ -240,7 +240,7 @@
 
 datadir = $(kde_mimedir)/image
 
-data_DATA = x-pcx.desktop x-xwd.desktop x-miff.desktop x-tga.desktop \
+data_DATA = x-xwd.desktop x-miff.desktop x-tga.desktop \
 	x-pict.desktop
 
 
@@ -302,7 +302,7 @@
 
 #>- DISTFILES = $(DIST_COMMON) $(DIST_SOURCES) $(TEXINFOS) $(EXTRA_DIST)
 #>+ 4
-KDE_DIST=x-pict.desktop x-xwd.desktop x-miff.desktop x-pcx.desktop x-tga.desktop 
+KDE_DIST=x-pict.desktop x-xwd.desktop x-miff.desktop x-tga.desktop 
 
 DISTFILES= $(DIST_COMMON) $(DIST_SOURCES) $(TEXINFOS) $(EXTRA_DIST) $(KDE_DIST)
 
