--- ./Makefile.in.orig	2001-03-17 06:08:33.000000000 +0200
+++ ./Makefile.in	2003-08-19 21:03:58.000000000 +0300
@@ -55,7 +55,7 @@
 install_shared:	shared install_common
 		install -m 755 liblockfile.so \
 			$(ROOT)$(libdir)/liblockfile.so.$(VER)
-		ln -s liblockfile.so.$(VER) $(ROOT)$(libdir)/liblockfile.so
+		ln -sf liblockfile.so.$(VER) $(ROOT)$(libdir)/liblockfile.so
 		if test "$(ROOT)" = ""; then @LDCONFIG@; fi
 
 install_common:diff -Nru3 trunk-1418/package/base/gdb/gdb.conf trunk/package/base/gdb/gdb.conf
