--- INSTALL.orig	2005-03-12 02:50:00.000000000 +0100
+++ INSTALL	2005-04-22 09:46:41.000000000 +0200
@@ -205,7 +205,7 @@
 {
   var=$1
   
-  defdir="/usr/local/Adobe/Acrobat7.0"
+  defdir="/opt/acroread"
   case "$OSname" in
     SunOS)
       case "$OSrelease" in
@@ -364,7 +364,7 @@
   CustomPlatformTar="$4"
   
   
-  OutputLicense "$ReadLicense"
+#  OutputLicense "$ReadLicense"
   ReadLicense=""
 
   PrintRequiredFree "$ReadTar" "$SearchTar" "$CustomTar" \
@@ -374,6 +374,7 @@
   if [ -z "$ReadInstallDir" ] ; then
     GetInstallDirectory ReadInstallDir
   fi
+  mkdir -p $ReadInstallDir
 
   echo $ReadInstallDir
 
@@ -410,7 +411,7 @@
 
 AcroSerNum=""
 AcroCompName=""
-ReadInstallDir=""
+ReadInstallDir="/opt/acroread"
 case "${ScriptDirectory}" in
   /*) ;;
   .) ScriptDirectory="$CurrentDirectory" ;;
