diff --git a/eeprom/decode-dimms b/eeprom/decode-dimms
index 787b6f5..87af7e8 100755
--- a/eeprom/decode-dimms
+++ b/eeprom/decode-dimms
@@ -27,7 +27,7 @@
 # the legacy eeprom driver (in the kernel tree since v2.6.0). For kernels
 # older than 2.6.0, the eeprom driver can be found in the lm-sensors 2
 # package. For DDR4, you need the ee1004 driver (in the kernel tree since
-# kernel v4.20).
+# kernel v4.20). For DDR5 you need the spd5118 driver.
 #
 # References:
 # PC SDRAM Serial Presence
@@ -2622,7 +2622,8 @@ sub get_dimm_list
 	if ($use_sysfs) {
 		@drivers = ('eeprom',
 			    'at24',
-			    'ee1004');	# DDR4
+			    'ee1004', # DDR4
+			    'spd5118');	# DDR5
 	} else {
 		@drivers = ('eeprom');
 		$dir = '/proc/sys/dev/sensors';
@@ -2647,7 +2648,8 @@ sub get_dimm_list
 				next unless defined $attr &&
 					    ($attr eq "eeprom" ||
 					     $attr eq "spd" ||
-					     $attr eq "ee1004");	# DDR4
+					     $attr eq "ee1004" ||	# DDR4
+					     $attr eq "spd5118");	# DDR5
 			} else {
 				next unless $file =~ /^eeprom-/;
 			}
@@ -2659,7 +2661,7 @@ sub get_dimm_list
 	}
 
 	if (!$opened) {
-		print STDERR "No EEPROM found, try loading the eeprom, at24 or ee1004 module\n";
+		print STDERR "No EEPROM found, try loading the eeprom, at24, ee1004 or spd5118 module\n";
 		exit;
 	}
 
