--- gkrellm-2.3.4/Makefile.vanilla	2025-03-25 20:26:36.050399296 +0100
+++ gkrellm-2.3.4/Makefile	2025-03-25 20:23:48.822727906 +0100
@@ -247,7 +247,7 @@
 	(cd po && ${MAKE} all)
 	(cd src && ${MAKE} \
 		EXTRAOBJS= SYS_LIBS="-lkvm" \
-		SMC_LIBS="-L/usr/X11R6/lib -lSM -lICE -Wl,-R/usr/X11R6/lib" \
+		SMC_LIBS="`pkg-config --libs sm`" \
 		gkrellm )
 	(cd server && ${MAKE} \
 		EXTRAOBJS= SYS_LIBS="-lkvm" gkrellmd )
@@ -256,7 +256,7 @@
 	(cd po && ${MAKE} all)
 	(cd src && ${MAKE} \
 		EXTRAOBJS= SYS_LIBS="-lkvm -pthread" \
-		SMC_LIBS="-L/usr/X11R6/lib -lSM -lICE -R/usr/X11R6/lib" \
+		SMC_LIBS="`pkg-config --libs sm`" \
 		gkrellm )
 	(cd server && ${MAKE} \
 		EXTRAOBJS= SYS_LIBS="-lkvm -pthread" gkrellmd )
--- gkrellm-2.3.4/src/Makefile.vanilla	2025-03-25 20:26:45.033138701 +0100
+++ gkrellm-2.3.4/src/Makefile	2025-03-25 20:29:31.206674094 +0100
@@ -20,7 +20,7 @@
 MANDIRMODE ?= 755
 INSTALL ?= install
 LINK_FLAGS ?= -Wl,-E
-SMC_LIBS ?= -L/usr/X11R6/lib -lSM -lICE
+SMC_LIBS ?= $(shell pkg-config --libs sm ice)
 
 X11_LIBS ?= -lX11
 
@@ -85,13 +85,13 @@
 GKRELLM_INCLUDES = gkrellm.h gkrellm-public-proto.h $(SHARED_PATH)/log.h
 
 PKG_INCLUDE = `$(PKG_CONFIG) --cflags gtk+-2.0 gthread-2.0`
-PKG_LIB = `$(PKG_CONFIG) --libs gtk+-2.0 gthread-2.0`
+PKG_LIB = `$(PKG_CONFIG) --libs gtk+-2.0 gthread-2.0 gmodule-2.0`
 
 FLAGS = -O2 -I.. -I$(SHARED_PATH) $(PKG_INCLUDE) $(GTOP_INCLUDE) $(PTHREAD_INC) \
  ${NTLM_INCLUDES} -DGKRELLM_CLIENT
 
 LIBS = $(PKG_LIB) $(GTOP_LIBS) $(SMC_LIBS) $(SYS_LIBS) $(SSL_LIBS) $(SENSORS_LIBS) \
- $(NTLM_LIBS) $(X11_LIBS)
+ $(NTLM_LIBS) $(X11_LIBS) -lm
 
 ifeq ($(debug),1)
     FLAGS += -g
@@ -188,11 +188,11 @@
 
 netbsd1:
 	$(MAKE) EXTRAOBJS= SYS_LIBS="-lkvm" \
-		SMC_LIBS="-L/usr/X11R6/lib -lSM -lICE -Wl,-R/usr/X11R6/lib" \
+		SMC_LIBS="`pkg-config --libs sm`" \
 		gkrellm
 netbsd2:
 	$(MAKE) EXTRAOBJS= SYS_LIBS="-lkvm -pthread" \
-		SMC_LIBS="-L/usr/X11R6/lib -lSM -lICE -R/usr/X11R6/lib" \
+		SMC_LIBS="`pkg-config --libs sm`" \
 		gkrellm
 
 openbsd:
