diff -urN -x '*~' lesstif-0.93.36-orig/doc/lessdox/clients/Makefile.in lesstif-0.93.36/doc/lessdox/clients/Makefile.in
--- lesstif-0.93.36-orig/doc/lessdox/clients/Makefile.in	2002-08-06 11:52:13.000000000 +0200
+++ lesstif-0.93.36/doc/lessdox/clients/Makefile.in	2003-04-30 15:08:51.000000000 +0200
@@ -144,6 +144,7 @@
 manonedir = $(exec_prefix)/man/man1
 manthreedir = $(exec_prefix)/man/man3
 manfivedir = $(exec_prefix)/man/man5
+manext = x
 
 manone_DATA = mwm.1 xmbind.1 lesstif.1 uil.1 ltversion.1
 manthree_DATA = 
@@ -221,16 +222,16 @@
 	@list='$(manfive_DATA)'; for p in $$list; do \
 	  if test -f "$$p"; then d=; else d="$(srcdir)/"; fi; \
 	  f="`echo $$p | sed -e 's|^.*/||'`"; \
-	  echo " $(manfiveDATA_INSTALL) $$d$$p $(DESTDIR)$(manfivedir)/$$f"; \
-	  $(manfiveDATA_INSTALL) $$d$$p $(DESTDIR)$(manfivedir)/$$f; \
+	  echo " $(manfiveDATA_INSTALL) $$d$$p $(DESTDIR)$(manfivedir)/$${f}$(manext)"; \
+	  $(manfiveDATA_INSTALL) $$d$$p $(DESTDIR)$(manfivedir)/$${f}$(manext); \
 	done
 
 uninstall-manfiveDATA:
 	@$(NORMAL_UNINSTALL)
 	@list='$(manfive_DATA)'; for p in $$list; do \
 	  f="`echo $$p | sed -e 's|^.*/||'`"; \
-	  echo " rm -f $(DESTDIR)$(manfivedir)/$$f"; \
-	  rm -f $(DESTDIR)$(manfivedir)/$$f; \
+	  echo " rm -f $(DESTDIR)$(manfivedir)/$${f}$(manext)"; \
+	  rm -f $(DESTDIR)$(manfivedir)/$${f}$(manext); \
 	done
 manoneDATA_INSTALL = $(INSTALL_DATA)
 install-manoneDATA: $(manone_DATA)
@@ -239,16 +240,16 @@
 	@list='$(manone_DATA)'; for p in $$list; do \
 	  if test -f "$$p"; then d=; else d="$(srcdir)/"; fi; \
 	  f="`echo $$p | sed -e 's|^.*/||'`"; \
-	  echo " $(manoneDATA_INSTALL) $$d$$p $(DESTDIR)$(manonedir)/$$f"; \
-	  $(manoneDATA_INSTALL) $$d$$p $(DESTDIR)$(manonedir)/$$f; \
+	  echo " $(manoneDATA_INSTALL) $$d$$p $(DESTDIR)$(manonedir)/$${f}$(manext)"; \
+	  $(manoneDATA_INSTALL) $$d$$p $(DESTDIR)$(manonedir)/$${f}$(manext); \
 	done
 
 uninstall-manoneDATA:
 	@$(NORMAL_UNINSTALL)
 	@list='$(manone_DATA)'; for p in $$list; do \
 	  f="`echo $$p | sed -e 's|^.*/||'`"; \
-	  echo " rm -f $(DESTDIR)$(manonedir)/$$f"; \
-	  rm -f $(DESTDIR)$(manonedir)/$$f; \
+	  echo " rm -f $(DESTDIR)$(manonedir)/$${f}$(manext)"; \
+	  rm -f $(DESTDIR)$(manonedir)/$${f}$(manext); \
 	done
 manthreeDATA_INSTALL = $(INSTALL_DATA)
 install-manthreeDATA: $(manthree_DATA)
@@ -257,16 +258,16 @@
 	@list='$(manthree_DATA)'; for p in $$list; do \
 	  if test -f "$$p"; then d=; else d="$(srcdir)/"; fi; \
 	  f="`echo $$p | sed -e 's|^.*/||'`"; \
-	  echo " $(manthreeDATA_INSTALL) $$d$$p $(DESTDIR)$(manthreedir)/$$f"; \
-	  $(manthreeDATA_INSTALL) $$d$$p $(DESTDIR)$(manthreedir)/$$f; \
+	  echo " $(manthreeDATA_INSTALL) $$d$$p $(DESTDIR)$(manthreedir)/$${f}$(manext)"; \
+	  $(manthreeDATA_INSTALL) $$d$$p $(DESTDIR)$(manthreedir)/$${f}$(manext); \
 	done
 
 uninstall-manthreeDATA:
 	@$(NORMAL_UNINSTALL)
 	@list='$(manthree_DATA)'; for p in $$list; do \
 	  f="`echo $$p | sed -e 's|^.*/||'`"; \
-	  echo " rm -f $(DESTDIR)$(manthreedir)/$$f"; \
-	  rm -f $(DESTDIR)$(manthreedir)/$$f; \
+	  echo " rm -f $(DESTDIR)$(manthreedir)/$${f}$(manext)"; \
+	  rm -f $(DESTDIR)$(manthreedir)/$${f}$(manext); \
 	done
 tags: TAGS
 TAGS:
diff -urN -x '*~' lesstif-0.93.36-orig/doc/lessdox/functions/Makefile.in lesstif-0.93.36/doc/lessdox/functions/Makefile.in
--- lesstif-0.93.36-orig/doc/lessdox/functions/Makefile.in	2002-08-06 11:52:13.000000000 +0200
+++ lesstif-0.93.36/doc/lessdox/functions/Makefile.in	2003-04-30 15:08:34.000000000 +0200
@@ -144,6 +144,7 @@
 manonedir = $(exec_prefix)/man/man1
 manthreedir = $(exec_prefix)/man/man3
 manfivedir = $(exec_prefix)/man/man5
+manext = x
 
 manone_DATA = 
 manthree_DATA = 
@@ -221,16 +222,16 @@
 	@list='$(manfive_DATA)'; for p in $$list; do \
 	  if test -f "$$p"; then d=; else d="$(srcdir)/"; fi; \
 	  f="`echo $$p | sed -e 's|^.*/||'`"; \
-	  echo " $(manfiveDATA_INSTALL) $$d$$p $(DESTDIR)$(manfivedir)/$$f"; \
-	  $(manfiveDATA_INSTALL) $$d$$p $(DESTDIR)$(manfivedir)/$$f; \
+	  echo " $(manfiveDATA_INSTALL) $$d$$p $(DESTDIR)$(manfivedir)/$${f}$(manext)"; \
+	  $(manfiveDATA_INSTALL) $$d$$p $(DESTDIR)$(manfivedir)/$${f}$(manext); \
 	done
 
 uninstall-manfiveDATA:
 	@$(NORMAL_UNINSTALL)
 	@list='$(manfive_DATA)'; for p in $$list; do \
 	  f="`echo $$p | sed -e 's|^.*/||'`"; \
-	  echo " rm -f $(DESTDIR)$(manfivedir)/$$f"; \
-	  rm -f $(DESTDIR)$(manfivedir)/$$f; \
+	  echo " rm -f $(DESTDIR)$(manfivedir)/$${f}$(manext)"; \
+	  rm -f $(DESTDIR)$(manfivedir)/$${f}$(manext); \
 	done
 manoneDATA_INSTALL = $(INSTALL_DATA)
 install-manoneDATA: $(manone_DATA)
@@ -239,16 +240,16 @@
 	@list='$(manone_DATA)'; for p in $$list; do \
 	  if test -f "$$p"; then d=; else d="$(srcdir)/"; fi; \
 	  f="`echo $$p | sed -e 's|^.*/||'`"; \
-	  echo " $(manoneDATA_INSTALL) $$d$$p $(DESTDIR)$(manonedir)/$$f"; \
-	  $(manoneDATA_INSTALL) $$d$$p $(DESTDIR)$(manonedir)/$$f; \
+	  echo " $(manoneDATA_INSTALL) $$d$$p $(DESTDIR)$(manonedir)/$${f}$(manext)"; \
+	  $(manoneDATA_INSTALL) $$d$$p $(DESTDIR)$(manonedir)/$${f}$(manext); \
 	done
 
 uninstall-manoneDATA:
 	@$(NORMAL_UNINSTALL)
 	@list='$(manone_DATA)'; for p in $$list; do \
 	  f="`echo $$p | sed -e 's|^.*/||'`"; \
-	  echo " rm -f $(DESTDIR)$(manonedir)/$$f"; \
-	  rm -f $(DESTDIR)$(manonedir)/$$f; \
+	  echo " rm -f $(DESTDIR)$(manonedir)/$${f}$(manext)"; \
+	  rm -f $(DESTDIR)$(manonedir)/$${f}$(manext); \
 	done
 manthreeDATA_INSTALL = $(INSTALL_DATA)
 install-manthreeDATA: $(manthree_DATA)
@@ -257,16 +258,16 @@
 	@list='$(manthree_DATA)'; for p in $$list; do \
 	  if test -f "$$p"; then d=; else d="$(srcdir)/"; fi; \
 	  f="`echo $$p | sed -e 's|^.*/||'`"; \
-	  echo " $(manthreeDATA_INSTALL) $$d$$p $(DESTDIR)$(manthreedir)/$$f"; \
-	  $(manthreeDATA_INSTALL) $$d$$p $(DESTDIR)$(manthreedir)/$$f; \
+	  echo " $(manthreeDATA_INSTALL) $$d$$p $(DESTDIR)$(manthreedir)/$${f}$(manext)"; \
+	  $(manthreeDATA_INSTALL) $$d$$p $(DESTDIR)$(manthreedir)/$${f}$(manext); \
 	done
 
 uninstall-manthreeDATA:
 	@$(NORMAL_UNINSTALL)
 	@list='$(manthree_DATA)'; for p in $$list; do \
 	  f="`echo $$p | sed -e 's|^.*/||'`"; \
-	  echo " rm -f $(DESTDIR)$(manthreedir)/$$f"; \
-	  rm -f $(DESTDIR)$(manthreedir)/$$f; \
+	  echo " rm -f $(DESTDIR)$(manthreedir)/$${f}$(manext)"; \
+	  rm -f $(DESTDIR)$(manthreedir)/$${f}$(manext); \
 	done
 tags: TAGS
 TAGS:
diff -urN -x '*~' lesstif-0.93.36-orig/doc/lessdox/widgets/Makefile.in lesstif-0.93.36/doc/lessdox/widgets/Makefile.in
--- lesstif-0.93.36-orig/doc/lessdox/widgets/Makefile.in	2002-08-06 11:52:14.000000000 +0200
+++ lesstif-0.93.36/doc/lessdox/widgets/Makefile.in	2003-04-30 15:08:01.000000000 +0200
@@ -144,6 +144,7 @@
 manonedir = $(exec_prefix)/man/man1
 manthreedir = $(exec_prefix)/man/man3
 manfivedir = $(exec_prefix)/man/man5
+manext = x
 
 manone_DATA = 
 manthree_DATA = \
@@ -247,16 +248,16 @@
 	@list='$(manfive_DATA)'; for p in $$list; do \
 	  if test -f "$$p"; then d=; else d="$(srcdir)/"; fi; \
 	  f="`echo $$p | sed -e 's|^.*/||'`"; \
-	  echo " $(manfiveDATA_INSTALL) $$d$$p $(DESTDIR)$(manfivedir)/$$f"; \
-	  $(manfiveDATA_INSTALL) $$d$$p $(DESTDIR)$(manfivedir)/$$f; \
+	  echo " $(manfiveDATA_INSTALL) $$d$$p $(DESTDIR)$(manfivedir)/$${f}$(manext)"; \
+	  $(manfiveDATA_INSTALL) $$d$$p $(DESTDIR)$(manfivedir)/$${f}$(manext); \
 	done
 
 uninstall-manfiveDATA:
 	@$(NORMAL_UNINSTALL)
 	@list='$(manfive_DATA)'; for p in $$list; do \
 	  f="`echo $$p | sed -e 's|^.*/||'`"; \
-	  echo " rm -f $(DESTDIR)$(manfivedir)/$$f"; \
-	  rm -f $(DESTDIR)$(manfivedir)/$$f; \
+	  echo " rm -f $(DESTDIR)$(manfivedir)/$${f}$(manext)"; \
+	  rm -f $(DESTDIR)$(manfivedir)/$${f}$(manext); \
 	done
 manoneDATA_INSTALL = $(INSTALL_DATA)
 install-manoneDATA: $(manone_DATA)
@@ -265,16 +266,16 @@
 	@list='$(manone_DATA)'; for p in $$list; do \
 	  if test -f "$$p"; then d=; else d="$(srcdir)/"; fi; \
 	  f="`echo $$p | sed -e 's|^.*/||'`"; \
-	  echo " $(manoneDATA_INSTALL) $$d$$p $(DESTDIR)$(manonedir)/$$f"; \
-	  $(manoneDATA_INSTALL) $$d$$p $(DESTDIR)$(manonedir)/$$f; \
+	  echo " $(manoneDATA_INSTALL) $$d$$p $(DESTDIR)$(manonedir)/$${f}$(manext)"; \
+	  $(manoneDATA_INSTALL) $$d$$p $(DESTDIR)$(manonedir)/$${f}$(manext); \
 	done
 
 uninstall-manoneDATA:
 	@$(NORMAL_UNINSTALL)
 	@list='$(manone_DATA)'; for p in $$list; do \
 	  f="`echo $$p | sed -e 's|^.*/||'`"; \
-	  echo " rm -f $(DESTDIR)$(manonedir)/$$f"; \
-	  rm -f $(DESTDIR)$(manonedir)/$$f; \
+	  echo " rm -f $(DESTDIR)$(manonedir)/$${f}$(manext)"; \
+	  rm -f $(DESTDIR)$(manonedir)/$${f}$(manext); \
 	done
 manthreeDATA_INSTALL = $(INSTALL_DATA)
 install-manthreeDATA: $(manthree_DATA)
@@ -283,16 +284,16 @@
 	@list='$(manthree_DATA)'; for p in $$list; do \
 	  if test -f "$$p"; then d=; else d="$(srcdir)/"; fi; \
 	  f="`echo $$p | sed -e 's|^.*/||'`"; \
-	  echo " $(manthreeDATA_INSTALL) $$d$$p $(DESTDIR)$(manthreedir)/$$f"; \
-	  $(manthreeDATA_INSTALL) $$d$$p $(DESTDIR)$(manthreedir)/$$f; \
+	  echo " $(manthreeDATA_INSTALL) $$d$$p $(DESTDIR)$(manthreedir)/$${f}$(manext)"; \
+	  $(manthreeDATA_INSTALL) $$d$$p $(DESTDIR)$(manthreedir)/$${f}$(manext); \
 	done
 
 uninstall-manthreeDATA:
 	@$(NORMAL_UNINSTALL)
 	@list='$(manthree_DATA)'; for p in $$list; do \
 	  f="`echo $$p | sed -e 's|^.*/||'`"; \
-	  echo " rm -f $(DESTDIR)$(manthreedir)/$$f"; \
-	  rm -f $(DESTDIR)$(manthreedir)/$$f; \
+	  echo " rm -f $(DESTDIR)$(manthreedir)/$${f}$(manext)"; \
+	  rm -f $(DESTDIR)$(manthreedir)/$${f}$(manext); \
 	done
 tags: TAGS
 TAGS:
