--- libdc1394-1.0.0/examples/grab_partial_image.c.orig	2006-01-22 16:58:25.000000000 +0200
+++ libdc1394-1.0.0/examples/grab_partial_image.c	2006-01-22 16:59:27.000000000 +0200
@@ -213,7 +213,7 @@
     /*---------------------------------------------------------------------
      *  output elapsed time
      *---------------------------------------------------------------------*/
-    elapsed_time = (float)(times(&tms_buf) - start_time) / CLK_TCK;
+    elapsed_time = (float)(times(&tms_buf) - start_time) / CLOCKS_PER_SEC;
     printf( "got frame %d. elapsed time: %g sec ==> %g frames/second\n",
             i, elapsed_time, (float)i / elapsed_time);
   }
