--- ./src/examine_client.c.vanilla	2004-06-21 22:08:42.991214496 +0200
+++ ./src/examine_client.c	2004-06-21 22:09:21.631340304 +0200
@@ -25,6 +25,9 @@
 Ecore_Config_Ipc_Server_List *examine_client_server;
 examine_prop   *prop_list;
 
+void examine_client_list_props_cb(void);
+void examine_client_get_val_cb(void);
+
 /*****************************************************************************/
 
 
