--- vlc-0.7.0/modules/demux/mkv.cpp.orig	2004-03-04 12:11:41.000000000 +0100
+++ vlc-0.7.0/modules/demux/mkv.cpp	2004-03-04 12:12:47.000000000 +0100
@@ -633,7 +633,7 @@
                         else  if( EbmlId( *el3 ) == KaxCodecPrivate::ClassInfos.GlobalId )
                         {
                             KaxCodecPrivate &cpriv = *(KaxCodecPrivate*)el3;
-                            cpriv.ReadData( p_sys->es->I_O() );
+                            cpriv.ReadData( p_sys->es->I_O(), SCOPE_ALL_DATA );
 
                             tk.i_extra_data = cpriv.GetSize();
                             if( tk.i_extra_data > 0 )
@@ -856,7 +856,7 @@
                         {
                             KaxSeekID &sid = *(KaxSeekID*)el;
 
-                            sid.ReadData( p_sys->es->I_O() );
+                            sid.ReadData( p_sys->es->I_O(), SCOPE_ALL_DATA );
 
                             id = EbmlId( sid.GetBuffer(), sid.GetSize() );
                         }
