--- gconf-editor-2.8.1/docs/C/Makefile.in~	2004-10-01 13:45:47.000000000 +0200
+++ gconf-editor-2.8.1/docs/C/Makefile.in	2004-10-15 19:31:55.040055488 +0200
@@ -335,7 +335,8 @@
 install-data-hook-omf:
 	$(mkinstalldirs) $(DESTDIR)$(omf_dest_dir)
 	for file in $(omffile); do \
-		$(INSTALL_DATA) $$file.out $(DESTDIR)$(omf_dest_dir)/$$file; \
+		[ -f $$file.out ] && file=$$file.out; \
+		$(INSTALL_DATA) $$file $(DESTDIR)$(omf_dest_dir)/$$file; \
 	done
 	-scrollkeeper-update -p $(scrollkeeper_localstate_dir) -o $(DESTDIR)$(omf_dest_dir)
 
