--- ./Makefile.orig	2003-03-25 10:29:49.000000000 +0100
+++ ./Makefile	2003-07-13 14:43:55.000000000 +0200
@@ -4,6 +4,9 @@
 LDFLAGS =
 LIBS = -L/usr/X11R6/lib -lX11 -lXtst
 
+PREFIX = /usr
+DESTBIN = $(PREFIX)/sbin
+
 all: sonypid
 
 sonypid: sonypid.o
@@ -16,3 +19,10 @@
 clean:
 	rm -f sonypid.o sonypid *~
 	rm -f sonypi.h
+
+install:
+	install -m 755 -d $(DESTBIN)
+	install -m 755 sonypid $(DESTBIN)
+
+uninstall:
+	rm -f $(DESTBIN)/sonypid
