--- kdelibs-3.5.7/configure.in.orig	2007-06-05 08:57:10.000000000 +1000
+++ kdelibs-3.5.7/configure.in	2007-06-05 08:58:29.000000000 +1000
@@ -1885,12 +1885,12 @@
         GSSAPI_RPATH="-R ${GSSAPI_ROOT}/lib"
       fi
     fi
-    if test -f ${GSSAPI_ROOT}/include/gssapi/gssapi.h ; then
-      gssapi_flavor=MIT
-      GSSAPI_LIBS="${GSSAPI_LIBS}-lgssapi_krb5 -lkrb5 -lk5crypto -lcom_err ${LIBRESOLV}"
-    else
+    if test -f ${GSSAPI_ROOT}/include/heim_err.h ; then
       gssapi_flavor=HEIMDAL
       GSSAPI_LIBS="${GSSAPI_LIBS}-lgssapi -lkrb5 -lasn1 -lcrypto -lroken -lcrypt ${LIBRESOLV}"
+    else
+      gssapi_flavor=MIT
+      GSSAPI_LIBS="${GSSAPI_LIBS}-lgssapi_krb5 -lkrb5 -lk5crypto -lcom_err ${LIBRESOLV}"
     fi
   fi
   if test "x$gssapi_flavor" = "xMIT" ; then
