mail to Chris Evans <chris@scary.beasts.org>

--- ./vsf_findlibs.sh.org	2002-10-28 01:12:53.000000000 +0100
+++ ./vsf_findlibs.sh	2003-04-02 15:01:56.000000000 +0200
@@ -1,7 +1,13 @@
 #!/bin/sh
 # Cheesy hacky location of additional link libraries.
 
-locate_library() { [ ! "$1*" = "`echo $1*`" ]; }
+locate_library() { 
+    while [ "$1" ]; do
+	for i in $1* ; do [ -f "$i" ] && return 0; done
+	shift
+    done
+    return 1
+}
 
 if (grep -q "#define VSF_BUILD_TCPWRAPPERS" builddefs.h); then
   echo "-lwrap";
