--- sysstat-6.0.0/sa.h.vanilla	2005-09-04 11:24:23.000000000 +0000
+++ sysstat-6.0.0/sa.h	2005-09-04 16:44:57.000000000 +0000
@@ -489,7 +489,7 @@
 					 struct stats_one_cpu *);
 extern char	   *get_devname(unsigned int, unsigned int, int);
 extern void	    init_bitmap(unsigned char [], unsigned char, unsigned int);
-extern void	    init_stats(struct file_stats [], unsigned int [][]);
+/*extern void	    init_stats(struct file_stats *, unsigned int [][]);*/
 extern int	    next_slice(unsigned long long, unsigned long long,
 			       struct file_hdr *, int, long);
 extern int	    parse_sar_opt(char * [], int, unsigned int *,
