diff -Nur sysstat-5.0.0.orig/build/CONFIG sysstat-5.0.0/build/CONFIG
--- sysstat-5.0.0.orig/build/CONFIG	2003-11-05 20:45:35.000000000 +0100
+++ sysstat-5.0.0/build/CONFIG	2003-11-29 02:34:02.000000000 +0100
@@ -2,7 +2,7 @@
 # (C) 2000 Sebastien GODARD <sebastien.godard@wanadoo.fr>
 
 # Directories
-PREFIX = /usr/local
+PREFIX = /usr
 SA_DIR = /var/log/sa
 MAN_DIR = ${PREFIX}/man
 CLEAN_SA_DIR = n
diff -Nur sysstat-5.0.0.orig/build/CONFIG.def sysstat-5.0.0/build/CONFIG.def
--- sysstat-5.0.0.orig/build/CONFIG.def	2002-05-05 18:54:39.000000000 +0200
+++ sysstat-5.0.0/build/CONFIG.def	2003-11-29 02:33:43.000000000 +0100
@@ -2,7 +2,7 @@
 # (C) 2000 Sebastien GODARD <sebastien.godard@wanadoo.fr>
 
 # Directories
-PREFIX = /usr/local
+PREFIX = /usr
 SA_DIR = /var/log/sa
 MAN_DIR = ${PREFIX}/man
 CLEAN_SA_DIR = n
diff -Nur sysstat-5.0.0.orig/build/Configure.sh sysstat-5.0.0/build/Configure.sh
--- sysstat-5.0.0.orig/build/Configure.sh	2003-01-02 21:32:11.000000000 +0100
+++ sysstat-5.0.0/build/Configure.sh	2003-11-29 02:34:30.000000000 +0100
@@ -10,10 +10,10 @@
 
 # Installation directory
 
-PREFIX=`${ASK} 'Installation directory:' "/usr/local" "prefix"`
+PREFIX=`${ASK} 'Installation directory:' "/usr" "prefix"`
 if [ ! -d ${PREFIX} ]; then
-	echo "WARNING: Directory ${PREFIX} not found: Using default (/usr/local)."
-	PREFIX=/usr/local
+	echo "WARNING: Directory ${PREFIX} not found: Using default (/usr)."
+	PREFIX=/usr
 fi
 
 # System Activity directory
diff -Nur sysstat-5.0.0.orig/sa.h sysstat-5.0.0/sa.h
--- sysstat-5.0.0.orig/sa.h	2003-11-05 20:42:36.000000000 +0100
+++ sysstat-5.0.0/sa.h	2003-11-29 02:31:58.000000000 +0100
@@ -118,7 +118,7 @@
 #define NET_SOCKSTAT	"/proc/net/sockstat"
 #define SADC		"sadc"
 #define SADC_PATH	"/usr/lib/sa/sadc"
-#define SADC_LOCAL_PATH	"/usr/local/lib/sa/sadc"
+#define SADC_LOCAL_PATH	"/usr/lib/sa/sadc"
 #define LOADAVG		"/proc/loadavg"
 #define VMSTAT		"/proc/vmstat"
 
