--- ./xpfe/bootstrap/mozilla.in.vanilla	2005-09-21 12:45:56.000000000 +0200
+++ ./xpfe/bootstrap/mozilla.in	2005-09-21 12:48:39.000000000 +0200
@@ -119,6 +119,17 @@
   fi
 fi
 
+#
+## Make sure that we set the plugin path for backwards compatibility
+##
+if [ "$MOZ_PLUGIN_PATH" ]
+then
+  MOZ_PLUGIN_PATH=$MOZ_PLUGIN_PATH:$moz_libdir/plugins
+else
+  MOZ_PLUGIN_PATH=$moz_libdir/plugins
+fi
+export MOZ_PLUGIN_PATH
+
 script_args=""
 moreargs=""
 debugging=0
