
The usual gcc-3.4 fixes.

  - Rene Rebe <rene@rocklinux.org>

diff -ur commoncpp2-1.0.13-orig/include/cc++/thread.h commoncpp2-1.0.13/include/cc++/thread.h
--- commoncpp2-1.0.13-orig/include/cc++/thread.h	2003-06-07 01:16:45.000000000 +0200
+++ commoncpp2-1.0.13/include/cc++/thread.h	2004-05-04 18:03:06.000000000 +0200
@@ -1080,7 +1080,7 @@
 	static Thread* _main;
 
 	Thread *_parent;
-	enum Cancel _cancel;
+	Cancel _cancel;
 	Semaphore *_start;
 
 	// private data
