--- notecase-1.9.8/src/lib/IniFile.cpp.old	2009-03-09 14:42:39.000000000 +0100
+++ notecase-1.9.8/src/lib/IniFile.cpp	2009-03-09 14:44:10.000000000 +0100
@@ -9,6 +9,7 @@
 #include "IniFile.h"
 #include <algorithm>    //find() function
 #include <string>
+#include <cstdio>
 #include "File64.h"
 #include "FilePath.h"
 #include "debug.h"
@@ -121,14 +122,14 @@
     for(int i=0; i<size; i++)
     {
         //write line with section name
-        snprintf(szBuffer, sizeof(szBuffer), "[%s]\n", m_contents[i].m_name.c_str());
+        std::snprintf(szBuffer, sizeof(szBuffer), "[%s]\n", m_contents[i].m_name.c_str());
 		file.Write(szBuffer, strlen(szBuffer));
 
         int count = m_contents[i].m_Keys.size();
         for(int j=0; j<count; j++)
         {
             //write "key = value"
-            snprintf(szBuffer, sizeof(szBuffer), "%s=%s\n",
+            std::snprintf(szBuffer, sizeof(szBuffer), "%s=%s\n",
                 m_contents[i].m_Keys[j].m_name.c_str(),
                 m_contents[i].m_Keys[j].m_value.c_str());
 			file.Write(szBuffer, strlen(szBuffer));
@@ -210,7 +211,7 @@
 bool IniFile::SetValue(const char *szSection, const char *szKey, const int nValue)
 {
     char szNumber[30] = "";
-    snprintf(szNumber, sizeof(szNumber), "%d", nValue);
+    std::snprintf(szNumber, sizeof(szNumber), "%d", nValue);
 
     return SetValue(szSection, szKey, szNumber);
 }
