diff -ruN nx-X11/config/cf/site.def nx-X11-new/config/cf/site.def
--- nx-X11/config/cf/site.def	2002-02-27 01:51:12.000000000 +0100
+++ nx-X11-new/config/cf/site.def	2004-03-25 13:12:17.926673776 +0100
@@ -72,23 +72,25 @@
 #ifdef AfterVendorCF
 
 #ifndef ProjectRoot
-#define ProjectRoot /usr/X11R6
+#define ProjectRoot /opt/nx
 #endif
 
 /*
  * On some platforms, some things may be installed outside of ProjectRoot
  * by default.  To disable this, uncomment the following line.
  *
-#define NothingOutsideProjectRoot YES
  */
+#define NothingOutsideProjectRoot YES
+#define FontconfigFontsConfDir ProjectRoot/etc/fonts
+#define LinkGLToUsrInclude NO
 
 /*
  * Set EtcX11Directory if you want config file links installed under
  * ProjectRoot rather than in /etc/X11.  See also HasVarDirectory,
  * UseEtcX11 and UseSeparateConfDir.
  *
-#define EtcX11Directory ProjectRoot/etc
  */
+#define EtcX11Directory ProjectRoot/etc
 
 
 /* Only set HasXdmAuth to YES if you have a Wraphelp.c file. */
