--- lua-5.1-alpha/src/lstrlib.c.orig	2005-09-04 14:50:17.000000000 +0200
+++ lua-5.1-alpha/src/lstrlib.c	2005-09-04 15:09:09.000000000 +0200
@@ -781,6 +781,33 @@
   return 1;
 }
 
+static int str_split (lua_State *L) {
+  size_t slen;
+  const char *s = luaL_checklstring(L, 1, &slen);
+  size_t seplen;
+  const char *sep = luaL_optlstring(L, 2, " \t\n\r\v\f", &seplen);
+
+  size_t pos,p,i;
+  size_t t=1;
+
+  lua_newtable(L);
+
+  /* loop over string */
+  for(pos=0,p=0; p<slen; p++) {
+    /* loop over seperators */
+    for(i=0; i<seplen; i++) { 
+      if (s[p] == sep[i]) {
+	if (pos-p > 1) { /* FIXME really skip empty elements? */
+	  lua_pushlstring(L, s+pos, p-pos);
+	  lua_rawseti(L, -2, t++);
+	}
+	pos = p+1;
+      }
+    }
+  }
+
+  return 1;
+}
 
 static const luaL_Reg strlib[] = {
   {"byte", str_byte},
@@ -798,6 +825,7 @@
   {"reverse", str_reverse},
   {"sub", str_sub},
   {"upper", str_upper},
+  {"split", str_split},
   {NULL, NULL}
 };
 
