--- lua-5.1/src/lstrlib.c.orig	2006-03-08 16:50:15.000000000 +0100
+++ lua-5.1/src/lstrlib.c	2006-03-08 16:52:44.000000000 +0100
@@ -815,6 +815,33 @@
   return 1;
 }
 
+static int str_split (lua_State *L) {
+  size_t slen;
+  const char *s = luaL_checklstring(L, 1, &slen);
+  size_t seplen;
+  const char *sep = luaL_optlstring(L, 2, " \t\n\r\v\f", &seplen);
+
+  size_t pos,p,i;
+  size_t t=1;
+
+  lua_newtable(L);
+
+  /* loop over string */
+  for(pos=0,p=0; p<slen; p++) {
+    /* loop over seperators */
+    for(i=0; i<seplen; i++) {
+      if (s[p] == sep[i]) {
+      if (pos-p > 1) { /* FIXME really skip empty elements? */
+         lua_pushlstring(L, s+pos, p-pos);
+         lua_rawseti(L, -2, t++);
+       }
+       pos = p+1;
+      }
+    }
+  }
+
+  return 1;
+}
 
 static const luaL_Reg strlib[] = {
   {"byte", str_byte},
@@ -832,6 +859,7 @@
   {"reverse", str_reverse},
   {"sub", str_sub},
   {"upper", str_upper},
+  {"split", str_split},
   {NULL, NULL}
 };
 
