--- llvm-16.0.1.src/lib/MC/ELFObjectWriter.cpp.vanilla	2023-04-10 20:58:25.759786704 +0200
+++ llvm-16.0.1.src/lib/MC/ELFObjectWriter.cpp	2023-04-10 20:59:49.456781201 +0200
@@ -1449,7 +1449,8 @@
     const MCSection &SecB = SymB.getSection();
     if (&SecB != &FixupSection) {
       Ctx.reportError(Fixup.getLoc(),
-                      "Cannot represent a difference across sections");
+		      Twine("symbol '") + SymB.getName() +
+                      "' can not represent a difference across sections: " + SecB.getName());
       return;
     }
 
