--- ./gcl-2.6.5/o/sfaslbfd.c.orig	2004-09-25 00:00:24.804187230 +0200
+++ ./gcl-2.6.5/o/sfaslbfd.c	2004-09-25 00:00:46.287771708 +0200
@@ -263,7 +263,7 @@
 
     current=round_up(current,1<<s->alignment_power);
 
-    current+=s->_raw_size;
+    current+=s->rawsize; 
 
   }
   curr_size=(unsigned long)current;
@@ -288,7 +288,7 @@
 
     m=round_up(m,1<<s->alignment_power);
     s->output_section->vma=(unsigned long)m;
-    m+=s->_raw_size;
+    m+=s->rawsize;
 	     
   }
 
@@ -353,7 +353,7 @@
 					     v,0,q)) 
        FEerror("Cannot get relocated section contents\n",0);
 
-     memcpy((void *)(unsigned long)s->output_section->vma,v,s->_raw_size);
+     memcpy((void *)(unsigned long)s->output_section->vma,v,s->rawsize);
      
    }
  }
