diff -ru xcdroast-0.98alpha14-/src/io.c xcdroast-0.98alpha14/src/io.c
--- xcdroast-0.98alpha14-/src/io.c	2003-05-29 15:14:42.000000000 +0200
+++ xcdroast-0.98alpha14/src/io.c	2003-06-15 22:42:15.000000000 +0200
@@ -5507,7 +5507,11 @@
 	}
 
 	/* build command line */
-	get_wrap_path_cdrecord(tmp3);
+	if (curset.isProDVD) {
+        	get_wrap_path("CDRECORDPRODVD", tmp3);
+	} else {
+        	get_wrap_path("CDRECORD", tmp3);
+	}
 	g_snprintf(line,MAXLINE,"%s %s -atip 2>&1",
 		tmp3,tmp);
 
@@ -5518,6 +5522,11 @@
                 g_error("popen error\n");
         }
 
+	/* syntax for dvd-atip is different */
+	if (is_dvdwriter(curset.writer_devnr)) {
+		found_atip = 1;
+	}
+
 	errcount = 0;
         for (;;) {
                 if (fgets(line,MAXLINE,fpin) == NULL) 
