diff -ru xcdroast-0.98alpha14-/src/io.c xcdroast-0.98alpha14/src/io.c
--- xcdroast-0.98alpha14-/src/io.c	2003-06-15 22:44:04.000000000 +0200
+++ xcdroast-0.98alpha14/src/io.c	2003-06-15 22:46:10.000000000 +0200
@@ -6614,7 +6614,7 @@
 				read_output_ctrl = 4;
 			} 
 			/* check Medium error/buffer underrun */
-			if (strncmp(tmp2,"Input/output error",18)) {
+			if (strncmp(tmp2,"Input/output error",18) == 0) {
 				read_output_ctrl = 5;
 			}
 		}
