Rene Rebe: This fixes compilation with gcc-3.3

--- glibc-2.3.2/stdio-common/sscanf.c	2002-08-10 20:09:08.000000000 +0200
+++ glibc-2.3.2/stdio-common/sscanf.c	2003-05-13 09:46:19.000000000 +0200
@@ -27,12 +27,10 @@
 /* Read formatted input from S, according to the format string FORMAT.  */
 /* VARARGS2 */
 int
-sscanf (s, format)
-     const char *s;
-     const char *format;
+sscanf (const char *s, const char *format, ... )
 {
-  va_list arg;
   int done;
+  va_list arg;
 
   va_start (arg, format);
   done = __vsscanf (s, format, arg);
