--- glibc-2.3.2/sysdeps/unix/sysv/linux/sparc/sys/ucontext.h.orig	2003-08-03 19:45:12.000000000 +0000
+++ glibc-2.3.2/sysdeps/unix/sysv/linux/sparc/sys/ucontext.h	2003-08-03 19:47:29.000000000 +0000
@@ -19,6 +19,8 @@
 #ifndef _SYS_UCONTEXT_H
 #define _SYS_UCONTEXT_H	1
 
+#ifdef __GLIBC__
+
 #include <features.h>
 #include <signal.h>
 #include <bits/wordsize.h>
@@ -267,4 +269,5 @@
   } ucontext_t;
 
 #endif /* __WORDSIZE == 32 */
+#ennif /* __GLIBC__ */
 #endif /* sys/ucontext.h */
