--- linux-fusion/Makefile.orig	2000-01-01 00:00:00.000000000 -0300
+++ linux-fusion/Makefile	2005-01-31 22:10:27.000000000 -0300
@@ -1,6 +1,7 @@
-KERNEL_MODLIB     = /lib/modules/$(shell uname -r)
+LINUXVERSION     ?= $(shell uname -r)
+KERNEL_MODLIB     = /lib/modules/$(LINUXVERSION)
 KERNEL_SOURCE     = $(KERNEL_MODLIB)/build
-KERNEL_PATCHLEVEL = $(shell uname -r | cut -d . -f 2)
+KERNEL_PATCHLEVEL = $(shell echo $(LINUXVERSION) | cut -d . -f 2)
 #KERNEL_PATCHLEVEL = $(shell grep 'PATCHLEVEL =' $(KERNEL_SOURCE)/Makefile | cut -d ' ' -f 3)
 
 SUB = linux/drivers/char/fusion
@@ -14,7 +15,7 @@
 		SUBDIRS=`pwd`/$(SUB) modules
 
 install:
-	(if test $(KERNEL_PATCHLEVEL) = 4; then cp $(SUB)/fusion.o $(KERNEL_MODLIB); else cp $(SUB)/fusion.ko $(KERNEL_MODLIB); fi) && depmod -ae
+	(if test $(KERNEL_PATCHLEVEL) = 4; then cp $(SUB)/fusion.o $(KERNEL_MODLIB); else cp $(SUB)/fusion.ko $(KERNEL_MODLIB)/misc; fi)
 
 clean:
 	find $(SUB) -name *.o -o -name *.ko -o -name .*.o.cmd -o \
