--- busybox-1.01/include/applets.h	2005-08-17 03:29:15.000000000 +0200
+++ busybox-1.01/include/applets.h.orig	2005-11-01 01:03:25.461831208 +0100
@@ -326,7 +326,7 @@
 	APPLET(length, length_main, _BB_DIR_USR_BIN, _BB_SUID_NEVER)
 #endif
 #ifdef CONFIG_FEATURE_INITRD
-	APPLET_NOUSAGE("linuxrc", init_main, _BB_DIR_ROOT, _BB_SUID_NEVER)
+	APPLET_NOUSAGE("linuxrc", init_main, _BB_DIR_SBIN, _BB_SUID_NEVER)
 #endif
 #ifdef CONFIG_LN
 	APPLET(ln, ln_main, _BB_DIR_BIN, _BB_SUID_NEVER)
