--- ./src/Makefile.org	1999-11-03 01:47:10.000000000 +0100
+++ ./src/Makefile	2003-04-06 21:52:34.000000000 +0200
@@ -6,8 +6,8 @@
 CFLAGS		= ${CWARN} ${COPT} ${CDEF} ${CDBG}
 LDFLAGS		= 
 LIBS            = -lpthread
-INSTDIR		= /usr/local/sbin
-MANDIR		= /usr/local/man/man8
+INSTDIR		= /usr/sbin
+MANDIR		= /usr/man/man8
 SHELL		= /bin/sh
 
 all:	dnrd
@@ -18,8 +18,8 @@
 
 install: dnrd
 	strip dnrd
-	install -m 0751 -o root -g root -s dnrd ${INSTDIR}
-	install -m 0644 -o root -g root -s ../doc/dnrd.8 ${MANDIR}
+	install -m 0751 -o root -g root dnrd ${INSTDIR}
+	install -m 0644 -o root -g root ../doc/dnrd.8 ${MANDIR}
 
 sources = args.c cache.c common.c dns.c lib.c main.c master.c \
 	  query.c relay.c sig.c tcp.c udp.c
