--- acl-2.2.27/setfacl/Makefile.orig	2005-01-26 19:27:04.000000000 -0800
+++ acl-2.2.27/setfacl/Makefile	2005-01-26 19:27:15.000000000 -0800
@@ -37,7 +37,7 @@
 CFILES = setfacl.c do_set.c sequence.c parse.c
 HFILES = sequence.h parse.h
 
-LLDLIBS = $(LIBACL) $(LIBATTR) $(LIBMISC)
+LLDLIBS = $(LIBACL) $(LIBATTR) $(LIBMISC) /usr/lib/libintl.a
 LTDEPENDENCIES = $(LIBACL) $(LIBMISC)
 
 default: $(LTCOMMAND)
