--- ./configure.orig	2005-01-04 00:18:33.000000000 -0300
+++ ./configure	2005-01-04 00:20:07.000000000 -0300
@@ -945,6 +945,12 @@
 target_configdirs=`echo ${target_libraries} ${target_tools}`
 build_configdirs=`echo ${build_libs} ${build_tools}`
 
+case "$target" in
+  *-*-*-uclibc*)
+    build_modules=
+    ;;
+esac
+
 ################################################################################
 
 srcname="gnu development package"
--- ./configure.in.orig	2005-01-04 00:18:42.000000000 -0300
+++ ./configure.in	2005-01-04 00:20:27.000000000 -0300
@@ -192,6 +192,12 @@
 target_configdirs=`echo ${target_libraries} ${target_tools}`
 build_configdirs=`echo ${build_libs} ${build_tools}`
 
+case "$target" in
+  *-*-*-uclibc*)
+    build_modules=
+    ;;
+esac
+
 ################################################################################
 
 srcname="gnu development package"
