--- gc6.4/configure.orig	2005-02-02 04:44:52.000000000 +0000
+++ gc6.4/configure	2005-02-02 04:47:08.000000000 +0000
@@ -9453,7 +9453,7 @@
   ac_status=$?
   echo "$as_me:$LINENO: \$? = $ac_status" >&5
   (exit $ac_status); }; }; then
-  ac_cv_lib_unwind_backtrace=yes
+  ac_cv_lib_unwind_backtrace=no
 else
   echo "$as_me: failed program was:" >&5
 sed 's/^/| /' conftest.$ac_ext >&5
--- gc6.4/include/gc.h.orig	2005-02-02 04:53:13.000000000 +0000
+++ gc6.4/include/gc.h	2005-02-02 04:53:43.000000000 +0000
@@ -470,12 +470,6 @@
 
 #ifdef __linux__
 # include <features.h>
-# if (__GLIBC__ == 2 && __GLIBC_MINOR__ >= 1 || __GLIBC__ > 2) \
-     && !defined(__ia64__)
-#   ifndef GC_HAVE_BUILTIN_BACKTRACE
-#     define GC_HAVE_BUILTIN_BACKTRACE
-#   endif
-# endif
 # if defined(__i386__) || defined(__x86_64__)
 #   define GC_CAN_SAVE_CALL_STACKS
 # endif
