diff -r -u bsd-games-2.16/dm/Makefrag bsd-games-2.16.patched/dm/Makefrag
--- bsd-games-2.16/dm/Makefrag	2004-01-27 21:33:19.000000000 +0000
+++ bsd-games-2.16.patched/dm/Makefrag	2005-05-12 18:28:57.000000000 +0000
@@ -27,7 +27,7 @@
 # SUCH DAMAGE.
 
 # Add -DLOG if you want logging
-dm_DEFS := $(GETLOADAVG_DEFS) -D_GNU_SOURCE -DSUPPORT_UTMPX -DSUPPORT_UTMP # -DLOG
+dm_DEFS := $(GETLOADAVG_DEFS) -D_GNU_SOURCE -DSUPPORT_UTMP # -DLOG
 dm_DIRS := $(GAMESDIR) $(MAN8DIR) $(MAN5DIR)
 
 dm_all:	dm/dm dm/dm.8 dm/dm.conf.5
diff -r -u bsd-games-2.16/hunt/hunt/hunt.c bsd-games-2.16.patched/hunt/hunt/hunt.c
--- bsd-games-2.16/hunt/hunt/hunt.c	2004-02-08 22:29:06.000000000 +0000
+++ bsd-games-2.16.patched/hunt/hunt/hunt.c	2005-05-12 18:29:32.000000000 +0000
@@ -51,7 +51,7 @@
 static struct termios saved_tty;
 # endif
 # include	<unistd.h>
-# include	<ifaddrs.h>
+/*# include	<ifaddrs.h> */
 
 # include	"hunt.h"
 
diff -r -u bsd-games-2.16/hunt/huntd/bsd.h bsd-games-2.16.patched/hunt/huntd/bsd.h
--- bsd-games-2.16/hunt/huntd/bsd.h	2003-12-17 02:46:45.000000000 +0000
+++ bsd-games-2.16.patched/hunt/huntd/bsd.h	2005-05-12 18:32:50.000000000 +0000
@@ -32,7 +32,7 @@
  */
 
 # if defined(BSD_RELEASE) && BSD_RELEASE >= 43
-# define	BROADCAST
+/*# define	BROADCAST*/
 # define	SYSLOG_43
 # endif
 # if defined(BSD_RELEASE) && BSD_RELEASE == 42
