--- ./src/Makefile.in.orig	2004-02-07 23:01:51.000000000 -0300
+++ ./src/Makefile.in	2004-02-07 23:06:44.000000000 -0300
@@ -209,7 +209,7 @@
 printf_LDADD = $(LDADD) @POW_LIB@ @LIBICONV@
 
 # If necessary, add -lm to resolve use of floor, rint, modf.
-seq_LDADD = $(LDADD) @SEQ_LIBM@
+seq_LDADD = $(LDADD) @POW_LIB@ @SEQ_LIBM@
 
 # If necessary, add -lm to resolve the `pow' reference in lib/strtod.c
 # or for the fesetround reference in programs using nanosec.c.
