--- emacs-21.3/configure.orig	2005-01-23 17:17:09.000000000 -0800
+++ emacs-21.3/configure	2005-01-23 17:27:56.000000000 -0800
@@ -1649,6 +1649,7 @@
       *-esix* )			opsys=esix ;;
       *-xenix* )		opsys=xenix ;;
       *-linux-gnu* )		opsys=gnu-linux ;;
+      *-linux-uclibc* )		opsys=gnu-linux ;;
       *-sco3.2v4* )		opsys=sco4 ; NON_GNU_CPP=/lib/cpp  ;;
       *-sco3.2v5* )		opsys=sco5 
 				NON_GNU_CPP=/lib/cpp
