--- gcc-3.4.3/gcc/mklibgcc.in.orig	2005-01-28 19:38:19.000000000 -0800
+++ gcc-3.4.3/gcc/mklibgcc.in	2005-01-28 19:38:40.000000000 -0800
@@ -398,7 +398,7 @@
 
   libgcc_a_objs="$libgcc_objs $libgcc_st_objs"
 
-  if [ "@libgcc_visibility@" = yes -a "$SHLIB_LINK" ]; then
+#  if [ "@libgcc_visibility@" = yes -a "$SHLIB_LINK" ]; then
     libgcc_a_objs=
     echo ""
     echo "libgcc/${dir}/stacknote.s: stmp-dirs"
@@ -414,7 +414,7 @@
       echo '	( $(NM_FOR_TARGET) '${SHLIB_NM_FLAGS} ${o}' | $(AWK) '\''NF == 3 && $$2 !~ /^[UN]$$/ { print "\t.hidden", $$3 }'\''; cat libgcc/${dir}/stacknote.s ) | $(GCC_FOR_TARGET) $(LIBGCC2_CFLAGS) '${flags}' -r -nostdinc -nostdlib -o $@ '${o}' -xassembler -'
       libgcc_a_objs="${libgcc_a_objs} ${oS}"
     done
-  fi
+#  fi
 
   # Each of these .a files depends on stmp-dirs.  It would seem that
   # this dependency is redundant, since each of the object files
