--- gcc-4.3.1/libstdc++-v3/config/locale/uclibc/time_members.cc.vanilla	2008-07-20 21:37:44.000000000 +0200
+++ gcc-4.3.1/libstdc++-v3/config/locale/uclibc/time_members.cc	2008-07-20 21:37:55.000000000 +0200
@@ -35,6 +35,7 @@
 // Written by Benjamin Kosnik <bkoz@redhat.com>
 
 #include <locale>
+#include <stdlib.h> // For free()
 #include <bits/c++locale_internal.h>
 
 #ifdef __UCLIBC_MJN3_ONLY__
--- gcc-4.3.1/libstdc++-v3/config/locale/uclibc/ctype_members.cc.vanilla	2008-07-20 21:32:55.000000000 +0200
+++ gcc-4.3.1/libstdc++-v3/config/locale/uclibc/ctype_members.cc	2008-07-20 21:36:18.000000000 +0200
@@ -47,21 +47,19 @@
 {
   // NB: The other ctype<char> specializations are in src/locale.cc and
   // various /config/os/* files.
-  template<>
-    ctype_byname<char>::ctype_byname(const char* __s, size_t __refs)
-    : ctype<char>(0, false, __refs) 
-    { 		
-      if (std::strcmp(__s, "C") != 0 && std::strcmp(__s, "POSIX") != 0)
-	{
-	  this->_S_destroy_c_locale(this->_M_c_locale_ctype);
-	  this->_S_create_c_locale(this->_M_c_locale_ctype, __s); 
-#ifdef __UCLIBC_HAS_XLOCALE__
-	  this->_M_toupper = this->_M_c_locale_ctype->__ctype_toupper;
-	  this->_M_tolower = this->_M_c_locale_ctype->__ctype_tolower;
-	  this->_M_table = this->_M_c_locale_ctype->__ctype_b;
-#endif
-	}
-    }
+  ctype_byname<char>::ctype_byname(const char* __s, size_t __refs)
+  : ctype<char>(0, false, __refs) 
+  {    
+    if (std::strcmp(__s, "C") != 0 && std::strcmp(__s, "POSIX") != 0)
+      {
+       this->_S_destroy_c_locale(this->_M_c_locale_ctype);
+       this->_S_create_c_locale(this->_M_c_locale_ctype, __s); 
+      }
+  }
+
+  ctype_byname<char>::~ctype_byname()
+  { }
+
 
 #ifdef _GLIBCXX_USE_WCHAR_T  
   ctype<wchar_t>::__wmask_type
--- gcc-4.3.1/libstdc++-v3/config/locale/uclibc/codecvt_members.cc.vanilla	2008-07-20 21:29:58.000000000 +0200
+++ gcc-4.3.1/libstdc++-v3/config/locale/uclibc/codecvt_members.cc	2008-07-20 21:32:21.000000000 +0200
@@ -33,6 +33,8 @@
 
 // Written by Benjamin Kosnik <bkoz@redhat.com>
 
+#include <cstdlib>  // For MB_CUR_MAX
+#include <climits>  // For MB_LEN_MAX
 #include <locale>
 #include <bits/c++locale_internal.h>
 
--- gcc-4.3.1/libstdc++-v3/config/locale/uclibc/messages_members.cc.vanilla	2008-07-20 21:36:34.000000000 +0200
+++ gcc-4.3.1/libstdc++-v3/config/locale/uclibc/messages_members.cc	2008-07-20 21:36:43.000000000 +0200
@@ -33,6 +33,7 @@
 
 // Written by Benjamin Kosnik <bkoz@redhat.com>
 
+#include <stdlib.h> // For free()
 #include <locale>
 #include <bits/c++locale_internal.h>
 
--- gcc-4.3.1/libstdc++-v3/config/locale/uclibc/monetary_members.cc.vanilla	2008-07-20 21:37:19.000000000 +0200
+++ gcc-4.3.1/libstdc++-v3/config/locale/uclibc/monetary_members.cc	2008-07-20 21:37:32.000000000 +0200
@@ -34,6 +34,7 @@
 // Written by Benjamin Kosnik <bkoz@redhat.com>
 
 #include <features.h>
+#include <stdlib.h> // For free()
 #ifdef __UCLIBC_HAS_LOCALE__
 #define _LIBC
 #include <locale>
--- gcc-4.3.1/libstdc++-v3/config/locale/uclibc/c++locale_internal.h.vanilla	2008-07-20 21:41:14.000000000 +0200
+++ gcc-4.3.1/libstdc++-v3/config/locale/uclibc/c++locale_internal.h	2008-07-20 21:42:47.000000000 +0200
@@ -30,6 +30,7 @@
 // Written by Jakub Jelinek <jakub@redhat.com>
 
 #include <bits/c++config.h>
+#include <stdlib.h>
 #include <clocale>
 
 #ifdef __UCLIBC_MJN3_ONLY__
