--- gcc-4.3.1/gcc/config/arm/linux-eabi.h.orig	2008-07-20 20:22:08.000000000 +0200
+++ gcc-4.3.1/gcc/config/arm/linux-eabi.h	2008-07-20 20:22:23.000000000 +0200
@@ -61,7 +61,7 @@
 /* Use ld-linux.so.3 so that it will be possible to run "classic"
    GNU/Linux binaries on an EABI system.  */
 #undef  GLIBC_DYNAMIC_LINKER
-#define GLIBC_DYNAMIC_LINKER "/lib/ld-linux.so.3"
+#define GLIBC_DYNAMIC_LINKER "/lib/ld-uClibc.so.0"
 
 /* At this point, bpabi.h will have clobbered LINK_SPEC.  We want to
    use the GNU/Linux version, not the generic BPABI version.  */
