--- libsafe-2.0-16/src/intercept.c.orig	2005-01-22 12:53:10.000000000 -0800
+++ libsafe-2.0-16/src/intercept.c	2005-01-22 12:47:15.000000000 -0800
@@ -78,8 +78,11 @@
 typedef char *(*getwd_t) (char *buf);
 typedef char *(*gets_t) (char *s);
 typedef char *(*realpath_t) (char *path, char resolved_path[]);
-typedef int (*_IO_vfscanf_t) (_IO_FILE *s, const char *format, _IO_va_list
-	argptr, int *errp);
+typedef int (*_IO_vfscanf_t)
+ (FILE *s,
+   const char *format,
+   va_list argptr,
+   int *errp);
 
 
 
@@ -1443,7 +1446,7 @@
 }
 
 
-int _IO_vfscanf (_IO_FILE *s, const char *format, _IO_va_list argptr, int *errp)
+int _IO_vfscanf (FILE *s, const char *format, va_list argptr, int *errp)
 {
     static _IO_vfscanf_t real_IO_vfscanf = NULL;
     int res, save_count;
