--- mm2.7/src/richmail/richlex.c.orig	2005-01-25 12:45:17.000000000 -0800
+++ mm2.7/src/richmail/richlex.c	2005-01-25 12:45:19.000000000 -0800
@@ -49,10 +49,12 @@
 int CorrectionEnabled = 1;	/* Zero if correction has been disabled */
 int RichtextLessThanFlag = 0;	/* Non-zero to turn on multi-byte '<' hack */
 
+#if 0
 #ifndef AMIGA
 extern	int fgetc ();
 extern	int fputc ();
 #endif
+#endif
 
 int (*RichtextGetc) () = fgetc;	/* Function to call to get characters */
 int (*RichtextPutc) () = fputc; /* Function to call to put characters */
