--- ./ex/filter.c.orig	2005-01-19 00:27:53.000000000 -0300
+++ ./ex/filter.c	2005-01-19 00:28:48.000000000 -0300
@@ -342,7 +342,6 @@
 	const char *cmd;
 	int okpipe;
 {
-	extern const char *const sys_siglist[];
 	size_t len;
 	int nf, pstat;
 	char *p;
@@ -372,7 +371,7 @@
 		len = strlen(p);
 		msgq(sp, M_ERR, "%.*s%s: received signal: %s%s",
 		    MIN(len, 20), p, len > 20 ? " ..." : "",
-		    sys_siglist[WTERMSIG(pstat)],
+		    strsignal(WTERMSIG(pstat)),
 		    WCOREDUMP(pstat) ? "; core dumped" : "");
 		if (nf)
 			FREE_SPACE(sp, p, 0);
