--- parted-1.6.20/debug/clearfat/Makefile.in.orig	2005-02-02 05:28:16.000000000 +0000
+++ parted-1.6.20/debug/clearfat/Makefile.in	2005-02-02 05:26:15.000000000 +0000
@@ -167,7 +167,7 @@
 	$(INCLUDES) $(AM_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS)
 CCLD = $(CC)
 LINK = $(LIBTOOL) --mode=link $(CCLD) $(AM_CFLAGS) $(CFLAGS) \
-	$(AM_LDFLAGS) $(LDFLAGS) -o $@
+	$(AM_LDFLAGS) $(LDFLAGS) -o $@ $(INTLLIBS)
 CFLAGS = @CFLAGS@
 DIST_SOURCES = $(clearfat_SOURCES)
 DIST_COMMON = Makefile.am Makefile.in
