--- php-4.3.7/ext/standard/dns.c.orig	2003-09-02 00:37:46.000000000 +0200
+++ php-4.3.7/ext/standard/dns.c	2004-11-23 11:59:36.973971768 +0100
@@ -325,15 +325,23 @@
 	cp = (u_char *)&ans + HFIXEDSZ;
 	end = (u_char *)&ans +i;
 	for (qdc = ntohs((unsigned short)hp->qdcount); qdc--; cp += i + QFIXEDSZ) {
+#if HAVE_DN_SKIPNAME
 		if ((i = dn_skipname(cp, end)) < 0 ) {
 			RETURN_FALSE;
 		}
+#else
+		RETURN_FALSE;
+#endif
 	}
 	count = ntohs((unsigned short)hp->ancount);
 	while (--count >= 0 && cp < end) {
+#if HAVE_DN_SKIPNAME
 		if ((i = dn_skipname(cp, end)) < 0 ) {
 			RETURN_FALSE;
 		}
+#else
+		RETURN_FALSE;
+#endif
 		cp += i;
 		GETSHORT(type, cp);
 		cp += INT16SZ + INT32SZ;
