--- ./src/Makefile.in.orig	2005-01-19 23:36:21.000000000 -0300
+++ ./src/Makefile.in	2005-01-19 23:44:29.000000000 -0300
@@ -192,10 +192,11 @@
 AM_CFLAGS = -Wall -DLOCALEDIR=\"/usr/share/locale\"
 bin_SCRIPTS = pstree.x11
 fuser_SOURCES = fuser.c comm.h signals.c signals.h loop.h
+fuser_LDADD = @LTLIBINTL@
 killall_SOURCES = killall.c comm.h signals.c signals.h 
-killall_LDADD = @FLASK_LIB@
+killall_LDADD = @FLASK_LIB@ @LTLIBINTL@
 pstree_SOURCES = pstree.c comm.h
-pstree_LDADD = @TERMCAP_LIB@ @FLASK_LIB@
+pstree_LDADD = @TERMCAP_LIB@ @FLASK_LIB@ @LTLIBINTL@
 BUILT_SOURCES = signames.h
 EXTRA_DIST = signames.c 
 CLEANFILES = signames.h pstree.x11
