--- rpm-4.1/db3/configure.orig	2005-01-26 12:23:31.000000000 -0800
+++ rpm-4.1/db3/configure	2005-01-26 12:35:58.000000000 -0800
@@ -12,7 +12,7 @@
 ARGS="`echo $* | sed -e 's% [^ ]*CFLAGS=[^ ]*%%' -e 's% -[^-][^ ]*%%g' -e 's%--cache-file=.*$%%'`"
 
 CC="$CC" CFLAGS="$CFLAGS" $db_dist/configure $ARGS \
-	--enable-shared --enable-static --enable-rpc \
+	--enable-shared --enable-static \
 	--with-uniquename=_rpmdb --srcdir=$db_dist
 
 # XXX hack to get db-3.3.4 to configure properly
