--- shadow-4.0.6/libmisc/login_access.c.orig	2003-05-05 23:44:15.000000000 +0200
+++ shadow-4.0.6/libmisc/login_access.c	2004-11-23 09:21:22.672325008 +0100
@@ -192,6 +192,7 @@
 static int
 netgroup_match (const char *group, const char *machine, const char *user)
 {
+#ifdef NIS
 	static char *mydomain = 0;
 
 	if (mydomain == 0) {
@@ -202,6 +203,10 @@
 	}
 
 	return innetgr (group, machine, user, mydomain);
+#else
+	syslog(LOG_ERR, "NIS netgroup support not configured");              
+	return (NO);
+#endif
 }
 
 /* user_match - match a username against one token */
