--- ./Makefile.orig	2004-05-23 00:51:39.000000000 +0200
+++ ./Makefile	2004-05-23 00:52:08.000000000 +0200
@@ -83,8 +83,8 @@
 	$(MAKE) USE_NETIF=0 SOLARIS=1 arping1-make
 
 install:
-	install -c arping /usr/local/bin/arping
-	install arping.8 /usr/local/man/man8/arping.8
+	install -c arping /usr/bin/arping
+	install arping.8 /usr/man/man8/arping.8
 
 arping.o: arping.c
 	$(CC) -Wall $(CFLAGS) -c `libnet-config --defines` `libnet-config --cflags` arping.c
