--- ./plug-ins/common/Makefile.in.orig	2004-06-17 23:49:47.332120072 +0200
+++ ./plug-ins/common/Makefile.in	2004-06-17 23:50:30.527553368 +0200
@@ -1351,6 +1351,7 @@
 	$(libgimpcolor)							\
 	$(libgimpbase)							\
 	$(GLIB_LIBS)							\
+	$(GTK_LIBS)							\
 	$(RT_LIBS)							\
 	$(INTLLIBS)
 
@@ -1532,6 +1533,7 @@
 	$(libgimpcolor)							\
 	$(libgimpbase)							\
 	$(GLIB_LIBS)							\
+	$(GTK_LIBS)							\
 	$(RT_LIBS)							\
 	$(INTLLIBS)
 
