--- gpm-1.20.6/src/daemon/open_console.c.vanilla	2018-08-08 18:05:02.531843476 +0000
+++ gpm-1.20.6/src/daemon/open_console.c	2018-08-08 18:05:17.019843258 +0000
@@ -27,6 +27,7 @@
 #include <linux/serial.h>           /* for serial console check */
 #include <asm/ioctls.h>            /* for serial console check */
 
+#include <sys/sysmacros.h>
 
 #include "headers/message.h"        /* messaging in gpm  */
 #include "headers/daemon.h"         /* daemon internals  */
--- gpm-1.20.6/src/prog/gpm-root.y.vanilla	2018-08-08 18:07:26.435841312 +0000
+++ gpm-1.20.6/src/prog/gpm-root.y	2018-08-08 18:07:50.655840948 +0000
@@ -1197,7 +1197,7 @@
    /* reap your zombies */
    childaction.sa_handler=reap_children;
 #if defined(__GLIBC__)
-   __sigemptyset(&childaction.sa_mask);
+   sigemptyset(&childaction.sa_mask);
 #else /* __GLIBC__ */
    childaction.sa_mask=0;
 #endif /* __GLIBC__ */
