--- ./squashfs-tools/mksquashfs.c.vanilla	2018-08-09 22:36:31.642299130 +0000
+++ ./squashfs-tools/mksquashfs.c	2018-08-09 22:36:46.814298902 +0000
@@ -60,6 +60,7 @@
 #include <endian.h>
 #include <sys/sysinfo.h>
 #endif
+#include <sys/sysmacros.h>
 
 #include "squashfs_fs.h"
 #include "squashfs_swap.h"
--- ./squashfs-tools/unsquashfs.c.vanilla	2018-08-09 22:36:53.938298795 +0000
+++ ./squashfs-tools/unsquashfs.c	2018-08-09 22:37:09.986298554 +0000
@@ -35,6 +35,7 @@
 #include <sys/types.h>
 #include <sys/time.h>
 #include <sys/resource.h>
+#include <sys/sysmacros.h>
 #include <limits.h>
 #include <ctype.h>
 
