Index: grub-9999/grub-core/disk/lvm.c
===================================================================
--- grub-9999.orig/grub-core/disk/lvm.c
+++ grub-9999/grub-core/disk/lvm.c
@@ -482,6 +482,8 @@ grub_lvm_detect (grub_disk_t disk,
 
 		      if (seg->node_count != 1)
 			seg->stripe_size = grub_lvm_getvalue (&p, "stripe_size = ");
+		      else
+			seg->stripe_size = 0;
 
 		      seg->nodes = grub_zalloc (sizeof (*stripe)
 						* seg->node_count);
