--- curl-7.10.4/configure.orig	2003-04-16 00:04:15.000000000 +0200
+++ curl-7.10.4/configure	2003-04-16 00:04:46.000000000 +0200
@@ -10797,13 +10797,7 @@
 
 echo "$as_me:$LINENO: checking if argv can be written to" >&5
 echo $ECHO_N "checking if argv can be written to... $ECHO_C" >&6
-if test "$cross_compiling" = yes; then
-  { { echo "$as_me:$LINENO: error: cannot run test program while cross compiling
-See \`config.log' for more details." >&5
-echo "$as_me: error: cannot run test program while cross compiling
-See \`config.log' for more details." >&2;}
-   { (exit 1); exit 1; }; }
-else
+
   cat >conftest.$ac_ext <<_ACEOF
 #line $LINENO "configure"
 /* confdefs.h.  */
@@ -10847,7 +10841,6 @@
 
 fi
 rm -f core core.* *.core gmon.out bb.out conftest$ac_exeext conftest.$ac_objext conftest.$ac_ext
-fi
 
 
 
