--- ./configure.orig	2003-01-15 10:17:11.000000000 +0100
+++ ./configure	2003-01-15 10:21:20.000000000 +0100
@@ -10,6 +10,11 @@
 ## M4sh Initialization.  ##
 ## --------------------- ##
 
+# These are not set if no libz is available.
+#
+export CONTENT_ENCODING_TRUE=
+export CONTENT_ENCODING_FALSE='#'
+
 # Be Bourne compatible
 if test -n "${ZSH_VERSION+set}" && (emulate sh) >/dev/null 2>&1; then
   emulate sh
