--- ./Makefile.in.orig	2005-01-11 10:51:33.000000000 -0300
+++ ./Makefile.in	2005-01-11 10:52:37.000000000 -0300
@@ -434,7 +434,7 @@
 	-test -z "$(bin_PROGRAMS)" || rm -f $(bin_PROGRAMS)
 ht$(EXEEXT): $(ht_OBJECTS) $(ht_DEPENDENCIES) 
 	@rm -f ht$(EXEEXT)
-	$(CXXLINK) $(ht_LDFLAGS) $(ht_OBJECTS) $(ht_LDADD) $(LIBS)
+	$(CXXLINK) $(ht_OBJECTS) $(ht_LDADD) $(LIBS) $(ht_LDFLAGS)
 
 mostlyclean-compile:
 	-rm -f *.$(OBJEXT)
