diff -ruN unison-2.9.20/ubase/util.ml unison-2.9.20-avm/ubase/util.ml
--- unison-2.9.20/ubase/util.ml	2002-08-19 23:15:04.000000000 +0200
+++ unison-2.9.20-avm/ubase/util.ml	2003-11-10 22:46:21.000000000 +0100
@@ -217,7 +217,7 @@
     close_in c;
     match l with
       "Darwin" -> true
-    | "linux" | "FreeBSD" | "SunOS" -> false
+    | "Linux" | "FreeBSD" | "SunOS" -> false
     | _ ->
         Printf.printf
          "Warning: 'uname' returned unrecognized result (%s) in Util.isOSX.\n"
