--- strace-4.5.16/mem.c.vanilla	2007-11-09 11:03:22.000000000 +0000
+++ strace-4.5.16/mem.c	2007-11-09 11:03:33.000000000 +0000
@@ -544,7 +544,7 @@
 	return 0;
 }
 
-#if defined(ALPHA) || defined(FREEBSD) || defined(IA64) || defined(SUNOS4) || defined(SVR4)
+#if defined(ALPHA) || defined(FREEBSD) || defined(IA64) || defined(SUNOS4) || defined(SVR4) || defined (SPARC) || defined (SPARC64)
 int
 sys_getpagesize(tcp)
 struct tcb *tcp;
--- strace-4.5.16/syscall.c.vanilla	2007-11-09 10:58:28.000000000 +0000
+++ strace-4.5.16/syscall.c	2007-11-09 10:58:28.000000000 +0000
@@ -42,6 +42,8 @@
 #include <sys/syscall.h>
 #include <sys/param.h>
 
+#include <linux/syscall.h>
+
 #if HAVE_ASM_REG_H
 #if defined (SPARC) || defined (SPARC64)
 #  define fpq kernel_fpq
--- strace-4.5.16/defs.h.vanilla	2007-11-09 10:58:28.000000000 +0000
+++ strace-4.5.16/defs.h	2007-11-09 10:58:28.000000000 +0000
@@ -196,14 +196,15 @@
 
 #ifdef LINUXSPARC
 #include <linux/a.out.h>
-#include <asm/psr.h>
 #define PERSONALITY0_WORDSIZE 4
 #define PERSONALITY1_WORDSIZE 4
 #undef  SUPPORTED_PERSONALITIES
 #if defined(SPARC64)
+#include <asm/psrcompat.h>
 #define SUPPORTED_PERSONALITIES 3
 #define PERSONALITY2_WORDSIZE 8
 #else
+#include <asm/psr.h>
 #define SUPPORTED_PERSONALITIES 2
 #endif /* SPARC64 */
 #endif /* LINUXSPARC */
