--- ./src/Timer.hh.orig	2004-05-14 11:14:47.822257792 +0200
+++ ./src/Timer.hh	2004-05-14 11:16:53.078215992 +0200
@@ -98,9 +98,9 @@
   ~_timer_queue(void) {}
 
   void release(const _Tp& value) {
-    c.erase(std::remove(c.begin(), c.end(), value), c.end());
+    this->c.erase(std::remove(this->c.begin(), this->c.end(), value), this->c.end());
     // after removing the item we need to make the heap again
-    std::make_heap(c.begin(), c.end(), comp);
+    std::make_heap(this->c.begin(), this->c.end(), this->comp);
   }
   bool empty(void) const { return _Base::empty(); }
   size_t size(void) const { return _Base::size(); }
