--- ./gas/config/tc-mips.c.gprel	Fri Dec  6 17:38:22 2002
+++ ./gas/config/tc-mips.c	Fri Dec  6 17:39:09 2002
@@ -13396,7 +13396,16 @@ tc_gen_reloc (section, fixp)
       && (code == BFD_RELOC_GPREL16 || code == BFD_RELOC_MIPS16_GPREL)
       && reloc->addend != 0
       && mips_need_elf_addend_fixup (fixp))
-    reloc->addend += S_GET_VALUE (fixp->fx_addsy);
+    {
+      /* gbl: in o32 GPREL16 is always inplace, but in n32 and 64
+         it isn't. I wish someone would get rid of this "fix". */
+      reloc_howto_type *howto;
+
+      howto = bfd_reloc_type_lookup (stdoutput, fixp->fx_r_type);
+      if (howto->partial_inplace) {
+       reloc->addend += S_GET_VALUE (fixp->fx_addsy);
+      }
+    }
 #endif
 
   /* To support a PC relative reloc when generating embedded PIC code
