--- ./panel/xfce_support.c.orig	2003-08-07 22:15:03.000000000 +0300
+++ ./panel/xfce_support.c	2003-11-09 16:55:51.000000000 +0200
@@ -955,10 +955,13 @@
 static gboolean
 delayed_exec(ActionCommand *command)
 {
-	real_exec_cmd(command->cmd, command->in_terminal, command->use_sn,
-			command->silent);
+	if (command->cmd != NULL)
+	{
+	    real_exec_cmd(command->cmd, command->in_terminal, command->use_sn,
+			    command->silent);
+	    g_free(command->cmd);
+	}
 
-	g_free(command->cmd);
 	g_free(command);
 
 	return(FALSE);
