--- hping2-rc3/bytesex.h.orig	2006-10-27 16:33:15.000000000 -0700
+++ hping2-rc3/bytesex.h	2006-10-27 16:38:58.000000000 -0700
@@ -7,7 +7,7 @@
 #ifndef ARS_BYTESEX_H
 #define ARS_BYTESEX_H
 
-#if 	defined(__i386__) \
+#if 	defined(__i386__) || __BYTE_ORDER == __LITTLE_ENDIAN \
 	|| defined(__alpha__) \
 	|| (defined(__mips__) && (defined(MIPSEL) || defined (__MIPSEL__)))
 #define BYTE_ORDER_LITTLE_ENDIAN
@@ -15,7 +15,7 @@
 	|| defined (__sparc__) \
 	|| defined (__sparc) \
 	|| defined (__PPC__) \
-	|| defined (__BIG_ENDIAN__) \
+	|| defined (__BIG_ENDIAN__) || __BYTE_ORDER == __BIG_ENDIAN \
 	|| (defined(__mips__) && (defined(MIPSEB) || defined (__MIPSEB__)))
 #define BYTE_ORDER_BIG_ENDIAN
 #else
