--- ./src/tla/tests/test-dir-regexps.sh	2004-01-22 04:08:02.000000000 +0100
+++ /tmp/tmp.vIxwK26605	2004-05-24 12:34:14.000000000 +0200
@@ -36,7 +36,7 @@
 echo "unrecognized ^attrapca$" > .arch-inventory
 test "$(echo $(tla inventory -u))" = "@rapca attrapca"
 echo "source ^@.*$" > .arch-inventory
-test "$(echo $(tla inventory -s --all | head -4))" = \
+test "$(echo $(tla inventory -s --all | head -n 4))" = \
      ".arch-ids/.arch-inventory.id .arch-inventory @rapca attrapca"
 end_test
 
