--- /home/fizz/dev/games/crimson/tools/cfed.cpp	Fri Apr 16 17:56:33 2004
+++ ./tools/cfed.cpp	Mon May 17 18:57:12 2004
@@ -105,7 +105,7 @@
   char *uset = NULL, *tset = NULL;
   string outname;
 
-  if ( SDL_Init( SDL_INIT_VIDEO ) < 0 ) {
+  if ( SDL_Init( 0 ) < 0 ) {
     cerr << "Error: Could not init SDL. " << SDL_GetError() << endl;
     return 1;
   }
--- /home/fizz/dev/games/crimson/src/common/surface.cpp	Fri Feb 27 19:54:12 2004
+++ ./src/common/surface.cpp	Mon May 17 18:59:50 2004
@@ -91,7 +91,7 @@
 int Surface::DisplayFormat( void ) {
   SDL_Surface *new_s;
 
-  if ( s_surface == 0 ) return -1;
+  if ( s_surface == 0 || !SDL_WasInit( SDL_INIT_VIDEO ) )  return -1;
   if ( (new_s = SDL_DisplayFormat( s_surface )) == 0 ) return -1;
   SDL_FreeSurface( s_surface );
   s_surface = new_s;
