--- webCDwriter-2.6.7/CDWserver/Store.cpp	2003-07-05 13:20:11.000000000 +0200
+++ webCDwriter-2.6.7-t2/CDWserver/Store.cpp	2005-01-14 10:12:36.848763864 +0100
@@ -32,7 +32,7 @@
 
 	maxOpenSessions = config.getMaxOpenSessions();
 	sessionsNum = 0;
-	sessions = new (Session *)[maxOpenSessions];
+	sessions = new Session *[maxOpenSessions];
 	int i;
 	for (i = 0; i < maxOpenSessions; i++)
 		sessions[i] = NULL;
