--- ./Makefile.orig	2005-02-03 19:21:42.000000000 -0300
+++ ./Makefile	2005-02-03 19:23:32.000000000 -0300
@@ -91,15 +91,15 @@
 
 install:        ipvsadm
 		if [ ! -d $(SBIN) ]; then $(MKDIR) -p $(SBIN); fi
-		$(INSTALL) -m 0755 -s ipvsadm $(SBIN)
-		$(INSTALL) -m 0755 ipvsadm-save $(SBIN)
-		$(INSTALL) -m 0755 ipvsadm-restore $(SBIN)
+		$(INSTALL) -m 0755 -s ipvsadm $(SBIN)/ipvsadm.old
+		$(INSTALL) -m 0755 ipvsadm-save $(SBIN)/ipvsadm-save.old
+		$(INSTALL) -m 0755 ipvsadm-restore $(SBIN)/ipvsadm-restore.old
 		[ -d $(MAN) ] || $(MKDIR) -p $(MAN)
-		$(INSTALL) -m 0644 ipvsadm.8 $(MAN)
-		$(INSTALL) -m 0644 ipvsadm-save.8 $(MAN)
-		$(INSTALL) -m 0644 ipvsadm-restore.8 $(MAN)
+		$(INSTALL) -m 0644 ipvsadm.8 $(MAN)/ipvsadm.old.8
+		$(INSTALL) -m 0644 ipvsadm-save.8 $(MAN)/ipvsadm-save.old.8
+		$(INSTALL) -m 0644 ipvsadm-restore.8 $(MAN)/ipvsadm-restore.old.8
 		if [ -d $(INIT) ]; then \
-		  $(INSTALL) -m 0755 ipvsadm.sh $(INIT)/ipvsadm; \
+		  $(INSTALL) -m 0755 ipvsadm.sh $(INIT)/ipvsadm.old; \
 		fi
 
 clean:
